<HTML>
<head>
	<title>SyMAP System Guide</title>
	<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
	<style>
a.wt {color:inherit;}
table {font-size:100%}
tr {font-size:100%}
td {font-size:100%}
tt {font-size:13px}
pre {font-size:13px}
a {font-size:100%}
</style>
</head>

<body style="font-family:Verdana, Arial, Helvetica, sans-serif;font-size:100%">

	<a name=top></a>
	<table align="center" style="border: 1px solid #999999" cellspacing="0" cellpadding="0">
	  <tr><td width="100%" style="background-color: #CCCCCC;">
	  <a href="http://www.arizona.edu">
	  <img border="0" alt="The University of Arizona" title="The University of Arizona" src="./img/UA.gif"></a>
	  </td></tr>

	  <tr><td>
		  <table width="750" border="0" align="center" cellpadding="0" cellspacing="0">
		    <tr><td height="60" width="161" align="left" valign="bottom">
       				<img alt="" title="" src="img/agcol_logo.png">
       			</td>
		    	<td height="60" width="428" align="left" valign="bottom" style="padding:0px 0px 10px 30px">
			    	<font face="Verdana, Arial, Helvetica, sans-serif" color="#000066" size="5">
				    <b>SyMAP System Guide</b>
			    	</font> 
		    	</td>
		    	<td height="60" width="161" align="center" valign="bottom">&nbsp;
        		</td>
		    </tr>
		    
		    <tr><td height="21" colspan="3" align="left" valign="top">
			    <table width="100%" border="0" cellspacing="0" cellpadding="3" height="18">
			    	<tr><td align="left" bgcolor="#666666">
<span style="font-size:75%;color:white">
<a href="http://www.agcol.arizona.edu" class="wt">AGCoL</a> |
<a href="http://www.agcol.arizona.edu/software/symap" class="wt">SyMAP Home</a> |
<a href=index.html class="wt">Index</a> |
System Guide |
<a href=UserGuide.html class="wt">User Guide</a> |
<a href=Tour.html class="wt">Tour</a> |
<a href=TroubleShoot.html class="wt">Troubleshooting </a>
</span>
					</td></tr>
				</table>
			</td></tr>
			
	        <tr><td colspan="3" valign="top" align="left" style='padding:20px 15px 0px 15px;font-size:85%'>
								
The following discusses building a SyMAP v5 database. This document
has been updated as of v5.0.7; see <a href=Release.html>Release Notes</a> for changes.

<h4>Contents</h4>
<ul>
	<li><a href=#overview>Overview and Publications</a>
	<li>Getting Started
		<ul>
			<li><a href=#start>Steps for finding synteny</a>
			<li><a href=#machine>System Requirements</a>
			<li><a href=#installation>Installation</a>
			<li><a href=#demo>Running the demo</a>
				<ul>
				<li><a href=#draft>Ordering the demo draft</a>
				</ul>
			<ul>
				<li><a href=Demo.html target="_blank">Querying the demo</a>
			</ul>
			</ul>
	<li>New project
		<ul>
			<li><a href=#mysql>MySQL and Parameters</a>
			<li><a href=#time>Runtime and Memory</a>
			<li><a href=#create>Creating a New Project</a>
			<ul>
				<li><a href=#params2>Parameters</a>
			</ul>
			<li>Sequence project
			<ul>
				<li><a href=#seqprep>Preparing the Sequences</a> 
				<li><a href=#annot>Annotation files</a> - convert from NCBI or Ensembl
				<li><a href=#draft2>Draft sequence</a>
				<li><a href=#self>Self alignments</a>
				<li><a href=#dir>Directory structure</a>
			</ul>

			<li>Alignments and external programs
				<ul>
					<li><a href=#ext>Alignment executables
					<li><a href=#mum>Running MUMmer from the command line</a>
				</ul>
			<li>FPC project
				<ul>
					<li><a href=#fpc>Working with FPC Files</a>
				</ul>
		</ul>
	<li><a href=#alg>How SyMAP Works</a>
	<li><a href=#refs>References</a>
</ul>


<a name="overview">
<h1>Overview and Publications</h1>
SyMAP is a system for computing, displaying, and analyzing syntenic alignments between 
<font color=green><b>medium-to-high divergent eukaryotic genomes</b></font>. It does <b>not</b> work well
for very similar genomes or bacterial genomes.
<p>Its features include the following (for a pictorial introduction, see the 
<a href="Tour.html" target="_blank">Tour</a>): 
<p>
<table>
	<tr>
		<td align='left' valign='top'>
		<table cellpadding=7>
		
		<tr>
			<td>
				Find synteny between two sequenced genomes with optional annotation.
			</td>
		</tr>
		<tr>
			<td>
				Draft sequence ordering by synteny, i.e. 
				<font color=green><b>align a draft genome to a fully sequence (not draft-to-draft</b></font>).
			</td>
		</tr>
		<tr>
			<td>
				Find synteny between an FPC map and sequenced genome.
			</td>
		</tr>
		<tr>
			<td>
				For multiple selected synteny pairs, display using dot plot, circular, and side-by-side. 
			</td>
		</tr>
		<tr>
			<td>
				Complete annotation-based queries with construction of cross-species gene families.
			</td>
		</tr>
		</table>
	</td>
	<td align='left' valign='middle'>
			<table cellspacing=1 cellpadding=0 border=1>
				<tr>
					<td>
						<a href="img/3d.png"><img src="img/3d.png" width=100 height=100></a>
					</td>
					<td>
						<a href="img/2d.png"><img src="img/2d.png" width=100 height=100></a>
					</td>
					<td>
						<a href="img/circmap.png"><img src="img/circmap.png" width=100 height=100></a>
					</td>

				</tr>
				<tr>
					<td>
						<a href="img/dotplot.png"><img src="img/dotplot.png" width=100 height=100></a>
					</td>
					<td>
						<a href="img/mgr.png"><img src="img/mgr.png" width=100 height=100></a>
					</td>
					<td>
						<a href="img/querysetup.png"><img src="img/querysetup.png" width=100 height=100></a>
					</td>

				</tr>
			</table>
		</td> 
	</tr>
</table>
<p>

<a name="pubs">
<h4>Publications</h4>

The back-end processing of SyMAP, including the synteny block 
and anchor filtering algorithms, is described in the following 
two publications. A sketch of the algorithms is also provided <a href="#alg">here</a>.

<pre>
        C. Soderlund, M. Bomhoff, and W. Nelson (2011) 
        SyMAP: A turnkey synteny system with application to plant genomes.
        Nucleic Acids Research 39(10):e68.
</pre>
<p>
<pre>
        C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
        SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps 
        Genome Research 16:1159-1168.
</pre>

SyMAP is freely distributed software, however
<font color=green><b><i>if you use SyMAP results in published research, you must cite one
or both of these articles along with any external programs you used within SyMAP such as
MUMmer<sup>1,2</sup></i></font></b>


<p>&nbsp;
<a name=start></a>
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h1>Getting Started</h1>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<h2>Steps for finding synteny</h2>
Follow the steps below to get started with SyMAP.
Note that each SyMAP application window has a separate Help
button providing further details on the use of its functions. 
<p>

<table cellpadding=7 >
	<tr>
		<td align='left' valign='top' >
			1. 
		</td>		
		<td align='left' valign='top' >
			Use a Linux or Mac machine.  
		</td>
		<td align='left' valign='top'>
			It needs to have Java v1.7 or later, and sufficient processing power.
			See <a href="#machine">system requirements</a>. 
		</td>
	</tr>
	
	<tr>
		<td align='left' valign='top'>
			2. 
		</td>
		<td align='left' valign='top'>
			Set up MySQL.
		</td>
		<td align='left' valign='top'> See <a href="#mysql">MySQL</a>.
		</td>
	</tr>
	<tr>
		<td align='left' valign='top'>
			3.  
		</td>
		<td align='left' valign='top'>
			Download SyMAP. 
		</td>
		<td align='left' valign='top'>
			Installation is a simple unzip. See <a href="#installation">installation</a>.
		</td>
	</tr>
	<tr>
		<td align='left' valign='top'>
			4.
		</td>
		<td align='left' valign='top'>
			Run the demo.
		</td>
		<td align='left' valign='top'>
			Highly recommended. See <a href="#demo">running the demo.</a>
		</td>
	</tr>
	<tr>
		<td align='left' valign='top'>
			5.  
		</td>
		<td align='left' valign='top'>
			Prepare sequences and annotation.  
		</td>
		<td align='left' valign='top'>
			Sequences can be in one or many files and can be masked or unmasked.
			See <a href="#seqprep">preparing the sequences</a>.
			Annotation format is gff3; see <a href="#annot">annotation files</a>.
		</td>
	</tr>
	

	<tr>
		<td align='left' valign='top'>
			6. 
		</td>
		<td align='left' valign='top'>
			Load the files into SyMAP.
		</td>
		<td align='left' valign='top'>
			The SyMAP interface makes this easy; see
			<a href="#create">creating a new project</a>.
		</td>
	</tr>
	<tr>
		<td align='left' valign='top'>
			7. 
		</td>
		<td align='left' valign='top'>
			Compute alignments and synteny.
		</td>
		<td align='left' valign='top'>
			This is also easy through the SyMAP interface. See
			<a href="#time">runtime and memory</a>.
		</td>
	</tr>
	<tr>
		<td align='left' valign='top'>
			8. 
		</td>
		<td align='left' valign='top'>
			View results.
		</td>
		<td align='left' valign='top'>
			Detailed
			description of the user interface is in the <a href="UserGuide.html" target="_blank">User Guide</a>.
		</td>
	</tr>
</table>
<p>If you are working with FPC files, see the section on <a href="#fpc">FPC</a>. 

<p>&nbsp;<p>
<a name="machine">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>System Requirements</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

SyMAP v5.0.6 has been tested on the following:
<table border=1 cellspacing=5 cellpadding=5>
<tr><td>Centos Linux amd64<td>MariaDB v10.4.12<td>Java 1.8
<tr><td>MacOS x86_64 (Catalina 10.15.4)<td>MySQL v8.0.17<td>Java 14.0
<tr><td>MacOS x86_64 (Maverick 10.9.5)<sup>*</sup><td>MySQL v5.6.21<td>Java 1.7
</table>
<sup>*</sup><small>The /ext directory of external executables does not work with the defaults: 
see <a href="TroubleShoot.html#mac" target="_blank"> Trouble Shoot Mac</a>.</small>

<p>The jar files have been compiled with Java 1.7, which is upward compatible. 
<p>
For performing large alignments (e.g. 1Gb genomes or more) it is essential
to have multiple CPUs, a 64-bit computer, and at least 5Gb of RAM
for each CPU that you intend to use. (Note that you can set the
number of CPUs for SyMAP to use). 
<p>
For viewing alignments, CPU and memory needs are typically negligible, unless
you are performing queries on more than 4-5 genomes at once. 


<a name="installation">
<h2>Installation</h2>

Installation simply consists of unzipping the download package, using
the command
<pre>     &#62 tar -xzvf symap_5.tar.gz</pre>
This can be done anywhere and creates a directory called <tt>symap_5</tt>. You can
move this directory later if desired.
<p>
To run SyMAP, enter the <tt>symap_5</tt> directory and run the command:
<pre>     &#62 ./symap 
</pre>

<p>MUMmer<sup>1,2</sup> (sequence alignment), Blat<sup>3</sup> (FPC alignment)
and MUSCLE<sup>7</sup> (SyMAP Queries) are provided with the package.

<p>&nbsp;
<a name="demo">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>Running the Demo</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


If you have not used SyMAP before, it is essential to run the demos. After you have 
installed MySQL, do the following:
<ol>
<li>Change into the <tt>symap_5</tt> directory. 
<p>
<li>Edit <tt>symap.config</tt> and enter database 
and host information (see <a href=#mysql>MySQL</a>), e.g. <tt>db_name=symapDemo</tt>.
<p>
<li>From the command line, type <tt>./symap</tt>. 
</ol>
<p>The first time you run SyMAP, it will create the database with information written to the terminal, e.g.
<pre>
Creating database 'symapDemo' (jdbc:mysql://localhost/symapDemo?characterEncoding=utf8).
</pre>
<br>It will check your MySQL variables; if there are any "Suggested" changes,
see <a href="TroubleShoot.html#msqslow" target="_blank">Trouble Shoot MySQL</a>. It will also
check that the provided executables (e.g. MUMmer); if it shows any problems,
see <a href="TroubleShoot.html#ext" target="_blank">Trouble Shoot External</a>.
<p>
<table>
<tr><td align="left" valign="top">
The Project Manager window opens showing the four demo projects provided with the SyMAP packages.
Check "Demo-Seq" and "Demo-Seq2".
</td>
<td align="left" valign="middle">
	<a href="img/projman1.png"><img src="img/projman1.png" width=400 border=1></a>
</td>
</tr>

<tr><td colspan=2 height=40>&nbsp;</td></tr>

<tr><td align="left" valign="top">

A link "Load All Projects" will be displayed in the top of the right panel; select it to 
load the projects, which will take several minutes. 
<p>If loading the "Demo-Seq" takes
<i>more than a few minutes</i>, you may need to adjust the MySQL parameters, see
<a href="TroubleShoot.html#msqslow" target="_blank">TroubleShoot MySQL</a>.  
<p>When done, the Manager will look as shown in the image on the right.
<p>In the "Available Syntenies" table, click the cell for the "Demo_Seq2" row and the "Demo_Seq" column.  
Then click the "Selected Pair" button to start the alignment. 
</td>
					
<td align="left" valign="middle">
<a href="img/projman4.png"><img src="img/projman4.png" width=400 border=1></a>
</td>
</tr>

<tr><td colspan=2 height=40>&nbsp;</td></tr>
				
<tr>
<td align="left" valign="top">
<p>The Synteny & Alignment should take less than 30 minutes with one CPU (15 minutes with two CPUs). 
<p>When done, the table will have a checkbox, signifying that the synteny is available for viewing. 
Click the checked cell, which will enable the viewing buttons.  
</td>
					
<td align="left" valign="middle">
<a href="img/projman5.png"><img src="img/projman5.png" width=400 border=1></a>
</td>
</tr>
				
<tr><td colspan=2 height=40>&nbsp;</td></tr>
				
<tr>
<td align="left" valign="top">
Click "Summary" to view the v5 summary shown on the right; <i>there may be slight difference in the number of anchors because the results are slightly different when run when different numbers of CPUs</i>.  
<p>To view the other interfaces, see <a href="Demo.html" target="_blank">Demo Query</a>.
</td>
<td align="left" valign="middle">
<a href="img/projsum.png"><img src="img/projsum.png" width=400 border=1></a>
</td>
</tr>
</table>						
				
<a name="draft">
<h3>Ordering the demo draft</h3>

The remainder of the demo relates to draft sequence, i.e. unanchored
shotgun sequence. 
					
<table>
<tr><td colspan=2 height=40>&nbsp;</td></tr>
	<tr>
		<td>
Load the Demo-Draft project, following
the same steps used to load the previous two projects. 

<p>Under the Demo-Draft listing, you will see
the parameter "Order against: demo_seq2". With this setting,
the Demo-Draft contigs will be ordered using synteny to
Demo-Seq2.  
		</td>
		<td align="left" valign="middle">
<a href="img/demodraft1.png"><img src="img/demodraft1.png" width=400 border=1></a>
		</td>
	</tr>
	<tr><td colspan=2 height=40>&nbsp;</td></tr>
	<tr>
		<td align="left" valign="top">
Run the Alignment & Synteny, where the alignment should take less than 30 minutes with one CPU.
When done, open the Summary for the pair, as shown on the right; <i>there
may be slight difference in the number of anchors because the results are
slightly different when run when different numbers of CPUs.</i>

See the first dot plot in <a href="Demo.html#draft" target="_blank">Demo-draft</a>.
		</td>
		<td align="left" valign="middle">
<a href="img/projsum2.png"><img src="img/projsum2.png" width=400 border=1></a>
		</td>
	</tr>
</table>

<p>The ordering algorithm changes the order of the draft contigs in the database, but does
not change the sequence files on disk. However, it writes the following files:
<p>1. <i>File of ordered contigs:</i> It writes the order of the contigs along with
whether they should be flipped to a file called <tt>/data/seq/Demo_seq2/ordering.csv</tt>.

<p>2. <i>Fasta files of ordered sequences: </i>
It creates sequence files from the ordered contigs that are flipped when appropriate,
which are put into a new project
with the suffix "<tt>_ordered</tt>", as shown on the right. The chromosome names correspond
to the order-against project (e.g. Demo_seq2), and the third chromosome is 'chr0'
which contains all draft sequences that were not placed.
<p>
<table>
<tr><td>
As shown in the image on the left, a new project has been added. Click <tt>Demo_Draft-ordered</tt> and
load it.
Running the Alignment&Synteny
between the <tt>_ordered</tt> project and Demo_seq2 will provide a more coherent display
(see the second dot plot for <a href="Demo.html#draft" target="_blank">Demo-draft</a>). 
<td><td align="center" valign="middle">
<a href="img/demodraft3.png"><img src="img/demodraft3.png" width=400 border=1></a>

</table>

<p><b>Further demos:</b>
<ol>
<li>There is also Demo-FPC that can be aligned to Demo-Seq
following the same set of steps as described above. 
<li>The section on <a href=#self>Self alignments</a>
discusses performing a demo_seq self alignment.
</ol>
This is the end of the demo.
<p>&nbsp;<p>
<!-------------------------------------------------------->

<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h1>New Project</h1>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<a name="mysql">
<h2>MySQL and parameters</h2>

If your machine does not have MySQL or MariaDB, download and install it. For example,
MySQL can be downloaded from <a href="http://dev.mysql.com" class="ext" target="_blank">dev.mysql.com</a>.
On a personal MacOS, simply download the '.dmg' file and following the instructions. 
On a work server, the system administrator may need to install it.

<p>
The MySQL installation does not need to be on the machine where
you will do the computations or view the results, as long as
it is on an accessible network. Once the server is ready, fill out
the database parameters in the <tt>symap.config</tt> file in the 
main SyMAP directory, as described below. 

<p>
<font color=red>Important Note</font>: The default settings of MySQL are poorly suited for large-scale
data storage. You will want to adjust the parameters 
<tt>innodb_buffer_pool_size</tt> and <tt>innodb_flush_log_at_trx_commit</tt> as described in
<a href="TroubleShoot.html#msqslow" target="_blank">Trouble Shoot MySQL</a>.

<p>
<a name="params">
<h4>Database Parameters</h4>
Parameters for accessing the MySQL database
should be set in the <tt>symap.config</tt> file in the main symap directory, as follows:
<p>
<table width=650 rules=all cellpadding=4 border=1 align=center>
	<tr>
		<td colspan=2 align=center >
		<i>Database Parameters</i>
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_name</tt></td>
		<td>
			Name of the MySQL database, which SyMAP will create when it first reads
			<tt>symap.config</tt>. It is standard to start the name with <tt>symap</tt>,
			e.g <tt>symapDemo</tt>.
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_server</tt></td>
		<td>
			The machine hosting the MySQL database, e.g. <tt>myserver.myschool.edu</tt>. If using
			your local machine, enter <tt>localhost</tt>.
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_adminuser</tt></td>
		<td>
			MySQL username of a user with sufficient privileges to create a database. Needed for
			creating and updating alignments.
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_adminpasswd</tt></td>
		<td>
			Password of the admin user.	
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_clientuser</tt></td>
		<td>
			MySQL username of a user with read-only access. This is only necessary if you
			want a machine to run SyMAP as read-only.
		</td>
	</tr>
	<tr>
		<td align="left" valign="top" ><tt>db_clientpasswd</tt></td>
		<td>
			Password of the client user.
		</td>
	</tr>
</table>

<p>Example <tt>symap.config</tt>.
<pre>
db_name             = symapDemo
db_server           = localhost
db_adminuser        = root
db_adminpasswd      = &lt;password&gt;
db_clientuser       = 
db_clientpasswd     = 
</pre>
To use an alternative file than <tt>symap.config</tt>, use the "-c" command line argument, e.g.
<pre>
&gt;./symap -c symapTmp.config
</pre>
This is useful if you have multiple SyMAP databases.
<a name="time"></a>
<h2>Runtime and Memory</h2>

The largest component of SyMAP execution time is in running
MUMmer<sup>1,2</sup> (or BLAT<sup>3</sup>). The typical runtime is one CPU-hour per
target sequence (or group of chromosomes, since SyMAP
will group shorter sequences together for efficiency). 
<p>
For example, to align rice (12 chromosomes, 370Mb) to maize (10 chromosomes, 2Gb)
 required 1 hour, 3 minutes using 8 CPUs
with 2.3Ghz speed. SyMAP used one CPU per maize chromosome to align the enter rice chromosome
against each of the 10 maize chromosomes (i.e. used 10 CPUs).
<p>
The memory usage of MUMmer is typically 5G per CPU, however it can
be as high as 10G for very long or repetitive chromosomes. 

<p>&nbsp;<p>
<a name="create">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>Creating a New Project</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

Start symap (i.e. <tt>./symap</tt>). All projects listed under <tt>/data/fpc</tt> are FPC projects
and all projects listed under <tt>/data/seq</tt> are Sequence projects. 
These sub-directories of <tt>/data/fpc</tt> and <tt>/data/seq</tt> will be listed on the
left panel under <b>Projects</b>. You may create the project as discussed in <a href=#dir>Directory
structure</a>, or you may create a new project via the SyMAP interface.
<p>
<table>
<tr><td  align="left" valign="top">
To create a new project via the SyMAP interface, press the "Add Project" button
at the lower left. Enter the project name beside "Name:"; if the project type is "fpc", change the dropdown
beside "Type:". The Help
button on the dialog provides further information. The following will discuss a type "sequence" project and
everything is similar for the "fpc" project.
<p>&nbsp;
<td width="50%">
<a href="img/projnew.png"><img src="img/projnew.png" width=400 border=1></a>

<p>
<tr><td  align="left" valign="top">
After saving the new project, it appears in the Projects list on the left, but
it is still an empty shell. A directory will be made under the <tt>/data/seq</tt>, 
e.g. for the project added on the right, a directory will be created called <tt>/data/seq/foobar</tt>. 
Check its box and it will appear in the Summary section (right hand side).
<td width="50%">
<a href="img/projnew2.png"><img src="img/projnew2.png" width=400 border=1></a>
</table>

<a name="params2">
<h4>Parameters</h4>

<table>
<tr>
<td  align="left" valign="top">
<i>For a project:</i> Click the <u>Parameters</u> link to open the Parameters window shown in the image on the right.
<p>
On the Parameters window you will add the filenames for the sequences and annotations for the project,
as well as setting other parameters if desired. The Help button on this window
provides the necessary details. 
<p><i>Defaults:</i> If the annotations are put in the sub-directory <tt>/annotations</tt> and
the sequences are put in the subdirectory <tt>/sequence</tt>, it is not necessary to enter where
the annotations and sequences exist, e.g. the default directories for the project <tt>foobar</tt>
are as follows:
<pre>
/data/seq/foobar/annotations
/data/seq/foobar/sequences
</pre>

<td><a href="img/parameters1.png"><img src="img/parameters1.png" width=400 border=1></a>
</table>

<p>
After setting the parameters, the project is ready to be loaded.  
<p>
<table>
<tr>
<td  align="left" valign="top">
<i>Alignment & Synteny</i> parameters can be changed by selecting a project pair cell 
followed by the "Parameters" button, which brings up with window shown on the right. 
See the "Help" for a description of the parameters. 

<p>These parameters generally do not need to be changed. 

<p><tt>Blat</tt> is the alignment program for FPC to Seq. 
The <tt>MUMmer</tt> program <tt>PROmer</tt> is the default program
for sequence alignments for different genomes. The <tt>MUMmer</tt> program <tt>NUCmer</tt> is the default program
for sequence self-alignments. 
<td valign="top">
<a href="img/alignLine.png"><img src="img/alignLine.png"  width=400 border=1></a>
<br><a href="img/parameters2.png"><img src="img/parameters2.png"  width=350 border=1></a>
</table>


<!-------------------------------------------------------->
<p>&nbsp;<p>
<a name="seqprep">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>Sequence project</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<h3>Preparing the Sequences</h3>
The first decision with whole-genome sequence is whether to used repeat masked sequences. 
<ul>
<li>Masking reduces alignment time and false-positive hits, 
but also runs a risk of concealing true hits due to inaccurate masking.
<li>SyMAP does not perform the repeat-masking so must be done with another program.
However, you may obtain masked sequences from NCBI or Ensembl. 
NCBI provides soft-masked sequences where the 
<tt>scripts/ConvertNCBI</tt> script will convert it to hard-masked, and Ensembl
provides both soft and hard-masked sequences (see <a href="./convert">Convert</a>). 
<li>
Masking is not really necessary unless the genome is highly repetitive
<i>and</i> those repeats are shared with other genomes being aligned. 
(Repeats cause particular trouble for self-alignments,
see <a href="#self">self-alignments in SyMAP</a>).
<li>Occasionally, MUMmer fails aligning sequences which can be fixed by masking
(see <a href="TroubleShoot.html#mask" target="_blank">Trouble Shoot Mask</a>)
</ul>
<p>
Another masking option which is available if you have gene
annotation is to mask out everything but the annotated genes. You
can enable the "mask_genes" option on in the Project's Parameters window (shown <a href=#params2>above</a>);
turn it on <i>before</i> doing the alignments.
<p>
Note that sequence files should be in FASTA format and the name of a sequence
is the string immediately following the "&gt;", e.g. 
<pre>
&gt;Chr3  Oryza Sativa
GAATTCGAATTTGGGTAATGCTAATCAATACAGGTCAAAATCTATGTATTGAGTGGAATATACTGCAAAGTAATTACCTT
CTTCCAAAGGAAAGCATTCCTTCTCTCTTGTGGGACTAGCAGATGATCTCGCAGCCAAGACGTGACCACCCAAGGCTCAC
...
</pre>
In this example, the sequence name is "Chr3" and the Chr3 sequence follows. The additional information
"Oryza Sativa" is ignored.
<p>
<b><font color=green>Three things are important in naming sequences for SyMAP</font></b>:
<table cellpadding=4 style="margin-left:10px;margin-top:5px;" >
	<tr>
		<td align='left' valign='top' >
			A. 
		</td>		
		<td align='left' valign='top' >
			<u>Sequence names can contain only letters, numbers, and underscores. </u>
		</td>
	</tr>
	<tr>
		<td align='left' valign='top' >
			B. 
		</td>		
		<td align='left' valign='top' >
			<u>The sequence names must <i>exactly</i> match those 
			used in the annotation files (first column), or the
			annotations will not be loaded. </u> 
		</td>
	</tr>
	<tr>
		<td align='left' valign='top' >
			C. 
		</td>		
		<td align='left' valign='top' >
			Use a consistent prefix such as "Chr" for all sequences, followed
			by a short number; set 'grp_prefix' to the prefix in Project's Parameters window (shown <a href=#params2>above</a>).
			If there is not a consistent prefix, you may leave the 'grp_prefix' blank; 
			beware, this can have unintended results, so should be avoided if possible.
		</td>
	</tr>
	
</table>

<a name="annot">
<h3>Annotation files</h3>

Annotation files should be in 
<a href="http://www.sequenceontology.org/gff3.shtml" target="_blank">gff3 format</a>. 
The first column (seqid) must exactly match the sequence names in
the fasta files. The third column (type) determines how SyMAP
uses the entry. Types "gene", "exon", "CDS", "centromere", and "gap"
are recognized (other entries are ignored). "CDS" and "exon" are
treated equivalently in SyMAP. 
<p>
The last column (attributes) contains "tag=value" pairs describing
the annotation. You can set which attributes to use, or use all
those occurring more than a certain number of times; open the 
Project's Parameters window (shown <a href=#params2>above</a>), look for parameter "annot_keywords". 
<p>
<i><b>Important:</b> Only the annotation on "gene" entries is shown in the
displays or used for searching. For entries "exon", "CDS", "gap", and "centromere",
only the coordinates from the gff file are used; the annotation text is not read.   
</i>

<p><i><b>NCBI files:</b></i> A Java script (<tt>scripts/ConvertNCBI.class</tt>) has been provided that 
converts NCBI genome fasta files and gff annotation files
into the format that works best with SyMAP. 
See the <a href="convert/ncbi.html" target="_blank">documentation</a> for instructions. 

<p><i><b>Ensembl files:</b></i> A Java script (<tt>scripts/ConvertEnsembl.class</tt>) 
has been provided that converts Ensembl genome fasta files and gff3 annotation files
into the format that works best with SyMAP. 
See the <a href="convert/ensembl.html" target="_blank">documentation</a> for instructions.


<a name="draft2">
<h3>Draft sequence</h3> 

If you are ordering the draft sequence against a closely related genome, see <a href=#draft>demo draft</a> on how to proceed. 

<p>If you are not ordering the draft sequence,
and if the draft sequence is in too many sequence pieces, 
then (1) it takes a long time for the MUMmer comparisons, (2)
the display is very cluttered, and 
(3) the blocks display does not work right. Limit the number of sequence pieces
by setting <tt>min_size</tt> in the parameters window to only load the largest 150 sequences; 
there is a script called <tt>scripts/lenFasta.pl</tt> which
will print out all the lengths; set the min_size to the 150th length. However, even 150 are a lot of blocks to view
so you might want to start with the largest 50, merge them, then repeat.


<a name="self">
<h3>Self Alignments</h3>
<table>
<tr>
<td>To perform self-synteny, select the cell for the same project followed by 
"Selected Pair". The "All Pairs" option does not include self-synteny.
<p>
By default, SyMAP uses the MUMmer 'nucmer' program for self-alignments. Each chromosomes
is compared to every other chromosome including itself. 
<td><a href="img/self.png"><img src="img/self.png" width=200 border=1></a>
</table>
 
<p>The Alignment & Synteny "Parameters" popup has an option to set "Self Args",
which only uses the "Self Args" parameter(s) when comparing the chromosome sequence file
to itself. In v4.2, the "--maxmatch" parameter was always used, but now it
is up to the user as to whether they want to add it to "Self Args" (it can greatly
increase the execution time).
 
<p>Reasoning for using "--maxmatch": 
 MUMmer ordinarily 
seeds its alignments with unique matches, which eliminates the possibility
of off-diagonal seeds in the alignment of a chromosome to itself. To overcome
this problem, individual chromosome self-alignments can use
the MUMmer parameter -maxmatch, which removes the uniqueness requirement at
the cost of greatly increased noise. The extra noise is then filtered to
a large extent  by the default SyMAP filters, but the diagonal squares of
the dot plot will still have more noise visible than the off-diagonal.


<p>&nbsp;<p>
<a name="dir">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h3>Directory structure</h3>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

For SyMAP v5, the default location for sequence and annotation files is:
<pre>
/data/seq/&lt;project-name&gt;/sequence
/data/seq/&lt;project-name&gt;/annotation
</pre>
You may do one of the following:
<ol>
<li>Create these sub-directories under <tt>/data/seq</tt> and put your files there.
<p>
<li>Create these sub-directories under <tt>/data/seq</tt> and use soft links to point to the file locations,
e.g.
<pre>
cd data/seq
mkdir foobar
cd foobar
ln -s &lt;location of directory of sequence files&gt; sequence
ln -s &lt;location of directory of annotation files&gt; annotation
</pre>
<li>Use the project parameter window to enter the location of the sequences and optional annotation.
</ol>
For options 1 and 2, it is not necessary to enter the locations of the files in the parameter window since
both use the default locations. Beware, executing the "Remove project from disk" (depending on
location), can remove the data files also.
<p>
The file results of the alignment and synteny computations are is as follows:
<pre>
/data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/align
/data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/final
</pre>

<a name="ext">
<h2>Alignments and external programs</h2>

<a name="ext">
<h3>Alignment executables</h3>
The alignment programs are provided in the <tt>symap/ext</tt> directory. There are 
executables for 64-bit Linux and 64-bit MacOS. SyMAP will
select the correct directory for the machine you are running from, i.e. you
do not need to do anything. Typically, these will work just fine;
but if they do not, see <a href="TroubleShoot.html#ext" target="_blank">Trouble Shoot External</a>
and the following section.

<a name="mum">
<h3>Running MUMmer from the command line</h3>
A compile version of MUMmer v3 is included, where the latest release is MUMmer v4. A compiled
version of MUMmer v4 cannot be included because it has hard-coded path names. 
As discussed in <a href="TroubleShoot.html#ext" target="_blank">Trouble Shoot External</a>, sometimes
when MUMmer v3 fails, MUMmer v4 will work .
If you need to run MUMmer v4 using the command line, do the following:
<ul>
<li>Distant genomes:
<ul>
<li>Use promer
<li>The output of promer is input to "show-coords -dlkTH".
</ul>
<li>Self synteny:
<ul>
<li>Use nucmer. For the same chromosome against itself, use the parameter "--maxmatch".
<li>The output of nucmer is input to "show-coords -dlTH".
</ul>
<li>Result files:
<ul>
<li>The result files must have suffix ".mum" 
<li>Put the ".mum" files in the directory <tt>data/seq_results/&lt;proj1-to-proj2&gt;/align</tt>
<li>In the <tt>&lt;proj1-to-proj2&gt;/align</tt> directory, execute:
<pre>touch all.done</pre>
This creates a file, which indicates to SyMAP that the alignments are done and to process
the files in the directory ending with ".mum".
</ul>
<li>When you run "Selected Pair", SyMAP will recognize the files and use them to build the synteny blocks.
</ul>
Example of compiling mummer4:
<pre>
cd ext/mummer4
tar xvf mummer-4.0.0rc1.tar.gz
mkdir macRC1
cd macRC1
pwd		# prints path to terminal, <i>copy path</i>
cd ../mummer-4.0.0rc1
./configure --prefix=<i>copied path</i>
make 
make install
</pre>
NOTE: You cannot move the mummer directory after creating it because the absolute paths are hard-coded.

<p>Example of using mummer4 for loaded projects "foo" and "bar":
<pre>
cd data/seq_results
mkdir foo_to_bar
mkdir foo_to_bar/align
touch foo_to_bar/align/all.done
cd ../../ext/mummer4/macRC1
./bin/promer ../../../data/seq/foo/sequence/genomic.fna ../../../data/seq/bar/sequence/genomic.fna
./bin/show-coords -dlkTH out.delta >foobar.mum
mv foobar.mum ../../../data/seq_results/foo_to_bar/align
</pre>
The /align directory can have any number of .mum files, that is, you can align the chromosome files against
each other separately.
<a name="mum4">
<h4>Using MUMmer v4 within SyMAP</h4>
Will be added for next release.

<p>&nbsp;<p>	 
<a name="fpc">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>FPC project</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
<h3>Working with FPC Files</h3>

SyMAP can also align genomes represented by an FPC physical
map, by first aligning the BACs using BAC-end sequences or marker
sequences. If you plan to work with FPC alignments, the first
step is to run the provided demo "Demo-FPC". Align it to "Demo-Seq2"
using the same steps as described <a href="#demo">above</a>, and 
explore the various displays. 
<p>
Creating an FPC project is the same as for a 
<a href="#create">sequence project</a> except that you choose
the type "fpc", and then the Project Parameters window
has some different parameters. The Parameters window is where
you will enter the FPC file, and your fasta files of marker
and BAC-end sequences.
<p>
Note that the BAC-end sequence names must be exactly the clone
names used in FPC, with extension "r" or "f" labeling the 
strand. In other words, if the FPC map has a clone
"a0435B26" then the BES for that clone can be named
"a0435B26f" or "a0435B26r".
<p>
The BES and marker alignments in an FPC project are performed using 
BLAT<sup>3</sup>. The running time is typically
several times longer than that of MUMmer (described <a href="#time">here</a>), but
the memory usage is much lower.  
<p>
The default locations are:
<pre>
/data/fpc/&lt;project-name&gt;/&lt;name&gt;.fpc;
/data/fpc/&lt;project-name&gt;/bes
/data/fpc/&lt;project-name&gt;/mrk
</pre>
The options for sequence files under <a href=#dir>directory structure</a> applies to FPC files using
these default locations.
<p>


<a name="alg">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>How SyMAP Works</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

This section provides a brief overview of the SyMAP processing steps; for
more, see the SyMAP published papers<sup>5,6</sup>. The processing
has four phases:
<div style="padding:10px 0px 0px 20px;">
<i>Alignment:</i> <br>
The sequences are written to disk<sup>*</sup>, with gene-masking
if desired. In the alignment, one species is "query" and the other is 
"target". If one project is FPC, that is the query; if both are sequence,
the query is the one with alphabetically the first name. The query sequences
are written into one large file, while smaller target sequences are grouped
into larger fasta files of size up to 70Mb, for more efficient processing
in MUMmer. 
<p>
<i>Anchor Clustering:</i><br>
The raw anchor set consists of the hits found by MUMmer or BLAT.
These are first clustered into gene, or putative-gene hits. This is
done by clustering the hit regions on each sequence, and then defining new "gene"
hits which connect these regions. For example if three separate 
exons hit between two genes, they will be clustered into one "gene"
hit having a combined score equal to the sum of the raw hit scores.
Clustering is by gene if the hits overlap annotation, otherwise, it uses
a max separation 1kb, creating "putative gene" regions.
<p>
<i>Anchor Filtering:</i><br>
The clustered "gene anchors" are now filtered using a version of 
reciprocal-best filtering which is adapted for retaining duplications and
gene families. For each pair of genes (or putative genes) which is 
connected by a clustered anchor, the retained anchors must be among
 the top two anchors by score on both sides (top-2 allows for one
ancestral whole-genome duplication). An anchor will also be retained if its
score is at least 80% of that of the 2nd-best anchor on each side (this
allows for retention of gene family anchors). These filter parameters
may be adjusted through the Alignment & Synteny Parameters window. 
<p>
<i>Synteny Block Detection:</i><br>
After the clustered anchors are loaded into the database,  the synteny
synteny block algorithm runs. This algorithm looks for approximately-collinear
sequences of anchors, subject to several parameters including (A) Number
of anchors; (B) Collinearity of the anchors; (C) Amount of "noise" in the 
surrounding region (to help reject false-positive chains). Criterion A can
be adjusted in the Alignment & Synteny Parameters window.
<p>
<sup>*</sup> Note that the sequences are re-written from the database to the
disk for three reasons: (A)To allow re-grouping for efficiency; (B) To ensure elimination
of invalid characters; (C) To mask non-gene regions, if desired. This also ensures that
sequences names will match those in the database, and prevents problems caused by
moving the source sequences on disk.
</div>

<a name="refs">
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h2>References</h2>
</td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<sup>1</sup>
Kurtz, S., Phillippy, A., Delcher, A.L., Smoot, M., Shumway, M., Antonescu, C., Salzberg, S.L. 
(2004) Versatile and open software for comparing large genomes, Genome Biology, 5:R12 

<p>
<sup>2</sup>
Marçais, G., A.L. Delcher, A.M. Phillippy, R. Coston, S.L. Salzberg, A. Zimin (2018)
MUMmer4: A fast and versatile genome alignment system,  PLoS computational biology, 14(1): e1005944.

<p>
<sup>3</sup>
Kent, J. (2002) BLAT--the BLAST-like alignment tool, Genome Research 12:656-64. 

<p>
<sup>4</sup>
Krzywinski, M., J. Schein, I. Birol, J. Connors, R. Gascoyne, D. Horsman, S. Jones, M. Marra (2009)
Circos: An information aesthetic for comparative genomics. Genome Research doi:10.1101/gr.092759.109.

<p>
<sup>5</sup>
Soderlund, C., Nelson, W., Shoemaker, A., and Paterson, A.(2006)
SyMAP: A system for discovering and viewing syntenic regions of FPC maps.
Genome Res. 16:1159-1168. 
<p>
<sup>6</sup>
Soderlund, C., Bomhoff, M., and Nelson, W. (2011)
SyMAP: A turnkey synteny system with application to multiple large duplicated plant sequenced genomes.
Nucleic Acids Res V39, issue 10, e68.
<p>
<sup>7</sup> 
Edgar, R (2004) 
MUSCLE: a multiple sequence alignment method with reduced time and space complexity.
BMC Bioinformatics 113.

				</td>
		</tr>

</table>
	
	<table width="100%" border="0" align="center" cellpadding=2 cellspacing=0 style="margin-top:40px">
		<tr>
			<td bgcolor="#DDDDDD" align="center">
			<font face="Verdana, Arial, Helvetica, sans-serif" size="2">Email
			Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></font>
			</td>
		</tr>
	</table>
	  
	</td>
</tr>
</table>
</BODY>
</HTML>
