<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<title>SyMAP Tour</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">

<style>
	
a.wt {color:inherit;}
table {font-size:100%}
tr {font-size:100%}
td {font-size:100%}
tt {font-size:13px}
pre {font-size:13px}
a {font-size:100%}
.hang {text-indent:-20px; margin-left:20px;}
	
		li {margin-top:3px;}

		h3 {margin-top:10px;}
</style>
</head>
<a name=top></a>	
<body style="font-family:Verdana, Arial, Helvetica, sans-serif;font-size:100%">
<!--#include virtual="/common/uaheader.html" -->

<table  align="center" width=850 bgcolor="#FFFFFF" cellpadding="0" cellspacing="0" class="bodytext" style="border: 1px solid #999999" >

  <tr> 
    <td colspan="3" valign="top" align="left"  > 

	<table align="center"  cellspacing="0" cellpadding="0" width="100%">
	  

	  <tr>
	  <td>
		  	<table width="850" border="0" align="center" cellpadding="0" cellspacing="0">
		    <tr>
    	  	<td height="60" width="161" align="left" valign="bottom">
       			<img alt="" title="" src="img/agcol_logo.png">
       		</td>
		    <td height="60" width="428" align="left" valign="bottom"
				style="padding:0px 0px 10px 30px">
			    <font face="Verdana, Arial, Helvetica, sans-serif" color="#000066" size="5">
				    <b>SyMAP Tour</b>
			    </font> 
		    </td>
		    <td height="60" width="161" align="center" valign="bottom">&nbsp;
        	</td>
		    </tr>

		    <tr>
		    <td colspan="3" align="left">
			    <table width="100%" border="0" cellspacing="0" cellpadding="3" height="18">
			    <tr>
			    <td align="left" bgcolor="#666666">
<span style="font-size:75%;color:white">
<a href="http://www.agcol.arizona.edu" class="wt">AGCoL</a> |
<a href=../index.html class="wt">SyMAP Home</a> |
<a href=SystemGuide.html class="wt">System Guide</a> |
<a href=UserGuide.html class="wt">User Guide</a> |
Tour |
<a href=TroubleShoot.html class="wt">Troubleshooting</a>
</span>				</td>
				</tr>
          		</table>
			</td>
			</tr>
			</table>
		</td>
		</tr>
	</table>
	</td>
  </tr>
  <tr>
	<td align="center">
	<table width='100%' >
		<tr>
		<td align='left' style='padding:20px 15px 0px 15px;font-size:85%'>



<table align='center' width='100%' >
	<tr>
		<td align='left' >

Following is a quick tour of the many displays and queries available through SyMAP.

<p>
<h3>Contents</h2>
<table >
	<tr>

<td valign="top" align="left" style="padding-left:20px;">
Management:
<li style="margin-top:5px;"><a href="#mgr">Project Manager</a></li>
<li style="margin-top:5px;"><a href="#exp">Chromosome Explorer</a></li>


<td valign="top" align="left" style="padding-left:20px;">
High-level views:
<li style="margin-top:3px;"><a href="#circ">Circle</a></li>
<li style="margin-top:3px;"><a href="#dp">Dot Plot</a></li>
<li style="margin-top:3px;"><a href="#blk">Block</a></li> 
<li style="margin-top:3px;"><a href="#3d">3D View</a></li> 		
	

<td valign="top" align="left" style="padding-left:20px;">
Detailed views:
<li style="margin-top:3px;"><a href="#2d3">2D - 3 Chr</a></li> 
<li style="margin-top:3px;"><a href="#2d2">2D - 2 Chr</a></li> 
<li style="margin-top:3px;"><a href="#2dc">2D Closeup</a></li>
<li style="margin-top:3px;"><a href="#base">Base View</a></li>
</td>

<td valign="top" align="left" style="padding-left:20px;">
Analysis:
<li style="margin-top:5px;"><a href="#query">Query Setup</a></li>
<li style="margin-top:3px;"><a href="#qres">Query Result</a></li>
<li style="margin-top:3px;"><a href="#tab">Gene Families</a></li>
<li style="margin-top:3px;"><a href="#ma">Multiple Alignment</a></li>
</td>	
		
</tr>
</table>
<p>

<a name="mgr"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Project Manager</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

After the synteny database has been build 
(see <a href=SystemGuide.html>System Guide</a>), the synteny can be viewed
from the manager. All projects in the database will be listed on the left panel.
<p>
<table>
<tr><td valign=top>
Selecting projects on the left panel shows them on the right panel. A check mark in the
<b>Available Syntenies</b> matrix indicates pairs that have computed synteny that can be viewed.

<p>By selecting a cell with synteny, 
the "Display for Selected Pair:" buttons are activated, which 
display synteny views for the <i><b>two-genome</b></i>. 

<p>The "Display for All 
Projects" provides <i><b>multi-genome</b></i> synteny views for all projects in the
matrix.
<td><a href="img/main.png"><img border=1 src="img/main.png" width=550></a>
</table>


<p>
<a name="exp"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Chromosome Explorer</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The Chromosome Explorer provides <i>multi-chromosome</i> for <i>multi-genome</i>.
<p>
<table><tr><td  valign=top>

Multiple chromosomes can be selected
from different species on the left-hand panel. 
<p>The default display is the Circle view shown. 
From here, one can select Dot Plot, 2D or 3D (if available).

<p>For <a href=#circ>Circle</a>,
<a href=#dp>Dot Plot</a> and <a href=#3D>3D</a>,  
multiple chromosomes can be selected and viewed (as shown below).
<p>For <a href=#2d>2D</a>, 
it is best to only have two or three chromosomes selected (as
shown below).

<td><a href="img/explorer.png"><img border=1 src="img/explorer.png" width=550 class="imgbreaks"></a>
</table>
<p>

<a name="circ"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Circle View</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The Circle view is available for <i>two-genome</i> and <i>multi-chromosome</i>.
<p>The left
image shows all chromosomes for maize and rice.
The right image shows one maize chromosome (3),
two sorghum chromosomes (3, 8) and two rice chromosomes (1, 5). 

<table><tr>
<td  valign=top><a href="img/circle-two-genome.png">
	<img src='img/circle-two-genome.png' border='1' width=350></a>
<td  valign=top><a href="img/circle-multi-chr.png">
	<img src='img/circle-multi-chr.png' border='1' width=400></a>
</table>

<a name="dp"></a>

<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Dot Plot</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The Dot Plot view is available for <i>two-genome</i> and <i>multi-chromosome</i>.
<p>The left
image shows all chromosomes for maize and rice.
The right image shows one maize chromosome (3, reference),
two sorghum chromosomes (3, 8) and two rice chromosome (1, 5). 
Synteny blocks are outlined in blue, and one can click to zoom further and
arrive at a more detailed <a href="#2d">2D display</a>. 

<table><tr>
<td  valign=top><a href="img/dot-two-genome.png">
	<img src='img/dot-two-genome.png' border='1' width=400></a>
<td  valign=top><a href="img/dot-multi-chr.png">
	<img src='img/dot-multi-chr.png' border='1' width=350></a>
</table>


<a name="blk"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Block View</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The Block View is a <i>two-genome</i> view.
<p>The synteny blocks are colored by chromosome  (see color key at top).
<p>
<table>
<tr>
<td><a href="img/block.png"><img src="img/block.png"  border=1 width=500></a>
<td  valign=top>
Clicking a chromosome pops up a window of the chromosome as shown below. Clicking
a block pops up a <a href="#2d">2D display</a>.
<p><a href="img/block2.png"><img src="img/block2.png"  border=1 width=200></a>
</table>
<p>

<a name="3d"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>3D View</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<table>
<tr><td  valign=top>The 3D is a <i>multi-chromosome</i> view. 
<p>The 3D view is not available from the web or on systems that do not have the 3D
Java libraries installed.
<td><a href="img/3D-chr.png">
	<img src="img/3D-chr.png" border=1 width=550 class="imgbreaks"></a>
</table>
<p>

<a name="2d3"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>2D 3-Chr</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<table>
<tr>
<td  valign=top>
This is a <i>three-chromosome</i> view.
<p>The dark blue down the center of the chromosomes are annotated genes.
The brown lines between chromosomes are hits, where only hits that are part
of a synteny block are shown here.
<td><a href="img/2d-multi-chr.png"><img src='img/2d-multi-chr.png' border=1 width=550></a>
</table>

<a name="2d2"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>2D 2-Chr</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


<table>
<td  valign=top>
This is a <i>two-chromosome</i> view.

<p>The Maize Chr3 and Rice Chr1 were displayed, then zoomed into the top part (i.e.
the top 7.97 Mb from the above image). It was then 
flipped (note that it is inverted in the above image), resulting the view on the right.
<td><a href="img/2d-2chr.png"><img src='img/2d-2chr.png' border=1 width=550></a>
</table>
<p>


<a name="2dc"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>2D Closeup</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<table>
<td  valign=top>

Zooming in further and turning on the annotation results in the view on the
right.
<td><a href="img/2d-closeup.png"><img src='img/2d-closeup.png'  border=1 width=550></a>
</table>
<p>

<a name="base"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Base View</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<table>
<tr><td  valign=top>It is possible to zoom in even further and see a basepair-level view
of the region, with Needleman-Wunsch alignments generated on the fly.
<td><a href="img/baseview.png"><img src='img/baseview.png' border=1 width=550> 
</table>
<p>

<a name="query"></a>

<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Query Interface</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

SyMAP includes powerful search features to help make the most of synteny information. 
Aligned regions across the selected species are automatically grouped into "putative gene
families". The filters allow you to locate the families of interest based on annotation, 
level of synteny, and presence or absence in different lineages. 
<p>
<a href="img/query_query.png"><img src="img/query_query.png" border=1 width=550></a>

<a name="qres"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Query Results</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The query results table has full flexibility in column selection, column ordering, 
and save/reload of searches. For example if you find an interesting family that is
not annotated, you can easily save all the relevant sequences as a fasta file for further study.  
<p>
<a href="img/queryres1.png"><img src="img/queryres1.png" border=1 width=550></a>

<a name="tab"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Gene Families</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The hits are grouped into gene families. 
<br>The PgeneF number is a unique number assigned to the gene family.
<br>The PgFSize number is how many unique instances are in the family.
<br>The #RGN tells how many times the gene occurs for the species. 
<br>Example: For PgeneF=20 of PgFSize=6, Rice #RGN=2 and its two set of Start coordinates 
are very different, whereas Maize #RGN=1 and its Start coordinates are all about the same.
<p>
<a href="img/querytab1.png"><img src="img/querytab1.png" border=1 width=550></a>

<a name="ma"></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left><h3>Multiple alignment</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
Another unique option is to send the selected sequences directly to Muscle for multiple
alignment, to reveal the patterns in the sequence alignments between the species. 
SyMAP parses and displays the MUSCLE (Edgar 2004 NAR:32) results:
<p>
<a href="img/multalign.png"><img src="img/multalign.png" border=1 width=550></a>
<p>

</td>
</tr>
</table>

</td>
</tr>
</table>

