<html>
<head>
	<title>SyMAP Trouble Shoot</title>
	<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
	<style>
a.wt {color:inherit;}
table {font-size:100%}
tr {font-size:100%}
td {font-size:100%}
tt {font-size:13px}
pre {font-size:13px}
a {font-size:100%}
code {font-family: courier, monospace;  font-size: 16px;font-weight: 300;font-variant: small-caps;}
	</style>
</head>
<body style="font-family:Verdana, Arial, Helvetica, sans-serif;font-size:100%">


<a name=top></a>	
<table align="center" style="border: 1px solid #999999" cellspacing="0" cellpadding="0">
	<tr><td width="100%" style="background-color: #CCCCCC;">
	<a href="http://www.arizona.edu">
	<img border="0" alt="The University of Arizona" title="The University of Arizona" src="./img/UA.gif"></a>
	</td></tr>
	
	<tr><td>
		<table width="750" border="0" align="center" cellpadding="0" cellspacing="0">
		  	<tr><td height="60" width="161" align="left" valign="bottom">
       						<img alt="" title="" src="./img/agcol_logo.png">
       			</td>
		    	<td height="60" width="428" align="left" valign="bottom" style="padding:0px 0px 10px 30px" size=5>
					<font face="Verdana, Arial, Helvetica, sans-serif" color="#000066" size="5">
				    	<b>SyMAP Trouble Shoot</b>
					</font>
		    	</td>
		    	<td height="60" width="161" align="center" valign="bottom">&nbsp;
        		</td>
		    </tr>
		    <tr><td height="21" colspan="3" align="left" valign="top">
			    <table width="100%" border="0" cellspacing="0" cellpadding="3" height="18">
			    	<tr><td align="left" bgcolor="#666666">
<span style="font-size:75%;color:white">
<a href="http://www.agcol.arizona.edu" class="wt">AGCoL</a> |
<a href="http://www.agcol.arizona.edu/software/symap" class="wt">SyMAP Home</a> |
<a href=index.html class="wt">Index</a> |
<a href=SystemGuide.html class="wt">System Guide</a> |
<a href=UserGuide.html class="wt">User Guide</a> |
<a href=Tour.html class="wt">Tour</a> |
Troubleshooting
</span>
					</td></tr>
				</table>
			</td></tr>
			
			
	        <tr><td colspan="3" valign="top" align="left" style='padding:20px 15px 0px 15px;font-size:85%'>
	        
<a name="top"></a>
<i>Please send questions and suggestions to symap@agcol.arizona.edu.</i><br>
<p>As of SyMAP v5.0.6, the Applet has been removed and will no longer be supported; 
it is in earlier SyMAP versions, and can
be obtained at <a href="https://github.com/csoderlund/SyMAP">github<a/>.

<h4>Contents</h4>
<p>
<ul>
<li>Memory
<ul>
<li><a href="#not_enough_memory1">Not enough memory</a></li>
</ul>
<br>
<li>MySQL suggested settings
<ul>
<li><a href="#msqslow">Slow to load sequence and annotation</a>
<li><a href="#slow">Slow queries</a>
</ul>
<br>
<li>MySQL 
<ul>
<li><a href="#database_error">Database connect errors</a> ("Unable to connect","Error running SQL Command",...)</li>
<li><a href="#mariadb">MariaDB cannot connect</a>
<li><a href="#innodb">The MySQL database does not support Innodb tables</a>
<li><a href="#root">Please read "Security" section of the manual to find out how to run mysqld as root</a> 
<li><a href="#msqcase"> Can't create table  (errno: 121)</a>
</ul>
<br>
<li>External programs
<ul>
<li><a href="#exec">Not executable</a>
<li><a href="#run">Fails to execute</a>
</ul>
<br>
<li>Alignment
<ul>
<li><a href="#status">Status window disappeared</a>
<li><a href="#glib">/lib64/libc.so.6: version `GLIBC_2.7' not found </a>
<li><a href="#mumFail">MUMmer failed</a></li>
<li><a href="#mum64">MUMmer/PROmer can't handle large sequences</a>	
<li><a href="#noanch">No anchors loaded</a>
<li><a href="#bes">BES not loaded</a>
</ul>
</ul>

</ul>

<hr>
<h2>Memory</h2>
<a name="#not_enough_memory1">
<h3>Not enough memory</h3>
For very large projects, or queries involving multiple
genomes (e.g. 5 or more), the Project Manager may run out of memory.
To fix this problem, run symap with the -m parameter to increase the maximum
memory used from the default 2048 MB to 4096 MB.:
<pre><b>     &#62 ./symap -m 4096</b></pre>
Alternatively, you can edit the file <tt>symap</tt> and change
<pre>
my $maxmem="2048m"; # change this line to use "4096m" 
</pre>

You will also need to be using a 64-bit processor. 

<p><a href="#top">Return to top</a><br>

<!---------------------------------------------------------!>
<hr>
<h2>MySQL suggested settings</h2>
<a name="msqslow">
<h3>Slow to load sequence and annotation</h3>

Two MySQL settings are especially important for SyMAP performance (and generally for InnoDB table
performance). 
<p>
<table width=650 cellspacing=3 cellpadding=2 border=1 rules=all>
	<tr>
		<td align=left valign=top><tt>innodb_buffer_pool_size</td>
		<td>
The default is too low for most purposes. You should set this to around 1Gb if possible (note that 
the units are bytes). 
		</td>
	</tr>
	<tr>
		<td align=left valign=top><tt>innodb_flush_log_at_trx_commit=0</td>
		<td>
The default setting is 1, which results in very slow uploading of data.
		</td>
	</tr>
</table>
<p>
To check the values, start mysql and type:
<pre>
show variables like "innodb_buffer_pool_size"; 
show variables like "innodb_flush_log_at_trx_commit"; 
</pre>
To change the variables, then type:
<pre>
set global innodb_buffer_pool_size=1073741824;
set global innodb_flush_log_at_trx_commit=0;
</pre>
Alternatively, You can set these in the MySQL configuration file <tt>my.cnf</tt> and restart MySQL. 
Note that <tt>my.cnf</tt> is typically located at <tt>/etc/my.cnf</tt>, on both Linux and Mac.

<p>Different machines and MySQL installations can produce different results with these variables.
If the SyMAP demo-seq to demo-seq2 load and synteny computations seem slow, 
try different combinations of these two variables to see what performs the fastest inserts.

<a name="slow">
<h3>Slow queries</h3>
<table width=600 cellspacing=3 cellpadding=2 border=1 rules=all>
	<tr>
		<td align=left valign=top><tt>max_allowed_packet</td>
		<td>
This should be at least 500M if possible (the units are bytes). 
		</td>
	</tr>	
</table>
<p>
To check the value, start mysql and type:
<pre>show variables like "max_allowed_packet"; </pre>
To change this variable, then type:
<pre>set global max_allowed_packet=1073741824;</pre>
<p><a href="#top">Return to top</a><br>

<!---------------------------------------------------------!>
<hr>
<h2>MySQL</h2>
<a name="database_error">
<h3>Database connect errors</h3>
After startup, if the program is left idle for some time then its
database connection will close and it will show a complicated-looking error on 
next usage, or it may say this:
<pre>
com.mysql.jdbc.CommunicationsException: Communications link failure due to underlying exception:

** BEGIN NEXTED EXCEPTION **
</pre>
In this case the solution is to restart the program. 
<p>
On startup, error messages mentioning the database or SQL probably mean that the database connection
could not be established or was lost. A message at startup means SyMAP was not able
to connect to the database using the provided information in <tt>symap.config</tt>. First, make sure
your <tt>db_adminuser</tt> and <tt>db_adminpasswd</tt> are correct.

<p>These days, the best way to find the MySQL problems is to search the internet. However,
the following are some things to check for connection problems on SyMAP startup:

<p><u>Check the database privileges</u>:
The database users specified in <tt>symap.config</tt> may not have necessary access to the database.
If you running the read/write <tt>symap</tt>, then the relevant user is <tt>db_adminuser</tt>. If 
you are running a read-only session using <tt>viewSymap</tt>, it will use the
read-only <tt>db_clientuser</tt> or read-write <tt>db-adminuser</tt>. In either case the user needs
to have database access privileges <i>from the machine where you are working</i>. 

Note that MySQL privileges are granted specifically to certain users on certain machines. You can test 
the user access by using the MySQL command line client <tt>mysql</tt>, if it is installed
on your machine, or by looking at the privilege table using an administration interface
such as PHPMyAdmin. The <tt>mysqlaccess</tt> tool is also useful, if it is installed
on your machine. 

<p><u>Check the MySQL configuration</u>:
The MySQL configuration file (<tt>/etc/my.cnf</tt>) should <b>not</b> contain any
of the following lines that prevent remote access to the database.
<tt>
<pre>     
	bind_address=127.0.0.1
	skip_networking 
</pre></tt>

<p><u>Make sure the port is visible</u>:
If the database is on a different computer, test that its port 3306 is visible
from your computer:
<pre>     telnet &lt;server address&gt; 3306
</pre>

If it does not make a connection then either the server is not running, or 
it is set to run on a non-standard port, or the
port is blocked by a firewall. Contact a system administrator. (Note, 
to get out of telnet type "<tt>^]</tt>".)

<a name="mariadb">
<h3>MariaDB cannot connect </h3>
If you are using MariaDB (post version 10.4.12) and get the error:
<pre>
Unable to connect to the mysql database on localhost; are username and password correct?
</pre>
Try adding the following lines to the <tt>/etc/my.cnf</tt>:
<pre>
[mysqld]
character-set-server=utf8mb4
</pre>
And restart MariaDB with <tt>systemctl restart mariadb</tt>. 
<p>Thanks to Lori for this solution! 

<a name="innodb">
<h3>The database does not support Innodb tables</h3>
This can happen in a couple of ways ways. Look in the mysql log for clues.  
<li> MySQL started with the "skip-innodb" flag. Solution is to remove this. 
<li> InnoDB log file problems. This can happen if you change the log file size, without
deleting the log file. Solution: shut down mysql, remove the innodb log file, and restart mysql. 


<a name="root">
<h3>Please read "Security" section of the manual to find out how to run mysqld as root</h3> 
You are launching SyMAP as the unix root user, which is blocked by MySQL 
as a security hazard. Solution is simply to use a regular non-root user account.  

<a name="msqcase">
<h3>Can't create table  (errno: 121)</h3>
This can occur if you try to create a SyMAP database whose name differs only by
case from an existing database, e.g. "symapTest" when "symaptest" already exists. 

<p><a href="#top">Return to top</a><br>

<!--------------------------------------------------------!>

<hr>
<a name="ext">
<h2>External programs</h2>

The SyMAP package comes with compiled versions of MUMmer (sequence alignment),
Blat (FPC alignment) and Muscle (SyMAP Queries). These programs are in the
<tt>symap_5/ext</tt> directory. There are compiled versions for 64-bit Linux
and 64-bit MacOS. For example, in the <tt>symap_5/ext/mummer</tt> directory,
there are sub-directories <tt>linux64</tt> and <tt>mac</tt>. SyMAP will
select the correct directory for the machine you are running from, i.e. you
do not need to do anything. 

<a name="exec"></a>
<h3>Not executable</h3>
When SyMAP creates a database, it (1) checks the MySQL variables, and (2) checks
that the external programs are executable. If you see a message like:
<pre>
***Error - file is not executable: ext/mummer/mac/promer
</pre>
Execute:
<pre>
> chmod 755 ext/mummer/mac/promer
</pre>

<h4>MacOS 10.15 (and possibly earlier versions)</h4>
Any executable that has not been okayed by Apple get the error message
"cannot be opened because the developer cannot be verified". Hence, you need
to use the Finder to enter all <tt>/ext/&lt;program&gt;/mac</tt> directories,
select each program, then choose to Open it; it will run the program in a terminal
window and be saved as a program that can thereafter be run.

<a name="run"></a>
<h3>Fails to execute</h3>
If when "Alignment&Synteny" is executing, you get a message to the terminal like:
<pre>
Error running command: /Users/cari/Workspace/github/symap/ext/mummer/mac/promer  
-p data/seq_results/demo_seq_to_demo_seq2/align/demo_seq_cc.demo_seq2_f2.promer 
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq2/demo_seq2_f2.fa 
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq/demo_seq_cc.fa
</pre>
One of the following should elucidate the problem (say the problem occurred when running <tt>promer</tt> on MacOS):
<ol>
<li>Enter the <tt>logs</tt> directory, then the directory of the pair (e.g. <tt>demo_seq_to_demo_seq2</tt>). View one
of the <tt>.log</tt> file, which contains the output of <tt>promer</tt>.
<li>Copy the command (beside "Error running command:") from the terminal, and past it on a new terminal line to execute.
This shows <tt>promer</tt> output directly on the terminal.
<li>Enter the <tt>ext/mummer/mac</tt> directory (or the appropriate one), and execute the command at the command line,
e.g <tt>./promer</tt>. 
</ol>
All the above give additional that may show the problem and lead to the fix. 
<p>
Otherwise, you may need to compile the program on your machine. In this case,
replace the existing executables with your compiled ones. For example, if you compile MUMmer,
look at the respective <tt>ext/mummer/mac</tt> directory and copy the
corresponding programs in. The <tt>promer</tt> and <tt>nucmer</tt> are perl scripts that
have paths hard-coded at the top, which need to be changed to <tt>./ext/mummer/mac/</tt>.

<h4>Early versions of MacOS</h4>
The MacOS executables were compiled on a MacOS 10.15. These will not work on old versions such as MacOS 10.9;
however, the directories with the the <tt>pre506</tt> suffixes. If you do the following:
<pre>
mv mac mac_506
mv mac_pre506 mac
</pre>
These executables compiled ona MacOS 10.9 may work  on your Mac.

<h4>32-bit</h4>
It seems that 32-bit machines are history. However, the <tt>linux</tt> directories have executables
assembled on a 32-bit machine. 

<p><a href="#top">Return to top</a><br>

<!--------------------------------------------------------!>
<hr>


<h2>Alignment</h2>
If MUMmer or Blat fail to run, see the above section.

<a name="status">
<h3>Status window disappeared</h3>
Occasionally the status window disappears when running an alignment. 
Clicking the top of the project manager sometimes
brings it back. Otherwise, just check the output to the terminal,
which tells you when it is done. You may
have to restart <tt>symap</tt> to see the results.


<a name="mumFail">
<h3>MUMmer failed</h3>
Try using a hard-masked genome sequence. Ensembl supplies both soft-masked and
hard-masked genome sequences. NCBI supplies soft-masked sequences, where the
supplied ConvertNCBI script (scripts/ConvertNCBI) can convert the file to hard-mask.

<a name="mum64">
<h3>MUMmer/PROmer cannot handle large sequences</h3>
If the alignment doesn't produce results and you see errors indicating
that PROmer cannot produce the .aaqry file, then the mummer binary may
not be 64-bit. You can tell for sure by looking at the output lines like
<pre>
# (maximum reference length is 536870908)
# (maximum query length is 4294967295)
</pre>
If you see small numbers like this, it is not 64-bit. For 64-bit, the numbers
are extremely large, much larger than any possible genome.
To build a 64-bit mummer, you must set  
<pre>
#define SIXTYFOURBITS 
</pre>
in 
src/kurtz/libbasedir/types.h (and, of course,
compile on a 64-bit machine). We fixed this problem in the 3.4 version of SyMAP, 
for lintel64, but if you're building for a different platform you'll have
to make this fix. 

<a name="noanch">
<h3>No anchors loaded</h3>
This message indicates that SyMAP could not parse the MUMmer or BLAT output files
to load the anchors. 
<p>
The most likely reason is that MUMmer or BLAT failed to run.
Usually this is because the platform (linux 32/64 bit, or mac)
was not detected properly. First try re-starting symap using
the '-32, -64, -linux, -mac' parameters to specify your platform 
(run 'symap -h' for list of parameters). 
<p>
The next likely cause is trying to run a whole-chromosome MUMmer alignment on a 32-bit
machine. MUMmer usually uses more than 2Gb per thread, which is not
possible on a 32-bit machine. If you suspect this is the problem, try running
on a 64-bit machine. If you compiled MUMmer yourself, check 
<a href="#mum64">the above section</a>.
<p>
If that does not resolve the problem, The next step is to look in the project
.log file in the /logs directory for 
more detailed error messages. Then, you should look at the MUMmer log
files (extension .mum.log) in the directory
<tt>
<pre>
logs/&lt;proj1_to_proj2&gt;
</pre>
</tt>
These contain the messages printed by MUMmer, and should show why it failed to run.  
<p>

<a name="bes">
<h3>FPC: BES not loaded, or BES hits not loading</h3>
Generally, this is caused by a name formatting error. Make sure the BES
names are formatted as clone names with extension "r" or "f". The clone
names should <i>exactly match</i> those in FPC, and there should be no other prefix or suffix. 
Do not use ".r" or ".f". 

<p><a href="#top">Return to top</a><br>
<p>&nbsp;
</div>
</table>
  <table width="100%" border="0" align="center">
    <tr>
      <td bgcolor="#DDDDDD" align="center">
       <font face="Verdana, Arial, Helvetica, sans-serif" size="2">Email
       Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></font></td>

    </tr>
  </table>
<br>

		</td>
	</tr>

</table>
</BODY>
</HTML>








