<HTML>
<HEAD>
	<TITLE>SyMAP FPC 2D </TITLE>
	<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
	<style>
a.wt {color:inherit;}
table {font-size:100%}
tr {font-size:100%}
td {font-size:100%}
tt {font-size:13px}
pre {font-size:13px}
a {font-size:100%}
	</style>
</head>
<body style="font-family:Verdana, Arial, Helvetica, sans-serif;font-size:100%">

	<a name=top></a>
	<table align="center" style="border: 1px solid #999999" cellspacing="0" cellpadding="0">
	  <tr><td width="100%" style="background-color: #CCCCCC;">
		<a href="http://www.arizona.edu">
		<img border="0" alt="The University of Arizona" title="The University of Arizona" src="./img/UA.gif"></a>
	  </td></tr>

	  <tr>
	  <td>
		  <table width="750" border="0" align="center" cellpadding="0" cellspacing="0">
		    <tr>
    	  <td height="60" width="161" align="left" valign="bottom">
       		<img alt="" title="" src="img/agcol_logo.png">
       	</td>
		    <td height="60" width="428" align="left" valign="bottom"
				style="padding:0px 0px 10px 30px">
			    <font face="Verdana, Arial, Helvetica, sans-serif" color="#000066" size="5">
				    <b>SyMAP FPC 2D</b>
			    </font> 
		    </td>
		    <td height="60" width="161" align="center" valign="bottom">&nbsp;
        </td>
		    </tr>

		    <tr>
		    <td height="21" colspan="3" align="left">
			    <table width="100%" border="0" cellspacing="0" cellpadding="3" height="18">
			    <tr>
			    <td align="left" bgcolor="#666666">
<span style="font-size:75%;color:white">
<a href="http://www.agcol.arizona.edu" class="wt">AGCoL</a> |
<a href="http://www.agcol.arizona.edu/software/symap" class="wt">SyMAP Home</a> |
<a href=index.html class="wt">Index</a> |
<a href=SystemGuide.html class="wt">System Guide</a> |
<a href=UserGuide.html class="wt">User Guide</a> |
<a href=Tour.html class="wt">Tour</a> |
<a href=TroubleShoot.html class="wt">Troubleshooting</a> 
</span>
				</td>
			</tr>
          </table>
			    </td>
			    </tr>

          <tr>
    	    <td colspan="3" valign="top" align="left" style='padding:20px 15px 0px 15px;font-size:85%'>

<p>
This guide provides detailed FPC-to-Seq instructions for the SyMAP <b>2D</b> user interface, which
includes some duplicated information from <a href=UserGuide.html target="_blank">User Guide</a> since
the FPC track and Sequence track are displayed together.
All instructions for the other displays in the <a href=UserGuide.html target="_blank">User Guide</a> 
apply to FPC-to-Seq unless stated otherwise.

<h2>Contents</h2>
<ul>
<li><a href="#explorer">Chromosome Explorer</a>
<li><a href="#alignment_display_2d">2D Display</a>
<ul>
<li><a href="#sequence_track">Sequence Track and Filter</a>
<li><a href="#block_track">FPC Block Track and Filter</a>
<li><a href="#contig_track">FPC Contig Track and Filter</a>
</ul>
<li><a href="#caveat">FPC Caveats</a>
</ul>


<a name=alignment_display_2d></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>2D Display</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
<div class="lrpad">


The 2D display is activated from the Explorer by
clicking the <b>2D</b> button, 
or by selecting a block in the Dot Plot.
<p>

Below is a image depicting an <b>FPC-to-sequence comparison</b> for Maize FPC
(chromosome 3) to Rice (chromosome 1) to Sorghum (chromosome 3). 
Note that rice chr1 is the "reference" 
chromosome, hence is placed in between the others.
<p>
<a href="img/alignment2D.png"><img border=1 src="img/alignment2D.png" width=650></a>
<p>
The alignment window displays "tracks" (drawn as rectangles) of sequence information, showing synteny  "hits" (drawn as lines) between tracks.
<p>
There are three types of tracks:
<ul>
<li><a href="#sequence_track">Sequence Track</a> - sequenced chromosome or draft sequenced contig.<br>
<li><a href="#contig_track">FPC Contig Track</a> - single contig from FPC map.<br>
<li><a href="#block_track">FPC Block Track</a> - group of contigs from FPC map.<br>
</ul>
Left click in a track shows the most commonly used filters and features (e.g.
flipping the region). They are also on the filter menu above each track
(e.g. Sequence Filter). See the Filter descriptions below.
<p>



<a name="control"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h3>Control Panel and Navigation</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The control panel provides buttons that allow for navigation between views, scaling, color customization, and help.
<p>
<a href="img/controlpanel.png"><img src="img/controlpanel.png"></a>
<p>

<b>History (Home, &lt;, &gt;)</b> SyMAP retains a record of the prior views (like a web browser). 
The history navigation back (<b><big>&lt;</big></b>) and 
forward (<b><big>&gt;</big></b>) buttons allow you to 
move back and forth through the previous views.  
The <b>Home</b> button returns to the initial view.
<p>

<b>Zoom Buttons (-, +)</b> 
The zoom buttons allow for quickly expanding and narrowing the view region. 
 The minus (<b><big>-</big></b>) button shrinks the view region by 50%, keeping
the same center, while the plus (<b><big>+</big></b>) button doubles the view region,
again keeping the same center.  
<p>

<b>Scale Button</b> 
The Scale button 
(right of <b><big>+</big></b>) resizes the tracks so that they are in the same scale
(base pairs per pixel) as the reference sequence track in the view.
<p>


<a name="mouse_function_selector"></a>
<b>Mouse Function Selector</b> 
This drop-down selector assigns a function to the mouse's left button
click and drag actions:
<ul>
<li>Base View - open the base alignment view for the selected region.</li>
<li>Zoom - zoom to the selected region.</li>
<li>Zoom All - zoom the track to the selected region and all other tracks with hits in the region.</li>
</ul>

<b>Colors</b> Opens a menu for customizing colors.
<p>
<b>Help</b> Opens this manual in a web browser.

<p>
<b>Additional navigation:</b>
<table border=1 rules="rows, cols" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr>
	<td nowrap><b>Resize Track</b>&nbsp;</td>
	<td>Drag bottom of track</td>
	<td>Position mouse at bottom of track (resize cursor appears), hold down left 
	mouse button, and move mouse.</td></tr>
<tr>
	<td><b>Scroll Track</b>&nbsp;</td>
	<td>Mouse wheel</td>
	<td>Position mouse over track and use mouse wheel.</td></tr>
<tr>
	<td><b>Filter</b>&nbsp;</td>
	<td>Right mouse button</td>
	<td>Position mouse over track or white space between tracks, and click right mouse button.</td></tr>
</table>

<p>


<a name="main"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h3>Tracks in Main Panel</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The main panel is divided into "tracks", 
where each track represents an FPC map or sequence. Each track
is aligned to the tracks to its immediate left or right, and the 
individual pieces of evidence for synteny (referred to as anchors or "hits") are drawn as 
lines between the two tracks, color-coded by type. Above each track 
is a "filter" button which allows configuration of the content displayed for 
that track (e.g., annotations), and above the space between each pair of tracks
is a "hit filter" button that allows selection of the types of hits displayed 
between the tracks on either side (e.g., only show syntenic hits).

<a name="sequence_track"></a>
<h4>Sequence Track</h4>

A <b>Sequence</b> track represents a contiguous piece of sequenced DNA, 
typically for a sequenced chromosome. The sequence length
and the coordinates of the displayed region are indicated above and below the track. 
<p>
Annotation data of the following types may be loaded into SyMAP and displayed.
(Note that display colors <a href=#control>can be changed</a>.) 
<p>
<table cellspacing=2 cellpadding=3 border=1 rules=all width=600>
	<tr>
		<td valign=top align=left>Annotation Type</td>
		<td valign=top align=left>How Displayed</td>
	</tr>
	<tr>
		<td valign=top align=left>Gaps</td>
		<td valign=top align=left>Red band across the chromosome</td>
	</tr>	
	<tr>
		<td valign=top align=left>Centromere</td>
		<td valign=top align=left>Blue "X" across the chromosome</td>
	</tr>	
	<tr>
		<td valign=top align=left>Predicted genes and exons</td>
		<td valign=top align=left>
			Annotation strip in center of chromosome (see <a href=#seqtrackimg>image</a> )</td>
	</tr>
	<tr>
		<td valign=top align=left>
			Genetically anchored ("Framework") markers (FPC maps only)</td>
		<td valign=top align=left>See images <a href=#block_track>below</a></td>
	</tr>
</table>

<p>
The sequence information displayed can be configured via the 
<a href="#sequence_filter">Sequence Filter</a> by clicking the 
<b>Sequence Filter</b> button above the sequence track, or by right-clicking the
mouse over the sequence track.
<p>
The example below shows a sequence-to-sequence alignment.

<p>
<a name="seqtrackimg">
<ul>
<li>The black lines down the center of the Brachypodium chromosome indicate predicted
genes; this annotation file did not include exons.
The black lines with heavier blue bars interspersed down the center of the rice chromosome indicate exon/intron
predictions; the annotation file included both genes and exons. 
<table><tr>
<td>

<li>The alternating brown/gray lines indicate anchor clusters. The solid brown lines indicate
a single anchor (or multiple anchors very close together), while the gray lines are spaces with no alignment.
<p>
<li>The ruler on the right side of the sequence rectangle 
shows the relative BP position along the sequence. 
<p>
<li>Additional information about each
annotation may be seen by choosing "Show Description for Annotations" from the Sequence Filter.
</ul>
<td><a href="img/sequencetrack1.png"><img src="img/sequencetrack1.png" width=400></a>
</table>

<p>

<a name="sequence_filter"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h4>Sequence Filter</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


The Sequence Filter allows the user to select the type of information shown for the 
sequence.  It is accessed via the <b>Sequence Filter</b> button above the sequence track, 
or by right-clicking the mouse over the sequence track.
<p>

<table><tr>
<td>
<b>Start and End</b>
The positions of the sequence display can be set via the 
corresponding text boxes.
The units of the values entered can be selected from the accompanying drop down menus 
(BP, KB, MB, GB).
<p>
<b>Full Sequence</b> Sets the start and end positions of
the sequence display to encompass the whole chromosome.
<p>
<b>Flip</b> Reverses the orientation of the sequence track.
<p>
<b>Show Genes</b>
Enables/disables the display of gene/exon annotations 
 along the sequence. 
<p>
<b>Show Framework Markers</b> Typically genetic markers or radiation hybrid markers.
This option enables/disables the display of framework markers (drawn as solid green 
rectangles) along the sequence.

<td><a href="img/sequencefilter.png"><img src="img/sequencefilter.png" width=300></a>
</table>


<p>
<b>Show Gaps</b> Enables/disables the display of sequence gaps (drawn as solid red 
rectangles) along the sequence.
<p>
<b>Show Centromere</b> Enables/disables the display of the centromere (drawn as a cyan "X") on 
the sequence.
<p>
<b>Show Ruler</b> Enables/disables the display of the sequence ruler along the right side 
of the sequence.
<p>
<b>Show Description for Annotations</b> Enables/disables the display of the annotation descriptions along the 
right side of sequence. NOTE: this only works if you are zoomed in close enough that they can clearly
be displayed.
<p>
<b>Show Hit Score Line</b> Enables/disables the display of the score line next to each hit along the 
sequence.  The length of the line represents the magnitude of the % Identity value for 
the hit.
<p>
<b>Show Hit Score Value</b> Enables/disables the display of the score value next to each hit along the 
sequence.  The score value corresponds to the % Identity value for the hit.
<p>
<b>Show Hit Length</b> Enables/disables the display of the hit length line next to each hit along the 
sequence.  The hit length line denotes the start and end points of the hits relative
to the sequence (may not be visible for tiny hits).
<p>

<a name="block_track"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h4>FPC Block Track</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


<table><tr>
<td valign=top>
A <b>Block</b> track shows a block of FPC contigs, which is simply 
a list of contigs arranged end-to-end. This track
type allows alignments to be displayed which stretch across more than one
FPC contig. Individual <a href="#clones">clones</a> are not shown in a
block track, but <a href="#markers">markers</a> are shown, and hits from
both clones and markers to neighboring tracks are shown.
<p>
Clicking on one of the contig rectangles of a block track
brings up the contig track display for that particular contig. The filter options
that are shared with the Contig track are maintained.
<b>Block Title</b>
A Block track may simply be a group of contigs, in which case the block title is the 
project name (e.g. "Maize").  Or a block track may contain a <b>synteny block</b>,
identified by SyMAP software, in which 
case the Block Title includes the Synteny Block Number.  The Synteny Block Number 
consists of three fields:
<pre>
Project1Chromosome.Project2Chromosome.BlockID
</pre>
In the screenshot above, the Synteny Block Number "6.1.1" refers to Maize Chromosome 6 
mapped to Rice Chromosome 1, synteny block #1 of this pairing.

<td><a href="img/blocktrack.png"><img src="img/blocktrack.png" border=1 width=150></a>
</table>
<p>


<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h4>
 Filter</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The Block Filter allows the user to select the type of information shown for the 
block.  It is accessed via the <b>Block Filter</b> button above the block track, 
or by right-clicking the mouse over the block track. See the <a href="#markers">Markers</a> section, for more information on markers.
<p>
<table><tr>
<td>
<b>Marker Name</b> Allows control over which marker names/hits to display.  
A marker is specified by entering a search string for the marker name and 
selecting the appropriate setting in the show/hide drop-down. A wild-card
character "*" may be used at either end, e.g. "*SSR*".
<p>
<b>Show No Marker Names</b> Hide all marker names.
<p>
<b>Show Only Marker Names With Hits</b> Only show the names of markers which have a
 hit somewhere along the chromosome currently being viewed. These hits
may not be visible if the view has been zoomed.
<p>
<b>Show Only Marker Names With Visible Hits</b> Only show the names of markers which have hits visible in the
current view.

<td><a href="img/blockfilter.png"><img src="img/blockfilter.png" width=300></a>
</table>

<p>
<b>Show All Marker Names</b> Display the names of all markers on the block, regardless of hits.
<p>
<b>Contig Set</b> Allows the user to define which contigs to display. 
Multiple contigs and ranges may be specified, for example
"1,2,3,5-20,31,47" will display contigs 1,2,3,31,47, and all contigs between 5 and 20. 
The contigs are displayed in the order given. 
<p>
<b>Flip</b> Flip the entire block, reversing the
display order of the contigs.
<p>

<a name="contig_track"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h4>FPC Contig Track</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<p>
<table><tr>
<td valign=top>
The <b>Contig </b> track shows a detailed view of the <a href="#clones">clones</a> 
and <a href="#markers">markers</a> that make up the selected contig. Hits from
clones to neighboring tracks are shown as lines going to the clones. If the 
hit is from a BES on the clone, the line is drawn to one end of the clone, while
if the hit is from a marker attached to that clone, the hit is drawn to the middle
of the clone. Because a marker may be attached to many clones, a hit from
a particular marker is drawn from a red "marker join dot" to the side of the 
contig, and the join dot is then connected to each of the clones to which
the marker is attached.
<p>
Since a contig track is usually accessed by choosing a contig from a block
track, a <b>Block View</b> link has been provided at the top of the contig
track which will return to the previous block view.  The filter options that
are shared with the Block track are maintained when using this link.
<p>
The contig information displayed can be controlled via the 
<a href="#contig_filter">Contig Filter</a>.

<td><a href="img/contigtrack.gif"><img src="img/contigtrack.gif" width=400 border=1></a>
</table>
<p>

<p>
<a name="clones"></a>
<b>Clones</b> 
are represented as short vertical lines.  Dots on the ends of the
line represent a BES.
<p>
<table border=1 rules="rows, cols" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr><td></td><td><b>Dot Color</b></td><td><b>Meaning</b></td></tr>
<tr><td align="center"><img src="img/clonedot_open.png"></td><td>open</td><td>the BES has no hit</td></tr>
<tr><td align="center"><img src="img/clonedot_light.png"></td><td>light purple</td><td>the BES has a hit</td></tr>
<tr><td align="center"><img src="img/clonedot_black.png"></td><td>black</td><td>the BES has a hit within a block and has no hits showing</td></tr>
<tr><td align="center"><img src="img/clonedot_dark.png"></td><td>dark purple</td><td>the BES has a hit within a block and has a hit showing</td></tr>

</table> 
<p>
<img src="img/markerhit.png">A green line in the middle of the clone going towards
another track is a <a href="#markerhit">marker hit</a>.
<p>
<img src="img/inversehit.png">Gold-colored clones are clones having both BES hits  
 with the same orientation. This can indicate an inversion breakpoint within the clone.
<p>
Moving the mouse over a clone will give a description in the status bar and highlight 
the clone and its hits. Moving the mouse over a clone's BES gives more information
on that BES's hits.
<p>
The clone information displayed can be configured via the 
<a href="#contig_filter">Contig Filter</a> by clicking the 
<b>Contig Filter</b> button above the contig track, or by right-clicking the
mouse over the contig track.

<p>
<a name="markers"></a>
<b>Markers</b>
are shown next to Block tracks and Contig tracks. Moving the mouse over
a marker name will highlight the corresponding hits. Clicking on a marker's name will
keep the marker highlighted until another marker is clicked or the same marker is 
clicked again.
<p>
Markers are colored based on their hit information and if they are shared.
<p>
<table border=0 style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr><td></td><td>Not Shared</td><td>Shared</td></tr>
<tr><td>No Hits</td><td bgcolor="#BBBBBB"></td><td bgcolor="#A0A0FF"></td></tr>
<tr><td>Hits</td><td bgcolor="#FFAAAA"></td><td bgcolor="#5A5AFF"></td></tr>
<tr><td>Current Hits</td><td bgcolor="#149646"></td><td bgcolor="#0000FF"></td></tr>
</table>
<p>
<table border=1 rules="rows, cols" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr>
	<td><b>Select Marker</b></td>
	<td>Left mouse button</td>
	<td>position mouse over marker and click left mouse button.</td></tr>	
<tr>
	<td><b>Highlight Marker/Clone</b></td>
	<td></td>
	<td>position mouse over marker/clone.</td></tr>	
</table>
<p>

<a name="beshit"></a>
<b>BES Hit</b> is signified by a purple line between two tracks.

<p>
<a name="markerhit"></a>
<b>Marker Hit</b> is signified by a green line between tracks.  If a
marker of the same name can be found on another track in the view, the marker hit
will be blue.  When viewing a contig in the Contig view, the hits
from the same marker on different clones are joined together at a red
dot. Moving the mouse over the red dot will highlight the marker's name.

<p>
<a name="fingerprinthit"></a>
<b>Clone Fingerprint Hit</b> is signified by a black line between two tracks.

<a name="contig_filter"></a>
<b>Contig Filter</b>
allows the user to select the type of information shown for the 
currently visible contig.  It augments the Block Filter with options specific to clones.  
It is accessed via the <b>Contig Filter</b> button above the contig track, 
or by right-clicking the mouse over the contig track.
<p>

<table><tr>
<td>
<b>Change Contig</b> a quick way to change the currently viewed contig.
<p>
<b>Clone Name:</b> The Clone names and hits can be shown, hidden, or highlighted through the Clone
Name text box. A clone name can be searched on by entering a search string
for the clone name and choosing to show, hide, or highlight those clones that match it.
A wildcard "*" may be used at the beginning or end.
<p>
<b>Clone with Remarks:</b> Clones can be filtered or highlighted based on the remarks attached to them in the FPC file.
For convenience, all remarks present in the contig are listed in the selection box, 
and one or more may be selected for filtering (to select multiple, hold down the 
Control key while clicking).  Clones containing at least one of the remarks selected
are filtered or highlighted based on the option selected above the remarks selection box.
<p>
<b>Start and End</b> The portion of the contig to be displayed can be adjusted through the start and end text boxes.
The desired start and end points are entered in CB units.
<p>
<a name="width"></a>
<b>Width of Contig</b> The width of the contig can be adjusted using the slider.
<p>
<b>Show Clone Names</b> The user can choose to show clone names by selecting this
check box.  The width of the contig can than be adjusted to 
give sufficient space for the names.
<p>
<b>Flip the Contig</b> Reverse the order in which the clones are displayed.
<td><a href="img/contigfilter.png"><img src="img/contigfilter.png" width=300></a>
</table>

<p>
<b>Show Only Clones with current BES paired hits</b> Show only those clones for which both BES have a hit showing in the current view.
<p>
<b>Show Only Clones with BES paired hits</b> Show only those clones for which both BES have hits (filled-in circles).
<p>


<a name="hit"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h4>Hit Filter</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


The Hit Filter menu allows the user to select which types of hits are displayed,
and what stringency of alignment is required. There are several types of hits which may be shown or hidden:
<ul>
<li>Marker hits</li> 
<li>BES hits</li> 
<li>Synteny hits, i.e. those found by SyMAP analysis to be part of a synteny block. </li> 
<li>Gene hits, i.e. hits to an annotated gene region.</li> 
</ul>

<a name="hitfilter">
<p>
<table><tr>
<td>
<p>
<b>E-Value</b>: Hit types can be filtered by E-Value using the slider. 
When the slider is all the way to the left, all hits are shown.
<p>
<b>% Identity</b>: Hit types can be filtered by percent identity using this slider.
<p>
<b>Show Only Shared Marker Hits</b>: This option applies to a 3-track (FPC/Seq/FPC) view. When selected, only marker hits 
for markers that are shared between the two FPC tracks are shown.
<p>
<b>Show Marker Join Dot</b>: The join dot is the red dot connecting all of the clone hits of a marker. 
This option allows the user to show the marker hits without the join dot. 
When this option is selected, marker hits will be drawn directly from the 
centers of the clones to hit locations on the sequence. 
This can make synteny more visible, but it also adds
some redundant lines for markers which are attached to many clones. 
 

<td><a href="img/hitfilter.png"><img src="img/hitfilter.png" withd=300></a>
</table>
<p>
<b>Show Only Synteny Hits</b>: Shows only hits that are part of a synteny block.
<p>
<b>Show Only Hits to Genes</b>: Show only hits which intersect a gene annotation. 
<p>
<b>Show Only Non-Gene Hits</b>: Show only hits which do not intersect a gene annotation. 
<p>
<b>Show All Hits</b>: shows all known hits.
<p>




<a name="base"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h4>Base View</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>


Select "Base View" from the "Mouse" dropdown. The base view displays a detailed diagram of the selected hit.
<p>
<a href="img/closeup.png"><img src="img/closeup.png"></a>
<p>
A base view of the hits along a sequence track can be brought up by dragging the 
mouse along the sequence and releasing when the desired range is highlighted. The range 
selected is increased if necessary to show the full length of the markers and BESs 
involved. If the user selected range contains blank areas (no marker or BES hits) the 
range is reduced to include only the genes with hits. There is a maximum range of 50Kb 
that can be selected for the base view, so zooming in on an area first may be 
necessary.
<p>

If there are multiple hits in the selected region, than the base view of those hits 
appears in a new dialog.  This view consists of a ruler along the top showing the area 
of the sequence covered, the hits, and the genes.
<p>

<b>Hits</b><br>
Hits are displayed as lines with an arrow on one end showing the direction of the hit.  
Marker hits are green, while BES hits are purple.Clicking on a hit brings up the hit's 
BLAST view in the bottom of the dialog. The hit presently shown in the BLAST view is 
shown in gray.  The full length of the marker/BES is shown as a grey dotted line above 
the hits.A vertical red line along a hit represents a mismatch .  A vertical green line 
along a hit represents a deletion. An arrow pointing down (i.e. 'v') along a hit 
represents an insertion.
<p>

<b>Genes</b><br>
Annotated genes are displayed below the hits.  Exons are represented by a blue box, and 
the leftmost or rightmost exon box will have an arrow tip indicating whether the gene 
is on the + strand (right-pointing) or - strand (left-pointing) relative to the 
sequence. The system attempts to expand the view to show the full gene that has 
hits. Genes that overlap this gene will be partially shown.
<p>

</div>

<a name=caveat></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left><h2>FPC Caveats</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
<div class="lrpad">

<ul>
<li>If contigs are not anchored to chromosome, they will be on chromosome 0. If chromosome 0 is part of any blocks, they are only be shown on the block display.
<li>The block display only shows the FPC chromosomes aligned to the sequence chromosomes; the reverse does not work.
</ul>

</div>
</table>
  <table width="100%" border="0" align="center">
    <tr>
      <td bgcolor="#DDDDDD" align="center">
       <font face="Verdana, Arial, Helvetica, sans-serif" size="2">Email
       Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></font></td>

    </tr>
  </table>
<br>

		</td>
	</tr>

</table>
</BODY>
</HTML>
