<HTML>
<HEAD>
	<TITLE>SyMAP User Guide</TITLE>
	<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
	<style>
a.wt {color:inherit;}
table {font-size:100%}
tr {font-size:100%}
td {font-size:100%}
tt {font-size:13px}
pre {font-size:13px}
a {font-size:100%}
		.emph {font-weight:bold;}
		.semph {font-weight:bold;color:red;font-style:italic;}
		.hang {margin-left:5px; text-indent:-5px;}
		.rpad{padding-right:20px;}		
		.lrpad{padding-left:10px;padding-right:20px;}
		li {margin-top:2px;padding-left:2px}
		.smbreak {margin-top:20px;}
		.lgbreak {margin-top:40px;}
		.imgbreaks {margin-top:20px;margin-bottom:20px;}
		.codesec{padding-left:40px;margin-top:13px;margin-bottom:13px;font-weight: 500; font-family: courier, monospace;  font-size: 14px; color: #000000;}
	</style>
</head>
<body style="font-family:Verdana, Arial, Helvetica, sans-serif;font-size:100%">
<!--#include virtual="/common/uaheader.html" -->

	<a name=top></a>
	<table align="center" style="border: 1px solid #999999" cellspacing="0" cellpadding="0">
	  <tr>
	  </tr>

	  <tr>
	  <td>
		  <table width="750" border="0" align="center" cellpadding="0" cellspacing="0">
		    <tr>
    	  <td height="60" width="161" align="left" valign="bottom">
       		<img alt="" title="" src="img/agcol_logo.png">
       	</td>
		    <td height="60" width="428" align="left" valign="bottom"
				style="padding:0px 0px 10px 30px">
			    <font face="Verdana, Arial, Helvetica, sans-serif" color="#000066" size="5">
				    <b>SyMAP User Guide</b>
			    </font> 
		    </td>
		    <td height="60" width="161" align="center" valign="bottom">&nbsp;
        </td>
		    </tr>

		    <tr>
		    <td height="21" colspan="3" align="left">
			    <table width="100%" border="0" cellspacing="0" cellpadding="3" height="18">
			    <tr>
			    <td align="left" bgcolor="#666666">
<span style="font-size:75%;color:white">
<a href="http://www.agcol.arizona.edu" class="wt">AGCoL</a> |
<a href=../index.html class="wt">SyMAP Home</a> |
<a href=SystemGuide.html class="wt">System Guide</a> |
User Guide |
<a href=Tour.html class="wt">Tour</a> |
<a href=TroubleShoot.html class="wt">Troubleshooting</a> 
</span>
				</td>
			</tr>
          </table>
			    </td>
			    </tr>

          <tr>
    	    <td colspan="3" valign="top" align="left" style='padding:20px 15px 0px 15px;font-size:85%'>

<p>
Document updated 26-Jan-20. This guide is for the Seq-to-Seq (sequence) synteny user interface.
Note, the term <i>pseudo</i>
and <i>sequence</i> are used interchangably. 
<ol> 
<li>Most of the contents also apply to the FPC-to-Seq view (unless stated
otherwise). Specific information is provided in <a href=UserGuideFPC.html target="_blank">FPC specific</a>.
<li>For detailed information on installation, see
the <a href="SystemGuide.html" target="_blank">System Guide</a>. 
<li>For a quick visual introduction, see the <a href="Tour.html" target="_blank">Tour</a>.
<li>Please send bug reports and suggestions to symap@agcol.arizona.edu.</i>
</ol>

<h2>Contents</h2>
The Dot Plot and Circle displays are available from various places.
<ul>
<li>A. Main Display
<li>B. Display for Selected Pair (Two-Genome Views):
	<ul>
	<li><a href="#dotplot_display">Dot Plot (1)</a> <!- # ref in code -!>
		<ul>
		<li><a href="#dotplot_control">Dotplot Control Panel
		<li><a href="#dotplot_filter">Dotplot Filters
	</ul>

	<li><a href="#block">Block Display</a>
	<li><a href="#circle">Circle Display (1)</a>
	<li><a href="#sum">Summary</a>
	</ul>

<li>C. Display for All Projects (Multi-Genome Views):
	<ul>
	<li><a href="#explorer">Chromosome Explorer</a> (Multi-Chromosome views)
		<ul>
		<li><a href="#alignment_display_circ">Circle Display (2)</a>
		<li><a href="#dot2">Dot Plot (2)</a>
		<li><a href="#alignment_display_3d">3D Display (if available)</a> <!- # ref in code -!>
		<li><a href="#alignment_display_2d">2D Display</a> <!- # ref in code -!>
			<ul>
			<li><a href="#control">Control Panel and Navigation</a>
			<li><a href="#main">Tracks in Main Panel</a>
				<ul>
				<li><a href="#sequence_track">Sequence Track and Filter</a>
				<li><a href="#hit">Hit Filter</a>	
				<li><a href="UserGuideFPC.html" target="_blank">FPC Track and Filter</a>
				</ul>
			<li><a href="#base">Base View</a>
			
			</ul>
		</ul>
	<li><a href="#dot3">Dot Plot (3)</a>
	<li><a href="#annotsearch">SyMAP Queries </a>(Seq-to-Seq only)
	</ul>
<li><a href="#caveat">Self-synteny Displays</a>
<li><a href="#data_download">Data Download</a>
<li><a href="#image">Print and Help</a>
</ul>

The SyMAP User Agreement requires you cite one of the following articles 
if you use SyMAP in a paper, poster, or presentation.

<b><pre>
        C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
        SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps 
        Genome Research 16:1159-1168.
</pre></b>
<b><pre>
        C. Soderlund, M. Bomhoff, and W. Nelson (2010) 
        SyMAP: A turnkey synteny system with application to plant genomes.
        Nucleic Acids Res 39(10):e68.
</pre></b>

<h1>A. Main display</h1>
The following commands result in the Manager image similar to the one shown below:
<table border=1 cellspacing=3>
<tr><td><i>v5</i><td><i>v4.2</i><td><i>Description</i>
<tr><td>viewSymap<td>symap -r -no3d<td>Read only, no 3D
<tr><td>viewSymap3D<td>symap -r <td>Read only, with 3D
</table>
<p>
<table>
<tr><td>
All projects in the database will be listed on the left panel.
<p>Selecting projects on the left panel shows them on the right panel. A check mark in the
<b>Available Syntenies</b> table indicates pairs that have computed synteny that can be viewed.
<p>By selecting a cell with synteny, the <b>Display for Selected Pair:</b> buttons are activated.

<td><a href="img/main.png"><img border=1 src="img/main.png" width=500></a>
</table>

<h1>B. Display for Selected Pair</h1>
The four buttons beside this label are "whole genome" displays for the selected pair (highlighted yellow).
<!------------------------------------------------------------!>
<a name=dotplot_display></a>
<table width="95%" cellspacing=0 cellpadding=0 class='lgbreak'>
<tr><td align=left>
<h2>Dot Plot (Two Genome)</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

Dots represent anchors (also referred to as <i>hits</i>). A blue box indicates a
<i>Synteny Block</i>. The horizontal and vertical grid lines
delineate the chromosomes (or linkage groups) of the two genomes being compared.
<p>
<table>
<tr>
<td>
 The
individual grid boxes (cell) for each chromosome pair are referred to as <i>Regions</i>, and clicking
on a region brings up a larger-scale <i>Region view</i>.  
<p>Within the region view:
<ul>
<li>Click on a blue box or drag the mouse to highlight a sub-region.</li>
<li>Then, click on the highlighted sub-region to bring up a SyMAP 2D view for that sub-region.</li> 
<li>This is illustrated in the <a href=#dot2>chromosome Dotplot</a>.
</ul>

<td><a href="img/dot-two-genome.png"><img border=1 src="img/dot-two-genome.png" width=500></a>
</table>

<a name=dotplot_control></a>
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h3>Dotplot Control Panel</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
<a href="img/controlpanel2.png"><img border="1" src="img/controlpanel2.png"></a>
<p>
<b>Home</b> Resets the zoom selection to the full grid view of the entire dot plot.
<p>
<b>Zoom</b> The plus (<b><big>+</big></b>) and minus (<b></big>-</big></b>) buttons shrink or grow the display.
<p>
<b>Scale</b> The check box resizes the species displayed on the y-axis according to the current scale of
the species displayed on the x-axis.
<p>
<b>Reference</b> The drop-down menu allows the reference species on the x-axis to be changed.
<p>
<b>Size</b> The slider increases/decreases the size of the dots in the display.


<a name=dotplot_filter></a>
<table width="95%" cellspacing=0 cellpadding=0 class='smbreak'>
<tr><td align=left>
<h3>Dotplot Filters</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<table>
<tr><td>
The Filter image on the right indicates the default settings.
<p><b>Show Only Block Hits</b> 
A <i>block hit</i> is a hit that the SyMAP synteny-finding algorithm
used to create a synteny block. 
<p>
<b>Show All Hits</b> 
Displays all hits. Note that the SyMAP hit loading algorithm removes
some repetitive hits from the raw anchor set, so some hits from the raw BLAT
or MUMmer files will not be shown. 
<p>
<b>Highlight Block Hits</b> 
Highlight hits that make up a synteny block in blue.
<p>
<b>Show Blocks</b> 
This option draws blue rectangles around the synteny blocks.
<p>
<b>Show Empty Regions</b> 
When deselected, hides regions that do not contain any synteny blocks.
<td><a href="img/dotplot_filter.png"><img src="img/dotplot_filter.png"></a>
</table>
<!-------------------------------------------------!>
<a name=block></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Blocks Display (Two Genome)</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The image below shows the block display for Maize and Rice.
The 10 Maize chromosomes are color coded as shown at the top, and aligned to
the 12 rice chromosomes. The Rice to Maize block view can be viewed by
selecting <u>reverse</u>.
<table>
<tr>
<td><a href="img/block.png"><img src="img/block.png"  border=1 width=500></a>
<td>Clicking a chromosome pops up a window of the chromsome as shown below. Clicking
a block pops up a 2D view.
<p><a href="img/block2.png"><img src="img/block2.png"  border=1 width=170></a>

</table>
<!-------------------------------------------------!>
<a name=circle></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Circle Display (Two Genome)</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The circle display shows the chromosomes arranged in a circle, 
with synteny blocks shown as colored
ribbons between the chromosomes. 
The color of a ribbon comes from one of the two chromosomes it connects; for example,
in the image below, the Rice Chr1 is red and the corresponding ribbon is red.
Clicking on a species names causes its colors to be used preferentially (<i>sometimes it
is necessary to click at the very beginning or end of the name for this to work</i>).
Clicking on a chromosome (e.g. the the red part of the circle for Rice Chr 1) 
causes ribbons from that chromosome to be shown on top. 
<p>
<table>
<tr><td>
The three controls on the left above the circle image control various aspects of the display. 
The first two buttons (<b>+, -</b>) change the size, while the third rotates the image. 
<p>
<b>Scale to genome size:</b> This checkbox causes the chromosomes to be drawn proportionately to their
actual length in basepairs. Otherwise, each species is allocated the same amount of space,
e.g. for two species each one gets exactly half the circle. 
<p>
<b>Show all blocks:</b> 
The drop-down menu has "Show inverted blocks", "Show non-inverted blocks",
and "Two-color scheme", where this last option shows inverted=green, non-inverted=red.
<p>
<b>Self-align: </b> This checkbox controls whether blocks are shown between chromosomes 
of the same species (if there is a check mark in the projects self-synteny cell indicating
its been computed). 
<p>
<b>Rotated text:</b> Depending on the system, it may be easier (or harder) to read with
the text rotated.

<td><a href="img/circle-two-genome.png">
	<img src="img/circle-two-genome.png"  border=1 width=450></a>
</table>
<p><i>Note, if you are running SyMAP over an X-windows session on a remote server,
the circle display will draw very slowly. For this case, the 3D display (if available) is 
recommended (or, use VNC or screen sharing to the remote server). </i>
<!-------------------------------------------------!>
<a name=sum></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Summary (Two Genome)</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The <b>Help</b> button on this view explains the different columns.
<p><a href="img/sum-two-genome.png">
	<img src="img/sum-two-genome.png"  border=1 width=650></a>
<!-------------------------------------------------!>

<h1>C. Display for All Projects</h1>
Select two or more projects and the "Chromosome Explorer" and "Dotplot" buttons
will be activated. 

<a name=explorer></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Chromosome Explorer</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

<p>
The Explorer is a two-panel display in which the left panel is used to select specific chromosomes, while the right panel 
shows the synteny for the selected chromosomes. Four different right-panel views are available: 
<a href="#alignment_display_circ">circle</a>,
<a href="#alignment_display_2d">2D</a>,
<a href="#dotplot_display">Dotplot</a>,
and <a href="#alignment_display_3d">3D</a> (the 3D is not available through the web and only on systems that have the Java 3D libraries). 
Below is a snapshot of the Explorer showing the circle format, which is the initial view.

<a name="circfig">
<a href="img/explorer.png">
	<img border=1 src="img/explorer.png" width=650 class="imgbreaks"></a>

<p><b>Left panel: </b>The left panel controls which species and chromosomes are shown:
<ul>
<li>Click a chromosome rectangle to add it to the display in the right panel. Click it again to remove
it.
<li>Click the chromosome number above a chromosome to make it the <i>reference</i>. (The 
choice of reference sequence does not matter for the circle view,
but it is important for the 2D, 3D, and Dot Plot views; see below). 
<li> Adding and removing chromosomes affects the 3D and Circle views instantly. 
To add/remove from the 2D or Dot Plot
views, open either 3D or Circle view, make the change, and then re-open the 2D or Dot Plot.
</ul>  
<p>
Instructions: As the mouse moves over various
components, instructions for that component are shown in the Instruction box. 
<p>

The four buttons in the lower-left corner of the window change the view in the right panel, 
where the choices are 
<a href="#alignment_display_3d">3D</a>,
<a href="#alignment_display_2d">2D</a>,
<a href="#dotplot_display">Dotplot</a>,
 and <a href="#alignment_display_circ">Circle</a>.
<p>
Download Blocks: Exports a table of all of the synteny block co-ordinates for the selected species; see also <a href="#data_download">Data Download</a>.
<p>
Click the minus (<b><big>-</big></b>) button next to a project name or the Instruction box, 
and that section will be hidden;
the minus sign will change to a plus.  
Click the plus (<b><big>+</big></b>) button to show the project or Instruction box..


<p><b>Right panel:</b> The right panel shows the synteny display for the species and chromosomes selected on the left. 
There are four views: <a href="#alignment_display_3d">3D</a> (if available),
<a href="#alignment_display_2d">2D</a>,
<a href="#dotplot_display">Dotplot</a>,
 and <a href="#alignment_display_circ">Circle</a>.
The first three modes are
essentially overview modes, while the 2D view allows zooming in to details, all the way
to the basepair level if desired.
Each right-panel display format has its own controls; see the individual sections for details.

<!-----------------------------------------------------------------!>
<a name=alignment_display_circ></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>Circle Display (Multi-chromosome)</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

This view is shown in the above image. 
All features are described in the first <a href=#circle>circle section</a>,
except it is chromosome-based instead of genome-based.

<!-----------------------------------------------------------------!>
<a name=dot2></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>Dot Plot (Multi-chromosome)</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
All features are described in the first <a href=#dotplot_display>dot plot section</a>,
except that it is chromosome-based instead of genome-based.
Selecting a region followed by a synteny block (or selected region)
brings up the 2D view.
<table>
<tr><td><a href="img/dot-chr.png">
	<img border=1 src="img/dot-chr.png" width=350 class="imgbreaks"></a>
<td><a href="img/dot-chr-2d.png">
	<img border=1 src="img/dot-chr-2d.png" width=350 class="imgbreaks"></a>
</table>
<!-----------------------------------------------------------------!>
<a name=alignment_display_3d></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>3D Display (if available)</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The 3D is not available through the web and only on systems that have the Java 3D libraries.
<p>
The 3D display shows the selected chromosomes arrayed at equal spacing around the reference, which is
placed in the center (see figure below):

<table>
<tr><td>
The synteny blocks are shown as colored ribbons, with direct synteny blocks colored red, and inverted
blocks colored green. (Note that the direct and inverse designations are for the overall block, but blocks may
contain smaller-scale regions which are inverted relative to the block as a whole.)
<p>
 Drag the mouse over
the 3D view to rotate, move, or zoom the view, depending on which of the three Navigation buttons at the top is 
selected. For example, if the leftmost button is selected, as shown in the figure, then dragging with the
mouse causes the figure to rotate. 
<td><a href="img/3D-chr.png">
	<img src="img/3D-chr.png" border=1 width=500 class="imgbreaks"></a>
</table>
<!--------------------------------------------------------------------!>
<a name=alignment_display_2d></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>2D Display (Multi-chromosome)</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The 2D display is activated from the <a href="#explorer">Explorer</a> by
clicking the <b>2D</b> button (or by selecting a region or block from a dot plot).
<p>
<table>
<tr><td>
On the right is an image depicting an alignment from Rice Chr10 to Maize Chr1 to Sorghum Chr1.
Note that Maize Chr1 is the <i>reference</i> 
chromosome, hence is placed in between the others.
<p>
The alignment window displays <i>tracks</i> (drawn as rectangles) of sequence information, 
showing synteny <i>hits</i> (drawn as lines) between tracks.
<p>
Left click in a track shows the most commonly used filters and features (e.g.
flipping the region). They are also on the filter menu above each track
(e.g. Sequence Filter). See the Filter descriptions below.
<td><a href="img/alignment2Ds.png">
	<img border=1 src="img/alignment2Ds.png" width=550></a>
</table>


<a name="control"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h4>Control Panel and Navigation</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The control panel provides buttons that allow for navigation between views, scaling, color customization, and help.
<p>
<a href="img/controlpanel.png"><img src="img/controlpanel.png"></a>
<p>
<b>History (Home, &lt;, &gt;)</b> SyMAP retains a record of the prior views (like a web browser). 
The history navigation back (<b><big>&lt;</big></b>) and 
forward (<b><big>&gt;</big></b>) buttons allow you to 
move back and forth through the previous views.  
The <b>Home</b> button returns to the initial view.
<p>
<b>Zoom Buttons (-, +)</b> 
The zoom buttons allow for quickly expanding and narrowing the view region. 
 The minus (<b><big>-</big></b>) button shrinks the view region by 50%, keeping
the same center, while the plus (<b><big>+</big></b>) button doubles the view region,
again keeping the same center.  
<p>
<b>Scale Button</b> 
The Scale button 
(right of <b><big>+</big></b>) resizes the tracks so that they are in the same scale
(base pairs per pixel) as the reference sequence track in the view.
<p>
<a name="mouse_function_selector"></a>
<b>Mouse Function Selector</b> 
This drop-down selector assigns a function to the mouse's left button
click and drag actions:
<ul>
<li>Base View - open the base alignment view for the selected region (see <a href=#base>Base View</a>.</li>
<li>Zoom - zoom to the selected region.</li>
<li>Zoom All - zoom the track to the selected region and all other tracks with hits in the region.</li>
</ul>
<b>Colors</b> Opens a menu for customizing colors.
<p>
<b>Additional navigation:</b>
<table border=1 cellspacing=3 rules="rows, cols" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr>
	<td nowrap><b>Resize Track</b>&nbsp;</td>
	<td>Drag bottom of track</td>
	<td>Position mouse at bottom of track (resize cursor appears), hold down left 
	mouse button, and move mouse.</td></tr>
<tr>
	<td nowrap><b>Scroll Track</b>&nbsp;</td>
	<td>Mouse wheel</td>
	<td>Position mouse over track and use mouse wheel.</td></tr>
<tr>
	<td><b>Filter</b>&nbsp;</td>
	<td>Right mouse button</td>
	<td>Position mouse over track or white space between tracks, and click right mouse button.</td></tr>
</table>


<a name="main"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h4>Tracks in Main Panel</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The main panel is divided into <i>tracks</i>, 
where each track represents a sequence or FPC map (see <a href=UserGuideFPC.html target="_blank">
FPC specific</a> for details of the FPC tracks). Each track
is aligned to the track oh its immediate left or right, and the 
individual pieces of evidence for synteny (referred to as <i>anchors</i> or <i>hits</i>) are drawn as 
lines between the two tracks. Above each track 
is a "filter" button which allows configuration of the content displayed for 
that track (e.g. annotations), and above the space between each pair of tracks
is a "hit filter" button that allows selection of the types of hits displayed 
between the tracks on either side (e.g. only show syntenic hits).

<a name="sequence_track"></a>
<h4><i>Sequence Track and Filter</i></h4>

A Sequence track may represent a sequenced chromosome, linkage group or scaffold. 
The sequence length
and the coordinates of the displayed region are indicated above and below the track. 
<p>
Annotation data of the following types may be loaded into SyMAP and displayed.
(Note that display colors <a href=#control>can be changed</a>.) 
<p>
<table cellspacing=2 cellpadding=3 border=1 rules=all width=600>
	<tr>
		<td valign=top align=left><i>Annotation Type</i></td>
		<td valign=top align=left><i>Display</i></td>
	</tr>
	<tr>
		<td valign=top align=left>Gaps</td>
		<td valign=top align=left>Red band across the chromosome</td>
	</tr>	
	<tr>
		<td valign=top align=left>Centromere</td>
		<td valign=top align=left>Blue "X" across the chromosome</td>
	</tr>	
	<tr>
		<td valign=top align=left>Predicted genes and exons</td>
		<td valign=top align=left>
			Annotation strip in center of chromosome (see <a href=#seqtrackimg>image</a> )</td>
	</tr>
</table>

<p>
The sequence information displayed can be configured via the 
<a href="#sequence_filter">Sequence Filter</a> by clicking the 
<b>Sequence Filter</b> button above the sequence track, or by right-clicking the
mouse over the sequence track.
<p>
The example below shows a sequence-to-sequence alignment, which was derived by right-clicking
at the top of the region, pulling down the mouse and releasing in order to view a small region.

<p>
<a name="seqtrackimg">
<table>
<tr><td>
The black lines with heavier blue bars interspersed down the center indicate exon/intron
predictions. A solid black line would indicate the absense of introns in the annotation.
<p>The alternating brown/gray lines indicate anchor clusters. The solid brown lines indicate
a single anchor (or multiple anchors very close together), while the gray lines are spaces with no alignment.
<p>The ruler on the right side of the sequence rectangle 
shows the relative BP position along the sequence. 
<p>Additional information about each
annotation may be seen by choosing "Show Description for Annotation" from the Sequence Filter.

<td><center><a href="img/2d-closeup.png">
	<img src="img/2d-closeup.png" border=1 width=500></a></center>
</table>

<p>

<a name="sequence_filter"></a>
The Sequence Filter allows the user to select the type of information shown for the 
sequence.  It is accessed via the <b>Sequence Filter</b> button above the sequence track, 
or by right-clicking the mouse over the sequence track.
<p>
<b>Start and End</b>
The positions of the sequence display can be set via the 
corresponding text boxes.
The units of the values entered can be selected from the accompanying drop down menus 
(BP, KB, MB, GB).
<p>
<table>
<tr><td>

<p>
<b>Full Sequence</b> Sets the start and end positions of
the sequence display to encompass the whole chromosome.
<p>
<b>Flip</b> Reverses the orientation of the sequence track.
<p>
<b>Show Genes</b>
Enables/disables the display of gene/exon annotations 
 along the sequence.  
<p>
<b>Show Framework Markers</b> Typically genetic markers or radiation hybrid markers.
This option enables/disables the display of framework markers (drawn as solid green 
rectangles) along the sequence.
<p>
<b>Show Gaps</b> Enables/disables the display of sequence gaps (drawn as solid red 
rectangles) along the sequence.
<p>
<b>Show Centromere</b> Enables/disables the display of the centromere (drawn as a cyan "X") on 
the sequence.
<p>

<td>
<a href="img/sequencefilter.png">
	<img src="img/sequencefilter.png" width=300></a>
</table>
<p>
<b>Show Description for Annotations</b> Enables/disables the display of the annotation descriptions along the 
right side of sequence. NOTE: this only works if you are zoomed in close enough that they can clearly
be displayed.
<p>
<b>Show Ruler</b> Enables/disables the display of the sequence ruler along the right side 
of the sequence.
<p>
<b>Show Hit Length</b> Enables/disables the display of the hit length line next to each hit along the 
sequence.  The hit length line denotes the start and end points of the hits relative
to the sequence (may not be visible for tiny hits).
<p>
<b>Show Hit Score Line</b> Enables/disables the display of the score line next to each hit along the 
sequence.  The length of the line represents the magnitude of the % Identity value for 
the hit.
<p>
<b>Show Hit Score Value</b> Enables/disables the display of the score value next to each hit along the 
sequence.  The score value corresponds to the % Identity value for the hit.

<a name="hit"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h4><i>Hit Filter</i></h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
<a name="hitfilter">
The Hit Filter menu allows the user to select which types of hits are displayed.
<table>
<tr><td>
<p>
The types of filter are as follows:

<p>
<b>Show Only Synteny Hits</b> Shows only hits that are part of a synteny block.
<p>
<b>Show Only Hits to Genes</b> Show only hits which intersect a gene annotation. 
<p>
<b>Show Only Non-Gene Hits</b> Show only hits which do not intersect a gene annotation. 
<p>
<b>Show All Hits</b> hows all known hits.

<td><a href="img/hitfilterS.png">
	<img src="img/hitfilterS.png" width=300></a>
</table>

<a name="base"></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h4>Base View</h4></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

Select "Base View" from the "Mouse" drop-down, then select a region of maximum 50kb. 
To select a region,  drag the 
mouse along the sequence and release when the desired range is highlighted.  
<p>
<a href="img/baseview.png">
	<img src="img/baseview.png" width=750 border=1></a>
<p>
If there are multiple hits in the selected region, than the base view of those hits 
appears in a new dialog.  This view consists of a ruler along the top showing the area 
of the sequence covered, the hits, and the genes.
<p>
<b>Hits</b><br>
Hits are displayed as lines with an arrow on one end showing the direction of the hit.  
Clicking on a hit brings up the hit's 
BLAST view in the bottom of the dialog. The hit presently shown in the BLAST view is 
shown in gray.  A vertical red line along a hit represents a mismatch .  
A vertical green line 
along a hit represents a deletion. An arrow pointing down (i.e. 'v') along a hit 
represents an insertion.
<p>
<b>Genes</b><br>
Annotated genes are displayed below the hits.  Exons are represented by a blue box, and 
the leftmost or rightmost exon box will have an arrow tip indicating whether the gene 
is on the + strand (right-pointing) or - strand (left-pointing) relative to the 
sequence. The system attempts to expand the view to show the full gene that has 
hits. Genes that overlap this gene will be partially shown.

<!--------------------------------------------------------------------!>
<a name=dot3></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>Dot Plot (Multi-genome)</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
All selected projects will be shown in the Dot Plot. The interface is the same as
discussed in <a href=#dotplot_display>Dot Plot (Two Genome)</a>. The dot plot in the image
shows three genomes.
<p><a href="img/wg-dp.png"><img border=1 src="img/wg-dp.png" width=600></a>
<!--------------------------------------------------------------------!>
<a name=annotsearch></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h3>SyMAP Queries</h3></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>
The query interface only works for multi-genome seq-to-seq synteny.
This interface has two basic
functions:
<br>
A. Locate homologous (or paralogous) regions based on annotation and location
<br>
B. Create putative gene families spanning multiple species, and apply family-based filters
<p>
Some sample queries which are possible through the interface:
<p>
<i>&bull; Find un-annotated regions on one genome which are homologous to regions on another genome
which have already been annotated. </i>
<p>
<i> &bull; Find putative gene families which are present in one lineage but absent in another.</i>
<p>
<i>&bull; Identify the likely orthologs on genome A of a given gene on genome B, by using synteny blocks
or exactly collinear pairs to filter out probable paralogs.  </i>
<p>
<i> &bull; Find orphan genes which are specific to one genome.</i>
<p>
To open the query interface, first select two or more sequence projects in the Project Manager (note, the
queries are not supported for FPC maps). Then press the "SyMAP Queries" button to open the interface:
<p>
<a href="img/query_oview.png"><img src="img/query_oview.png" width=550 border=1></a>
<p>
The Overview window (above) lists the projects which were selected for querying. To set up a query, open the 
"Query Setup" window by clicking on its title in the left panel. This opens the window below:
<p>
<a href="img/query_query.png"><img src="img/query_query.png" width=550 border=1></a>
<p>
Here you set up the query and then press "Do Search" to execute it. <p> 
The query occurs in three stages, corresponding to the section numbers
on the Query Page. Each stage has its own filters, as follows:

<h4>1. Filter hits</h4>

The first stage is to retrieve hits (anchors) from the database, based on 
filters such as chromosome and annotation string. Note that each anchor
connects two species and hence represents a <i>pair</i> of putatively-related
regions on the two species. Filters are as follows:

<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>Annotation Description</b>
<p>
Enter a search string that may be found in any value for any annotation keyword 
(e.g. the gene name or description). Hits will be returned which overlap a matching annotation
on either side. 
<p>
<b>Chromosome/location filters</b>
<p>
Each selected species has a row in which you can select to search either all
its sequences, or a particular sequence or basepair region.
<p>
<b>Only synteny hits</b>
<p>
Return only hits which are part of a syntenic block. This
helps to screen false positive hits, but can also conceal true hits that are not
part of a detectable syntenic block. (Note that SyMAP hits are already filtered
during loading using a reciprocal-top-2 filter.)
<p>
<b>Only hits in a collinear gene pair</b>
<p>
Return only hits which are part of a pair of aligning genes having no intervening
non-aligning genes. This relates to the RunSize column, which is one of the optional
columns of the Query Result table. RunSize shows the size of collinear chain which 
a hit is contained in, hence this checkbox is equivalent to RunSize at least 2. 
Note that a collinear chain is not the same as a SyMAP synteny block, because blocks
require at least 7 anchors, and are allowed to have intervening genes which do not align.
<p>
<b>Show orphan genes</b>
<p>
If checked, then the results will show all genes which <i>do not</i> overlap any of the hits
matching the search criteria (i.e, the hits which would be shown in the table if this
checkbox were not checked). Note that the genes must also meet the search criteria,
e.g. matching the annotation query string and the chromosome location requirements. 

</div>

<p>

<h4>2. Filter putative gene families (PgeneFs)</h4>

Using the hits that pass the Stage 1 filters, SyMAP constructs putative gene families (PgeneFs) 
spanning the selected species. This is done by grouping hits whose hit regions overlap on 
at least one genome. <i>Note, if you have more than 6 species selected, this stage can take
an hour or more.</i>
<p>
Each PgeneF is given a number, which is shown in the Query Results table (column name "PgeneF"). 
The size of the PgeneF is also shown (column "PgFSize"). 
<p>
Filters at this stage apply to the PgeneF as a whole:

<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>Include/Exclude</b>
<p>
These filters permit searching for gene families shared by one group of species but not
present in another. 
<p>
If a species is checked to include, then the PgeneF will only be retained if it includes
at least one hit which hits that species. 
<p>
If a species is checked to exclude, then the PgeneF will be discarded if any of its
hits are to that species. 
<p>
<b>No annotation to included species</b>
<p>
Find PgeneFs which are not yet annotated. A PgeneF will be discarded if it is annotated on any of
the species which are checked in the <b>Include</b> line.
<p>
<b>Complete linkage of included species</b>
<p>
Require the PgeneF to be fully linked, i.e. for each pair of species A and B in the group,
there must be a hit linking A to B.  
</div>

<p>

<h4>3. Filter displayed hits</h4>

The query returns hits between all pairs of the selected species; however, you may
only be interested in seeing those which hit certain species. To achieve this,
select those species in the <b>Include</b> row of Filter Stage 2, and also check
the box below, <b>Show only hits to the included species</b>. Only those hits will be shown,
although the PgeneF numbers will reflect groupings created using all hits. 

<p>
<h4>SyMAP Query Result Table</h4>
<p>
When the query is complete, the Query Results page opens showing the table of results:
<p>

<a href="img/query_results.png"><img src="img/query_results.png" width=550 border=1></a>

<p>
The table contains all the hits (anchors) resulting from the query. Each hit connects
two species and you can see the respective chromosomes and start/end locations of 
the hits, as well as gene annotations overlapped by the hits. 
<p>
Note that the table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty.
If the query specified orphan genes, then each row represents one gene and shows data only
for one species.
<p>
The entries in the table may be selected; to select more than one
entry, use "ctrl-click". You can sort the columns by clicking on the column name, 
and rearrange them by dragging the header boxes. 

<h4>Button functions</h4>

<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>Show Synteny</b>
<p>
Brings up a SyMAP 2-track view for each selected entry, so the hits
can be seen in their full chromosomal context. The hits are initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.  
<p>
<b>Save CSV</b>
<p>
Saves the selected hits using the selected set of columns to a CSV format suitable for import into Excel.
<p>
<b>Save for Reload</b>
<p>
Saves the selected hits in a CSV format which can be reloaded back into SyMAP 
later (click "Results" on the lefthand pane,
and use the "Load Saved Query" button). The saved table includes <i>all</i> possible columns, not just
those currently selected for display. 
<p>
<b>Save as Fasta</b>
<p>
Sequences from the selected hits are written to a Fasta file. Both sides of each hit
are written, using the start/end coordinates shown in the table.
<p>
<b>Align sequences</b>
<p>
Sequences for the selected hits are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32). If no selection is made then the whole table is used. <i>Note that
alignments of more than 10 sequences can take considerable time.</i>
</div>

<h4>Results Table Columns</h4>
You can add/remove columns using the "Select Columns" button at the bottom. This
opens up a column-selection section, shown here:
<p>
<a href="img/query_results_column.png">
	<img src="img/query_results_column.png" width=550 border=1></a>
<p>
 Following
are the meanings of the columns:
<p>
<table border='1' cellpadding='5' rules=all>
<tr>
	<td align="center" colspan=2>
		<i>General</i>
	</td>

</tr>
<tr>
	<td align="left" valign="top">
		<b>Row</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Row number within the table
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>PgeneF</b>
	</td>
	<td align="left"  valign="top" width="450" >
		PgeneF number. All hits in a given PgeneF have the same number. Note that 
		the number is generated during the search and won't necessarily be the same
		in a different search.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>PgFSize</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Size of the PgeneF which contains this hit.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>HitIdx</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Database index of the hit. This is used in the "Save for Reload" function.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>BlockNum</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Synteny block containing this hit (if any). The format is <chr1>.<chr2>.N, where
		runs from one to the number of blocks between those two chromosomes.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>BlockScore</b>
	</td>
	<td align="left"  valign="top" width="450" >
		The number of anchors which comprise the synteny block.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>RunSize</b>
	</td>
	<td align="left"  valign="top" width="450" >
		If the hit is contained in an uninterrupted collinear sequence of shared
		genes, RunSize shows the length of these sequence. If you have chosen
		the option "Only hits in a collinear gene pair", then RunSize will be two
		or greater for all returned hits.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>HitIdx</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Database index of the hit. This is used in the "Save for Reload" function.
	</td>
</tr>
<tr>
	<td align="center" colspan=2>
		<i>Location: one row for each species</i>
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>Chr</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Chromosome (or draft sequence number) where the hit is aligned
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>Start/End</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Start/end of the hit region (note that hits are clustered during anchor loading, so hit regions
		are not necessarily single MUMmer alignments).
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>#RGN</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Shows the number of distinct regions on this species which are included in the 
		PgeneF containing this hit. For example, if a gene from Species A expanded by 
		tandem duplication to 5 genes on Species B, then the PgeneF will have PgFSize=5,
		and #RGN will be one for Species A and 5 for species B.
	</td>
</tr>
<tr>
	<td align="center" colspan=2>
		<i>Annotation: one row for each species</i>
	</td>
</tr>
<tr>
	<td align="left" colspan=2>
	The key/value pairs from the annotation gff attributes sections are shown here 
	They can be different for each species, e.g. one may have "ID" while another uses "Name" for the 
	gene names. "All_anno" shows all of the attributes joined together.  
	</td>
</tr>
</table>

<h4>Statistics</h4>
This section shows overall statistics for the query results, for each species. 
The meaning of the terms is as follows: 
<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>hits</b>
<p>
Number of hits involving that species.
<p>
<b>distinct regions</b>
<p>
Number of distinct regions covered on that species (see the definition of the "#PGF" column above). 
<p>
<b>annotated</b>
<p>
Out of the number of distinct regions, how many have annotation.
<p>
<b>orphans</b>
<p>
Number of orphan genes returned for this species.
<p>
<b>chrs</b>
<p>
Number of chromosomes (or draft sequences) this species has, for reference.
</div>


<a name=caveat></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Self-synteny Display</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

The Chromosome Explorer does not work for a given chromosome self-alignment 
(e.g. it will not show Chr1 aligned to Chr1).
<p>However, from the Project Manager, select the self-synteny cell followed 
by Dot Plot. From the Dot Plot, any block or region
can be selected to show the results in the Chromosome Explorer 2D display.

<a name=data_download></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Data Download</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

There are two ways to download data for SyMAP synteny blocks, individual anchors, and annotations:
<ul>
<li> Through the <a href="#explorer">Explorer</a>. Select the species of interest,
 open the Explorer, and the download button is at the lower left.
 This exports a table of all the computed synteny block co-ordinates for all the
 selected species, including their self-alignments, if those were computed. 
<li> Through the <a href="#annotsearch">Query interface</a>. For example,
 choosing two species and executing a query with the default parameters
 will show a table of all the anchors found between the two species, plus 
their annotations and synteny block membership, if any. 
</ul>

<a name=image></a>
<table width="95%" cellspacing=0 cellpadding=0 class="lgbreak">
<tr><td align=left>
<h2>Print and Help</h2></td><td align=right valign=top class="toptxt"><a href="#top">Go to top</a></td></tr>
</table>

Most displays shown have a <img border=1 src="img/print.png" width=50> set of icons.
<p>The left icon is for printing the image. However, this does not work with all versions
of Java. If it does not work, use the system "Screen Capture" (all the images in this
document were created with screen capture, along with the images in the SyMAP publications).

<p>The right
icon brings up the this web page, typically to the correct section (obviously, there needs
to be an internet connection).


<p>&nbsp;
</table>
  <table width="100%" border="0" align="center">
    <tr>
      <td bgcolor="#DDDDDD" align="center">
       <font face="Verdana, Arial, Helvetica, sans-serif" size="2">Email
       Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></font></td>

    </tr>
  </table>
<br>

		</td>
	</tr>

</table>
</BODY>
</HTML>
