<!DOCTYPE html>
<html>
<head>
<title>SyMAP NCBI</title>
<style>
	a:link {text-decoration: none;color:#6666CC;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.ty    {border: 1px  solid black; border-spacing: 3px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; }
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="../img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					NCBI convert to SyMAP
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="../img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="../img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="../Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="../SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="../UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="../Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

This document corresponds to the ConvertNCBI release with SyMAP v5.0.2
and updated for SyMAP v5.1.2 (dated August-2022).
It has been tested on these <a href=index.html class="ext" target=_blank>genomes</a>.
<p>
NCBI supplies FASTA formatted files for genome sequence and GFF3 formatted files for the annotation,
where FASTA and GFF3 files are the input to SyMAP.
<i>However, using them directly can cause problems</i>.
The following provides a simple scheme to produce only the files necessary.

<p>
<b>Contents</b>

<table>
<tr>
<td style="vertical-align: text-top;">
	<table>
	<tr><td>1.</td><td> <a href="#dl">Download</a></td></tr>
	<tr><td>2.</td><td> <a href="#cv">Convert files</a></td></tr>
	<tr><td>3.</td><td> <a href="#load">Load files into SyMAP</a></td></tr>
	</table>
</td>
<td style="vertical-align: text-top;">
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;5.</td><td> <a href="#scaf">Scaffolds</a></td></tr>
	<tr><td>&nbsp;&nbsp;&nbsp;6.</td><td> <a href="#edit">Editing the script</a></td></tr>
	<tr><td>&nbsp;&nbsp;&nbsp;4. </td><td><a href="#what">What the ConvertNCBI script does</a></td></tr>
	</table>
</td>
</table>

<a id="dl"></a>
<h2>Download</h2>

<ol>
<li>Go to <a href="https://www.ncbi.nlm.nih.gov" class="ext" target="_blank">NCBI</a>.
<li>As shown in Fig 1:
<ol type="A">
<li>Select "Genome" from the pull-down at the top.
<li>Enter you genome name followed by "Search". You should see a page similar to Fig 2.
</ol>

<li>The following are two approaches to download the FASTA and GFF files:
<a id="A"></a>
<ol type="A">
<li>As shown in Fig 2:
<ol type="i">
<li>Use the <font color=darkblue><b>genome</b></font> link beside the "Download sequences in FASTA format for genome,..."
<li>Use the <font color=darkblue><b>GFF</b></font> link beside the "Download genome annotation in GFF...".
<li>Process as described in <a href="#cv1">Convert A</a>.
</ol>

<a id="B"></a>
<li>As shown in Fig 2 and Fig 3:
<ol type="i">
<li>Go to the <font color=darkblue><b>NCBI Datasets</b></font> page (link at bottom of Fig 2).
This will bring up a window as shown in the background of Fig 3.
<li>Select the assembly you want, then click on the Action dots, which brings up a menu; select
Download. That brings up a window as shown in Fig 3.
<li>Select the file source, then  genome FASTA and annotation GFF for download.
<li>Process as described in <a href="#cv2">Convert B</a>.
</ol>

</ol>
</ol>
<p>If NCBI only provides the Genbank
format, try using <a href="https://pypi.org/project/bioconvert/" class="ext" target="_blank">bioconvert</a> to convert to GFF3, followed
by the SyMAP converter. Alternatively, try <a href="https://bio.tools/genbank2gff3" class="ext" target="_blank">Galaxy</a>.


<p><a href="img/NCBI-search.png"><img src="img/NCBI-search.png" alt="" style="border: 1px solid black; width: 750px"></a>
<br><b>Fig 1. </b> Search the NCBI site.
<p>&nbsp;
<p><a href="img/NCBI-genome.png"><img src="img/NCBI-genome.png" alt="" style="border: 1px solid black; width: 500px"></a>
<br><b>Fig 2. </b> Download A: download the genome and GFF files.
<p>&nbsp;
<p><a href="img/NCBI-genome-DS.png"><img src="img/NCBI-genome-DS.png" alt="" style="border: 1px solid black; width: 500px"></a>
<br><b>Fig 3. </b> Download B: from NCBI Datasets, select the genome and GFF files.

<!------------------------------------------------------->
<h3>ConvertNCBI optional flags</h3>
<table class="ty" style="width:90%;">
<tr><td><b>Flag</b><td><b>Description</b><td><b>Details</b><td><b>Default</b>
<tr><td>-s<td>Include Scaffolds<td>Include scaffolds in the output. See section <a href="#scaf">Scaffolds</a><td>No scaffolds
<tr><td>-m<td>Hard-mask<td>NCBI genome sequences are soft-masked (generally for download A only), which is changed to hard masked<td>Leave as soft-mask
<tr><td>-l<td>Use linkage groups<td>Search 'linkage' instead of 'chromosome'<td>Use chromosomes
<tr><td>-v<td>Verbose<td>Print out header lines of skipped sequences<td>No print
</table>

<p><i><u>Beware</u></i>: There are variations in the text associated with the FASTA ">" header lines. The rules
used by this script are as follows:
<ol>
<li>If the ">" line contains "chromosome N", where N={number, X, Y or roman numeral}, than it is identified as such.
<li>If the ">" line contains "scaffold" or "contig", than it is identified as a scaffold (if "chromosome" is
also on the line, it is still treated as a scaffold).
<li>All other ">" entries are considered "other".
</ol>
It is a real good idea to execute the following on your fasta file:
<pre>
   zgrep ">" [prefix]._genomic.fna.gz
</pre>
This shows you the lines and confirms that the above rules will work for your species. If not,
then you will need to edit the script to work (or ask me to at symap@agcol.arizona.edu).

<!---------------------------------------------------------------->
<a id="cv"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Convert files</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The following conversions were tested on the <i>Oryza Sativa</i> NCBI files on 15-Aug-22.
The download A genome sequence was soft-masked, but the download B was hard-masked.

<a id="cv1"></a>
<h3>Convert files from FASTA and GFF (download A)</h3>
<ol>
<li>Go to the <tt>symap_5/data/seq</tt> directory.
<li>Make a subdirectory for your species and move the FASTA and GFF files
into the directory.
<li>From the <tt>seq</tt> directory, type the following at the command line to copy the <tt>ConvertNCBI</tt> script:
<pre>cp ../../scripts/ConvertNCBI*.class .
chmod 755 *.class
</pre>

<li>Execute
<pre>java ConvertNCBI &lt;species&gt;</pre>
</ol>

<h3>Example</h3>
From the <tt>symap_5</tt> directory:
<pre>
&gt; cd data/seq
&gt; mkdir riceA
&gt; cd riceA
&gt; mv ~/Download/GCF_001433935.1_IRGSP-1.0_genomic.fna.gz .
&gt; mv ~/Download/GCF_001433935.1_IRGSP-1.0_genomic.gff.gz .
&gt; cd ..
&gt; cp ../../scripts/ConvertNCBI*.class .
&gt; chmod 755 *.class
&gt; java ConvertNCBI riceA
</pre>
This results in the following contents:
<pre>
data/seq
	ConvertNCBI.class
	ConvertNCBI$Gene.class
data/seq/riceA/
      GCF_001433935.1_IRGSP-1.0_genomic.fna.gz
      GCF_001433935.1_IRGSP-1.0_genomic.gff.gz
      annotation/
         anno.gff
         gap.gff
      sequence/
         genomic.fna
</pre>
The output gives useful details of the annotation (e.g. see riceA <a href=riceA.html class="ext" target="_blank">details</a>);
if the details do not appear right, you may need to <a href="#edit">edit</a> the script for your genomes.
<!--------------------------------------------------------------->
<a id="cv2"></a>
<h3>Convert files from ncbi_dataset.zip (download B)</h3>
This approach does not seem to soft-mask the sequence. Also, you need to check the resulting directory
to make sure there is just one sub-directory.
<p>Follow all steps from Download A, except for step 2, do the following:
<ol start=2>
<li>Make a subdirectory for your species and move the <tt>ncbi_dataset.zip</tt> file to the species directory and unzip it.
</ol>

<h3>Example</h3>
From the <tt>symap_5</tt> directory:
<pre>
&gt; cd data/seq
&gt; mkdir riceB
&gt; cd riceB
&gt; mv ~/Download/ncbi_dataset.zip .
unzip ncbi_dataset.zip
Archive:  ncbi_dataset.zip
  inflating: README.md
  inflating: ncbi_dataset/data/assembly_data_report.jsonl
  inflating: ncbi_dataset/data/GCA_001433935.1/chr1.fna
  inflating: ncbi_dataset/data/GCA_001433935.1/chr2.fna
  ...
  inflating: ncbi_dataset/data/GCA_001433935.1/genomic.gff
  inflating: ncbi_dataset/data/GCA_001433935.1/sequence_report.jsonl
  inflating: ncbi_dataset/data/dataset_catalog.json

&gt; cd ..
&gt; cp ../../scripts/ConvertNCBI*.class .
&gt; chmod 755 *.class
&gt; java ConvertNCBI riceB
</pre>
This results in the following contents (some NCBI files not listed):
<pre>
data/seq
	ConvertNCBI.class
	ConvertNCBI$Gene.class
data/seq/riceB/
	ncbi_dataset.zip
	ncbi_dataset/data/GCA_001433935.1/
		chr1.fna
		...
	annotation/
		anno.gff
		gap.gff
	sequence/
		genomic.fna
</pre>
The output gives useful details of the annotation (e.g. see riceB <a href=riceB.html class="ext" target="_blank">details</a>);
if the details do not appear right, you may need to <a href="#edit">edit</a> the script for your genomes.

<!----------------------------------------------------------------->
<a id="load"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Load files into SyMAP</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The above scenario puts the files in the default SyMAP directories.
When you start up SyMAP, you will see your projects listed on the left of the panel
(e.g <a href="../SystemGuide.html#demoseq" class="ext" target="_blank">as shown for demos</a>). Check the projects
you want to load, which will cause them to be shown on the right of the symap window and continue
as described in the <a href="../SystemGuide.html" class="ext" target="_blank">System Guide</a>.

<!----------------------------------------------------------------->
<a id="scaf"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Scaffolds</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

By default, the <tt>ConvertNCBI</tt> script creates the <tt>genomic.fna</tt> file with only the chromosomes.
However, you can have it also include the scaffolds by using the "-s" flag, e.g. using Oryza Sativa
<pre>
java ConvertNCBI rice -s
</pre>
This will include all chromosomes (prefix 'C') and scaffolds (prefix 's') in the <tt>genomic.fna</tt> file.
Beware, there can be many tiny scaffolds. If they all aligned in SyMAP, it causes the display to be very cluttered.
Hence, it is best to just align the largest ones (e.g. the longest 30); merge them if possible, then try
the smaller ones. You should set the following SyMAP project parameters:
<ol>
<li><tt>grp_prefix</tt> needs to be blank as there is no common prefix now.
<li><tt>min_size</tt> should be set to only load the largest scaffolds. To determine the value to use, run the
<tt>lenFasta.pl</tt> script,
e.g. from the <tt>seq</tt> directory and using rice as an example:
<pre>
cp ../../scripts/lenFasta.pl .
perl lenFasta.pl riceA/sequence/genomic.fna
</pre>
</ol>
As of 12-Aug-22 (GCF_001433935.1_IRGSP-1.0_genomic.fna),
rice has 58 sequences where 12 are chromosomes, 43 are scaffolds
and 3 are other. The script outputs all their
sorted lengths followed by the following table:
<pre>
Read riceA/sequence/genomic.fna and print sorted lengths
Read 55 sequences
   N    Length  Seqid
   1  43270923  >C01 NC_029256.1
   2  36413819  >C03 NC_029258.1
   3  35937250  >C02 NC_029257.1
...
Values for min_len (assuming no duplicate lengths):
#Seqs  min_len
   10 27531856
   20    19457
   30    11447
   40    10311
   50     7140
</pre>

To align the top 30 sequences (12 chromosomes, 18 of the largest scaffolds),
this says to set <tt>min_size</tt> to 11447.

<a id="edit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Editing the script</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

You may want the make changes such as what attributes are included. Therefore, the <tt>ConvertNCBI.java</tt>
code is supplied in the <tt>scripts</tt> directory. It is very simply written, it does not use external libraries
and only uses common programming techniques.
<p>Once you make your changes, execute:
<pre>
   javac ConvertNCBI.java
</pre>
You will need to have JDK installed to use the 'javac' command.
<!------------------------------------------------------------->
<a id="what"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>What the ConvertNCBI script does</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The following occurs in the <tt>data/seq/&lt;project directory name&gt;</tt> where "project directory name"
is the argument supplied to <tt>ConvertNCBI</tt>. The following is for Download A,
but works similarly for Download B.
It assumes no parameters are set (e.g. -s for scaffolds).

<ul>
<li>Reads the file ending in '.fna.gz' (or '.fna') and writes a new file called <tt>sequence/genomic.fna</tt> with the following changes:
<ol>
<li>Sequences must have the word "chromosome" in their ">" header line in order to be copied (unless -l or -s flags).
<li>The header line is replaced with ">ChrN", where N comes from the header line "chromosome N", e.g.
"chromosome 1", etc.

For example,
<pre>
>NC_029256.1 Oryza sativa Japonica Group cultivar Nipponbare chromosome 1, IRGSP-1.0
</pre>
is replaced with:
<pre>
>Chr1  NC_029256.1
</pre>
<li>Gaps of &gt;30,000 are written to the <tt>annotation/gap.gff</tt> file (30,000 is hard-coded in ConvertNCBI script).
</ol>
<br>&nbsp;
<li>Reads the file ending in 'gff.gz' (or .gff) and writes the file <tt>annotation/anno.gff</tt>. The
<a href="http://www.sequenceontology.org/gff3.shtml" class="ext" target="_blank">gff3 format</a>
has 9 columns, where the first is the 'seqid', the third is the 'type' (e.g. feature 'gene'), the
last column is a semicolon-delimited keyword=value attribute list. The input file is processed as follows:

<ol>
<li>The <i>type=gene</i> with attribute <i>gene-biotype=protein-coding</i> are processed.
The gene line is written to the <tt>anno.gff</tt> file with the following changes:

<ol type="i">
<li>The first column 'seqid' is replace with the 'ChrN' value assigned when reading the '.fna' file.
<li>A subset of the attributes are written, where the product keyword value is modified as follows:
<ol type='a'>
<li>If there are multiple mRNA lines for a gene where the values are different, they are concatenated together.
<li>If there are multiple mRNA lines for a gene where
the only difference is the variant, then only the variant difference is show, e.g.
<pre>
product=monocopper oxidase-like protein SKU5%2C transcript variant X2, X1, X3
</pre>
</ol>
</ol>

<li>The first <i>type=mRNA</i> line for a gene is written to the <tt>anno.gff</tt> file followed
by its <i>type=exon</i>, where they each are written with the new seqid and a subset of the
attributes.
</ol>
</ul>

<p><a href="#top">Go to top</a>
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
