<!DOCTYPE html>
<html>
<head>
<title>SyMAP NCBI</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.ty    {border: 1px  solid black; border-spacing: 3px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; }
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="../img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					NCBI convert to SyMAP
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="../img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="../img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="../Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="../SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="../UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="../Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->


NCBI supplies FASTA formatted files for genome sequence and GFF3 formatted files for the annotation,
where FASTA and GFF3 files are the input to SyMAP.
<i>However, using them directly can cause problems</i>.
The following provides a simple scheme to produce only the files necessary.
This has been tested on these <a href=index.html class="ext" target=_blank>genomes</a>.

<p>

<table>
<tr><td colspan=2><b>Contents</b>
<tr>
<td style="vertical-align: text-top;">
	<table>
	<tr><td>&nbsp;&nbsp;</td><td> <a href="#dl">Download</a></td></tr>
	<tr><td>&nbsp;&nbsp;</td><td> <a href="#cv">Convert files</a></td></tr>
	<tr><td>&nbsp;&nbsp;</td><td><a href="#opts">&nbsp;&nbsp;&nbsp;Options</a></td></tr>
	<tr><td>&nbsp;&nbsp;</td><td><a href="#scaf">&nbsp;&nbsp;&nbsp;Scaffolds</a></td></tr>

	</table>
</td>
<td style="vertical-align: text-top;">
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td> <a href="#load">Load files into SyMAP</a></td></tr>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td> <a href="#edit">Editing the script</a></td></tr>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td><a href="#what">What the ConvertNCBI script does</a></td></tr>
	<tr><td>&nbsp;<td>&nbsp;
	</table>
</td>
</table>

These instructions will use <i>Brassica oleracea</i> (wild cabbage) as an example.

<a id="dl"></a>
<h2>Download</h2>
Go to <a href="https://www.ncbi.nlm.nih.gov" class="ext" target="_blank">NCBI</a>.
<ol>
<li>As shown in Fig 1:
<ol type="A">
<li>Select <ttl>Genome</ttl> from the pull-down at the top.
<li>Enter you genome name followed by <ttl>Search</ttl>.
</ol>
<li>As shown in Fig 2:
<ol type="A">
<li>From the list of files, select the <ttl>Assembly</ttl> you want, e.g. <ttx>BOL</ttx>
 (Genbank GCA_000695525.1 and RefSeq GCF_000695525.1).
<li>Select <ttl>Download Package</ttl>.  As shown, a window will popup with options.
<li>Select the <ttx>RefSeq</ttx><sup>1</sup> format with
the <ttx>FASTA</ttx> and <ttx>GFF</ttx> files.
<li>Select <ttl>Download</ttl>. By default, a file called <ttx>ncbi_dataset.zip</ttx> will be downloaded.
</ol>
</ol>
<sup>1</sup><ttb>ConvertNCBI</ttb> does NOT work with Genbank files, only RefSeq.

<p><a href="img/NCBI-search.png"><img src="img/NCBI-search.png" alt="" style="border: 1px solid black; width: 750px"></a>
<br><b>Fig 1. </b> Search the NCBI site.
<p>&nbsp;

<p><a href="img/NCBI-download.png"><img src="img/NCBI-download.png" alt="" style="border: 1px solid black; width: 500px"></a>
<br><b>Fig 2. </b> Select the RefSeq genome FASTA and GFF files.

<!---------------------------------------------------------------->
<a id="cv"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Convert files</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ol>
<li>Go to the <tt>symap_5/data/seq</tt> directory.
<li>Make a subdirectory for your species, move <ttx>ncbi_dataset.zip</ttx> to it, and unzip it, e.g.
<pre>
<ttl>symap_5/data/seq&gt;</ttl> mkdir cabb
<ttl>symap_5/data/seq&gt;</ttl> cd cabb
<ttl>symap_5/data/seq/cabb&gt;</ttl> mv ~/Download/ncbi_dataset.zip .
<ttl>symap_5/data/seq/cabb&gt;</ttl> unzip ncbi_dataset.zip
Archive:  ncbi_dataset.zip
  inflating: README.md
  inflating: ncbi_dataset/data/data_summary.tsv
  inflating: ncbi_dataset/data/assembly_data_report.jsonl
  inflating: ncbi_dataset/data/GCF_000695525.1/GCF_000695525.1_BOL_genomic.fna
  inflating: ncbi_dataset/data/GCF_000695525.1/genomic.gff
  inflating: ncbi_dataset/data/dataset_catalog.json
</pre>

<li>From the <tt>seq</tt> directory, type the following at the command line to copy the <tt>ConvertNCBI</tt> script:
<pre>
<ttl>symap_5/data/seq&gt;</ttl> cp ../../scripts/ConvertNCBI*.class .
<ttl>symap_5/data/seq&gt;</ttl> chmod 755 *.class
</pre>

<li>Execute
<pre><ttl>symap_5/data/seq&gt;</ttl> java ConvertNCBI &lt;species&gt;</pre>
e.g.
<pre><ttl>symap_5/data/seq&gt;</ttl> java ConvertNCBI cabb</pre>
</ol>

The script will output statistics to the terminal,
e.g. <a href="./cabbNCBI.html" class="_ext" target="_blank">cabbage output</a>.
The results in the cabb directory is as follows:
<pre>
<ttl>symap_5/data/seq/cabb&gt;</ttl> <i>ls -hlG</i>
total 264008
-rw-------@ 1 cari  staff   1.6K Jul 23  2024 README.md
drwxr-xr-x  4 cari  staff   128B Jul 23 09:54 annotation/
drwxr-xr-x@ 3 cari  staff    96B Jul 23 08:19 ncbi_dataset/
-rw-r--r--@ 1 cari  staff   126M Jul 23 07:11 ncbi_dataset.zip
-rw-r--r--  1 cari  staff   373B Jul 23 09:56 params
drwxr-xr-x  3 cari  staff    96B Jul 23 09:54 sequence/
</pre>
You may remove everything but <ttx>annotation/</ttx>, <ttx>sequence/</ttx> and <ttx>params</ttx>.
However, you may want to keep the original <ttx>data/seq/cabb/ncbi_dataset</ttx>.
<p>In the SyMAP <a href="../SystemHelp.html#start" class="ext" target="_blank">project parameters</a> window,
enter "Chr" for the <ttl>Group prefix</ttl>.

<!------------------------------------------------------->
<a id="opts"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Options</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="ty" style="width:90%;">
<tr><td><b>Flag</b><td><b>Description</b><td><b>Details</b><td><b>Default</b>
<tr><td>-s<td>Include Scaffolds<td>Any sequence not labeled a "chromosome" will be written as a scaffold. See section <a href="#scaf">Scaffolds</a><td>No scaffolds
<tr><td>-m<td>Hard-mask<td>NCBI genome sequences are soft-masked (generally for download A only), which is changed to hard masked<td>Leave as soft-mask
<tr><td>-p<td>1st protein name<td>1st protein name for the gene will be at the end of the <tt>Attributes</tt> using the <tt>protein=</tt> keyword.
This can be searched using the <ttb>Queries</ttb><td>Do not include
<tr><td>-pa<td>All protein names<td>Same as above, but all protein names for the gene will be included in the <tt>protein=</tt> keyword. <td>Do not include
<tr><td>-l<td>Use linkage groups<td>Search 'linkage' instead of 'chromosome'<td>Use chromosomes
<tr><td>-v<td>Verbose<td>Print out header lines of skipped sequences<td>No print
</table>

<p><i><u>Beware</u></i>: There are variations in the text associated with the FASTA ">" header lines. The rules
used by this script are as follows:
<ol>
<li>If the ">" line contains "chromosome N", where N={number, X, Y or roman numeral}, than it is identified as such.
<li>If the ">" line contains "scaffold" or "contig", than it is identified as a scaffold (if "chromosome" is
also on the line, it is still treated as a scaffold).
<li>All other ">" entries are considered "other".
</ol>
It is a real good idea to execute the following on your fasta file:
<pre>
   zgrep ">" [prefix]._genomic.fna.gz
</pre>
This shows you the lines and confirms that the above rules will work for your species. If not,
then you will need to edit the script to work (or ask me to at symap@agcol.arizona.edu).

<!----------------------------------------------------------------->
<a id="scaf"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Scaffolds</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

By default, the <tt>ConvertNCBI</tt> script creates the <tt>genomic.fna</tt> file with only the chromosomes.
However, you can have it also include the scaffolds by using the "-s" flag, e.g. using cabbage
<pre>
<ttl>symap_5/data/seq&gt;</ttl> java ConvertNCBI cabb -s
</pre>
The results for cabbage are shown <a href="./cabbNCBIscaf.html" class="_ext" target="_blank">here</a>.
This will include all chromosomes (prefix 'C') and scaffolds (prefix 's') in the <tt>genomic.fna</tt> file.
Beware, there can be many tiny scaffolds. If they all aligned in SyMAP, it causes the display to be very cluttered.
Hence, it is best to just align the largest ones (e.g. the longest 30); merge them if possible, then try
the smaller ones. You should set the following SyMAP <a href="../SystemHelp.html#start" class="ext" target="_blank">project parameters</a>:
<ol>
<li><tt>grp_prefix</tt> needs to be blank as there is no common prefix now.
<li><tt>min_size</tt> should be set to only load the largest scaffolds. To determine the value to use, run the
<tt>lenFasta.pl</tt> script,
e.g. from the <tt>seq</tt> directory and using rice as an example:
<pre>
<ttl>symap_5/data/seq&gt;</ttl> cp ../../scripts/lenFasta.pl .
<ttl>symap_5/data/seq&gt;</ttl> perl lenFasta.pl cabb/sequence/genomic.fna
</pre>
</ol>
As of 23-July-2024 (GCF_000695525),
cabbage has 9 chromosome sequences and 32,877 scaffolds
The script outputs all their
sorted lengths followed by a summary table of lengths:
<pre>
Read cabb/sequence/genomic.fna and print sorted lengths
Read 32886 sequences
   N    Length  Seqid
   1  64984695  >C03 NC_027750.1
   2  54679868  >C09 NC_027756.1
   3  53719093  >C04 NC_027751.1
   4  52886895  >C02 NC_027749.1
   5  48366697  >C07 NC_027754.1
   6  46902585  >C05 NC_027752.1
   7  43764888  >C01 NC_027748.1
   8  41758685  >C08 NC_027755.1
   9  39822476  >C06 NC_027753.1
  10    550871  >s01  NW_013617415.1
...(list rest of scaffolds with lengths)

Values for min_len (assuming no duplicate lengths):
#Seqs  min_len
   10   550871
   20   213381
   30   154937
   40   131235
   50   101387
   60    86265
   70    70649
   80    65803
   90    61068
  100    58362
</pre>

To align the top 30 sequences (9 chromosomes, 21 of the largest scaffolds),
this says to set <tt>min_size</tt> to 154937.

<!----------------------------------------------------------------->
<a id="load"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Load files into SyMAP</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The above scenario puts the files in the default SyMAP directories.
When you start up SyMAP, you will see your projects listed on the left of the panel
(e.g <a href="../SystemGuide.html#demoseq" class="ext" target="_blank">as shown for demos</a>).
<ol>
<li>Check the projects
you want to load, which will cause them to be shown on the right of the symap window.
<li>For the project you want to load, open the <a href="../SystemHelp.html#start" class="ext" target="_blank">project parameters</a> window
to enter the appropriate values.
<li>The select <ttl>Load Project</ttl>.
</ol>

<!----------------------------------------------------------------->
<a id="edit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Editing the script</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

You may want the make changes such as what attributes are included. Therefore, the <tt>ConvertNCBI.java</tt>
code is supplied in the <tt>scripts</tt> directory. It is very simply written, it does not use external libraries
and only uses common programming techniques.
<p>Once you make your changes, execute:
<pre>
   javac ConvertNCBI.java
</pre>
You will need to have JDK installed to use the 'javac' command.
<!------------------------------------------------------------->
<a id="what"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>What the ConvertNCBI script does</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The following occurs in the <tt>data/seq/&lt;project directory name&gt;</tt> where "project directory name"
is the argument supplied to <tt>ConvertNCBI</tt>.
The following assumes no options are set.

<ul>
<li>Reads the file ending in '.fna.gz' (or '.fna') and writes a new file called <tt>sequence/genomic.fna</tt> with the following changes:
<ol>
<li>Sequences must have the word "chromosome" in their ">" header line in order to be copied (unless -l or -s flags).
<li>The header line is replaced with ">ChrN", where N comes from the header line "chromosome N", e.g.
"chromosome 1", etc.

For example,
<pre>
>NC_029256.1 Oryza sativa Japonica Group cultivar Nipponbare chromosome 1, IRGSP-1.0
</pre>
is replaced with:
<pre>
>Chr1  NC_029256.1
</pre>
<li>Gaps of &gt;30,000 are written to the <tt>annotation/gap.gff</tt> file (30,000 is hard-coded in ConvertNCBI script).
</ol>
<br>&nbsp;
<li>Reads the file ending in 'gff.gz' (or .gff) and writes the file <tt>annotation/anno.gff</tt>. The
<a href="http://www.sequenceontology.org/gff3.shtml" class="ext" target="_blank">gff3 format</a>
has 9 columns, where the first is the 'seqid', the third is the 'type' (e.g. feature 'gene'), the
last column is a semicolon-delimited keyword=value attribute list. The input file is processed as follows:

<ol>
<li>The <i>type=gene</i> with attribute <i>gene-biotype=protein-coding</i> are processed.
The gene line is written to the <tt>anno.gff</tt> file with the following changes:

<ol type="i">
<li>The first column 'seqid' is replace with the 'ChrN' value assigned when reading the '.fna' file.
<li>A subset of the attributes are written, where the product keyword value is modified as follows:
<ol type='a'>
<li>If there are multiple mRNA lines for a gene where the values are different, they are concatenated together.
<li>If there are multiple mRNA lines for a gene where
the only difference is the variant, then only the variant difference is show, e.g.
<pre>
product=monocopper oxidase-like protein SKU5%2C transcript variant X2, X1, X3
</pre>
</ol>
</ol>

<li>The first <i>type=mRNA</i> line for a gene is written to the <tt>anno.gff</tt> file followed
by its <i>type=exon</i>, where they each are written with the new seqid and a subset of the
attributes.
<br>&nbsp;
<li>If the "-p" flag is set, then the <tt>ID=</tt> from the first <i>type=CDS</i> for the gene is extracted, and the "cds-" removed. From the files
I have tested, this name corresponds with the ">" name in the NCBI protein file. Likewise for the "-pa" flags, but all unique ID's from the
gene's <i>type=CDS</i> records are made into a comma-delimited list; note, this can cause a long list.
</ol>
</ul>

<p><a href="#top">Go to top</a>
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
