<!DOCTYPE html>
<html>
<head>
<title>SyMAP NCBI</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: Arial, monospace; font-size: 15px; font-weight: 500; color: DarkSlateGrey; }
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.ty    {border: 1px  solid black; border-spacing: 3px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; }
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="../img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					NCBI convert to SyMAP
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="../img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="../img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="../Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./index.html" class="white">Input</a> <mark class="white">|</mark>
						<a href="../SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="../UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="../Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->


NCBI supplies FASTA formatted files for genome sequence and GFF3 formatted files for the annotation,
where FASTA and GFF3 files are the input to SyMAP. Though they can be loaded directly into SyMAP (since v5.5.7),
it is better to convert them.


<p><b>Contents</b>
<table>
<tr>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td>
<td style="vertical-align: text-top;">
	<table>
	<tr><td><a href="#dl">Download</a>
	<tr><td><a href="#cv">Convert files</a>
	<tr><td><a href="#opts">&nbsp;&nbsp;&nbsp;Options</a>
	<tr><td><a href="#scaf">&nbsp;&nbsp;&nbsp;Scaffolds</a>
	</table>
</td>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td>
<td style="vertical-align: text-top;">
	<table>
	<tr><td> <a href="#load">Load files into SyMAP</a></td></tr>
	<tr><td> <a href="#edit">Editing the script</a></td></tr>
	<tr><td><a href="#what">What the ConvertNCBI script does</a></td></tr>
	</table>
</td>
</table>
<p><b>Reasons to convert files</b>
<ol>
<li>The NCBI name for a chromosome or scaffold are long and crowd the interface. <ttx>Convert</ttx> uses
"ChrN" instead for chromosomes only (where the 'Chr' prefix can be removed on load);
when scaffolds are included it uses the prefixes 'C' for chromosome and 's' for scaffold.
<li>It has an option to produce a hard masked sequence.
<li>Only the 'protein-coding' genes are processed.
<li>Gene attributes:
<ul>
<li>Only the gene attributes of 'ID' and 'Name' are retained from the gene.
<li>It has an added 'product' attribute, which combines the 'product' from all mRNAs for the gene.
<li>It has an optional 'protein' attribute, which uses the 1st or all protein name(s) from the CDS(s).
</ul>
<li>If it has problems converting the file(s), then <ttp>symap</ttp> will have problems loading them;
the script can be edited for your particular files. (Note: NCBI formats are not totally consistent,
so this may not take everything into account; check your files with the <ttl>xToSymap</ttl>
<ttx>Summarize files</ttx> function).
</ol>
The following instructions will use <i>Brassica oleracea</i> (wild cabbage) as an example.

<a id="dl"></a>
<h2>Download</h2>
Go to <a href="https://www.ncbi.nlm.nih.gov" class="ext" target="_blank">NCBI</a>.
<ol>
<li>As shown in Fig 1:
<ol type="A">
<li>Select <ttl>Genome</ttl> from the pull-down at the top.
<li>Enter you genome name followed by <ttl>Search</ttl>.
</ol>
or, go to <a href="https://www.ncbi.nlm.nih.gov/refseq/annotation_euk/all/"  class="ext" target="_blank">All RefSeq</a> and select your genome.

<li>As shown in Fig 2:
<ol type="A">
<li>From the list of files, select the <ttl>Assembly</ttl> you want.
<li>Select <ttl>Download Package</ttl>.  As shown, a window will popup with options.
<li>Select the <ttx>RefSeq</ttx><sup>1</sup> format with
the <ttx>FASTA</ttx> and <ttx>GFF</ttx> files.
<li>Select <ttl>Download</ttl>. By default, a file called <ttx>ncbi_dataset.zip</ttx> will be downloaded.
</ol>
</ol>
<sup>1</sup><ttb>ConvertNCBI</ttb> does NOT work with Genbank files, only RefSeq.

<p><a href="img/NCBI-search.png"><img src="img/NCBI-search.png" alt="" style="border: 1px solid black; width: 750px"></a>
<br><b>Fig 1. </b> Search the NCBI site.
<p>&nbsp;

<p><a href="img/NCBI-download.png"><img src="img/NCBI-download.png" alt="" style="border: 1px solid black; width: 500px"></a>
<br><b>Fig 2. </b> Select the RefSeq genome FASTA and GFF files.

<!---------------------------------------------------------------->
<a id="cv"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Convert files</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ol>
<li>Go to the <tt>symap_5/data/seq</tt> directory.
<li>Make a subdirectory for your species, move <ttx>ncbi_dataset.zip</ttx> to it, and unzip it, e.g.
<pre>
<ttl>symap_5/data/seq&gt;</ttl> mkdir cabb
<ttl>symap_5/data/seq&gt;</ttl> cd cabb
<ttl>symap_5/data/seq/cabb&gt;</ttl> mv ~/Download/ncbi_dataset.zip .
<ttl>symap_5/data/seq/cabb&gt;</ttl> unzip ncbi_dataset.zip
Archive:  ncbi_dataset.zip
  inflating: README.md
  inflating: ncbi_dataset/data/data_summary.tsv
  inflating: ncbi_dataset/data/assembly_data_report.jsonl
  inflating: ncbi_dataset/data/GCF_000695525.1/GCF_000695525.1_BOL_genomic.fna
  inflating: ncbi_dataset/data/GCF_000695525.1/genomic.gff
  inflating: ncbi_dataset/data/dataset_catalog.json
</pre>
<li>Start the <a href="./index.html#if" class="ext" target="_blank">xToSymap</a> program,
select the appropriate options (described below),
then select <ttl>Convert</ttl>. The FASTA file must end in ".fna" and the annotation file
must end in ".gff" (the NCBI defaults).
</ol>

The program will output statistics to the terminal and a log file,
e.g. <a href="./logs/cnvCabbN.html" class="ext" target="_blank">cabbage output</a>.
The results in the cabb directory is as follows:
<pre>
<ttl>symap_5/data/seq/cabb&gt;</ttl> <i>ls -hlG</i>
total 264008
-rw-------@ 1 cari  staff   1.6K Jul 23  2024 README.md
drwxr-xr-x  4 cari  staff   128B Jul 23 09:54 annotation/
drwxr-xr-x@ 3 cari  staff    96B Jul 23 08:19 ncbi_dataset/
-rw-r--r--@ 1 cari  staff   126M Jul 23 07:11 ncbi_dataset.zip
-rw-r--r--  1 cari  staff   373B Jul 23 09:56 params
drwxr-xr-x  3 cari  staff    96B Jul 23 09:54 sequence/
</pre>
You may remove everything but <ttx>annotation/</ttx>, <ttx>sequence/</ttx> and <ttx>params</ttx>.
However, you may want to keep the original <ttx>data/seq/cabb/ncbi_dataset</ttx>.


<!------------------------------------------------------->
<a id="opts"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Options</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="ty" style="width:90%;">
<tr><td><b>Option</b><td><b>Description</b><td><b>Default</b>

<tr><td><ttl>Hard Mask</ttl>
    <td>NCBI genome sequences are soft-masked (generally for download A only), which is changed to hard masked
    <td>Leave as soft-mask
<tr><td><ttl>1st protein</ttl>
    <td>A new attribute called <tt>protein=</tt> will be added which has the value of the first CDS.
        This can be searched using the <ttb>Queries</ttb>
    <td>Do not include
<tr><td><ttl>All proteins</ttl>
	<td>Same as above, but all protein names for the gene will be included in the <tt>protein=</tt> keyword.
	<td>Do not include
<tr><td><ttl>Include scaffold</ttl><sup>1</sup>
    <td>Scaffolds will be included in FASTA and GFF. See section <a href="#scaf">Scaffolds</a>
    <td>No scaffolds
<tr><td><ttl>Include Mt/Pt</ttl>
    <td>Mt/Pt chromosomes will be included in FASTA and GFF. Only the first occurence will be included.
    <td>No Mt/Pt
<tr><td><ttl>Only prefix</ttl><sup>2</sup><td>Only sequences with the specified prefix will be processed.
	<td>None
<tr><td><ttl>Verbose</ttl>
     <td>Print out header lines of skipped sequences
     <td>No print
</table>
<sup>1</sup>You may use <ttl>Include scaffold</ttl>, and then limit the input on the <ttp>symap</ttp>
<ttl>Load</ttl> by setting
<br>&nbsp;&nbsp;<ttl>Minimal length</ttl> in the project's
<a href="../SystemHelp.html#projParams" class="ext" target="_blank">Parameters</a>.
<br><sup>2</sup>For situations needing <ttl>Only prefix</ttl>,
see <a href="./index.html#except" class="ext" target="_blank">exceptions</a>.


<p><i><u>Rules</u></i>: There are variations in the text associated with the FASTA ">" header lines. The rules
used by this script are as follows:
<ol>
<li>If <ttl>Only prefix</ttl> is not blank, sequences are filtered out if the <ttx>seqid</ttx> does not
start with the prefix. Then all the following apply to the non-filtered sequences:
<br>&nbsp;
<li>Chromosomes: The name starts with 'NC_'.
	 <ul>
	 <li>The exception is that Mt/Pt chromosomes will not be output unless <ttl>Include Mt/Pt</ttl> is selected.
	  <br>Mt/Pt: header line contains the word 'mitochondrion', 'mitochondrial', 'plastid' or 'chloroplast'.
	  <li>Chromosomes are always output unless <ttl>Only prefix</ttl> is set, and the prefix does not match.
	 <li>Output <ttx>Seqid</ttx>: If the ">" line contains "chromosome N", where N={number, X, Y or roman numeral}, than this number is used prefixed
	 by 'Chr' or 'C' (if scaffolds are included).
	 <br>Otherwise, the word following 'chromosome' is used (e.g. C1).

	 </ul>
<br>
<li>Scaffolds: The name starts with 'NW_' or 'NT_'.
	<ul>
	<li>They will only be output if <ttl>Include scaffold</ttl> is selected.
	<li>Output <ttx>Seqid</ttx>: 'Scaf' followed by a consecutive number.
	</ul>
<br>
<li>Unknown: All other ">" entries are considered "unknown".
	<ul>
	<li>They will only be output if <ttl>Only prefix</ttl> matches.
	<li>Output <ttx>Seqid</ttx>: 'Unk' followed by a consecutive number.
	</ul>
</ol>
See <a href="./index.html#sf" class="ext" target="_target">Summarize</a> to help determine how to set the options for your input.


<!----------------------------------------------------------------->
<a id="scaf"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Scaffolds</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

By default, the <ttl>Convert</ttl> option creates the <tt>genomic.fna</tt> file with only the chromosomes.
However, you can have it also include scaffolds by selecting <ttl>Include Scaffolds</ttl>.

The results for cabbage are shown <a href="./logs/cnvCabbNscaf.html" class="ext" target="_blank">here</a>.
This will include all chromosomes (prefix 'C') and scaffolds (prefix 's') in the <tt>genomic.fna</tt> file.
Beware, there can be many tiny scaffolds. If they all aligned in SyMAP, it causes the display to be very cluttered.
Hence, it is best to just align the largest ones (e.g. the longest 30); merge them if possible, then try
the smaller ones. You should set the following SyMAP project's
<a href="../SystemHelp.html#projParams" class="ext" target="_blank">Parameters</a>:
<ol>
<li><tt>Group prefix</tt> needs to be blank as there is no common prefix now.
<li><tt>Minimum size</tt> should be set to only load the largest scaffolds.
</ol>

<p>As of 23-July-2024 (GCF_000695525),
cabbage has 9 chromosome sequences and 32,877 scaffolds. This is too many individual sequences
to load, so just load the largest. You can calculate what to set the minimum size to using
the <ttp>xToSymap</ttp> <a href="./index.html#ol" class="ext" target="_blank">Lengths</a> button.
It outputs all their
sorted lengths followed by a summary table of lengths:
<pre>
Read cabb/sequence/genomic.fna and print sorted lengths
Read 32886 sequences
   N    Length  Seqid
   1  64984695  >C03 NC_027750.1
   2  54679868  >C09 NC_027756.1
   3  53719093  >C04 NC_027751.1
   4  52886895  >C02 NC_027749.1
   5  48366697  >C07 NC_027754.1
   6  46902585  >C05 NC_027752.1
   7  43764888  >C01 NC_027748.1
   8  41758685  >C08 NC_027755.1
   9  39822476  >C06 NC_027753.1
  10    550871  >s01  NW_013617415.1
...(list rest of scaffolds with lengths)

Values for parameter 'Minimal length' (assuming no duplicate lengths):
#Seqs  min_len
   10   550871
   20   213381
   30   154937
   40   131235
   50   101387
   60    86265
   70    70649
   80    65803
   90    61068
  100    58362
</pre>

To align the top 30 sequences (9 chromosomes, 21 of the largest scaffolds),
this says to set <ttx>Minimum length</ttx> to 154937.

<!----------------------------------------------------------------->
<a id="load"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Load files into SyMAP</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The above scenario puts the files in the default SyMAP directories.
When you start up SyMAP, you will see your projects listed on the left of the panel
(e.g <a href="../SystemGuide.html#demoseq" class="ext" target="_blank">as shown for demos</a>).
<ol>
<li>Check the projects
you want to load, which will cause them to be shown on the right of the <ttp>symap</ttp> window.
<li>For the project you want to load,
open the project's <a href="../SystemHelp.html#projParams" class="ext" target="_blank">Parameters</a> window
to enter the appropriate values.
<li>The select <ttl>Load Project</ttl>.
</ol>

<!----------------------------------------------------------------->
<a id="edit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Editing the script</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The NCBI FASTA files are not consistent in their header lines. Hence, the parsing could
be incorrect. They may also not be consistent in the GFF files, but I have not found one.
Anyway, if it is not parsing the FASTA file correctly, edit the program
<a href="./index.html#edit" class="ext" target="_blank">as described  here.</a>
<!------------------------------------------------------------->
<a id="what"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>What the ConvertNCBI script does</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<u>FASTA</u>: Reads the file ending in '.fna.gz' (or '.fna') and writes a new file called <tt>sequence/genomic.fna</tt> with the following changes:
<ol>
<li>Sequences must start with the "NC_", "NW_" or "NT_" names,
or have the word "chromosome" or "scaffold" in their ">" header line
in order to be copied. Scaffolds are only processed if the <ttl>Include scaffolds</ttl> is selected.
<li>The header line is replaced with ">ChrN", where N comes from the header line "chromosome N", e.g.
"chromosome 1", etc.

For example,
<pre>
>NC_029256.1 Oryza sativa Japonica Group cultivar Nipponbare chromosome 1, IRGSP-1.0
</pre>
is replaced with:
<pre>
>Chr1  NC_029256.1
</pre>

<li>Gaps of &gt;30,000 are written to the <tt>annotation/gap.gff</tt> file (this value can be changed
in the <ttl>xToSymap</ttl> interface).
<li>If <ttl>Hard mask</ttl> is selected, all lower case bases are changed to 'N'.
</ol>

<p><u>GFF</u>: Reads the file ending in 'gff.gz' (or .gff) and writes the file <tt>annotation/anno.gff</tt>. The
<a href="http://www.sequenceontology.org/gff3.shtml" class="ext" target="_blank">gff3 format</a>
has 9 columns, where the first is the 'seqid', the third is the 'type' (e.g. feature 'gene'), the
last column is a semicolon-delimited keyword=value attribute list. The input file is processed as follows:

<ol>
<li>The <i>type=gene</i> with attribute <i>gene-biotype=protein-coding</i> are processed.
The gene line is written to the <tt>anno.gff</tt> file with the following changes:

<ol type="i">
<li>The first column 'seqid' is replace with the 'ChrN' value assigned when reading the '.fna' file.
<li>A subset of the attributes are written, where the product keyword value is modified as follows:
<ol type='a'>
<li>If there are multiple mRNA lines for a gene where the values are different, they are concatenated together.
<li>If there are multiple mRNA lines for a gene where
the only difference is the variant, then only the variant difference is show, e.g.
<pre>
product=monocopper oxidase-like protein SKU5%2C transcript variant X2, X1, X3
</pre>
</ol>
</ol>

<li>The first <i>type=mRNA</i> line for a gene is written to the <tt>anno.gff</tt> file followed
by its <i>type=exon</i>, where they each are written with the new seqid and a subset of the
attributes.
<br>&nbsp;
<li>If <ttl>1st protein</ttl> is selected, then the <tt>ID=</tt> from the first <i>type=CDS</i> for the gene is extracted,
and the "cds-" removed. From the files
I have tested, this name corresponds with the ">" name in the NCBI protein file.
If <ttl>All proteins</ttl> is selected, but all unique ID's from the
gene's <i>type=CDS</i> records are made into a comma-delimited list; note, this can cause a long list.
</ol>

<p><a href="#top">Go to top</a>
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
