<!DOCTYPE html>
<html>
<head>
<title>SyMAP Other inputs</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="../style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="https://csoderlund.github.io">
		<img src="../img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;Other Inputs</p>

</div>

<!------ right side header and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav"> <!-- header same in all pages -->
    <br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="../img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="../img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="../Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="../SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="../input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="../SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="../UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="../Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">
<p>
This page covers using FASTA and GFF files from sources other than NCBI or Ensembl.
<h3>Requirements</h3>
The requirements are listed in <a href="./index.html#seqprep" class="ext" target="_blank">Sequence files</a> and
<a href="./index.html#seqprep" class="ext" target="_blank">Annotation files</a>.

<p>Run <a href="./index.html#sf" class="ext" target="_blank">Summarize</a> to
make sure your files have the right requirements for input to SyMAP; if they do, you may directly input them to
SyMAP.

<p>The <ttp>xToSymap</ttp> conversion programs cannot be of aid to you if your GFF file does not meet the
requirements specified. You will need to write your own conversion program, or you can send email
to <ttl>cas1@arizona.edu</ttl> and I will do it for you.
<p>
The main reason for conversion is to provide shorten chromosome/scaffold names. If your sequence names are long,
it really clutters everything; find a way to edit your files to shorten them.
You may be able to use one of the <ttp>xToSymap</ttp> conversion programs for your data.

<h3>xToSymap NCBI and Ensembl conversions</h3>
<p>Some important differences between the NCBI and Ensembl files:
<ol>
<li><u>Protein_coding attribute</u>: NCBI uses <ttx>gene_biotype=</ttx> whereas Ensembl uses  <ttx>biotype=</ttx>. Since
this is the keyword that says whether the gene is "protein_coding", it is <b>VERY</b> important.
<br>&nbsp;
<li><u>Description attribute</u>: Early NCBI files did not have a <ttx>description=</ttx> keyword for the gene, and only had <ttx>product=</ttx>
keywords for the mRNAs (now it has both). In constrast, Ensembl only has the <ttx>description=</ttx> keyword for the gene.
<br>&nbsp;
<li><u>File suffix</u>: NCBI uses the suffixes ".fna" and ".gff" whereas Ensembl uses ".fa" and ".gff3". The respective conversions requires
the correct suffix; this prevent running the wrong conversion.
</ol>
When you run <ttb>Summarize</ttb> on recent files, you will see as its last line:
<pre>
Input type: NCBI chromosome prefix
            NCBI mRNA 'product' keyword; NCBI 'gene_biotype' keyword
</pre>
or
<pre>
Input type: Ensembl like 'Number, X, Y, Roman'
            Ensembl gene 'description' keyword; Ensembl 'biotype' keyword
</pre>
The NCBI chromosome prefix is precisely "NC_". However, if your file has its own prefix, but otherwise adheres to the NCBI keywords, you can
enter a <ttx>Only prefix</ttx> with the NCBI convert.

<p>The Ensembl <ttx>#, X, Y, I</ttx> is when the FASTA line starts with the chromosome indicates,
<br>e.g. '>1' or '>X' or '>XI';

<p>When you run <ttb>Summarize</ttb>, if your results match NCBI or Ensembl, that will imply you can use one of them.
<p>For details on the conversion, see
<a href="./ncbi.html#what" class="ext" target="_blank">NCBI</a> and <a href="./ensembl.html#what" class="ext" target="_blank">Ensembl</a>.
<!---- END BODY -->
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p class="trailer">  Email Comments To: <a href="mailto:cas1@arizona.edu">cas1@arizona.edu</a></p>
</div>
</div>
</body>
</html>



