<!DOCTYPE html>
<html>
<head>
<style>
.ty {border: 1px solid black; border-spacing: 0px;  border-collapse: collapse;}
.ty td {border: 1px solid black; padding: 3px;}
.tx {border: 0px; border-spacing: 0px; }
.tx td {border: 0px; padding: 3px;}
</style>
</head>
<body style="padding:20px;font-family:arial;">

<div style="width:500px;">

<h2>Sequence Project Parameters</h2>
<p>
This page provides all the setup parameters for a sequence project.
Parameters are saved to the <tt>data/seq/&lt;project-name&gt/params</tt> file.
When the project is loaded, the parameters are also saved to the database.

<p>The most important parameters are the following:
<p>
<b>sequence_files</b>
<p>
Select the input FASTA sequence file(s), or directories
of sequence files. There must be at least
one sequence file; a file can contain multiple sequences.
<b>Default location:</b> <tt>data/seq/&lt;project-name&gt/sequence</tt> (i.e. if no "sequence_files"
are entered, SyMAP will look in the default directory).
<p>

<b>anno_files</b>
<p>
Select the input GFF3-formatted annotation files corresponding to your sequences. Note, using
a GFF3 file directly can cause problems;
see the <a href="https://csoderlund.github.io/SyMAP/SystemGuide.html#annot">SystemGuide.html</a>.
Annotation is optional but highly recommended.
<b>Default location</b>: <tt>data/seq/&lt;project-name&gt/annotation</tt> (i.e. if no "anno_files"
are entered, SyMAP will look in the default directory).
<p>
<b>grp_prefix</b>
<br>The term "Group" is used for any sequence type, e.g. chromosome, scaffold, contig. It is
used interchangeably with the term "sequence".
<table class="tx">
<tr><td valign="top">1.
<td>The parameter effects the display, and is applied during the loading of the sequences.
If provided, this allows SyMAP to remove the prefix from the chromosome or contig
names and use the remaining part as a shorter name (e.g. "100" instead of "contig_100").
<tr><td valign="top">2.
<td>All sequences in the FASTA file must have this prefix. If the sequences have a mix of prefixes, leave this blank.
<tr><td  style="vertical-align: text-top;">3.
<td><i>Long names really clutter the display</i>; it is best to use a consistent prefix that SyMAP can remove,
otherwise, use really short prefixes, e.g. 's' for scaffold, 'c' for chromosome.
<tr><td valign="top">4.
<td>You may remove the prefix after the project is loaded. For example, if your sequences had
names "Scaf1", "Scaf2", etc, and "Scaf" was NOT entered as the grp_prefix before load, you may
enter it later and it will be removed from all sequence names. However, a prefix cannot be added
to the sequence names.
</table>

<!-- <b>grp_sort</b><p> How your sequences should be ordered, either numerically, alphabetically, or in the order
found in the fasta file (this only works if they are all in one file). Note that you
can change the ordering from "file" to one of the others, without reloading the files.
-->
<p>
<b>Other parameters</b>
<p>The following table is broken into section on what needs
to be re-run following the parameter change.

<!----------------------------------------------------------------------->
<table class="ty">
<tr>
	<td><b>Parameter</b>
	<td><b>Description</b>
	<td><b>Default Value</b>
<tr><td colspan=3><i><b>Nothing needs to be re-run - new values take IMMEDIATE effect</b></i>
<tr>
	<td valign="top">category
	<td style="width: 400px;">
		Category label for the project. This is only used in the
		Summary list of the Project Manager window.
	<td>Uncategorized
<tr>
	<td valign="top">display_name
	<td valign="top">A reader-friendly name for the project. It can contain any characters
		and multiple words, but shorter names will work better in the displays.
	<td>Defaults to the project directory name.
<tr>
	<td valign="top">abbrev_name
	<td valign="top">A name (must be exactly 4 characters) to be used in the column headings for SyMAP Query.
	<td>Defaults to the last 4 characters of the display_name.
<tr>
	<td valign="top">description
	<td>Description of the project. This is only shown in the Project Manager Summary
		section.
	<td>&nbsp;
<tr>
	<td valign="top">grp_type
	<td>How to refer to the individual sequences, e.g. as "Chromosome" or "Scaffold" or "Contig"
	<td valign="top">Chromosome
<tr>
	<td valign="top">min_display_size_bp
	<td>For the dotplot display, in case some of your sequences are much shorter than others. Set this value to a
		minimum basepair size and all sequences will be scaled to appear at least this large
		shown in a dotplot.
	<td valign="top">0
<tr>
	<td valign="top">
		annot_kw_mincount
	<td>This applies to the annotation attributes columns shown in the SyMAP Query results table. See the
		Annotation section below.
	<td valign="top">50
<tr><td colspan=3><i><b>Reload Project</b> (All: remove existing alignment files), <b>Alignment&Synteny</b></i>
<tr>
	<td valign="top">grp_prefix
	<td>Explained above
	<td>Chr
<tr>
	<td valign="top">min_size
	<td>Minimum length of sequence to load; smaller sequences will be ignored. Note that annotations
		for ignored sequences will also be ignored, but some error messages will print.
	<td valign="top">100000
<tr><td colspan=3><i><b>Reload annotation, Alignment&Synteny</b> (if alignment files exist, only Synteny will be run)</i>.
<tr>
	<td valign="top">
		annot_keywords
	<td>
	This can be used to reduce the annotation attribute keywords shown in the 2D display and SyMAP Query table.
	See the Annotation section below.
	<td>&nbsp;
<tr><td colspan=3><i><b>Clear, Alignment&Synteny</b> (clear removes the alignment files)</i>
<tr>
	<td valign="top">mask_all_but_genes
	<td>Mask out all non-genic parts of the sequences before running MUMmer (gene annotation must
		be provided). This can save time but prevents non-annotated anchors from being found.
	<td valign="top">no
<tr><td colspan=3><i><b>Alignment&Synteny</b> (if alignment files exist, only Synteny will be run)</i>.
<tr>
	<td valign="top">order_against
	<td>For draft contig sets, this allows you to order them using synteny to
		one of the other projects. The ordering occurs automatically after the synteny alignment is run,
		resulting in the following two:
		<ul>
		<li>A file showing the ordering is printed into the "order_against" project directory called "ordered.cvs".
		<li>A new "_ordered" project is created,
		in which the draft contigs are concatenated into chromosome sequences based on the synteny;
		these can be aligned to the "order_against" project in order to get a less fragmented
		alignment.
		</ul>
	<td valign="top">None
</table>

<h3>Annotation</h3>
The gene annotation is shown on the 2D display and as columns in the SyMAP Query results table. The annotation
comes from the last column, called attributes, for each gene in the GFF file. The attributes is a keyword=value list,
e.g.
<pre>
ID=gene-AT1G01010;Name=NAC001;ID=rna-NM_099983.2;product=NAC domain containing protein 1
</pre>
Defaults: Generally, all genes in a file have the same keywords, in which case, use the defaults. This will cause the
entire attribute to be shown for the gene in the 2D display, and the SyMAP Query table would have columns for each
keyword that has over 50 occurrences. In the example above, the columns would be ID, Name and product (the second
ID will be ignored).
<p>
If there are many different keywords in the attribute list, this causes too many columns in the SyMAP Query table.
This can be reduced by one (or both) of the following:
<p>
annot_kw_mincount: If there are many different keywords in the attribute list, set this such that only the highest
occurrences will be in columns.
<p>
annot_keywords: The keyword/value pairs to be saved for each gene can be limited by listing
the desired keywords here (separated by commas).  Using this approach, it will also reduce the above of
annotation description per gene. For example, if the keywords ID and product were entered, the Name=value
would not be part of any gene annotation.

</div>
</body>
</html>
