<html>
<body style="padding:20px;font-family:arial;">

<div style="width:500px;">

<h2>Sequence Project Parameters</h2>
<p>
This page provides all the setup parameters for a sequence project.
The most important are the following:
<p>
<b>sequence_files</b>
<p>
Select the fasta-formatted sequence files, or directories
of sequence files, to upload to the project. There must be at least
one sequence file. Note that a file can contain multiple sequences. 
Default: data/seq/&lt;project-name&gt/sequence (if no "sequence_files"
are entered, SyMAP will look in the default directory).
<p>

<b>anno_files</b>
<p>
Select gff3-formatted annotation files corresponding to your sequences. SyMAP currently
recognizes annotations of type <tt>gene</tt>,<tt>exon</tt>,<tt>centromere</tt>,<tt>gap</tt>,
and <tt>CDS</tt> (treated the same as <tt>exon</tt>). The type is read from column 3 of
the gff file. Annotation is optional but highly recommended. 
Default: data/seq/&lt;project-name&gt/annotation (if no "anno_files"
are entered, SyMAP will look in the default directory).
<p>
<b>grp_prefix</b>
<ul>
<li>
The parameter effects the display, but is applied during the loading of the sequences.
If provided, this allows SyMAP to remove the prefix from the chromosome or contig 
names and use the remaining part as a shorter name (e.g. "100" instead of "contig_100").
All sequences must have this prefix.
<li>If the sequences have a mix of prefixes, leave this blank.
<li><i>Long names really clutter the display</i>; it is best to use a consistent prefix that SyMAP can remove,
otherwise, use really short prefixes, e.g. 's' for scaffold, 'c' for chromosome.
<li>NOTE: If you change this after load, you need to re-load and re-run all computations!
</ul>

<!-- <b>grp_sort</b><p> How your sequences should be ordered, either numerically, alphabetically, or in the order
found in the fasta file (this only works if they are all in one file). Note that you
can change the ordering from "file" to one of the others, without reloading the files.
--!>
<p>
<b>Other parameters</b>
<p>The following table is broken into section on what needs 
to be re-run following the parameter change.
<table border='1' cellpadding='5' rules=all >
<tr>
	<td align="left"><b>Parameter</b></td>
	<td align="left" ><b>Description</b></td>
	<td align="left"><b>Default Value</b></td>
</tr>
<tr><td colspan=3><i>Nothing needs to be re-run</i></td>
<tr>
	<td align="left" valign="top">category</td>
	<td align="left"  valign="top" width="400" >
		Category label for the project. This is only used in the 
		Summary list of the Project Manager window.
	</td>
	<td align="left"  valign="top" >Uncategorized
	</td>
</tr>
<tr>
	<td align="left" valign="top">display_name
	</td>
	<td align="left"  valign="top" width="400" >
		A reader-friendly name for the project. Can contain any characters 
		and multiple words, but shorter names will work better in the displays.
	</td>
	<td align="left"  valign="top" >defaults to the name used to create the project (typically the directory name)</td>
</tr>
<tr>
	<td align="left" valign="top">description</td>
	<td align="left"  valign="top" width="400" >
		Description of the project. This is only shown in the Project Manager Summary 
		section.
	</td>
	<td align="left"  valign="top" >&nbsp;</td>
</tr>
<tr>
	<td align="left" valign="top">grp_type
	</td>
	<td align="left"  valign="top" width="400" >
		How to refer to the individual sequences, e.g. as "Chromosome" or "LG" or "Contig" 
	</td>
	<td align="left"  valign="top" >Chromosome</td>
</tr>
<tr>
	<td align="left" valign="top">min_display_size_bp
	</td>
	<td align="left"  valign="top" width="400" >
		For the dotplot display, in case some of your sequences are much shorter than others. Set this value to a 
		minimum basepair size and all sequences will be scaled to appear at least this large
		shown in a dotplot.
	</td>
	<td align="left"  valign="top" >0</td>	
</tr>

<tr><td colspan=3><i>Reload Project (remove existing alignment files), Alignment&Synteny</i></tr>
<tr>
	<td align="left" valign="top">grp_prefix</td>
	<td align="left"  valign="top" width="400" >Explained above	
	</td>
	<td align="left"  valign="top" >Chr</td>
</tr>
<tr>
	<td align="left" valign="top">min_size</td>
	<td align="left"  valign="top" width="400" >
		Minimum length of sequence to load; smaller sequences will be ignored. Note that annotations
		for ignored sequences will also be ignored, but some error messages will print.
	</td>
	<td align="left"  valign="top" >100000</td>
</tr>



<tr><td colspan=3><i>Reload annotation, Alignment&Synteny (if annotation files exist, only Synteny will be run)</i>.
<tr>
	<td align="left" valign="top">
		annot_keywords
	</td>
	<td align="left"  valign="top" width="400" >
		SyMAP parses keyword/value pairs from the attributes field (last column) of the annotation
		gff files, which are entered into the database as the "Annotation Description". 
		The keyword/value pairs to be saved in the database can be limited by listing 
		the desired keywords here (separated by commas).
	</td>
	<td align="left"  valign="top" >
		
	</td>	
</tr>
<tr>
	<td align="left" valign="top">
		annot_kw_mincount
	</td>
	<td align="left"  valign="top" width="400" >
		For the SyMAP Query, all keywords are search UNLESS they have less than this many occurrences,
		e.g. if only one gene has a given keyword, then it is not included on the SyMAP query search, however,
		it will still be in the gene's annotation description. 
	</td>
	<td align="left"  valign="top" >
		0
	</td>	
</tr>

<tr><td colspan=3><i>Remove alignment files, Alignment&Synteny</i>
<tr>
	<td align="left" valign="top">
		mask_all_but_genes
	</td>
	<td align="left"  valign="top" width="400" >
		Mask out all non-genic parts of the sequences before running MUMmer (gene annotation must
		be provided). This can save time but prevents non-annotated anchors from being found. 
	</td>
	<td align="left"  valign="top" >
		no
	</td>	
</tr>

<tr><td colspan=3><i>Alignment&Synteny (if alignment files exist, only Synteny will be run)</i>.
<tr>
	<td align="left" valign="top">
		order_against
	</td>
	<td align="left"  valign="top" width="400" >
		For draft contig sets, this allows you to order them using synteny to 
		one of the other projects. The ordering occurs automatically after the synteny alignment is run,
		resulting in the following two:
		<ul>
		<li>A file showing the ordering is printed into the "order_against" project directory called "ordered.cvs". 
		<li>A new "_ordered" project is created,
		in which the draft contigs are concatenated into chromosome sequences based on the synteny;
		these can be aligned to the "order_against" project in order to get a less fragmented 
		alignment.
		</ul>
	</td>
	<td align="left"  valign="top" >
		None
	</td>
</tr>

</table>

</div>
</body>
</html>