<html>
<body style="padding:20px;font-family:arial;">

<i>If you are looking for query result column definitions, scroll down.</i>
<h1>SyMAP Query Results</h1>
For more information, see <a href="https://csoderlund.github.io/SyMAP/UserGuide.html#queries">User Guide</a>.
<p>The results table list the hits (anchors) resulting from the query, where each hit connects
two species.  If the query was for single genes, then each row represents one gene and shows data
only for one genome.

<h2>Button functions</h2>

<u>Selected:</u> One row can be selected.
<p><b>Show</b>
<br>A popup will show the information for all columns of the selected row.

<p><b>Align</b>
<br>Select one or more rows: All sequences (two per row) are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32) and shown in the main panel. The alignment is left in the <tt>muscle</tt>
sub-directory and is over-written with each new alignment.

<p><b>View 2D</b>
<br>Select one row: Brings up a SyMAP 2-track view for the selected entry.
<table border=1>
<tr><td nowrap>Region (kb) <td>The hits are initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.
The default filter is "Show all hits".
<tr><td nowrap>Synteny block <td>The entire block for this hit will be shown. The default filter is
"Show Block Hits".
<tr><td nowrap>Collinear set <td>If the hit is in a collinear set, the set will be shown. The default filter
is "Highlight Collinear Set; Show Block and Set Hits".
</table>
High: If this is checked, the selected hit will be highlighted in the 2D display using the hover color.

<p><u>Table (or selected):</u> One or more rows can be selected for the following exports,
or if no rows are selected, the table is exported.

<p><b>Export CSV</b>
<br>Exports the rows
using the selected set of columns to a CSV format suitable for import into Excel.

<p><b>Export Fasta</b>
<br>
Export the sequences from the selected hits in Fasta format. Both sides of each hit
are written, using the start/end coordinates shown in the table.
<br>NOTE: this is a very slow function and takes minutes if many rows are selected.
<!----------------------------------------------------->
<h2>Results Table</h2>

The table shows the hits (or single genes) resulting from the query.
<ul>
<li>The table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty.
<li>If the hit aligns to a gene on either/both ends,
columns of information about the gene(s) is included; however, if the hit aligns to overlapping
genes, only the gene with the best overlap is provided.
<li>If the results are for single genes, then only one species information is provided and the
other species columns are blank.
</ul>

<h3>Columns</h3>
You can sort the columns by clicking on them, and rearrange them by dragging the header
boxes. You can add/remove columns using the "Select Columns" button at the bottom. Following
are the meanings of the columns:
<p>The "Single genes" tables only have the Gene Location and Annotation columns.
<table cellpadding=3>
<tr>
	<td colspan=3><i>General</i>
<tr>
	<td valign="top"><b>Row</b><td>&nbsp;
	<td>Row number within the table
<tr>
	<td valign="top"><b>Block</b><td>&nbsp;
	<td>Synteny block containing this hit (if any). The format is C1.C2.N, where
		C1 and C2 are the chromosome numbers and N is the block number for the two chromosomes.
<tr>
	<td valign="top" nowrap><b>Block Score</b><td>&nbsp;
	<td>The number of hits which comprise the synteny block.
<tr>
	<td valign="top"><b>Collinear</b><td>&nbsp;
	<td>Collinear set containing this hit (if any). The format is C1.C2.Size.Set (e.g. 1.2.5.100; there are 5
		adjacent gene hits in set 100 on Chr1 to Chr2).
<tr>
	<td valign="top"><b>PgeneF</b><td>&nbsp;
	<td>If <b>Compute PgeneF</b>: PgeneF number. All hits in a given PgeneF have the same number. Note that
		the number is generated during the search and won't necessarily be the same
		in a different search.
<tr>
	<td valign="top"><b>PgFSize</b><td>&nbsp;
	<td>If <b>Compute PgeneF</b>: Size of the PgeneF which contains this hit. Sort on this column to view the
		other rows with same PgeneF.
<tr>
	<td valign="top"><b>Hit#</b><td>&nbsp;
	<td>The hit number, which are sequential numbers for the chromosome pair. This is shown on the "Chromosome Explorer" when
		the mouse is over the hit line.
<tr><td valign="top"><b>Hit %Id</b><td>&nbsp;
	<td>Percent identity of the alignment.The value of the "Identity" column from the MUMmer file.
<tr><td valign="top"><b>Hit %Sim</b><td>&nbsp;
	<td>Percent similarity of the alignment (as determined by the BLOSUM scoring matrix).
	The value of the "Similarity" column from the MUMmer file.
<tr><td valign="top"><b>Hit #Merge</b><td>&nbsp;
	<td>The number of clustered hits into one merged hit.
<tr><td valign="top"><b>Hit St</b><td>&nbsp;
	<td>If "=", both hit ends are to the same strand; if "!=", they are to different strands.
<tr>
	<td colspan=3>
		<i>Location: one row for each species</i>
<tr>
	<td valign="top"><b>Chr</b><td>&nbsp;
	<td>Chromosome (or draft sequence number) where the hit is aligned

<tr>
	<td valign="top" nowrap><b>Gstart/Gend/Gs</b><td>&nbsp;
	<td>Gstart/Gend of the annotated gene. The Gst is the strand (+/-).
<tr>
	<td valign="top"><b>Gene#</b><td>&nbsp;
	<td>The gene number is C.#.{a-z}. The C is the chromosome number.
	The # is the sequential number along the chromosome.
	If a run of genes overlap, they receive the
	same gene number with different suffixes {a-z, a2-z2, etc}. This is shown on the Chromosome Explorer
	"Annotation Description" and when the mouse hoovers over the gene.
<tr>
	<td valign="top" nowrap><b>Hstart/Hend</b><td>&nbsp;
	<td>Hstart/Hend of the hit region.
		<br><u>Merged hits</u>: Hits are clustered during anchor loading, so hit regions
		are not necessarily single MUMmer alignments.
<tr>
	<td colspan=3><i>Annotation: one row for each species</i>
<tr>
	<td colspan=3>The keywords for the annotations of each species are listed.
	They can be different for each species, e.g. one may have "ID" while another uses "Name" for the
	gene names. "All_anno" shows all of the attributes joined together. The keywords used can be limited
	using the <tt>Project Parameters</tt> (see <tt>annot_kw_mincont</tt> and <tt>annot_keywords</tt>).
</table>


<h2>Statistics</h2>
This section shows overall statistics for the query results. Most of the
statistics are self-explanatory except the following:
<p><b>Annotated</b> and <b>Genes</b>: The first is the number of hits that
overlap one or more genes, where a gene can have multiple hits.
The second is the number of Genes with at least one hit.
<p><b>Regions</b>: This statistic
is only shown if <b>Compute PgeneF</b> was checked, and it is
the number of distinct regions covered on that species.

</body>
</html>
