<html>
<body style="padding:20px;font-family:arial;">

<div style="width:500px;">
<i>If you are looking for query result column definitions, scroll down.</i>
<h2>SyMAP Query Results</h2>
<p>
This page shows the table of hits (anchors) resulting from the query. Each hit connects
two species and you can see the respective chromosomes and start/end locations of 
the hits, as well as gene annotations overlapped by the hits. (If the query
was for orphan genes, then each row represents one gene and shows data
only for one genome).
<p>
Note that entries in the table may be selected. To select more than one
entry, use "ctrl-click". 
<p>
<h3>Button functions</h3>

<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>Show Synteny</b>
<p>
Brings up a SyMAP 2-track view for each selected entry, so the hits
can be seen in their full chromosomal context. The hits are initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.  
<p>
<b>Save CSV</b>
<p>
Saves the selected hits using the selected set of columns to a CSV format suitable for import into Excel.
<p>
<b>Save for Reload</b>
<p>
Saves the selected hits in a CSV format which can be reloaded back into SyMAP 
later (click "Results" on the lefthand pane,
and use the "Load Saved Query" button). The saved table includes <i>all</i> possible columns, not just
those currently selected for display. 
<p>
<b>Save as Fasta</b>
<p>
Sequences from the selected hits are written to a Fasta file. Both sides of each hit
are written, using the start/end coordinates shown in the table.
<p>
<b>Align sequences</b>
<p>
Sequences for the selected hits are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32). 
</div>

<p>

<h3>Results Table</h3>

The table shows the hits (or orphan genes) resulting from the query. Each row is a single
hit or orphan gene. Note that the table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty (and
an orphan gene only has one non-empty species). 
<p>
You can sort the columns by clicking on them, and rearrange them by dragging the header
boxes. You can add/remove columns using the "Select Columns" button at the bottom. Following
are the meanings of the columns:
<p>
<table border='1' cellpadding='5' rules=all >
<tr>
	<td align="center" colspan=2>
		<i>General</i>
	</td>

</tr>
<tr>
	<td align="left" valign="top">
		<b>Row</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Row number within the table
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>PgeneF</b>
	</td>
	<td align="left"  valign="top" width="450" >
		PgeneF number. All hits in a given PgeneF have the same number. Note that 
		the number is generated during the search and won't necessarily be the same
		in a different search.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>PgFSize</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Size of the PgeneF which contains this hit.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>HitIdx</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Database index of the hit. This is used in the "Save for Reload" function.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>BlockNum</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Synteny block containing this hit (if any). The format is <chr1>.<chr2>.N, where
		runs from one to the number of blocks between those two chromosomes.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>BlockScore</b>
	</td>
	<td align="left"  valign="top" width="450" >
		The number of anchors which comprise the synteny block.
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>RunSize</b>
	</td>
	<td align="left"  valign="top" width="450" >
		If the hit is contained in an uninterrupted collinear sequence of shared
		genes, RunSize shows the length of these sequence. If you have chosen
		the option "Only hits in a collinear gene pair", then RunSize will be two
		or greater for all returned hits.
	</td>
</tr>
<tr>
	<td align="center" colspan=2>
		<i>Location: one row for each species</i>
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>Chr</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Chromosome (or draft sequence number) where the hit is aligned
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>Start/End</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Start/end of the hit region (note that hits are clustered during anchor loading, so hit regions
		are not necessarily single MUMmer alignments).
	</td>
</tr>
<tr>
	<td align="left" valign="top">
		<b>#RGN</b>
	</td>
	<td align="left"  valign="top" width="450" >
		Shows the number of distinct regions on this species which are included in the 
		PgeneF containing this hit. For example, if a gene from Species A expanded by 
		tandem duplication to 5 genes on Species B, then the PgeneF will have PgFSize=5,
		and #RGN will be one for Species A and 5 for species B.
	</td>
</tr>
<tr>
	<td align="center" colspan=2>
		<i>Annotation: one row for each species</i>
	</td>
</tr>
<tr>
	<td align="left" colspan=2>
	The key/value pairs from the annotation gff attributes sections are shown here 
	They can be different for each species, e.g. one may have "ID" while another uses "Name" for the 
	gene names. "All_anno" shows all of the attributes joined together.  
	</td>
</tr>
</table>

<h3>Statistics</h3>
This section shows overall statistics for the query results, for each species. 
The meaning of the terms is as follows: 
<div style="width:485px;margin-top:10px; margin-left:15px;">
<b>hits</b>
<p>
Number of hits involving that species.
<p>
<b>distinct regions</b>
<p>
Number of distinct regions covered on that species (see the definition of the "#PGF" column above). 
<p>
<b>annotated</b>
<p>
Out of the number of distinct regions, how many have annotation.
<p>
<b>orphans</b>
<p>
Number of orphan genes returned for this species.
<p>
<b>chrs</b>
<p>
Number of chromosomes (or draft sequences) this species has, for reference.


</div>

</div>
</body>
</html>