<html>
<body style="padding:20px;font-family:arial;">

<i>If you are looking for query result column definitions, scroll down.</i>
<h1>SyMAP Query Results</h1>
<p>
This page shows the table of hits (anchors) resulting from the query, where each hit connects
two species.  If the query
was for orphan genes, then each row represents one gene and shows data
only for one genome.
<h2>Button functions</h2>

<u>For selected:</u> As detailed for the following two:

<p><b>Align sequences</b>
<br>Select one or more rows: All sequences (two per row) are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32) and shown in the main panel.

<p>NOTE: This is especially
useful to view the similarity of putative gene families (PgeneF),
as follows: check "Compute PgeneF" and "Run Query", show the
PgeneF and PgFsize columns, sort on PgeneF, then select the rows for PgeneF to
align.

<p><b>Show Synteny</b>
<br>Select one row: Brings up a SyMAP 2-track view for the selected entry. The hits are initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.


<p><u>For table (or selected):</u> One or more rows can be selected for the following two exports,
or if no rows are selected, the table is exported.

<p><b>Export CSV</b>
<br>Exports the rows
using the selected set of columns to a CSV format suitable for import into Excel.

<p><b>Export Fasta</b>
<br>
Export the sequences from the selected hits in Fasta format. Both sides of each hit
are written, using the start/end coordinates shown in the table.
<br>NOTE: this is a very slow function and takes minutes if many rows are selected.


<h2>Results Table</h2>

The table shows the hits (or orphan genes) resulting from the query. Each row is a single
hit or orphan gene. Note that the table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty (and
an orphan gene only has one non-empty species).
<p>
You can sort the columns by clicking on them, and rearrange them by dragging the header
boxes. You can add/remove columns using the "Select Columns" button at the bottom. Following
are the meanings of the columns:
<p>
<table cellpadding=3>
<tr>
	<td colspan=3>
		<i>General</i>
	</td>

</tr>
<tr>
	<td valign="top"><b>Row</b></td>
	<td>&nbsp;</td>
	<td>Row number within the table</td>
</tr>
<tr>
	<td valign="top"><b>PgeneF</b></td>
	<td>&nbsp;</td>
	<td>
		If <b>Compute PgeneF</b>: PgeneF number. All hits in a given PgeneF have the same number. Note that
		the number is generated during the search and won't necessarily be the same
		in a different search.
	</td>
</tr>
<tr>
	<td valign="top"><b>PgFSize</b></td>
	<td>&nbsp;</td>
	<td>
		If <b>Compute PgeneF</b>: Size of the PgeneF which contains this hit. Sort on this column to view the
		other rows with same PgeneF.
	</td>
</tr>
<tr>
	<td valign="top"><b>HitIdx</b></td>
	<td>&nbsp;</td>
	<td>
		Database index of the hit. This is shown on the "Chromosome Explorer" when
		the mouse is over the hit line.
	</td>
</tr>
<tr>
	<td valign="top"><b>Block</b></td>
	<td>&nbsp;</td>
	<td>
		Synteny block containing this hit (if any). The format is C1.C2.N, where
		C1 and C2 are the chromosome numbers and N is the block number for the two chromosomes.
	</td>
</tr>
<tr>
	<td valign="top" nowrap><b>Block Score</b></td>
	<td>&nbsp;</td>
	<td>
		The number of hits which comprise the synteny block.
	</td>
</tr>
<tr>
	<td valign="top"><b>RunSize</b></td>
	<td>&nbsp;</td>
	<td>
		If the hit is contained in an uninterrupted collinear set of shared
		genes, RunSize shows the length of the set. This column is used for the
		"Collinear gene" query.
	</td>
</tr>
<tr>
	<td colspan=3>
		<i>Location: one row for each species</i>
	</td>
</tr>
<tr>
	<td valign="top"><b>Chr</b></td>
	<td>&nbsp;</td>
	<td>
		Chromosome (or draft sequence number) where the hit is aligned
	</td>
</tr>
<tr>
	<td valign="top" nowrap><b>Start/End</b></td>
	<td>&nbsp;</td>
	<td>
		Start/end of the hit region.
		<br><u>NOTE1</u>: Hits are clustered during anchor loading, so hit regions
		are not necessarily single MUMmer alignments).
		<br><u>NOTE2</u>: If the table contains single project genes (e.g. orphan or all),
		    then the coordinates are the gene coordinates.
	</td>
</tr>
<tr>
	<td valign="top"><b>Gene#</b></td>
	<td>&nbsp;</td>
	<td>The order number along the chromosome of the gene. This is shown on the Chromosome Explorer
	"Annotation Description" and when the mouse hoovers over the gene.
	</td>
</tr>
<tr>
	<td colspan=3>
		<i>Annotation: one row for each species</i>
	</td>
</tr>
<tr>
	<td colspan=3>
	The keywords for the annotations of each species are listed.
	They can be different for each species, e.g. one may have "ID" while another uses "Name" for the
	gene names. "All_anno" shows all of the attributes joined together.
	</td>
</tr>
</table>

<h3>Statistics</h3>
This section shows overall statistics for the query results. Most of the
statistics are self-explanatory except the following:
<p><b>Annotated</b> and <b>Genes</b>: The first is the number of hits that
overlap one or more genes, where a gene can have multiple hits.
The second is the number of Genes with at least one hit.
<p><b>Regions</b>: This statistic
is only shown if <b>Compute PgeneF</b> was checked, and it is
the number of distinct regions covered on that species.

</body>
</html>
