<!DOCTYPE html>
<html>
<head>
<title>runSingleTCW</title>
<style>
.ty    		{border: 1px; padding: 3px; border-spacing: 3px; }
.ty td 		{border: 1px; padding: 3px; border-spacing: 3px}
</style>
</head>
<body style="padding:20px;font-family:arial;">

<div style="width:500px;">

<h2>Explain Summary</h2>
<i>Clustered hits</i> have n&ge;1 MUMmer hits that may have gaps between them or overlap.
<br><i>Subhit</i> is a MUMmer hit that is part of a clustered hit.
<br><i>Synteny Block</i> is composed of clustered hits.

<p>Abbrev: Gene (GN), Clustered hit (CH), SubHit (SH), Synteny Block (SB), Cluster hit in block (CHB)
<br>If the second project does not have a value, then it is the same for the two projects.

<table class="ty">
	<tr>
		<td colspan=2 align=left>
			<b>Genome and Annotation</b>
	<tr>
		<td align=left valign=top>#Genes
		<br>~Cover
		<br>Avg
		<br>Max
		<td>Number of genes
		<br>Approximate coverage of genes; it does not take into account overlapping genes
		<br>Average length of genes
		<br>Length of the longest gene
	<tr>
		<td align=left valign=top>#Exons<br>~Cover<br>Avg<br>Max
		<td>Number of exons
		<br>Approximate coverage of exons; it does not take into account overlapping exons
		<br>Average length of exons
		<br>Length of the longest exon
	<tr>
		<td align=left valign=top>#Seqs
		<td>Total number of sequences (e.g. Chr) loaded for the project
	<tr>
		<td align=left valign=top>Length
		<td>Total length of the genome sequence
<!--------------------------------------------------------->
	<tr>
		<td colspan=2 align=left>
			<b>Clustered hits - table 1</b>
	<tr>
		<td align=left valign=top>#Clusters
		<td>Number of clustered hits
	<tr>
		<td align=left valign=top>#Rev
		<td>Number of clusters with opposite strand hits, i.e. +/- and -/+
	<tr>
		<td align=left valign=top>CHnBlock
		<td>Percent of cluster hits in blocks

	<tr>
		<td align=left valign=top>g2
		<td>Number of cluster hits whose alignment covers a gene on both sides
	<tr>
		<td align=left valign=top>g1
		<td>Number of cluster hits whose alignment covers a gene on one side
	<tr>
		<td align=left valign=top>g0
		<td>Number of cluster hits whose alignment covers zero genes
	<tr>
		<td align=left valign=top>#SubHits
		<td>Total number of hits within the clustered hits
	<tr><td align=left valign=top>Avg
		<td>Average number of subhits per cluster

	<tr>
		<td colspan=2 align=left>
			<b>Clustered hits - table 2</b>
	<tr>
		<td align=left valign=top>GNwCH<br>(CHwGN)
		<td>Percent of genes with at least one clustered hit aligning to it
		<br>(Percent of clustered hits intersecting with at least one gene annotation)

	<tr>
		<td align=left valign=top>*Cover CH<br>(SH)
		<td>Percent of the genome sequence covered by clustered hits
		<br>(Percent of the genome sequence covered by subhits)

	<tr>
		<td align=left valign=top>SH Size range columns
		<td>Number of subhits having alignment lengths in the given ranges
<!--------------------------------------------------------->
	<tr>
		<td colspan=2 align=left>
			<b>Synteny Blocks</b>
	<tr>
		<td align=left valign=top>#Blocks
		<td>Total number of synteny blocks
	<tr>
		<td align=left valign=top>Inv
		<td>Number of inverted blocks; note, blocks may contain local regions of opposite orientation
	<tr>
		<td align=left valign=top>GNwCHB
		<td>Percent of genes overlapping a clustered hit in a block
	<tr>
		<td align=left valign=top>*Cover SB (&ge;2x)
		<td>Percent of genome sequence covered by synteny blocks
		<br>(Percent of genome sequence covered by two or more synteny blocks, i.e. mapping to
				a putative duplication in the other genome)
	<tr>
		<td align=left valign=top>Size range columns
		<td>Number of blocks whose genome coverage is in the given range
<!--------------------------------------------------------->
	 <tr>
		<td colspan=2 align=left>
			<b>Collinear Sets</b>
	<tr>
		<td align=left valign=top>Size range columns
		<td>Number of collinear sets within the size range
</table>
<br>The "*Cover" coverage statistics take into account overlapping hits and blocks.
</div>
</body>
</html>
