<!DOCTYPE html>
<html>
<head>
<title>SyMAP 2D Align</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}

    a.ext {font-weight: bold;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; text-align: right}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	ttp  {font-family: Helvetica, monospace; font-size: 14px; font-weight: 600;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP 2D Align
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>

				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<a id="alignment_display_3d"></a>
<h2>Details of 2D Alignment</h2>

<p><u>SyMAP clusters hits as follows</u>:
The raw anchor set are the hits found by MUMmer. Multiple hits are clustered
into a single hit when the hits are close together.
The hits in a cluster are called <i>subhits</i>; the subhits of a cluster are either
all on opposite strands or the same strands.

<p><u>Alignment:</u> To build the synteny, SyMAP uses Promer by default for analysis of two different genomes.
Promer translates sequences, uses a fast suffix tree alignment algorithm, then converts the coordinates back to
nucleotide. To display the alignment in SyMAP, SyMAP performs a semi-global dynamic programming (DP) algorithm on the nucleotide sequences using the MUMmer coordinates,
which can produce a slightly different alignment. This document discusses the following:
<ul>
<li><a href="#olap">Overlap</a>: The hits can overlap.
<li><a href="#trim">Trim</a>: There can be non-matching sequence at the beginning or end of the DP alignment.
SyMAP trims this sequence before display.
<li><a href="#strand">Strand</a>: If the gene strands are the same, the hit strands will be the same (=); likewise
if the gene strands are different, the hit strands will be different (!=). See <a href="#strand">Strand</a> for
more information on when they are different.
</ul>

<i>Click on any image to view a closeup of it.</i>
<!-------------------------------------------------------------->
<a id="olap"></a>
<h3>Overlap hits</h3>

<table>
<tr><td>
<a href="img/align2D.png">
<img src="img/align2D.png" alt="2D" style="border: 1px solid black; width: 300px"></a>
<p>The <ttx>Hit Information</ttx> image on the right titled "Hit #662" shows the coordinates of the subhits for
the hit shown in the 2D image above.
	<ol>
	<li>The "L" and "R" correspond to the 2D left track and right track, as shown in the 2D image above.
	<li>The coordinates are relative to the input genome sequences.
	<li>Both lists are sorted by start coordinate.
	<li>The subhit columns # in the two table of subhits align to each, e.g. the rows #2 align to each other.
	The sequence name that is alphabetically lower (e.g. arab&lt;cabb) is numbered 1-N; the opposite sequence
	is ordered to match the first.

	<li>Though two subhits may overlap (negative gap), they will not align to the same region on the opposite chromosome.
	</ol>
<td>
<p><a href="img/alignHitInfo1.png">
<img src="img/alignHitInfo1.png" alt="hitInfo" style="border: 1px solid black; width: 200px"></a>
</table>
Cabbage shows the distance between #7 and #6 as -16.
Viewing the Cabbage alignment, #662.7 and #662.6 overlap but they do not overlap for A.thal.

<table><tr>
<td valign="top">
<small>Selected A.thal alignment</small>
<br><a href="img/baseview.png">
<img src="img/baseview.png" alt="Align graphics" style="border: 1px solid black; width: 360px"></a>
<td><small>Selected Cabbage alignment</small>
<br><a href="img/baseview1.png">
<img src="img/baseview1.png" alt="Align graphics" style="border: 1px solid black; width: 360px"></a>
</table>

<p><b>Extreme example:</b>
<table>
<tr>
<td valign="top"><a href="img/alignHitInfo2.png">
	<img src="img/alignHitInfo2.png" alt="hitInfo2" style="border: 1px solid black; width: 220px"></a>
<td>&nbsp;
<td>
	The <ttl>Hit Info</ttl> image on the left shows that A.thal has its hit # in sequential order
	whereas Cabbage has its hit # in non-sequential order; as mentioned earlier, the two hit #1 align, etc.
	For both, the coordinates are listed
	in descending order, though Cabbage has many of its hits overlapping.
	<p>


</table>

<p><a href="#top">Go to Top</a>

<!-------------------------------------------------------------->
<a id="olap"></a>
<h3>Trimmed alignments</h3>
SyMAP extracts the sequence between the coordinates provided by MUMmer and aligns the sequence.
When the sequences do not align fully, the alignment is trimmed.
The left side of the image below required some trimming for Hit #1.6 but not for Hit #1.5. The right
image shows Hit #1.6 without trimming.

<p><a href="img/alignTrim.png">
<img src="img/alignTrim.png" alt="Trimmed align" style="border: 1px solid black; width: 750px"></a>
<p>
The MUMmer results for Hit #1.6 are:
<pre>
[S1]    [E1]    [S2]    [E2]   [LEN 1][LEN 2][% IDY] [% SIM]
77766	77939   4268    4095   174     174    91.38   94.83
</pre>
Note the the %Id for MUMmer is 91.38, and is 92.9 for the trimmed SyMAP dynamic programming result. These differences
are a natural results of the variation in algorithms. Most hits do not need trimming, or only 1 base at the beginning or
end.

<p>If you want to view the alignment without trimming, you can run <ttb>viewSyMAP</ttb> with the "-a" flag.

<p><a href="#top">Go to Top</a>
<a id="strand"></a>
<h3>Strand</h3>
<u>Detail of <ttx>Algo1</ttx></u>. When the synteny is created, the hits are clustered and assigned to genes by a
<a href="SystemHelp.html#clust" class="ext" target="_blank">clustering algorithm</a>, either using
<ttx>Algo1</ttx> or <ttx>Algo2</ttx>.
For <ttx>Algo1</ttx>, occasionally the strand is wrong, which only effects the <ttb>Align</ttb> view, as shown below:
<table>
<tr><td><a href="img/alignStrand.png">
<img src="img/alignStrand.png" alt="Align strand" style="border: 1px solid black; width: 400px">
<td><ttx>Algo1</ttx> only: The A.thal gene is on the negative strand, but its aligned hit is shown on the postive strand. Most of these
occurrences have been fixed in v5.4.8, but a few will have slipped through.
</table>

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

