<!DOCTYPE html>
<html>
<head>
<title>SyMAP Conserved</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}

    a.ext {font-weight: bold;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; text-align: right}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	ttp  {font-family: Helvetica, monospace; font-size: 14px; font-weight: 600;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Conserved genes
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>

				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<!-------------------------- --------------------------------------->

<h3>Highlight conserved genes</h3>


<table>
	<tr><td>
	As stated in <a href="#sequence_filter">Sequence Filters</a>, there are two extra option for the first <i>reference</i> track,
	as shown on the right. Referring to <ttx>Hit g2x2</ttx>, if there are 3 tracks,
	the genes conserved across the 3 tracks will be highlighted, as shown below.
	<td><a href="img/seqFilterRefOnly.png">
		<img src="img/seqFilterRefOnly.png" alt="sequence filter reference" style="border: 1px solid black; width: 250px"></a>
</table>

<p><a href="img/seqFilterConserved.png">
	<img src="img/seqFilterConserved.png" alt="sequence filter conserved" style="border: 1px solid black; width: 700px"></a>
<p>The above image has the settings <ttb>Sequence Filter</ttb> <ttx>Conserved</ttx> and the <ttb>Hit Filter</ttb> <ttx>Hit =2 genes</ttx>.
Hence, the pink hit-wires and cyan exons indicate the conserved genes, and
the green hit-wires have genes at both ends but are not conserved across
the 3 tracks.
<p><font color=green>Caveats:</font>
<ol>
<li>For a 3-chromosome view, say genes A-B-C are conserved. The hit for A-B may not overlap the hit for B-C. The algorithm only
checks the A-B hit and B-C hit, but not the possible A-C hit.
<li>There may be different numbers of highlighted genes on the different tracks or hit-wires between tracks. This is because multiple hits can align to a gene, and one gene can align
to two different genes on the opposing genome.
<li>When there are overlapping genes, a hit is assigned to just one of them, which
causes some conserved genes to have incorrect pairing.
<li>There are occasional tiny hits that just barely overlap genes; the current algorithm marks them as conserved genes.
<li>Currently, the hits may align strictly in introns (especially for the long introns of mammalian genomes); the current
algorithm marks these as conserved.
</ol>
Even with these caveats, I rarely see a pairing that seems 'iffy' with plant genomes, though the overlap patterns
and long introns of mammalian genomes cause some questionable results. I will be working on improving the hit-to-gene assignment
for the next release. As always, feedback is welcome!

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

