<!DOCTYPE HTML>
<html>
<head>
<title>SyMAP Demo</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="https://csoderlund.github.io">
		<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;Results</p>
<p>
<a href="#reg">Demo Results</a>
<a href="#two">&nbsp;&nbsp;Complete</a>
<a href="#draft">&nbsp;&nbsp;Draft</a>
<a href="#self">&nbsp;&nbsp;Self-synteny</a>
<a href="#slog">&nbsp;&nbsp;symap.log</a>
<a href="#synteny">Synteny Results</a>
<p>&nbsp;
<a href="#top">Go to top</a>
</div>

<!------ right side header and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav"> <!-- header same in all pages -->
<br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="./Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="./SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">
<p><center><b><big>Demo and Synteny Results</big></b></center>
<br>
<table class="tz"><tr> <!----- top index  ------->
<td><a href="#reg">&nbsp;&nbsp;&nbsp;Demo_Seq-Demo_Seq2 Results</a><td>
<td><a href="#synteny">Synteny Results</a>&nbsp;&nbsp;&nbsp;&nbsp;
</table>
<p>The first section provides results from the supplied demo, which
is used by <a href="SystemGuide.html" class="ext">System Guide</a> in discussion of running the demo.
<p>The second section describes synteny results using the different options; it uses a chromosome
from <i>Arabidopsis thaliana</i> and one from <i>Brassica rapa</i>.

<a id="reg"></a>
<h2>Demo Results</h2>
<table class="tz"><tr>
<td><a href="#two">Two complete genomes</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#draft">Draft</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#self">Self-synteny</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#slog">symap.log</a>&nbsp;&nbsp;&nbsp;&nbsp;
</table>
<p>This demo was created from real data &gt;15 years ago, and is VERY obsolete - it is strictly for demonstration.

<a id="two"></a>
<h3>Two complete genomes: Demo-Seq to Demo-Seq2</h3>
<table class="tx">
	<tr><td>&nbsp;<td><small>Click an image to view a closeup.</small>
	<tr><td style="vertical-align: text-top;">
	The demo that is provided with the SyMAP tarball has the projects <ttx>demo_seq</ttx>, <ttx>demo_seq2</ttx>
	and <ttx>demo_draft</ttx>.
	<p>The project <ttx>demo_draft..demo_seq2</ttx> is created with ordered contigs,
	as described in <a href="#draft">Draft</a>.
	<p>This section discusses the
	<ttx>demo_seq</ttx> to <ttx>demo_seq2</ttx> synteny, which represent complete genome sequences.
	<td><a href="img/demoProjects.png">
    	<img src="img/demoProjects.png" alt="demo projects" style="border: 1px solid black; width: 450px"></a>
</table>

<p>There are a variety of options, as discussed in
<a href="SystemHelp.html" class="ext" target="_blank">Pair Parameters</a>.
The following shows the <ttp>Dot Plot</ttp> and <ttp>Summary</ttp> using the following:
<p>
<ul>
<li><ttl>Cluster Algo1 (original modified)</ttl> or <ttl>Cluster Algo2 (exon-intron)</ttl>: The second
is recommended when there is good gene annotation provided.
<p>
<li><ttl>Synteny Original</ttl> or <ttl>Synteny Strict</ttl>:
The second is recommended for complete genome sequence; it does not perform well for draft.
</ul>
<p>

<a id="dot"></a>
<table class="tx">
<tr><td colspan="2"><ttl>Cluster Algo1 (original modified)</ttl> and <ttl>Synteny Original</ttl>
<tr><td style="vertical-align: text-top;"><a href="img/demoSum.png">
       <img src="img/demoSum.png" alt="demoSum" style="border: 1px solid black; width: 350px"></a>
     <td style="vertical-align: text-top;"><a href="img/demoDotplot.png">
	     <img src="img/demoDotplot.png" alt="demoDotplot" style="border: 1px solid black; width: 400px"></a>

<tr><td colspan="2"><ttl>Cluster Algo2 (exon-intron)</ttl> and <ttl>Synteny Strict</ttl>
<tr><td style="vertical-align: text-top;"><a href="img/demoSum2.png">
       <img src="img/demoSum2.png" alt="demo Sum algo2" style="border: 1px solid black; width: 350px"></a>
     <td style="vertical-align: text-top;"><a href="img/demoDotplot2.png">
	     <img src="img/demoDotplot2.png" alt="demoDotplot2" style="border: 1px solid black; width: 400px"></a>
<tr><td colspan="2">There is very little difference between the <ttp>Dotplots</ttp>. The differences are
more pronounced on the <ttp>Chromosome Explorer 2D</ttp> and <ttp>Queries</ttp>. The
difference in the synteny algorithm are obvious on the more complicated examples, as shown in
<a href="#synteny">Synteny</a>.
</table>

<!----------------------------------------------------------->
<p>
<a id="draft"></a> <!-- ref -->
<h3>Draft: Demo-Draft to Demo-Seq2</h3>

<table style="width: 800px;">
<tr><td colspan="2">The <tt>Demo-draft</tt> was run against the <tt>Demo-Seq2</tt> with its <ttx>order_against</ttx> parameter set to <tt>Demo_Seq2</tt>.
The resulting dot plot is shown on the right. Note that the 43 contigs
are ordered, but still fragmented.
<tr><td style="vertical-align: text-top;">
<a href="img/demoSumDraft.png">
    <img src="img/demoSumDraft.png" alt="demoSumDraft" style="border: 1px solid black; width: 340px"></a>
<td><a href="img/demoDraft2.png">
    <img src="img/demoDraft2.png" alt="demoDraft2" style="border: 1px solid black; width: 400px"></a>
<tr><td>&nbsp;
<tr><td colspan=2>When the ordering algorithm is run, it creates a new project called <tt>Demo-Draft.Demo_Seq2</tt>,
where the 43 contigs are written into three chromosomes; the first two correspond
to chr1 and chr3 of <tt>Demo_Seq2</tt>, and chr0 contains the contigs that do not align.

<p>The new project <tt>Demo-Draft.Demo_Seq2</tt>
was aligned with <tt>Demo_Seq2</tt>, resulting in the dot plot shown on the right.
<tr><td style="vertical-align: text-top;">
<a href="img/demoSumOrd.png">
    <img src="img/demoSumOrd.png" alt="demoSumOrd" style="border: 1px solid black; width: 340px"></a>
<td><a href="img/demoDraftOrd.png">
    <img src="img/demoDraftOrd.png" alt="demoDraftOrd" style="border: 1px solid black; width: 400px"></a>
</table>

<!-------------------------------------------------------------------->
<!----------------------------------------------------------->
<a id="self"></a> <!-- ref -->
<h3>Self-synteny: Demo-Seq</h3>

<table class="tx">
<tr><td colspan="2"><tt>Demo-Seq</tt> has been run against itself.
There are a few blocks, but these blocks are probably too small to be meaningful.
The blocks below the diagonal are mirrored above
the diagonal, e.g. 3.5.1 and 5.3.1 are the same.
<tr> <td style="vertical-align: text-top; width: 670px">
<a href="img/demoSumSelf.png">
    <img src="img/demoSumSelf.png" alt="demo Sum Self" style="border: 1px solid black; width: 340px"></a>
<td><a href="img/demoSelf.png">
    <img src="img/demoSelf.png" alt="demoSelf" style="border: 1px solid black; width: 400px"></a>
</table>
<!-------------------------------------------------------------------->
<!---slog ref'ed----------------------------------------------------------------->
<a id="slog"></a>
<h2>symap.log</h2>

The following 4 sections show the contents of the <ttx>logs/&lt;proj-to-proj&gt;/symap.log</ttx> file:
<table class="tz"><tr>
<td><a href="#algo1off">Algo1 !verbose</a>
<td><a href="#algo1on">Algo1 verbose</a>
<td><a href="#algo2off">Algo2 !verbose</a>
<td><a href="#algo2on">Algo2 verbose</a>
<td><a href="#top">Go to top</a>
</table>
<p>The parameters selected are:
<p>
<pre>
 Use previous 2 MUMmer files 17-Aug-25 13:45   CPUs: 2
 Cluster Algo1 (modified original) or Cluster Algo2 (exon-intron)
  Synteny Strict
</pre>
The 2 MUMmer files indicates the chromosome were not split and <ttx>Concat</ttx> was
used. These two facts along with the number of CPUs can make subtle differences in
the output.
<p>
<!-------------------------------------------------------------------->
<a id="algo1off"></a>
<b>Algo1 !verbose</b>

<pre><small>
Start calculating cluster hits for demo_seq and demo_seq2
   Alignment files in data/seq_results/demo_seq_to_demo_seq2/align/
   Scan files to create candidate genes from hits
    20,005 Total scanned hits from 2 files
   Scan files to cluster and filter hits
    11,000 Total cluster hits
     6,816 Filtered hits to save
   Finding Collinear sets
       717 Collinear sets
Finish clustering hits               8s

Finding synteny for demo_seq and demo_seq2
     	17 Blocks
Finish Synteny                       2s
</small></pre>

<!-------------------------------------------------------------------->
<a id="algo1on"></a>
<b>Algo1 verbose</b>

<pre><small>
Start calculating cluster hits for demo_seq and demo_seq2
   Alignment files in data/seq_results/demo_seq_to_demo_seq2/align/
   Loading annotations
    10,108 demo_seq genes
    10,137 demo_seq2 genes
   Scan files to create candidate genes from hits
    14,058 scanned PROMER demo_seq_cc.demo_seq2_f1.mum
     5,947 scanned PROMER demo_seq_cc.demo_seq2_f2.mum
    20,005 Total scanned hits from 2 files
   Scan files to cluster and filter hits
     6,923 clustered demo_seq_cc.demo_seq2_f1.mum
     4,077 clustered demo_seq_cc.demo_seq2_f2.mum
    11,000 Total cluster hits
Filter hits
     8,556 for demo_seq
     7,350 for demo_seq2
     6,816 Filtered hits to save
Save results
   Save filtered hits
   Save hit to gene
     6,564  (6,375) for demo_seq
     6,564  (12,749) for demo_seq2
   Finding Collinear sets
       717 Collinear sets
     1,969 Updates
   Finish Collinear            0s
Finish clustering hits               8s

Finding synteny for demo_seq and demo_seq2
     6,817 Total hits
         4 group-x-group pairs to analyze
            7 Blocks chr3-chr1     Orient:  Same   3   Diff   4
            1 Blocks chr3-chr3     Orient:  Same   0   Diff   1
            1 Blocks chr5-chr1     Orient:  Same   0   Diff   1
            8 Blocks chr5-chr3     Orient:  Same   4   Diff   4
        17 Blocks
           10 Collinear Sets not in blocks (20 total hits)
Finish Synteny                       2s
</small></pre>
<!-------------------------------------------------------------------->
<a id="algo2off"></a>
<b>Algo2 !verbose</b>
<pre><small>
Start calculating cluster hits for demo_seq and demo_seq2
   Alignment files in data/seq_results/demo_seq_to_demo_seq2/align/
   Final totals    Raw hits 20,005    Files 2                             .
     4,747 Clusters   Both genes     4,713   One gene        34   No gene         0
     6,914 Filtered   Both genes     3,263   One gene     2,453   No gene     1,198   Pile hits 0
   Finding Collinear sets
       635 Collinear sets
Finish clustering hits               9s

Finding synteny for demo_seq and demo_seq2
        14 Blocks
Finish Synteny                       1s
</small></pre>

<!-------------------------------------------------------------------->
<a id="algo2on"></a>
<b>Algo2 verbose</b>

<pre><small>
Start calculating cluster hits for demo_seq and demo_seq2
   Alignment files in data/seq_results/demo_seq_to_demo_seq2/align/
   Load genes for Algo2
      Demo_Seq2  [Avg Min Max] [Gene 2,968  120   89k] [Intron   430    1   87k] [Exon   294    3 4,935] Cov 46.5%
      Demo_Seq   [Avg Min Max] [Gene 2,575   63  508k] [Intron   373    1  501k] [Exon   379    1 6,153] Cov 57.8%
   Load demo_seq_cc.demo_seq2_f1.mum  Hits 14,058 (EQ 2,893 NE 11,165)
        3,035 Clusters chr3-chr1     Both genes  3,022   One gene     13   No gene      0
          489 Clusters chr5-chr1     Both genes    484   One gene      5   No gene      0
   Load demo_seq_cc.demo_seq2_f2.mum  Hits 5,947 (EQ 2,366 NE 3,581)
          424 Clusters chr3-chr3     Both genes    417   One gene      7   No gene      0
          799 Clusters chr5-chr3     Both genes    790   One gene      9   No gene      0
   Final totals    Raw hits 20,005    Files 2
     4,747 Clusters   Both genes     4,713   One gene        34   No gene         0
     6,914 Filtered   Both genes     3,263   One gene     2,453   No gene     1,198   Pile hits 0
   Finding Collinear sets
       635 Collinear sets
     1,714 Updates
   Finish Collinear            0s
Finish clustering hits               9s

Finding synteny for demo_seq and demo_seq2
     4,747 Total hits
         4 group-x-group pairs to analyze
            7 Blocks chr3-chr1     Orient:  Same   3   Diff   4
            1 Blocks chr3-chr3     Orient:  Same   0   Diff   1
            1 Blocks chr5-chr1     Orient:  Same   0   Diff   1
            5 Blocks chr5-chr3     Orient:  Same   1   Diff   4
        14 Blocks
            6 Collinear Sets not in blocks (12 total hits)
Finish Synteny                       1s
</small></pre>
<!--------------------------------------------------------------->
<a id="synteny"></a>
<h2>Synteny Results</h2>
The following is from comparing <i>Arabidopsis thaliana</i> chromosomes 1 and 2 with
<i>Brassica rapa</i> chromosomes 1 and 7. The <ttx>Clustering hits Algo2 (exon-intron)</ttx> was
used in all cases.
<p>
<table>
<tr><td colspan=2><u><ttx>Original</ttx> vs <ttx>Strict</ttx></u>: Both allow a mix of inverted and non-inverted hits in a block, but
<ttx>Strict</ttx> only allows small inversions within a non-inverted block and vice-versa (unless
<ttx>Merge</ttx> is used).
The <ttx>Strict</ttx> blocks tend to not have the tails of dots that have bigger gaps.
<tr><td><ttx>Original</ttx><br><a href="img/dp.ab.orig.all.png">
    <img src="img/dp.ab.orig.all.png" alt="Arab Brap orig" style="border: 1px solid black; width: 380px"></a>
    <td><ttx>Strict</ttx><br><a href="img/dp.ab.strict.all.png">
    <img src="img/dp.ab.strict.all.png" alt="Arab Brap strict" style="border: 1px solid black; width: 380px"></a>

<tr><td><ttx>Original</ttx> chr1&amp;chr7<br><a href="img/dp.ab.orig.part.png">
    <img src="img/dp.ab.orig.part.png" alt="Arab Brap orig 1-7" style="border: 1px solid black; width: 380px"></a>
    <td><ttx>Strict</ttx> chr1&amp;chr7<a href="img/dp.ab.strict.part.png">
    <img src="img/dp.ab.strict.part.png" alt="Arab Brap strict 1-7" style="border: 1px solid black; width: 380px"></a>
<tr><td colspan=2>&nbsp;
<tr><td colspan=2><u><ttx>Orient</ttx></u>: This requires all hits in a block to be in the same orientation.
This options fits with what some other software packages consider 'synteny'.
<tr><td><ttx>Original Orient</ttx> chr1&amp;chr7<br><a href="img/dp.ab.orig.or.png">
    <img src="img/dp.ab.orig.or.png" alt="Arab Brap orig orient 1-7" style="border: 1px solid black; width: 380px"></a>
    <td><ttx>Strict Orient</ttx> chr1&amp;chr7<a href="img/dp.ab.strict.or.png">
    <img src="img/dp.ab.strict.or.png" alt="Arab Brap strict orient 1-7" style="border: 1px solid black; width: 380px"></a>
<tr><td colspan=2>&nbsp;
<tr><td colspan=2><u><ttx>Merge Overlap</ttx></u>: Only overlapping blocks can be merged. This will
have little effect if <ttx>Orient</ttx> has been used since mixed blocks cannot be merged.
<tr><td><ttx>Original Merge Overlap</ttx> chr1&amp;chr7<br><a href="img/dp.ab.orig.olap.png">
    <img src="img/dp.ab.orig.olap.png" alt="Arab Brap orig overlap 1-7" style="border: 1px solid black; width: 380px"></a>
    <td><ttx>Strict Merge Overlap</ttx> chr1&amp;chr7<a href="img/dp.ab.strict.olap.png">
    <img src="img/dp.ab.strict.olap.png" alt="Arab Brap strict overlap 1-7" style="border: 1px solid black; width: 380px"></a>
<tr><td colspan=2>&nbsp;
<tr><td colspan=2><u><ttx>Merge Close</ttx></u>: Blocks that are close can be merged.
<tr><td><ttx>Original Merge Close</ttx> chr1&amp;chr7<br><a href="img/dp.ab.orig.close.png">
    <img src="img/dp.ab.orig.close.png" alt="Arab Brap orig close 1-7" style="border: 1px solid black; width: 380px"></a>
    <td><ttx>Strict Merge Close</ttx> chr1&amp;chr7<a href="img/dp.ab.strict.close.png">
    <img src="img/dp.ab.strict.close.png" alt="Arab Brap strict close 1-7" style="border: 1px solid black; width: 380px"></a>
</table>
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<!---- END BODY -->
<p class="trailer">  Email Comments To: <a href="mailto:cas1@arizona.edu">cas1@arizona.edu</a></p>
</div>

</div>
</body>
</html>
