<!DOCTYPE html>
<html>
<head>
<title>SyMAP Details</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
    a.ext {font-weight: bold;}
     mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 3px; }

   ttp  {font-family: Arial, monospace; font-size: 14px; font-weight: 600; color: #556B2F; }
	ttb  {font-family: SanSerif, monospace; font-size: 14px; font-weight: 600; color: #454545;}
	ttl  {font-family: Serif, monospace; font-size: 14px; font-weight: 600; color: DarkSlateGrey;}
	ttx  {font-family: courier, monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
	tt {font-size:13px}
	pre {font-size:13px; margin-top: 0;}
    ul { margin-top: 0; }
    p { margin-bottom: 0; }

	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Details
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>

				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./input/index.html" class="white">Input</a>&nbsp;
						<a href="./SystemGuide.html" class="white">System Guide</a>&nbsp;
						<a href="./SystemHelp.html" class="white">Parameters</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a>&nbsp;
						<a href="./Query.html" class="white">Queries</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->
<b>Contents</b>
<table>
<tr>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#col">Collinear sets</a>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#con">Highlight conserved genes</a>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#hit">Hit popup</a>
</table>
<a id="col"></a>
<h2>Collinear sets</h2>
Using the <ttb>Hit Filter</ttb>, or right click in the hit region, set the highlighting for <ttx>Collinear Sets</ttx>.
As shown below, the number of highlighted collinear hits and sets will be shown in
the <ttl>Information</ttl> box.
<br>
<a href="img/colHigh.png">
<img src="img/colHigh.png" alt="Collinear Sets" style="border: 1px solid black; width: 650px"></a>
<p>
<table>
<tr><td  style="vertical-align: text-top;">
In the image on the right, two different sets are shown (green highlighted hits is the 1st set,
pink is the 2nd). They were broken up by a gene on each chromosome
that do not have a hit.
<p>A collinear set either has all genes to the same strand (+/+, -/-) or different (+/-, -/+)

<p>In the image on the right, all genes in both sets are on the same strand.

<p>In the image below, all genes are on opposite strands.
<br><a href="img/colRHits.png"><img src="img/colRHits.png" alt="Reverse hits" style="border: 1px solid black; width: 270px"></a>
<p>&nbsp;
<p>The collinear set algorithm does not consider the amount of overlap of the hit to a gene, or the similarity
of the hit sequences.

<td><a href="img/colSet2.png">
<img src="img/colSet2.png" alt="2 Collinear Sets" style="border: 1px solid black; width: 450px"></a>
</table>

<p>The following are examples of interpreting the 2D view:
<table>
<tr><td>
Sometimes a hit looks like it overlaps a gene, but the gene is actually in the gap, so it is not hit.
This is illustrated on the right, where the gene at the bottom appears hit by the pink hit-wire,
but in fact are spanned by the gap gray area. Note that the
hit popup shows it is a nn type, which hits no genes.
<td><a href="img/colG0HitsSpan.png">
<img src="img/colG0HitsSpan.png" alt="Clustered hit" style="border: 1px solid black; width: 300px"></a>
<tr><td>&nbsp;<td>&nbsp;
<tr><td   style="vertical-align: text-top;">
The image on the right shows one collinear set with 6 brown hit-wires that are not part of the set. The following
explains them, using the symbols g0 (hits 0 genes), g1 (hits 1 gene), g2 (hits paired genes, at
least one on each chromosome).
<p>
<table class="ty">
<tr><td>1<td>g1<td>Ignored, not part of the set since it does not hit paired genes.
<tr><td>2<td>g0<td>Ignored.
<tr><td>3<td>g1<td>Ignored since the same gene has a hit to a paired gene.
<tr><td>4<td>g0<td>Ignored.
<tr><td>5<td>g1<td>This ends the set since it does not hit a paired gene.
<tr><td>6<td>g2<td>This could start a new set, but is not followed by another paired gene.
</table>

<p>&nbsp;<p>As shown earlier, sometimes two genes look as one, so the <ttl>Gene Delimiter</ttl> option is important to
view these.
<br><a href="img/colSetGL.png">
	<img src="img/colSetGL.png" alt="Collinear Highlight and Gene Lines" style="border: 1px solid black; width: 250px"></a>
<td><a href="img/colG1Hits.png">
<img src="img/colG1Hits.png" alt="G1 hits" style="border: 1px solid black; width: 300px"></a>
</table>

<!--------------------------------------------------------------------->
<a id="con"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Highlight conserved genes</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/seqFilterRefOnly.png">
		<img src="img/seqFilterRefOnly.png" alt="sequence filter reference" style="border: 1px solid black; width: 400px"></a>

<p>As stated in <a href="UserGuide.html#sequence_filter" class="ext" target="_blank">Sequence Filters</a>,
when there are more than 2 tracks displayed in <ttb>2D</ttb>,
	there will be two extra options for the first <i>reference</i> track,
	as shown on the above. Referring to <ttx>Hit g2x2</ttx>, if there are 3 tracks,
	the genes conserved across the 3 tracks will be highlighted, as shown below.


<p><a href="img/seqFilterConserved.png">
	<img src="img/seqFilterConserved.png" alt="sequence filter conserved" style="border: 1px solid black; width: 700px"></a>
<p>The above image has the settings <ttb>Sequence Filter</ttb> <ttx>Hit g2x2</ttx> and the <ttb>Hit Filter</ttb> highlight <ttx>Hit =2 genes</ttx>.
Hence, the pink hit-wires and cyan exons indicate the conserved genes, and
the green hit-wires have genes at both ends but are not conserved across
the 3 tracks.
<p><font color=green>Caveats:</font>
<ol>
<li>For a 3-chromosome view, say it shows genes A-B-C are conserved. The algorithm only
checks the A-B hit and B-C hits exist, but does not the possible A-C hit.
<li>There may be different numbers of highlighted genes on the different tracks or hit-wires between tracks. This is because multiple hits can align to a gene, and one gene can align
to two different genes on the opposing genome.
<li>When there are overlapping genes, a hit is assigned to just one of them, which
causes some conserved genes to have incorrect pairing.
<li>There are occasional tiny hits that just barely overlap genes; the current algorithm marks them as conserved genes.
<li>Currently, the hits may align strictly in introns (especially for the long introns of mammalian genomes); the current
algorithm marks these as conserved.
</ol>
Even with these caveats, I rarely see a pairing that seems 'iffy' with plant genomes, though the overlap patterns
and long introns of mammalian genomes cause some questionable results. As always, feedback is welcome!

<!--------------------------------------------------------------------->
<a id="hit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Hit popup</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Default view:
<ul>
<li>Both lists are sorted by start coordinate.
<li>The # column in the two tables align to each, e.g. the rows #2 align to each other.
<li>The sequence name that is alphabetically lower (e.g. arab&lt;cabb) is the <i>query</i> and
the other is the <i>target</i>. The query is numbered 1-N; the target
	is ordered to match the query.
<li>Two subhits may overlap (Gap&lt;=0) on one chromosome but not the other.
</ul>

<h3>Merge</h3>
<table><tr>
<td><a href="img/2DhitMerge.png">
		<img src="img/2DhitMerge.png" alt="2D Hit Merge" style="border: 1px solid black; width: 500px"></a>
<td style="vertical-align: text-top;">In the merge view, hits with gap&lt;=0 are merged with the hit they overlap with.
<p>The numbered hits  are no longer 1-to-1, i.e. they do not necessarily align to each other like they do in the un-merged view.
<p><ttx>Totals</ttx> are provided for the <ttx>Len</ttx> and <ttx>Gap</ttx>.
<p>The merge view corresponds better to the visual view on the 2D track.
<p>The existence of the <ttl>Merge</ttl> is a quick indication that there are overlapping hits.
</table>

<h3>Order</h3>
<table><tr>
<td><a href="img/2DhitOrder.png">
		<img src="img/2DhitOrder.png" alt="2D Hit Order" style="border: 1px solid black; width: 450px"></a>
<td style="vertical-align: text-top;">
As stated above, each numbered hit corresponds to the same number in the opposite list,
e.g. the two #1 aligns. An '#' on the far right implies that the order numbers are not sequential.
<p>This option produces another popup where the target list is sorted by '#', but then the
coordinates are no longer sorted.
<p>The existence of the <ttl>Order</ttl> is a quick indication that there are <i>disordered</i> hits.

</table>

<p><a href="#top">Go to Top</a>
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

