<!DOCTYPE html>
<html>
<head>
<title>SyMAP MUMmer</title>
<style>
	:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: 'Lucida Console', monospace; font-size: 16px; font-weight: 500; font-variant: small-caps;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
    a.ext {font-weight: bold;}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					Using MUMmer from SyMAP
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

This document discusses how to deal with one or more MUMmer alignments failing.  This document refers to the v5.0.8. The corresponding
mummer documentation is
<a href="http://mummer.sourceforge.net" class="ext" target="_blank">v3</a> and
<a href="https://mummer4.github.io" class="ext" target="_blank">v4</a>.

<h4>Contents</h4>
<ul>
	<li><a href="#ext">Executables</a>
	<li><a href="#log">Finding the problem</a>
	<li><a href="#mem">Out of Memory</a>
	<ul>
		<li><a href="#cat">Limit CPUs and uncheck Concat</a>
		<li><a href="#mask">Not-masked or Soft-masked</a>
		<li><a href="#fail1">Fail immediately</a>
		<li><a href="#fail2">One or more alignments fail</a>
	</ul>

	<li><a href="#mum4">Using MUMmer4 from SyMAP</a>
	<li><a href="#cmd">Running MUMmer from the command line</a>
	<li><a href="#help">Getting Help</a>
	<li><a href="#mac">MacOS details</a>
</ul>
If none of the following helps, contact symap@agcol.arizona.edu.
<p>
<a id="ext"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Executables</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The alignment programs are provided in the <tt>symap/ext</tt> directory. There are
executables for 64-bit Linux and 64-bit MacOS. SyMAP will
select the correct directory for the machine you are running from, i.e. you
do not need to do anything. When SyMAP creates a database, it (1) checks the MySQL variables, and (2) checks
that the external programs are executable. If you see a message like:
<pre>
***Error - file is not executable: ext/mummer/mac/promer
</pre>
Execute:
<pre>
> chmod 755 ext/mummer/mac/promer
</pre>
Execute the program from the command line to make sure it works on you machine, e.g.
<pre>
>./ext/mummer/mac/promer

  USAGE: promer  [options]  &lt;Reference&gt;  &lt;Query&gt;

  Try './ext/mummer/mac/promer -h' for more information.
</pre>
The above shows that the promer code will execute on my MacOS.

<p>
<a id="log"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Finding the problem</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
If the MUMmer alignment fails, inspect the log files.
The log files are as follows:
<pre>
symap_5/
   error.log   # a SyMAP error will write its trace data into this file and list failed MUMmer

   logs/
      &lt;dbName&gt;_LOAD.log	# this keeps track of data loaded or removed from the database.
      &lt;project-name1&gt;-to-&lt;project-name2&gt;/ # one directory per project-to-project alignment
         &lt;alignment&gt;.log	# one file per alignment
         symap.log              # this keeps most of the output shown on the terminal for this alignment
     e.g.
       symapDemo_load.log
       demo_seq_to_demo_seq2/
         demo_seq_cc.demo_seq2_f1.log
         demo_seq_cc.demo_seq2_f2.log
         symap.log
</pre>
<ol>
<li>If an alignment is listed as failed in the <tt>error.log</tt> file,
the corresponding <tt>&lt;alignment&gt;.log</tt> file will contain the MUMmer error.
<li>If there
is no error listed in the <tt>error.log</tt>, check the <tt>symap.log</tt> and the <b>tail</b>
of the alignment files.
<li>
If the error is not found in the log files or it is not clear, try the following:
<ul>
<li>Copy the command from the terminal (or log file), and paste it on a new terminal line to execute, e.g.
<pre>
ext/mummer/mac/promer  -p data/seq_results/demo_seq_to_demo_seq2/align/demo_seq_cc.demo_seq2_f2.promer
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq2/demo_seq2_f2.fa
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq/demo_seq_cc.fa
</pre>
<li>This shows <tt>promer</tt> output directly on the terminal.
</ul>
</ol>
<a id="mem"></a>
<h2>Out of memory</h2>
A MUMmer failure is typically from insufficient memory.
The following error typically indicates a memory problem.
<pre>
Alignment program error code: 141
20220512|075853|6007| ERROR: mummer and/or mgaps returned non-zero, please file a bug report
</pre>
This error will probably not show at the terminal; read the above section.
<a id="cat"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Limit CPUs and uncheck Concat</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

There is no straight-forward way to know if you have enough memory as it depends on the
size and complexity of the two genomes being compared. If you think memory may be tight,
you can limit the amount used at one time by limiting the number of CPUs, as each
CPU uses a considerable amount of memory (e.g. 4 CPUs could collectively use 20GB of memory at once).
<p><a href="img/concat.png"><img src="img/concat.png" alt="concat" style="border: 1px solid black; width: 400px"></a>
<p>
By default, all sequences of the 1st genome are concatenated into one file and then
multiple files are created from the 2nd genome to be searched against the first;
the multiple files may have two or more sequential chromosomes concatenated together to create
files that are &gt;60M.  If memory is limited, you can uncheck <ttl>Concat</ttl> so that multiple files are
created for each genomes, and all files from the 1st genome are searched against the 2nd genome.
The following statistics are from comparing Arabidopsis thaliana (119M) against Brassica rapa (297M)
on a MacOS using 1 CPU.
<table>
<tr><td><u>Concatenated</u>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><u>Not concatenated</u>
<tr>
<td>
<pre>
48819 hits
  334 synteny blocks
46319 gene hits
38334 synteny hits

Finished in 1 hr 7 min 41 sec
</pre>
<td>

<td>
<pre>
48846 hits
  334 synteny blocks
46348 gene hits
38345 synteny hits

Finished in 1 hr 34 min 29 sec
</pre>
</table>

<p>
<a id="mask"></a>
<h3>Not-masked or Soft-masked</h3>
A memory problem can occur if the genome sequence is not masked or only soft-masked.
Either: (1) change the sequence to hard-masked, or (2) set the SyMAP parameter
<ttx>mask_all_but_genes</ttx> to <tt>yes</tt>.
<p>

<a id="fail1"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Fails immediately</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

If an alignment fails immediately, and if the last line of the first <tt>&lt;alignment&gt;.log</tt> file is:
<pre>
1: PREPARING DATA
</pre>
the reason is probably that your machine does not have near enough memory as MUMmer could not even
prepare the data. Try it again with unchecked <ttb>Concat</ttb>; if that does not work, you need more memory.

<a id="fail2"></a>
<h3>One or more fails</h3>

Sometimes just one or a few of the alignment processes will fail. You will see a line such as:
<pre>
Error: Running command: /Users/cari/Workspace/symap_5/ext/mummer/mac/promer
-p data/seq_results/demo_seq_to_demo_seq2/align/demo_seq_cc.demo_seq2_f2.promer
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq2/demo_seq2_f2.fa
data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq/demo_seq_cc.fa
</pre>

<table>
<tr>
<td><a href="img/runDialog.png"><img src="img/runDialog.png" alt="runDialog" style="border: 1px solid black; width: 420px"></a>
<td>
You will see the failure on the dialog box as is
shown on the left. The remaining processes will continue. When all processes are complete,
you will see a "?" for the pair as shown on the below.
<p>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="img/runResults.png"><img src="img/runResults.png" alt="runResults" style="border: 1px solid black; width: 220px"></a>
<p>Select the "?" followed by
<ttb>Selected Pair</ttb> and it will complete the failed processes.
</table>
<p>
A preferred way is to set CPUs to a low number if the genomes are large, that way,
it will likely finish the first time.
<p>

<a id="mum4"></a>
<h2>Using MUMmer4 from within SyMAP</h2>
Sometimes when MUMmer v3 fails, MUMmer v4 will work.
MUMmer4 is included in SyMAP package with a fix to promer (make sure you get v5.0.8 or later).
Enter the <tt>ext/mummer4</tt> directory and follow the instructions in the README.
<p>My observations on MacOS indicate that MUMmer v4 takes more memory
but less time than MUMmer v3.
<p>

<a id="cmd"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Running MUMmer from the command line</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

If you need to run MUMmer using the command line from some other machine, do the following:
<ul>
<li>The query must be alphanumerically less than reference. For the <tt>promer</tt> command:
<pre>
USAGE: promer  [options]  &lt;Reference&gt;  &lt;Query&gt;
e.g.
promer proj2 proj1
</pre>
<li>Execute:
	<ul>
	<li>For different genomes:
		<ul>
		<li>Use promer
		<li>The output of promer is input to "show-coords -dlkTH".
		</ul>
	<li>For Self synteny:
		<ul>
		<li>Use nucmer.
		<li>The output of nucmer is input to "show-coords -dlTH".
		</ul>
	</ul>
<br>
<li>Result files:
<ul>
<li>The result files must have suffix ".mum"
<li>Put the ".mum" files in the directory <tt>data/seq_results/&lt;proj1-to-proj2&gt;/align</tt>
<li>In the <tt>&lt;proj1-to-proj2&gt;/align</tt> directory, execute:
<pre>touch all.done</pre>
This creates a file, which indicates to SyMAP that the alignments are done and to process
the files in the directory ending with ".mum".
</ul>
<br>
<li>When you run <ttb>Selected Pair</ttb>, SyMAP will recognize the files and use them to build the synteny blocks.
</ul>

<p><u>Example of using mummer for loaded projects <tt>demo_seq</tt> and <tt>demo_seq2</tt>:</u>
If you try this with <tt>demo_seq</tt> and <tt>demo_seq2</tt>, make sure there is no current alignment in the
database or disk for your project (e.g. <ttb>Clear Pair</ttb>).

<p><tt>Demo_seq</tt> has chr3 and chr5, <tt>demo_seq2</tt>  has chr1 and chr3; these will be compared separately against each other.
<tt>Demo_seq</tt>  is alphanumerically less than <tt>demo_seq2</tt> , so it is first in the directory name
but second in the command arguments (this is how SyMAP knows which
project Chr3 belongs to). The commands are as follows:

<pre>
   cd data/seq_results
   mkdir demo_seq_to_demo_seq2
   mkdir demo_seq_to_demo_seq2/align
   touch demo_seq_to_demo_seq2/align/all.done
   cd ../..
   ext/mummer/mac/promer data/seq/demo_seq2/sequence/chr1.seq data/seq/demo_seq/sequence/chr3.seq
   ext/mummer/mac/show-coords -dlkTH out.delta >chr1chr3.mum
   mv chr1chr3.mum data/seq_results/demo_seq_to_demo_seq2/align
   rm out.delta
</pre>
This would need to be repeated 3 more times for <tt>chr1chr5.mum, chr3chr3.mum, chr3chr5.mum</tt>
(suggestion: put the commands
in a script). The resulting <tt>align</tt> directory will have:
<pre>
   all.done	 chr1chr3.mum	chr1chr5.mum	chr3chr3.mum	chr3chr5.mum
</pre>
When symap is started and <tt>demo_seq</tt> and <tt>demo_seq2</tt> selected, there will be a "A" in their cell; select it followed by
<ttb>Selected Pair</ttb> and it will load the alignments and compute the synteny.
<p>

<a id="help"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Getting Help</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

If none of these suggestions fix your problem, email symap@agcol.arizona.edu with the following files
(refer to the files discussed in <a href="#log">Finding the problem</a>):
<ol>
<li><tt>error.log</tt>
<li><tt>logs/&lt;project-name1&gt;-to-&lt;project-name2&gt;/symap.log</tt>
<li>The <tt>_fn.log</tt> file from the alignment that failed, where n is the alignment number
and the prefix before the "_" is <tt>&lt;project-name1&gt;_cc.&lt;project-name2&gt;</tt>.
<li>Any output to the terminal (either copy and paste into the email, or send a screen capture)
</ol>

<a id="mac"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>MacOS details</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<h4>Executable on MacOS</h4>
Any executable that has not been okayed by Apple get the error message
<pre>
cannot be opened because the developer cannot be verified
</pre>
Hence, you need
to use the Finder to enter all <tt>/ext/&lt;program&gt;/mac</tt> directories,
select each program, then choose to Open it; it will run the program in a terminal
window and be saved as a program that can thereafter be run.

<h4>Early versions of MacOS</h4>
The MacOS executables were compiled on a MacOS 10.15. They will not work on old versions such as MacOS 10.9.
Try the following:
<pre>
cd ext/mummer
mv mac mac_506
mv mac_pre506 mac
cd ../muscle
mv muscle muscle_506
mv muscle_pre506 muscle
</pre>
These executables compiled on MacOS 10.9 may work on your Mac.
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

