<!DOCTYPE html>
<html>
<head>
<title>SyMAP MUMmer</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="https://csoderlund.github.io">
		<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;MUMmer</p>
<p>
<a href="#mfiles">MUMmer files</a>
<a href="#log">Finding the problem</a>
<a href="#mem">Out of Memory</a>
<a href="#cmd">MUMmer from the command line</a>
<a href="#help">Getting Help</a>
<p>&nbsp;
<a href="#top">Go to top</a>
</div>

<!------ right side header and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav"> <!-- header same in all pages -->
    <br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="./Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="./SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">
<p>This document discusses how to deal with one or more MUMmer alignments failing.
The corresponding MUMmer documentation is
<a href="http://mummer.sourceforge.net" class="ext" target="_blank">v3</a> and
<a href="https://mummer4.github.io" class="ext" target="_blank">v4</a>.
<p>
<table class="center">

<tr><td style="vertical-align: text-top;" class="tx">
	<table>
	<tr><td><a href="#mfiles">MUMmer files</a>
	<tr><td><a href="#log">Finding the problem</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#mum4">Using MUMmer4</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#ext">Executables</a>
	</table>
<td>&nbsp;&nbsp;&nbsp;
<td style="vertical-align: text-top;" class="tx">
	<table>
	<tr><td><a href="#mem">Out of Memory</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#cat">Limit CPUs and uncheck Concat</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#mask">Not-masked or Soft-masked</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#fail2">One or more alignments fail</a>
	</table>
<td>&nbsp;&nbsp;&nbsp;
<td style="vertical-align: text-top;">
	<table>
	<tr><td><a href="#cmd">MUMmer from the command line</a>
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#example">Example</a>
	<tr><td><a href="#mum">MUMmer parameters</a>
	<tr><td><tr><td><a href="#help">Getting Help</a>
	</table>
</table>

<!-------------------------------------------------------------->
<a id="mfiles"></a>
<h2>MUMmer files</h2>

When <ttp>symap</ttp> executes MUMmer, the resulting alignment files are in:
<pre>
   /data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/align
</pre>
For example,
<pre>
   data/seq_results/demo_seq_to_demo_seq2/align> ls
   all.done				demo_seq_cc.demo_seq2_f2.mum
   demo_seq_cc.demo_seq2_f1.mum		demo_seq_cc.demo_seq2_f2.mum.done
   demo_seq_cc.demo_seq2_f1.mum.done
</pre>

<p>All MUMmer files but those with the <tt>.mum</tt> suffix are removed by <ttp>symap</ttp>. If you prefer them not to be removed,
use the "-mum" command line parameter, i.e.
<pre> ./symap -mum</pre>

<p>The log files are in the <tt>/logs/&lt;project-name1&gt;-to-&lt;project-name2&gt</tt> directory.

<!------------------------------------------------------------->
<a id="log"></a>
<h2>Finding the problem</h2>

<table class="tz" style="width:70%"><tr>
<td><a href="#mum4">Using MUMmer4</a>
<td><a href="#ext">Executables</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<p>If the MUMmer alignment fails:
<ul>
<li>Check that the sequences loaded correctly using the <ttb>View</ttb> option on the <ttp>Manager</ttp>.
<li>If the files loaded correctly, check the <a href="#ext">Executables</a>.
<li>If the executables are okay, inspect the log files.
</ul>
<h3>Log files</h3>
<p>Using <tt>p1</tt> = <tt>project-name1</tt> and <tt>p2</tt> = <tt>project-name2</tt>

<pre>
   symap_5/
     error.log   # a SyMAP error will write its trace data into this file and list failed MUMmer

     logs/
       &lt;p1&gt;-to-&lt;p2&gt;/         # one directory per project-to-project alignment
          &lt;p1_cc.p2_f1&gt;.log  # MUMmer terminal output - one file per MUMmer process
          &lt;p1_cc.p2_f2&gt;.log  #    fn is n=1,2... for number of processes, e.g f2 is 2nd process
          symap.log          # keeps most of the SyMAP output shown on the terminal for this A&S
</pre>

e.g. <tt>p1</tt> = <tt>demo_seq</tt> and <tt>p2</tt> = <tt>demo_seq2</tt>
<pre>
       demo_seq_to_demo_seq2/
         demo_seq_cc.demo_seq2_f1.log  # MUMmer output directed to this file for process 1
         demo_seq_cc.demo_seq2_f2.log  # MUMmer output directed to this file for process 2
         symap.log                     # SyMAP terminal output
</pre>

&rarr; If an alignment is listed as failed in the <tt>error.log</tt> file,
the corresponding <tt>&lt;p1_cc.p2_fn&gt;.log</tt> file will contain the MUMmer error
(see <a href="img/mummerLog.png">example mummer log</a>).
<p>&rarr;
If the error is not found in the log files or the error is not clear, try the following:
Copy the command from the terminal (or log file), and paste to a terminal to execute.
The command will look similar to the following, though it will all be on one line:
<p>
<pre>
  > ext/mummer/mac/promer
    -p data/seq_results/demo_seq_to_demo_seq2/align/demo_seq_cc.demo_seq2_f2.promer
     data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq2/demo_seq2_f2.fa
     data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq/demo_seq_cc.fa
</pre>
This shows <tt>promer</tt> output directly on the terminal and provides a better description of the problem.

<!----------------------------------------------------->
<a id="mum4"></a>
<h3>Using MUMmer4</h3>

Sometimes when MUMmer v3 fails, MUMmer v4 will work.
MUMmer4 is included in SyMAP package with a fix to <tt>promer</tt>.
Enter the <tt>ext/mummer4</tt> directory and follow the instructions in the README.

<!------------------------------------------------------->
<a id="ext"></a>
<h3>Executables</h3>

The alignment programs are provided in the <tt>symap/ext</tt> directory. There are
64-bit executables for Linux, MacOS x86_64, and MacOS M4. SyMAP will
select the correct directory for the machine you are running from, i.e. you
do not need to do anything. See <a href="SystemGuide.html#ext" class="ext" target="_blank">System Guide</a> for details.

<p>When SyMAP creates a database, it (1) checks the MySQL variables, and (2) checks
that the external programs are executable. If you see a message like:
<pre>
  ***Error - file is not executable: ext/mummer/mac/promer
</pre>
Execute:
<pre>
  > chmod 755 ext/mummer/mac/promer
</pre>
Execute the program from the command line to make sure it works on you machine, e.g.
<pre>
  >./ext/mummer/mac/promer

    USAGE: promer  [options]  &lt;Reference&gt;  &lt;Query&gt;

    Try './ext/mummer/mac/promer -h' for more information.
</pre>
The above shows that the promer code will execute on my MacOS.

<p><font color="red"><u>Executable on MacOS</u></font>:
Any executable that has not been okayed by Apple results in the error message
<pre>
  Apple could not verify "prepro" is free of malware that may harm your mac...
</pre>

See <a href="TroubleShoot.html#macos" class="ext" target="_blank">MacOS External</a> for the fix.

<!-------------------------------------------------->
<a id="mem"></a>
<h2>Out of memory</h2>

<table class="tz"><tr>
<td><a href="#cat">Limit CPUs and<br>uncheck Concat</a>
<td><a href="#mask">Not-masked or<br>Soft-masked</a>
<td><a href="#fail2">One or more<br>alignments fail</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>A MUMmer failure is typically from insufficient memory.

<p>If an alignment fails immediately, and if the last line of the first <tt>&lt;alignment&gt;.log</tt> file is:
<pre>
  1: PREPARING DATA
</pre>
the reason is probably that your machine does not have near enough memory as MUMmer could not even
prepare the data.

<p>Additionally, the following errors typically indicates a memory problem:
<pre>
  Alignment program error code: 141
  20220512|075853|6007| ERROR: mummer and/or mgaps returned non-zero, please file a bug report
</pre>
or
<pre>
#..........................ERROR: mummer and/or mgaps returned non-zero, please file a bug report
Alignment program error code: 1
</pre>
The error code will appear on the terminal and the MUMmer log file,
but not in <tt>symap/logs/&lt;..&gt;/symap.log</tt>.

<!----------------------------------------------------->
<a id="cat"></a>
<h3>Limit CPUs and uncheck Concat</h3>

There is no straight-forward way to know if you have enough memory as it depends on the
size and complexity of the two genomes being compared.
If you think memory may be tight or MUMmer produced an error as shown in the above section,
first try running again with reduced <ttl>CPUs</ttl> and unchecked <ttl>Concat</ttl>:
<ol>
<li>On the <ttb>Project Manager</ttb> panel, limit the number of CPUs, as each
CPU uses a considerable amount of memory (e.g. 4 CPUs could collectively use 20GB of memory at once).
See <a href="SystemHelp.html#cpu" class="ext" target="_blank">CPU</a> for further information.
<p><li>
In the <ttb>Project Parameters</ttb> panel, uncheck <ttl>Concat</ttl> to reduce the
size of the input files to MUMmer.
See <a href="SystemHelp.html#concat" class="ext" target="_blank">Concat</a> for a description of
this option.
</ol>
<!----------------------------------------------------->
<a id="mask"></a>
<h3>Not-masked or Soft-masked</h3>
A memory problem can occur if the genome sequence is not masked or only soft-masked.
Try either of the following:
<ul><li>Change the sequence to hard-masked (see <a href="input/index.html#seqprep" class="ext" target="_blank">Sequence files</a>).
<li>Set the <a href="SystemHelp.html#align2" class="ext" target="_blank">Pair Parameters</a> parameter to <ttx>Mask</ttx>
all sequence but the genes.
</ul>

<!----------------------------------------------------->
<a id="fail2"></a>
<h3>One or more fails</h3>

Sometimes just one or a few of the alignment processes will fail. You will see a line such as:
<pre>
  Error: Running command: /Users/cari/Workspace/symap_5/ext/mummer/mac/promer
   -p data/seq_results/demo_seq_to_demo_seq2/align/demo_seq_cc.demo_seq2_f2.promer
   data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq2/demo_seq2_f2.fa
   data/seq_results/demo_seq_to_demo_seq2/tmp/demo_seq/demo_seq_cc.fa
</pre>

<table>
<tr>
<td><a href="img/runDialog.png"><img src="img/runDialog.png" alt="runDialog" style="border: 1px solid black; width: 420px"></a>
<td>
You will see the failure on the dialog box as is
shown on the left. The remaining processes will continue. When all processes are complete,
you will see a "?" for the pair as shown on the below.
<p>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="img/runResults.png"><img src="img/runResults.png" alt="runResults" style="border: 1px solid black; width: 220px"></a>
<p>Select the "?" followed by
<ttb>Selected Pair</ttb> and it will complete the failed processes.
</table>

<p><u>Incomplete alignment</u>:
If SyMAP completed the alignment, e.g. the demo <tt>/align</tt> directory will have the following files:
<p><pre>
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:28 all.done
-rw-r--r--@ 1 cari  staff   1.2M Apr 10 10:28 demo_seq_cc.demo_seq2_f1.mum
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:28 demo_seq_cc.demo_seq2_f1.mum.done
-rw-r--r--@ 1 cari  staff   541K Apr 10 10:27 demo_seq_cc.demo_seq2_f2.mum
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:27 demo_seq_cc.demo_seq2_f2.mum.done
</pre>
<ol>
<li>The <tt>all.done</tt> indicates that the all alignments completed.
<li>If <tt>all.done</tt> does not exist, SyMAP will perform any alignments that do NOT have a corresponding <tt>mum.done</tt>.
If the <ttl>Concat</ttl> setting has been switched between the previous and current run, this will not work correctly.
<li>If the user supplied the alignments (<a href="MUMmer.html#cmd">Supply MUMmer files</a>),
there may not be <tt>mum.done</tt> files but there should be an <tt>all.done</tt>, so it is assumed they are all done.
</ol>
<!------------------------------------------------------------------>
<a id="cmd"></a>
<h2>Running MUMmer from the command line</h2>
<table class="tz" style="width:40%"><tr>
<td><a href="#cmd">Command line</a>
<td><a href="#example">Example</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<h3>Command line</h3>

If you need to run MUMmer using the terminal command line from some other machine, do the following:


<ol>
<li><u>The naming of your files and their order into MUMmer is VERY important.</u> Everything will mess up
if this is done wrong!
<ul>
<li>It uses the <ttx>Directory</ttx> name of the two projects in the <tt>data/seq</tt> directory.
<li>Say the projects are in <tt>data/seq/arab</tt> and <tt>data/seq/brap</tt>.
<li>The directory name <tt>arab</tt>
is <i>alphanumerically less</i> than <tt>brap</tt>, so it is <tt>proj1</tt> and <tt>brap</tt> is <tt>proj2</tt>.
</ul>
<p>
<li>If your chromosomes are large, split the sequence file into chromosome files using <a href="input/index.html#sp" class="ext" target="_blank">xToSymap</a>.
Process each chromosome file for the first genome with each chromosome file for the second genome.
<p>
<li>Create a directory <tt>data/seq_results/&lt;proj1-to-proj2&gt;/align</tt>.
<p>
<li>Running MUMmer: The query must be alphanumerically less than reference. For the <tt>promer</tt> command:
<pre>
USAGE: promer  [options]  &lt;Reference&gt;  &lt;Query&gt;
e.g.   promer proj2 proj1
</pre>
<p>
See the commands for the <b>Example</b> below. Replace the demo names with your project/chromosome names.
If you have many chromosome pairs to process,
put them in a script, e.g. <a href="input/logs/mummer.html">demo script</a>.
Then execute the script from the main symap directory.

<p>To view the MUMmer parameters, see <a href="SystemHelp.html#mum" class="ext" target="_blank">MUMmer parameters</a>.
	<ul>
	<li>Promer: The output of promer is input to "show-coords -dlkT".
	<li>Nucmer: The output of nucmer is input to "show-coords -dlT".
	</ul>

<p>
<li>Result files:
<ul>
<li>The result files must have suffix ".mum"
<p>
<li>The ".mum" files must be in the directory <tt>data/seq_results/&lt;proj1-to-proj2&gt;/align</tt>.
<p>
<li>In the <tt>&lt;proj1-to-proj2&gt;/align</tt> directory, execute:
<tt>touch all.done</tt>
<br>This creates a file, which indicates to SyMAP that the alignments are done and to process
the files in the directory ending with ".mum".
</ul>
<br>
<li>When you run <ttb>Selected Pair</ttb>, SyMAP will recognize the files and use them to build the synteny blocks.
</ol>

<a id="example"></a>
<h3>Example</h3>
This example will use MUMmer for the loaded projects <tt>demo_seq</tt> and <tt>demo_seq2</tt>.
<br>If you want to try this and you have already run the demo, remove the synteny results
from the database using the <ttp>symap</ttp>
<ttb>Clear Pair</ttb> along with the directory <tt>data/seq_results/demo_seq_to_demo_seq2</tt>.

<p>Input:
<ul>
<li>Project <tt>demo_seq</tt> has file <tt>genomic.fa</tt> (containing 2 chromosomes).
<li>Project <tt>demo_seq2</tt> has files <tt>chr1.seq.gz</tt> and <tt>chr3.seq</tt>.
<li>Note that <tt>demo_seq</tt> is alphanumerically less than <tt>demo_seq2</tt>.
</ul>
<p>The commands are as follows:
<p>
<pre>
   gunzip data/seq/demo_seq2/sequence/chr1.seq.gz  # MUMmer does not process zipped files
   cd data/seq_results
   mkdir demo_seq_to_demo_seq2
   mkdir demo_seq_to_demo_seq2/align
   touch demo_seq_to_demo_seq2/align/all.done

   # from symap directory, execute the following (the indented line are part of the previous line)
   ext/mummer/mac/promer  -p data/seq_results/demo_seq_to_demo_seq2/align/results.promer
         data/seq/demo_seq2/sequence/chr1.seq data/seq/demo_seq/sequence/genomic.fa
   ext/mummer/mac/show-coords
        -dlkTH data/seq_results/demo_seq_to_demo_seq2/align/results.promer.delta
         &gt;data/seq_results/demo_seq_to_demo_seq2/align/seq1chr1.mum
</pre>
See <a href="input/logs/mummer.html">script</a> for the full set of MUMmer commands to process all sequence data.
<br>It is easiest to run the demo by copying the contents of the script
into a file (e.g. <tt>./symap/demo.txt</tt>),
<br>then from the <tt>symap</tt> directory, execute it (e.g. <tt>sh demo.txt</tt>).

<p>When <tt>./symap</tt> is started, select <tt>demo_seq</tt> and <tt>demo_seq2</tt>.
There will be a "A" in their cell; select it followed by
<ttb>Selected Pair</ttb> and it will load the alignments and compute the synteny.

<p>This demo has been fully tested with <ttp>symap</ttp> v5.7.9.

<!------------------------------------------------------------------------>
<a id="mum"></a>
<h2>MUMmer parameters</h2>

When setting parameters for MUMmer in
<a href="SystemHelp.html#over2" target="_blank" class="ext">pairs parameters</a>,
they are <b>NOT</b> checked for correctness.
<p>
To see the parameters for the default MUMmer V3 on MacOS, from the <tt>symap</tt> directory:
<pre>
./ext/mummer/mac/mummer -h
./ext/mummer/mac/promer -h
./ext/mummer/mac/nucmer -h
</pre>
To see the parameters for the default MUMmer V3 on Linux:
<pre>
./ext/mummer/linux/mummer -h
./ext/mummer/linux/promer -h
./ext/mummer/linux/nucmer -h
</pre>

If  you compiled V4 in the /ext directory:
<pre>
./ext/mummer4/m4/bin/mummer -h
./ext/mummer4/m4/bin/promer -h
./ext/mummer4/m4/bin/nucmer -h
</pre>
<p>After running MUMmer, all alignment files are removed except the ".mum" file; to prevent removal,
<br> execute
<tt>./symap -mum</tt>.
<!----------------------------------------------------->
<a id="help"></a>
<h2>Getting Help</h2>

If none of these suggestions fix your problem,
email <a href="mailto:cas1@arizona.edu">cas1@arizona.edu</a>
with the following files (described in <a href="#mfiles">MUMmer files</a>):
<ol>
<li><tt>error.log</tt>
<li><tt>logs/&lt;p1&gt;-to-&lt;p2&gt;/symap.log</tt>
<li><tt>logs/&lt;p1&gt;-to-&lt;p2&gt;/&lt;p1_cc.p2.fn&gt;.log</tt> where n is the process number
<li>Any output to the terminal (either copy and paste into the email, or send a screen capture)
</ol>
For example, email the terminal output and:
<pre>
  symap/error.log
  symap/logs/demo_seq_to_demo_seq2/symap.log
  symap/logs/demo_seq_to_demo_seq2/demo_seq_cc.demo_seq2_f2.log
</pre>
<!---- END BODY -->
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p class="trailer">  Email Comments To: <a href="mailto:cas1@arizona.edu">cas1@arizona.edu</a></p>
</div>
</div>
</body>
</html>
