<!DOCTYPE html>
<html>
<head>
<title>SyMAP Queries</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
    a.ext {font-weight: bold;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; text-align: left}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
    ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: Cambria, monospace; font-size: 14px; font-weight: 500;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Queries
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>

				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<b>Content:</b>

<table>
<tr>
<td><a href="#face"><u>&gt;Instructions</u></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#query"><u>&gt;Query Setup</u></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#result"><u>Query Result Panel</u></a>&nbsp;&nbsp;
<td><a href="#rtab"><u>&gt;Results</u></a>

<tr>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#gen">General</a>
<td><a href="#tab">Result table</a>&nbsp;&nbsp;&nbsp;&nbsp;
<td>&nbsp;

<tr>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#sin">Single gene</a>
<td><a href="#col">Columns</a>&nbsp;&nbsp;&nbsp;&nbsp;
<td>&nbsp;

<tr>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#pair">Pair hits</a>
<td>&nbsp;&nbsp;&nbsp;
	<a href="#pairCol">Pair</a>,
	<a href="#singleCol">Single</a>,
	<a href="#save">Auto-save</a>&nbsp;&nbsp;&nbsp;&nbsp;
<td>&nbsp;&nbsp;&nbsp;

<tr>
<td>&nbsp;&nbsp;&nbsp;
<td><a href="#pfs">Gene Groups</a>
<td><a href="#but">Top button</a>
<td>&nbsp;&nbsp;&nbsp;

<tr>
<td>&nbsp;&nbsp;&nbsp;
<td>&nbsp;
<td><a href="#stat">Statistics</a>
<td>&nbsp;&nbsp;&nbsp;

</table>
<p>See <a href="./UserGuide.html#term" class="ext" target="_blank">Terminology</a> in the User Guide,
especially the <ttx>Cluster Hit Algo1 versus Algo2</ttx> description.

<hr>
<a id="face"></a>
<h2>&gt;Instructions</h2>
To open the query interface, first select two or more sequence projects in the <a href="UserGuide.html#mgr">Project Manager</a>.
Then select the <ttb>Queries</ttb> button to open the queries interface.

<p>
<a href="img/query_oview.png"><img src="img/query_oview.png" alt="query_oview" style="border: 1px solid black; width: 500px"></a>
<p>
The <ttl>Instructions</ttl> window (above) lists the projects which were selected for querying. The last
line states what the <ttx>Olap</ttx> column represents; if one or more project pairs used <ttx>Algo1</ttx>, it will be gene overlap;
if all project pairs used <ttx>Algo2</ttx>, it will be exon overlap

<p>Open the <ttl>Query Setup</ttl> window by clicking on its tab in the left panel.

<a id="query"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>&gt;Query Setup</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b>Sections:</b>
<a href="#gen">1. General</a>&nbsp;&nbsp;&nbsp;
<a href="#sin">2. Single genes</a>&nbsp;&nbsp;&nbsp;
<a href="#pair">3. Pair Hits</a>&nbsp;&nbsp;&nbsp;
<a href="#pfs">4. Gene Groups</a>

<p>Set up the desired filters and then select <ttb>Run Search</ttb> to execute it.
When the query is complete, the <a href="#result">query result panel</a> will be displayed.
<p>
<a id="qresults"></a>
<a href="img/query_query.png"><img src="img/query_query.png" alt="query_query" style="border: 1px solid black; width: 550px"></a>

<a id="rules"></a>
<p><b>Rules</b>

<table>
<tr><td style="vertical-align: text-top;">
1. All queries are on the <ttl>Pair hits</ttl> unless <ttl>Single</ttl> is selected.

<p>2. All <ttl>Pair hits</ttl> queries return a row per <i>hit pair</i>, i.e.
aligned region between to two chromosomes from two projects.

<td><a href="img/query_table_rows.png"><img src="img/query_table_rows.png" alt="query_table_rows" style="border: 1px solid black; width: 450px"></a>
</table>

<table>
<tr><td style="vertical-align: text-top;">
3. All <ttx>Hit#</ttx> are uniquely numbered for a chromosome pair, e.g. there will be a hit #3 on all chromosome pairs
that have at least 3 hits.
All <ttx>Gene#</ttx> are sequentially numbered per chromosome, e.g. there will be a
gene #3 on all chromosomes that have at least 3 non-overlapping genes.

<p>4. <ttx>Major</ttx> and <ttx>Minor</ttx> genes:
<br>a. A hit can align to multiple genes on either or both chromosomes; it is assigned to the best (<ttx>major</ttx>) gene on each end.
<br>b. For example, in the image on the right, the burgundy gene is the major gene as the hit (green line)
fits to it best,
whereas the blue single exon gene is the minor gene.
<br>c. When queries involve genes, only the hits to the major genes will be shown
except for the following queries, which will also list the minor genes:
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<ttx>Every*</ttx>, <ttx>Hit#</ttx>,
<ttx>Gene#</ttx> when no suffix is provided, and <ttx>Multi</ttx> with <ttx>minor</ttx> checked.

<p>5. Most filters can be used in conjunction with other
filters; options will be disabled if they cannot be used with a selected filter.

<td>
<a href="img/queryMultOlap.png">
<img src="img/queryMultOlap.png" alt="Hit overlaps two genes" style="border: 1px solid black; width: 35px"></a>
</table>


<a id="gen"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>1. General</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr>
<td><ttl>Annotation Description</ttl>
<br>Enter a substring: the entire annotation string (i.e. column <ttl>All_Anno</ttl>) will be searched
for the substring. Hits will be returned that align to the genes with the corresponding annotation.
<td><a href="img/query_gen.png">
<img src="img/query_gen.png" alt="General" style="border: 1px solid black; width: 400px"></a>
</table>


<p><ttl>Location</ttl>
<table class="tx">
<tr><td style="vertical-align: text-top;"><ttl>Chr</ttl><td>&nbsp;
	<td>Select a specific chromosome for the species. This is the most common query to use in
	conjunction with others.
<tr><td style="vertical-align: text-top;"><ttl>From</ttl><td>&nbsp;
	<td>The <ttx>Start</ttx> coordinates for the selected chromosome will be &gt;= this number.
<tr><td style="vertical-align: text-top;"><ttl>To</ttl><td>&nbsp;
	<td>The <ttx>End</ttx> coordinates for the selected chromosome will be &lt;= this number.
</table>
<br>
For <ttx>Pair hits</ttx>, the <i>hit</i> coordinates must pass the <ttl>From</ttl> and <ttl>To</ttl> input integers.
<br>For <ttx>Single genes</ttx>, the <i>gene</i> coordinates must pass the <ttl>From</ttl> and <ttl>To</ttl> input integers.
<p>It is valid to only enter the <ttl>From</ttl> or <ttl>To</ttl>, or no value for either (leave blank).

<!--------------------------------------------------->
<a id="sin"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2. Single gene</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<table><tr>
<td>The <ttl>Single</ttl> queries produces rows of genes; there is no hit or block
information since the rows do not represent hits.
<td><a href="img/query_single.png">
<img src="img/query_single.png" alt="Single" style="border: 1px solid black; width: 400px"></a>
</table>
Options:
<ul>
<li><ttl>Orphan genes (no hits)</ttl>
<br>
Genes that do not a have a hit and meet the additional filters.
The orphan genes are relative to the projects shown on the <ttl>Instruction</ttl> page.
For example, if species X, Y and Z have synteny computed between all pairs,
but only X and Y are selected, the orphan genes for X would be those with no hits to Y.
If X,Y&Z are selected, the orphan genes for X would be those with no hits to Y and Z.
<br>&nbsp;
<li><ttl>All genes (w/o hits)</ttl>, i.e. genes with and without hits
<br>This shows all genes that meet the additional filters, regardless if they have a hit or not.
There is always the same set of genes for a project, regardless of synteny.
</ul>

<p><u>Allowed filters for singles</u>: both options can be restricted by:
<ul>
<li>Using the <ttl>Project</ttl> pull-down, select a project.
<li>If a project is selected, then a chromosome and optional location can be selected.
<li>An <ttl>Annotation Description</ttl> may be entered.
</ul>
<!--------------------------------------------------->
<a id="pair"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3. Pair hits</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table><tr><td>
Each hit connects two <i>species (projects)</i> and hence represents a <i>pair</i> of aligned
regions for two of the selected species. Filters are as follows:
<td><a href="img/query_pair.png">
<img src="img/query_pair.png" alt="Pair" style="border: 1px solid black; width: 400px"></a>
</table>

<p><ttl>In Block (Synteny Hit)</ttl>
<table class="tx">
<tr><td  style="vertical-align: text-top;">Yes<td><td>&nbsp;<td>Only hits that are part of a synteny
block will be returned. All hits will have a value for the <ttx>Block</ttx> column.
<tr><td style="vertical-align: text-top;">No<td><td>&nbsp;<td>Only hits that are NOT part of a synteny
block will be returned. No hits will have a value for the <ttx>Block</ttx> column.
</table>

<a id="genehit"></a>
<p><ttl>Annotated (Gene Hit)</ttl>
<table class="tx">
<tr><td style="vertical-align: text-top;">Every<td><td>&nbsp;
<td>Only hits that align to a gene on one or both sides of the hit will be shown.
The <ttx>Gene#</ttx> columns will list the best gene.
<tr><td style="vertical-align: text-top;">Every*<td><td>&nbsp;
<td>This is like the <ttx>Every</ttx> option, but a hit may be listed multiple times if it aligns
to overlapping genes; in other words, ALL genes will be shown if they have any alignment.
The <ttx>Gene#</ttx> for non-best hits will be suffixed with an "*" and the <ttx>Htype</ttx> is
unknown.

<tr><td style="vertical-align: text-top;">One<td><td>&nbsp;
<td>Only hits that align to a gene on ONE end will be shown.
The <ttx>Gene#</ttx> column will show the best gene for the end with the hit.

<tr><td style="vertical-align: text-top;">Both<td><td>&nbsp;
<td>Only hits that align to genes on BOTH end will be shown.
The <ttx>Gene#</ttx> column will show the best gene for both ends of the hit.

<tr><td style="vertical-align: text-top;">None<td><td>&nbsp;<td>Only hits that do NOT align to a gene
on either end will be shown. The <ttx>Gene#</ttx> column will not have a value for either end of the hit.
</table>
<u>Location</u>: if a chromosome is selected, every hit row must be on the selected chromosome, but the "every" or "one"
gene may be on the opposite chromosome.

<p><ttl>Collinear size</ttl>
<table class="tx">
<tr><td style="white-space: nowrap;vertical-align: text-top;"><td>&gt;= [=, &gt;=] N<td>&nbsp;
<td>List all hits in collinear sets that have size &gt;= N or  = N or &lt;= N, respectively.
<br>The text box must have an integer &gt;0;

<tr><td style="vertical-align: text-top;"><td>Ignore<td>&nbsp;<td>Do not filter on collinear set sizes.
</table>
<br>
See <a href="Collinear.html">Collinear</a>, which explains the SyMAP collinear sets.

<p><ttl>Block#</ttl>
<br>Enter a single block number (<u>the <ttx>Block</ttx> column is formatted
Chr.Chr.Block#</u>). Use it in conjunction with the chromosome pull-downs. For example,
if you select Chr 1 from the first project, Chr 2 from a second project, and enter block=3, you will see
hits in block 1.2.3 from the two project chromosomes you selected.

<p><ttl>Collinear Set#</ttl>
<br>Enter a set number (<u>the <ttx>Collinear</ttx> column is formatted Chr.Chr.Size.Set#</u>). All hits in a collinear set have the same Set#, which is the last number of the <ttx>Collinear</ttx> column.
This can be used in conjunction with the chromosome pull-downs (as for blocks).

<p><ttl>Hit#</ttl>
<br>The hits are numbered for each chromosome pair, e.g. there will be a Hit#1 for every chromosome pair that
has hits. If a hit aligns to multiple genes, both major and minor genes will be shown.

<a id="genenum"></a>
<p><ttl>Gene#</ttl>
<br>Enter a <ttx>Gene#</ttx> number (<u>the <ttx>Gene#</ttx> column is formatted Chr.Gene#.suffix</u>
where the suffix may be blank); <i>do not include the 'Chr' number</i>.
<br>&nbsp;&bull; If only a number is entered, all genes with the numeric prefix will be displayed (including minor hits).
<br>&nbsp;&bull; If a number.suffix is entered, the exact gene will be displayed (but no minor hits).
<br>This query results in values for the <ttx>Grp#</ttx> and <ttx>GrpSize</ttx> columns; this in turn
allows the <ttl>View 2D</ttl> <a href="#group">Group</a> option to be used.

<!--------------------------------------------------->
<a id="pfs"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Gene Groups</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
These two options are computed on the fly.
They produce query results with values for the <ttx>Grp#</ttx> and <ttx>GrpSize</ttx> columns; this in turn
allows the <ttl>View 2D</ttl> <a href="#group">Group</a> option to be used.

<a id="multi"></a>
<h4>4.1 <ttl>Multi-hit Genes&gt;= N</ttl></h4>
<a href="img/query_multi.png">
<img src="img/query_multi.png" alt="Multi" style="border: 1px solid black; width: 450px"></a>

<p>List all hits genes that have &gt;= N hits, where multiple hits can join the same two genes.
<br>The <u>target</u> gene refers to the gene with &gt;= N hits to the <u>opposite</u> species.
<br>The options are as follows:
<p>
<table style="tx">
<tr><td>&nbsp;<td><ttl>Exon</ttl>
	<td>&nbsp;<td>(<tt>Algo2</tt> only) The hits must be to exons in the target gene.
<tr><td>&nbsp;<td><ttl>Minor*</ttl>
	<td>&nbsp;<td>Include minor hits on either chromosome.
<tr><tr><td colspan=2><ttl>Opposite</ttl>
<tr><td>&nbsp;<td style="white-space: nowrap; vertical-align: text-top;"><ttl>Same Chr</ttl>
	<td>&nbsp;<td>The &gt;=N hits must all be on the same opposite chromosome.
<tr><td>&nbsp;<td style="vertical-align: text-top;"><ttl>Tandem</ttl>
    <td>&nbsp;<td>(Annotated species only): The &gt;=N hits must be to a tandem array of genes.
	<br><ttl>Same Chr</ttl> is automatic with this option.
</table>

<p>Good associated filters:
<table style="tx">
<tr><td>&nbsp;<td style="white-space: nowrap;vertical-align: text-top;"><ttl>Location <ttx>Chr:</ttx><td>&nbsp;
     <td>When used with a specified chromosome,
     restricts the target genes with &gt;= N hits to the specified chromosome.
     For example:
     <ul>
     <li>If species<sup>1</sup> Chr02 is selected, there will only be gene
     groups shown from Chr02 to any chromosome in species<sup>2</sup>.
     <li>If both species<sup>1</sup> Chr02
     and species<sup>2</sup> Chr03 are selected, all gene groups from
     species<sup>1</sup> Chr02 to species<sup>2</sup> Chr03 will be shown, and
     vice versa.
     </ul>
<tr><td>&nbsp;<td style="white-space: nowrap;vertical-align: text-top;"><ttl>Annotated</ttl> <ttx>Both</ttx>
    <td>&nbsp;<td>Each target gene must have its N hits to genes on the opposite species (not needed with the
    <ttl>Tandem</ttl> option).
</table>

<p>Table can list the same hit multiple times:
<table>
<tr><td style="vertical-align: text-top;">For all other queries, a hit will only be listed once for a chromosome pair; however, that is not
the case for this query. That is because gene X and gene Y may be connected by a hit, where both gene X
and gene Y have &gt;=N, so both groups need to be shown.
<p>An example is shown on the right, where the
two highlighted rows are the same, but in two different groups (grp#1 for gene 5.1458 and grp#4 for gene 3.638).
<td><a href="img/query_multi_results.png">
<img src="img/query_multi_results.png" alt="multi-hit results" style="border: 1px solid black; width: 250px"></a>
</table>

<a id="group_ex"></a>
<u>View 2D Group Example:</u>
The image below shows the two groups, where the group hits are highlighted in magenta; these were
produced using the <ttl>View 2D</ttl> option <a href="#group">Group</a>.
<br><a href="img/query_multi_2D.png">
<img src="img/query_multi_2D.png" alt="multi-hit 2D" style="border: 1px solid black; width: 700px"></a>

<h4>4.2 <ttl>PgeneF (putative gene families)</ttl></h4>
<i>Note</i>: this has not been tested for a long time except superficially (i.e. making sure it has the same results as the previous
release).

<table><tr><td>
Using the hits that pass the other filters, SyMAP constructs putative gene families (PgeneFs)
spanning the selected species. This is done by grouping hits which overlap on at
least one genome.
<td><a href="img/query_pgenef1.png">
<img src="img/query_pgenef1.png" alt="PgeneF two" style="border: 1px solid black; width: 250px"></a>
<tr>
<td>Additional options are provided when &gt;2 species are selected.
<i>Note, if you have more than 6 species selected, this stage can take
an hour or more.</i>
<td><a href="img/query_pgenef.png">
<img src="img/query_pgenef.png" alt="PgeneF" style="border: 1px solid black; width: 250px"></a>
</table>

<p>
Each PgeneF is given a number, which is shown in the Query Results table (column name <ttx>PgeneF</ttx>).
The size of the PgeneF is also shown (column <ttx>PgFSize</ttx>).
<p>
Filters using the PgeneF values:

<p><ttl>Include/Exclude</ttl>
<br>
These filters permit searching for gene families shared by one group of species but not
present in another.
<p>
If a species is checked to include, then the PgeneF will only be retained if it includes
at least one hit which hits that species.
<p>
For &gt;2 species only: If a species is checked to exclude,
then the PgeneF will be discarded if any of its hits are to that species.


<p><ttl>For the included species:</ttl>

<table class="tx">

<tr><td style="vertical-align: text-top;"><ttl>No annotation to the included species</ttl>
<td>Find PgeneFs which are not yet annotated. A PgeneF will be discarded if it is annotated on any of
the species which are checked in the <ttl>Include</ttl> line.

<tr><td style="vertical-align: text-top;"><ttl>Complete linkage if included species</ttl>
<td>For &gt;2 species only: Require the PgeneF to be fully linked, i.e. for each pair of species A and B in the group,
there must be a hit linking A to B.

<tr><td style="vertical-align: text-top;"><ttl>At least one hit for each included species</ttl>
<td>For &gt;2 species only: Only PgeneF hits will be shown if they have hits to the included species,
although the PgeneF numbers will reflect groupings created using all hits.
</table>
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<a id="result"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Query Result Panel</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b>Sections:</b>
<a href="#tab">1. Results table</a>&nbsp;&nbsp;&nbsp;
<a href="#col">2. Columns</a>&nbsp;&nbsp;&nbsp;
<a href="#but">3. Top buttons</a>&nbsp;&nbsp;&nbsp;
<a href="#stat">4. Statistics</a>

<!----------------------------------------------------->
<a id="tab"></a>
<h3>1. Results Table</h3>

<small>A pair hits table</small>
<br><a href="img/query_result.png">
<img src="img/query_result.png" alt="query results" style="border: 1px solid black; width: 600px"></a>

<ul>
<li><ttx>Pair Hits</ttx>:
<ul>
<li>The table contains columns for <i>all</i> of the selected species, but each hit only connects two species, and the other species columns are empty.
<li>Each <ttx>Hit#</ttx> is only listed once unless minor genes are included (see <a href="#rules">Rules</a>).
<li>A gene may be listed more than once if multiple hits align to it with a best overlap (i.e. major gene with multiple hits).
</ul>
<p>
<li><ttx>Single genes</ttx>:
<ul>
<li>If the query specified <ttl>Single</ttl> genes, then each row represents one gene and shows data only
for one species.
</ul>
</ul>
<p>You can sort the columns by clicking the column name in the table, and rearrange them by dragging the
column name. You can add/remove columns using the <ttb>Select Columns</ttb> button at the bottom.

<a id="col"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2. Columns</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b>Sections:</b>
<a href="#pairCol"> Pair hits columns</a>&nbsp;&nbsp;&nbsp;
<a href="#singleCol"> Single gene columns</a>&nbsp;&nbsp;&nbsp;
<a href="#save"> Auto-save columns</a>

<p>
<a href="img/query_columns.png">
<img src="img/query_columns.png" alt="query columns" style="border: 1px solid black; width: 600px"></a>

<table class="tx">
<tr><td>The buttons on the bottom will be <ttb>Select Columns</ttb> and <ttb>Hide Stats</ttb>. If
<ttb>Select Columns</ttb> is selected, it changes to <ttb>Hide Columns</ttb> and the
<ttb>Hide Stats</ttb> is replaced with <ttb>Clear</ttb> and <ttb>Defaults</ttb>.
<td><a href="img/query_lower_buttons.png"><img src="img/query_lower_buttons.png" alt="query results" style="border: 1px solid black; width: 250px"></a>
</table>

<p>In the column panel shown above, <i>hover</i> over a column name to see its brief description. Following
are the full descriptions of the columns.

<a id="pairCol"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>2.1 Pair hits columns</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx">
<tr>
	<td colspan=3><ttl>General</ttl>
<tr>
	<td style="vertical-align: text-top;"><ttx>Row</ttx><td>&nbsp;
	<td>Row number within the table
<tr>
	<td style="vertical-align: text-top;"><ttx>Block</ttx><td>&nbsp;
	<td>Synteny block containing this hit (if any). The format is Cn.Cm.Block#, where
		Cn and Cm are the chromosome numbers.
<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Block Hits</ttx><td>&nbsp;
	<td>The number of hits which comprise the synteny block.
<tr>
	<td style="vertical-align: text-top;"><ttx>Collinear</ttx><td>&nbsp;
	<td>Collinear set containing this hit (if any). The format is Cn.Cm.Size.Set# (e.g. 1.2.5.100; there are 5
		adjacent gene hits in set# 100 on Chr1 to Chr2).
<tr>
	<td style="vertical-align: text-top;"><ttx>Grp#</ttx><td>&nbsp;
	<td>If <ttl>Gene#, Multi-hit gene, PgeneF</ttl>: These three queries produces groups of hits,
	where each group has a group number. These numbers are generated during the search so will not be the same for different filter settings.
<tr>
	<td style="vertical-align: text-top;"><ttx>GrpSize</ttx><td>&nbsp;
	<td>If <ttl>Gene#, Multi-hit gene, PgeneF</ttl>: Size of the group for the corresponding <ttx>Grp#</ttx>.
<tr>
	<td style="vertical-align: text-top;"><ttx>Hit#</ttx><td>&nbsp;
	<td>The hit number, which are sequential numbers for the chromosome pair.
	This number is shown on the <ttp>Chromosome Explorer</ttp> when the mouse is over the hit line.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Id</ttx><td>&nbsp;
	<td>Percent identity of the alignment. The value of the "Identity" column is from the MUMmer file.
	If the hit has subhits, then this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Sim</ttx><td>&nbsp;
	<td>Percent similarity of the alignment (as determined by the BLOSUM scoring matrix).
	The value of the "Similarity" column is from the MUMmer file. If the hit has subhits, then
	this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit #Subs</ttx><td>&nbsp;
	<td>The number of subhits in a clustered hit.
<tr><td style="vertical-align: text-top;"><ttx>Hit St</ttx><td>&nbsp;
	<td>If "=", both hit ends are to the same strand; if "!=", they are to different strands.
<tr><td style="vertical-align: text-top;"><ttx>Hit Cov</ttx><td>&nbsp;
	<td>The length of the subhits within a clustered hit are summed taking into accounts overlaps; the longest summed subhits
	of both sides is shown. See <i>Clustered Hits</i> in <a href="UserGuide.html#term">Terminology</a>.
<tr><td style="vertical-align: text-top;"><ttx>Hit Type</ttx><td>&nbsp;
	<td>There are two alternative algorithms for clustering the hits on database creation.
	<br><ttx>Algo1</ttx>: g2 (two genes), g1 (one gene), g0 (no genes).
	<br><ttx>Algo2</ttx>: E is exon, I is Intron, n is intergenic. There will be 2 characters, one for each project,
	where the 1st letter goes with the alphabetically lesser project name; e.g 'EI' indicates the hit covers A.thal exon(s)
	and Cabbage intron(s).
<tr>
	<td colspan=3>
		<ttl>Gene&Hit Info</ttl>: one row for each species
<tr>
	<td style="vertical-align: text-top;"><ttx>Chr</ttx><td>&nbsp;
	<td>Chromosome of the hit.

<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Gstart/Gend/Gst</ttx><td>&nbsp;
	<td>Start and end of the annotated gene. The <ttx>Gst</ttx> is the strand (+/-).
<tr>
	<td style="vertical-align: text-top;"><ttx>Gene#</ttx><td>&nbsp;
	<td>The gene number is C.#.{a-z}. The C is the chromosome number.
	The # is the sequential number along the chromosome.
	If a run of genes overlap, they receive the
	same gene number with different suffixes {a-z, a2-z2, etc}. This is shown on the <ttp>Chromosome Explorer</ttp>
	"Annotation Description" and when the mouse hoovers over the gene.
<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hstart/Hend</ttx><td>&nbsp;
	<td>Start and end of the hit region.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hlen</ttx><td>&nbsp;
	<td>Hend-Hstart+1
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Olap</ttx><td>&nbsp;
	<td>The value depends on which <ttx>Cluster Hit</ttx> algorithm was used.
	<br><ttx>Algo1</ttx>: If <i>any</i> of the project pairs used <ttx>Algo1</ttx>, then this column will be
	the <i>gene overlap</i> for all project pairs.
	<br><ttx>Algo2</ttx>: If <i>all</i> of the project pairs used <ttx>Algo2</ttx>, then this column will be
	the <i>exon overlap</i>.
	<br><font color=green>NOTE:</font> For gene overlap, <ttx>Algo2</ttx> takes into account gaps between the subhits, whereas
	<ttx>Algo1</ttx> does not.
<tr>
	<td colspan=3><ttl>Annotation</ttl>: one row for each species
<tr>
	<td colspan=3>The keywords for the annotations of each species are listed;
	they can be different for each species.
	See <a href="SystemHelp.html#annot" class="ext" target="_blank">Project Parameters</a>
	for modifying the keywords shown. The <ttx>Anno Key Count</ttx> can be modified at any time using <tt>symap</tt> (not <tt>viewSymap</tt>).
</table>

<!---------------------------------------------------------------------------->
<a id="singleCol"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>2.2. Single genes columns</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The single genes tables only have the
<ttl>Gene Info</ttl> and <ttl>Annotation</ttl> columns, with one additional column, as follows.
<table class="tx">
<tr><td style="vertical-align: text-top;"><ttx>NumHits</ttx>
<td>This is the number of hits to the gene in the ENTIRE databases, except for SELF synteny.
<br>For example, if X, Y
and Z species have all been compared for synteny, but only Y and Z are being queried,
the <ttl>Orphans</ttl> for Y can have <ttx>NumHits</ttx>&gt;0 if it hits X.
</table>

<table><tr>
<td><small>Single result from 3 selected species</small>
<br><a href="img/query_single_result.png">
<img src="img/query_single_result.png" alt="query single results" style="border: 1px solid black; height: 360px"></a>
<td><small>Single result from 2 selected species</small>
<br><a href="img/query_single_result2.png">
<img src="img/query_single_result2.png" alt="query single results2" style="border: 1px solid black; height: 360px"></a>
</table>

<p>In the above tables, the A.thal gene #2 is the 2-species table because it has not hit to Cabbage; but it is not in
the 3-species table because it does have a hit to A.thal. The values of <ttx>NumHits</ttx> is the same in both. That is,
they both are showing orphan genes relative to the selected set, but the <ttx>NumHits</ttx> lets the user know if its
orphan over all the computed sets.
<p>The other query on singles is for <ttl>All genes</ttl>, which provides the same columns.
<a id="save"></a>
<h4>2.3 Auto-save columns</h4>
The columns selection is saved in a file called <tt>.symap_saved_props</tt> in the user's home directory.
If you only have one SyMAP database, there is no need to read the following:

<p>Say the last database D<sub>n</sub> queried had N species,
and the current database D<sub>m</sub> has M species, where D<sub>n</sub> and D<sub>m</sub> are different
SyMAP databases:
<ol>
<li>If N=M: the <ttx>General</ttx> and <ttx>Gene&Hit Info</ttx> columns will
be set according to the previous settings, but the <ttx>Annotation</ttx> columns may be wrong.
<li>If N&gt;M or N&lt;M: the <ttx>General</ttx> will be according to the previous settings,
all <ttx>Gene&Hit Info</ttx> columns will be set like the first D<sub>n</sub> species, and no
<ttx>Annotation</ttx> columns will be set.
</ol>

<a id="but"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3. Top button functions</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/query_buttons.png"><img src="img/query_buttons.png" alt="query buttons" style="border: 1px solid black; width: 600px"></a>
<p><ttb>Show</ttb>: For the selected row, a popup will show all columns and associated information for the hit.
The text in the popup can be copied.

<a id="muscle"></a>
<table class="tx">
<tr><td style="vertical-align: text-top;">
<ttb>Align</ttb>:
Select one or more rows. The sequences of the selected hit(s) are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32). The figure on the right shows the MUSCLE alignment of
two genes from B.rapa and genes from Cabbage and A.thal.
<td>
<a href="img/query_muscle.png"><img src="img/query_muscle.png" alt="query muscle" style="border: 1px solid black; width: 450px"></a>
</table>

<p>&nbsp;

<table class="tx">
<tr><td style="vertical-align: text-top;">
<td style="vertical-align: text-top;">
<ttb>View 2D</ttb>: This displays the <ttp>2D</ttp> view for the selected entry. The following 4 views
are provided using the drop-down (the image on the right shows a Collinear set of 5).

<p>
<small>Columns for 3 of the 4 options:</small><br><a href="img/query_columns_2D.png"><img src="img/query_columns_2D.png" alt="query 2D columns" style="border: 1px solid black; width: 280px"></a>
<td><a href="img/query_synteny.png"><img src="img/query_synteny.png" alt="query synteny" style="border: 1px solid black; width: 450px"></a>
</table>
<a id="group"></a>
<table class="ty">
	<tr><td>Option<td>Column*<td>Selected Hit<td>Display Filter
	<tr><td style="white-space: nowrap;"><ttl>Region (kb)</ttl>
	<td>N/A
	<td>The hit is padded to each side by the amount indicated (default 50kb).
	<td><ttx>Show all hits</ttx>.

	<tr><td style="white-space: nowrap;"><ttl>Synteny block</ttl>
	<td><ttx>Block</ttx>
	<td>The entire block for the hit will be shown.
	<td><ttx>Show Block Hits</ttx>.

	<tr><td style="white-space: nowrap;"><ttl>Collinear set</ttl>
	<td><ttx>Collinear</ttx>
	 <td>The collinear set for the hit will be shown.
	 <p><ttx>Highlight Collinear Set</ttx> will be automatically selected.
	<td><ttx> Show Block and Set Hits</ttx>.

	<tr><td style="white-space: nowrap;"><ttl>Group</ttl>
	<td><ttx>Grp#</ttx>
	<td>The group for the hit will be shown, where they are required to be on the
	same chromosome pair as the selected row.
	<p>The hits are highlighted in the "Special" color, e.g. <a href="#group_ex">Example</a>.
	See <a href="#high">High</a> below for disabling the highlighting.

	<td><ttx>Show all hits</ttx>
</table>
*<small>The selected row must have a value for the column.</small>
<p>
After the initial display, the <ttp>2D</ttp> view can be changed as described in the
<a href="UserGuide.html#control" class="ext" target="_blank">User Guide</a>.

<p><a id="high"></a>
If the <ttl>High</ttl> checkbox is selected, the selected hit is highlighted in the
"Special" color (default magenta), and can be changed
in the <a href="UserGuide.html#wheel"  class="ext" target="_blank">Color Icon</a>.
The coloring can also be turned off by selecting the <ttl>2D Hit Filter</ttl> <ttx>Hit popop (or Query)</ttx>
option.

<p>&nbsp;
<a id="export"></a>
<p><ttb>Export</ttb>: One or more rows can be selected for the following exports;
or if no rows are selected, the entire table is exported.
<table class="tx">
<tr><td>
<p><ttl>CSV</ttl>:
Export the rows using the selected set of columns to a CSV format suitable for import into Excel.
<p><ttl>HTML</ttl>:
Export the rows using the selected set of columns to a HTML format suitable for viewing on the web
(e.g. <a href="./img/query_export.html"  class="ext" target="_blank">Example</a>).
<p>
<ttl>FASTA</ttl>: (Pairs Only) Sequences from the rows are written to a FASTA file.
Both sides of each hit are written using the start/end coordinates shown in the table.
NOTE: this is a very slow function and takes minutes if many rows are selected.
<td>
<a href="img/query_export.png"><img src="img/query_export.png" alt="result export" style="border: 1px solid black; width: 260px"></a>
</table>
The <ttl>Include Row column</ttl> option is available because this column is alway present, but
it may be desirable to not include it in the output.

<a id="stat"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Statistics</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Statistics for the query results are shown at the bottom of the results table. They can be
hidden by selecting <ttl>Hide Stats</ttl>.
<p><a href="img/query_stats.png"><img src="img/query_stats.png" alt="result stats" style="border: 1px solid black; width: 450px"></a>

<p>Most of the statistics are self-explanatory except the following:
<p><ttx>Annotated</ttx> and <ttx>Genes</ttx>: The first is the number of hits that
overlap one or more genes, where a gene can have multiple hits.
The second is the number of Genes with at least one hit.
<p><ttx>Groups</ttx>: This statistic
is only shown if <ttl>Multi-hit genes</ttl> was checked, and it is
the number of distinct groups per species (along with the last <ttx>Grp#</ttx> for the species).
<p><ttx>Regions</ttx>: This statistic
is only shown if <ttl>PgeneF</ttl> was checked, and it is
the number of distinct regions covered on that species.

<a id="rtab"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>&gt;Results</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/query_result_tab.png"><img src="img/query_result_tab.png" alt="query results tab" style="border: 1px solid black; width: 600px"></a>
<p>All query results are listed under the <ttb>&gt;Results</ttb> tab on the left. Clicking this tab shows the table of queries illustrated
above. Query results can be displayed by clicking a result on the left panel, or double-clicking it in the list of results table shown above.
<p>The only way to remove query results from the left tab is by selecting them in this table followed by <ttl>Remove Selected Query</ttl>.
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

