<!DOCTYPE html>
<html>
<head>
<title>SyMAP Queries</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
    a.ext {font-weight: bold;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; text-align: left}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
    ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: Arial, monospace; font-size: 15px; font-weight: 500; color: DarkSlateGrey; }
	xx	{line-height: 2}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr><td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
					<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Queries
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
					<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
				<br>
			<tr><td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr><td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./input/index.html" class="white">Input</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
				</table>
		</table>
	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<b>Content:</b>
<table>
<tr>
<td style="white-space: nowrap; vertical-align: text-top;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td style="white-space: nowrap; vertical-align: text-top;">
	<a href="#instruct"><u><small>&gt;</small> Instructions</u></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td style="white-space: nowrap; vertical-align: text-top;">
	<table>
		<tr><td><a href="#query"><u><small>&gt;</small> Query Setup</u></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
		<tr><td><a href="#rules">Rules</a>, <a href="#genen">Gene#</a>
		<tr><td><a href="#gen">General</a>
		<tr><td><a href="#sin">Single genes</a>
		<tr><td><a href="#pair">Pair hits</a>
		<tr><td><a href="#pfs">Compute Groups</a>
		<tr><td>&nbsp;&nbsp;&nbsp;<a href="#multi">Multi-gene</a>, <a href="#clhit">Cluster</a>
	</table>
<td>&nbsp;&nbsp;&nbsp;
<td style="white-space: nowrap; vertical-align: text-top;">
	<table>
		<tr><td><a href="#result"><u>Result Panel</u></a>&nbsp;&nbsp;
		<tr><td><a href="#tab">Result table</a>&nbsp;&nbsp;&nbsp;&nbsp;
		<tr><td><a href="#col">Columns</a>&nbsp;&nbsp;&nbsp;&nbsp;
		<tr><td>&nbsp;&nbsp;&nbsp;<a href="#pairCol">Pair</a>, <a href="#singleCol">Single</a>, <a href="#save">Auto-save</a>&nbsp;&nbsp;&nbsp;&nbsp;
		<tr><td><a href="#stat">Statistics</a>
		<tr><td><a href="#but">Top buttons</a>
		<tr><td>&nbsp;&nbsp;&nbsp;<a href="#msaAlign">MSA, <a href="#view2D">View 2D</a>, <a href="#report">Gene Report</a>...
	</table>
<td  style="white-space: nowrap; vertical-align: text-top;">
	<table>
	<tr><td><a href="#rtab"><u><small>&gt;</small> Results</u></a>
	<tr><td>&nbsp;
	<tr><td>&nbsp;
	<tr><td>Extra topic:
	<tr><td>&nbsp;&nbsp;&nbsp;<a href="#msp">3-chromosomes</a>
	</table>
</table>

<p>See <a href="Release.html#v562" class="ext" target="_blank">Release</a> for the latest v5.6.1-v5.6.5 changes.
&nbsp;&nbsp;See <a href="./UserGuide.html#term" class="ext" target="_blank">Terminology</a> in the User Guide.
<!------------------------------------------------------------------------------>
<hr>
<a id="instruct"></a>
<h2><small>&gt;</small> Instructions</h2>
To open the query interface, first select two or more sequence projects in the
<a href="UserGuide.html#mgr" class="ext" target="_blank">Project Manager</a>.
Then select the <ttb>Queries</ttb> button at the lower right to open the queries interface.

<table>
<tr><td>
The <ttl>Instructions</ttl> window (on right) lists the projects which were selected for querying.
<p>The <ttl>Notes</ttl> states what the <ttx>Olap</ttx> column represents; if one or more project pairs used <ttx>Algo1</ttx>, it will be gene overlap;
if all project pairs used <ttx>Algo2</ttx>, it will be exon overlap.
<p>Open the <ttl>Query Setup</ttl> window by clicking on its tab in the left panel.
<td><a href="img/query_oview.png"><img src="img/query_oview.png" alt="query_oview" style="border: 1px solid black; width: 500px"></a>
</table>

<!-------------------QUERY SETUP --------------------------------------------------->
<a id="query"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2><small>&gt;</small> Query Setup</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

&nbsp;&nbsp;&nbsp;&nbsp;<a href="#rules">1. Rules</a> (<a href="#genen">Gene#</a>,
<a href="#pseudo">Pseudo</a>, <a href="#stop">Wait/Stop</a>)&nbsp;&nbsp;&nbsp;
<a href="#gen">2. General</a>&nbsp;&nbsp;&nbsp;
<a href="#sin">3. Single genes</a>&nbsp;&nbsp;&nbsp;
<a href="#pair">4. Pair Hits</a>&nbsp;&nbsp;&nbsp;
<a href="#pfs">5. Compute Groups</a>
<p>

<a id="rules"></a>
<h3>1. Rules</h3>

<table>
<tr><td style="vertical-align: text-top;">
Set up the desired filters and then select <ttb>Run Query</ttb> to execute it.
When the search is complete, the <a href="#result">table result panel</a> will be displayed.

<p>1. Most filters can be used in conjunction with other
	filters; options will be disabled if they cannot be used with a selected filter.
<p>2. The <ttl>Single</ttl> queries returns rows without a hit, and only list genes for one project.
<p>3. All <ttl>Pair hits</ttl> queries return a row per <i>hit pair</i>, i.e.
	aligned region between to two chromosomes from two projects.

<td><a href="img/query_query.png"><img src="img/query_query.png" alt="query_query" style="border: 1px solid black; width: 550px"></a>
</table>

<b>Hit#</b>
<table>
<tr><td>All <ttx>Hit#</ttx> are uniquely numbered for a chromosome pair,
	e.g. there will be a <ttx>Hit#</ttx> 3 on all chromosome pairs that have hits.

<td><a href="img/query_table_rows.png"><img src="img/query_table_rows.png" alt="query_table_rows" style="border: 1px solid black; width: 450px"></a>
</table>

<a id="genen">
<table style="width: 95%"><tr><td style="text-align: left">
<b>Gene#</b>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr><td style="vertical-align: text-top;">
	All genes have a <ttx>Gene#</ttx>, which are sequentially numbered per chromosome, e.g. there will be a
	<ttx>Gene#</ttx> 3 on all chromosomes that have at least 3 non-overlapping genes. Overlapping
	genes have the same number, but different suffixes, e.g. 4.b.

	<p><ttl>Major</ttl> and <ttl>Minor</ttl> genes:
	<ol type='a'>
	<li>A hit can align to multiple genes on either or both chromosomes; it is assigned to the best (<ttx>major</ttx>) gene on each end.
	<li>For example, in the image on the right, the pink gene is the major gene as the hit (pink line)
	fits it best, whereas the burgundy gene is the minor gene.
	<li>When queries involve genes, only the hits to the major genes will be shown
	except for the following queries, which will also list the minor genes:
	<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<ttl>Every*</ttl>, <ttl>Hit#</ttl>,
	<ttl>Gene#</ttl>, and <ttl>Multi</ttl> with <ttl>Minor*</ttl> checked.
	</ol>
	<td><a href="img/queryMultOlap.png">
	<img src="img/queryMultOlap.png" alt="Hit overlaps two genes" style="border: 1px solid black; width: 35px"></a>
</table>

<a id="pseudo">
<table style="width: 95%"><tr><td style="text-align: left">
<ttl>Pseudo</ttl>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ttx>Un-annotated</ttx>:
A hit that does not overlap a gene (annotated from the .gff file) is un-annotated (<i>pseudo gene</i>). It may be numbered
or un-numbered, depending on the <a href="SystemHelp.html#pseudo" class="ext" target="_blank">parameter</a> setting when the database was built.
The <a href="#instruct">Instruction</a> wil say whether they are numbered or not.
<p>
<table><tr><td>
Not numbered: The <ttx>Gene#</ttx> column will show the chromosome followed by ".~".
<p>Numbered: These are numbered sequentially, where the counts start after the annotated gene numbers.
For example, if the last Gene# for Chr03 is 5550 (e.g. 3.5550.), the first pseudo gene number will be 6000 and
the column will show 3.6000.~.

<td><a href="img/query_pseudo.png"><img src="img/query_pseudo.png" alt="Pseudo genes" style="border: 1px solid black; width:  350px"></a>
</table>
If numbered, the pseudo genes can be included in <a href="#clhit">Clusters</a> and <a href="#report">Report</a>.

<a id="stop">
<p><b>Wait/Stop Query:</b>
<p>&rarr; <i>Wait for the current query to stop running before starting another!</i>
<br>It is not possible to start a new query from the same panel, but you could start one from a different table;
<i>they both will fail, so do NOT do this!</i>

<table class="tx">
<tr><td>
&rarr;<ttl>Stop</ttl>
<br>When a query is running, you will see a status line and the <ttl>Stop</ttl> button. You can stop
the query which will take the right hand view to the <a href="#rtab">Results Panel</a>.
<p>It cannot stop the database search, but once the search is done, it stops all subsequent processing.
You will typically see the status line change 3 or 4 times before showing the table.
<td><a href="img/query_stop.png">
	<img src="img/query_stop.png" alt="Query Stop" style="border: 1px solid black; width: 150px"></a>
	<br>
	<a href="img/query_stop2.png">
	<img src="img/query_stop2.png" alt="Query Stop" style="border: 1px solid black; width: 140px"></a>
</table>

<a id="gen"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2. General</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr>
<td><ttl>Annotation Description</ttl>
<br>Enter a substring: the entire annotation string (i.e. column <ttl>All_Anno</ttl>) will be searched
for the substring.
Hits will be shown that have at least one gene, of two possible, with the annotation.

<td><a href="img/query_gen.png">
<img src="img/query_gen.png" alt="General" style="border: 1px solid black; width: 400px"></a>
</table>

<p><ttl>Location</ttl>
<table class="tx">
<tr><td style="vertical-align: text-top;"><ttl>Chr</ttl><td>&nbsp;
	<td>Select a specific chromosome for the species.
<tr><td style="vertical-align: text-top;"><ttl>From</ttl><td>&nbsp;
	<td>The <ttx>Hit Start</ttx> coordinates for the selected chromosome will be &gt;= this number.
<tr><td style="vertical-align: text-top;"><ttl>To</ttl><td>&nbsp;
	<td>The <ttx>Hit End</ttx> coordinates for the selected chromosome will be &lt;= this number.
</table>
<br>It is valid to enter only the <ttl>From</ttl> or <ttl>To</ttl>, or leave both blank.

<p>The <ttl>From</ttl> and <ttl>To</ttl> are disabled for
<ttl>Single genes</ttl>, <ttl>Block#</ttl>, <ttl>Collinear set#</ttl>, <ttl>Hit#</ttl>, and <ttl>Gene#</ttl>.

<!--------------------------------------------------->
<a id="sin"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3. Single genes</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<table><tr>
<td>The <ttl>Single</ttl> queries produces rows of genes; there is no hit or block
information since the rows do not represent hits.
<td><a href="img/query_single.png">
<img src="img/query_single.png" alt="Single" style="border: 1px solid black; width: 400px"></a>
</table>
Options:
<ul>
<li><ttl>Orphan genes (no hits)</ttl>
<br>
Genes that do not a have a hit and meet the additional filters.
The orphan genes are relative to the projects shown on the <ttl>Instruction</ttl> page.
For example, if species X, Y and Z have synteny computed between all pairs,
but only X and Y are selected, the orphan genes for X would be those with no hits to Y.
If X,Y&Z are selected, the orphan genes for X would be those with no hits to Y and Z.
<br>&nbsp;
<li><ttl>All genes (w/o hits)</ttl>, i.e. genes with and without hits
<br>This shows all genes that meet the additional filters, regardless if they have a hit or not.
There is always the same set of genes for a project, regardless of synteny.
</ul>

<a id="active"></a>
<table><tr>
<td><u>Unselect species</u>: When the <ttl>Single</ttl> or <ttl>Gene#</ttl> is selected, the check boxes beside the species names will be
activated. In order to view the genes from just one species, deactivate the others. Additionally,
a single chromosome can be selected, but location cannot be entered.
<td><a href="img/query_active.png">
<img src="img/query_active.png" alt="Single" style="border: 1px solid black; width: 400px"></a>
</table>

<!--------------------------------------------------->
<a id="pair"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Pair hits</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table><tr><td>
Each hit connects two <i>species (projects)</i> and hence represents a <i>pair</i> of aligned
regions for two of the selected species.
<p>Filters are as follows:
<td><a href="img/query_pair.png">
<img src="img/query_pair.png" alt="Pair" style="border: 1px solid black; width: 400px"></a>
</table>

<p><ttl>In Block (Synteny Hit)</ttl>
<table class="tx">
<tr><td  style="vertical-align: text-top;">Yes<td><td>&nbsp;<td>Only hits that are part of a synteny
block will be returned. All hits will have a value for the <ttx>Block</ttx> column.
<tr><td style="vertical-align: text-top;">No<td><td>&nbsp;<td>Only hits that are NOT part of a synteny
block will be returned. No hits will have a value for the <ttx>Block</ttx> column.
</table>

<a id="genehit"></a>
<p><ttl>Annotated (Gene Hit)</ttl>
<table class="tx">
	<tr><td style="vertical-align: text-top;">
		Every<td>&nbsp;
		<td>Only hits that align to a gene on one or both sides of the hit will be shown.
		The <ttx>Gene#</ttx> columns will list the best gene.
	<tr><td style="vertical-align: text-top;">
		Every*<td>&nbsp;
		<td>This is like the <ttx>Every</ttx> option, but a hit may be listed multiple times if it aligns
		to minor genes. The minor <ttx>Gene#</ttx> will be suffixed with an "*" and the <ttx>Hit Type</ttx> is unknown.
	<tr><td style="vertical-align: text-top;">
		One<td>&nbsp;
		<td>Only hits that align to a gene on ONE end will be shown.
		The <ttx>Gene#</ttx> column will show the best gene for the end with the hit.
	<tr><td style="vertical-align: text-top;">
		Both<td>&nbsp;
		<td>Only hits that align to genes on BOTH ends will be shown.
		The <ttx>Gene#</ttx> column will show the best gene for both ends of the hit.
	<tr><td style="vertical-align: text-top;">
		None<td>&nbsp;
		<td>Only hits that do NOT align to a gene on either end will be shown.
		The <ttx>Gene#</ttx> column will not have a value for either end of the hit.
	<tr><td colspan=3>If a species chromosome is selected with <ttl>One</ttl> and a species chromosome selected,
the hit must be on the species chromosome, but the "one" gene can be on either end. Same for
<ttl>Every</ttl> or <ttl>Every*</ttl>.
</table>

<p><ttl>Collinear size</ttl>
<table class="tx">
	<tr><td style="white-space: nowrap;vertical-align: text-top;">
		&gt;= [=, &gt;=] N
		<td>&nbsp;
		<td>List all hits in collinear sets that have size &gt;= N or  = N or &lt;= N, respectively.
	<tr><td style="vertical-align: text-top;">
		Ignore
		<td>&nbsp;
		<td>Do not filter on collinear set sizes.
	<tr><td colspan=3>See <a href="Details.html" class="ext" target="_blank">Collinear</a>, which explains the SyMAP collinear sets.
</table>

<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p>For the following 4 filters, <i>do not include the 'Chr' number</i>. Instead use the chromosome pull-downs
to narrow the search to a specific chromosome, as exampled in <ttl>Block#</ttl>.
<p><ttl>Block#</ttl>
<br>Enter a single block number (<u>the <ttx>Block</ttx> column is formatted Chr.Chr.Block#</u>).
<br>This will display all hits with this block number from any chromosome pair.
<br>Example using the chromosome pull-downs: if you select Chr 1 from the first project, Chr 2 from a second project, and enter block=3, you will see
hits in block 1.2.3.

<p><ttl>Collinear Set#</ttl>
<br>Enter a set number (<u>the <ttx>Collinear</ttx> column is formatted Chr.Chr.Size.Set#</u>).
<br>This will display all hits with this set number from any chromosome pair. See Block example.

<p><ttl>Hit#</ttl>
<br>Enter a hit number. Both major and minor gene hits will be shown.
<br>This will display all hits with this number from any chromosome pair. See Block example.

<a id="genenum"></a>
<p><ttl>Gene#</ttl>
<br>Enter a <ttx>Gene#</ttx> number (<u>the <ttx>Gene#</ttx> column is formatted Chr.Gene#.suffix</u>, where suffix may be blank).
<br>A gene will only show if it has a hit. All hits with the <ttx>Gene#</ttx> on either end will be shown.
<br>If a gene has a suffix:
<br>&nbsp;&bull; If only a number is entered, all genes with the numeric prefix will be displayed (including minor hits).
<br>&nbsp;&bull; If a number.suffix is entered, the exact gene will be displayed, including minor hits.
<br>Select a chromosome:
<br>&nbsp;&bull; The hit has to have one end to the chromosome, but the <ttx>Gene#</ttx> can be on either end.
<br>&nbsp;&bull; By unselecting the other species (see <a href="#active">Unselect species</a>), the <ttx>Gene#</ttx> will only be on the specified end.
<br>Group:
<br>&nbsp;&bull; When a gene hits multiple places on the same opposite chromosome, the hits are put in a group
<br>&nbsp;&nbsp;&nbsp;(see <ttx>Group</ttx> column), which can be viewed with the
<ttl>2D</ttl> <a href="#group">Group</a> pull-down option (similar to <a href="#group_ex">Multi-hit</a>).
<!--------------------------------------------------->
<a id="pfs"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Compute Groups</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
&nbsp;&nbsp;&nbsp;<b>Sections:</b> &nbsp;&nbsp;&nbsp;
<a href="#multi">1. Multi-hit genes</a>&nbsp;&nbsp;&nbsp;
<a href="#clhit">2. Cluster hits</a>&nbsp;&nbsp;&nbsp;

<p>These two options are computed on the fly.
They produce query results with values for the <ttx>Group</ttx> column; this in turn
allows the <ttl>View 2D</ttl> <a href="#group">Group</a> option to be used.

<a id="multi"></a>
<h4>4.1. <ttl>Multi-hit genes</ttl></h4>
<a href="img/query_multi.png">
<img src="img/query_multi.png" alt="Multi" style="border: 1px solid black; width: 450px"></a>

<p>List all genes that have &gt;= N hits <i>to the same species</i>. To view the groups merged,
see <a href="#rmulti">Multi-hit gene report</a>.
<ul>
<li>The <u>target</u> gene refers to the gene with &gt;= N hits to the <u>opposite</u> species.
<li>The target must be a gene, but it can be to non-gene hits on the opposite species.
<li>Used with the following filters:
<ul>
<li><ttl>Both</ttl> (<ttx>Annotated Genes</ttx>): if selected, the opposite species must also have a gene hit.
<li><ttl>Chr</ttl>: If one or more chromsomes are selected, the target genes will only be on those chromosomes.
</ul>
<li>See <a href="#rmulti">Multi-hit Gene Report...</a> for a good way to see a condensed view of the results.
</ul>
<br>The options are as follows:
<p>
<table style="tx">
<tr><td>&nbsp;<td><ttl>Minor*</ttl>
	<td>&nbsp;<td>Include minor hits on either chromosome.
<tr><td>&nbsp;<td style="vertical-align: text-top;">
	<ttl>Tandem</ttl>
    <td>&nbsp;<td>(Annotated species only): The &gt;=N hits must be to a tandem array of genes.
	<br><ttl>Same Chr</ttl> is automatic with this option. Numbered pseudos will not be included.
<tr><td>&nbsp;<td style="white-space: nowrap; vertical-align: text-top;">
	<ttl>Same Chr</ttl>
	<td>&nbsp;<td>The &gt;=N hits must all be on the same opposite chromosome.
<tr><td>&nbsp;<td style="white-space: nowrap; vertical-align: text-top;">
	<ttl>Diff Chr</ttl>
	<td>&nbsp;<td>The &gt;=N hits may be on any set of opposite chromosomes for a given species.
</table>

<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
	<tr><td style="vertical-align: text-top;">
	The table can list the same hit multiple times, because gene X and gene Y may be connected by a hit, and both genes
	have &gt;=N hits, so the hit in both groups need to be shown.
	<p>An example is shown on the right, where Hit#881 is in both Atha Group#3 and Brap Group#15.
	The image below shows the two groups, where the group hits are highlighted in magenta; these were
produced using the <ttl>View 2D</ttl> option <a href="#group">Group</a> (selecting any hit in the group results in the
same 2D display).
	<td><a href="img/query_multi_results.png">
	<img src="img/query_multi_results.png" alt="multi-hit results" style="border: 1px solid black; width: 250px"></a>
</table>

<a id="group_ex"></a>
<br><a href="img/query_multi_2D.png">
<img src="img/query_multi_2D.png" alt="multi-hit 2D" style="border: 1px solid black; width: 700px"></a>

<!------------ Cluster --------------------->
<a id="clhit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.2. <ttl>Cluster genes</ttl></h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
&rarr;<i>Note: this algorithm was written for v5.6.3.</i>
<ul>
<li>This algorithm finds clusters overlapping genes. This is, every gene in the cluster aligns to at least one
other gene in the cluster, and all genes that align to any gene in the cluster are included. For example, for
a cluster of genes {A,B,C,D} each has a hit with at least one gene in the cluster and no hits to a gene not in the
cluster.
<li>Each cluster is given a group number, which is shown in the Query Results table (column name <ttx>Group</ttx>).
<li>These can be pre-filtered by species chromosome,  etc.
<li>If the un-annotated are not numbered (see <a href="#pseudo">Pseudo</a>) only hits with both genes will be included.
<li>If the un-annotated are numbered, the pseudo genes will be included, that is, the default is to consider every
hit with at least one gene. In this case, the <ttl>One Gene</ttl>
and <ttl>Both genes</ttl> filters can be used in conjunction with <ttl>Cluster</ttl>.
</ul>

<table class="tx"><tr><td>
	<tr>
	<td>If there are only two species, the only filter will be <ttl>&gt;=N</ttl>.
	<br>For more than 2 species, the filters are shown on the right.
	<td>&nbsp;
	<td><a href="img/query_clust.png">
	<img src="img/query_clust.png" alt="Cluster3" style="border: 1px solid black; width: 280px"></a>
</table>

<table class="tx">
<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>&gt;=N</ttl>
	<td>List all clusters that have &gt;= N hits.
<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>Include</ttl>
	<td> The resulting clusters may have at least one gene from all included species.
<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>Exclude</ttl>
	<td> The resulting clusters will contain no genes from the excluded species.
</table>

<p><u>Alternative algorithm</u>:
<br><ttl>PgeneF</ttl>: If you start <ttx>viewSymap</ttx> with the "-g" flag, you will see the
filters from the original PgeneF cluster algorithm.
&rarr;<i>Note: the PgeneF feature has not been tested for a long time except superficially
(i.e. making sure it has the same results as the previous release)</i>. It remains available
until the <ttl>Cluster gene</ttl> algorithm is enhanced to include non-gene hits.
<table class="tx">
<tr><td style="vertical-align: text-top;">
If you start <ttx>viewSymap</ttx> with the "-g" flag, you will see the
filters as shown on the right. The one addition to v5.6.3 is the &gt;= N hits filter.
<td><a href="img/query_clust1.png">
	<img src="img/query_clust1.png" alt="Cluster3" style="border: 1px solid black; width: 360px"></a>
</table>
<ttl>No gene</ttl> only shows clusters without annotation. The second two filters only occur
if there is &gt;2 species. From the original documentation: <ttl>Linkage</ttl>: Require the PgeneF to be fully linked,
i.e. for each pair of species A and B in the group, there must be a hit linking A to B.
<ttl>At least one</ttl>: Only PgeneF hits will be shown if they have
hits to the included species, although the PgeneF numbers will reflect groupings created
using all hits.
<hr>
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<a id="result"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Result Panel</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

&nbsp;&nbsp;&nbsp;<b>Sections:</b> &nbsp;&nbsp;&nbsp;
<a href="#tab">1. Results table</a>&nbsp;&nbsp;&nbsp;
<a href="#col">2. Columns</a>&nbsp;&nbsp;&nbsp;
<a href="#stat">3. Statistics</a>&nbsp;&nbsp;&nbsp;
<a href="#but">4. Top buttons</a>&nbsp;&nbsp;&nbsp;

<!----------------------------------------------------->
<a id="tab"></a>
<h3>1. Results Table</h3>

<small>A pair hits table</small>
<br><a href="img/query_result.png">
<img src="img/query_result.png" alt="query results" style="border: 1px solid black; width: 600px"></a>

<ul>
<li><ttx>Pair Hits</ttx>:
	<ul>
	<li>The table contains columns for <i>all</i> of the selected species, but each hit only connects two species, and the other species columns are empty.
	<li>Each <ttx>Hit#</ttx> is only listed once unless minor genes are included (see <a href="#rules">Rules</a>).
	<li>A gene may be listed more than once if it has multiple major hits.
	</ul>
<p>
<li><ttx>Single genes</ttx>:
	<ul>
	<li>If the query specified <ttl>Single</ttl> genes, then each row represents one gene and shows data only
	for one species.
	</ul>
</ul>

<p>You can sort the columns by clicking the column name in the table, and rearrange them by dragging the
column name. You can add/remove columns using the <ttb>Select Columns</ttb> button at the bottom.

<a id="col"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2. Columns</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

&nbsp;&nbsp;&nbsp;<b>Sections:</b>&nbsp;&nbsp;&nbsp;
<a href="#pairCol">1. Pair hits columns</a>&nbsp;&nbsp;&nbsp;
<a href="#singleCol">2. Single gene columns</a>&nbsp;&nbsp;&nbsp;
<a href="#save">3. Auto-save columns</a>

<p><a href="img/query_columns.png">
<img src="img/query_columns.png" alt="query columns" style="border: 1px solid black; width: 600px"></a>

<table class="tx">
	<tr><td>The buttons on the bottom will be <ttl>Select Columns</ttl> and <ttl>Hide Stats</ttl>. If
	<ttl>Select Columns</ttl> is selected, it changes to <ttl>Hide Columns</ttl> and the
	<ttl>Hide Stats</ttl> is replaced with the 3 buttons explained below.
	<td><a href="img/query_lower_buttons.png"><img src="img/query_lower_buttons.png" alt="query results" style="border: 1px solid black; width: 250px"></a>
	</table>
	<table class="tx">
	<tr><td><ttl>Clear</ttl><td>Clears the selection of all columns except <ttx>Row#</ttx>.
	<tr><td><ttl>Defaults</ttl><td>Selects the default columns, which are shown in the image above. If the collinear
	query was run, this will include the <tt>Collinear</tt>. If the group
	query was run, this will include the <tt>Group</tt>.
	<tr><td><ttl>Arrange</ttl><td>Arranges similar columns, putting the gene columns first.
</table>
<p>In the column panel shown above, <i>hover</i> over a column name to see its brief description. Following
are the full descriptions of the columns.

<a id="pairCol"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>2.1 Pair hits columns</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx">
<tr><td colspan=3><ttl>General</ttl>
<tr><td style="vertical-align: text-top;"><ttx>Row</ttx><td>&nbsp;
	<td>Row number. This column does not sort.
<tr><td style="vertical-align: text-top;"><ttx>Block</ttx><td>&nbsp;
	<td>The synteny block containing this hit (if any). The format is Chr.Chr.Block#, where
		the two "Chr" are chromosome numbers.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Block Hits</ttx><td>&nbsp;
	<td>The number of hits which comprise the synteny block.
<tr><td style="vertical-align: text-top;"><ttx>Collinear</ttx><td>&nbsp;
	<td>The collinear set containing this hit (if any). The format is Chr.Chr.Size.Set# (e.g. 1.2.5.100; there are 5
		adjacent gene hits in set# 100 on Chr1 to Chr2).
<tr><td style="vertical-align: text-top;"><ttx>Group</ttx><td>&nbsp;
	<td><ttl>Gene#, Multi-hit gene, Cluster</ttl>: These three queries produces groups of hits,
	where each group has a group number and size. The format is Size.Group#; clicking the column heading sorts by Group#.
	These numbers are generated during the search so will not be the same for different filter settings.

<tr><td style="vertical-align: text-top;"><ttx>Hit#</ttx><td>&nbsp;
	<td>The number assigned to the hit. They are sequential along the chromosome of the  alphabetically lesser species, e.g. Arab&lt;Brap.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Id</ttx><td>&nbsp;
	<td>Percent identity of the alignment. The value of the "Identity" column is from the MUMmer file.
	If the hit has subhits, then this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Sim</ttx><td>&nbsp;
	<td>Percent similarity of the alignment (as determined by the BLOSUM scoring matrix).
	The value of the "Similarity" column is from the MUMmer file. If the hit has subhits, then
	this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit #Subs</ttx><td>&nbsp;
	<td>The number of subhits in a clustered hit.
<tr><td style="vertical-align: text-top;"><ttx>Hit St</ttx><td>&nbsp;
	<td>The strand for the two ends of the end separated by a "/".
<tr><td style="vertical-align: text-top;"><ttx>Hit Cov</ttx><td>&nbsp;
	<td>The summed subhits within a clustered hit taking into accounts overlaps. The summed subhits
	are usually different for the two sides; this will be the longest.
<tr><td style="vertical-align: text-top;"><ttx>Hit Type</ttx><td>&nbsp;
	<td>There are two alternative algorithms for clustering the hits on database creation, which assign
	different hit types, as follows:
	<br><ttx>Algo1</ttx>: g2 (two genes), g1 (one gene), g0 (no genes).
	<br><ttx>Algo2</ttx>: E is exon, I is Intron, n is intergenic. There will be 2 characters, one for each gene,
	where the 1st letter goes with the alphabetically lesser project name;
	<br> e.g 'EI' would indicate the hit covers A.thal exon and Cabbage intron.
<tr><td colspan=3>
		<ttl>Gene&Hit Info</ttl>: one row for each species
<tr><td style="vertical-align: text-top;"><ttx>Chr</ttx><td>&nbsp;
	<td>Chromosome of the hit.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Gstart/Gend/Gst</ttx><td>&nbsp;
	<td>Start and end of the annotated gene. The <ttx>Gst</ttx> is the strand (+/-).
<tr><td style="vertical-align: text-top;"><ttx>Gene#</ttx><td>&nbsp;
	<td>The gene number is C.#.{a-z}. The C is the chromosome number.
	The # is the sequential number along the chromosome.
	If a run of genes overlap, they receive the same gene number with different suffixes {a-z, a2-z2, etc}.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hstart/Hend</ttx><td>&nbsp;
	<td>Start and end of the hit region.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hlen</ttx><td>&nbsp;
	<td>Hend-Hstart+1
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Olap</ttx><td>&nbsp;
	<td>The value depends on which <ttx>Cluster Hit</ttx> algorithm was used.
	<br><ttx>Algo1</ttx>: If <i>any</i> of the project pairs used <ttx>Algo1</ttx>, then this column will be
	the <i>gene overlap</i>.
	<br><ttx>Algo2</ttx>: If <i>all</i> of the project pairs used <ttx>Algo2</ttx>, then this column will be
	the <i>exon overlap</i>.

<tr><td colspan=3><ttl>Annotation</ttl>: one row for each species
<tr><td colspan=3>The keywords for the annotations of each species are listed;
	they can be different for each species.
	See <a href="SystemHelp.html#annot" class="ext" target="_blank">GFF Attributes</a>
	for modifying the keywords shown. The <ttx>Anno Key Count</ttx> can be modified at any time using <tt>symap</tt> (not <tt>viewSymap</tt>).
</table>

<!---------------------------------------------------------------------------->
<a id="singleCol"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>2.2. Single genes columns</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The single genes table only has the
<ttl>Gene Info</ttl> and <ttl>Annotation</ttl> columns, with one additional column, as follows.
<table class="tx">
	<tr><td style="vertical-align: text-top;"><ttx>NumHits</ttx>
	<td>This is the number of hits to the gene in the ENTIRE databases, except for SELF synteny.
</table>
For example,
<br>&rarr; if Arab, Brap and Cabb species have all been compared for synteny,
<br>&rarr; and only Arab-Brap or Arab-Cabb are being queried, they will have some rows with <ttx>NumHits</ttx>&gt;1,
<br>&rarr; and the Arab-Brap-Cabb query will have all <ttx>NumHits</ttx>=0.
<br>This is illustrated below, where gene# 1.2 is the first table
because it does NOT have a hit to Cabb, but it has <ttx>NumHits</ttx>=1 because it DOES have a hit to Brap.
<p>
<table><tr>
	<td><i>Arab-Cabb orphans</i>
	<br><a href="img/query_single_result2.png">
	<img src="img/query_single_result2.png" alt="query single results2" style="border: 1px solid black; height: 130px"></a>
	<td>&nbsp;
	<td><i>Arab-Brap orphans</i>
	<br><a href="img/query_single_result.png">
	<img src="img/query_single_result.png" alt="query single results" style="border: 1px solid black; height: 130px"></a>
	<td>&nbsp;
	<td><i>Arab-Brap-Cabb orphans</i>
	<br><a href="img/query_single_result3.png">
	<img src="img/query_single_result3.png" alt="query single results" style="border: 1px solid black; height: 130px"></a>
</table>

<a id="save"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>2.3 Columns and order shown</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

During a SyMAP session, when you display a new table,
it will use the columns and order  from the last table created or modified (add/remove columns).
<p>The selected columns are saved between sessions (described below), but the order is not.

<h4>2.4 Auto-save columns</h4>
The columns selection is saved in a file called <tt>.symap_saved_props</tt> in the user's home directory so
that the next time you <ttp>viewSymap</ttp>, the table will show the same column (but in their default order).
<p>If you have multiple SyMAP databases, when you change between them the columns displayed are relative to
the last SyMAP database queried (they may seem some what random to a different SyMAP database).

<a id="stat"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3. Statistics</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Statistics for the query results are shown at the bottom of the results table. They can be
hidden by selecting <ttl>Hide Stats</ttl>.
<p><a href="img/query_stats.png"><img src="img/query_stats.png" alt="result stats" style="border: 1px solid black; width: 450px"></a>

<p>Most of the statistics are self-explanatory except the following:
<p><ttx>Annotated</ttx> and <ttx>Genes</ttx>: The first is the number of hits that
overlap one or more genes, where a gene can have multiple hits.
The second is the number of Genes with at least one hit.
<p><ttx>Groups</ttx>: This statistic is only shown if the <ttl>Group</ttl> column is populated.
<br>For <ttl>Multi-hit Genes</ttl>: there is a column with the number of groups with each of the other species.
If there is over 2 species, this will also include the highest group number in parenthesis. For example,
<table>
<tr><td>
<pre>
  A.thal  Groups: #50 (50)    #142 (192)
  B.rapa  Groups: #618 (810)  #984 (1794)
  Cabbage Groups: #847 (2641) #495 (3136)
</pre>
<td>&nbsp;&nbsp;
<td>There are 3,136 total groups.
<br>A.thal has 50 with B.rapa and 142 with Cabbage,
where the first set is 1-50, the second is 51-192, etc.
</table>

<!--------------------------------------------------------------------->
<a id="but"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Top buttons</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

&nbsp;&nbsp;&nbsp;<b>Sections</b>&nbsp;&nbsp;&nbsp;
	<a href="#showRow">1. Show</a>&nbsp;&nbsp;&nbsp;
	<a href="#msaAlign">2. MSA...</a>&nbsp;&nbsp;&nbsp;
	<a href="#view2D">3. View 2D</a>&nbsp;&nbsp;&nbsp;
	<a href="#export">4. Export...</a>&nbsp;&nbsp;&nbsp;
	<a href="#report">5. Gene Report...</a>&nbsp;&nbsp;&nbsp;
	<a href="#search">6. Search...</a>
<p>
<a href="img/query_buttons.png"><img src="img/query_buttons.png" alt="query buttons" style="border: 1px solid black; width: 600px"></a>
<p>The <ttl>Unselect All</ttl> unselects any selected rows.

<a id="showRow"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.1 Show</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
For the selected row, a popup will show all columns and associated information for the hit.
The text in the popup can be copied.

<a id="msaAlign"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.2 MSA...</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Below is an image of a multiple alignment. The right half shows the sequence, where the
first row is the consensus: A 'n' (purple) is used where there is no majority base.
Lowercase (green) is used if there is a least one mismatch. The scroll bar is not shown, as it is on the bottom of the panel.
<a href="img/query_msa.png"><img src="img/query_msa.png" alt="query MSA" style="border: 1px solid black; width: 770px"></a>
<p>Select one or more rows from the query table. The sequences of the selected hit(s) are written out and a multiple alignment is created
using either MAFFT (Katoh 2013 MBA:30) or MUSCLE (Edgar 2004 NAR:32). MAFFT can be run with
multiple CPUs and the optional <tt>--auto</tt> option, where <tt>--auto</tt> takes longer but allows MAFFT
to determine the best algorithm to use.
<table class="tx">
<tr><td style="vertical-align: text-top;">
<table>
<tr><td><ttl>Merge overlapping hits</ttl><td>Instead of repeating the alignment of identical regions,
	overlapping hits will be merged. The above shows that Atha.01 has 5 merged hits.
<tr><td><ttl>Use gapless hits</ttl><td>Multiple subhits can be clustered into one hit, where there
are gaps between them. The subhits are concatenated together with 5 N's in between each
non-overlapping subhit. This is important for long genes as found in humans, where the are
often to exons with large gaps covering the introns.
<tr><td><ttl>None of the above</ttl><td>Align from the start coordinate to the end coordinate
of each end of each selected hit.
<tr><td><ttl>Trim</ttl><td>There can be a long set of bases at the beginning or end
of one sequence, these will be removed.
</table>

<td><a href="img/query_align.png"><img src="img/query_align.png" alt="query align menu" style="border: 1px solid black; width: 200px"></a>
</table>

MAFFT is considerably faster than MUSCLE on long sequences; however, MAFFT can take a long
time if aligning over 300,000 bases, which is why the <ttl>Gapless</ttl> option exists.
<table class="tx">
<tr><td><a href="img/query_align_banner.png"><img src="img/query_align_banner.png" alt="query MSA banner" style="border: 1px solid black; width: 420px"></a>
<td>While aligning, the line at the top of the panel will state the number
of bases aligned as shown on the right.
<p>The <ttl>Stop</ttl> does not stop MAFFT or MUSCLE, that must be done manually.
</table>

<a id="msaMultP">
<p>General Notes:
<br>&rarr; <i>Wait for the current MSA to stop running before starting another!</i>
<br>It is not possible to start a new MSA from the same table, but you could start one from a different table;
<i>they both will fail, so do NOT do this!</i>
<p>&rarr; <ttl>Stop</ttl>: This removes the tab from the left panel, but it does <u>not stop</u> the execution of MAFFT or MUSCLE.
<ul>
<li>Their processes will be called <tt>muscle</tt> for MUSCLE and <tt>distbfast</tt> for MAFFT; MAFFT
may have some other process running, but <tt>distbfast</tt> this seems to be the one that takes the most time/memory.
<li>To stop: you can type <ttl>top</ttl> at the command line, copy the 'process number' beside <tt>distbfast</tt> or <tt>muscle</tt>,
then at the terminal, type <tt>kill -p 'process number'</tt>.
<li>Make sure <tt>distbfast</tt> or <tt>muscle</tt> are stopped, as they take a lot of memory!
<br>However, if you do not see it, then it has completed.
</ul>
<p>MAFFT Notes:
<br>&rarr; All executables were removed that did not seem necessary. However, I may have removed
one(s) that are used in certain unusual situations when <tt>--auto</tt> is used. Try again with <tt>--auto</tt> unchecked.
<br>&rarr; On Mac, if you are using MAFFT <tt>--auto</tt>, you may get a "Cannot be verified" for <tt>dndpre</tt>.
See <a href="TroubleShoot.html#macos">MacOS and external programs</a> to fix this.

<a id="view2D"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.3 View 2D</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/query_2D.png"><img src="img/query_2D.png" alt="query 2D" style="border: 1px solid black; width: 300px"></a>
<p>This displays the <ttl>2D</ttl> view for the selected entry (see <a href="#msp">3-chromosomes</a>).
The region displayed can be specified by the drop-down beside the <ttl>View 2D</ttl> button, as follows:

<a id="group"></a>
<table class="ty">
	<tr><td>Option<td>Column*<td>Selected Hit<td>Display Filter
	<tr><td style="white-space: nowrap;"><ttl>Region</ttl>
	<td>N/A
	<td>The hit is padded to each side by the amount indicated in the kb text box.
	<td><ttx>Show all hits</ttx>.

<tr><td style="white-space: nowrap;"><ttl>Collinear</ttl>
	<td><ttx>Collinear</ttx>
	 <td>The entire collinear set of hits for the selected hit will be shown.
	<td><ttx> Show Block and CoSets</ttx>, where<ttx>CoSets</ttx> is collinear sets.

	<tr><td style="white-space: nowrap;"><ttl>Block</ttl>
	<td><ttx>Block</ttx>
	<td>The entire synteny block for the selected hit will be shown.
	<td><ttx>Show Block Hits</ttx>.

	<tr><td style="white-space: nowrap;"><ttl>Group-chr</ttl>
	<td><ttx>Group</ttx>
	<td>The entire group of hits from the <u>two chromosomes</u> of the selected hit will be shown.
	<td><ttx>Show all hits</ttx>
</table>
*<small>The selected row must have a value for the column.</small>

<p><a id="high"></a>
<table>
	<tr><td><ttl>High</ttl> checkbox
	<td>If selected, the selected hit is highlighted in the <tt>Popup-query</tt> color (default magenta).
	The coloring can also be turned off by selecting the <ttl>2D Hit Filter</ttl> <ttx>Hit popop (or Query)</ttx> option.
	<tr><td><ttl>Gene#</ttl> checkbox
	<td>If selected, the <ttx>Gene#</ttx> will be shown beside each gene in the 2D display, else the <ttx>Annotation</ttx> box
	will be shown (see 2D image on lower right).
</table>

<table style="width: 95%"><tr><td style="text-align: left">
<u>Example:</u>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx">
<tr><td style="vertical-align: text-top;">
	<td style="vertical-align: text-top;">
	<p>The image on the right shows a Collinear set of 5.
	After the initial display, the <ttl>2D</ttl> view can be changed as described in the
	<a href="UserGuide.html#control" class="ext" target="_blank">User Guide</a>.
	<p>&nbsp;
	<p>The table below shows results when the <ttx>Group</ttx> column has a value
	(using <ttx>Gene#</ttx>, <ttx>Multi-gene</ttx> or <ttx>Cluster</ttx> search). The <ttl>Group</ttl> option
	can be used with these rows.
	<p><a href="img/query_columns_2D.png"><img src="img/query_columns_2D.png" alt="query 2D columns" style="border: 1px solid black; width: 280px"></a>

	<td><a href="img/query_synteny.png"><img src="img/query_synteny.png" alt="query synteny" style="border: 1px solid black; width: 450px"></a>
</table>

<a id="export"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.4 Export...</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
One or more rows can be selected for the following exports; or if no rows are selected, the entire table is exported.

<table class="tx">
<tr><td>
	<p><ttl>CSV</ttl>:
	Export the rows using the selected set of columns to a CSV format suitable for import into Excel.
	<p><ttl>HTML</ttl>:
	Export the rows using the selected set of columns to a HTML format suitable for viewing on the web
	(e.g. <a href="./img/query_export.html"  class="ext" target="_blank">Example</a>).
	<p>
	<ttl>FASTA</ttl>: (Pairs Only) For each row, the two hit sequences from the <ttx>Hstart</ttx> to <ttx>Hend</ttx>
	are written to file. NOTE: this is a very slow function and takes minutes if many rows are selected.
	<p>The <ttl>Include Row column</ttl> option is available because this column is alway present, but
it may be desirable to not include it in the output.
	<td>
	<a href="img/query_export.png"><img src="img/query_export.png" alt="result export" style="border: 1px solid black; width: 260px"></a>
</table>

<!------------------------------------------------------------------------->
<!-------------------------REPORT------------------------------------------------>
<a id="report"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.5 Gene Report...</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
&nbsp;&nbsp;&nbsp;<b>Contents</b>&nbsp;&nbsp;&nbsp;
	<a href="#rint">Interface</a>&nbsp;&nbsp;&nbsp;
	<a href="#rgene">Gene report</a>&nbsp;&nbsp;&nbsp;
	<a href="#rcoset">Collinear report</a>&nbsp;&nbsp;&nbsp;
	<a href="#rmulti">Multi-hit gene report</a>&nbsp;&nbsp;&nbsp;
	<a href="#rclust">Cluster gene report</a>&nbsp;&nbsp;&nbsp;

<ul>
<li>The report is on the genes in the query table; <i>hits without paired genes are ignored</i>.
<li>Unless the query is for <ttl>Collinear</ttl>, <ttl>Multi-hit</ttl> or <ttl>Cluster genes</ttl>, a <ttl>Gene</ttl>
report will be produced.
<li>This is most relevant when used with &gt;2 species. <i><u>It has not been tested for &gt;4 species.</u></i>
<li><ttl>Gene</ttl>, <ttl>Collinear</ttl> and <ttl>Multi-hit genes</ttl> produce a reference-based report whereas
the <ttl>Cluster genes</ttl> does not; however <ttl>Cluster genes</ttl> shares most of the same options.
</ul>
<a id="rint"></a>
<h4>Reference-based Interface</h4>
<p>The menu on the lower-right shows the options that will always exist.
There may be additional options in between the middle lines,
depending on the condition used to create the table, as discussed below.
<table><tr><td>
<ttl>Reference</ttl>: In all cases, a reference must be selected; all reference genes are listed in a column,
and any genes it links to (has a hit) are referenced in the other species columns.

<p><ttl>Gene Annotation Columns</ttl>: One or more annotation keywords can be entered in a comma-delimited list (e.g. <tt>product, ID</tt>); the
keywords must be found in the <tt>All_Anno</tt> keyword column. A column will be created containing the values of all entered keywords.
Columns can be entered for any of the species (not just the reference).
<td style="vertical-align: text-top;"><a href="img/query_report_menu.png">
<img src="img/query_report_menu.png" alt="report menu gene" style="border: 1px solid black; width: 280px"></a>
</table>
<table class="ty">
<tr><td colspan=2><ttx>Gene Annotation Columns</ttx>
<tr><td><ttl>Width</ttl><td>Determines the length of the annotation before wrapping around; relevant to long descriptions.
<tr><td><ttl>Truncate</ttl><td> Truncate long description using the width specified.
<tr><td><ttl>Gene#</ttl><td>Show the <ttx>Gene#</ttx> before the gene annotation text.
<tr><td colspan=2><ttx>Show</ttx>
<tr><td><ttl>Border</ttl><td>Draw a border around each cell of the table. Otherwise, there will only be a line
before each group of genes.

<tr><td colspan="2"><ttx>Create</ttx>
<tr><td><ttl>Popup</ttl><td>Displays a panel of the results, which will look just like viewing the HTML file.
<tr><td style="white-space: nowrap;"><ttl>HTML File</ttl><td> Writes a file that can be viewed as a web page.
It is written in a human readable form such that anyone with HTML knowledge can edit it.
<tr><td style="white-space: nowrap;"><ttl>TSV File</ttl><td> Writes a tab-separated-values file that can be viewed with <tt>Excel</tt> or
any editor.
</table>
<p><i>Additional options for &gt;2 species</i>: The following options will be shown between the middle lines if there
are &gt;2 species. The options are explained with the respective report.
<p><i>Link</i>: This term is used to indicate that there is a hit between two species, typically non-reference species.
<p><a href="img/query_report_options.png">
<img src="img/query_report_options.png" alt="report menu options" style="border: 1px solid black; width: 720px"></a>
<p>For all reports:
<ul>
<li>The reference column is in <i>italics</i>. In the TSV file, the reference header has an '*'.
<li>The SyMAP <ttx>Gene#</ttx> is used in the report, which provides the chromosome number and order of genes.
<li>The annotations correspond to the genes listed for the species.
A "---" indicates that there is a gene with no corresponding annotation.
If there is more than one gene in a column, the corresponding annotations will be separated
by a ";".
</ul>
The following discussion will all apply to results from 3 or 4 species.
<!--------------------------------------->
<a id="rgene"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Gene Report</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<table class="ty">
<tr><td style="white-space: nowrap;" colspan="3"><ttb>Gene</ttb> - additional options for &gt;2 species
	<tr><td style="white-space: nowrap;">
		<ttx>All species</ttx><td><ttl>Per Row</ttl><td>Only show rows with all species.
	<tr><td>&nbsp;<td><ttl>Link+</ttl><td>Only show rows with all species and at least one link.
	<tr><td>&nbsp;<td style="white-space: nowrap;"><ttl>All Link</ttl>
		<td>Only show rows with links between all species (this often produces no results).
	<tr><td><ttx>Show</ttx><td><ttl>Links</ttl>
		<td>Show the links. Since <ttl>All Link</ttl> is completely linked, this can be unchecked.
</table>
<br>The report lists all genes from the reference species and what genes each aligns to.
	The image below shows the top of two different <ttl>All Link</ttl> reports, one with
	Hsa17 selected as the reference and the other is with Mus11 selected.
<p>Rows are always merged that share the same reference. The top line shows there are 15 single rows
	 (one gene for each species) and 0 multi rows (multiple non-reference genes for at least one species).
	 There are 2 multi rows in Mus11, where one is shown in the 3rd row.
	<p>A "=" indicates that the gene is linked to the gene next to it (ignoring the Reference column);
	a "+" indicates it is linked to the gene over one column (only happens when there are 4 species).
<p><a href="img/query_report_gene.png">
   <img src="img/query_report_gene.png" alt="report A.thal ref" style="border: 1px solid black; width: 700px"></a>

<!--------------------------------------->
<a id="rcoset"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Collinear Report</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Collinear sets are grouped to show the union of overlapping sets for the reference gene.
<p><table class="ty">
<tr><td colspan="3"><ttb>Collinear</ttb> - additional options for &gt;2 species
	<tr><td style="white-space: nowrap;"><tt>All species</tt><td><ttl>Per Union</ttl><td>The reference gene must be in a collinear set with every species.
	<tr><td><ttx>Show</ttx><td><ttl>Collinear</ttl><td>Displays the collinear set beside each gene.
</table>
<table class="tx">
<tr><td  style="vertical-align: text-top;">
	<p>The report on the right shows the first union of the A.thal collinear sets.
	<p>This was generated with <ttl>Per Union</ttl> checked.

	<p>The '1 [3 sets]' indicates it is the first union with 3 collinear sets.
	<p>For the non-reference gene columns, by default, the collinear set size and number (size.setN) is shown.

	<p>For example, row 4:
		<p><table class="ty">
		<tr><td><i>Gene#</i><td style="white-space: nowrap;"><i>Species Chr</i><td style="white-space: nowrap;"><i>Collinear set</i>
		<tr><td>59<td style="white-space: nowrap;">A.thal Chr01<td>6.6, 8.104, 8.1
		<tr><td>5153.a<td style="white-space: nowrap;">B.rapa Chr09<td style="white-space: nowrap;">6.6
		<tr><td>28<td style="white-space: nowrap;">B.rapa Chr10<td style="white-space: nowrap;">8.104
		<tr><td>1<td style="white-space: nowrap;">Cabbage Chr05<td style="white-space: nowrap;">8.1
		</table>
	<p>The other genes in each collinear set are obvious since they share the same size.setN, e.g. #8.1 is shown in 8 rows.

<td><a href="img/query_report_coset.png">
  <img src="img/query_report_coset.png" alt="report A.thal ref" style="border: 1px solid black; width: 380px"></a>
</table>

<!---- Multi-gene Group ---->
<p><a id="rmulti"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Multi-hit gene report</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Shows reference genes that align to at least N genes, where N is the number input on the query panel.

<p><table class="ty">
<tr><td colspan=3><ttb>Multi-hit gene</ttb> - additional options for &gt;2 species
	<tr><td><ttx>All species</ttx><td><ttl>Per Row</ttl><td>Only show rows with all species.
	<tr><td>&nbsp;<td><ttl>Ignore</ttl><td>Show all rows.
	<tr><td><ttx>Show</ttx><td><ttl>Group</ttl><td>Show the group (size-number) above the respective.
</table>
<p>The <ttl>Group</ttl> (size-number) is shown above each group.
This used the <ttl>Gene#</ttl> option so it prefixes the annotations.
<p><a href="img/query_report_multi.png"><img src="img/query_report_multi.png" alt="report A.thal ref" style="border: 1px solid black; width: 700px"></a>

<!---- Cluster Group ---->
<p><a id="rclust"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.5.5 Cluster hit report</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx">
<tr><td>
The cluster report has no reference, as shown on the right (i.e. no radio buttons for selection).
<p><table class="ty">
<tr><td colspan=3><ttb>Cluster gene</ttb> - additional options (for any number of species)
	<tr><td><ttx>Display</ttx><td><ttl>Linked rows</ttl>
		<td>Each row will be linked genes. Hence, a gene can be listed in multiple rows if it
		has multiple unique linkage patterns.
	<tr><td>&nbsp;<td><ttl>Unlinked unique</ttl>
		<td>Only the unique genes are listed along with how many links they have. There is no
		indication of what genes are linked (but this is faster then the above).
<tr><td><ttx>Show</ttx><td><ttl>Group</ttl><td>Show the group (size-group number).
</table>
<td><a href="img/query_report_menu2.png">
  <img src="img/query_report_menu2.png" alt="report cluster options" style="border: 1px solid black; width: 270px"></a>
</table>
In the two images below, the <ttl>Unlinked unique</ttl> is shown on the left and the <ttl>Linked rows</ttl> is
shown on the right.
Each cluster will have all species since the <ttl>Cluster gene</ttl> query included them all (i.e <ttx>Inc(Hsa,Mus,Pan)</ttx>), but not
all genes will be linked; the <ttl>Linked rows</ttl> shows the genes that are linked.
<p><ttl>Unlinked unique</ttl>: Cluster #1 (7-1) shows that there is only one Hsa gene, which has 4 links;
likewise for the one Mus gene; however, there are 5 Pan genes. If there are only two species shown, then this
is an easier way to view the results.
<p><ttl>Linked rows</ttl>: Every gene listed in a row is linked. The first row of 7-1 is the only one that
is linked across all species. There are 229 fully linked rows in the 177 clusters; each one has the number
of fully linked rows in parenthesis next the cluster label. This is the most informative when &gt;2 species
are included.
<p>
<a href="img/query_report_clust.png">
  <img src="img/query_report_clust.png" alt="report cluster" style="border: 1px solid black; width: 700px"></a>

<!--------------------------------------------------------------------------->
<a id="search"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>4.6 Search...</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx">
<tr><td>The <ttl>Search...</ttl> produces a popup as shown on the right.
<p>The columns in the popup
will vary depending on the columns displayed in the table. This will only show
the columns that are searchable. Besides the top 4 shown,
the species <ttx>Gene#</ttx> and all of the <ttx>Gene Annotation</ttx> columns are searchable.
<p><ttl>Search</ttl> results in the table positioning to the row with the first occurrence
of the search string, and the row will be selected.
<table class="ty">
<tr><td><ttl>Last#</ttl><td>This searches on the last number of a "." or "-" delineated string,
i.e. <ttx>Block</ttx>, <ttx>Collinear</ttx>, <ttx>Group</ttx> and <ttx>Gene#</ttx>.
<br>For a <ttx>Gene#</ttx> with a character suffix (e.g. 1.1.a for chr01, gene# 1.a), enter the 1.a.
<tr><td><ttl>Exact</ttl><td>This searches for the exact string.
<tr><td><ttl>Substring</ttl><td>This searches for a substring, which is most relevant for
the <ttx>Gene Annotation</ttx> of 'desc' or 'product'.
</table>
<td><a href="img/query_search.png"><img src="img/query_search.png" alt="Query search" style="border: 1px solid black; width: 230px"></a>
</table>

<hr>
<!--------------------------------------------------------------------------->
<a id="rtab"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2><small>&gt;</small> Results</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/query_result_tab.png"><img src="img/query_result_tab.png" alt="query results tab" style="border: 1px solid black; width: 600px"></a>
<p>All query results are listed under the <ttb>&gt; Results</ttb> tab on the left.
Clicking this tab shows the table of queries illustrated above.
Query results can be displayed by clicking a result on the left panel,
or double-clicking it in the list of results table shown above.
<p>The only way to remove query results from the left tab is by selecting them in this table followed
by <ttl>Remove Selected</ttl>, all remove all with <ttl>Remove All</ttl>.

<!---------------------------------------------------------------->
<a id="msp"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3-chromosomes</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

&rarr; <a href="#report">Report</a> is only way to view more than two species in a query row (described
immediately above).
<br>&rarr;As shown here, it is possible to bring up the 2D display with 3 chromosomes from the query table (released v5.6.2).
<p>
<table>
	<tr>
	<td style="vertical-align: text-top;">
	Select two rows with the following requirements:
	(1) Both genes must exist in each row.
	(2) There must be a shared gene.
	<br><a href="img/query_2D_2row.png"><img src="img/query_2D_2row.png" alt="Select 2 rows" style="border: 1px solid black; width: 280px"></a>
	<p>The two chromosomes with the unique genes may be from the same species or different species.
	<br><a href="img/query_2D_2rowb.png"><img src="img/query_2D_2rowb.png" alt="Select 2 rows" style="border: 1px solid black; width: 240px"></a>
	<p>Select <ttl>View 2D</ttl> with either the <ttx>Region</ttx> and <ttx>Collinear</ttx> options. The
	image on the right used the <ttx>Collinear</ttx> options.
	<p>&nbsp;<p>
	<a href="Details.html#con" class="ext" target="_blank">2D Highlight Conserved Genes</a>
shows how to highlight shared or unique genes in the 2D display of 3 chromosomes.
	<td><a href="img/2d-3d-coset.png"><img src="img/2d-3chr-coset.png" alt="report A.thal ref" style="border: 1px solid black; width: 480px"></a>
</table>




<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>
</td></tr>
</table>
</body>
</html>

