<!DOCTYPE html>
<html>
<head>
<title>SyMAP Queries</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
    a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
    a.ext {font-weight: bold;}
    .tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 5px; text-align: left}
    mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
    ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: Helvetica, monospace; font-size: 14px; font-weight: 600;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Queries
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>

				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

Content:
<table class=tx>
<tr>
<td><a href="#face"><u>Interface</u></a>
<br>&nbsp;&nbsp;&nbsp;
<br>&nbsp;&nbsp;&nbsp;
<br>&nbsp;&nbsp;&nbsp;
<br>&nbsp;&nbsp;&nbsp;
<td>&nbsp;
<td><a href="#query"><u>Queries</u></a>
<br><a href="#gen">General</a>
<br><a href="#sin">Single gene</a>
<br><a href="#pair">Pair hits</a>
<br><a href="#pfs">PgeneFs</a>
<td>&nbsp;
<td><a href="#result"><u>Results</u></a>
<br><a href="#tab">Result table</a>
<br><a href="#pairCol">Pair hits columns</a>
<br><a href="#singleCol">Single gene columns</a>
<br><a href="#save">Auto-save columns</a>

<td style="vertical-align: text-top;">&nbsp;
<br><a href="#but">Top button functions</a>
<br><a href="#stat">Statistics</a>
</table>
See <a href="./UserGuide.html#term" class="ext" target="_blank">Terminology</a> in the User Guide.

<hr>
<a id="face"></a>
<h2>Interface</h2>
To open the query interface, first select two or more sequence projects in the Project Manager.
Then select the <ttb>Queries</ttb> button to open the queries interface.

<p>
<a href="img/query_oview.png"><img src="img/query_oview.png" alt="query_oview" style="border: 1px solid black; width: 500px"></a>
<p>
The <ttl>Instructions</ttl> window (above) lists the projects which were selected for querying. The last
line states what the <ttx>Olap</ttx> column represents; if one or more project pairs used <ttx>Algo1</ttx>, it will be gene overlap;
if all project pairs used <ttx>Algo2</ttx>, it will be exon overlap
(see <a href="./UserGuide.html#term2" class="ext" target="_blank">Cluster Hit Algo1 versus Algo2</a>).

<p>Open the <ttl>Query Setup</ttl> window by clicking on its tab in the left panel.
Set up the desired filters and then select <ttb>Run Search</ttb> to execute it.
When the query is complete, the <a href="#result">table of results</a> will be displayed.

<a id="query"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Queries</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p>
<a id="qresults"></a>
<a href="img/query_query.png"><img src="img/query_query.png" alt="query_query" style="border: 1px solid black; width: 550px"></a>

<a id="rules"></a>
<p><b>Rules</b>
<table>
<tr><td style="vertical-align: text-top;">
1. All queries are on the <ttl>Pair hits</ttl> unless <ttl>Single</ttl> is selected.
<p>2. All <ttl>Pair hits</ttl> queries return a row per <i>hit pair</i>, i.e.
aligned region between to two chromosomes from two projects.
<p>3. A hit can align to multiple genes on either or both chromosomes. Only the major gene hit(s) will be shown
except for the following queries: <ttx>Annotation</ttx>, <ttx>Every*</ttx>, <ttx>Hit#</ttx>,
<ttx>Gene#</ttx> when no suffix is provided, and <ttx>Multi</ttx> with <ttx>minor</ttx> checked, which
will also show the minor hit genes (see <a href="UserGuide.html#term2"  class="ext" target="_blank"> Major vs minor</a>).
For example, in the image on the right, the red highlighted gene is the major gene for the highlighted hit, whereas the purple
single exon gene is a minor gene.
<p>4. Most filters can be used in conjunction with other
filters; options will be disabled if they cannot be used with a selected filter.
<td>
<a href="img/queryMultOlap.png"><img src="img/queryMultOlap.png" alt="Hit overlaps two genes" style="border: 1px solid black; width: 35px"></a>
</table>

<a id="gen"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>1. General</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ttl>Annotation Description</ttl>
<br>Enter a substring: the entire annotation string (i.e. column <ttl>All_Anno</ttl>) will be searched
for the substring. Hits will be returned that align to the genes with the corresponding annotation.

<p><ttl>Chromosome/location</ttl>
<table class="tx">
<tr><td style="vertical-align: text-top;">Chr<td>&nbsp;
	<td>Select a specific chromosome for the species. This is the most common query to use in
	conjunction with others.
<tr><td style="vertical-align: text-top;">From<td>&nbsp;
	<td>The <ttx>Start</ttx> coordinates for the selected chromosome will be &gt;= this number.
<tr><td style="vertical-align: text-top;">To<td>&nbsp;
	<td>The <ttx>End</ttx> coordinates for the selected chromosome will be &lt;= this number.
<tr><td>&nbsp;<td>&nbsp;<td>
For <ttx>Pair hits</ttx>, the <i>hit</i> coordinates must pass the start and end.
<br>For <ttx>Single genes</ttx>, the <i>gene</i> coordinates must pass start and end.

</table>

<!--------------------------------------------------->
<a id="sin"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2. Single gene</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The <ttl>Single</ttl> queries produces rows of genes; there is no hit or block
information since the rows do not represent pairs.

Options:
<ul>
<li><ttl>Orphan genes (no hits)</ttl>
<br>
Genes that do not a have a hit and meet the additional filters.
The orphan genes are relative to the projects shown on the <ttl>Instruction</ttl> page.
For example, if species X, Y and Z have synteny computed between all pairs,
but only X and Y are selected, the orphan genes for X would be those with no hits to Y.
If X,Y&Z are selected, the orphan genes for X would be those with no hits to Y and Z.
<br>&nbsp;
<li><ttl>All genes (w/o hits)</ttl>, i.e. genes with and without hits
<br>This shows all genes that meet the additional filters, regardless if they have a hit or not.
There is always the same set of genes for a project, regardless of synteny.
</ul>

<p><u>Allowed filters for singles</u>: both options can be restricted by:
<ul>
<li>Using the <ttl>Project</ttl> pull-down, select a project.
<li>If a project is selected, then a chromosome and optional location can be selected.
<li>A <ttl>Annotation Description</ttl> may be entered.
</ul>
<!--------------------------------------------------->
<a id="pair"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3. Pair hits</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Each hit connects two <i>species (projects)</i> and hence represents a <i>pair</i> of aligned
regions for two of the selected species. Filters are as follows:

<p><ttl>In Block (Synteny Hit)</ttl>
<table class="tx">
<tr><td  style="vertical-align: text-top;">Yes<td><td>&nbsp;<td>Only hits that are part of a synteny
block will be returned. All hits will have a value for the <ttx>Block</ttx> column.
<tr><td style="vertical-align: text-top;">No<td><td>&nbsp;<td>Only hits that are NOT part of a synteny
block will be returned. No hits will have a value for the <ttx>Block</ttx> column.
</table>

<a id="genehit"></a>
<p><ttl>Annotated (Gene Hit)</ttl>
<table class="tx">
<tr><td style="vertical-align: text-top;">Every<td><td>&nbsp;
<td>Only hits that align to a gene on one or both sides of the hit will be shown.
The <ttx>Gene#</ttx> columns will list the best gene.

<tr><td style="vertical-align: text-top;">Every*<td><td>&nbsp;
<td>This is like the <ttx>Every</ttx> option, but a hit may be listed multiple times if it aligns
to overlapping genes; in other words, ALL genes will be shown if they have any alignment.
The <ttx>Gene#</ttx> for non-best hits will be suffixed with an "*" and the <ttx>Htype</ttx> is
unknown.

<tr><td style="vertical-align: text-top;">One<td><td>&nbsp;
<td>Only hits that align to a gene on ONE side of the will be shown.
The <ttx>Gene#</ttx> columns will show the best gene for the side with the hit.

<tr><td style="vertical-align: text-top;">Both<td><td>&nbsp;
<td>Only hits that align to genes on BOTH sides will be shown.
The <ttx>Gene#</ttx> columns will show the best gene for both sides of the hit.

<tr><td style="vertical-align: text-top;">None<td><td>&nbsp;<td>Only hits that do NOT align to a gene
on either side of the will be shown. The <ttx>Gene#</ttx> columns will not have a value for either side of the hit.
</table>

<p><ttl>Collinear size &gt;= N</ttl>
<br>List all hits in collinear sets that have size &gt;= N.
See <a href="Query.html" target="_blank" class="ext">Collinear</a>, which explains
the SyMAP collinear sets. There are also options to select for collinear sets with size = N or &lt;= N.

<p><ttl>Multi &gt;= N</ttl>
<br>List all hits that align to genes with &gt;= N hits. Check <ttl>Minor*</ttl> to include minor hits.
<br>This query with <ttl>Minor*</ttl> checked is the same as <ttl>Every*</ttl> with a hit limit.
<br>This can be used in conjunction with other queries, e.g. <ttl>Multi &gt;= 5</ttl> and
<ttl>Both</ttl> will show all g2 hits with at least 5 hits.  Used in conjunction with <ttl>PgeneF</ttl> produces a nice grouping.

<p><ttl>Block#</ttl>
<br>Enter a single block number (<u>the <ttx>Block</ttx> column is formatted
Chr.Chr.Block#</u>). Use it in conjunction with the chromosome pull-downs. For example,
if you select Chr 1 from the first project, Chr 2 from a second project, and enter block=3, you will see
hits in block 1.2.3 from the two project chromosomes you selected.

<p><ttl>Collinear Set#</ttl>
<br>Enter a set number (<u>the <ttx>Collinear</ttx> column is formatted Chr.Chr.Size.Set#</u>). All hits in a collinear set have the same Set#, which is the last number of the <ttx>Collinear</ttx> column.
This can be used in conjunction with the chromosome pull-downs (as for blocks).

<p><ttl>Hit#</ttl>
<br>The hits are numbered for each chromosome pair, e.g. there will be a Hit#1 for every chromosome pair that
has hits. If a hit aligns to multiple genes, both major and minor genes will be shown.

<p><ttl>Gene#</ttl>
<br>Enter a <ttx>Gene#</ttx> number (<u>the <ttx>Gene#</ttx> column is formatted Chr.Gene#.suffix</u>
where the suffix may be blank); <i>do not include the 'Chr' number</i>.
<ul>
<li>If only a number is entered, all genes with the numeric prefix will be displayed (including minor hits).
<li>If a number.suffix is entered, the exact gene will be displayed (but no minor hits).
<li>If a <ttl>Chr:</ttl> is selected from the pull-down, the search will be confined to the genes on that specie's chromosome.
</ul>
<!--------------------------------------------------->
<a id="pfs"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>4. Filter putative gene families (PgeneFs)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
<i>Note</i>: this has not been tested for a long time except superficially (i.e. making sure it has the same results as the previous
release).
<p>This is only computed if the <ttl>Computed PgeneF</ttl> is checked.
Using the hits that pass the other filters, SyMAP constructs putative gene families (PgeneFs)
spanning the selected species. This is done by grouping hits which overlap on at
least one genome. <i>Note, if you have more than 6 species selected, this stage can take
an hour or more.</i>
<p>
Each PgeneF is given a number, which is shown in the Query Results table (column name <ttx>PgeneF</ttx>).
The size of the PgeneF is also shown (column <ttx>PgFSize</ttx>).
<p>
Filters using the PgeneF values:

<p><ttl>Include/Exclude</ttl>
<br>
These filters permit searching for gene families shared by one group of species but not
present in another.
<p>
If a species is checked to include, then the PgeneF will only be retained if it includes
at least one hit which hits that species.
<p>
For &gt;2 species only: If a species is checked to exclude,
then the PgeneF will be discarded if any of its hits are to that species.

<p><ttl>For the included species:</ttl>

<table class="tx">

<tr><td style="vertical-align: text-top;"><ttl>No annotation to the included species</ttl>
<td>Find PgeneFs which are not yet annotated. A PgeneF will be discarded if it is annotated on any of
the species which are checked in the <ttl>Include</ttl> line.

<tr><td style="vertical-align: text-top;"><ttl>Complete linkage if included species</ttl>
<td>For &gt;2 species only: Require the PgeneF to be fully linked, i.e. for each pair of species A and B in the group,
there must be a hit linking A to B.

<tr><td style="vertical-align: text-top;"><ttl>At least one hit for each included species</ttl>
<td>For &gt;2 species only: Only PgeneF hits will be shown if they have hits to the included species,
although the PgeneF numbers will reflect groupings created using all hits.
</table>
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<a id="result"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Results</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p>
<a href="img/query_result.png"><img src="img/query_result.png" alt="query results" style="border: 1px solid black; width: 600px"></a>
<table class="tx">
<tr><td>The buttons on the bottom will be <ttb>Select Columns</ttb> and <ttb>Hide Stats</ttb>. If
<ttb>Select Columns</ttb> is selected, it changes to <ttb>Hide Columns</ttb> and the
<ttb>Hide Stats</ttb> is replaced with <ttb>Clear</ttb> and <ttb>Defaults</ttb>.
<td><a href="img/query_lower_buttons.png"><img src="img/query_lower_buttons.png" alt="query results" style="border: 1px solid black; width: 250px"></a>
</table>
<p>Click <a href="img/query_single_result.png">Single Result Table</a> for an example when all three projects
are being queried versus <a href="img/query_single_result2.png"> Results</a> when two of the three are being
queried; note that the <ttl>NumHits</ttl> is something &gt;0 because there are hits to <tt>Arab</tt> that are not
taking into account since it was not part of the query set.

<!----------------------------------------------------->
<a id="tab"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Results Table</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The results table lists the hits (anchors) resulting from the query, where each hit connects
two species.  If the query was for single genes, then each row represents one gene and shows data
only for one species.

<p>Specifics of rows of <ttx>Pair hits</ttx>:
<ul>
<li>Generally a hit only aligns to one gene, possibly on both sides. If there are overlapping genes,
it will show the hit aligned to overlapping genes if any of the following queries are used:
<ttx>Every*, Annotation, Hit#</ttx>, or <ttx>Gene#</ttx> (see <a href="#rules">Rules</a> for a complete explanation).

<li>A gene may be listed more than once if multiple hits align to it with a best overlap.
</ul>

<p>Specifics of columns:
<ul>
<li><ttx>Pair Hits</ttx>: The table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty.
<li><ttx>Single genes</ttx>: If the query specified <ttl>Single</ttl> genes, then each row represents one gene and shows data only
for one species.
</ul>

<a id="col"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Columns</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Sections:
<a href="#pairCol"> Pair hits columns</a>&nbsp;&nbsp;&nbsp;
<a href="#singleCol"> Single gene columns</a>&nbsp;&nbsp;&nbsp;
<a href="#save"> Auto-save columns</a>
<p>You can sort the columns by clicking on them, and rearrange them by dragging the header
boxes. You can add/remove columns using the <ttb>Select Columns</ttb> button at the bottom.

<p>In the columns panel, <i>hover</i> over a column name to see its brief description. Following
are the full descriptions of the columns:
<a id="pairCol"></a>
<h4>Pair hits columns</h4>
<table class="tx">
<tr>
	<td colspan=3><ttl>General</ttl>
<tr>
	<td style="vertical-align: text-top;"><ttx>Row</ttx><td>&nbsp;
	<td>Row number within the table
<tr>
	<td style="vertical-align: text-top;"><ttx>Block</ttx><td>&nbsp;
	<td>Synteny block containing this hit (if any). The format is Cn.Cm.Block#, where
		Cn and Cm are the chromosome numbers.
<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Block Hits</ttx><td>&nbsp;
	<td>The number of hits which comprise the synteny block.
<tr>
	<td style="vertical-align: text-top;"><ttx>Collinear</ttx><td>&nbsp;
	<td>Collinear set containing this hit (if any). The format is Cn.Cm.Size.Set# (e.g. 1.2.5.100; there are 5
		adjacent gene hits in set# 100 on Chr1 to Chr2).
<tr>
	<td style="vertical-align: text-top;"><ttx>PgeneF</ttx><td>&nbsp;
	<td>If <ttl>Compute PgeneF</ttl>: PgeneF number. All hits in a given PgeneF have the same number. Note that
		the number is generated during the search and won't necessarily be the same
		in a different search.
<tr>
	<td style="vertical-align: text-top;"><ttx>PgFSize</ttx><td>&nbsp;
	<td>If <ttl>Compute PgeneF</ttl>: Size of the PgeneF which contains this hit. Sort on this column to view the
		other rows with same PgeneF.
<tr>
	<td style="vertical-align: text-top;"><ttx>Hit#</ttx><td>&nbsp;
	<td>The hit number, which are sequential numbers for the chromosome pair.
	This number shown on the <ttp>Chromosome Explorer</ttp> when the mouse is over the hit line.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Id</ttx><td>&nbsp;
	<td>Percent identity of the alignment.The value of the "Identity" column from the MUMmer file.
	If the hit has subhits, then this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit %Sim</ttx><td>&nbsp;
	<td>Percent similarity of the alignment (as determined by the BLOSUM scoring matrix).
	The value of the "Similarity" column from the MUMmer file. If the hit has subhits, then
	this is an approximation.
<tr><td style="vertical-align: text-top;"><ttx>Hit #Subs</ttx><td>&nbsp;
	<td>The number of sub-hits in a clustered hit.
<tr><td style="vertical-align: text-top;"><ttx>Hit St</ttx><td>&nbsp;
	<td>If "=", both hit ends are to the same strand; if "!=", they are to different strands.
<tr><td style="vertical-align: text-top;"><ttx>Hit Cov</ttx><td>&nbsp;
	<td>The length of the subhits within a clustered hit are summed taking into accounts overlaps; the longest summed subhits
	of both sides is shown. See <i>Clustered Hits</i> in <a href="UserGuide.html#term" class="ext" target="_blank">Terminology</a>.
<tr><td style="vertical-align: text-top;"><ttx>Hit Type</ttx><td>&nbsp;
	<td>There are two alternative algorithms for clustering the hits on database creation.
	<br>&#10139;<ttx>Algo1</ttx>: g2 (two genes), g1 (one gene), g0 (no genes); exon/intron information is not known.
	<br>&#10139;<ttx>Algo2</ttx>: E is exon, I is Intron, n is intergenic. There will be 2 characters, one for each project,
	where the 1st letter goes with the alphabetically lesser project name. For example, 'EE' indicates the hit covers an exon on
	both sides.
<tr>
	<td colspan=3>
		<ttl>Gene&Hit Info</ttl>: one row for each species
<tr>
	<td style="vertical-align: text-top;"><ttx>Chr</ttx><td>&nbsp;
	<td>Chromosome of the hit.

<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Gstart/Gend/Gst</ttx><td>&nbsp;
	<td>Start and end of the annotated gene. The <ttx>Gst</ttx> is the strand (+/-).
<tr>
	<td style="vertical-align: text-top;"><ttx>Gene#</ttx><td>&nbsp;
	<td>The gene number is C.#.{a-z}. The C is the chromosome number.
	The # is the sequential number along the chromosome.
	If a run of genes overlap, they receive the
	same gene number with different suffixes {a-z, a2-z2, etc}. This is shown on the <ttp>Chromosome Explorer</ttp>
	"Annotation Description" and when the mouse hoovers over the gene.
<tr>
	<td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hstart/Hend</ttx><td>&nbsp;
	<td>Start and end of the hit region.
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Hlen</ttx><td>&nbsp;
	<td>Hend-Hstart+1
<tr><td style="vertical-align: text-top; white-space: nowrap;"><ttx>Olap</ttx><td>&nbsp;
	<td>The value depends on which <ttx>Cluster Hit</ttx> algorithm was used.
	<br>&#10139;<ttx>Algo1</ttx>: If any of the project pairs used <ttx>Algo1</ttx>, then this column will represent
	the gene overlap for all project pairs. <font color=green>NOTE:</font> <u><ttx>Algo2</ttx> takes into account gaps between the subhits, whereas
	<ttx>Algo1</ttx> does not</u>.
	<br>&#10139;<ttx>Algo2</ttx>: If all of the project pairs used <ttx>Algo2</ttx>, then this column will represent
	the exon overlap.
<tr>
	<td colspan=3><ttl>Annotation</ttl>: one row for each species
<tr>
	<td colspan=3>The keywords for the annotations of each species are listed;
	they can be different for each species.
	See <a href="SystemHelp.html#annot" class="ext" target="_blank">Project Parameters</a>
	for modifying the keywords shown
</table>


<a id="singleCol"></a>
<h4>Single genes columns</h4>
<p>The single genes tables only have the
<ttl>Gene Info</ttl> and <ttl>Annotation</ttl> columns, with one additional column, as follows.
<table class="tx">
<tr><td style="vertical-align: text-top;"><ttx>NumHits</ttx>
<td>This is the number of hits to the gene in the ENTIRE databases, except for SELF synteny.
<br>For example, if X, Y
and Z species have all been compared for synteny, but only Y and Z are being queried,
the <ttl>Orphans</ttl> for Y can have <ttx>NumHits</ttx>&gt;0 if it hits X. This
is illustrated <a href="img/query_single_result.png">Single genes</a>.
</table>

<a id="save"></a>
<h4>Auto-save columns</h4>
The columns selection is saved in a file called <tt>.symap_saved_props</tt> in the user's home directory.
If you only have one SyMAP database, there is no need to read the following:

<p>Say the last database D<sub>n</sub> queried had N species,
and the current database D<sub>m</sub> has M species, where D<sub>n</sub> and D<sub>m</sub> are different
SyMAP databases:
<ol>
<li>If N=M: the <ttx>General</ttx> and <ttx>Gene&Hit Info</ttx> columns will
be set according to the previous settings, but the <ttx>Annotation</ttx> columns may be wrong.
<li>If N&gt;M or N&lt;M: the <ttx>General</ttx> will be according to the previous settings,
all <ttx>Gene&Hit Info</ttx> columns will be set like the first D<sub>n</sub> species, and no
<ttx>Annotation</ttx> columns will be set.
</ol>
This feature was implemented in v5.3.2.
<a id="but"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Top button functions</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/query_buttons.png"><img src="img/query_buttons.png" alt="query buttons" style="border: 1px solid black; width: 600px"></a>
<p><ttb>Show</ttb>: For the selected row, a popup will show all columns and associated information for the hit.
The text in the popup can be copied.

<table class="tx">
<tr><td style="vertical-align: text-top;">
<ttb>Align</ttb>:
Select one or more rows. The sequences of the selected hit(s) are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32). The figure on the right shows the MUSCLE alignment of
two genes from B.rapa and genes from Cabbage and A.thal.
<td>
<a href="img/query_muscle.png"><img src="img/query_muscle.png" alt="query muscle" style="border: 1px solid black; width: 450px"></a>
</table>

<ttb>View 2D</ttb>: This displays the <ttp>2D</ttp> view for the selected entry. The following 3 views
are provided using the drop-down (the image below shows a Collinear set of 5).
<table class="tx">
<tr><td style="vertical-align: text-top;">

<table class="ty">
<tr><td style="white-space: nowrap;"><ttl>Region (kb)</ttl>
<td>The selected hit is padded to each side by the amount indicated (default 50kb).
The default filter is <ttl>Show all hits</ttl>.
<tr><td style="white-space: nowrap;"><ttl>Synteny block</ttl>
<td>The entire block for this hit will be shown. The default filter is <ttl>Show Block Hits</ttl>.
<tr><td style="white-space: nowrap;"><ttl>Collinear set</ttl> <td>If the hit is in a collinear set, the set will be shown.
The default filter is <ttl>Highlight Collinear Set; Show Block and Set Hits</ttl>.
</table>

<td><a href="img/query_synteny.png"><img src="img/query_synteny.png" alt="query synteny" style="border: 1px solid black; width: 450px"></a>
</table>
After the initial display, the <ttp>2D</ttp> view can be changed as described in the
<a href="UserGuide.html#control" class="ext" target="_blank">User Guide</a>.
<p><ttl>High</ttl>: If this is selected, the selected hit is highlighted in red;
the selected color is the same as the hover color, and can be changed
in the <a href="UserGuide.html#wheel"  class="ext" target="_blank">Color Icon</a>.

<a id="export"></a>
<p><ttb>Export</ttb>: One or more rows can be selected for the following exports;
or if no rows are selected, the entire table is exported.
<table class="tx">
<tr><td>
<p><ttl>CSV</ttl>:
Export the rows using the selected set of columns to a CSV format suitable for import into Excel.
<p><ttl>HTML</ttl>:
Export the rows using the selected set of columns to a HTML format suitable for viewing on the web
(e.g. <a href="./img/query_export.html"  class="ext" target="_blank">Example</a>).
<p>
<ttl>FASTA</ttl>: (Pairs Only) Sequences from the rows are written to a FASTA file.
Both sides of each hit are written using the start/end coordinates shown in the table.
NOTE: this is a very slow function and takes minutes if many rows are selected.
<td>
<a href="img/query_export.png"><img src="img/query_export.png" alt="result export" style="border: 1px solid black; width: 260px"></a>
</table>
The <ttl>Include Row column</ttl> option is available because this column is alway present, but
it may be desirable to not include it in the output.

<a id="stat"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Statistics</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
Statistics for the query results are shown at the bottom of the results table. They can be
hidden by selecting <ttl>Hide Stats</ttl>.
<p><a href="img/query_stats.png"><img src="img/query_stats.png" alt="result stats" style="border: 1px solid black; width: 450px"></a>

<p>Most of the statistics are self-explanatory except the following:
<p><ttx>Annotated</ttx> and <ttx>Genes</ttx>: The first is the number of hits that
overlap one or more genes, where a gene can have multiple hits.
The second is the number of Genes with at least one hit.
<p><ttx>Regions</ttx>: This statistic
is only shown if <ttl>Compute PgeneF</ttl> was checked, and it is
the number of distinct regions covered on that species.


<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

