<!DOCTYPE html>
<html>
<head>
<title>SyMAP Release Notes</title>
<link rel="stylesheet" href="style.css">
<style>
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: Black;}
	a:link {text-decoration: none;color:#6666CC;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Release Notes
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<h3>Release: v5.1.6 (20Sept22)</h3>
<ul>
<li>Blocks
<ul>
<li>For the single chromosome view: Inverted blocks have dotted outline.
<li>The windows are better sized.
<li>Fixed rare bug: For the initial multi-chromosome view: If there are many chromosomes, they were being cutoff; now the scroll bar allows them all to be shown.
</ul>
<li>2d Display
<ul>
<li>The full gene description can be viewed by right clicking on a gene annotation in the track;
see <a href="UserGuide.html#view" target="_blank" class="ext">UserGuide</a>.
<li>Hover over a hit may include "Inv" (inverted block) and will include "gNrM" (N exon hits, M colinear run size);
see <a href="UserGuide.html#view" target="_blank" class="ext">UserGuide</a>;
</ul>
<li>SyMAP Query
<ul>
<li>Add general columns <ttl>Hit %Id</ttl>, <ttl>Hit %Sim</ttl>, <ttl>Hit #Merge</ttl>
<li>Add species column <ttl>Len</ttl> which is the length of the hit (or merged hit).
<li>Bug Fix: <ttl>Show Synteny</ttl>: the hover over the hit only showed the correct hit <ttl>Information</ttl>
for those below the selected hit - that is fixed.
<li>Gene# of 0 will sort to bottom of table with a '-'.
<li>MUSCLE alignment: (1) Fixed rare bug. (2) The alignment files will be left in the <ttl>muscle</ttl>
directory, but over-written with each new alignment.
</ul>
<li>Alignment&Synteny: (2 tiny things)
<ul>
<li>If promer or nucmer are over-rided in the <ttl>Parameters</ttl>, it will say on the Summary page (unless the
alignment is re-used).
<li>Fixed rare bug: if the <ttl>Parameters</ttl> are changed before <i>any</i> alignment has been done, there is an error
because the <ttl>data/seq_results</ttl> directory has not been created.
</ul>
</ul>


<h3>Release: v5.1.5 (13Sept22)</h3>
<ul>
<li>Online Docs - the biggest change of this release is an updating of the documentation.
<li>Alignment & Synteny
<ul>
<li>Clear Pair provides option to "only delete synteny from database" or "delete synteny and remove files".
<li>The MUMmer %Sim column is now loaded in addition to the %Id column.

<li>The Hit# will be assigned sorted on the query start coordinate.
</ul>
<li>Chromosome Explorer
<ul>
<li>Hover over Gene: the gene information will be shown when hovering anywhere within the gene's rectangle.
(It used to only show Gene information if the mouse was hovered directly over the center.)
<li>The Hit information now has %Id and %Sim, plus the number of merged hits.
</ul>
<li>Blocks - the chromosome blocks view has a smaller window.
<li>demo_seq2 - Removed duplicate genes (worked okay, but technically wrong).
<li>Developers
<ul>
<li>If all hits are removed (e.g. Clear Pair), the index will be reset so that the next loaded hit will have a HitIdx (Hit #)
that starts at 1.
<li>The alignment files may have header lines. Any line that does not start with a digit will be ignored.
<li>Removed some dead tables from Schema and changed a few datatypes.
</ul>
</ul>

<h3>Release: v5.1.4 (4Sept22)</h3>

<ul>
<li>SyMAP Query
<ul>
<li>Filters:
<ol>
<li>Added a <u>General</u> section for the annotation and location, since they can work with all other queries.
Made the <u>Single</u> project (Orphan genes and all genes) a separate section.
<li>Add a filter to <u>Show genes w/o hits</u> for a project, which shows all genes.
<li>Changed the <u>Orphan genes</u> filter to show orphans relative to the chosen projects (it was across all synteny
project, even projects selected for the SyMAP query).
<li>The Hit# query did not work in some situations.
</ol>
<li>Improved the query summary. Also, the location had been left off the summary for most filters.
<li>Replaced the result table <u>#Gene</u> column: it was computed per query, whereas the replacement is stored in the database.
The replacement is the order number of the gene along the chromosomes, which stays constant across all queries
and is shown on the Chromosome Explorer Annotation and the hover of the gene.

</ul>
<li>Chromosome Explorer
<ul>
<li>Sequence Filter: If a wrong coordinate was entered, the program exited.
<li>Alignment Bug Fix: only the first part of the alignment was working (fairly recent bug).
</ul>
</ul>

<h3>Release: v5.1.3 (30Aug22)</h3>
There is no major new development, but small interface improvements and SyMAP Query has a small bug fix.
<ul>
<li>Main window
<ul>
<li>Projects listed on left are now in alphabetical order per category. If the project is loaded
into the database, the load date is shown to its right. If it has computed synteny, the number of syntenies is in brackets.
<li>The symap window is centered, along with the Chromosome Explorer and SyMAP Query window.
<li>viewSyMAP: only shows the projects that have synteny computed.
</ul>
<li>Load Annotation
<ul>
<li>grp_prefix: This can now be removed from existing projects, i.e. if the 'Chr' was left on,
so the 2D display is clustered with Chr01, Chr02, etc, then the 'Chr' can be removed.
<li>annot_kw_mincount: the Help made no sense for this. It has been upgraded to do the following:
the SyMAP Query shows the GFF attribute columns that have &gt;=annot_kw_mincount occurrences.
It can be changed at any time through the symap parameters window.
<li>When the annotation is loaded and the attribute (column 9) is parsed for keywords, the only keywords
considered are those followed by a '=' (it used to allow spaces between keyword and value).
</ul>

<li>2D chromosome view
<ul>
<li>The popup annotation is now a smaller constant sized scrollable window.
<li>The annotation hover and popup coordinates are more readable.
</ul>

<li>SyMAP Query
<ul>
<li>The Block query is now on just the block number along with the selection of chromosomes.
<li>Small changes:
<ul>
<li>The Results Panel includes the number of rows in the query.
<li>The query will not run if there is an error in the input, and there is better error-checking.
<li>The window is smaller and the columns panel in the table is fixed width.
<li>When the full block name is shown, any leading zeros are removed, e.g. 01.03.2&rarr;1.3.2.
</ul>
<li>Bug fix: The Orphan query with one or more chromosome selected gave wrong results (it was obvious); moreover,
if a start or stop was entered, it gave an error. This is fully functional now.
</ul>

<li>Convert
<ul>
<li>Both scripts work if there is not a gff file (they were requiring it).
<li>For both scripts, the bases that are not ACGT are summarized.
<li>ConvertNCBI - anything that does not have "chromosome" in its ">" line is considered a scaffold.
(ConvertEnsembl already does this).
</ul>
</ul>


<h3>Release: v5.1.2 (19Aug22)</h3>
The biggest changes are with the <tt>ConvertNCBI</tt> and <tt>ConvertEnsembl</tt> scripts.
<ul>
<li>Annotation
<ul>
<li>Input format:
<ol type='i'>
<li>The <tt>ConvertNCBI</tt> and <tt>ConvertEnsembl</tt> scripts created separate <tt>gene.gff</tt> and <tt>exon.gff</tt> files; now they
create one file. The <tt>ConvertEnsembl</tt> was not producing a <tt>gap.gff</tt> file, and now it is. Both have been
updated to be aware of more input variations and provide a better summary. See <a href=convert class="ext" target="_blank">Convert</a> for details.
<li>The Demo_seq has separate <tt>gene.gff</tt> and <tt>exon.gff</tt> files; now it has one file.
<li>SyMAP can still load separate gene and exon files, but its association of exon to gene is not as accurate.
</ol>

<li>Schema change
<ol type='i'>
<li>An existing database will be updated the first time it is viewed.
<li>This adds a MySQL field to associate Exons with their Gene; for an existing database,
the annotation would need to be reloaded for this to happen and be displayed (see 2d Gene Popup below).
</ol>
<li>Load Annotation
<ol type='i'>
<li>The code was treating 'CDS' as 'exon', now only exons are loaded.
<li>The attributes were not being loaded for Exon; now, the "parent" attribute is loaded for the
exon and shown when it is moused over.
<li>Quotes are removed from attributes.
<li>The exons are now being mapped to the parent gene to be used in the 2D display.
</ol>
</ul>
<li>Chromosome Explorer
<ul>
<li>Tiny bug fix: the "Circ" button get stuck in un-active mode.
<li>2D display
<ul>
<li>Tiny bug fix: For a single contiguous hit for a flipped chromosome, the hit length graphic was
not aligned right.
<li>When the Gene or Exon is moused over, it will include the number of exons or exon#, respectively.
<li>When the hit line is moused over, it will include the project name, start and end coordinates.
<li>2d Gene Popup: Includes the exons for the gene.
</ul>
</ul>
<li>SyMAP Query
<ul>
<li>Add warnings for bad block or collinear input.
</ul>
<li>Developers
<ul>
<li>The code is now Java v8 compliant (no depreciated statements).
</ul>
</ol>
</ul>

<h3>Release: v5.1.1 (31July22)</h3>

These changes are to 'tidy up' things; nothing major.
<ul>

<li>Scripts:
<ul>
<li>The <tt>symap</tt> and <tt>viewSymap</tt> scripts have been changed from Perl to shell.
<li>The <tt>viewSymap3D</tt> script has been removed, but the <a href="./3D.html" class="ext" target="_blank">3D</a> documentation explains
how you could recreate the 3D view.
</ul>
<li>Data directory
<ul>
<li>The <tt>data/fpc</tt> has been tarred and zipped (<tt>fpc.tar.gz</tt>) since it is rarely used anymore.
If the <tt>data/fpc</tt> directory does not exist, then the BLAT args are not shown on
the "Param" window. The FPC-Seq alignments still work.
<li>If the <tt>data</tt> directory does not exist, it is created: (1) on "Add Project",
(2) when redoing an alignment in the database. (The directory will exist if using <tt>symap_5.tar.gz</tt>,
but it will not exist if starting from the code base).
</ul>
<li>Slight display change
<ul>
<li>The Summary shows the following: (1) for a project, if it used mask_genes or
a different min_size then 100000, (2) for an alignment, if the "mummer_path" is set,
or "concat" is unset, or the promer/nucmer parameters are changed. Only creates/updates on
running the synteny.
</ul>
<li>Small changes
<ul>
<li><tt>symap -v</tt> will display important MySQL variables and make suggestions.
<li>It does not try to load hidden files.
<li>It does not allow "cancel" on remove operations.
<li>It displays length of log files (they are appended to, so can become large).
<li>The project parameters <tt>annot_keywords</tt> and <tt>annot_kwy_mincount</tt> are now being saved to the database
so their values are shown correctly in the project parameters window on re-display.
<li>A proper message is given when <tt>viewSymap</tt> tries to read a databases (<tt>symap.config</tt>) that does not exist.
<li>A stacktrace will not be printed to the terminal, only to the error.log (the e.getMessage() sometimes includes a partial
stack).
<li>Developers: removed more dead code.
</ul>
</ul>
Note: this was first released on 27July22, then again on 31July22 with a slight change.


<h3>Release: v5.1.0 (12June22)</h3>
The documentation has
been moved from www.agcol.arizona.edu/software/symap to csoderlund.github.io/SyMAP.
Any references to the online documentation within the code has been updated.
For pre-v5.1 releases, the www.agcol.arizona.edu/software/symap/docs still exists.

<p>The html Help files can now have URL links in them.

<h3>Release: v5.0.8 (1June22)</h3>
<ol>
<li>A new option "Concat" has been added; when this is unchecked, it will not create a
concatenation of the first genome to run against all chromosomes of the second, instead
it perform a chromosome-by-chromosome search.
<li><u>MUMmer4</u> can be used by compiling it on your machine from the /ext directory
and adding its path to the "symap.config" file. See <a href=MUMmer.html>Running MUMmer</a>.
<li>When computing alignment&amp;synteny, all stacktrace errors are now written into error.log.
<li>The database name is prefixed to the "logs/LOAD.log" file name.
<li>A new document has been created for when there is problems running
<a href=MUMmer.html>MUMmer</a> from SyMAP.
<li>ConvertNCBI script was writing the scaffold sequences to end of the last chromosome (this
did not make the annotation or synteny wrong - just extra sequence).
</ol>

<h3>Release: v5.0.7 (4Feb22)</h3>
<ol>
<li><u>MUMmer</u> has been updated since the last SyMAP release; the code in the /ext/mummer4 directory
has been updated for the latest MUMmer. The documentation has been updated accordingly
(see <a href=MUMmer.html>Running MUMmer</a>).
<li><u>ConvertNCBI</u>: (1) Include ncbi_dataset format. (2) Convert hex to character.
<li><u>Parameters</u>: (1) The Load would not work if any numeric parameter was set to blank.
(2) For Alignment&amp;Synteny "Parameters", the Global button was removed as it did not work
except under weird conditions.
<li><u>Help</u>: the viewer Help pages link to the pages at www.agcol.arizona.edu/software/symap. The
approach for Mac was becoming obsolete, so it was changed. For Linux, more supported browsers were added.
<li>Improve documentation.
<li>Developers: Removed all Applet code (was disabled previously, but code was in everything).
</ol>

<!----------------------------------------------------------->
<h3>Release: v5.0.6 (2June20)</h3>
<ol>
<li>Some changes were necessary for SyMAP to run with MySQL v8.
	<ul><li>The schema needed a change, which will occur on first reading of database.
	</ul>
<li>SyMAP works on MacOS Catalina 10.15
	<ul><li>It checks each external program to see if it is executable, as this is not automatic with 10.15
	(see <a href="TroubleShoot.html#ext">Trouble Shoot External Programs</a>).
	</ul>
<li>SyMAP can be compiled with Java v14 (some changed were necessary for this).
<li>The Applet has been removed and will no longer be supported. (It is in earlier SyMAP versions, and can
be obtained at <a href="https://github.com/csoderlund/SyMAP">github</a>).
<li>The built-in MySQL database on Linux is no longer supported.
<li>The Alignment&amp;Synteny has a couple checks for 'ordered_against' before processing:
	(1) Is the order_against the same as the other selected project.
	(2) Does the ordered directory already exist.
</ol>

<h3>Release: v5.0.5 (2May20)</h3>

<ol>
<li>Ordering algorithm: The main change is a rewrite of the ordering algorithm, which orders draft contigs
against a completed sequence closely related genome. The previous algorithm sometimes
would put many of the placed contigs incorrectly into 'chrUNK', though the
ordering in the <tt>anchored.csv</tt> file was correct.

<li>SyMAP Manager:
<ul>
<li>Show #genes and #gaps in summary
<li>Add Project: On Linux, the "Add" button would stay disabled, this has been fixed.
</ul>

<li>Chromosome Explorer: Mouse over a hit and the block number will be shown along with the other information.

<li>SyMAP Query:
The "Align Sequence" display now show locations in text box under alignment.
The v5.0.4 introduced a bug where the row selections display were not right, which
is fixed.

<li>Small changes:
<ul>
<li>Running 'symap -v' will check the MySQL variables needed to run at a reasonable speed,
and will output to terminal any variables that need changing.
<li>Block View: When viewing the blocks using VNC, the chromosome that the
	blocks were aligned to was not visible, which has been fixed.
<li>All Pairs button: Is only enabled if there is anything that will be executed when selected.
<li>Load Project: If a file of 100's of draft sequences are loaded, this was flagged as an error (but
still loaded); it no longer called an error, though a message is written to terminal.
</ul>
</ol>

<h3>Release: v5.0.4 (18Apr20)</h3>
<ol>
<li>SyMAP Query has had major changes:
<ol>
<li>It is faster, e.g. finding the hits in synteny blocks for 5 plant genomes did take 9m:32s
and now takes 0m:09s.
<li>The query page has changed to allow more versatile queries.
<li>The PgeneF is only computed if the "Compute PgeneF" option is checked; otherwise, there
is no changes to this function or options.
<li>Some of many minor changes:
<ul>
<li>Selecting a chromosome for a species (or multiples) now only shows hits that involve the
selected chromosomes.
<li>The #RGN column has been removed and a new "Has Gene" column has been added. Note, all species
with a gene overlapping the hit should have 'annotation', but this makes it easier to see in a glance.
Plus, if some genes do not have an ID or some other keyword, it can be confusing looking at the
annotation.
<li>The "Save for Reload" and "Reload Results" have been removed, as it is so much faster, this
does not seem necessary anymore.
<li>Both the Overview and Table Summary have small changes.
</ul>
</ol>
<li>scripts/ConvertNCBI has been updated to also create a gap file.
<li>scripts/symap.sql has been removed and the schema has been incorporated into the Java code.
<li>SyMAP Summary: Displays major parameters for projects that are not loaded.
<li>Chromosome Explorer 2D View: The mouse over provides location information in the "Instruction" box,
plus:
<ul>
<li>The following functions can be applied to the selected region of a sequence track:
"Zoom All Tracks", "Zoom Selected Tracks", "Align (Max 30000)", "Show Sequence", where the
first three are not new, but the labels changed for clarity.
<li>The last one, "Show Sequence", has
been added as there was no way to see the underlying sequence; it is simply a popup of the
selected sequence.
<li>The "Align" view has been slightly changed to add comma's in large numbers and the length of the sequence.
</ul>
<li>The symapApplet.jar was not working in recent releases, which has been fixed.
</ol>

<h3>Release: v5.0.3 (11mar20)</h3>
<ol>
<li><i>Chromosome Explorer 2D view</i>. The "Annotation Description" can be right-click to view
a pop-up of the description; the benefit is that sometimes part of the description is
buried, and the pop-up can be copied.
<li><i>SyMap Queries</i>. (1) The filters set are shown at the top of the table.  (2) The "Only
Orphan Genes" option can now only be used with the "Annotation Description", "Location", and
"Include" options.
<li>If access to MySQL does not work, a message is printed with the Java error message.
</ol>

<h3>Release: v5.0.2 (24feb20)</h3>
<ol>
<li>scripts/ConvertNCBI: (1) Produce a hard-masked genome sequence (input is soft-masked) with
-m flag. (2) Includes the mRNA ID in the product attribute. (3) The product attribute has
been improved.
<li>SyMAP Bug fix for using chromosome and scaffolds in the same project.
<br>The following did not work:
Say the genomic  sequence input was a mix of chromosomes (prefix 'Chr')
and scaffolds (prefix 's'), and the Project parameters were prefix 'Chr' and minsize '10000',
where the minsize filtered out annotated scaffolds, SyMAP would quit if there was
annotation on filtered-out scaffolds; this has been fixed. Also, the documentation
was not correct for mixing sequence types (e.g. prefix 'Chr' and 's'); the Project parameter
'grp_prefix' should be set to blank.

</ol>

<h3>Release: v5.0.1 (16feb20)</h3>

<ul>
<li>Load Annotation
	<ul>
	<li>In the Params interface for a project, the user can set what keywords from the
column 9 "Attributes" field of the gff file to load; this was not working correctly
and has now been fixed.
	<li>The type (column 3) of mRNA was being loaded as a gene, which causes duplicate records;
now mRNA's are ignored.
</ul>
<li>Alignment &amp; Synteny
	<ul>
	<li>Though it was not possible to make MUMmer v4 part of SyMAP, it is possible
to install MUMmer v4 in the /ext/mummer directory and it will work. Instructions
are provided in <a href="SystemGuide.html">System Guide</a>.
	<li>The nucmer "--maxmatch" parameter was automatically used for self-synteny when
a sequence file was being compared with itself; this would result in an extremely
long execution time for large files (e.g. chromosomes). Consequently, this
parameter has been removed, but can now be set by the user in the "Parameter" interface.
	<li><i>Pair Parameters</i>: (1) The user can set parameters in "Self Args", which are only
used when a chromosome sequence is being compared to itself. (2) This interface
was very confusing as to when parameters were saved and what would be used;
this should be clearer now, and is explained in the "Help" for the Parameter interface.
	<li>If only part of the alignments finished and it ended pre-maturely, when run again,
it will not redo any finished alignments.
	</ul>
<li>Internally, goobs of commented out code has been removed.
</ul>

<p><a href="#top">Go to top</a>

<h3>Release v5.0 (22jan20)</h3>
New projects cannot be added to v4.2 databases, but they can be queried (e.g. dotplot).
The significant change
is with the file structure for the data, where the old file structure will not
work with this new release as shown further under <a href="#ds">Directory structure</a>.


<ul>
<li>General
	<ul>
	<li>Error messages (and stack traces) are written to error.log
	<li>There was one symap script with various flags, now there is four symap scripts
	with less flags, as follows:
	<table>
	<tr><td><i>Script</i><td><i>Build</i><td><i>View</i><td><i>Has 3D view</i>
	<tr><td><tt>./symap</tt> <td>Yes<td>Yes<td>No
	<tr><td><tt>./symap3D</tt> <td>Yes<td>Yes<td>Yes
	<tr><td><tt>./viewSymap</tt> <td>No<td>Yes<td>No
	<tr><td><tt>./viewSymap</tt> <td>No<td>Yes<td>Yes
	</table>
	<li>The symap -c command line parameter can be used to
	specify a configuration file other than 'symap.config'
	<li>The Summary list of all selected projects has been enhanced.
	</ul>

<li>Query database
	<ul>
	<li>The "Display for Selected Pair:" popup Summary is now available for self-synteny and FPC-Seq
	</ul>

<li>Build database
	<ul>
	<li>Log files: The log files are no longer in the results directories, instead they are in
	the top level <tt>/logs</tt> directory, as follows:
		<ol>
		<li>The output from the loads are concatenated to the file <tt>/logs/LOAD.log</tt>.
		<li>The output during alignment is in the <tt>/logs/&lt;proj1-to-proj2&gt;</tt> directory.
		<li>The output to the terminal from the alignment and synteny computation is
written in the <tt>/logs/&lt;proj1-to-proj2&gt;symap.log</tt> file.
		</ol>
	</ul>
<li>Load project
	<ul>
	<li>The input files can be gzipped (end with .gz)
	<li>The Parameters menu uses the previous location for the file/directory finder.
	<li>Add ConvertEnsembl.java to convert an Ensembl file to SyMAP annotation file.
	</ul>

<li>Alignment &amp; Synteny
<ul>
	<li>The trace output to terminal is much more informative.
	<li>The alignments for self-synteny have been optimized, which also
  changes the results (e.g. demo_seq now has some tiny blocks were before
  it had none).
	<li>The number of CPUs can be set in the interface, or either of the following:
		<ol>
		<li>Use the -p command line parameter, e.g. <tt>./symap -p 12</tt>.
		<li>Set the 'nCPUs' configuration parameter in the symap.config file, e.g. <tt>nCPUs=12</tt>.
		</ol>

	<li>Much of the code for building the syntenies has been re-written, but the results
	are basically the same with a few tiny differences.
</ul>

<a id="ds"></a>
<li>Directory structure: The following are the directory changes, where they
are under <tt>/data</tt>.
	<table>
	<tr><td><i>New</i><td><i>Old</i><td><i>Description</i>
	<tr><td>/seq
		<td>/pseudo
		<td>Directory for sequence projects
	<tr><td>/seq/&lt;project&gt;/sequence
		<td>/pseudo/&lt;project&gt;/sequence/pseudo
		<td>Default directory for sequences
	<tr><td>/fpc_results
		<td>/fpc_pseudo
		<td>Directory of FPC to sequence project results
	<tr><td>/seq_results
		<td>/seq_pseudo
		<td>Directory of two sequence project results
	<tr><td>/seq_results/&lt;proj1-to-proj2&gt;/align
		<td>/pseudo_pseudo/&lt;proj1-to-proj2&gt;/anchor
		<td>Directory of alignments
	</table>
</ul>


<p>

<p><a href="#top">Go to top</a>

<p>
<hr>
<p>
<h3>Release: Jan2018 (build 120)</h3>
<p>Release date: 1/16/18 - updated ConvertNCBI.java script.

<p>Release date: 1/10/18
<ul>
<li>A few more improvements on the applet code.
<li>Renamed some functions to make them clearer and updated the help.
<li>Improved the Help on the parameters page.
<li>Fixed a rare memory bug (occured on a self-synteny; requires the database to be re-created).
<li>The Image button does not work for Java v9, so a popup tells the user.
<li>Added a Java script called ConvertNCBI.java, as documented in <a href=ncbi target="_blank">NCBI to SyMAP</a>.
</ul>
<b>Release date: 12/28/17 (build 119)</b>
<ul>
<li>The applet code has been fixed to work better.
<li>The FPC to sequence block view had a bug that was fixed (though 'reverse' does not work).
</ul>
<b>Release date: 12/24/17 (build 118)</b>
<ul>
<li>There was a limit of 75 sequenced regions per genome, which the block view did not work after that.
<br>The limit has been changed to 150 and a popup message occurs if that limit is passed.
<br>A script (scripts/lenFasta.pl) is supplied to provided the sorted lengths, where you should set the
min_size parameter to the 150th length.
</ul>
<b>Release date: 11/7/17 (build 117)</b>
<ul>
<li>Fixed problem that prevented symap from running with new JDK v9.
</ul>
<b>Release date:  1/7/16 (build 116)</b>
<ul>
<li> Fix display sizing errors in circle view display
<li>Fix database error for long project names (max project name remains 40)
</ul>
<p><a href="#top">Go to top</a>

<h3>Version 4.2</h3>
Release date:  4/23/2014
<p><b>New Features:</b>
<ul>
<li> New, unified web access applet interface. Now you can present a unified SyMAP web display of some or all
of the projects in your database using just one small HTML page (no CGI). Applet interfaces are also
available for each of the sub-functions (dotplot, circle view, etc.) enabling customized web displays.
</ul>
<p>
<b>Modified Features:</b>
<ul>
<li> The Explorer has been updated to be more usable for draft projects, i.e. when there are many
sequences that have not been ordered and anchored to pseudomolecules. Specifically, it will now show
up to 450 sequences, and to decrease clutter it shows only sequences that actually have a synteny block.
<li> The prior web system files have been removed, including the cgi and supporting perl modules.
<li> The circle view now draws all text labels horizontally, due to rendering problems
with rotated text on several platforms (most severely on Mac). The rotated text is still available through a
checkbox option.
It is no longer necessary to run the perl install script to install web functionality.
</ul>
<h3>Version 4.1</h3>
Release date:  2/20/2014<br>
<p><b>New Features:</b>
<ul>
<li> Publication-quality image saving. A save button has been added to every graphical window allowing saving to
a wide variety of formats, including scalable vector formats (svg, eps, pdf) which may be resized clearly to any size. (Many thanks to the developers of the FreeHEP Vector Graphics library, and especially Mark Donszelmann, for making this possible!)
<li> Download of block co-ordinates. Choose the species of interest, open the Explorer, and the download button is at the lower left.
</ul>
<p>
<b>Modified Features:</b>
<p>
<ul>
<li> Applet is signed with a DigiCert certificate, removing the security blocks which have been increasing
in recent Java versions.
<li> MUMmer binaries for Mac OSX have been replaced by 64-bit versions. The 32-bit versions which had been
supplied could not access more than 2G RAM, preventing alignment of many chromosomes.

</ul>
<p>
<h3>Version 4.0</h3>
Release date:  7/15/2012<br>
<p><b>New Features:</b>
<ul>
<li>Project Manager Enhancements
<ul>
<li> Create new projects through the Manager
<li> No need to create directories or copy files
<li> Alignment &amp; Synteny parameters also available
</ul>
<li>Query Page Enhancements
<ul>
<li> Compute putative gene families across species
<li> Many additional filters
<li> Create and view Muscle multiple-alignment for selected results
</ul>
<li>Additional Features
<ul>
<li>Script 3track_figure.pl makes 3-genome figure for publication (see online
<a href="Tour.html#svg">Tour</a>;
script added to 4.0 package post-release)
</ul>
</ul>
<p>
<b>Modified Features:</b>
<ul>
<li> Write contig anchoring information to a file (<tt>order_against</tt> mode)
<li> No longer creates the "_ordered" project output from <tt>order_against</tt>
<li><tt>grp_prefix</tt> parameter now optional (but recommended). <tt>grp_sort</tt> eliminated.
<li> Self-alignments improved by using -maxmatch option of MUMmer for individual chromsome self-alignments
<li> Top-2 filtering modified to accept also hits within 80% of the 2nd best score, in order to
retain gene-family hits and allow identification of families through the query interface.
</ul>
<p>
<h3>Version 3.5</h3>
Release date:  1/24/12<br>
<p><b>New Features:</b>
<ul>
<li>Dynamic circular display
<ul>
<li> Circular-style view allowing addition/removal of species and chromosomes
<li> Replaces 3D view in the web applet (removing problems of MacOS support for 3D)
<li> 3D still available in standalone mode
</ul>
<li> Annotation query interface (Java-based, upgrades the web CGI version)
<li> Project parameters editing window
<li> Unsigned web applet (no approval popups, if database and webserver have same host)
<li> For draft sequence using <tt>order_against</tt>, create ordered and anchored versions of the project
</ul>
<p><b>Modified Features:</b>
<ul>
<li> Added contig-flipping to draft sequence <tt>order_against</tt> function
<li> Added draft sequence ordering demo
<li> Improved command-line launch script
</ul>
<p>
<h3>Version 3.4</h3>
Release date:  12/14/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Query files catenated before running MUMmer, for substantial increase in speed.
<li>Batch alignment buttons added to Project Manager.
<li>Improvements to handling of unordered scaffolds, including:
<ul>
<li>Scaffolds can all be in one fasta file.
<li>Added min_size parameter, to specify minimum size scaffold to load.
<li>Added order_against parameter, to specify reference sequence for ordering scaffolds
	(the order only affects the dotplot display)
</ul>
<li>Minor changes to CGI page displays to better accomodate unordered.
<li>Fix compilation error in 64-bit MUMmer binary, which prevented use of very long pseudomolecules.
</ul>


<h3>Version 3.3</h3>

Release date:  10/4/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Added "no_overlapping_blocks" parameter
<li>Added printout of block and synteny anchor tables to "/results" directory
<li>Remove "Contig0" (singleton clones) from FPC displays
<li>Adjustments to unannotated sequence clustering to obtain larger clusters,
    not dependent on the order of scanning of MUMmer output files.
<li>When clicking a block in the dot plot, 2D view will open with hit filter setting "show only synteny hits"
<li>Some bug-fixes to search.cgi
</ul>

<p>

<h3>Version 3.2</h3>
Release date:  6/9/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Inversions are shown on the 3D view in green, where un-inverted are shown in red.
<li>Anchor chains are not merged.
<li>The dot plot background is white and the synteny blocks blue.
<li>SyMAP automatically only uses one processor unless specified otherwise with the -p option.
<li>The whole genome dot plot is available from the project manager.
<li>When running an alignment, a summary is listed to the terminal and to a log file.
<li>A lot of error checking has been added. However, every platform can have variations, which
we may not have accounted for. <font color=green>So PLEASE, let us know if you have any problems and we will
work with you to it.</font>
</ul>

<h3>Version 3.1</h3>
Release date:  2/12/10<br>
This release includes major enhancements to the Dot Plot view.<br>
<br>
<b>New Features:</b>
<ul>
<li>Multi-genome Dot Plot: multiple species can be displayed on the y-axis against a common reference species.</li>
<li>Dot Plot view integrated into 3D viewer, now the species/chromosomes displayed in the Dot Plot can be changed on-the-fly.</li>
<li>Dot Plot reference species on the x-axis can be changed dynamically via a new drop-down menu.</li>
<li>Main web page redesigned.</li>
</ul>

<p><br>


<h3>Version 3.0</h3>
Release date:  1/7/10<br>
This major release includes many new features, in particular the 3D viewer and Project Manager.<br>
<br>
<b>New Features in User Interface:</b>
<ul>
<li>Multi-genome 3D viewer.</li>
<li>Circular CGI view.</li>
<li>Gene search CGI.</li>
<li>Support for sequence-to-sequence alignment.<li>
<li>Option to flip sequence.</li>
<li>Zoom into sequence or hit regions using the mouse scroll wheel.</li>
<li>Three-track CGI blocks view.</li>
<li>Improved annotation display and addition of URL-embedding capability.</li>
</ul>
<b>New Features in Back-End Processing:</b>
<ul>
<li>Completely rewritten in Java with major performance improvement.</li>
<li>Addition of Project Manager GUI to simplify and automate back-end processing, sequence alignment, and synteny analysis.</li>
<li>Support for sequence-to-sequence alignment.<li>
<li>Hits are now clustered based on gene annotation and drawn as "ribbons" to indicate locations of sub-blocks.</li>
</ul>


<h3>Version 2.0</h3>
Release date:  5/18/07<br>
This release includes many new features for the alignment view.  The dot plot view was
not changed.<br>

<br>
<b>New Features Block view (CGI):</b><br>
<ul>
<li> <i>View reverse</i> option: reverse which of the two genomes is the reference genome.
<li> Chromosome view: the blocks are color coded to indicate which chromosome they are from. Also, the number of anchors is displayed over each block.
</ul>

<b>New Features on Java interface:</b><br>
<ul>
<li> Filters:
<ul>
<li> For a given track (Block, Hit, Sequence) hover the mouse over the track and
right click. A menu of the most common filters is shown;
this is much easier than opening up the Filter window.
<li> From this window, you can go directly to the help for the given track.
There is also a 'Navigation Help', as there are many ways to resize, scroll, etc.
</ul>
<li> Sequence track:
<ul>
<li> <i>Show hit score bar</i> is a new filter that
shows the hits down the length of the sequence as a histogram showing how strong each hit is.

<li> <i>Show Hit Score Value</i> is a new filter that shows the %identity.
Note that the sequence graphic is twice as wide to show the score bar and score value.
<li> The sequence track can be scrolled by moving the mouse wheel when the
mouse is position over the sequence track.
This retains the same zoom and filters, but moves up or down the lenght of the sequence (i.e. pseudomolecule).
<li> The gene annotation now shows the intron/exon structure.
</ul>
<li> Hit track:
<ul>
<li> <i>Show Only Gene Hits (contained)</i> is a new filter that turns off the
display of all hits unless they are contained in an annotated gene.
<li> <i>Show Only Gene Hits (overlap)</i> is a new filter that turns off the
display of all hits unless they overlap an annotated gene (but is not contained in a gene).

<li> <i>Show Only Non-Gene</i> is a new filter that turns off the
display of all hits unless they are NOT overlapping or contained in a gene.
<li> Hit lines are highlighted in red when mouse positioned over a hit.
</ul>
<li> Block track:
<ul>
<li> When hovering over a contig, the contig information at the bottom of the
display includes all the chromosomes that the contig hits.
</ul>
<li>Bug Fixes:
<ul>
<li> Genes drawn incorrectly when sequence range within gene boundary.<br>
<li> Fixed bug with pseudomolecule filter.<br>
<li> Certain queries were failing when used with MySQL 5.0.26 due to comma precedence change.<br>
</ul>
</ul>
<b>New Features in Back-End Processing:</b><br>
<ul>
<li> Support for mixed letter/number chromosome names and arbitrary ordering (see example in demo-fpc params file)
<li> Pseudomolecule sequences loaded in 1Mb chunks, removing database problems with large text fields
<li> Filtering optimized in anchors.pl to reduce memory use in loading large blat outputs
</ul>


<h3>Version 1.0</h3>
Release date:  9/1/06<br>
First full public release.

<p><a href="#top">Go to top</a>

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
