<!DOCTYPE html>
<html>
<head>
<title>SyMAP Release Notes</title>
<link rel="stylesheet" href="style.css">
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
	ttx  {font-family: 'Helvetica', monospace; font-size: 14px; font-weight: 500; color: DarkSlateGrey;}
    ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
    ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
    ttp  {font-family: Arial, monospace; font-size: 15px; font-weight: 500; color: DarkSlateGrey; }
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr><td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP Release Notes
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br><a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr><td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./input/index.html" class="white">Input</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
				</table>
		</table>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->
For existing databases, there are no database or build changes since <a href="#v560">v5.6.0</a>; see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>.
<br>This will no longer run on 32-bit machines, starting with v5.5.1.
<br>You must have Java v17.0.11 or later, starting with SyMAP v5.5.0.
<h3>Summary of recent major releases</h3>
<table class="tx">
<tr><td><a href="#v560">v5.6.0</a><td>(23-Jan-25)<td>Improve <ttx>Cluster Hit</ttx> algorithm <ttx>Algo2</ttx>,
which is aware of the exon-intron structure.
<tr><td><a href="#v557">v5.5.7</a><td>(8-Aug-24)<td>Add <ttl>xToSymap</ttl> and improve ease of input.
<tr><td><a href="#v556">v5.5.6</a><td>(14-July-24)<td>Improved the collinear set algorithm.
<tr><td><a href="#v540">v5.4.0</a><td>(6-Apr-23)<td>Improved <ttx>Cluster Hit</ttx> algorithm <ttx>Algo1</ttx> for the detection of gene homology.
</table>

<a id="v565"></a>
<h3>Release: v5.6.5 (24-Apr-25)</h3>
Updates for Queries: Support for non-annotated hit ends (pseudo genes) bug fixes
for display of minor genes and a few other little things.

<p>For existing databases: (1) Run <tt>./symap -pseudo</tt>.
(2) Select the project pair and run A&S.
<p>Updates:
<ul>
<li><ttl>A&S</ttl>
<ul>
	<li><ttl>Number Pseudo</ttl>: A new option to number the non-annotated hit ends.
	This will created 'pseudo genes' that will be numbered sequentially and end with a "~".
	This is only useful in <ttl>Queries</ttl>.
	See <a href="SystemHelp.html#pseudo" class="ext" target="_blank">System Help</a>
	and <a href="Query.html#pseudo" class="ext" target="_blank">Queries</a>.
	<li>Small change: If '<tt>Remove alignments from disk</tt>' is to be performed, it prompts the user
	a second time to be sure.
	<li>The <ttl>Algo2</ttl> -wsp option to print the wrong strand hits had quit working.
</ul>
<li><ttl>Queries</ttl>
<ul>
	<li><ttl>Cluster Genes:</ttl>
	If <ttx>Number Pseudo</ttx> was used, un-annotated will be included in the clusters.
	<br>Updated to make the resulting cluster have at least one gene from each
	included species.
	<li><ttl>Gene Report</ttl>: If <ttx>Number Pseudo</ttx> was used, un-annotated  will be included
	for the non-reference genes.
	<li>Shorten query summary: Use species abbreviations instead of full display name.
	The Cluster will use 'Inc(All)' instead of listing all species.
	<li>Bug fixes for displaying <i>minor</i> genes: Two minor genes could be incorrectly paired in the table.
	<br>From v5.6.4: (1) A gene could be paired with what appeared
	to be an unannotated end, i.e. there would be no gene information, but the minor hit information
	was correct. (2) The <ttx>Every+</ttx> would not work with the <ttx>Multi-hit</ttx> option.
	<li>Bug fix from v5.6.4: If the user input was incorrect, it was stuck in "Running" mode.
	The <ttx>Multi-gene</ttx> report would show rows that were not reference if the report was run,
	another query was done and the user came back to this one.
</ul>
</ul>

<a id="v564"></a>
<h3>Release: v5.6.4 (10-Apr-25)</h3>
Minor release for Queries: two new small features, one bug fix (from v563), and few other small things.
<ul>
<li><ttl>Queries</ttl>
	<ul>
	<li><ttl>Export</ttl>: Bug fix from v5.6.3 - the <ttx>Export</ttx> button on
	the popup was not visible.
	<li>When the columns are changed by selection, the table stayed scrolled to the same row.
	<li><a href="Query.html#search" class="ext" target="_blank">Search</a>.
	New feature to search on string for a selected column.
	<li>Table Panel:
	<ul>
		<li>The <ttl>Stop</ttl> did nothing; it works now.
		<li>If multiple queries are running at once, they will fail;
		see <a href="Query.html#stop" class="ext" target="_blank">Wait/Stop Query</a>.
		<li>If a different table tab was clicked while one was building, it produced an error message though
		there was no problem.
	</ul>
	<li>MSA panel:
	<ul>
		<li>The <ttl>Cancel Alignment</ttl> did nothing; it works now and is called <ttl>Stop</ttl>.
		<li>If multiple MSA are running at once, they will fail;
		see <a href="Query.html#msaMultP" class="ext" target="_blank">MSA General Notes</a>.
		<li>It would appear in whatever panel was being shown; now that does not happen.
		<li>The MSA tab on the left is now incomplete until it is MSA panel is visible.

	</ul>
	</ul>
<br>
<li><small>Developers:
Rename SyMAPQueryFrame -&gt; QueryFrame; New file UtilSearch. Put 2D methods in UtilSelect.
</small>
</ul>

<a id="v563"></a>
<h3>Release: v5.6.3 (5-Apr-25)</h3>
Improvements to the <ttb>Queries</ttb>: <ttl>MSA Align</ttl>,
<ttl>Gene Pair Report...</ttl> and replaced <ttl>PgeneF</ttl> with a <ttl>Cluster hits</ttl> algorithm.
Two bug fixes.
<p>If you are currently using SyMAP, you only need to download <ttx>symap.jar</ttx>.
However, if you want to run MAFFT, you need to  download <ttx>symap_5.tar.gz</ttx>, extract the
<ttx>ext/mafft</ttx> directory and move it to your <ttx>ext</ttx> directory.

<ul>

<li><ttl>Queries</ttl>
<ul>
<li>Bug fix on Linux: crashed on searching for a given collinear set.
<li><ttl>MSA</ttl>: See <a href="Query.html#msaAlign" ext="class" target="_blank">MSA</a>.
	<ul>
	<li>MUSCLE is very slow, so MAFFT has been added which is many times faster.
	<li>Bug fix: negative strand hits were not reversed complemented.
	<li>There is a popup menu with the options:
		(1) MUSCLE or MAFFT; for MAFFT, use Auto and enter #CPU.
		(2) Options for merging overlapping hits, or removing gaps between subhits.
		(3) Option to trim aligned sequence.
	<li>The consensus sequences now uses a 'n' if there are multiple different bases in the
	column and no majority. The consensus is lowercase if there is any mismatch in the column.
	<li>The source files and resulting alignment are written in sub-directory <tt>queryAlign</tt>.
	The results are sorted according to the lower table of information. The results are
	in a split pane.
	</ul>
<li><ttl>Cluster genes</ttl>: See <a href="Query.html#clhit" ext="class" target="_blank">Cluster genes</a>.
	<ul>
	<li>A new filter called <ttl>Cluster hits</ttl> has replaced <ttl>PgeneF</ttl> (which can still
	be run using <tt>viewSymap -g</tt>).
	</ul>
<li><ttl>Gene Pair Report...</ttl> See <a href="Query.html#report" ext="class" target="_blank">Report</a>.
<ul>
	<li><ttl>Multi-Hit Gene Report</ttl>:
	(1) This was confusing because groups can be created from non-gene hits,
	but the report only works with gene hits. This has been made more obvious.
	(2) The report is now sorted by <ttx>Gene#</ttx> instead of <ttx>Group</ttx>.
	<li><ttl>Cluster Hit Report</ttl>: This works for the <ttl>Cluster Hits</ttl> (formally <ttx>PgeneF</ttx>) query.
	<li>Values are saved for subsequent uses of a <ttl>Gene Report</ttl> for its respective <ttx>Query Results table</ttx>.
	This fixed an issue where the last query settings would be used for any report.
	<li>The length of the <ttx>Gene Annotation Columns</ttx> wrap-around length can be set,
	or truncated (relevant to <tt>Desc, Product</tt>). There is an option <ttl>Gene#</ttl> to
	prefix annotation with the Gene#. There is an option to not put borders around each cell.
	<li>The reference <tt>Gene#</tt> column is in italics for the HTML report.
	If there are duplicate <tt>Gene#</tt>, the 2nd has is followed by a "+".
</ul>
<li>The <ttl>Grp#</ttl> and <ttl>Grp Size</ttl> have been merged into one column:
<tt>Size-Grp#</tt>.
<li>Other:<small>
<ul>
<li>Rewrote the Instruction page since its HTML was obsolete in Java20.
Formatted the table Statistics better. If a chromosome was selected, the statistics did not show the Groups counts; it does now.
<li>The Hit Strand was changed from "=" and "!=" to the actual strands separated by "/".
The empty row/column value was changed from "-" to "." so as not to be confused with negative strand "-".
<li>The column <ttx>Defaults</ttx> have changed. <tt>Block</tt> and <tt>Collinear</tt> columns sort in the reverse direction.
<li>The <ttx>2D Group</ttx> only works for hits on the same two chromosomes as the selected hit; hence, it is renamed <tt>Group(chr)</tt>.
<ttx>2D 3-chromosomes</ttx>: There was an erroneous popup message if the 2nd of 3 chromosomes has start coordinate close to zero.
<li>Disabled Collinear when either Multi or Clust is selected.
</ul>
</small>

</ul>
<br>
<li><small><ttl>Developers:</ttl> Changes to <ttl>Queries</ttl>.
<ul>
<li> Removed most hard-coded symbols.
<li>Gene Pair Report: Allowed various length project names in the Report menu.
Major changes to UtilReport.java.
<li>Moved some of the MSA functions to a new file UtilSelect.java and renamed the
button to MSA. Change the MSA Export to use the file chooser and make directory Globals.
Add the number of bases aligned to the status panel.
<li>File name changes:
<br>&rarr;symapMultiAlign package:
Delete MultiLineTextPanel and AlignmentData;
Delete MenuMapper and LegendPanel (moved to MsaMainPanel);
Delete AlignmentPanelBase (MsaPanel was extend, now merged);
<br>Rename AlignmentViewPanel-&gt;MsaMainPanel; MultiAlignmentData-&gt;MsaRun; MultipleAlignmentPanel-&gt;MsaPanel.
symapMultiAlign package deleted and files moved to symap.closeup.
<br>&rarr;symapQuery package: FieldData-&gt;TableColumns; TableDataPanel-&gt;TableMainPanel; SortTable-&gt;TableSort.
TableReport-&gt;UtilReport; TableExport-&gt;UtilExport; SpeciesSelectPanel-&gt;SpeciesPanel;
ResultSummaryPanel->ResultsPanel.
New UtilSelect, UtilReportNR, ComputeClust.

<li>Non-queries changes: Manager: add extra info if the Display Name has been used.

</ul>
</ul>
<a id="v562"></a>
<h3>Release: v5.6.2 (24-Feb-25)</h3>
New <ttl>Queries</ttl> features for 3 species.

<ul>
<li><ttl>Queries</ttl>
	<ul>
	<li><ttl>View 2D</ttl>:
		<ul>
		<li>It will display <ttl>View 2D</ttl> with 3-tracks (chromosomes) if two rows are selected which have a common reference gene,
		and has 'both' genes in both rows. This works with the <ttx>Region</ttx> or <ttx>Collinear Set</ttx> options.
		See <a href="Query.html#msp" class="ext" target="_blank">3-chromosome 2D</a>.
		<li>A new <ttx>Gene#</ttx> checkbox for <ttl>View 2D</ttl> which shows the <ttx>Gene#</ttx> instead of the <ttx>Annotation</ttx>.
		</ul>
	<li><ttl>Report...</ttl>:
		<ul>
		<li>Add export to TSV.
		<li>Add <ttl>Row all</ttl> and <ttl>Collinear all</ttl> for when
		there is &gt;2 species, it requires all species to be present
		(see <a href="Query.html#report" class="ext" target="_blank">Report...</a>).
		<li>Many small improvements.
		</ul>
	</ul>
<br>
<li><small><ttl>Developers:</ttl> (1) Rewrote part of Query.tableReport.java.
(2) Small changes to the Query Result tab.</small>
</ul>
<a id="v561"></a>
<h3>Release: v5.6.1 (12-Feb-25)</h3>
Bug fixes, minor updates to A&S output, and improved <ttl>Queries</ttl>.
<ul>
<li><ttp>xToSymap</ttp>:
<ul><li>Bug fix: <ttl>NCBI convert</ttl>: if the gene did not have a description, it was suppose to use the first mRNA 'product', which
it was not doing.
</ul>
<br>
<li><ttl>A&S (Alignment&Synteny)</ttl>:
<ul>
<li>The default is to surpress output to the terminal and <tt>logs/{proj-to-proj}/symap.log</tt> except final summaries.
To show the verbose output, use <ttx>./symap -v</ttx> or use the parameters
<a href="SystemHelp#over2" class="ext" target="_blank">Verbose</a> checkbox.
See <a href="Demo.html#slog" class="ext" target="_blank">Demo examples</a>.
<li>Status lines, which repeatedly use the same line, are written to stderr instead of stdout.
<li>The <ttx>symap</ttx> or <ttx>viewSymap</ttx> flag to show settings for the important MySQL variables is now "-sql".
<li><ttl>Algo2</ttl> no longer quits if it encounters a sequence name not loaded when reading MUMmer.
</ul>
<br>

<li><ttl>Query</ttl>:
<ul>
<li>New <ttl>Arrange</ttl> button: When the column panel is being displayed, the new <ttl>Arrange</ttl> button groups similar
columns, putting the gene columns first.  See <a href="Query.html#col" class="ext" target="_blank">Arrange</a> option.
<li>The <ttx>Singles</ttx> query allowed a species to be selected; this has been replaced with a species-specific check mark
beside each species name. This query no longer allows <ttl>From</ttl> and <ttl>To</ttl> to be entered.
See <a href="Query.html#query" class="ext" target="_blank">Query</a>.
<li><ttx>Gene#</ttx> search:
<ul>
<li>The new species-specific check can be used for this search also.
<li>Search speedup.
<li>The search will now match genes with minor hits, e.g. 5.b.
<li>The <ttx>Grp#</ttx> column computation has been changed as follows: for a given gene that
hits multiple places on the same opposite chromosome, they are put in a group so that they can all be
viewed at once with the <a href="Query.html#view2D" class="ext" target="_blank">2D Group</a> option.
</ul>
<li><ttx>Multi</ttx>: removed the <ttx>Exon</ttx> option and make the <ttx>Opposite</ttx> options radio buttons.
<li>Better error checking.
</ul>
<li><ttl>2D</ttl>: Better positioning of <ttx>Gene#</ttx> when the <ttl>Gene#</ttl> sequence option is selected.

<br>&nbsp;
<small>
<li>Little changes: (1) Query title does not use 'Chr' if number has a prefix. (2) Query Summary replaced Category with #Genes.
(3) In 2D, 'Set' changed to 'CoSet'. (4) For A&S, rename 'Aligning Projects' to 'Synteny'.

<li>Minor bug fixes:
<ol>
<li>If "NUCmer only" had been set, it was not showing in <ttx>Pair Parameters</ttx> on restart.
<li>For the <ttx>Singles</ttx> <tt>numHits</tt> column, it was not being updated
if a project was removed from the database.
<li> The <ttl>Queries</ttl> <ttx>Summary</ttx> had the incorrect <ttx>Genes</ttx> count (error introduced v560)
</ol>
<br>
<li>Developers: Changed TableDataPanel, TableExport, QueryPanel, SpeciesSelectPanel, PairParams to use Jcomp and have button tips.
</small>
</ul>
<!---------------------------------------------------->
<a id="v560"></a>
<h3>Release: v5.6.0 (23-Jan-25)</h3>
There were some major changes made to <ttl>Algo2</ttl> (first released <a href="#v546">v5.4.6</a>) and other minor changes.
<br>For existing databases, see <i>Update</i> in the items below.
<ul>
<li><ttb>A&S</ttb> <ttl>Algo2</ttl>:
<ul>
<li>The improved algorithm reduces:
<ol>
<li>the number of long clustered hits with huge gaps (happens with mammalian genomes),
<li>short hits,
<li>hits with little overlap to connected genes (this last change causes much smaller
<ttl>Summary</ttl> <ttx>GNwCH</ttx> and <ttx>GNwCHB</ttx> numbers).
</ol>
<li>G2 assignments do much better at assigning the hit's genes.
<br>G2 no longer allows wrong strand hits,
but the "-wsp" remains to print the wrong strand hits.
<br>G1 and G0 assignments now require less gaps between hits and a 'Hit len' ratio of &gt; 0.5.
<br>G1 no longer has minor assignments.
<li>Internal parameters are mostly set by automatic analysis of the gene models.
The user parameters have been changed from <ttx>Minimal matched bases</ttx> to <ttx>Scale</ttx>,
which is more intuitive to change, i.e. increase filters more, decrease filters less.
See <a href="SystemHelp.html#hint" class="ext" target="_blank">Algo2 Hints</a> for tested datasets and parameter changes.
<li>The gene overlap score (shown on the Gene popup) has been changed to be the
overlap between the hit and the gene without subtracting gaps (same as <ttl>Algo1</ttl>).
<li>The changes do not change the synteny blocks very much, however,
the reduction in short F+ hits can result in slightly better synteny blocks and less tiny blocks.
</ul>
<i>Update</i>: The <ttb>A&S</ttb> needs to be re-run to get this update.

<br>&nbsp;
<li><ttb>Queries</ttb>:
<ul>
<li>If the columns have been re-ordered in the last table that was created or had a column added/removed, the re-ordered
columns will be applied to the next created table.
<li>When sorting on <ttl>Gene#</ttl>, hits without a gene annotation will sort to the bottom.
<li>For a singles table: (1) Bug fix: the gene length (<ttx>glen</ttx>) column would not sort.
(2) The <ttx>NumHits</ttx> column has been moved to the 2nd column slot.
<li><ttx>Export</ttx>: Writes the SyMAP version and database name as a comment in the first line of the CSV and FASTA file,
and the title for the HTML file. The default for CSV and HTML is no row#.
<li>Start <ttx>viewSymap -q</ttx> to see the gene overlap scores instead of the exon scores for <ttx>Algo2</ttx>.
</ul>

<br>
<li><ttb>2D</ttb>
<ul>
<li>Scroll wheel: The scroll wheel had to be scrolled in the opposite direction for a flipped sequence.
Now it scrolls consistently regardless of flipped state.
<li>The algorithm for placing overlapping genes is slightly improved.
<li><ttl>Hit Filter</ttl>: An <ttx>And</ttx> and <ttx>Or</ttx> radio button has been added
next to the <ttx>Synteny Blocks</ttx> (see <a href="UserGuide.html#hitfilter" class="ext" target="_blank">Hit Filter</a>).
<li><ttl>Download Blocks</ttl>: See <a href="UserGuide.html#data_download" target="_blank" class="ext">Download</a>.
<ul>
<li>Has three options: the original <ttl>Blocks with gene info</ttl>, a new <ttl>Blocks only</ttl> and a new <ttl>Blocks summary</ttl>.
<li>The output is now sorted by the two chromosome numbers followed by the block number.

</ul>
<li>Hit popup: Two new conditional options have been added. (See <a href="Details.html#hit" class="ext" target="_blank">Hit popup</a>.)
<ol>
<li><ttl>Merge</ttl>: If there are negative overlaps, it will merge the hits.
<li><ttl>Order</ttl>: The target hits are listed by the query sorted order;
if the target hits are not strictly in sequential order, they will be sorted.
</ol>
The gap column was one off, which has been fixed.
</ul>
<p>
<li><ttb>Summary</ttb>: Added <ttx>#Genes</ttx> to the <ttx>Collinear set</ttx> row.
<br>&nbsp;&nbsp;&nbsp;Bug fix: "GNwCHB" was slightly off when a project is in multiple synteny pairs.
<br><i>Update</i>: Start up "<tt>symap -s</tt>"", then select <ttl>Summary</ttl>; the -s will cause it to regenerate the summary.

<p>
<li><ttb>Dotplot</ttb> and its <ttx>Summary</ttx>:
The stats showed 0% if &lt; 1%, now shows "&lt;1%"; and a few other little adjustments.

<p>
<li><ttx>Demo_seq</ttx>: 4 duplicates were removed from the <tt>data/seq/demo_seq/annotation/anno.gff</tt> file. SyMAP
works fine with duplicates, but they can result in non-deterministic results (i.e. sometimes the .a is selected
as the major gene and other times the .b is selected).

<br>&nbsp;
<li><ttx>JDBC</ttx>: The java-mySQL interface was updated to the latest v9.1.0. This fixed a bug on
Ubuntu that only allowed 'root' to be used as the username (db_adminuser).

<br>&nbsp;
<li>Developer:

<ul>
<li><small>Algo2 details</small>
<ul>
<li><small>It was the case that a hit could only be assigned to one cluster hit, but now it is allowed to
be assigned to multiple cluster hits when there are overlapping or close genes.
It is required that Clustered hits connecting two genes have the same orientation as the genes. </small>
<li><small>Many of the cutoffs are now approximated from the gene models.
The allowed size intron of a non-gene hit was way too big; now it
uses an approximation from the input gene introns.</small>
<li><small>The algorithm code has been largely rewritten and is much easier to follow. All constants are stored in Arg.java.</small>
<small>The average, max, min gene sizes are printed when genes are loaded,</small>.
</ul>

<li><small>In backend: (non-functional changes) </small>
<ul>
<li><small>Rename ProgSpec to be AlignRun; AnchorsPost to AnchorPost; AlgSynMain to DoAlignSynPair; LoadProj to DoLoadProj.
All Synteny methods go in the new package backend.synteny; rename Block to SyBlock; create new RWdb.
Moved collinear calculation from SyntenyMain to DoAlignSynPair. Move RW methods to RWdb.
Add a check for zero hits in chromosome pair when creating hit# in AnchorMain.</small>
</ul>

<li><small>Other</small>
<ul>
<li><small>Part of file closeup.SeqData was moved to a new file closeup.SeqDataInfo for the 2D popups.</small>
<li><small>Changed 'set GLOBAL' to 'set global' for database suggestions.</small>
<li><small>The ".first()" database command did not work with the new JDBC,
so was replaced with ".next()".</small>
<li><small>Cleaned up all command line options (created many changed files).
Remove the -wse to exclude wrong strand hits, as now Algo2 no longer uses them.
Remove the -y hit count only option, and the -s print stats. The -s now only regenerates the summary.
Add -ii for Query Overlap genes, 2D popup indices, and hit popup Extend column.</small>
<li><small>Query: Export: Prompt user if rows have been selected.
Query Table: cleaned up old code and some rearranging.</small>
</ul>
</ul>
</ul>

<a id="v559"></a>
<h3>Release: v5.5.9 (24-Sept-24)</h3>
This is mostly a documentation update, with a few tiny code updates.
<ul>
<li>SyMAP can be run with a "-mum" flag, which prevents any MUMmer files from being removed by SyMAP after alignment.
<li>On creating a database, symap checks the MySQL settings and external programs. It checks java minor version numbers.
<li><small>Developer: remove some obsolete code for orderAgainst, which removed an 'unchecked' call. Write Java version into symap.log.</small>
</ul>

<a id="v558"></a>
<h3>Release: v5.5.8 (1-Sept-24)</h3>
Small changes to <ttp>xToSymap</ttp> and some small <ttp>symap</ttp> interfaces changes.
<ol>
<li><ttp>xToSymap</ttp>:
<ol type='a'>
<li><ttl>Summarize</ttl>: Added output:
<ul>
<li>Output all attributes for gene. Compress the summary lengths counts.
If not <ttl>Verbose</ttl>, remark if the sequence appears to be hard masked.
<li>If <ttl>Verbose</ttl>, output base count; print header lines of &lt;=10 unique prefixes; print lengths of first 5 input sequences.
</ul>
<li><ttl>Split</ttl>: A new feature that will split the converted output into chromosome files.
<li><ttl>Summary</ttl>: Improved determining ">ID" prefixes.
<li><ttl>ConvertNCBI</ttl>: The mRNA <ttx>products</ttx> are no longer merged;
either the gene <ttx>description</ttx> is used or the 1st mRNA  <ttx>product</ttx>.
The 1st mRNA was a gene attributed with keyword "ID=", now it is a separate keyword "rnaID=".
The text "gene-" and "rna-" are removed from their respective IDs (to make name shorter).
The option for assigning a list of protein IDs corresponding to all mRNAs for the gene has been removed;
now only the 1st protein-id is listed.
See <a href="input/ncbi.html">NCBI</a> convert.
<li><ttl>ConvertEnsembl</ttl>: The gene has a new attribute of "rnaID=" with the 1st mRNA ID which
corresponds to the input exons. It now shares the option of having a gene attribute called "proteinID"
which is equal to the "protein=id" of the 1st CDS for the 1st mRNA.
See <a href="input/ensembl.html">Ensembl</a> convert.
Fixed a rare bug: This script removes the "[Source:..." text, but it would incorrectly remove any brackets.

</ol>
<br>
<li><ttp>symap</ttp>: Small interface changes:
<ul>
<li><ttl>Project Parameters</ttl>: Allow comma's in minimal length cutoff.
<li>The display name is checked for duplicates.
<li>The abbreviation is shown on the <ttl>Selected</ttl> summary and "Project" renamed to "Directory".
<li>Fixed a rare sort bug in <ttx>Algo2</ttx>.
</ul>
<br>
<li><ttp>viewSymap</ttp>:
<ul>
<li><ttl>Query</ttl>: Allow comma's in <ttx>Chr:</ttx> start or end.
The <ttx>Show</ttx> popup of a row's content has comma's in large numbers,
and breaks up long <ttx>All-Anno</ttx> lines.
</ul>
<br>
<li>Developers: Most of the <ttp>xToSymap</ttp> methods were simplified.
</ol>
<a id="v557"></a>
<h3>Release: v5.5.7 (8-Aug-24)</h3>
This release has added clarity and ease to loading the input files.
<ol>
<li>NCBI and Ensembl files
<ul>
<li><ttl>xToSymap</ttl>: A new <a href="./input/index.html" class="ext" target="_blank">interface</a> for the convert NCBI and Ensembl scripts,
and to replace the <tt>scripts/getLengths.pl</tt> script.
<li>The <ttl>Ensembl</ttl> conversion now recognizes a chromosome sequence if the word
'chromosome' is on the header line. It has different rules on what is output.
<li>The <ttl>NCBI</ttl> conversion now checks for "NT_", "NW_" and "NC_" prefixes (is
was just checking for word 'chromosome' or 'scaffold'). It has different rules on what is output.
</ul>
<li><ttl>Load Project</ttl>:
<ul>
<li>This will now load most NCBI and Ensembl files directly, though it is strongly recommended to convert them with <ttl>xToSymap</ttl>.
<li>The fasta file(s) are checked for a suffix of .fsa, .fa, .fna, .fasta, or .seq.
The gff file(s) must have a suffix of .gff or .gff3.
<li>In the GFF file, there must be a mRNA line for the following exons. The mRNA parent must match the gene ID, and
the exon parent must match the mRNA ID.
</ul>
<li><ttl>A&S</ttl>: <ttx>Algo 2</ttx> would crash on a rare case when sorting exons, which has been fixed.
<li>Demo: was updated to have correct GFF3 files.
<li>Documentation: has been improved for the input and build process.
<li>Developers: a new package called <ttl>xToSymap</ttl> has been added.
</ol>

<a id="v556"></a>
<h3>Release: v5.5.6 (15-July-24)</h3>
For existing databases, re-run the collinear algorithm as follows: Start <ttx>./symap -acs</ttx>, then select the pair and run
<ttb>Selected Pair (Redo)</ttb>; a popup will have you confirm "Collinear only" indicating that only this algorithm will be run.
<ul>
<li><ttl>A&S Collinear algorithm</ttl>: V5.5.6 finds more collinear sets and longer ones.

Pre-v5.5.6 gave obviously wrong results if a gene from the hit numbered project
(the hits are numbered along the alphabetically lesser project name) was part of two different collinear sets.
In my test sets, this happened a few times between Arabidopsis, Brassica rapa, and cabbage, but none for my partial
mammalian test set of humans, chimps, mouse and rabbit.

<br>&nbsp;
<li><ttl>Query</ttl>
<ul>
<li>New: A new <ttb>Report</ttb> button is at the top of the results table.
See <a href="Query.html#report" class="ext" target="_blank">Query Report</a>.

<li>Bug fix for <ttx>Multi-hit gene</ttx>: it would crash if there was no hits for a pair of
species.
<li>The results table was sorted by <ttx>Hit#</ttx>; now it is sorted by <ttx>Block</ttx> or <ttx>Collinear</ttx>
or <ttx>Chr</ttx>, and by <ttx>Hit#</ttx> within each group.
</ul>
<br>
<li>Developers: <ttx>TableExport.java</ttx>, <ttx>TableReport.java</ttx>
and <ttx>TmpRowData.java</ttx> are new files. The output of Load Annotation is clearer if there is
no annotation files.
</ul>

<a id="v555"></a>
<h3>Release: v5.5.5 (19-June-24)</h3>
For existing databases, if you ran <ttx>Algo2</ttx> on mammalian genomes, you may want to re-execute
the <ttl>Synteny</ttl> step to get better clustering of hits for very long genes.
<p>This release is mainly with upgrades for <ttb>Query</ttb> with an option to view tandem genes.
<ul>
<li><ttl>Alignment & Synteny</ttl>:
<ul>
<li><ttx>Algo2</ttx> for gene-to-gene hits, it was keeping hits separate when there was a very long gene with tiny exons;
it will now cluster them into one. For gene-to-non-gene, it is clustering hits into bigger clusters.
<li><ttx>Algo2</ttx> can take as input NUCmer files, but still does not do self-alignments. I tried
humans and chimpansee with PROmer vs NUCmer, and PROmer identified more genes.
</ul>

<li><ttl>Query</ttl>
<ul>
<li>Columns:
<ul>
<li>If the hit is not to a gene, the chromosome# will be shown for the <ttx>Gene#</ttx> column instead
of '-' (e.g. '3.-' indicates the hit is on chromosome 3).
<li>Columns <ttx>PgeneF</ttx> and <ttx>PgFSize</ttx> have been renamed <ttx>Grp#</ttx> and <ttx>GrpSize</ttx>,
respectively. These are assigned values by the queries <ttl>Gene#</ttl>,  <ttl>Multi-hit Gene</ttl>
and <ttl>PgeneF</ttl>.
</ul>

<li>New:
<ul>
<li>The <ttl>Gene#</ttl> filter results have assigned values for the columns <ttx>Grp#</ttx> and <ttx>GrpSize</ttx>,
where a group is associated with the species+chromosome.
<li>For the <ttl>Multi-hits</ttl> query (see <a href="Query.html#multi">Mulit-hit</a>):
<ol><li>New options:
<br><ttl>Exon</ttl>: only show hits if they overlap an exon in the target gene.
<br><ttl>Same chr</ttl>: the n hits for the target gene must be to the same opposite chromosome.
<br><ttl>Tandem</ttl>: the n hits for the target gene must be to tandem genes on the opposite chromosome.
<br><li>The columns <ttx>Grp#</ttx> and <ttx>GrpSize</ttx> are assigned values.
<li>Small change: If a chromosome is selected for all species, the hits must be on both chromosome (it was an 'either', now
its a 'both).
</ol>
<li>On the results table, the <ttl>View 2D</ttl> feature has a new option of <ttx>Group</ttx>
(see <a href="Query.html#group">View 2D</a>).
<ul>
<li>When the 2D is shown, all rows with the selected row <ttx>Grp#</ttx> and <ttx>Chr</ttx> will be highlighted.
<li>The selected hit (and group hits) are highlighted in a new "Special" color; this can be turned off
by selecting <ttl>High</ttl> on the <ttb>Query</ttb> interface before display; this release added the ability to turn the highlighting off by
de-selecting the <ttl>Hit Filter</ttl> <ttx>High popup</ttx> option.
</ul>
</ul>

<li>Interface:
<ul>
<li>The <ttl>Multi-hits</ttl> option has been grouped with the <ttl>PgeneF</ttl> option, and
now has a check mark in order to run.
<li><ttl>Collinear size</ttl> now has an <tt>Ignore</tt> so it does not depend on having a number in the
associated text box set to blank in order to ignore.
</ul>
<li>Small change: <ttl>Annotation Description</ttl> no longer automatically includes minor genes since this
can be used in conjunction with <ttl>Every*</ttl> to get that result.
<li>Bug fixes:
<ul>
<li>If a hit was to two minor genes (<ttx>Every*</ttx> option), the MUSCLE command would not work.
<li>If one of the genomes was not annotated, the <ttl>View 2D</ttl> did not work.
</ul>
</ul>
<li><ttl>Chromosome Explorer</ttl>:
<ul><li>New option: The <ttl>2D</ttl> display has a <ttl>Sequence Filter</ttl> of <ttx>Conserved</ttx>
for the reference chromosome, which was renamed to <ttx>g2x2</ttx> and option <ttx>g2x1</ttx> added, where the first highlights
all hit-wires and corresponding genes that are conserved across all 3 chromosomes, and the second highlights
the hit-wires and corresponding genes that are only found on one pair.
<li>The <ttx>Gene popup</ttx> coloring of the hit-wire has changed from the "hover" color to the
new "special" color.
</ul>
<li>Developers: <ttx>ComputeMulti.java</ttx> is a new file.
</ul>

<a id="v554"></a>
<h3>Release: v5.5.4 (7-May-24)</h3>

<ul>
<li><ttb>2D</ttb>
<ul>
<li><ttl>Show Annotation</ttl>: (1) This used a constant that was appropriate for
plant genes, but not mammalian; the constant is now computed based on the average
length of genes for each chromosome. (2) The boxes now have background light-gray and
border in the exon +/- color. (3) If the last line of the annotation is long, it
is now truncated.
<li>A tiny problem has been identified and documented in the <a href="./UserGuide.html">User Guide</a>;
search on "KNOWN PROBLEM" (this will be fixed for v5.5.5).
</ul>
<li><ttb>Circle</ttb>
<ul>
<li>The <ttx>Two-color all blocks</ttx> option, which colors the inverted and non-inverted
blocks two different colors, can now have its colors changes via the color palette icon.
</ul>
<li><ttb>View</ttb>
<ul>
<li>Add <ttx>AvgGeneLen</ttx> and <ttx>AvgExonLen</ttx> for the average gene and exon
length of each chromosome.
</ul>
</ul>
<a id="v553"></a>
<h3>Release: v5.5.3 (28-Apr-24)</h3>
Interface changes for the <ttb>Circle</ttb> display.
<br>Previous bug fixes: v5.5.1 had bug in 2D when no annotation. v5.5.2 had bug when genome&gt;2000M.
<br>This release has been very well tested (using these <a href="./SystemGuide.html#machine">datasets and machines</a>)!
<p><ttb>Circle</ttb> display
<ul>
<li>New color options so that they can be changed; see <a href="./UserGuide.html#circColors">Colors</a>.
<li>The circle now remains the same size between different datasets and options.
<li>Minor: (1) The mouse turns to a clickable finger when held over the
project name (instead of beneath it). (2) The rotate uses smaller increments.
<li>Minor 2D: (1) As long as the reference chromosome stays the same, chromosomes can be added/removed,
or a different view can be shown, and the circle display will
be the same according to the first set of top buttons. (3) A long project name
may be truncated if shown on the left, but if shown on the right, the full name can be viewed by
sliding the scroll bar.
</ul>

<a id="v552"></a>
<h3>Release: v5.5.2c (22-Apr-24)</h3>
22-Apr-24 bug fix release: A bug from v5.5.1 where 32-bit machines were no longer
supported, but I failed to test this on large genomes (big thanks to Ricardo for this bug report!).
<p>
15-Apr-24 bug fix release: A bug that caused 2D to crash when displaying
a sequence track with no annotation. This release still has date 31-Mar-24.
<p>
The control bars for <ttl>Circle</ttl>, <ttl>DotPlot</ttl> and <ttl>2D</ttl> have been made to
look and act in similar fashion. There are other small interface changes for clarity and new function.
<ul>

<li><ttb>Circle</ttb>
<ul>
<li><ttl>Control bar</ttl>: new icon buttons for rotate counter-clock-wise, Home
and <ttx><big>(i)</big></ttx> quick help.
<li>The scroll bar now works. The initial frame is smaller.
</ul>

<li><ttb>2D-display</ttb>
<ul>
<li><ttl>Control bar</ttl>: <ttx>Zoom</ttx> and <ttx>Show</ttx> buttons with popup menus replaced the
one <ttx>Selected</ttx> drop-down. The <ttx>Scale</ttx> icon is now a toggle.
<li><ttl>Sequence Filter and Popup</ttl>
	<ul>
	<li>New <ttx>Full</ttx> popup option to show the full sequence.
	<li>New <ttx>Info (i)</ttx> and gene <ttx>Clear</ttx> filter buttons; and tooltip on mono-texted items.
	<li>Instead of any change creating a history event, only a change to the coordinates does.
	</ul>
<li><ttl>Hit Filter Popup</ttl>: New <ttx>All Hits</ttx> popup option to show all hits.
<li><ttl>Gene Popup</ttl>: Now lists minor gene hits along with the major; add the block number for the hits.
</ul>
<p>
<li>Small changes:
<small>
<ul>
<li>Put <tt>Circle</tt> first on Manager.
<li><tt>Circle Control bar</tt>:
The <tt>Self-synteny</tt>, <tt>Reverse</tt> and <tt>Rotate</tt> are now toggle buttons,
and <tt>Scale</tt> is a toggle icon.
The <tt>Self-synteny</tt> option is only displayed if self-synteny has been
computed for at least one project displayed. The <tt>Rotate</tt> text button is moved next to the
rotate circle buttons.
The view drop-down has been changed to a popup menu.
<li><tt>Dot plot Control bar</tt>: The Scale option is now a toggle icon.
The Home button now includes scale, and disables when it is home.
<li><tt>2D</tt>: Information shows number of histories for &gt; and &lt;. The toggle shows on/off status.
<br>If a <tt>Hit Text</tt> option is selected, a prefix or suffix is added to the displayed number
indicating its meaning.
<br><tt>Hit Popup and Information</tt>: The first line has been reordered for clarity.
<br><tt>Hit filter</tt> did not have default None checked.
<br><tt>Sequence filter</tt> flipped option now waits until Save to flip (for consistency).
</ul>
</small>
<p>
<li>Developer
<small>
<ul>
<li>Removed props.ProjectPairs; renamed ProjectPool to be PropsDB and cleaned it up. It returns
isAlgo1 so that the gene popup knows whether to show the exon score.
<li><tt>sequence/Sfilter.java</tt> now contains all filter code. The KB drop-down were not sized; hence,
big on Linux.
<tt>history/History.java</tt> has all un-used code removed. All control bars are created as a row,
which makes the delimiter full depth.
</ul>
</ul>
</small>

<a id="v551"></a>
<h3>Release: v5.5.1 (14-Mar-24)</h3>
<p>This release has multiple small interface changes, as follows:
<ul>
<li><ttb>2D-display</ttb>
<ul>
<li><ttl>Control panel</ttl>: Has a new option to shrink the area between the tracks.
<li><ttl>Sequence Filter</ttl>:
<ol>

<li><ttx>Gene#</ttx>: There is a new filter to show the <ttx>Gene#</ttx> beside each gene.
<li><ttx>Hit %Id</ttx>: When the <ttx>Hit %Id</ttx> option is off, the length of the vertical
line is constant; when it is on, it is proportional to the %ID. When the <ttx>Hit Length</ttx>
is off, this line is not shown.

<li>(1) The filter window has been re-organized and the track number has been added to the title.
(2) Fixed a bug in setting the start and end coordinates, which occasionally did not work.
(3) If there was more than 3 tracks, the <ttx>Conserved</ttx> option was only available
for the first occurrence of the reference track; it is now present for all occurrences.
(4)
</ol>
<li>Small changes and fixes:
(1) When a 2D display was initiated from the dotplot, one track could be longer than the other;
this has been fixed (problem introduced in v5.5.0).
(2) The <ttl>Hit Filter</ttl> <ttx>Highlight Hit popup</ttx>, when turned off,
was still highlighting the hit gene(s); it no longer does.

</ul>
<br>
<li><ttb>Dot plot</ttb>
<ul>
<li>The <ttl>Stats</ttl> pull-down has been changed to an "i" icon button for the help, and
an "s" icon button for the statistics. Instead of writing the statistics to the terminal,
they are written to a popup (where the text can be copied).
</ul>
<br>
<li>Developer
<small>
<ul>
<li>Changed most 'long' to integers (there were still 32-bit machines when this was 1st developed).
<li>2D merge overlapping hits for display; it smooths out the hit length display a little for
highly conserved genomes.
<li>2D Control Panel changed Select pulldown to popup menu. Add space between control buttons
(was squashed on linux). Less space is used between tracks when there are &gt;2 tracks.
The display from the Dot plot is back to pre-v550 length. On gene popup, was not suppose to
show exon score for Algo1, but was showing 0.
</ul>
</small>

</ul>

<a id="v550"></a>
<h3>Release: v5.5.0 (3-Mar-24)</h3>
For existing databases, there are no changes (though there are for <a href="#v548">v5.4.8</a>).
<br>To use SyMAP v5.5.0, you must have Java v17 or later.
<p>
<ul>
<li>The <tt>symap.jar</tt> was compiled with Java version 1.8 (8); it is now compiled with Java 17, which
is more secure, faster and was released Sept 2021.
<li><ttb>2D display</ttb>
<ol>
<li>This display is much faster (really noticeable on slow machines or big chromosomes).
<li>The <ttl>Hit Length</ttl> line was only shown if the hit-wire
was shown; this is no longer the case.
<li>The <ttl>Stats</ttl> button was removed, and the <ttl>Help</ttl> was moved to a little "i" button,
which provides a popup of quick information.
</ol>
<li><ttb>Cluster Hits Algo1</ttb>
<ol>
<li>It no longer splits large genes since #2 above handles them better.
<li>The <ttl>Summary</ttl> now refers to this algorithm as <ttx>Algo1 (modified original)</ttx> and the alternative
as <ttx>Algo2 (exon-intron)</ttx>.
</ol>
<li>Developers: Lots of cleanup to remove OO spagetti code in 2d.
<small>
<ul>
<li>Recreate tracks on new display instead of trying to reuse.
<li>Move number.GenomeNumber to Sequence; remove the number class.
Changed all numbers in DrawingPanel from objects to long.
<li>Merge History class with HistoryControl, and remove HistoryObject.java.
	Move HistoryControl and HistoryListener to symap.drawingpanel; remove history class.
<li>Merge abstract Track.java with Sequence.java. Merge FSized2d.java with Frame2d.java.
Rename frame.TrackInfo to ChrInfo and frame.Mapper to LeftMap
</ul>
</small>
</ul>

<a id="v549"></a>
<h3>Release: v5.4.9 (20-Feb-24)</h3>
For existing databases, there are no changes (though there are for <a href="#v548">v5.4.8</a>).
<ul>
<li>v5.4.8 little fixes: (1) The new <ttl>Query</ttl> <ttx>Multi</ttx> option had a small bug due to it using ">" instead of ">=".
(2) The <ttl>Query</ttl> <ttx>Every*</ttx> option works for <ttx>Algo1</ttx> as of v5.4.8, but was still disabled.
(3) The <ttl>Summary</ttl> gave wrong message if <ttl>EI</ttl> or <ttl>En</ttl> were unchecked.
<li>The <ttl>Query</ttl> for selected chromosomes and locations was confusing. It is now consistent and properly documented.
<li>Improved documentation.
</ul>

<a id="v548"></a>
<h3>Release: v5.4.8 (13-Feb-24)</h3>
For existing databases, rerun the <ttb>Synteny</ttb> algorithm.

<p>This release fine tuned the <ttx>Cluster Hit Algo2</ttx> (first released in <a href="#v546">v5.4.6</a>)
and made some adjustments to <ttx>Cluster Hit Algo1</ttx> for new interface feature.

<ul>
<li><ttb>Synteny Clustering Algo2</ttb>:  See <a href="SystemHelp.html#clust">Cluster hits</a>.
<ol>
<li>The algorithm better differentiates between overlapping genes which have the same aligned hits;
one will be the <tt><u>major</u></tt> assignment and displayed for the hit,
and others will be <tt><u>minor</u></tt> assignments.
There is slight improvements to the assignments of clustered gene hits.
<li>The <ttx>Cluster Hit Algo2 Parameter</ttx> defaults check EI and En, along with EE.
<li>The g1 (one gene) and g0 (no genes) have improved assignments.
The rules for filtering have been fine-tuned.
<li>The algorithm for finding piles and selecting what clusters to retain has been improved.
<li>Two scores are computed, which are the percent of the clustered hit that aligns to exons and to the genes.

<li>The hit strands could be different from the two gene strands; these can be printed out with <tt>symap -wsp</tt>
or excluded with <tt>symap -wse</tt> .
<li>Previously, the hit strands could be assign "+/-" when it should be "-/+" and vice versa;
this resulted in the wrong sequence being
reversed in the <a href="UserGuide.html#align">2D-Align</a>; this has been fixed.
</ol>
<li><ttb>Synteny Clustering Algo1</ttb>:
<ol>
<li>Compute the percent of the clustered hit that aligns to the gene.
<li>Make an addition so that it works with the <ttl>Query</ttl> <ttx>Multi</ttx> command (otherwise,
all results will be shown twice).
<li>The same problem as shown in 7 above happens with <ttx>Algo1</ttx>. It is mostly fixed, though a few
slip through. See <a href="UserGuide.html#strand">Strand</a>.
</ol>
<li><ttl>2d-display</ttl>:
<ul>
<li>The <ttx>Annotation</ttx> (yellow boxes of text) have less information so more boxes can be viewed at a time.
<li>The gene popup text shows the start/end of each hit on the opposite chromosome in place of the subhits; it shows ALL hits
for the gene between the two chromosomes even if they are not visible;
it shows the gene overlap (and exon overlap for <ttx>Cluster Hit Algo2</ttx>). Both the hit and gene popups have some minor formatting changes for clarity.
See <a href="UserGuide.html#info">Gene and Hit Information</a>.
<li>The <ttl>Align</ttl> display has the <ttx>Gene#</ttx> along with the <ttx>Id</ttx> and <ttx>product (Desc)</ttx>.
</ul>

<li><ttl>Query</ttl>:
<ul>
<li>Add a column to show the percent exon overlap of the hit (<ttx>Cluster Hit Algo2</ttx>) or
the percent gene overlap (<ttx>Cluster Hit Algo1</ttx>).
<li>Search changes:
<ul>
<li>A new <ttx>Multi</ttx> gene annotation option has been added, which lists all genes with multiple hits.
<li>The <ttx>Every*</ttx> now works for both <ttx>Algo1</ttx> and <ttx>Algo2</ttx>, and the '+' was replaced with a '*' (so would
not be confused with strand).
<li>The <ttx>Hit#</ttx> query option shows all major and minor genes that overlap the hit, where the minor are suffixed
with an '*'.
<li>The <ttx>Gene#</ttx> query now allows a chromosome to be selected for each species, and shows
all major and minor hits.
<li>The <ttx>Annotation</ttx> search with one chromosome selected has been changed so that the annotation has to be on that chromosome
(it was allowing the annotation to be on the paired chromosome).
<li>The <ttx>Export</ttx> for .csv and .fasta allow appending to an existing file.
</ul>
</ul>

<li><ttl>ConvertEnsembl</ttl>: Was not replacing special characters, which has been added.
<li>Developers
<ul>
<li><small>The formatter for the yellow annotation box was re-written.</small>
<li><small>The Synteny clustering Algo2 was rewritten to be clearer.
The amount that a cluster hit overlaps the exon of a gene was an approximation, now it takes into account overlapping
subhits and removes the overlap from the calculation.</small>
</ul>
</ul>
<a id="v547"></a>
<h3>Release: v5.4.7 (14-Nov-23)</h3>
For existing databases: if using <ttx>Cluster Hit Algo 2</ttx>, rerun the <ttb>Synteny</ttb>.
<ul>
<li><ttl>Query</ttl>: A new <ttx>Every+</ttx> option for databases build with <ttx>Cluster Hit Algo 2</ttx> post-v546.
It shows all genes with a hit aligned to it, even the overlapping ones.
This also is used for the <ttx>Annotation</ttx> search.
<li><ttl>Alignment&amp;Synteny</ttl>: Fixed a bug in the new <ttx>Cluster Hit</ttx> algorithm; the bug did not effect the visible
results except to have a slight fractional difference in the <ttl>Summary</ttl> <ttx>GNwCH</ttx> statistics,
and the <ttl>Query</ttl> <ttx>NumHits</ttx> column on a few genes. The new <ttx>Every+</ttx> option is
disabled for pre-v5.6.7 database, as that shows this problem.
<li>Developers
<ul>
<li><small>Removed deprecated for Java v20: "exec(String) in Runtime has been deprecated" and "URL(String) in URL has been deprecated".
All export routines call Globals.getExport(). Set some gray buttons to white for linux.</small>
</ul>
</ul>

<a id="v546"></a>
<h3>Release: v5.4.6 (22-Oct-23)</h3>
<p>For existing databases: There will be a small database update on the first viewing.
For the new cluster hit algorithm, change the <a href="SystemHelp.html#clust">Cluster parameters</a> and rerun the <ttb>Synteny</ttb>.

<ul>
<li><ttl>Alignment &amp; Synteny</ttl>: A new <ttx>Cluster Hit</ttx> algorithm has been written to filter and cluster MUMmer hits;
the improvement over the original algorithm is that it has knowledge of the exon-intron structure and
identifies all gene pairs. See <a href="SystemHelp.html#clust">Cluster parameters</a>.

<li><ttl>Query</ttl> has a new column that states if a hit is to an exon-exon, exon-intron, etc. This is works
with the new cluster hit algorithm. This information is also on the <ttx>Hit Information</ttx>.
<li>Minor stuff
<ul>
<li><ttb>2D Sequence Filter</ttb> <ttl>Gene#</ttl> The full gene is highlighted instead of just the introns.
<li><ttb>Summary</ttb> The <ttx>Explain</ttx> did not format right with Java v20, which has
been fixed. The date at the top had "Oct" swapped with "Nov".
<li><ttb>A&S Cancel</ttb>: SyMAP was automatically removing any added hits or blocks, but for big databases, the removal
can take a long time, hence, there is a popup to request whether the user wants this done; otherwise, they may
prefer to just remove the database and start over.
<li><ttb>Summary</ttb> contains the number of MUMmer files read when using previous alignments (this provides a clue
as to whether <ttx>Concat</ttx> was used).
<li>The documentation stated that all genes with a hit could be shown in the query, which was not true; the documentation
has been updated.
</ul>
<li>Developers
<ul>
<li><small>The original <ttx>Cluster Hit</ttx> code has been moved to backend.anchor1. The SyProj was renamed to backend.anchor1.Proj, and now
only used by the <ttx>Cluster Hit</ttx> code. The new <ttx>Cluster Hit</ttx> code is in backend.anchor2.</small>
</ul>
</ul>

<a id="v545"></a>
<h3>Release: v5.4.5 (4-Aug-23)</h3>
<ul>
<li>Tested on <a href="SystemGuide.html#machine">Linode Ubuntu</a>.
<li><ttb>Chromosome Explorer 2D</ttb>
<ul>
<li><ttl>Sequence Filter</ttl>: See <a href="UserGuide.html#sequence_filter">Sequence Filter</a>
<ul>
<li><ttl>Gene#</ttl>: A new option of <ttl>Gene#</ttl> has been added, which will set the coordinates
of a 50kb region around the gene and highlights the gene's introns.
<li><ttl>Block#</ttl>, <ttl>Collinear#</ttl>: There are new options to show the <ttl>Block#</ttl> or <ttl>Collinear#</ttl> in the same way
as the  <ttl>Hit#</ttl> and <ttl>Hit %Id</ttl> are shown.
<li><ttl>Highlight Conserved Genes</ttl>: There is a new option to  <ttl>Highlight Conserved Genes</ttl>.
See <a href="UserGuide.html#conserved">Conserved</a>.
</ul>
<li><ttb>Hit Filter</ttb>
<ul>
<li>The <ttl>Highlight hit popup</ttl> can be on all the time (i.e.
it is not turned off if another highlight is turned on), and is highlighted in the hover color.
<li>When the hit popup highlights the hit-wire, it will also highlight the genes
assigned to it.
<li>The hit popup provides the <ttx>Gene#</ttx> of the genes that are being highlighted.
</ul>
</ul>

<li>Minor stuff
<small>
<ul>
<li><ttb>Query Table</ttb> Fixed tiny bug from v5.4.2 where the <ttl>Glen</ttl> column would not sort.
<li><ttb>Query Align</ttb> with MUSCLE, the <ttl>Cancel Alignment</ttl> button does not work, but now, it
gives a message on how the user can cancel the alignment.
<li><ttb>Block View</ttb> A clock shows when bringing up the 2D view (necessary for slow machines).
<li><ttb>2D Sequence Filter</ttb> (1) The highlighted popup gene now highlights the whole gene
(it was just the intron area and delimiter). (2) Setting sequence filters could cause consecutive duplicate views in History, i.e.
forward/back buttons, this has been fixed. (3) Old bug: when a track was reused with a different sequence,
the conditional checkboxes could be incorrectly disabled (still one problem left but need to rewrite masses of code
to fix).
</ul>
</small>

<li>Developer
<small>
<ul>
<li>Rename frame.TrackCom to frame.TrackInfo;
</ul>
</small>
</ul>

<a id="v544"></a>
<h3>Release: v5.4.4 (18-July-23)</h3>
<ul>
<li><ttb>Query</ttb>
<ul>
<li><font color=red>Bug fix for v5.4.3</font> The extension to make the <ttl>Gene#</ttl> search work with
suffixes (e.g. 5.a) had a bug when used without suffixes.
</ul>
<li><ttb>Chromosome Explorer 2D</ttb>
<ul>
<li><ttl>Sequence Filter</ttl>: There is a new option for highlighting a gene called <ttl>Popup</ttl>;
when this option is on, every time a gene popup window is created, the gene stays selected. The filter popup has
reduced number of options.
See <a href="UserGuide.html#popup">Popups</a> and <a href="UserGuide.html#sequence_filter">Sequence Filters</a>.
<li>The highlight <ttl>Hit popup</ttl> and <ttl>Gene popup</ttl> options are on by default.
</ul>
<li>Developers
<ul>
<li><small>Remove CloseupListener and DrawingPanelListener. Remove
DrawingPanelData (merge with DrawingPanel). Move TextData from util to sequence directory.  Moved frame SyMAP2d,
Frame2D, FSized2d, ControlPanel to drawingpanel because used exclusively by 2d. Rename AlignMain to AlgSynMain.
</small>
</ul>
</ul>

<a id="v543"></a>
<h3>Release: v5.4.3 (12-July-23)</h3>
There are numerous tiny changes, fixes, and two small new features.
<p>For existing databases: (1) View the database with <tt>./symap</tt> and okay the database update.
(2) Though it is not necessary to rerun the <ttb>Synteny</ttb> algorithm, there is a small improvement
for assigning hits to embedded genes.
<ul>
<li><ttb>Manager</ttb>
<ul>
<li><ttl>Synteny</ttl> For embedded genes, the algorithm was assigning the hit to the outer gene,
now it assigns it to the closest fit gene.
<li><small><i>Minor</i>: When <ttl>All Pairs</ttl> or <ttl>Selected Pair</ttl> is selected, the user will be asked to confirm
the processing. The panel is a little smaller. All frames show the database name.
Allow blank lines on annotation input. Add "Prefix" on manager summary.</small>
</ul>

<li><ttb>Chromosome Explorer 2D</ttb>
<ul>
<li><ttl>Hit Filter</ttl>: There is a new option for highlighting a hit-wire called <ttl>Popup</ttl>;
when this option is on, every time a hit-wire popup window is created, the hit-wire stays selected.
See <a href="UserGuide.html#popup">Popups</a> and <a href="UserGuide.html#hitfilter">Hit Filters</a>.
<li>When more than two sequence tracks are shown, the middle one now knows the hits on both
sides and can list them in the <ttl>Gene Popup</ttl> and <ttl>Align (Max 30kb)</ttl>.
<li><small><i>Minor</i>: The <ttl>Sequence Filter</ttl> options of <ttl>Show Text</ttl> now shows the
text on the outside of the sequence track rectangle. The panel is a little smaller.
The "Id %" is now shown on the <ttl>Hit Information</ttl>. Put <ttl>Info:</ttl> at beginning of controls.
Text popups stay in front, i.e. will not be buried behind panel. Improve the 2D information help. </small>
<li><small><i>Tiny bug fixes</i>: The end coordinate on this filter window always showed the end of the sequence
instead of the displayed end. If the <ttl>Dot Plot</ttl> was changed multiple times during
a given session of <ttl>ChrExp</ttl>, it sometimes messed up. The <ttl>Dot Plot</ttl> filters
could not be changed after the first view.</small>

</ul>
<li><ttb>Query</ttb>
<ul>
<li>The <ttl>Gene#</ttl> query option now allows a suffix; i.e. '100' will return all genes with this number
and any suffix, '100.b' will only return this gene.
<li><font color=red>Bug fix</font>: if an annotation query was performed that had a hit to a gene with embedded genes,
the gene and the overlapping genes would be returned.
<li><small><i>Minor</i>: Has a new button <ttl>Defaults</ttl> which sets the selected columns
to the defaults. In the result table, the <ttl>Gene#</ttl> is now the
first column.</small>
</ul>
<li><ttb>DotPlot</ttb>: <small><i>Minor</i>: The <ttl>Filters Defaults</ttl> option no longer closes
the window (behaves like the <ttb>2D</ttb> filters).</small>

<li>Developer
<ul><small>
<li>
The database 'props' table now keeps track of initial database version and updates.
The file util.Utilities has all the methods for
parsing the sequence 'tag'; the database tag field has a slightly different format.
Fix a tiny problem where Query would set a 2D static 'show annot' which would effect other 2D displays.
ShowContinue put 'Cancel' as the end button. Remove [Uniprot ID: xxx] from demo_seq/annotation/anno.gff.

<li>Moved track/* to sequence directory. Merged SequenceTrackData into TrackData.
Removed AnnotationData and PseudoData (was just moving data around). Rename PseudoPool to SeqPool.
Remove GroupSorter (code moved to SeqLoadMain) and improved. Change database field pseudo_hits.evalue to 'htype'
and rearrange the fields to be more logical.
</small>
</ul>
</ul>


<a id="v542"></a>
<h3>Release: v5.4.2 (28-June-23)</h3>
<ul>
<li><ttb>2D</ttb>
<ul>
<li>Redesigned the <ttl>Sequence Filter</ttl>.
<li>History only kept 10 places, now it keeps 50.
<li>Fixed bugs:
<br><small>(1) when the 2D was initiated from the DotPlot or Query, changing highlight
could change "Show" option. (2) Recent v541 DotPlot to 2d bug:
If a second block was selected for 2D, the genes would not be shown; with this fix, multiple 2D from Dotplot can be shown again.
</small>
</ul>
<li><ttb>Query</ttb>
<ul>
<li>Added an export option of HTML.
</ul>
<li>Developer:
<ul>
<li><small>The Abstract Filter file was removed and the Hit and Sequence filter codes rewritten without it.
Changed number of tracks from 100 to 30.</small>
</ul>
</ul>

<a id="v541"></a>
<h3>Release: v5.4.1 (22-June-23)</h3>
For existing SyMAP databases: there is a new column in <ttb>Query</ttb> called <ttl>NumHits</ttl>;
run <tt>symap -y</tt> to update all projects in the database.

<ul>
<li><ttb>Manager</ttb>
<ul>
<li><font color=red>Bad bug fix</font> for v5.4.0: Loading the hits would get a memory crash
if some chromosomes in a project had genes and other chromosomes did not.
<li>If only two projects are selected from the left, then their joint cell is automatically selected.
</ul>

<li><ttb>Dot Plot</ttb>
<ul>
<li><ttl>Info</ttl>: (see <a href="UserGuide.html#dotplot_display">Control Panel</a>)
A new pull down with options:
(1) <ttl>Stats</ttl>: in the <ttl>Information</ttl> area (short form),
(2) <ttl>Print</ttl>: printed to the terminal (long form), (3) <ttl>Help</ttl>.
<li>The dot colors can be changed with the Color Wheel (see <a href="UserGuide.html#image">Color Wheel</a>).

<li>Dotplot can no longer display multiple 2d from the same dotplot display. (v542 allows many again)
</ul>

<li><ttb>2D</ttb>
<ul>
<li><ttl>Info</ttl>: A new pull-down with options:
(1) <ttl>Stats</ttl>: Hovering in the gene space but not on a gene will
show summary info; hovering in the hit space but not a on hit will show hit summary
info. (2) <ttl>Help</ttl> Help will be shown instead of summary information.
<li>Display bug fix: If the "Show" filter was changed to show more hits, then changed back to show less,
the right click in hit empty space would keep showing hit-wire information.
</ul>

<li><ttb>Query</ttb>
<ul>
<li><ttl>Gene#</ttl> in the <ttl>Pair Hits</ttl> section: This query now works exactly like the other 3 in this row, i.e. you can set the
chromosome, and it returns full hits. NOTE: there will be no rows returned if there is no hit to
the gene.
<li><ttl>Single NumHits</ttl>: A column has been added to indicate whether there are any hits to the gene
in the entire database (not just those queried).
<li>Column Description: This was being shown on the bottom of display; it is now a tool tip.
<li><ttl>Clear Column</ttl>: This button has been added. It replaces the <ttl>Show Stats</ttl> button
when the columns are being shown.
</ul>

<li>Developer:
<ul>
<li><small>The MySQL connection has been rewritten to be more straight-forward and confirm that connections
are correctly closed. There is a new dedicated connection for any each of the displays (e.g. DotPlot, 2D, etc).
This removed 4 java files and replaced one.</small>
<li><small>MacOS Monterery no longer has "Times and Lucida", used by Serif, neither does JDK until v18, so
these were removed. However, removing this only occurrence does not get rid of the Java message.</small>
</ul>
</ul>

<a id="v540"></a>
<h3>Release: v5.4.0 (6-Apr-23)</h3>
This release focuses on improving the detection of gene homology in the <ttb>Synteny</ttb> computation.
<br>For existing databases, rerun <ttl>Alignment&Synteny</ttl> using existing MUMmer files.
<ul>
<li><ttb>Synteny</ttb>:
<ul>
<li>Creating clustered hits has been improved for long genes.
<li>Filtering hits has been updated to give preference when at least one end is a gene.
<li>Details of changes:
<small>
<br>&rarr; Long hits with embedded genes are split, which helps detect hits to the embedded genes
(15Oct24 disabled since graphics was fixed to show these).
<br>&rarr; When a candidate gene is created from hits, it is checked that it does not span a gene with no hits.
Candidate genes used to be created using the maximum length of the annotated genes; this is now maximum size 100k.
<br>&rarr; For finding the top N hits, the Max of query and target(summed subhits (minus overlaps)) of a clustered hit are used
instead of summed query subhits.
<br>&rarr; A clustered hit could have a mix of subhits that were on opposite strands and the same strand; this does not
happen anymore. (This is rare.)
<br>&rarr; %Identity was just using the query match; now it is Max(pct(query), pct(target)) where pct
is the approximate average %identity of the subhits.
<br>&rarr; %Similarity was the value of the first subhit of a clustered hit;
now it is computed like the %Identity.
<br>&rarr; Pair Parameters: The TopN and MinDots parameters are checked for integer and &gt;0.
When A&amp;S is executed, the non-default parameters are printed to the terminal immediately.
</small>
<li>These changes increased the number of gene homology hits for Humans to Mouse (Chrs 5,17,X) from
(68.6%, 57.0%) to (79.9%,66.3%), and for Humans to Chimpanzee (Chrs 5, X) from (53.9%, 54.6%) to
(69.9%, 56.7%).
<li>Note, these changes cause more hits to be saved, and hence, the synteny takes
a little longer. It may also remove or add small blocks due to additional intervening hits.
</ul>
<li><ttb>Summary</ttb>: See <a href="UserGuide.html#sum" class="ext" target="_blank">Summary</a>.
<ul>
<li>The summary has been rewritten in a textual format which can be copied with copy & paste.
<li>All hit and block statistics were computed in terms of <i>clustered hits</i>, which may have
gaps and overlaps between the subhits. Now many of the statistics are in terms of the subhits, and it is
clear as to whether clustered hits or their subhits are being used.
<li>Collinear sets have been added to the summary.
<li>The Project summary has exon stats, which have been removed from the <ttl>View</ttl>; the gene size ranges have
been removed.
<li>It states whether either input project had <ttl>Order Against</ttl> or <ttl>Mask genes</ttl> set.
</ul>

<li><ttb>Query</ttb>:
<ul>
<li><ttl>Hit Cov</ttl> is a new hit column, which is the maximum subhit coverage
between the two species connected by the clustered hit. The <ttl>Hit Cov</ttl> is computed by summing
the subhits and removing any overlap.
<li><ttl>Annotated (Gene Hit)</ttl> has an additional option of <ttl>Either</ttl>.
<li><ttl>Collinear size</ttl> now has the options of <ttl>&gt;=</ttl>, <ttl>=</ttl>, <ttl>&lt;=</ttl>.
<li><ttl>Export CSV</ttl> has an option to not include row number column. This is useful for comparing two
lists, where the row number can cause most rows to appear different. The quotes around columns have been
removed; any comma's in the descriptions are replaced with ";".

<li>The <ttl>Show</ttl> was disabled for singles; it has been enabled.
</ul>
<li><ttb>Chromosome Explorer</ttb>:
<ul><li>The text of the Hit popup has been slightly changed and the Subhit
coverage has been added (see <a href="UserGuide.html#popup" class="ext" target="_blank">Hit Popup</a>).
</ul>
<li>Developer:
<ul><li><small>The Hit assignment algorithm has been restructured.</small></ul>
</ul>
<a id="v535"></a>
<h3>Release: v5.3.5 (3-Mar-23)</h3>

<ul>
<li><ttl>Cancel</ttl>
<ul>
<li>The <ttl>Cancel</ttl> button did not always work well, now it does.
See <a href="SystemGuide.html#cancel" class="ext" target="_blank">Cancel</a>.
<li>From v5.3.4, if the <ttl>A&S</ttl> was cancelled or interrupted, it was not possible to restart the <ttl>A&S</ttl>
without reloading the projects.
</ul>
<li><ttl>Load project</ttl>
<ul>
<li>Any sequences that do not have the <ttl>Group prefix</ttl> will be ignored.
<li>The <ttl>Group prefix</ttl> is case-insensitive.
</ul>

<li><ttl>Align (Max 30000)</ttl>
<ul>
<li>The selected region was extended to the whole genes within it; now it just extends to the ends of the last overlapping exon of the selected region.
<li>The exons are labeled, e.g. Exon #11. If the end of the gene is not shown, a left or right arrow will be displayed
indicating that it continues. See <a href="UserGuide.html#align">Align</a>.
<li>The color icons now has Exon+ and Exon-, where the default colors correspond to the 2d display.
<li>Bug fix: when there were overlapping genes, the exons could be wrong for one of them.
</ul>
<li>Developer:
<ul><li><small>The Logger interface has been removed.
In the ManagerFrame, the load methods have been moved to LoadProj, and the remove methods have
been moved to RemoveProj. The AnchorsMain methods has been reworked for clarity, but provide same answer.
The debug statements were cleaned up.
</small></ul>
</ul>

<a id="v534"></a>
<h3>Release: v5.3.4 (19-Feb-23)</h3>
Existing SyMAP databases: if the project parameters seem wrong, bring up the <u>Parameters</u> window
and <ttl>Save</ttl>.
<ul>
<li>All "Help" in the SyMAP <ttb>Project Manager</ttb> has changed to <big>?</big> for online help.
The online has been improved for the parameters explanation;
see <a href="./SystemHelp.html" class="ext" target="_blank">Quick Guide and Parameters</a>.
<p>
<li>The <ttb>Project Parameters</ttb> window is clearer;
see <a href="./SystemHelp.html#projParams" class="ext" target="_blank">Project Parameters</a>.
Since parameters are written to file and database, they can get out of sync; this problem is reduced.
The popup messages are more concise.
<li>The <ttb>Pair Parameters</ttb> window is clearer;
see <a href="./SystemHelp.html#pairParams" class="ext" target="_blank">Pair Parameters</a>.
The parameters used to only be applied if A&S was run right after setting them; now they are
maintained across symap sessions.
<p>
<li>
Developers:
<ul><li><small>There has been a lot of renaming and restructuring. For example, the symap.projectmanager.common package has been shortened to symap.manager,
and a few of the classes in this package were renamed for clarity. All ChrExp and 2d display
classes have been renamed and moved to the frame package. A database package has been
created for the database specific classes. The project and pair parameter code has been redesigned.
</small></ul>
</ul>

<a id="v533"></a>
<h3>Release: v5.3.3 (24-Jan-23)</h3>
This release has various minor changes and additions.
<ul>
<li><ttb>Synteny&Alignment</ttb>
<ul>
<li>Synteny algorithm: The block start and end was assigned coordinates at the mid-point
of the first and last hit; this has been changed to the start of the first hit and end of
the last hit.
<li><ttl>Selected Pair Parameters</ttl>: The <ttl>No Overlapping Blocks</ttl> was removed because it does not work.
The <ttl>Summary</ttl> shows changes from defaults for the <ttl>MinDots</ttl>, <ttl>TopN</ttl> and <ttl>Merge Blocks</ttl> parameters.
</ul>
<li><ttb>Print Image</ttb>
<ul>
<li>Most SyMAP graphics have a print image icon, which was
not working for post Java V8; it now works for any version.
</ul>
<li><ttb>Dotplot</ttb>
<ul>
<li><ttl>Filter</ttl> - see <a href="UserGuide.html#dotplot_filter">DotPlot Filter</a>.
<ul>
<li>
New <ttl>Green</ttl>, <ttl>ID%</ttl>, <ttl>Mix</ttl> and <ttl>Show Block Numbers</ttl> options.
The dot size was moved from the main display to the filter. Add  <ttl><big>?</big></ttl> button.
<li>The filter display now freezes the other panels.
It no longer resets to default Filters on Home or selecting a cell.
</ul>

<li>The cells display all at once; before they were added one at a time.
<li><ttl>Home</ttl> now works for the 2-chromosome dotplot, which goes back to full view.
</ul>

<li>Developer:
<ul>
<li>Compiles under Java 17 with no warnings (removed Observable). Massively reduced and cleaned
Dotplot code. Removed many un-used files from the util package and moved some others to the package of
their sole use. Removed arrow from the end of the gene in the CloseUp. The <tt>classes_ext</tt> have
been cleaned of most unused classes and files.
</ul>
</ul>

<!----------------------------------------------------------->
<a id="v532"></a>
<h3>Release: v5.3.2 (14-Jan-23)</h3>
This release added <ttl><big>?</big></ttl> buttons for online help, and has some small <ttl>Query</ttl> and
<ttl>2D Color Icon</ttl> updates.
<ul>
<li>Load and synteny: <ttl>View</ttl> will show the sequence and annotation paths used along with their date.
If an existing alignment is used, then <ttl>Summary</ttl> will show the date that the alignment was run.
<li><ttl>Chromosome Explorer 2D</ttl>
<ul>
<li>Color Icon
<ul>
<li><ttl>Default</ttl>: Changed from resetting all colors to just resetting the colors for the current tab.
<li><ttl><big>?</big></ttl>: a new button that links to the online user guide for instructions.
<li>Fixed an old bug where the <ttl>Default</ttl> did not always work until <ttl>viewSymap</ttl> was restarted.
</ul>
<li>The <ttl>Sequence</ttl> and <ttl>Hit filter</ttl>, and the <ttl>Selected</ttl> pull-down,
have <ttl><big>?</big></ttl> links to their online help.
</ul>
<li><ttl>Queries</ttl>
<ul>
<li>The selected columns will be saved, as detailed in <a href="Query.html#save">Save Columns</a>.
<li>The <ttl>Help</ttl> has been changed to <ttl><big>?</big></ttl> for online documentation.
<li>A new html page has been created for the <ttl>Queries</ttl> at <a href="Query.html">Query</a>
<li>The columns would sometimes be displayed in a weird order, which has been fixed.
</ul>
<li>Developers: Created <tt>utils/Jhtml.java</tt> that has all html links and code in one place. The <tt>/properties</tt>
only contains colors and mainly those that can be changed by user.
</ul>

<a id="v531"></a>
<h3>Release: v5.3.1 (6-Jan-23)</h3>
Improvements to viewing the alignments of hits; see <a href="UserGuide.html#align">Alignment</a>.
<ol>
<li>2D
<ul>
<li><ttl>Align (Max 30,000)</ttl>
<ul>
<li>Each MUMmer subhit is now shown at the bottom as a full dynamic programming alignment (it was broken up
at each gap).
<li>The MUMmer subhits can overlap, so they are now staggered in the graphic display.
<li>The sequence is no longer converted to uppercase, i.e. if the sequence is soft-masked, it will show on the alignment.
<li>The hit color can be set in the color wheel.
</ul>

<li><ttl>Hit Information</ttl>:
<ol>
<li><ttl>Align Hit</ttl> <font color=green>New</font> :  When hover over a hit-wire and right-click, an information box pops up; that box now has
an <ttl>Align Hit</ttl> button that shows a text alignment in blast-style format.
<li>The Hits for each track are listed in descending start order, and the gap is negative if it overlaps with the previous.
</ol>

<li><ttl>Show Seq Options</ttl> <font color=green>New</font>: The <ttl>Show Sequence</ttl> is now <ttl>Show Seq Options</ttl> which has a popup menu that allows the user
to select whether to show the hit sequence, reverse hit, or region from the selected area.
<li><ttl>Sequence Filter</ttl>: Has a <font color=green>New</font> option <ttl>Show Hit#</ttl> which shows the hit# in the same location as the <ttl>%Id Value</ttl>
(only one of these two option can be active at a time).
<li>Tiny problems fixed:
(1) Fixed a problem where a sequence could not be selected for <ttl>Align</ttl> or <ttl>Show Sequence</ttl> if the zoom region was small.
(2) When the view went from 2D->Circle->2D, the "Selected:" would not be reset to "Zoom All".
(3) The arrow has been removed from the graphical align hit because it meaning was not clear.
(4) The term 'merged hits' was changed to 'clustered hits', where the cluster has sub-hits; the term
merged could be misleading.

</ul>

<li>DotPlot
<ul>
<li>Multiple 2d displays can occur now.
</ul>

<li>Developers: (1) Removed over 15 useless files of code. Removed used Cache from everything.
(2) Cleaned up the Mapper package hit code and Closeup package.
</ol>

<a id="v530"></a>
<h3>Release: v5.3.0 (10-Dec-22)</h3>
For existing SyMAP databases: the database schema will be updated on the first viewing of an existing database.

<p>This release does not have any new features, but FPC was removed from the schema and code, which is an major internal change.
<i>Typically the SyMAP software can view an earlier or later version SyMAP database, but that is
not the case with this release.</i>
<ol>
<li><ttl>Queries</ttl>: fixed a v5.1.9 bug in location filtering.
<li><ttl>Dotplot</ttl>: fixed a v5.1.9 bug; if a region was selected, the Hit Filter was set to show no hits.
<li><ttl>Github</ttl>:
<ul>
<li>Removed most of FPC from the code and Schema.
<br>
<li>Simplify classes_ext and Makefile
</ul>

</ol>
<a id="v521"></a>
<h3>Release: v5.2.1 (1-Dec-22)</h3>
For existing SyMAP databases: The colors in the circle view will be different for the 2nd project.
<ol>
<li><ttl>Circle view</ttl>
<ul>
<li>Whole genome:  <ttl>Reverse</ttl>: A new function that allows the reference to be reversed.
Also, the colors of a project's blocks stays consistent if compared to projects with different number of chromosomes (n&lt;=25).
<li>Chromosome view: The selected reference will have its colors shown by default.
<li>The project name that is reference is in bold italics.
</ul>

<li><ttl>Queries</ttl>:
<ul>
<li><ttl>High</ttl>: Added check box to unselect if the user does not want the selected hit highlighted.
<li>Improved the information for the MUSCLE alignment. See <a href="UserGuide.html#qview">Query Align</a>.
<li>Fixed <ttl>Clear Filters</ttl> to leave <ttl>Chr:</ttl> enabled.
</ul>
<li><ttl>SyMAP manager</ttl>:
<ul>
<li><ttl>View</ttl>: A new  link that brings up the chromosome information for a project.
<li>Collinear computation bug fixes: (1) self-chromosome can cause error so disabled. (2) If
a chromosome had no hits (can happen with small scaffolds), collinear was not computed.
</ul>

<li><ttl>2D Sequence Filter</ttl> bug: "Apply" could perform an unwanted change of boundaries on sequence.

<li>Developers: Removed 3D from code, classes_ext.tar.gz, Makefile.
Put in a check for no projects, as a user's trace showed it can happen.
Remove the Filtered interface. Removed more FPC code (Drawing Panel, Mapper Pool, SyMAP Frame).
Changed 2 Observers to PropertyChangeEvent. Reduce /props from 4 to 1 file.
</ol>

<!----------------------------------------------------------------->
<a id=v520></a>
<h3>Release: v5.2.0 (20-Nov-22)</h3>
For existing databases, see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>; the synteny
algorithm needs to be rerun to get the collinear changes.
<p>This release is concentrated on improving the algorithm and display of <u>collinear sets</u>;
see <a href="./Collinear.html">details</a>.
There is new viewing support for Collinear sets: see <a href="./UserGuide.html#hitfilter">2D Hit Filter</a>
and <a href="./UserGuide.html#qview">Queries</a>.

<ol>
<li>The <tt>symap</tt> and <tt>viewSymap</tt> scripts have been changed to access
<tt>SyMAPmanager</tt> instead of <tt>ProjectManagerFrame</tt>. NOTE: if you are updating, make sure you
get these new scripts!!

<li>Processing
<ul>
<li><u>Collinear sets</u>:
<ul>
<li>The collinear sets algorithm has been greatly improved for more accurate results.
The <tt>Alignment&Synteny</tt> needs to be rerun to compute the updated sets.
<li>Each collinear set has a unique number for the chromosome pair.
</ul>
<li>The <ttl>Hit#</ttl> used in Queries and the 2D display was the index into the database and could get very large.
A <tt>hit number</tt> is now computed during processing, which is a sequential number for the chromosome pair (e.g. there will be
a Hit#1 for each chromosome pair with hits).
</ul>

<li>Chromosome Explorer 2D display:
<ol>
<li><ttl>Hit Filter</ttl>:
<ol type="a">
<li>Has many new options, see <a href="./UserGuide.html#hitfilter">Hit Filters</a>
<li>The <tt>Cancel</tt> and <tt>Defaults</tt> buttons never did work right; they do now.
</ol>

<li><ttl>Color wheel</ttl>, for <ttl>Hits</ttl>, there are two new options:
<ol type="a">
<li><ttl>Hover</ttl> provides the color the hit-wire turns when the mouse hovers over it.
<li><ttl>Highlight1</ttl>  and <ttl>Highlight2</ttl> provides the highlight colors, which alternate
when Collinear sets or Synteny blocks are shown.
</ol>

<li><ttl>Sequence Filter</ttl>:
<ol type="a">
<li><ttl>Gene Delimiter</ttl>: This is a new option, which draws a black line over the top exon of each gene when
they are zoomed enough to see the exons. See <a href="./UserGuide.html#ex2d">2D view</a>.
</ol>
</ol>

<li>Queries
<ol>
<li>The <ttl>Show Synteny</ttl> has been changed to <ttl>View 2D</ttl> with the options (2nd two are new):
<br>(1) Region (kb), (2) Synteny Block, (3) Collinear Set. See <a href="./UserGuide.html#qview">Results</a>.
<li><ttl>Show</ttl>: This is a new option which produces a popup of all data for the selected row.
This allows copying any text.
<li>The collinear column displays the set as N.M, where N is the size and M is the unique set number.
<li><ttl>Gene#</ttl>: This is a new query, where one gene number can be entered (no suffix). This is
a good way to view all hits for a given gene.
<li><ttl>Hit St</ttl>: There is a new column, which is "=" if the two hit ends are equals, else it is "!=".
<li>A hit may align to overlapping genes, but only one hit pair is shown; now the best overlap for each
side is shown (before it was random).
</ol>

<li>Assorted minor stuff
<ul>
<li><tt>ConvertNCBI</tt> script has a new "-p" and "-pa" options to include as attributes the first protein or all protein names for the gene.
<li>Schema update: there is an schema update that occurs the first time an existing database is read.
<li>The <ttl>SyMAP Queries</ttl> button has been moved and renamed <ttl>Queries</ttl>.
The 2D Filter and Color Wheel panels will be displayed in the middle of screen.
The 2D <ttl>Information</ttl> box now shows the <ttl>Hit Filter</ttl> settings.
<li>The SyMAP version of running synteny is on the summary panel.
<li> All visible traces of FPC have been removed. From this release forward, FPC will no longer be supported.
However, there are still releases from AGCoL and Github that support it;
see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>.

</ul>
<li>Developers
<br>(1) All 'new Integer()' etc, have been removed.
(2) Schema update: The 'runum' has
been added to the hits table, which provided the set number for collinear pairs. The SyMAP version is now
stored in the projects and pairs table. The annotation table contains a new column for the number of hits.
(3) FPC code is all mixed in with the seq-to-seq; but it is no longer consistent with changes.
</ol>
<!--------------------------------------------------------------------->
<a id="v519"></a>
<h3>Release: v5.1.9 (26-Oct-22)</h3>
For existing databases, see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>.

<p>Release v5.1.7 to v5.1.9 greatly improved the <u>Gene# assignment algorithm</u> and the <u>2D gene placement</u>.
<p>Added -z option (obsolete): To execute the latest Gene# assignment algorithm, execute <tt>./symap -z</tt>,
then select <ttl>Reload Annotation</ttl>; the synteny algorithm does NOT need
to be rerun.

<h4>(25-Oct-22)</h4>
This release has some interface improvements to SyMAP Query and the 2D gene placement algorithm was improved.
<ul>
<li>SyMAP Project Parameters
<ul>
<li>The <ttl>Parameters</ttl> display has a new parameter
called <tt>abbrev_name</tt> which must be a 4 character abbreviation. This is used in the SyMAP Query
column headings. If one is not entered, then the last 4 characters of the <tt>display_name</tt>
is used.
<li>Bug fix: A <ttl>Save</ttl> after annotation was loaded would zero out the counts for annotation keywords, which
resulted in no annotation keywords (only All_anno) being shown in the <ttl>Query Columns</ttl>; this no longer happens.
</ul>
<li>2D display
<ul>
<li>The gene placement algorithm was improved.
</ul>
<li>SyMAP Query
<ul>
<li>Hit Pairs Table: The <ttl>Location</ttl> columns now include the Gene start/end/len and the strand assigned to the gene in the GFF file.
<li>Single Genes Table: Only shows the columns that are relevant to genes.
<li>Hover over column: the description is shown at the bottom.
</ul>
<li>Load Annotation: This computes the Gene#, and was allowing a gap of 3 without calling them overlap;
this has been changed to 0.
<li>Demos: A few small changes to the annotation files. Params updated.
<li>Developers: Cleaned-up trace and debug flags.
</ul>
<!--------------------------------------------------------------------->
<h3>Release: v5.1.8 (17-Oct-22)</h3>
For existing databases, see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>.

<p>This release continues to concentrates on making the 2D display provide more information.

<ul>
<li>Processing
<ul>
<li>Gene# assignment: the algorithm has been redone so that all overlapping and contained genes
have the same number with suffixes a-z; and if there are more than 26 overlapping genes,
then it starts over with numbers following
the a-z; e.g. Gene #100a, #100z,  #100a1, #100b1, ...
<li>Load sequence: the project will now be removed if the load fails.
</ul>

<li>2D display
<ul>
<li>The placement of overlapping is greatly improved; the improvements depend on reloading the annotation for v518
(though it works as before without reloading).
<li>When the mouse is in the gene space, it will only provide <ttl>Information</ttl> if the mouse is over a gene;
this makes it easier to run the mouse down the gene space to find a given gene number.
<li>The length of the exons is included in the gene tag, e.g. Gene #100a (Exons 3 2123bp)
<li>Right click in the yellow annotation box was not working for 2D from Dotplot or SyMAP Query,
so it has disabled since the right click on gene annotation works fine from everywhere.
</ul>
<li>SyMAP Query
<ul>
<li>The <ttl>Gene#</ttl> column now shows the number and suffix; it is of format chr.genenum.suffix.
For older databases, this does not provide useful information as it expect v5.1.7 format or later.
<li>Single: The logic has been changed so that if you select a project, then you can select a chromosome.
</ul>

</ul>
<!--------------------------------------------------------------------->
<h3>Release: v5.1.7 (12-Oct-22)</h3>
For existing databases, see <a href="SystemGuide.html#update" class="ext" target="_blank">Update</a>.
<p>This release concentrates on making the 2D display provide more information with different colors for
genes and hits.

<ul>
<li>Processing
<ul>
<li>Gene# assignment: have been numbered with the same gene number; now they are given a suffix, e.g.
Gene #1011a, Gene #1011b. For existing databases, the annotation must be re-loaded to get this feature; then
the synteny needs to be recomputed (existing MUMmer files will be used).
<li><ttc>ConvertEnsembl</ttc> has a new option "-u" to use the sequence names from the FASTA file.
</ul>

<li>Chromosome Explorer:
<ul>
<li>Colors (Color wheel in upper right corner):
<ul>
<li>Hits are now colored brown (hits have the same strand on both sequences) or green (different strands).
<li>Genes can now be colored according to if they are on the positive or negative strand,
where the default is dark blue (positive) and purple (negative).
<li>The FPC options are no longer shown if there is no FPC in the 2D display.
</ul>

<li>Annotation
<ul>
<li>Overlapping genes are now staggered, i.e. the exons of one overlapping gene will be displayed further out than the other.
<li>When right-clicking on a gene or its yellow text box, the results now show the hits to the gene
along with the gene information and exons.
<li>When right-click on a hit-wire (between two chromosomes), a popup of its merged hits is shown (or single hit).
<li>Tiny bug fix: Yellow text box: if one clicked on the outer edge, it would popup the wrong gene.
</ul>
</ul>

<li>SyMAP Query:
<ul>
<li>The Colinear genes was filtering on &gt; instead of &gt;=. It also has new filter of "=";
hence, you can set the chromosomes and the Colinear genes to get all colinear runs of N on a specific
pair of chromosomes (e.g. 3.1.9).
<li>SyMAP does a better job of making filters inactive if they will not be used or have not been "checked".
</ul>
</ul>

<!--------------------------------------------------------------------->
<a id="v516"></a>
<h3>Release: v5.1.6 (20-Sept-22)</h3>
This release has improvements to the Block display, 2D display and SyMAP Query.
<ul>
<li>Blocks
<ul>
<li>For the single chromosome view: Inverted blocks have dotted outline.
<li>The windows are better sized.
<li>Fixed rare bug: For the initial multi-chromosome view: If there are many chromosomes, they were being cutoff; now the scroll bar allows them all to be shown.
</ul>
<li>2D Display
<ul>
<li>The full gene description can be viewed by right clicking on a gene annotation in the track;
see <a href="UserGuide.html#view" target="_blank" class="ext">UserGuide</a>.
<li>Hover over a hit may include "Inv" (inverted block) and will include "gNrM" (N exon hits, M colinear run size);
see <a href="UserGuide.html#view" target="_blank" class="ext">UserGuide</a>;
</ul>
<li>SyMAP Query
<ul>
<li>Add general columns <ttl>Hit %Id</ttl>, <ttl>Hit %Sim</ttl>, <ttl>Hit #Merge</ttl>
<li>Add species column <ttl>Len</ttl> which is the length of the hit (or merged hit).
<li>Bug Fix: <ttl>Show Synteny</ttl>: the hover over the hit only showed the correct hit <ttl>Information</ttl>
for those below the selected hit - that is fixed.
<li>Gene# of 0 will sort to bottom of table with a '-'.
<li>MUSCLE alignment: (1) Fixed rare bug. (2) The alignment files will be left in the <ttl>muscle</ttl>
directory, but over-written with each new alignment.
</ul>
<li>Alignment&Synteny: (2 tiny things)
<ul>
<li>If promer or nucmer are over-rided in the <ttl>Parameters</ttl>, it will say on the Summary page (unless the
alignment is re-used).
<li>Fixed rare bug: if the <ttl>Parameters</ttl> are changed before <i>any</i> alignment has been done, there is an error
because the <ttl>data/seq_results</ttl> directory has not been created.
</ul>
</ul>

<!--------------------------------------------------------------------->
<h3>Release: v5.1.5 (13-Sept-22)</h3>
<ul>
<li>Online Docs - the biggest change of this release is an updating of the documentation.
<li>Alignment & Synteny
<ul>
<li>Clear Pair provides option to "only delete synteny from database" or "delete synteny and remove files".
<li>The MUMmer %Sim column is now loaded in addition to the %Id column.

<li>The Hit# will be assigned sorted on the query start coordinate.
</ul>
<li>Chromosome Explorer
<ul>
<li>Hover over Gene: the gene information will be shown when hovering anywhere within the gene's rectangle.
(It used to only show Gene information if the mouse was hovered directly over the center.)
<li>The Hit information now has %Id and %Sim, plus the number of merged hits.
</ul>
<li>Blocks - the chromosome blocks view has a smaller window.
<li>demo_seq2 - Removed duplicate genes.
<li>Developers
<ul>
<li>If all hits are removed (e.g. Clear Pair), the index will be reset so that the next loaded hit will have a HitIdx (Hit #)
that starts at 1.
<li>The alignment files may have header lines. Any line that does not start with a digit will be ignored.
<li>Removed some dead tables from Schema and changed a few datatypes.
</ul>
</ul>
<!--------------------------------------------------------------------->
<h3>Release: v5.1.4 (4-Sept-22)</h3>

<ul>
<li>SyMAP Query
<ul>
<li>Filters:
<ol>
<li>Added a <u>General</u> section for the annotation and location, since they can work with all other queries.
Made the <u>Single</u> project (Orphan genes and all genes) a separate section.
<li>Add a filter to <u>Show genes w/o hits</u> for a project, which shows all genes.
<li>Changed the <u>Orphan genes</u> filter to show orphans relative to the chosen projects (it was across all synteny
project, even projects selected for the SyMAP query).
<li>The Hit# query did not work in some situations.
</ol>
<li>Improved the query summary. Also, the location had been left off the summary for most filters.
<li>Replaced the result table <u>#Gene</u> column: it was computed per query, whereas the replacement is stored in the database.
The replacement is the order number of the gene along the chromosomes, which stays constant across all queries
and is shown on the Chromosome Explorer Annotation and the hover of the gene.

</ul>
<li>Chromosome Explorer
<ul>
<li>Sequence Filter: If a wrong coordinate was entered, the program exited.
<li>Alignment Bug Fix: only the first part of the alignment was working (fairly recent bug).
</ul>
</ul>
<!--------------------------------------------------------------------->
<h3>Release: v5.1.3 (30-Aug-22)</h3>
There is no major new development, but small interface improvements and SyMAP Query has a small bug fix.
<ul>
<li>Main window
<ul>
<li>Projects listed on left are now in alphabetical order per category. If the project is loaded
into the database, the load date is shown to its right. If it has computed synteny, the number of syntenies is in brackets.
<li>The symap window is centered, along with the Chromosome Explorer and SyMAP Query window.
<li>viewSyMAP: only shows the projects that have synteny computed.
</ul>
<li>Load Annotation
<ul>
<li>grp_prefix: This can now be removed from existing projects, i.e. if the 'Chr' was left on,
so the 2D display is clustered with Chr01, Chr02, etc, then the 'Chr' can be removed.
<li>annot_kw_mincount: the Help made no sense for this. It has been upgraded to do the following:
the SyMAP Query shows the GFF attribute columns that have &gt;=annot_kw_mincount occurrences.
It can be changed at any time through the symap parameters window.
<li>When the annotation is loaded and the attribute (column 9) is parsed for keywords, the only keywords
considered are those followed by a '=' (it used to allow spaces between keyword and value).
</ul>

<li>2D chromosome view
<ul>
<li>The popup annotation is now a smaller constant sized scrollable window.
<li>The annotation hover and popup coordinates are more readable.
</ul>

<li>SyMAP Query
<ul>
<li>The Block query is now on just the block number along with the selection of chromosomes.
<li>Small changes:
<ul>
<li>The Results Panel includes the number of rows in the query.
<li>The query will not run if there is an error in the input, and there is better error-checking.
<li>The window is smaller and the columns panel in the table is fixed width.
<li>When the full block name is shown, any leading zeros are removed, e.g. 01.03.2&rarr;1.3.2.
</ul>
<li>Bug fix: The Orphan query with one or more chromosome selected gave wrong results (it was obvious); moreover,
if a start or stop was entered, it gave an error. This is fully functional now.
</ul>

<li>Convert
<ul>
<li>Both scripts work if there is not a gff file (they were requiring it).
<li>For both scripts, the bases that are not ACGT are summarized.
<li>ConvertNCBI - anything that does not have "chromosome" in its ">" line is considered a scaffold.
(ConvertEnsembl already does this).
</ul>
</ul>

<!--------------------------------------------------------------------->
<h3>Release: v5.1.2 (19Aug22)</h3>
This release associates Exons with their genes; hence, your database needs to be
<a href="SystemGuide.html#update" class="ext" target="_blank">updated</a>.

<ul>
<li>Annotation
<ul>
<li>Input format:
<ol type='i'>
<li>The <tt>ConvertNCBI</tt> and <tt>ConvertEnsembl</tt> scripts created separate <tt>gene.gff</tt> and <tt>exon.gff</tt> files; now they
create one file. The <tt>ConvertEnsembl</tt> was not producing a <tt>gap.gff</tt> file, and now it is. Both have been
updated to be aware of more input variations and provide a better summary. See <a href=convert class="ext" target="_blank">Convert</a> for details.
<li>The Demo_seq has separate <tt>gene.gff</tt> and <tt>exon.gff</tt> files; now it has one file.
<li>SyMAP can still load separate gene and exon files, but its association of exon to gene is not as accurate.
</ol>

<li>Schema change
<ol type='i'>
<li>An existing database will be updated the first time it is viewed.
<li><i>This adds a MySQL field to associate Exons with their Gene</i>; for an existing database,
the annotation would need to be reloaded for this to happen and be displayed (see 2d Gene Popup below).
</ol>
<li>Load Annotation
<ol type='i'>
<li>The code was treating 'CDS' as 'exon', now only exons are loaded.
<li>The attributes were not being loaded for Exon; now, the "parent" attribute is loaded for the
exon and shown when it is moused over.
<li>Quotes are removed from attributes.
<li>The exons are now being mapped to the parent gene to be used in the 2D display.
</ol>
</ul>
<li>Chromosome Explorer
<ul>
<li>Tiny bug fix: the "Circ" button get stuck in un-active mode.
<li>2D display
<ul>
<li>Tiny bug fix: For a single contiguous hit for a flipped chromosome, the hit length graphic was
not aligned right.
<li>When the Gene or Exon is moused over, it will include the number of exons or exon#, respectively.
<li>When the hit line is moused over, it will include the project name, start and end coordinates.
<li>2d Gene Popup: Includes the exons for the gene.
</ul>
</ul>
<li>SyMAP Query
<ul>
<li>Add warnings for bad block or collinear input.
</ul>
<li>Developers
<ul>
<li>The code is now Java v8 compliant (no depreciated statements).
</ul>

</ul>

<h3>Release: v5.1.1 (31July22)</h3>

These changes are to 'tidy up' things; nothing major.
<ul>

<li>Scripts:
<ul>
<li>The <tt>symap</tt> and <tt>viewSymap</tt> scripts have been changed from Perl to shell.
<li>The <tt>viewSymap3D</tt> script has been removed, but the <a href="./3D.html" class="ext" target="_blank">3D</a> documentation explains
how you could recreate the 3D view.
</ul>
<li>Data directory
<ul>
<li>The <tt>data/fpc</tt> has been tarred and zipped (<tt>fpc.tar.gz</tt>) since it is rarely used anymore.
If the <tt>data/fpc</tt> directory does not exist, then the BLAT args are not shown on
the "Param" window. The FPC-Seq alignments still work.
<li>If the <tt>data</tt> directory does not exist, it is created: (1) on "Add Project",
(2) when redoing an alignment in the database. (The directory will exist if using <tt>symap_5.tar.gz</tt>,
but it will not exist if starting from the code base).
</ul>
<li>Slight display change
<ul>
<li>The Summary shows the following: (1) for a project, if it used mask_genes or
a different min_size then 100000, (2) for an alignment, if the "mummer_path" is set,
or "concat" is unset, or the promer/nucmer parameters are changed. Only creates/updates on
running the synteny.
</ul>
<li>Small changes
<ul>
<li><tt>symap -v</tt> will display important MySQL variables and make suggestions.
<li>It does not try to load hidden files.
<li>It does not allow "cancel" on remove operations.
<li>It displays length of log files (they are appended to, so can become large).
<li>The project parameters <tt>annot_keywords</tt> and <tt>annot_kwy_mincount</tt> are now being saved to the database
so their values are shown correctly in the project parameters window on re-display.
<li>A proper message is given when <tt>viewSymap</tt> tries to read a databases (<tt>symap.config</tt>) that does not exist.
<li>A stacktrace will not be printed to the terminal, only to the error.log (the e.getMessage() sometimes includes a partial
stack).
<li>Developers: removed more dead code.
</ul>
</ul>
Note: this was first released on 27July22, then again on 31July22 with a slight change.


<h3>Release: v5.1.0 (12June22)</h3>
The documentation has
been moved from www.agcol.arizona.edu/software/symap to csoderlund.github.io/SyMAP.
Any references to the online documentation within the code has been updated.
For pre-v5.1 releases, the www.agcol.arizona.edu/software/symap/docs still exists.

<p>The html Help files can now have URL links in them.

<h3>Release: v5.0.8 (1June22)</h3>
<ol>
<li>A new option "Concat" has been added; when this is unchecked, it will not create a
concatenation of the first genome to run against all chromosomes of the second, instead
it perform a chromosome-by-chromosome search.
<li><u>MUMmer4</u> can be used by compiling it on your machine from the /ext directory
and adding its path to the "symap.config" file. See <a href=MUMmer.html>Running MUMmer</a>.
<li>When computing alignment&amp;synteny, all stacktrace errors are now written into error.log.
<li>The database name is prefixed to the "logs/LOAD.log" file name.
<li>A new document has been created for when there is problems running
<a href=MUMmer.html>MUMmer</a> from SyMAP.
<li>ConvertNCBI script was writing the scaffold sequences to end of the last chromosome (this
did not make the annotation or synteny wrong - just extra sequence).
</ol>

<h3>Release: v5.0.7 (4Feb22)</h3>
<ol>
<li><u>MUMmer</u> has been updated since the last SyMAP release; the code in the /ext/mummer4 directory
has been updated for the latest MUMmer. The documentation has been updated accordingly
(see <a href=MUMmer.html>Running MUMmer</a>).
<li><u>ConvertNCBI</u>: (1) Include ncbi_dataset format. (2) Convert hex to character.
<li><u>Parameters</u>: (1) The Load would not work if any numeric parameter was set to blank.
(2) For Alignment&amp;Synteny "Parameters", the Global button was removed as it did not work
except under weird conditions.
<li><u>Help</u>: the viewer Help pages link to the pages at www.agcol.arizona.edu/software/symap. The
approach for Mac was becoming obsolete, so it was changed. For Linux, more supported browsers were added.
<li>Improve documentation.
<li>Developers: Removed all Applet code (was disabled previously, but code was in everything).
</ol>

<!----------------------------------------------------------->
<h3>Release: v5.0.6 (2June20)</h3>
<ol>
<li>Some changes were necessary for SyMAP to run with MySQL v8.
	<ul><li>The schema needed a change, which will occur on first reading of database.
	</ul>
<li>SyMAP works on MacOS Catalina 10.15
	<ul><li>It checks each external program to see if it is executable, as this is not automatic with 10.15
	(see <a href="TroubleShoot.html#ext">Trouble Shoot External Programs</a>).
	</ul>
<li>SyMAP can be compiled with Java v14 (some changed were necessary for this).
<li>The Applet has been removed and will no longer be supported. (It is in earlier SyMAP versions, and can
be obtained at <a href="https://github.com/csoderlund/SyMAP">github</a>).
<li>The built-in MySQL database on Linux is no longer supported.
<li>The Alignment&amp;Synteny has a couple checks for 'ordered_against' before processing:
	(1) Is the order_against the same as the other selected project.
	(2) Does the ordered directory already exist.
</ol>

<h3>Release: v5.0.5 (2May20)</h3>

<ol>
<li>Ordering algorithm: The main change is a rewrite of the ordering algorithm, which orders draft contigs
against a completed sequence closely related genome. The previous algorithm sometimes
would put many of the placed contigs incorrectly into 'chrUNK', though the
ordering in the <tt>anchored.csv</tt> file was correct.

<li>SyMAP Manager:
<ul>
<li>Show #genes and #gaps in summary
<li>Add Project: On Linux, the "Add" button would stay disabled, this has been fixed.
</ul>

<li>Chromosome Explorer: Mouse over a hit and the block number will be shown along with the other information.

<li>SyMAP Query:
The "Align Sequence" display now show locations in text box under alignment.
The v5.0.4 introduced a bug where the row selections display were not right, which
is fixed.

<li>Small changes:
<ul>
<li>Running 'symap -v' will check the MySQL variables needed to run at a reasonable speed,
and will output to terminal any variables that need changing.
<li>Block View: When viewing the blocks using VNC, the chromosome that the
	blocks were aligned to was not visible, which has been fixed.
<li>All Pairs button: Is only enabled if there is anything that will be executed when selected.
<li>Load Project: If a file of 100's of draft sequences are loaded, this was flagged as an error (but
still loaded); it no longer called an error, though a message is written to terminal.
</ul>
</ol>

<h3>Release: v5.0.4 (18Apr20)</h3>
<ol>
<li>SyMAP Query has had major changes:
<ol>
<li>It is faster, e.g. finding the hits in synteny blocks for 5 plant genomes did take 9m:32s
and now takes 0m:09s.
<li>The query page has changed to allow more versatile queries.
<li>The PgeneF is only computed if the "Compute PgeneF" option is checked; otherwise, there
is no changes to this function or options.
<li>Some of many minor changes:
<ul>
<li>Selecting a chromosome for a species (or multiples) now only shows hits that involve the
selected chromosomes.
<li>The #RGN column has been removed and a new "Has Gene" column has been added. Note, all species
with a gene overlapping the hit should have 'annotation', but this makes it easier to see in a glance.
Plus, if some genes do not have an ID or some other keyword, it can be confusing looking at the
annotation.
<li>The "Save for Reload" and "Reload Results" have been removed, as it is so much faster, this
does not seem necessary anymore.
<li>Both the Overview and Table Summary have small changes.
</ul>
</ol>
<li>scripts/ConvertNCBI has been updated to also create a gap file.
<li>scripts/symap.sql has been removed and the schema has been incorporated into the Java code.
<li>SyMAP Summary: Displays major parameters for projects that are not loaded.
<li>Chromosome Explorer 2D View: The mouse over provides location information in the "Instruction" box,
plus:
<ul>
<li>The following functions can be applied to the selected region of a sequence track:
"Zoom All Tracks", "Zoom Selected Tracks", "Align (Max 30000)", "Show Sequence", where the
first three are not new, but the labels changed for clarity.
<li>The last one, "Show Sequence", has
been added as there was no way to see the underlying sequence; it is simply a popup of the
selected sequence.
<li>The "Align" view has been slightly changed to add comma's in large numbers and the length of the sequence.
</ul>
<li>The symapApplet.jar was not working in recent releases, which has been fixed.
</ol>

<h3>Release: v5.0.3 (11mar20)</h3>
<ol>
<li><i>Chromosome Explorer 2D view</i>. The "Annotation Description" can be right-click to view
a pop-up of the description; the benefit is that sometimes part of the description is
buried, and the pop-up can be copied.
<li><i>SyMap Queries</i>. (1) The filters set are shown at the top of the table.  (2) The "Only
Orphan Genes" option can now only be used with the "Annotation Description", "Location", and
"Include" options.
<li>If access to MySQL does not work, a message is printed with the Java error message.
</ol>

<h3>Release: v5.0.2 (24feb20)</h3>
<ol>
<li>scripts/ConvertNCBI: (1) Produce a hard-masked genome sequence (input is soft-masked) with
-m flag. (2) Includes the mRNA ID in the product attribute. (3) The product attribute has
been improved.
<li>SyMAP Bug fix for using chromosome and scaffolds in the same project.
<br>The following did not work:
Say the genomic  sequence input was a mix of chromosomes (prefix 'Chr')
and scaffolds (prefix 's'), and the Project parameters were prefix 'Chr' and minsize '10000',
where the minsize filtered out annotated scaffolds, SyMAP would quit if there was
annotation on filtered-out scaffolds; this has been fixed. Also, the documentation
was not correct for mixing sequence types (e.g. prefix 'Chr' and 's'); the Project parameter
'grp_prefix' should be set to blank.

</ol>

<h3>Release: v5.0.1 (16feb20)</h3>

<ul>
<li>Load Annotation
	<ul>
	<li>In the Params interface for a project, the user can set what keywords from the
column 9 "Attributes" field of the gff file to load; this was not working correctly
and has now been fixed.
	<li>The type (column 3) of mRNA was being loaded as a gene, which causes duplicate records;
now mRNA's are ignored.
</ul>
<li>Alignment &amp; Synteny
	<ul>
	<li>Though it was not possible to make MUMmer v4 part of SyMAP, it is possible
to install MUMmer v4 in the /ext/mummer directory and it will work. Instructions
are provided in <a href="SystemGuide.html">System Guide</a>.
	<li>The nucmer "--maxmatch" parameter was automatically used for self-synteny when
a sequence file was being compared with itself; this would result in an extremely
long execution time for large files (e.g. chromosomes). Consequently, this
parameter has been removed, but can now be set by the user in the "Parameter" interface.
	<li><i>Pair Parameters</i>: (1) The user can set parameters in "Self Args", which are only
used when a chromosome sequence is being compared to itself. (2) This interface
was very confusing as to when parameters were saved and what would be used;
this should be clearer now, and is explained in the "Help" for the Parameter interface.
	<li>If only part of the alignments finished and it ended pre-maturely, when run again,
it will not redo any finished alignments.
	</ul>
<li>Internally, goobs of commented out code has been removed.
</ul>

<p><a href="#top">Go to top</a>

<h3>Release v5.0 (22jan20)</h3>
New projects cannot be added to v4.2 databases, but they can be queried (e.g. dotplot).
The significant change
is with the file structure for the data, where the old file structure will not
work with this new release as shown further under <a href="#ds">Directory structure</a>.


<ul>
<li>General
	<ul>
	<li>Error messages (and stack traces) are written to error.log
	<li>There was one symap script with various flags, now there is four symap scripts
	with less flags, as follows:
	<table>
	<tr><td><i>Script</i><td><i>Build</i><td><i>View</i><td><i>Has 3D view</i>
	<tr><td><tt>./symap</tt> <td>Yes<td>Yes<td>No
	<tr><td><tt>./symap3D</tt> <td>Yes<td>Yes<td>Yes
	<tr><td><tt>./viewSymap</tt> <td>No<td>Yes<td>No
	<tr><td><tt>./viewSymap</tt> <td>No<td>Yes<td>Yes
	</table>
	<li>The symap -c command line parameter can be used to
	specify a configuration file other than 'symap.config'
	<li>The Summary list of all selected projects has been enhanced.
	</ul>

<li>Query database
	<ul>
	<li>The "Display for Selected Pair:" popup Summary is now available for self-synteny and FPC-Seq
	</ul>

<li>Build database
	<ul>
	<li>Log files: The log files are no longer in the results directories, instead they are in
	the top level <tt>/logs</tt> directory, as follows:
		<ol>
		<li>The output from the loads are concatenated to the file <tt>/logs/LOAD.log</tt>.
		<li>The output during alignment is in the <tt>/logs/&lt;proj1-to-proj2&gt;</tt> directory.
		<li>The output to the terminal from the alignment and synteny computation is
written in the <tt>/logs/&lt;proj1-to-proj2&gt;symap.log</tt> file.
		</ol>
	</ul>
<li>Load project
	<ul>
	<li>The input files can be gzipped (end with .gz)
	<li>The Parameters menu uses the previous location for the file/directory finder.
	<li>Add ConvertEnsembl.java to convert an Ensembl file to SyMAP annotation file.
	</ul>

<li>Alignment &amp; Synteny
<ul>
	<li>The trace output to terminal is much more informative.
	<li>The alignments for self-synteny have been optimized, which also
  changes the results (e.g. demo_seq now has some tiny blocks were before
  it had none).
	<li>The number of CPUs can be set in the interface, or either of the following:
		<ol>
		<li>Use the -p command line parameter, e.g. <tt>./symap -p 12</tt>.
		<li>Set the 'nCPUs' configuration parameter in the symap.config file, e.g. <tt>nCPUs=12</tt>.
		</ol>

	<li>Much of the code for building the syntenies has been re-written, but the results
	are basically the same with a few tiny differences.
</ul>

<a id="ds"></a>
<li>Directory structure: The following are the directory changes, where they
are under <tt>/data</tt>.
	<table>
	<tr><td><i>New</i><td><i>Old</i><td><i>Description</i>
	<tr><td>/seq
		<td>/pseudo
		<td>Directory for sequence projects
	<tr><td>/seq/&lt;project&gt;/sequence
		<td>/pseudo/&lt;project&gt;/sequence/pseudo
		<td>Default directory for sequences
	<tr><td>/fpc_results
		<td>/fpc_pseudo
		<td>Directory of FPC to sequence project results
	<tr><td>/seq_results
		<td>/seq_pseudo
		<td>Directory of two sequence project results
	<tr><td>/seq_results/&lt;proj1-to-proj2&gt;/align
		<td>/pseudo_pseudo/&lt;proj1-to-proj2&gt;/anchor
		<td>Directory of alignments
	</table>
</ul>


<p>

<p><a href="#top">Go to top</a>

<p>
<hr>
<p>
<h3>Release: Jan2018 (build 120)</h3>
<p>Release date: 1/16/18 - updated ConvertNCBI.java script.

<p>Release date: 1/10/18
<ul>
<li>A few more improvements on the applet code.
<li>Renamed some functions to make them clearer and updated the help.
<li>Improved the Help on the parameters page.
<li>Fixed a rare memory bug (occured on a self-synteny; requires the database to be re-created).
<li>The Image button does not work for Java v9, so a popup tells the user.
<li>Added a Java script called ConvertNCBI.java, as documented in <a href=ncbi target="_blank">NCBI to SyMAP</a>.
</ul>
<b>Release date: 12/28/17 (build 119)</b>
<ul>
<li>The applet code has been fixed to work better.
<li>The FPC to sequence block view had a bug that was fixed (though 'reverse' does not work).
</ul>
<b>Release date: 12/24/17 (build 118)</b>
<ul>
<li>There was a limit of 75 sequenced regions per genome, which the block view did not work after that.
<br>The limit has been changed to 150 and a popup message occurs if that limit is passed.
<br>A script (scripts/lenFasta.pl) is supplied to provided the sorted lengths, where you should set the
min_size parameter to the 150th length.
</ul>
<b>Release date: 11/7/17 (build 117)</b>
<ul>
<li>Fixed problem that prevented symap from running with new JDK v9.
</ul>
<b>Release date:  1/7/16 (build 116)</b>
<ul>
<li> Fix display sizing errors in circle view display
<li>Fix database error for long project names (max project name remains 40)
</ul>
<p><a href="#top">Go to top</a>

<h3>Version 4.2</h3>
Release date:  4/23/2014
<p><b>New Features:</b>
<ul>
<li> New, unified web access applet interface. Now you can present a unified SyMAP web display of some or all
of the projects in your database using just one small HTML page (no CGI). Applet interfaces are also
available for each of the sub-functions (dotplot, circle view, etc.) enabling customized web displays.
</ul>
<p>
<b>Modified Features:</b>
<ul>
<li> The Explorer has been updated to be more usable for draft projects, i.e. when there are many
sequences that have not been ordered and anchored to pseudomolecules. Specifically, it will now show
up to 450 sequences, and to decrease clutter it shows only sequences that actually have a synteny block.
<li> The prior web system files have been removed, including the cgi and supporting perl modules.
<li> The circle view now draws all text labels horizontally, due to rendering problems
with rotated text on several platforms (most severely on Mac). The rotated text is still available through a
checkbox option.
It is no longer necessary to run the perl install script to install web functionality.
</ul>
<h3>Version 4.1</h3>
Release date:  2/20/2014<br>
<p><b>New Features:</b>
<ul>
<li> Publication-quality image saving. A save button has been added to every graphical window allowing saving to
a wide variety of formats, including scalable vector formats (svg, eps, pdf) which may be resized clearly to any size. (Many thanks to the developers of the FreeHEP Vector Graphics library, and especially Mark Donszelmann, for making this possible!)
<li> Download of block co-ordinates. Choose the species of interest, open the Explorer, and the download button is at the lower left.
</ul>
<p>
<b>Modified Features:</b>
<p>
<ul>
<li> Applet is signed with a DigiCert certificate, removing the security blocks which have been increasing
in recent Java versions.
<li> MUMmer binaries for Mac OSX have been replaced by 64-bit versions. The 32-bit versions which had been
supplied could not access more than 2G RAM, preventing alignment of many chromosomes.

</ul>
<p>
<h3>Version 4.0</h3>
Release date:  7/15/2012<br>
<p><b>New Features:</b>
<ul>
<li>Project Manager Enhancements
<ul>
<li> Create new projects through the Manager
<li> No need to create directories or copy files
<li> Alignment &amp; Synteny parameters also available
</ul>
<li>Query Page Enhancements
<ul>
<li> Compute putative gene families across species
<li> Many additional filters
<li> Create and view Muscle multiple-alignment for selected results
</ul>
<li>Additional Features
<ul>
<li>Script 3track_figure.pl makes 3-genome figure for publication (see online
<a href="Tour.html#svg">Tour</a>;
script added to 4.0 package post-release)
</ul>
</ul>
<p>
<b>Modified Features:</b>
<ul>
<li> Write contig anchoring information to a file (<tt>order_against</tt> mode)
<li> No longer creates the "_ordered" project output from <tt>order_against</tt>
<li><tt>grp_prefix</tt> parameter now optional (but recommended). <tt>grp_sort</tt> eliminated.
<li> Self-alignments improved by using -maxmatch option of MUMmer for individual chromsome self-alignments
<li> Top-2 filtering modified to accept also hits within 80% of the 2nd best score, in order to
retain gene-family hits and allow identification of families through the query interface.
</ul>
<p>
<h3>Version 3.5</h3>
Release date:  1/24/12<br>
<p><b>New Features:</b>
<ul>
<li>Dynamic circular display
<ul>
<li> Circular-style view allowing addition/removal of species and chromosomes
<li> Replaces 3D view in the web applet (removing problems of MacOS support for 3D)
<li> 3D still available in standalone mode
</ul>
<li> Annotation query interface (Java-based, upgrades the web CGI version)
<li> Project parameters editing window
<li> Unsigned web applet (no approval popups, if database and webserver have same host)
<li> For draft sequence using <tt>order_against</tt>, create ordered and anchored versions of the project
</ul>
<p><b>Modified Features:</b>
<ul>
<li> Added contig-flipping to draft sequence <tt>order_against</tt> function
<li> Added draft sequence ordering demo
<li> Improved command-line launch script
</ul>
<p>
<h3>Version 3.4</h3>
Release date:  12/14/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Query files catenated before running MUMmer, for substantial increase in speed.
<li>Batch alignment buttons added to Project Manager.
<li>Improvements to handling of unordered scaffolds, including:
<ul>
<li>Scaffolds can all be in one fasta file.
<li>Added min_size parameter, to specify minimum size scaffold to load.
<li>Added order_against parameter, to specify reference sequence for ordering scaffolds
	(the order only affects the dotplot display)
</ul>
<li>Minor changes to CGI page displays to better accomodate unordered.
<li>Fix compilation error in 64-bit MUMmer binary, which prevented use of very long pseudomolecules.
</ul>


<h3>Version 3.3</h3>

Release date:  10/4/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Added "no_overlapping_blocks" parameter
<li>Added printout of block and synteny anchor tables to "/results" directory
<li>Remove "Contig0" (singleton clones) from FPC displays
<li>Adjustments to unannotated sequence clustering to obtain larger clusters,
    not dependent on the order of scanning of MUMmer output files.
<li>When clicking a block in the dot plot, 2D view will open with hit filter setting "show only synteny hits"
<li>Some bug-fixes to search.cgi
</ul>

<p>

<h3>Version 3.2</h3>
Release date:  6/9/10<br>
<p><b>Modified Features:</b>
<ul>
<li>Inversions are shown on the 3D view in green, where un-inverted are shown in red.
<li>Anchor chains are not merged.
<li>The dot plot background is white and the synteny blocks blue.
<li>SyMAP automatically only uses one processor unless specified otherwise with the -p option.
<li>The whole genome dot plot is available from the project manager.
<li>When running an alignment, a summary is listed to the terminal and to a log file.
<li>A lot of error checking has been added. However, every platform can have variations, which
we may not have accounted for. <font color=green>So PLEASE, let us know if you have any problems and we will
work with you to it.</font>
</ul>

<h3>Version 3.1</h3>
Release date:  2/12/10<br>
This release includes major enhancements to the Dot Plot view.<br>
<br>
<b>New Features:</b>
<ul>
<li>Multi-genome Dot Plot: multiple species can be displayed on the y-axis against a common reference species.</li>
<li>Dot Plot view integrated into 3D viewer, now the species/chromosomes displayed in the Dot Plot can be changed on-the-fly.</li>
<li>Dot Plot reference species on the x-axis can be changed dynamically via a new drop-down menu.</li>
<li>Main web page redesigned.</li>
</ul>

<p><br>


<h3>Version 3.0</h3>
Release date:  1/7/10<br>
This major release includes many new features, in particular the 3D viewer and Project Manager.<br>
<br>
<b>New Features in User Interface:</b>
<ul>
<li>Multi-genome 3D viewer.</li>
<li>Circular CGI view.</li>
<li>Gene search CGI.</li>
<li>Support for sequence-to-sequence alignment.<li>
<li>Option to flip sequence.</li>
<li>Zoom into sequence or hit regions using the mouse scroll wheel.</li>
<li>Three-track CGI blocks view.</li>
<li>Improved annotation display and addition of URL-embedding capability.</li>
</ul>
<b>New Features in Back-End Processing:</b>
<ul>
<li>Completely rewritten in Java with major performance improvement.</li>
<li>Addition of Project Manager GUI to simplify and automate back-end processing, sequence alignment, and synteny analysis.</li>
<li>Support for sequence-to-sequence alignment.<li>
<li>Hits are now clustered based on gene annotation and drawn as "ribbons" to indicate locations of sub-blocks.</li>
</ul>


<h3>Version 2.0</h3>
Release date:  5/18/07<br>
This release includes many new features for the alignment view.  The dot plot view was
not changed.<br>

<br>
<b>New Features Block view (CGI):</b><br>
<ul>
<li> <i>View reverse</i> option: reverse which of the two genomes is the reference genome.
<li> Chromosome view: the blocks are color coded to indicate which chromosome they are from. Also, the number of anchors is displayed over each block.
</ul>

<b>New Features on Java interface:</b><br>
<ul>
<li> Filters:
<ul>
<li> For a given track (Block, Hit, Sequence) hover the mouse over the track and
right click. A menu of the most common filters is shown;
this is much easier than opening up the Filter window.
<li> From this window, you can go directly to the help for the given track.
There is also a 'Navigation Help', as there are many ways to resize, scroll, etc.
</ul>
<li> Sequence track:
<ul>
<li> <i>Show hit score bar</i> is a new filter that
shows the hits down the length of the sequence as a histogram showing how strong each hit is.

<li> <i>Show Hit Score Value</i> is a new filter that shows the %identity.
Note that the sequence graphic is twice as wide to show the score bar and score value.
<li> The sequence track can be scrolled by moving the mouse wheel when the
mouse is position over the sequence track.
This retains the same zoom and filters, but moves up or down the lenght of the sequence (i.e. pseudomolecule).
<li> The gene annotation now shows the intron/exon structure.
</ul>
<li> Hit track:
<ul>
<li> <i>Show Only Gene Hits (contained)</i> is a new filter that turns off the
display of all hits unless they are contained in an annotated gene.
<li> <i>Show Only Gene Hits (overlap)</i> is a new filter that turns off the
display of all hits unless they overlap an annotated gene (but is not contained in a gene).

<li> <i>Show Only Non-Gene</i> is a new filter that turns off the
display of all hits unless they are NOT overlapping or contained in a gene.
<li> Hit lines are highlighted in red when mouse positioned over a hit.
</ul>
<li> Block track:
<ul>
<li> When hovering over a contig, the contig information at the bottom of the
display includes all the chromosomes that the contig hits.
</ul>
<li>Bug Fixes:
<ul>
<li> Genes drawn incorrectly when sequence range within gene boundary.<br>
<li> Fixed bug with pseudomolecule filter.<br>
<li> Certain queries were failing when used with MySQL 5.0.26 due to comma precedence change.<br>
</ul>
</ul>
<b>New Features in Back-End Processing:</b><br>
<ul>
<li> Support for mixed letter/number chromosome names and arbitrary ordering (see example in demo-fpc params file)
<li> Pseudomolecule sequences loaded in 1Mb chunks, removing database problems with large text fields
<li> Filtering optimized in anchors.pl to reduce memory use in loading large blat outputs
</ul>


<h3>Version 1.0</h3>
Release date:  9/1/06<br>
First full public release.

<p><a href="#top">Go to top</a>

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
