<!DOCTYPE html>
<html>
<head>
<title>SyMAP System Guide</title>
<style>
	a:link {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:visited {color:#6666CC;text-decoration: none; background-color: transparent;}
	a:active {color:#6666CC;text-decoration: none; background-color: transparent;}
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: Black;}
	ttp  {font-family: 'Lucida Console', monospace; font-size: 16px; font-weight: 500; font-variant: small-caps;}
	a.current {font-family:verdana; font-size:12px; color:white; text-decoration: none;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.tx td {padding: 5px;}
	.ty    {border: 1px  solid black; border-spacing: 1px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 4px; }
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP System Guide
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="current">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="white">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

The following discusses building a SyMAP v5 database.
<ul>
<li>For release v5.0.8, a document on trouble shooting <a href="MUMmer.html#mum"  class="ext" target="_blank">MUMmer</a> alignments failures has been added.
<li>For release v5.1.1, 3D has been retired, but may still be usable (see <a href="3D.html"  class="ext" target="_blank">3D</a>).
<li>For release v5.2.0,  FPC (fingerprinted contigs) is totally retired from SyMAP.
To use FPC, see <a href="#fpc">FPC</a>.
</ul>
<h4>Contents</h4>
<ul>
	<li><a href="#overview">Overview and Publications</a>
	<li>Getting Started
		<ul>
			<li><a href="#start">Steps for finding synteny</a>
			<li><a href="#machine">System Requirements</a>
			<li><a href="#installation">Installation</a>
		</ul>
	<li>Demo
		<ul>
		<li><a href="#demo">Running the demo</a>
		<li><a href="#demoseq">Synteny between two genome sequences</a>
		<li><a href="#draft">Draft alignment and ordering</a>
		<li><a href="#demoself">Self alignment</a>
		</ul>

	<li>New project
		<ul>
			<li><a href="#mysql">MySQL and Parameters</a>
			<li><a href="#time">Runtime and Memory</a>
			<li><a href="#create">Creating a New Project</a>
			<ul>
				<li><a href="#params2">SyMAP Parameters</a>
				<li><a href="#seqprep">Sequences files</a>
				<li><a href="#annot">Annotation files</a> - convert from NCBI or Ensembl
			</ul>
			<li><a href="#draft2">Draft alignment and ordering</a>
			<li><a href="#self">Self alignment</a>
		</ul>
	<li>General
	<ul>
	<li><a href="#dir">Directory structure</a>
	<li><a href="#ext">External programs, MUMmer and FPC</a>
	<li><a href="#alg">How SyMAP Works</a>
	</ul>
<li><a href="#update">How to update SyMAP with a new release</a>
<li><a href="#refs">References</a>
</ul>

<a id="overview"></a>
<h2>Overview and Publications</h2>
SyMAP is a system for computing, displaying, and analyzing syntenic alignments between
<font color=green>medium-to-high divergent eukaryotic genomes</font>. It does <b>not</b> work well
for very similar genomes or bacterial genomes.
<p>Its features include the following (for a pictorial introduction, see the
<a href="Tour.html" class="ext" target="_blank">Tour</a>):
<p>
<table>
	<tr>
		<td style="vertical-align: text-top;">
		<table class="tx">
		<tr><td>Find synteny between two sequenced genomes with optional annotation.
		<tr><td>Draft sequence ordering by synteny, i.e.
				<font color=green>align a draft genome to a fully sequence (not draft-to-draft</font>).
		<tr><td>For multiple selected synteny pairs, display using dot plot, circular, and side-by-side.
		<tr><td>Complete queries on annotation, collinear genes, cross-species gene families, etc.
		</table>
	<td>
		<center><small>Click an image to see the closeup.</small></center>
		<table class="tx">
			<tr><td><a href="img/symapRosid.png"><img src="img/symapRosid.png" alt="viewRosid" style="width:120px;border: 1px solid black;"></a>
				<td><a href="img/2d-3chr.png"><img src="img/2d-3chr.png" alt="2d-3chr"       style="width:110px;border: 1px solid black;"></a>
				<td><a href="img/querySetup.png"><img src="img/querySetup.png" alt="querySetup" style="width:100px; border: 1px solid black;"></a>
			<tr><td><a href="img/dotplot-3g.png"><img src="img/dotplot-3g.png" alt="dotplot-3-genome"     style="width:120px; border: 1px solid black;"></a>
				<td><a href="img/block.png"><img src="img/block.png" alt="block-2-genome" style="width:110px; border: 1px solid black;"></a>
				<td><a href="img/circle-2g.png"><img src="img/circle-2g.png" alt="circle-2-genome"   style="width:90px; border: 1px solid black;"></a>
		</table>
</table>
<p>

<a id="pubs"></a>
<h4>Publications</h4>

The back-end processing of SyMAP runs MUMmer<sup>1,2</sup> for the alignments (included in the tarball) and computes
the synteny block from the alignment results. The SyMAP synteny algorithm is described in the following
two publications. A sketch of the algorithms is also provided <a href="#alg">here</a>.

<pre>
        C. Soderlund, M. Bomhoff, and W. Nelson (2011)
        SyMAP: A turnkey synteny system with application to plant genomes.
        Nucleic Acids Research 39(10):e68.
</pre>
<p>
<pre>
        C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
        SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps
        Genome Research 16:1159-1168.
</pre>

SyMAP is freely distributed software, however
<font color=green><i>if you use SyMAP results in published research, you must cite one
or both of the above articles along with the external program MUMmer<sup>1,2</sup></i></font>.

<br>&nbsp;
<a id=start></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Getting Started</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<h3>Steps for finding synteny</h3>
Follow the steps below to get started with SyMAP.
Note that each SyMAP application window has a separate <ttb>Help</ttb>
button providing further details on the use of its functions.
<p>

<table class="tx">
	<tr><td style="vertical-align: text-top;">1.
		<td style="vertical-align: text-top;">Use a Linux or Mac machine.
		<td style="vertical-align: text-top;">It needs to have Java v1.8 or later, and sufficient processing power.
			See <a href="#machine">system requirements</a>.
	<tr>
		<td style="vertical-align: text-top;">2.
		<td style="vertical-align: text-top;">Set up MySQL.
		<td style="vertical-align: text-top;"> See <a href="#mysql">MySQL</a>.
	<tr>
		<td style="vertical-align: text-top;">3.
		<td style="vertical-align: text-top;">Download SyMAP.
		<td style="vertical-align: text-top;">Installation is a simple unzip. See <a href="#installation">installation</a>.
	<tr>
		<td style="vertical-align: text-top;">4.
		<td style="vertical-align: text-top;">Run the demo.
		<td style="vertical-align: text-top;">Highly recommended. See <a href="#demo">running the demo.</a>
	<tr>
		<td style="vertical-align: text-top;">5.
		<td style="vertical-align: text-top;">Prepare sequences and annotation.
		<td style="vertical-align: text-top;">Sequences can be in one or many files and can be masked or unmasked.
			See <a href="#seqprep">preparing the sequences</a>.
			Annotation format is gff3; see <a href="#annot">annotation files</a>.
	<tr>
		<td style="vertical-align: text-top;">6.
		<td style="vertical-align: text-top;">Load the files into SyMAP.
		<td style="vertical-align: text-top;">The SyMAP interface makes this easy; see
			<a href="#create">creating a new project</a>.
	<tr>
		<td style="vertical-align: text-top;">7.
		<td style="vertical-align: text-top;">Compute alignments and synteny.
		<td style="vertical-align: text-top;">This is also easy through the SyMAP interface. See
			<a href="#time">runtime and memory</a>.
	<tr>
		<td style="vertical-align: text-top;">8.
		<td style="vertical-align: text-top;">View results.
		<td style="vertical-align: text-top;">Detailed
			description of the user interface is in the <a href="UserGuide.html" class="ext" target="_blank">User Guide</a>.
</table>
<p>If you are working with FPC files, see the section on <a href="#fpc">FPC</a>.

<br>
<a id="machine"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>System Requirements</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

SyMAP has been tested on the following:
<table class="ty">
<tr><td><i>Machine</i><td><i>MySQL</i><td><i>Java</i><td><i>Core</i><td><i>Memory</i><td><i>Purchased</i>
<tr><td>Centos Linux amd64<td>MariaDB v10.4.12<td>1.8<td>2.3 GHz 24-Core<td>128 GB<td>2011
<tr><td>MacOS x86_64 (Catalina 10.15.4)<td>MySQL v8.0.17<td>17.0<td>3.2 GHz 6-Core<td>64 GB<td>2018
<tr><td>MacOS x86_64 (Maverick 10.9.5)*<td>MySQL v5.6.21<td>1.8<td>2.4 GHz 5-Core<td>16 GB<td>2011
</table>
*<small>This is an old laptop. I could not download from Github, so got v5.0.8 from AGCoL. Then I had to replace
<tt>/ext/mummer/mac</tt> with <tt>/ext/mummer/mac_pre506</tt>. This machine is fine for viewing a SyMAP database,
but too small to run MUMmer in an acceptable amount of time.</small>

<p>The jar file has been compiled with Java 1.8, which is upward compatible.
<p>
For performing large alignments (e.g. 1Gb genomes or more) <i>it is essential
to have multiple CPUs and at least 5Gb of RAM for each CPU that you intend to use</i>. Note that you can set the
number of CPUs for SyMAP to use.
See <a href="MUMmer.html#mem" class="ext" target="_blank">MUMmer</a>
for information on insufficient memory.
<p>
For viewing alignments, CPU and memory needs are typically negligible, unless
you are performing queries on more than 4-5 genomes at once.

<a id="installation"></a>
<h3>Installation</h3>

Installation consists of unzipping the download package, using
the command
<pre>     &gt; tar -xf symap_5.tar.gz</pre>
This can be done anywhere and it creates a directory called <tt>symap_5</tt>. You can
move this directory later if desired. The contents are:
<pre>
   LICENSE   README   data/   ext/   java/
   scripts/  symap    symap.config   viewSymap
</pre>
The <tt>data/</tt> directory contains a <tt>/seq</tt> sub-directory,
which contains the demo files, and is the default location for all input sequence files.

<p>The <tt>ext/</tt> directory contains the external programs MUMmer<sup>1,2</sup> (sequence alignment)
and MUSCLE<sup>7</sup> (SyMAP Queries).

<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!--------- DEMO ---------->
<a id="demo"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Running the Demo</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

If you have not used SyMAP before, it is essential to run the demos. After you have
installed MySQL, do the following:
<ol>
<li>Change into the <tt>symap_5</tt> directory.
<p>
<li>Edit <tt>symap.config</tt> and enter database
and host information (see <a href="#mysql">MySQL</a>), e.g. <tt>db_name=symapDemo</tt>.
<p>
<li>From the command line, type <tt>./symap</tt>.
</ol>
<p>The first time you run SyMAP, it will create the database with information written to the terminal, e.g.
<pre>
Creating database 'symapDemo' (jdbc:mysql://localhost/symapDemo?characterEncoding=utf8).
</pre>
<br>It will check your MySQL variables; if there are any "Suggested" changes,
see <a href="TroubleShoot.html#msqslow" class="ext" target="_blank">Trouble Shoot MySQL</a>.
<p>It will also
check that the provided external programs (e.g. MUMmer) are executable; if it shows any problems,
see <a href="MUMmer.html#ext" class="ext" target="_blank">MUMmer Executables</a>.
For MacOS, see <a href="TroubleShoot.html#macos" class="ext" target="_blank">MacOS and external programs</a>.
<p>
<a id="demoseq"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Synteny between two genome sequences</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr><td style="vertical-align: text-top;">
The Project Manager window opens showing the three demo projects provided with the SyMAP tarball.
Check <ttl>Demo-Seq</ttl> and <ttl>Demo-Seq2</ttl>.
<td><a href="img/demoLoad.png"><img src="img/demoLoad.png" alt="demoLoad" style="border: 1px solid black; width: 400px"></a>

<tr><td style="colspan: 2; height: 40px;">&nbsp;

<tr><td style="vertical-align: text-top;">

A link <ttb>Load All Projects</ttb> will be displayed in the top of the right panel; select it to
load the projects, which will take several minutes.
<p>If loading the "Demo-Seq" takes
<i>more than a few minutes</i>, you may need to adjust the MySQL parameters, see
<a href="TroubleShoot.html#msqslow" class="ext" target="_blank">TroubleShoot MySQL</a>.
<p>When done, the Manager will look as shown in the image on the right.
<p>In the <ttb>Available Syntenies</ttb> table, click the cell for the "Demo_Seq2" row and the "Demo_Seq" column.
Then click the <ttb>Selected Pair</ttb> button to start the alignment.

<td>
<a href="img/demoMgr.png"><img src="img/demoMgr.png" alt="demoMgr" style="border: 1px solid black; width: 400px"></a>

<tr><td style="colspan: 2; height: 40px;">&nbsp;

<tr>
<td style="vertical-align: text-top;">
The <ttl>Alignment&amp;Synteny</ttl> takes less than 5 minutes on the MacOS 10.5 but could take up to 30 minutes
on a slow machine.
<p>When done, the table will have a checkbox, signifying that the synteny is available for viewing.
Click the checked cell, which will enable the viewing buttons.

<td>
<a href="img/demoTable.png"><img src="img/demoTable.png" alt="demoTable" style="border: 1px solid black; width: 400px"></a>

<tr><td style="colspan: 2; height: 40px;">&nbsp;
<tr>
<td style="vertical-align: text-top;">
Click <ttl>Summary</ttl> to view the summary shown on the right; <i>there may be slight differences in the #Anchors
because of different numbers of CPUs, MacOS vs Linux, etc (but the #Blocks come out the same)</i>.
<p>To view the other interfaces, see <a href="Demo.html" class="ext" target="_blank">Demo Results</a>.
<td>
<a href="img/demoSum.png"><img src="img/demoSum.png" alt="demoSum" style="border: 1px solid black; width: 400px"></a>
</table>

<a id="draft"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Draft alignment and ordering</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
	<tr><td style="vertical-align: text-top;">
Load the Demo-Draft project, following
the same steps used to load the previous two projects.

<p>Under the Demo-Draft listing, you will see
the parameter "Order against: demo_seq2". With this setting,
the Demo-Draft contigs will be ordered using synteny to
Demo-Seq2.
		<td>
<a href="img/demoDraft1.png"><img src="img/demoDraft1.png" alt="demoDraft1" style="border: 1px solid black; width: 400px"></a>
	<tr><td style="colspan: 2; height: 40px;">&nbsp;
	<tr>
		<td style="vertical-align: text-top;">
Run the <ttl>Alignment &amp; Synteny</ttl>, where the alignment should take less than 30 minutes with one CPU.
When done, open the Summary for the pair, as shown on the right; <i>there
may be slight difference in the number of anchors because the results are
slightly different when run when different numbers of CPUs.</i>

See the first dot plot in <a href="Demo.html#draft" class="ext" target="_blank">Demo-draft</a>.
		<td>
<a href="img/demoSum2.png"><img src="img/demoSum2.png" alt="demoSum2" style="border: 1px solid black; width: 400px"></a>

</table>

<p>The ordering algorithm changes the order of the draft contigs in the database, but does
not change the sequence files on disk. However, it writes the following files:
<p>1. <i>File of ordered contigs:</i> It writes the order of the contigs along with
whether they should be flipped to a file called <tt>/data/seq/Demo_seq2/ordering.csv</tt>.

<p>2. <i>Fasta files of ordered sequences: </i>
It creates sequence files from the ordered contigs that are flipped when appropriate,
which are put into a new project
with the suffix "<tt>_ordered</tt>", as shown in the image below. The chromosome names correspond
to the order-against project (e.g. Demo_seq2), and the third chromosome is 'chr0'
which contains all draft sequences that were not placed.
<p>

<table>
<tr><td style="vertical-align: text-top;">
As shown in the image on the left, a new project has been added. Click <tt>Demo_Draft-ordered</tt> and
load it.
Running the <ttl>Alignment&amp;Synteny</ttl>
between the <tt>_ordered</tt> project and Demo_seq2 will provide a more coherent display
(see the second dot plot for <a href="Demo.html#draft" class="ext" target="_blank">Demo-draft</a>).
Using the <u>parameters</u> window, the <tt>Demo_Draft-ordered</tt> name can be shortened.
<td><td>
<a href="img/demoDraft3.png"><img src="img/demoDraft3.png" alt="demoDraft3" style="border: 1px solid black; width: 400px"></a>
</table>

<a id="demoself"></a>
<h3>Self alignment</h3>

The section on <a href="#self">Self alignments</a> discusses the Demo_Seq project.

<br>
<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!--------------------- NEW PROJECT ----------------------------------->
<table style="width: 95%"><tr><td style="text-align: left">
<h2>New Project</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a id="mysql"></a>
<h3>MySQL and parameters</h3>

If your machine does not have MySQL or MariaDB, download and install it. For example,
MySQL can be downloaded from <a href="http://dev.mysql.com" class="ext" target="_blank">dev.mysql.com</a>.
On a personal MacOS, simply download the '.dmg' file and following the instructions.
On a work server, the system administrator may need to install it.

<p>
<font color=red>Important Note</font>: The default settings of MySQL are poorly suited for large-scale
data storage. You will want to adjust the parameters
<tt>innodb_buffer_pool_size</tt> and <tt>innodb_flush_log_at_trx_commit</tt> as described in
<a href="TroubleShoot.html#msqslow" class="ext" target="_blank">Trouble Shoot MySQL</a>.

<p>
<a id="params"></a>
Parameters for accessing the MySQL database
should be set in the <tt>symap.config</tt> file in the main symap directory, as follows:
<p>
<table class="ty">
	<tr><td colspan=2><i>Database Parameters</i></td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_name</tt></td>
		<td>Name of the MySQL database, which SyMAP will create when it first reads
			<tt>symap.config</tt>. It is standard to start the name with <tt>symap</tt>,
			e.g <tt>symapDemo</tt>.
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_server</tt></td>
		<td>
			The machine hosting the MySQL database, e.g. <tt>myserver.myschool.edu</tt>. If using
			your local machine, enter <tt>localhost</tt>.
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_adminuser</tt></td>
		<td>
			MySQL username of a user with sufficient privileges to create a database. Needed for
			creating and updating alignments.
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_adminpasswd</tt></td>
		<td>
			Password of the admin user.
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_clientuser</tt></td>
		<td>
			MySQL username of a user with read-only access. This is only necessary if you
			want a machine to run SyMAP as read-only.
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_clientpasswd</tt></td>
		<td>
			Password of the client user (if <tt>db_clientuser</tt> is non-blank).
		</td></tr>
</table>

<p>Example <tt>symap.config</tt>.
<pre>
db_name             = symapDemo
db_server           = localhost
db_adminuser        = &lt;adminid&gt;
db_adminpasswd      = &lt;password&gt;
db_clientuser       =
db_clientpasswd     =
</pre>
To use an alternative file than <tt>symap.config</tt>, use the "-c" command line argument, e.g.
<pre>
&gt;./symap -c symapTmp.config
</pre>
This is useful if you have multiple SyMAP databases.
<a id="time"></a>
<h3>Runtime and Memory</h3>

If SyMAP runs out of memory, see <a href="TroubleShoot.html#memory" class="ext" target="_blank">Trouble Shoot</a>.
If MUMmer runs out of memory, see <a href="MUMmer.html" class="ext" target="_blank">MUMmer</a>.

<p>The largest component of SyMAP execution time is in running
MUMmer<sup>1,2</sup>. The time and memory for MUMmer all depends on the size
of the genomes.
For example, to align rice (12 chromosomes, 370Mb) to maize (10 chromosomes, 2Gb)
required 1 hour and 3 minutes using 8 CPUs
with 2.3Ghz speed. SyMAP used one CPU per maize chromosome to align the enter rice chromosome
against each of the 10 maize chromosomes (i.e. used 10 CPUs).
<p>
The memory usage of MUMmer is typically 5G per CPU, however it can
be as high as 10G for very long or repetitive chromosomes. If MUMmer fails, it is often due
to insufficient memory, see the <a href="MUMmer.html" class="ext" target="_blank">MUMmer</a> document, which
explains how to determine the problem and ways around it.
<!---------------------------------------------------------->
<!---------------------------------------------------------->
<br>&nbsp;
<a id="create"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Creating a New Project</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Start symap (i.e. <tt>./symap</tt>).
The sub-directories of <tt>/data/seq</tt> will be listed on the
left panel under <b>Projects</b>. You may create the project as discussed in <a href="#dir">Directory
structure</a>, or you may create a new project via the SyMAP interface.
<p>
<table>
<tr><td style="vertical-align: text-top;">
To create a new project via the SyMAP interface, press the <ttb>Add Project</ttb> button
at the lower left. Enter the project name beside <ttl>Name:</ttl>; the <ttb>Help</ttb>
button on the dialog provides further information.
<p>&nbsp;
<td>
<a href="img/projNew.png"><img src="img/projNew.png" alt="projNew" style="border: 1px solid black; width: 400px"></a>

<p>
<tr><td style="vertical-align: text-top;">
After saving the new project, it appears in the Projects list on the left, but
it is still an empty shell. A directory will be made under the <tt>/data/seq</tt>,
e.g. for the project added on the right, a directory will be created called <tt>/data/seq/foobar</tt>.
Check its box and it will appear in the Summary section (right hand side).
<td>
<a href="img/projNew2.png"><img src="img/projNew2.png" alt="projNew2" style="border: 1px solid black; width: 400px"></a>
</table>

<a id="params2"></a>
<h3>SyMAP Parameters</h3>

<table>
<tr>
<td style="vertical-align: text-top;">
<i>For a project:</i> Click the <u>Parameters</u> link to open the Parameters window shown in the image on the right.
<p>The <ttb>Help</ttb> button on this window provides the necessary details; it is very important to read this <ttb>Help</ttb>
so that you fully understand the parameters.
<p>
<p><i>Defaults:</i> If the sequences are put in the project's subdirectory <tt>/sequence</tt>
and the annotations are put in the sub-directory <tt>/annotations</tt>,
it is not necessary to provide the location of these files, e.g. the default directories for the project <tt>foobar</tt>
are as follows:
<pre>
/data/seq/foobar/annotations
/data/seq/foobar/sequences
</pre>
<p>
After setting the project parameters, the project is ready to be loaded.

<td><a href="img/parameters1.png"><img src="img/parameters1.png" alt="parameters1" style="border: 1px solid black; width: 420px"></a>
</table>


<a id="paramsAS"></a>
<p>
<i>For a pair</i>: Select a pair cell followed by the <ttb>Parameters</ttb> button.

<table>
<tr><td style="vertical-align: text-top;">

<a href="img/alignLine.png"><img src="img/alignLine.png" alt="alignLine" style="border: 1px solid black; width: 400px"></a>
<p>The <i>Alignment &amp; Synteny</i> parameters window is shown on the right.
See the <ttb>Help</ttb> for a description of the parameters.

<p>These parameters generally do not need to be changed.

<p>The <tt>MUMmer</tt> program <tt>PROmer</tt> is the default program
for sequence alignments for different genomes. The <tt>MUMmer</tt> program <tt>NUCmer</tt> is the default program
for sequence self-alignments.

<td>
<a href="img/parameters2.png"><img src="img/parameters2.png" alt="parameters2" style="border: 1px solid black; width: 300px"></a>
</table>


<!-------------------------------------------------------->
<br>
<a id="seqprep"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Sequence files</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The first decision with whole-genome sequence is whether to used repeat masked sequences.
<ul>
<li>Masking reduces alignment time and false-positive hits,
but also runs a risk of concealing true hits due to inaccurate masking.
<li>SyMAP does not perform the repeat-masking so must be done with another program.
However, you may obtain masked sequences from NCBI or Ensembl.
NCBI provides soft-masked sequences where the
<tt>scripts/ConvertNCBI</tt> script will convert it to hard-masked, and Ensembl
provides both soft and hard-masked sequences (see <a href="./convert">Convert</a>).
<li>
Masking is not really necessary unless the genome is highly repetitive
<i>and</i> those repeats are shared with other genomes being aligned.
(Repeats cause particular trouble for self-alignments,
see <a href="#self">self-alignments in SyMAP</a>).
<li>Occasionally, MUMmer fails aligning sequences
(see <a href="MUMmer.html" class="ext" target="_blank">MUMmer</a>, which explains detection and solutions
for known reasons for failure).
</ul>
<p>
Another masking option which is available if you have gene
annotation is to mask out everything but the annotated genes. You
can enable the "mask_all_but_genes" option on the Project's Parameters window (shown <a href="#params2">above</a>);
turn it on <i>before</i> doing the alignments.
<p>
Note that sequence files should be in FASTA format and the name of a sequence
is the string immediately following the "&gt;", e.g.
<pre>
&gt;Chr3  Oryza Sativa
GAATTCGAATTTGGGTAATGCTAATCAATACAGGTCAAAATCTATGTATTGAGTGGAATATACTGCAAAGTAATTACCTT
CTTCCAAAGGAAAGCATTCCTTCTCTCTTGTGGGACTAGCAGATGATCTCGCAGCCAAGACGTGACCACCCAAGGCTCAC
...
</pre>
In this example, the sequence name is "Chr3" and the Chr3 sequence follows. The additional information
"Oryza Sativa" is ignored.
<p>
<font color=green>Important points in naming sequences for SyMAP</font>:
<table style="margin-left:10px;margin-top:5px;padding:5px">
	<tr>
		<td style="vertical-align: text-top;">A.
		</td>
		<td style="vertical-align: text-top;">
			<u>Sequence names can contain only letters, numbers, and underscores. </u>
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;">B.
		</td>
		<td style="vertical-align: text-top;">
			<u>The sequence names must <i>exactly</i> match those
			used in the annotation files (first column), or the
			annotations will not be loaded. </u>
		</td></tr>
	<tr>
		<td style="vertical-align: text-top;">C.
		</td>
		<td style="vertical-align: text-top;">
			Use a consistent prefix such as "Chr" for all sequences,
			then set 'grp_prefix' to the prefix in Project's Parameters window (shown <a href="#params2">above</a>).

		</td></tr>
	<tr>
		<td style="vertical-align: text-top;">D.
		</td>
		<td style="vertical-align: text-top;">
			If there is not a consistent prefix, you may leave the 'grp_prefix' blank;
			beware, this can have unintended results, so should be avoided if possible.
			<u><i>Make the names short</i></u> so they will not clutter the display. You may
			need to rename your sequence, in which case, it must be done in the FASTA
			and GFF files.
		</td></tr>
</table>
<!--------------------------------------------------------------------->
<a id="annot"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Annotation files</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Annotation files should be in
<a href="http://www.sequenceontology.org/gff3.shtml" class="ext" target="_blank">gff3 format</a>.
The first column (seqid) must exactly match the sequence names in
the fasta files. The third column (type) determines how SyMAP
uses the entry. Types "gene", "exon", "centromere", and "gap"
are recognized (other entries are ignored).
<p>
The last column (attributes) contains "keyword=value" pairs describing
the annotation.
<ul><li>For genes, all attributes are saved in the database for viewing.
You can set which attributes keywords to save; open the
Project's Parameters window (shown <a href="#params2">above</a>), look for parameter "annot_keywords".
<li>For exons, the "parent" keyword=value is saved for viewing.
</ul>
Note: As of v5.1.8, the genes and exons must be in one file.

<p><i><b>NCBI files:</b></i> A Java script (<tt>scripts/ConvertNCBI.class</tt>) has been provided that
converts NCBI genome fasta files and gff annotation files
into the format that works best with SyMAP.
See the <a href="convert/ncbi.html" class="ext" target="_blank">documentation</a> for instructions.

<p><i><b>Ensembl files:</b></i> A Java script (<tt>scripts/ConvertEnsembl.class</tt>)
has been provided that converts Ensembl genome fasta files and gff3 annotation files
into the format that works best with SyMAP.
See the <a href="convert/ensembl.html" class="ext" target="_blank">documentation</a> for instructions.

<!--------------------------------------------------------------------->
<a id="draft2"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Draft alignment and ordering</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p><u>If you are ordering the draft sequence</u> against a closely related sequenced genome, see <a href="#draft">demo draft</a> on how to proceed.
It is strongly suggested you run the run the demo! In a nutshell, the steps are:
<ol>
<li>Load both sequences.
<li>For the draft,
bring up the <a href="#params2">Parameters</a> window. Beside the <ttl>order_against</ttl> row is a
drop-down of all loaded projects; select your whole genome project.
<li>Run the <ttl>Alignment&Synteny</ttl>. At the end, you will see a new project with the suffix
<tt>_ordered</tt> in the left panel. It will contains:
<ul>
<li>A <tt>sequence</tt> directory with a ".fa" FASTA file.
Using the abbreviation WG for the fully sequence genome:
given a set of
contigs that align to the WG ChrN, the file will have the contigs ordered under ">ChrN" with 100bp of N's
in between each two contigs. The FASTA file will contain every WG chromosome that the draft aligns to. Any extra
sequences will be put in ">Chr0".
<li>An <tt>annotation</tt> directory with a ".gff" file that specifies
where the gaps are.
</ul>
<li>Load the <tt>_ordered</tt> project. Then run <ttl>Alignment&Synteny</ttl> between the whole genome
project and the new <tt>_ordered</tt> project.
</ol>
The original draft directory will have a new file called <tt>ordered.csv</tt> that will contain
the merged contigs and what chromosome number they are mapped to.

<p><u>If you are not ordering the draft sequence</u>,
and if the draft sequence is in too many sequence pieces,
then (1) it takes a long time for the MUMmer comparisons, (2)
the display is very cluttered, and
(3) the blocks display does not work right. Limit the number of sequence pieces
by setting <tt>min_size</tt> in the parameters window to only load the largest 150 sequences;
there is a script called <tt>scripts/lenFasta.pl</tt> which
will print out all the lengths; set the min_size to the 150th length. However, even 150 are a lot of blocks to view
so you might want to start with the largest 50, merge them, then repeat.

<!-------------------------------------------------------------------->
<a id="self"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Self Alignments</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr>
<td>To perform self-synteny, select the cell for the same project followed by
<ttb>Selected Pair</ttb>. The <ttb>All Pairs</ttb> option does not include self-synteny.
<td><a href="img/self.png"><img src="img/self.png" alt="self" style="border: 1px solid black; width: 230px"></a>
</table>
<p>
By default, SyMAP uses the MUMmer 'nucmer' program for self-alignments. Each chromosomes
is compared to every other chromosome including itself.

<p>The Alignment &amp; Synteny <a href="#paramsAS">Parameters</a> popup has an option to set "Self Args",
which only uses the "Self Args" parameter(s) when comparing the chromosome sequence file
to itself.

<p>The self-synteny of demo_seq shows two tiny blocks. A better demonstration is to download
Arabidopsis thaliana from NCBI, convert it with the <a href="./convert" target="_blank" class="ext">
NCBI convert script</a>, and run the self synteny. It took 16 minutes with one processor on a Mac Mini (2018) with
64GB main memory.

<p>The "--maxmatch" parameter was the SyMAP default, but now it is up to the user as to whether
they want to add it to "Self Args" (it can greatly increase the execution time).
Reasoning for using "--maxmatch": MUMmer ordinarily seeds its alignments with unique matches,
which eliminates the possibility of off-diagonal seeds in the alignment of a chromosome to itself.
To overcome this problem, individual chromosome self-alignments can use the MUMmer parameter -maxmatch,
which removes the uniqueness requirement at the cost of greatly increased noise. The extra noise is then
filtered to a large extent by the default SyMAP filters, but the diagonal squares of the dot plot will
still have more noise visible than the off-diagonal.

<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!-------GENERAL----------------------------------------------------------->
<h2>General</h2>
<a id="dir"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Directory structure</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

For SyMAP v5, the default location for sequence and annotation files is:
<pre>
/data/seq/&lt;project-name&gt;/sequence
/data/seq/&lt;project-name&gt;/annotation
</pre>
You may do one of the following:
<ol>
<li>Create these sub-directories under <tt>/data/seq</tt> and put your files there.
<p>
<li>Create these sub-directories under <tt>/data/seq</tt> and use soft links to point to the file locations,
e.g.
<pre>
cd data/seq
mkdir foobar
cd foobar
ln -s &lt;location of directory of sequence files&gt; sequence
ln -s &lt;location of directory of annotation files&gt; annotation
</pre>
<li>Use the project parameter window to enter the location of the sequences and optional annotation.
</ol>
For options 1 and 2, it is not necessary to enter the locations of the files in the parameter window since
both use the default locations. Beware, executing the "Remove project from disk" (depending on
location), can remove the data files also.
<p>
The file results of the alignment and synteny computations are is as follows:
<pre>
   /data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/align
   /data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/final
</pre>
After the database is complete, these can be removed. However, sometimes SyMAP version updates require
the project files to be reloaded and/or the synteny to be recomputed; if these files remain, the existing
MUMmer files will be used, which <i><b>saves a lot of time</b></i>.

<p>The log files are in the <tt>/logs</tt> directory, see
<a href="MUMmer.html#log" class="ext" target="_blank">Running MUMmer</a> for more details.

<br>&nbsp;
<a id="ext"></a>
<h3>External programs, MUMmer and FPC</h3>

<a id="exe"></a>
<h4>External programs</h4>
The alignment programs are provided in the <tt>symap/ext</tt> directory. There are
executables for 64-bit Linux and 64-bit MacOS. SyMAP will
select the correct directory for the machine you are running from, i.e. you
do not need to do anything.

<a id="mum"></a>
<h4>MUMmer with SyMAP details</h4>
As of SyMAP v5.0.8, all MUMmer details have been put in a separate document,
see <a href="MUMmer.html" class="ext" target="_blank">MUMmer</a>.
This includes trouble shooting when MUMmer fails, and running MUMmer outside of SyMAP.

<!------------------------------------------------------------------------->
<a id="fpc"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>FPC project</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

For working with FPC<sup>8,9</sup>, it is suggested you use release v5.0.8 from
<a href="http://www.agcol.arizona.edu/software/symap">AGCoL</a>.

<ol>
<li>It has the FPC demo files.
<li>It has BLAT<sup>3</sup>  in the <tt>/ext</tt> directory.
<li>It has the tar file <tt>doc.tar.gz</tt> of the documentation.
<li>The AGCoL documentation applies to this release. See AGCoL
<a href="http://www.agcol.arizona.edu/software/symap/doc/SystemGuide.html#fpc">System Guide</a>.
<li>This is the last release made from AGCoL, so the documentation will stay consistent.
<li>It is also available from <a href="https://github.com/csoderlund/SyMAP/releases">Github</a>.
</ol>
If you run into any problems, please do not hesitate to contact symap@agcol.arizona.edu.

<!----------------------------------------------------------------------->
<a id="alg"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>How SyMAP Works</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

This section provides a brief overview of the SyMAP processing steps; for
more, see the SyMAP published papers<sup>5,6</sup>. The processing
has four phases:
<div style="padding:10px 0px 0px 20px;">
<i>Alignment:</i> <br>
The sequences are written to disk<sup>*</sup>, with gene-masking
if desired. In the alignment, one species is "query" and the other is
"target".
The query is the one with alphabetically the first name. The query sequences
are written into one large file, while smaller target sequences are grouped
into larger fasta files of size up to 60Mb, for more efficient processing
in MUMmer. There is an option "Concat" that if unchecked, the query sequences
are treated the same as the target; this is useful if the query and target are very large
genomes.
<p>
<i>Anchor Clustering:</i><br>
The raw anchor set consists of the hits found by MUMmer.
These are first clustered into gene, or putative-gene hits. This is
done by clustering the hit regions on each sequence, and then defining new "gene"
hits which connect these regions. For example if three separate
exons hit between two genes, they will be clustered into one "gene"
hit having a combined score equal to the sum of the raw hit scores.
Clustering is by gene if the hits overlap annotation, otherwise, it uses
a max separation 1kb, creating "putative gene" regions.
<p>
<i>Anchor Filtering:</i><br>
The clustered "gene anchors" are now filtered using a version of
reciprocal-best filtering which is adapted for retaining duplications and
gene families. For each pair of genes (or putative genes) which is
connected by a clustered anchor, the retained anchors must be among
 the top two anchors by score on both sides (top-2 allows for one
ancestral whole-genome duplication). An anchor will also be retained if its
score is at least 80% of that of the 2nd-best anchor on each side (this
allows for retention of gene family anchors). These filter parameters
may be adjusted through the Alignment &amp; Synteny Parameters window.
<p>
<i>Synteny Block Detection:</i><br>
After the clustered anchors are loaded into the database,  the synteny
synteny block algorithm runs. This algorithm looks for approximately-collinear
sequences of anchors, subject to several parameters including (A) Number
of anchors; (B) Collinearity of the anchors; (C) Amount of "noise" in the
surrounding region (to help reject false-positive chains). Criterion A can
be adjusted in the Alignment &amp; Synteny Parameters window.
<p>
<sup>*</sup> Note that the sequences are re-written from the database to the
disk for three reasons: (A) To allow re-grouping for efficiency; (B) To ensure elimination
of invalid characters; (C) To mask non-gene regions, if desired. This also ensures that
sequences names will match those in the database, and prevents problems caused by
moving the source sequences on disk.
</div>

<!------------------------------------------------------------------------->
<a id="update"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>How to update SyMAP with a new release</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
If you have been working with SyMAP and have existing projects, you can update to a new SyMAP version by
downloading SyMAP, and:
<ul>
<li>Put <tt>symap_5.tar.gz</tt> in a permanent location and untar it.
<li>Replace the <tt>/data</tt> and <tt>symap.config</tt> from your previous SyMAP location to this new location.
<li>This approach is safest as it acquires all changes (e.g. scripts) except for changes to the demo files.
</ul>
or
<ul>
<li>Put <tt>symap_5.tar.gz</tt> in a temporary location and untar it.
<li>Move <tt>symap_5/java/jar/symap.jar</tt> to the <tt>java/jar</tt> location of your permanent SyMAP.
<li>Check to
see if there are any <tt>/scripts</tt> or <tt>/ext</tt> changes that need to also be copied over.
</ul>

<p>The following table shows what versions require action by the user.
<table class=ty>
<tr><td><b>Release</b><td><b>Changed files</b><sup>1</sup><td><b>Action by User</b>
<tr><td>v5.2.0 <td><tt>symap</tt>, <tt>viewSymap</tt>, <tt>scripts/ConvertNCBI.*</tt><td><ttl>Alignment&Synteny</ttl><sup>2</sup>
<tr><td>v5.1.9 to v5.1.7<td>Only the <tt>symap.jar</tt><td>See <sup>3</sup>. If you have customized hit or gene colors, you will need to reset them.
</table>
<ol>
<li>Always get the new <tt>java/jar/symap.jar</tt>.
<p>
<li>The <ttl>Alignment&Synteny</ttl> will use existing MUMmer files if they have not been removed.
<p>
<li>To update to the latest Gene# assignment, it depends on what version you have:
<table class=ty>
<tr><td>v5.1.2 to v5.1.7<td>Execute <tt>./symap -z</tt>, then select <ttl>Reload Annotation</ttl>; the synteny algorithm does NOT need to be re-run.
<tr><td>pre-v5.1.2<td>Delete database, <ttl>Load Project</ttl> and run <ttl>Alignment&Synteny</ttl>. The database does not have
to be deleted, but its cleaner to do so.
</table>
</ol>

<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!----------------------------------------------------------------------->
<a id="refs"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>References</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<sup>1</sup>
Kurtz, S., Phillippy, A., Delcher, A.L., Smoot, M., Shumway, M., Antonescu, C., Salzberg, S.L.
(2004). Versatile and open software for comparing large genomes, Genome Biology, 5:R12

<p>
<sup>2</sup>
Marcais, G., A.L. Delcher, A.M. Phillippy, R. Coston, S.L. Salzberg, A. Zimin (2018).
MUMmer4: A fast and versatile genome alignment system,  PLoS computational biology, 14(1): e1005944.

<p>
<sup>3</sup> Kent, J. (2002) BLAT--the BLAST-like alignment tool, Genome Research 12:656-64.
<p>
<sup>4</sup>
Krzywinski, M., J. Schein, I. Birol, J. Connors, R. Gascoyne, D. Horsman, S. Jones, M. Marra (2009).
Circos: An information aesthetic for comparative genomics. Genome Research doi:10.1101/gr.092759.109.

<p>
<sup>5</sup>
Soderlund, C., Nelson, W., Shoemaker, A., and Paterson, A.(2006).
SyMAP: A system for discovering and viewing syntenic regions of FPC maps.
Genome Res. 16:1159-1168.
<p>
<sup>6</sup>
Soderlund, C., Bomhoff, M., and Nelson, W. (2011).
SyMAP: A turnkey synteny system with application to multiple large duplicated plant sequenced genomes.
Nucleic Acids Res V39, issue 10, e68.
<p>
<sup>7</sup>
Edgar, R (2004).
MUSCLE: a multiple sequence alignment method with reduced time and space complexity.
BMC Bioinformatics 113.
<p>
<sup>8</sup>
Soderlund, C., S. Humphrey, A. Dunhum, and L. French (2000).
Contigs built with fingerprints, markers and FPC V4.7. Genome Research 10:1772-1787.
<p>
<sup>9</sup>
Engler, F., J. Hatfield, W. Nelson, and C. Soderlund (2003).
Locating sequence on FPC maps and selecting a minimal tiling path. Genome Research 13:2152:2163.

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

