<!DOCTYPE HTML>
<html>
<head>
<title>SyMAP System Guide</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="http://www.agcol.arizona.edu">
		<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;System Guide</p>
<p>
<a href="#intro">Introduction</a>
<a href="#require">Requirements</a>
<a href="#installation">Install SyMAP</a>
<a href="#demo">Demo</a>
<a href="#create">New Project</a>
<a href="#gen">General</a>
<p>&nbsp;
<a href="#top">Go to top</a>
</div>

<!------ right side index and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav">
    <br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="./Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="./SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">
<br>
<table class="tx center">
<tr><td colspan="2">This document discusses building a SyMAP v5 database. It always applies to the latest release.

<tr><td style="vertical-align: text-top;">
<u>
Referenced external docs</u>
<ul>
<li><a href="input/index.html" class="ext" target="_blank">Input</a>
<li><a href="SystemHelp.html" class="ext" target="_blank">Parameters and Interface</a>
<li><a href="MUMmer.html" class="ext" target="_blank">MUMmer with SyMAP</a>
<li><a href="Demo.html" class="ext" target="_blank">Demo</a>
<li><a href="TroubleShoot.html" class="ext" target="_blank">Trouble-shoot</a>
</ul>
1. <u>Introduction</u>
<ul>
<li style="margin: 0.4em 0;"><a href="#intro">Overview and Publications</a>
<li style="margin: 0.4em 0;"><a href="#steps">Steps for finding synteny</a>
</ul>
2. <u>Requirements</u>
<ul>
<li style="margin: 0.4em 0;"><a href="#require">System Requirements</a>
<li  style="margin: 0.4em 0;"><a href="#tested">Tested platforms and timing results</a>
<li style="margin: 0.4em 0;"><a href="#java">Perl, Java and MySQL</a>
</ul>
3. <u>Install SyMAP</u>
<ul>
<li style="margin: 0.4em 0;"><a href="#installation">Tarball</a>
<li style="margin: 0.4em 0;"><a href="#ext">External executables</a>	<!-- in SyMAP -->
<li style="margin: 0.4em 0;"><a href="#mysql">MySQL parameters for SyMAP</a>
</ul>
<td style="vertical-align: text-top;">
4. <u>Demo</u>
<ul>
<li style="margin: 0.4em 0;"><a href="#demo">Running the demo</a>
<li style="margin: 0.4em 0;"><a href="#demoseq">Two genome synteny</a>
<li style="margin: 0.4em 0;"><a href="#draft">Draft ordering</a>
<li style="margin: 0.4em 0;"><a href="#demoself">Self synteny</a>
</ul>
5. <u>New project</u> <!-- #create ref in code -->
<ul>
<li style="margin: 0.4em 0;"><a href="#create">Create a new project</a>
<li style="margin: 0.4em 0;"><a href="#dir">Input and directory structure</a>
<li style="margin: 0.4em 0;"><a href="#load">Load project</a>
<li style="margin: 0.4em 0;"><a href="#ans">Alignment&Synteny (A&S)</a>
<li style="margin: 0.4em 0;"><a href="#draft2">Draft ordering</a>
<li style="margin: 0.4em 0;"><a href="#self">Self synteny</a>
<li style="margin: 0.4em 0;"><a href="#cancel">Cancel</a>
</ul>
6. <u>General</u>
<ul>
	<li style="margin: 0.4em 0;"><a href="#gen">How to update SyMAP with a new release</a>
	<li style="margin: 0.4em 0;"><a href="#alg">How SyMAP Works</a>
	<li style="margin: 0.4em 0;"><a href="#fpc">FPC project</a>
	<li style="margin: 0.4em 0;"><a href="#refs">References</a>
</ul>
<tr><td colspan=2><b>Transcriptome Analysis and Comparative
Transcriptomes</b>, see <a href="https://csoderlund.github.io/TCW/" class="ext" target="_blank">TCW</a>.
</table>

<!---------------------------------------------------->
<hr>
<a id="intro"></a>
<h2>1. Introduction</h2>

<table class="tz">
<td><a href="#overview">Overview</a>
	<td><a href="#pubs">Publications</a>
    <td><a href="#steps">Steps for finding synteny</a>
    <td style="text-align: right"><a href="#top">Go to top</a>
</table>

<a id="overview"></a>
<h3>Overview</h3>

SyMAP is a system for computing, displaying, and analyzing syntenic alignments between
divergent eukaryotic genomes.  It is designed for the comparison of a few genomes at a time
(i.e. 2-4) where synteny is computed between each pair.
Its features include the following:
<table class="tx">
<tr><td>
<ul>
	<li><i>Compute</i>
	<ul>
		<li>Find synteny between two sequenced eukaryotic genomes with optional annotation.
		<li>Draft sequence ordering by synteny, i.e. align a draft genome to a fully sequence (not draft-to-draft).
		<li>Self synteny.
	</ul>
	<li><i>Query and view</i>
	<ul>
		<li>For multiple selected synteny pairs, display using dot plot, circular, and side-by-side.
		<li>Complete queries on annotation, collinear genes, multi-hit genes, etc.
	</ul>
</ul>
<td><a href="img/symapPlants.png"><img src="img/symapPlants.png" alt="viewPlants"
		style="width:220px;border: 1px solid black;"></a>
</table>

<center><small>Click an image to see the closeup.</small></center>
<table class="tx">
	<tr>
		<td><a href="img/dotplot-3g.png"><img src="img/dotplot-3g.png" alt="dotplot-3-genome"
		style="width:270px; border: 1px solid black;"></a>
		<td><a href="img/circle-2g.png"><img src="img/circle-2g.png" alt="circle-2-genome"
		style="width:230px; border: 1px solid black;"></a>
		<td><a href="img/2d-3chr.png"><img src="img/2d-3chr.png" alt="2d-3chr"
		style="width:280px;border: 1px solid black;"></a>
</table>

<!---------------------------------------------------->
<a id="pubs"></a>
<h3>Publications</h3>
SyMAP is freely distributed software, however <font color=green><b><i>if you use SyMAP results in published
research, you must cite one or both of the following articles along with the external
program MUMmer<sup>1,2</sup></i></b></font>.
<pre>
        C. Soderlund, M. Bomhoff, and W. Nelson (2011)
        SyMAP: A turnkey synteny system with application to plant genomes.
        Nucleic Acids Research 39(10):e68.
</pre>
<p><pre>
        C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
        SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps
        Genome Research 16:1159-1168.
</pre>
The back-end processing of SyMAP runs MUMmer<sup>1,2</sup> for the alignments (included in the tarball)
and computes the synteny block from the alignment results. The SyMAP synteny algorithm is described
in the above two publications, though there are many unpublished updates since publication.
<!---------------------------------------------------->
<a id="steps"></a>
<h3>Steps for finding synteny</h3>

The following three scripts are provided in the tar file.
<table class="tx">
<tr><td>&nbsp;<td><tt>./xToSymap</tt><td>&nbsp;Format files from NCBI and Ensembl into a SyMAP friendly format.
<tr><td>&nbsp;<td><tt>./symap</tt><td>&nbsp;Build the SyMAP synteny database; view and query
<tr><td>&nbsp;<td><tt>./viewSymap</tt><td>&nbsp;View and query the database results
</table>

<p>Follow the steps below to get started with SyMAP.
<p>

<table class="tx">
	<tr><td style="vertical-align: text-top;">1.
		<td style="vertical-align: text-top;">Use Linux or MacOS.
		<td style="vertical-align: text-top;">See <a href="#machine">system requirements</a>.
	<tr>
		<td style="vertical-align: text-top;">2.
		<td style="vertical-align: text-top;">Requirements.
		<td style="vertical-align: text-top;">Set up <a href="#java">Perl, Java and MySQL</a> .
	<tr>
		<td style="vertical-align: text-top;">3.
		<td style="vertical-align: text-top;">Install SyMAP.
		<td style="vertical-align: text-top;">It is a simple unzip; see <a href="#installation">Installation</a>
		and <a href="#mysql">SyMAP MySQL parameters</a>.
	<tr>
		<td style="vertical-align: text-top;">4.
		<td style="vertical-align: text-top;">Run the demo.
		<td style="vertical-align: text-top;">Highly recommended. See <a href="#demo">running the demo.</a>
	<tr>
		<td style="vertical-align: text-top;">5.
		<td style="vertical-align: text-top;">Prepare input files.
		<td style="vertical-align: text-top;">FASTA sequence and optional GFF annotation.
		<br>See <a href="input/index.html" class="ext" target="_blank">Input</a> and <a href="#create">Creating a new project</a>.
	<tr>
		<td style="vertical-align: text-top;">6.
		<td style="vertical-align: text-top;">Load the files into SyMAP.
		<td style="vertical-align: text-top;">Select <ttb>Load project</ttb> by the project name.
		<br>See <a href="#load">Loading project</a>.
	<tr>
		<td style="vertical-align: text-top;">7.
		<td style="vertical-align: text-top;">Compute alignments and synteny.
		<td style="vertical-align: text-top;">Select <ttb>Selected project</ttb> beside <ttl>Alignment&amp;Synteny</ttl>.
		<br>See <a href="#ans">Alignment&amp;Synteny</a>.
	<tr>
		<td style="vertical-align: text-top;">8.
		<td style="vertical-align: text-top;">View results.
		<td style="vertical-align: text-top;">See
		<a href="UserGuide.html" class="ext" target="_blank">User Guide</a> for a detailed
		description of viewing and querying the results.
</table>
<!-------------------------------------------------->
<hr>
<a id="require"></a>
<h2>2. Requirements</h2>

<table class="tz"><tr>
<td><a href="#machine">System<br>Requirements</a>
<td><a href="#tested">Tested<br>platforms</a>
<td><a href="#time">Datasets<br>and Timings</a>
<td><a href="#disk">Disk space</a>
<td><a href="#java">Perl, Java,<br>MySQL</a>
<td style="text-align: right"><a href="#top">Go to top</a>
</table>

<a id="machine"></a>
<h3>System Requirements</h3>
The machine must be a Linux or MacOS 64-bit machine with sufficient memory for your dataset.

<p>The largest component of SyMAP execution time is running MUMmer<sup>1,2</sup>.
The time and memory for MUMmer all depends on the genomes sizes, complexity and similarity.
If MUMmer fails, it is often due to insufficient memory; see the
<a href="MUMmer.html" class="ext" target="_blank">MUMmer</a> document, which
explains how to determine the problem and ways around it.  It explains how
to <a href="MUMmer.html#cmd"  class="ext" target="_blank">run MUMmer on a different machine</a> and port the
results to the <tt>symap_5/data/seq_results</tt> directory.
The results in <a href="#time">Tested datasets and timings</a> gives an idea of compute times and memory.

<p>If SyMAP runs out of memory, see <a href="TroubleShoot.html#memory" class="ext" target="_blank">Trouble Shoot</a>.

<p> For viewing alignments with <tt>viewSymap</tt>, CPU and memory needs are typically negligible, unless
you are performing queries on more than 4-5 genomes at once.

<a id="tested"></a>
<h3>Tested platform</h3>

SyMAP has been tested on the following:
<table class="ty" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr><td><i>Machine</i><td><i>MySQL</i><td><i>Java</i><td><i>Core (CPU)</i><td><i>Memory</i><td><i>Purchased</i>
<tr><td colspan=6><i>v5.6.6 and later</i>:
<tr><td>&nbsp;&nbsp;1. <ttl>MacOS M4</ttl>
	<td>MySQL v8.0.42
	<td>24 from <a href="https://www.oracle.com/java/" class="ext" target="_blank">Oracle</a>
	<td>M4 12-Core<td>&nbsp;48Gb<td>2025

<tr><td colspan=6><i>v5.4.1 and later</i>:

<tr><td>&nbsp;&nbsp;2. <ttl>MacOS x86_64</ttl><sup>2</sup>
	<td>MySQL v8.0.33,<br>MariaDB 11.0.2
	<td>8, 15, 17, 18, 20 from
	<br><a href="https://adoptium.net" class="ext" target="_blank">Adoptium</a> and <a href="https://www.oracle.com/java/" class="ext" target="_blank">Oracle</a>
	<td>3.2 GHz 6-Core<td>&nbsp;64Gb<td>2018

<tr><td>&nbsp;&nbsp;3. <a href="http://linode.com" class="ext" target="_blank"><ttl>Linode</ttl></a> (Ubuntu 22.04.2 LTS)<sup>3</sup>
	<td>MySQL 8.0.33<td>17<td>Nanode<td>&nbsp;&nbsp;1Gb<td>2023

<tr><td colspan=6><i>v5.4.0 and earlier</i>:
<tr><td>&nbsp;&nbsp;4. <ttl>Linux amd64</ttl> (Centos)<sup>4</sup>
	<td>MariaDB v10.4.12
	<td>1.8
	<td>2.3 GHz 24-Core
	<td>128Gb
	<td>2011
</table>

<a id="time"></a>
<h3>Tested datasets and timings</h3>

<ol>
<li>MacOS M4 with 48Gb has been tested with the following:

<ol type="i">
<li>The complete <i>Homo sapiens</i> (24 chrs, 3Gb) with <i>Mus musculus</i> (21 chrs, 2.7Gb).
<br>To get MUMmer to run successfully, it was necessary to uncheck
<a href="SystemHelp.html#concat" class="ext" target="_blank">Concat</a>.
<p>
<li>Chr 5,17,X,Y of <i>Homo sapiens</i> (478Mb) and <i>Pan troglodytes</i> (447Mb);
<br>Chr 11,13,X,Y of <i>Mus musculus</i> (504Mb); Chr 11,19,X of <i>Oryctolagus cuniculus</i> (257Mb).
<p>
<li><i>Arabidopsis thaliana</i> (5 chr, 119Mb), <i>Brassica rapa</i> (10 chr, 297Mb),
<i>Brassica oleracea</i> (9 chr, 447Mb).
<p>
<li>Peach <i>Prunus persica</i> (8 chr, 227Mb) and draft <i>Prunus yedoensis</i> (250 scaffolds, 408Mb).
<br>Pyeudo.Peach was the result of ordering Pyedo against peach (8 chr, 270Mb).
</ol>

<p><u>Times</u> are those reported in
the SyMAP output, which uses the Java system time functions (clock times
are greater than the Java CPU system times). Repeated execution can result in times
varying by approximately +/-5secs.  All the following results use 4 CPUs (thread/core).

<p><table>
	<tr><td><u>Species</u><td><u>MUMmer</u><td><u>SyMAP</u><td>&nbsp;&nbsp;&nbsp;<td><u>Size</u>
	<tr><td>Hsa x Mus<td><b>30h:45m</b><td> 1m:37s<td>&nbsp;<td>3Gb x 2.7Gb, large genomes
 	<tr><td>Hsa x Pan<td><b>16h:40m</b><td> 0m:35s<td>&nbsp;<td>478Mb x 447Mb, closely related
    <tr><td>Hsa x Mus<td>0h:55m<td> 0m:04s<td>&nbsp;<td>478Mb x 504Mb
	<tr><td>A.thal x B.oler<td>0h:31m<td> 1m:14s<td>&nbsp;<td>119Mb x 447Mb
	<tr><td>B.raba x B.oler<td><b>4h:23m</b><td> 4m:12s<td>&nbsp;<td>297Mb x 447Mb, closely related
	<tr><td>Peach x Pyedo<td>0h:44m<td> 6m:51s<td>&nbsp;<td>227Mb x 408Mb, Pyedo ordered by peach
	<tr><td>Peach x Pyedo.Peach<td>0h:40m<td> 0m:20s<td>&nbsp;<td>227Mb x 270Mb
</table>

<p><li> MacOS x86_64 with 64Gb has been tested on 2-4 above. For comparison, the following A&S was computed
with 4 CPUs; note the longer time for the synteny computation on the older Mac machine.

	<table>
	<tr><td><u>Species</u><td><u>MUMmer</u><td><u>SyMAP</u>
	<tr><td>A.thal x B.oler&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<td>0h:33m&nbsp;&nbsp;<td>4m:06s
	</table>

<p><li> Linode nanode with 1Gb was too small to run MUMmer, so the MUMmer demo result files were transfered to the
<tt>data/seq_results/demo_seq_to_demo_seq2</tt> directory.
This allowed all other features to be tested on the demo, including running the synteny algorithm
without the alignment. Also, two tiny input files were used to test MUMmer.

<p><li>Linux amd64 with 128Gb was used extensively on large plant genomes.
<br>To align rice x maize (400Mb x 2365Mb) required a total of <b>1h:3m</b> using 8 CPUs
</ol>

<a id="disk"></a>
<h3>Disk space</h3>

Given the amazing disk space available for all modern computers, disk space should not be a problem.
MUMmer produces a <tt>.mum</tt> and <tt>.delta</tt> file; SyMAP only uses the <tt>.mum</tt> file
so it removes the MUMmer alignment <tt>.delta</tt> file. For the full <i>Homo sapiens</i> x <i>Mus musculus</i>,
the resulting <tt>data/seq_results/Hsa_to_Mus/align</tt> directory was 573M.

<p>You can remove the <tt>data/seq_results/&lt;project1_to_project2&gt;/</tt> directory after SyMAP has finished the synteny
computation, but <i><b>it is strongly recommended</b></i> that you leave them if you have the space. There
are frequent SyMAP updates with improvements to the clustering and synteny computations; if you have
kept these files, then you can update your database in very little time (note the >30h for MUMmer
for Hsa x Mus versus 1m:37s for the synteny). Also, you can easily try different parameters for the clustering
and synteny stages.
<p>If you do not want SyMAP to remove the  <tt>.delta</tt>, use the "-mum" command line argument.

<!------------------------------------------------>
<a id="java"></a>
<h3>Perl, Java and MySQL</h3>

<u>Perl:</u> This is for MUMmer; see <a href="https://mummer4.github.io/install/install.html" class="ext" target="_blank">MUMmer manual</a>, section on
<ttb>Software Requirements</ttb>. It states that the following are required: Perl5 (5.6.0), sh,
sed, awk (the last 3 are standard on any linux-based machine).

<p><u>Java</u>: <i>You must have Java version 17.0.11 or later</i>.
The released <tt>symap.jar</tt> file has been compiled with Java 17.0.11, which is upward compatible.

<p><u>MySQL</u>: If your machine does not have MySQL or MariaDB, download and install it. For example,
MySQL can be downloaded from <a href="http://dev.mysql.com" class="ext" target="_blank">dev.mysql.com</a>.
On a personal MacOS, simply download the '.dmg' file and following the instructions.
On a work server, the system administrator may need to install it.

<p>
<font color=red>Important Note</font>: The default settings of MySQL are poorly suited for large-scale
data storage. You will want to adjust the parameters
<tt>innodb_buffer_pool_size</tt> and <tt>innodb_flush_log_at_trx_commit</tt> as described in
<a href="TroubleShoot.html#msqslow" class="ext" target="_blank">Trouble Shoot MySQL</a>.

<!--------------------------------------------->
<hr>
<a id="installation"></a>
<h2>3. Install SyMAP</h2>
<table class="tz"><tr>
<td><a href="#tar">Tarball</a>
<td><a href="#ext">Externals executables</a>
<td><a href="#mysql">MySQL parameters</a>
<td style="text-align: right"><a href="#top">Go to top</a>
</table>

<a id="tar"></a>
<h3>Tarball</h3>

Installation consists of unzipping the download tarball using the command
<pre>     &gt; tar -xf symap_5.tar.gz</pre>
This can be done anywhere and it creates a directory called <tt>symap_5</tt>. You can
move this directory later if desired. The contents are:
<br>
<pre>
   LICENSE   README   data/   ext/   java/
   scripts/  symap.config     symap  viewSymap  xToSymap
</pre>
<u>Data</u>: The <tt>data/</tt> directory contains a <tt>seq/</tt> sub-directory,
which contains the demo files, and is the default location for all input sequence files.
The <tt>symap</tt> script expects to find the <tt>data</tt> directory, the <tt>viewSymap</tt> does not.

<a id="ext"></a><!-- in SyMAP -->
<h3>External executables</h3>

The <tt>ext/</tt> directory contains the external programs
MUMmer<sup>1,2</sup> for sequence alignment,
and MAFFT<sup>6</sup> and MUSCLE<sup>7</sup> for
interactive MSA alignment (for <a href="./Query.html#msaAlign" class="ext" target="_blank">Queries</a>).

The directory contains:
<pre>
	README		mummer/		mummer4/	muscle/		mafft/
</pre>

Each has subdirectories:
<table class="ty" style="margin-left:30px">
<tr><td><i>Subdirectory</i><td><i>OS (Architecture)</i><td><i>Note</i>
<tr><td>lintel64<td>Linux<td>
<tr><td>mac<td>Mac OS X (x86_64)<td>
<tr><td>macM4<td>Mac OS X (M4 silicon)<td>No muscle executable
</table>
<p>SyMAP will determine which subdirectory to use.

<p>If you compile your own executables for a different machine (architecture), do the following:
<ol>
<li>Under mummer and mafft, make a directory with your machine name.
<li>Put the executables under this directory in the exact same way as shown for lintel64.
<li>In the symap configuration file (default <tt>symap.config</tt>), add a line
<pre>     arch={your directory architecture name}</pre>
</ol>
Note that I do not have an alternative machine to try this on, but it should work. Email me
at <ttl>symap@agcol.arizona.edu</ttl> if it does not.

<p>For MUMmer, see <a href="MUMmer.html#ext" class="ext" target="_blank">Executables</a>
and <a href="MUMmer.html#mum4" class="ext" target="_blank">Using MUMmer4</a>.
On MacOS, you may also need to view <a href="TroubleShoot.html#macos" class="ext" target="_blank">MacOS externals</a>.

<!---------------------------------------------------------->
<a id="mysql"></a>
<h3>MySQL parameters for SyMAP</h3>

<a id="params"></a>
Parameters for accessing the MySQL database
should be set in the <tt>symap.config</tt> file in the main symap directory, as follows:
<p>
<table class="ty">
	<tr><td colspan=2><i>Database Parameters</i>
	<tr>
		<td style="vertical-align: text-top;"><tt>db_name</tt>
		<td>Name of the MySQL database, which SyMAP will create when it first reads
			>symap.config</tt>. It is standard to start the name with <tt>symap</tt>,
			e.g <tt>symapDemo</tt>.
	<tr>
		<td style="vertical-align: text-top;"><tt>db_server</tt>
		<td>The machine hosting the MySQL database, e.g. <tt>myserver.myschool.edu</tt>. If using
			your local machine, enter <tt>localhost</tt>.
	<tr>
		<td style="vertical-align: text-top;"><tt>db_adminuser</tt>
		<td>MySQL username of a user with sufficient privileges to create a database. It is also
			necessary for loading, deleting and running synteny.
	<tr>
		<td style="vertical-align: text-top;"><tt>db_adminpasswd</tt>
		<td>Password of the admin user.
	<tr>
		<td style="vertical-align: text-top;"><tt>db_clientuser</tt>
		<td>Optional: MySQL username of a user with read-only access. This is only necessary if you
			want a machine to run <tt>viewSymap</tt> as read-only.
	<tr>
		<td style="vertical-align: text-top;"><tt>db_clientpasswd</tt>
		<td>Optional: Password of the client user (if <tt>db_clientuser</tt> is non-blank).
</table>

<p>Example <tt>symap.config</tt>.
<pre>
  db_name             = symapDemo
  db_server           = localhost
  db_adminuser        = &lt;adminid&gt;
  db_adminpasswd      = &lt;password&gt;
  db_clientuser       =
  db_clientpasswd     =
</pre>
To use an alternative file than <tt>symap.config</tt>, use the "-c" command line argument, e.g.
<pre>
  &gt;./symap -c symapTmp.config
</pre>
This is useful if you have multiple SyMAP databases.


<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!--------- DEMO ---------->
<hr>
<a id="demo"></a>
<h2>4. Demo</h2>

<table class="tz"><tr>
<td><a href="#rundemo">Running the demo</a>
<td><a href="#demoseq">Two genome synteny</a>
<td><a href="#draft">Draft ordering</a>
<td><a href="#self">Self synteny</a>
<td style="text-align: right"><a href="#top">Go to top</a>
</table>

<a id="rundemo"></a>
<h3>Running the Demo</h3>
If you have not used SyMAP before, it is essential to run the demos. After you have
installed MySQL, do the following:
<ol>
<li>Change into the <tt>symap_5</tt> directory.
<p>
<li>Edit <tt>symap.config</tt> and enter database
and host information (see <a href="#mysql">MySQL</a>).
<p>
<li>From the command line, type <tt>./symap</tt>.
</ol>
<p>The first time you run SyMAP, it will create the database with information written to the terminal, e.g.
<pre>
Creating database 'symapDemo' (jdbc:mysql://localhost/symapDemo?characterEncoding=utf8).
</pre>
<br>It will check your MySQL variables; if there are any "Suggested" changes,
see <a href="TroubleShoot.html#msqslow" class="ext" target="_blank">Trouble Shoot MySQL</a>.
<p>It will also
check that the provided external programs (e.g. MUMmer) are executable; if it shows any problems,
see <a href="MUMmer.html#ext" class="ext" target="_blank">Executables</a>.
For MacOS, you may also need <a href="TroubleShoot.html#macos" class="ext" target="_blank">MacOS externals</a>.
<p>

<a id="demoseq"></a>
<h3>Demo two genome synteny</h3>

<table>
<tr><td colspan=2>Executing <tt>./symap</tt> will bring up the <ttb>Manager</ttb> panel
as shown on the lower right; it will show the three demo projects provided with the SyMAP tarball.

<tr><td style="vertical-align: text-top;">
Check <ttx>Demo-Seq</ttx> and <ttx>Demo-Seq2</ttx> and they will be displayed on the right panel.
<p>A link <font color="blue"><u>Load All Projects</u></font> will be displayed at the top of the right panel; select it to
load the projects, which will take several minutes.
If loading the <ttx>Demo-Seq</ttx> takes
<i>more than a few minutes</i>, you may need to adjust the MySQL parameters, see
<a href="TroubleShoot.html#msqslow" class="ext" target="_blank">TroubleShoot MySQL</a>.

<td><small>Project Manager</small><br>
<a href="img/demoLoad.png"><img src="img/demoLoad.png" alt="demoLoad" style="border: 1px solid black; width: 400px"></a>

<tr><td style="colspan: 2; height: 40px;">&nbsp;
<tr><td style="vertical-align: text-top;">
When done, the <ttb>Manager</ttb> will look like the image shown on the right.
Your may verify the results by selecting the <a href="img/view.png">View</a> link.

<p>In the <ttl>Available Syntenies</ttl> table, the cell for  <ttx>Demo_Seq2</ttx> and <ttx>Demo_Seq</ttx> will
automatically be selected (green cell).

<p>Click the <ttb>Selected Pair</ttb> button to start the <ttl>Alignment&amp;Synteny</ttl>.
 <ttb>All Pairs</ttb> button can also be selected, as it will perform <ttl>Alignment&amp;Synteny</ttl> on all pairs
in the table.
<td>
<a href="img/demoMgr.png"><img src="img/demoMgr.png" alt="demoMgr" style="border: 1px solid black; width: 400px"></a>
<tr><td style="colspan: 2; height: 40px;">&nbsp;
<tr>
<td style="vertical-align: text-top;">
The <ttl>Alignment&amp;Synteny</ttl> takes less than 5 minutes on the MacOS 10.5 but could take up to 30 minutes
on a slow machine.
<p>When done, the table will have a checkmark (&#10003), signifying that the synteny is available for viewing.

<td>
<a href="img/demoTable.png"><img src="img/demoTable.png" alt="demoTable" style="border: 1px solid black; width: 400px"></a>

<tr><td style="colspan: 2; height: 40px;">&nbsp;
<tr>
<td style="vertical-align: text-top;">
Click <ttb>Summary</ttb> to view the summary shown on the right;
<i>there may be slight differences in the #Cluster hits</i>
because of different numbers of CPUs, MacOS vs Linux, etc (but the #Blocks come out the same).
<p>To view the other interfaces, see <a href="Demo.html" class="ext" target="_blank">Demo Results</a>.
<p>&nbsp;<p>
Once the alignments are computed, the A&S parameters can be experimented without
having to redo the alignments.
Try using the <ttl>Algorithm 2 (exon-intron)</ttl> for <i>Cluster Hits</i> by
clicking the <ttb>Parameters</ttb> button, which opens
the <a href="SystemHelp.html#pairParams" class="ext" target="_blank">Parameters</a> panel,
then select <ttl>Algorithm 2</ttl> and <ttb>Save</ttb>. Click <ttb>Selected Pairs (Redo)</ttb> which
will result in a summary similar to <a href="img/demoSum2.png" class="ext" target="_blank">this</a>.
<td>
<a href="img/demoSum.png"><img src="img/demoSum.png" alt="demoSum" style="border: 1px solid black; width: 400px"></a>
</table>

<a id="draft"></a>
<h3>Demo draft ordering</h3>

<table>
	<tr><td style="vertical-align: text-top;">
Load the <ttx>Demo-Draft</ttx> project. Select <ttx>Demo-Seq2</ttx> on the left side so it shows in the
<ttl>Available Syntenies</ttl>. Select the Demo-Draft/Demo-Seq2 cell.
Open the <a href="./SystemHelp.html#over2" class="ext" target="_blank">Parameters</a> panel;
at the bottom, select <ttx>Draf->Seq2</ttx>.
		<td>
<a href="img/parameters2c.png"><img src="img/parameters2c.png" alt="order parameters" style="border: 1px solid black; width: 320px"></a>
	<tr><td style="colspan: 2; height: 40px;">&nbsp;

	<tr><td style="vertical-align: text-top;">
Run the <ttl>Alignment&amp;Synteny</ttl>, where the alignment should take less than 10 minutes with one CPU.
When done, open the <ttb>Summary</ttb> for the pair which will be similar to what is shown on the right; as mentioned above, <i>there
may be slight difference in the number of Cluster hits.</i>
See the first dot plot in <a href="Demo.html#draft" class="ext" target="_blank">Demo-draft</a>.
<p>It is recommended that the
<a href="SystemHelp.html#clust" class="ext" target="_blank">Cluster Algo1</a> be used for ordering sequence.
		<td>
<a href="img/demoSumDraft.png"><img src="img/demoSumDraft.png" alt="demo Sum Draft" style="border: 1px solid black; width: 400px"></a>

</table>

<p>The ordering algorithm creates the following files and directories:
<p>1. <i>File of ordered contigs:</i> It writes the order of the contigs along with
whether they should be flipped to a file called <tt>/data/seq/demo_draft/demo_seq2_ordered.csv</tt>.

<p>2. <i>New project:</i> It creates a new project directory called
<tt>data/seq/draft_demo..demo_seq2</tt> (the naming allows <tt>demo_draft</tt> to unambiguously
be ordered against different genomes).
This directory will contain a sub-directory <tt>/sequence</tt> which contains the sequence file.

<p>3. <i>FASTA sequence file of ordered sequences: </i>
The sequence file will contain the ordered contigs that are flipped when appropriate.
The chromosome names correspond
to the order-against project (e.g. <ttx>demo_seq2</ttx>). Chromosome 'chr0'
contains all draft sequences that were not placed.
<p>
The new project is shown on the left side of the <ttb>Manager</ttb>.
The display name is <ttx>Demo_Draft.Demo_seq2</ttx>, click and load it.

<table>
<tr><td style="vertical-align: text-top;">
Running the <ttl>Alignment&amp;Synteny</ttl> between the <ttx>Demo_Draft.Demo_seq2</ttx> project
and <ttx>Demo_seq2</ttx> will provide a more coherent display (see the second dot plot for
<a href="Demo.html#draft" class="ext" target="_blank">Demo-Draft</a>).
Using the project's <a href="SystemHelp.html#projParams" class="ext" target="_blank">Parameters</a> panel,
the <ttx>Demo_Draft.Demo_Seq2</ttx> display name can be shortened.
<td><td>
<a href="img/demoDraft3.png"><img src="img/demoDraft3.png" alt="demoDraft3" style="border: 1px solid black; width: 400px"></a>
</table>

<a id="demoself"></a>
<h3>Demo self synteny</h3>

<table>
<tr>
<td>To perform self synteny, select the cell for <ttx>Demo_seq</ttx> row and column (it turns green) followed by
<ttb>Selected Pair</ttb>.
This computes very few blocks, as shown in
<a href="Demo.html#self" class="ext" target="_blank">Demo self synteny</a>.
<td><a href="img/self.png"><img src="img/self.png" alt="self" style="border: 1px solid black; width: 370px"></a>
</table>

<br>
<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!--------------------- NEW PROJECT ----------------------------------->
<!-- ref create in code --->

<hr>
<a id="create"></a>

<h2>5. New project</h2>
<table class="tz"><tr>
<td><a href="#create2">Create</a>
<td><a href="#dir">Input and<br>directory</a>
<td><a href="#load">Load<br>project</a>
<td><a href="#ans">Alignment<br>&amp; Synteny</a>
<td><a href="#draft2">Draft<br>ordering</a>
<td><a href="#self">Self<br>synteny</a>
<td><a href="#cancel">Cancel</a>
<td style="text-align: right"><a href="#top">Go to top</a>
</table>
<a id="create2"></a>
<h3>Create a new project</h3>

<u>Database name</u>: The parameter <ttx>db_name</ttx> (database name) in the <tt>symap.config</tt>
file that comes with the tar file is set to <ttx>symapDemo</ttx>.
Edit the <tt>symap.config</tt> file to set the database name to something more meaningful.
It is protocol to start the database name with 'symap'.

<p> The following gives an outline of the steps and important details
of the directory structure, linking to the appropriate sections
in <a href="./SystemHelp.html" class="ext" target="_blank">Interface and parameters</a> document, which
details all functions and parameters.
<p>

<a id="dir"></a>
<h3>Input and directory structure</h3>
<u>Input</u>:
See <a href="./input/index.html#projDir" class="ext" target="_blank">Input</a> for description of
the sequence (FASTA) and annotation (GFF) files. The
<a href="input/index.html#if" class="ext" target="_blank">xToSymap</a> interface takes as input NCBI or Ensembl files,
and outputs the best input files for SyMAP, along with automatically creating the correct directory
structure.

<p><u>Project-name</u>:
Each project has a <ttx>project-name</ttx>, which is also the project's directory name.
The <ttx>project-name</ttx> should be a short descriptive name; see the <a href="input/index.html#iloc" class="ext" target="_blank">
Input's Project directory</a> for important details.
The use of the <ttx>project-name</ttx> is shown below.

<p><u>Project directory structure</u>:

<p>Each project has a directory as follows:
<pre>
  /data/seq/&lt;project-name&gt;
</pre>

<p>The default location for sequence and annotation files is:
<pre>
  /data/seq/&lt;project-name&gt;/sequence
  /data/seq/&lt;project-name&gt;/annotation
</pre>
To indicate to the <ttp>symap</ttp> <ttb>Load project</ttb> function where the project's input files are,
do one of the following three:
<ol>
<li>Create  these sub-directories under <tt>/data/seq</tt> and put your files there,
<br>e.g using <ttx>project-name=foobar</ttx>:
<pre>
  cd data/seq
  mkdir foobar
  cd foobar
  mkdir sequence
  mkdir annotation
</pre>
Move your FASTA file(s) to <tt>data/seq/&lt;project-name&gt;/sequence</tt>
(e.g. <tt>data/seq/foobar/sequence</tt>)
and your optional GFF files(s) to <tt>data/seq/&lt;project-name&gt;/annotation</tt>
(e.g. <tt>data/seq/foobar/annotation</tt>)
<p>

<li>Create these sub-directories under <tt>/data/seq</tt> and use soft links to point to the file locations,
<br>e.g using <ttx>project-name=foobar</ttx>:
<pre>
  cd data/seq
  mkdir foobar
  cd foobar
  ln -s &lt;location of directory of sequence files&gt; sequence
  ln -s &lt;location of directory of annotation files&gt; annotation
</pre>
<li>Use the <ttb>Add project</ttb> button on the <ttp>symap</ttp> interface (lower-left corner) to
add the project name to the <tt>data/seq</tt> directory.
Use the <a href="SystemHelp.html#projParams" class="ext" target="_blank">project's parameter</a>
panel to enter the location of the sequences and optional annotation files into the <ttx>Sequence files</ttx>
and <ttx>Anno files</ttx> parameters.
</ol>
<p>For options 1 and 2, it is not necessary to enter the locations of the files in the
<a href="SystemHelp.html#projParams" class="ext" target="_blank"> project parameter</a> panel since
both use the default locations.
<p>
All sub-directories in <tt>data/seq</tt> are shown on the left-panel of the <ttp>symap</ttp> interface.
<table>
<tr><td style="vertical-align: text-top;">
<p>Each project is shown with its <ttx>Display name</ttx>, which is set in the project's parameters.
The <ttx>project-name</ttx> is shown on the right hand image as the <ttx>Directory name</ttx>, as they are the same.
<td>
<a href="img/projNew2.png"><img src="img/projNew2.png" alt="projNew2" style="border: 1px solid black; width: 450px"></a>
</table>
The <ttp>symap</ttp> program requires the directories to be present (and will create them if necessary), <ttp>viewSymap</ttp> does not.

<!------------ load ----------------------->
<a id="load"></a>
<h3>Load project</h3>
See <a href="./SystemHelp.html#projParams" class="ext" target="_blank">project parameters</a>
on setting the parameters.

See <a href="./SystemHelp.html#selLoad" class="ext" target="_blank">Load</a> for loading a project.

<table style="width: 95%"><tr><td style="text-align: left"><td style="text-align: right"><a href="#top">Go to top</a></table>
<!------------ ans ----------------------->
<a id="ans"></a>
<h3>Alignment&amp;Synteny</h3>

<table class="tx">
<tr><td style="vertical-align: text-top;">See <a href="./SystemHelp.html#pairParams" class="ext" target="_blank">Parameters</a>
on setting the pair parameters for the this step.

<p>See <a href="SystemHelp.html#cpu" class="ext" target="_blank">CPU and Verbose</a> for a description of these
two options.
<p>Then select <ttb>Selected Pair</ttb>
to run the Alignment&amp;Synteny.
<td><a href="img/demoTable2.png"><img src="img/demoTable2.png" alt="align and synteny" style="border: 1px solid black; width: 500px"></a>
</table>

<p><u>Alignment&amp;Synteny files</u>:
The result files are in the following directories:
<pre>
   /data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/align
   /data/seq_results/&lt;project-name1&gt;-to-&lt;project-name2&gt;/final
</pre>
As mentioned in <a href="#disk">Disk</a>, after the database is complete, these can be removed.
However, sometimes SyMAP version updates require
the project files to be reloaded and/or the synteny to be recomputed; if these files remain, the existing
MUMmer files will be used, which <i><b>saves a lot of time</b></i>.

<p>The log files are in the <tt>/logs</tt> directory, see
<a href="MUMmer.html#log" class="ext" target="_blank">Running MUMmer</a> for more details.

<p>See <a href="MUMmer.html"  class="ext" target="_blank">Using MUMmer with SyMAP</a>
for a discussion on how it works in SyMAP, trouble-shooting, and running MUMmer externally
(i.e. if your local machine does not have enough memory or CPUs, you may need to run it on
a bigger machine).

<p><u>Incomplete alignment</u>
If SyMAP completed the alignment, e.g. the demo <tt>/align</tt> directory will have the following files:
<p><pre>
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:28 all.done
-rw-r--r--@ 1 cari  staff   1.2M Apr 10 10:28 demo_seq_cc.demo_seq2_f1.mum
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:28 demo_seq_cc.demo_seq2_f1.mum.done
-rw-r--r--@ 1 cari  staff   541K Apr 10 10:27 demo_seq_cc.demo_seq2_f2.mum
-rw-r--r--@ 1 cari  staff     0B Apr 10 10:27 demo_seq_cc.demo_seq2_f2.mum.done
</pre>
<ol>
<li>The 'all.done' indicates that the all alignments completed.
<li>If 'all.done' does not exist, SyMAP will perform any alignments that do NOT have a corresponding 'mum.done'.
If the <ttl>Concat</ttl> setting has been switched between the previous and current run, this will not work correctly.
<li>If the user supplied the alignments (<a href="MUMmer.html#cmd"  class="ext" target="_blank">Supply MUMmer files</a>),
there may not be 'mum.done' files but there should be an 'all.done', so it is assumed they are all done.
</ol>

<!--------------------------------------------------------------------->
<a id="draft2"></a>
<h3>Draft alignment and ordering</h3>

<u>If you are ordering the draft sequence</u> against a closely related sequenced genome,
see <a href="#draft">demo draft</a> on how to proceed.
It is strongly suggested you run the demo! In a nutshell, the steps are:
<ol>
<li>Load both sequences.
<li>Open the pairs's <a href="SystemHelp.html#over2" class="ext" target="_blank">Parameters</a> panel.
At the bottom of the panel, select the radio button that indicates ordering your draft against
a complete sequence (e.g. Draf-&gtSeq2).

<li>Run the <ttl>Alignment&Synteny (A&S)</ttl>. At the end, you will see a new project with
the two <ttx>project-names</ttx> separated by ".." in the left panel. It will contain:

<ul>
<li>A <tt>sequence</tt> directory with a ".fa" FASTA file. Any scaffolds matching the <ttx>Order against</ttx>
genome will be assigned the same chromosome name. All scaffolds aligning to an <ttx>Order against</ttx> chromosome
will be appended together in order with 100 N's between each scaffold.
Any extra sequences will be put in ">Chr0".
<li>An <tt>annotation</tt> directory with a ".gff" file that specifies where the gaps are.
</ul>
<li>Load the new project. Then run <ttl>A&S</ttl> between this new project and the original
whole genome project.
</ol>
The original draft directory will have a new file with the suffix <tt>ordered.csv</tt> that will contain
the merged contigs and what chromosome number they are mapped to.

<p><i>If the draft sequence is in too many sequence contigs</i>,
it takes a long time for the MUMmer comparisons. Also,
the displays are very cluttered to the point of unreadable, but these will generally be merged in step 4 above, so the
new display is fine. Nevertheless, you may want to remove the smallest contigs. This can be done by
limiting the number of sequence contigs by setting <ttx>Minimum length</ttx> in the project's
<a href="SystemHelp.html#projParams" class="ext" target="_blank">Parameter</a> panel
to only load the largest 150 sequences.
To determine the minimal length, use the <ttl>Lengths</ttl> button in the
<a href="./input/index.html#ol" class="ext" target="_blank">xToSymap</a>
interface, which
will print out all the lengths; set the <ttx>Minimum length</ttx> to the 150th length.

<!-------------------------------------------------------------------->
<a id="self"></a>
<h3>Self synteny</h3>

<table>
<tr>
<td>To perform self synteny, select the cell for the same project (it turns green) followed by <ttb>Selected Pair</ttb>
(the <ttb>All Pairs</ttb> option does not include self synteny).
<td><a href="img/self.png"><img src="img/self.png" alt="self" style="border: 1px solid black; width: 370px"></a>
</table>

By default, SyMAP uses the MUMmer 'NUCmer' program for self-alignments. Each chromosomes
is compared to every other chromosome including itself.
<p>The Alignment&amp;Synteny <a href="SystemHelp.html#pairParams" class="ext" target="_blank">
Parameters</a> panel has an option to set <ttx>Self Args</ttx>,
which is only used when comparing the chromosome sequence file
to itself. Make sure that the <i>Cluster Hits</i> <ttx>Algorithm 1</ttx> option is selected.
<p>
<table class="tx">
<tr><td style="vertical-align: text-top;">
<p>A better demonstration than the demo is to download
Arabidopsis thaliana from NCBI, convert it with the
<a href="./input/ncbi.html" target="_blank" class="ext">
NCBI convert script</a>, and run the self synteny. It took 16 minutes with one processor on a Mac Mini (2018) with
64Gb main memory. The dot plot is shown on the right (click on the image for a closeup view).
<td><a href="img/arabSelf.png"><img src="img/arabSelf.png" alt="arab self"
style="border: 1px solid black; width: 370px"></a>
</table>

<a id="cancel"></a>
<h3>Cancel</h3>

<table class="tx">
<tr>
<td>The <ttb>Load</ttb> and <ttl>A&S</ttl> methods have a popup progress panel, as shown on
the right. There is a <ttb>Cancel</ttb> button on the bottom that can be clicked to cancel the execution;
it will remove the results from the database and exit.
<p>Occasionally, the <ttb>Cancel</ttb> will cause it to create an error, writing to the <tt>error.log</tt>
or to the terminal. This is not a problem, though you may need to remove the results yourself.
<p>If MUMmer is running when you <ttb>Cancel</ttb>, make sure there is "Error: Failed command:" line
to terminal for each MUMmer alignment that was running; if there is not, use the "top" linux command to
view the running processes and stop any MUMmers still running.
<td><a href="img/cancel.png"><img src="img/cancel.png" alt="self" style="border: 1px solid black; width: 350px"></a>
</table>
Also, see <a href="TroubleShoot.html#hang" class="ext" target="_blank">Trouble Shoot Hang</a>
<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!-------GENERAL----------------------------------------------------------->
<hr>
<a id="gen"></a>
<h2>6. General</h2>
<table class="tz"><tr>
<td><a href="#update">Update</a>
<td><a href="#alg">How SyMAP works</a>
<td><a href="#fpc">FPC</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#refs">References</a>
<td style="text-align: right"><a href="#top">Go to top</a>
</table>
<!------------------------------------------------------------------------->
<a id="update"></a>
<h3>How to update SyMAP with a new release</h3>
If you have been working with SyMAP and have existing projects:
<ul>
<li>If the <tt>symap.jar</tt> is available from the download site and there are only changes to it,
download it and replace the one in <tt>symap_5/java/jar</tt>.
</ul>
or if there are changes to more than the jar file:
<ul>
<li>Put the new <tt>symap_5.tar.gz</tt> in a permanent location and untar it.
<li>Replace the <tt>/data</tt> and <tt>symap.config</tt> from your previous SyMAP location to this new location.
<li>This approach is safest as it acquires all changes (e.g. scripts) except for changes to the demo files.
</ul>
or
<ul>
<li>Put the new <tt>symap_5.tar.gz</tt> in a temporary location and untar it.
<li>Move <tt>symap_5/java/jar/symap.jar</tt> to the <tt>java/jar</tt> location of your permanent SyMAP.
<li>Check to
see if there are any <tt>/scripts</tt> or <tt>/ext</tt> changes that need to also be copied over.
</ul>

The <ttl>Alignment&Synteny</ttl> will use existing MUMmer files if they have not been removed.

<!----------------------------------------------------------------------->
<a id="alg"></a>
<h3>How SyMAP Works</h3>

This section provides a brief overview of the SyMAP processing steps; for
more, see the SyMAP published papers<sup>4,5</sup>. The processing
has four phases:
<div style="padding:10px 0px 0px 20px;">
<i>Alignment:</i> <br>
The sequences are written to disk<sup>*</sup>, with gene-masking
if desired. In the alignment, one species is "query" and the other is
"target".
The query is the one with alphabetically the lesser name (e.g. A&lt;B). The query sequences
are written into one large file, while smaller target sequences are grouped
into larger FASTA files of size up to 60Mb, for more efficient processing
in MUMmer.

There is an option <ttl>Concat</ttl> that if unchecked, the query sequences
are treated the same as the target; i.e. generally there will be more sequence files to
 processed, but they will be smaller.
 See <a href="SystemHelp.html#concat" class="ext" target="_blank">Concat</a> for a description and timing results.
<p>
<i>Anchor Clustering and Filtering:</i><br>
The raw anchor set consists of the hits found by MUMmer, which are filtered and clustered for input to the synteny algorithm.
<p><u>Algorithm 1 (modified original)</u> is good for medium-to-high divergent genomes, aligning draft sequence, self synteny,
and genomes with little or no annotation.
The MUMmer hits are first clustered into gene, or putative-gene hits. This is
done by clustering the hit regions on each sequence, and then defining new "gene"
hits which connect these regions. For example if three separate
exons hit between two genes, they will be clustered into one "gene"
hit having a combined score equal to the sum of the raw hit scores.
Clustering is by gene if the hits overlap annotation, otherwise, it creates
"candidate genes" from hits that do no overlap annotation.
<p>The clustered "gene anchors" are then filtered using a version of
reciprocal-best filtering which is adapted for retaining duplications and
gene families. For each pair of genes (or putative genes) which is
connected by a clustered anchor, the retained anchors must be among
 the top two anchors by score on both sides (top-2 allows for one
ancestral whole-genome duplication). An anchor will also be retained if its
score is at least 80% of that of the 2nd-best anchor on each side (this
allows for retention of gene family anchors). These filter parameters
may be adjusted through the <ttl>Alignment&amp;Synteny</ttl> <a href="SystemHelp.html#over2" class="ext" target="_blank">Parameters</a> panel.

<p><u>Algorithm 2 (exon-intron)</u> is good for low-to-high divergent genomes with good annotation.
It directly maps hits to the exons and introns. Hits aligning to un-annotated regions are clustered separately.
There are many more parameters for
this approach, as the hits are filtered based on the parameter values.
<br>

<p>
<i>Synteny Block Detection:</i><br>
After the clustered anchors are loaded into the database,  the synteny
synteny block algorithm runs. This algorithm looks for approximately-collinear
sequences of anchors, subject to several parameters including (A) Number
of anchors; (B) Collinearity of the anchors; (C) Amount of "noise" in the
surrounding region (to help reject false-positive chains). Criterion A can
be adjusted in the <ttl>Alignment&amp;Synteny</ttl> <ttb>Parameters</ttb> panel.
<p>
<sup>*</sup> Note that the sequences are re-written from the database to the
disk for three reasons: (A) To allow re-grouping for efficiency; (B) To ensure elimination
of invalid characters; (C) To mask non-gene regions, if desired. This also ensures that
sequences names will match those in the database, and prevents problems caused by
moving the source sequences on disk.
</div>

<!------------------------------------------------------------------------->
<a id="fpc"></a>
<h3>FPC project</h3>

For working with FPC<sup>8,9</sup>, it is suggested you use release v5.0.8 from
<a href="http://www.agcol.arizona.edu/software/symap">AGCoL</a>.

<ol>
<li>It has the FPC demo files.
<li>It has BLAT<sup>10</sup>  in the <tt>/ext</tt> directory.
<li>It has the tar file <tt>doc.tar.gz</tt> of the documentation.
<li>The AGCoL documentation applies to this release. See AGCoL
<a href="http://www.agcol.arizona.edu/software/symap/doc/SystemGuide.html#fpc" class="ext" target="_blank">System Guide</a>.
<li>This is the last release made from AGCoL, so the documentation will stay consistent.
<li>It is also available from <a href="https://github.com/csoderlund/SyMAP/releases"  class="ext" target="_blank">Github</a>.
</ol>
If you run into any problems, please do not hesitate to contact symap@agcol.arizona.edu.

<!---------------------------------------------------------->
<!---------------------------------------------------------->
<!----------------------------------------------------------------------->
<a id="refs"></a>
<h3>References</h3>

<sup>1</sup>
Kurtz, S., Phillippy, A., Delcher, A.L., Smoot, M., Shumway, M., Antonescu, C., Salzberg, S.L.
(2004). Versatile and open software for comparing large genomes, Genome Biology, 5:R12
<p>
<sup>2</sup>
Marcais, G., A.L. Delcher, A.M. Phillippy, R. Coston, S.L. Salzberg, A. Zimin (2018).
MUMmer4: A fast and versatile genome alignment system,  PLoS computational biology, 14(1): e1005944.
<p>
<sup>3</sup>
Krzywinski, M., J. Schein, I. Birol, J. Connors, R. Gascoyne, D. Horsman, S. Jones, M. Marra (2009).
Circos: An information aesthetic for comparative genomics. Genome Research doi:10.1101/gr.092759.109.
<p>
<sup>4</sup>
Soderlund, C., Nelson, W., Shoemaker, A., and Paterson, A.(2006).
SyMAP: A system for discovering and viewing syntenic regions of FPC maps.
Genome Res. 16:1159-1168.
<p>
<sup>5</sup>
Soderlund, C., Bomhoff, M., and Nelson, W. (2011).
SyMAP: A turnkey synteny system with application to multiple large duplicated plant sequenced genomes.
Nucleic Acids Res V39, issue 10, e68.
<p>
<sup>6</sup>
Katoh, Standley (2013).
MAFFT multiple sequence alignment software version 7: improvements in performance and usability.
Molecular Biology and Evolution 30:772-780.
<p>
<sup>7</sup>
Edgar, R (2004).
MUSCLE: a multiple sequence alignment method with reduced time and space complexity.
BMC Bioinformatics 113.
<p>
<sup>8</sup>
Soderlund, C., S. Humphrey, A. Dunhum, and L. French (2000).
Contigs built with fingerprints, markers and FPC V4.7. Genome Research 10:1772-1787.
<p>
<sup>9</sup>
Engler, F., J. Hatfield, W. Nelson, and C. Soderlund (2003).
Locating sequence on FPC maps and selecting a minimal tiling path. Genome Research 13:2152:2163.
<p>
<sup>10</sup> Kent, J. (2002) BLAT--the BLAST-like alignment tool, Genome Research 12:656-64.

<!---- END BODY -->
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p class="trailer">  Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></p>
</div>
</div>
</body>
</html>
