<!DOCTYPE html>
<html>
<head>
<title>SyMAP System Parameters</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="http://www.agcol.arizona.edu">
		<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;Parameters</p>
<p>
<a href="#build">Build database</a>
<a href="#projParams">Project Parameters</a>
<a href="#pairParams">Pair Parameters</a>
<p>&nbsp;
<a href="#top">Go to top</a>
</div>

<!------ right side header and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav"> <!-- header same in all pages -->
    <br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="./Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="./SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">
<p>
The original SyMAP was written for diverse plant genomes with short introns, but has been modified to work
for the long introns of mammalian genomes, and less diverse genomes.

<p>
<table class="tx center">
<tr><td><a href="#start"><u>Start SyMAP</u></a>
<tr><td style="vertical-align: text-top;">
1. <a href="#build"><u>Build database</u></a>
<ol type="a">
<li style="margin: 0.4em 0;"><a href="#selLoad">Selected</a>
<li style="margin: 0.4em 0;"><a href="#avail">Available Syntenies</a>
<li style="margin: 0.4em 0;"><a href="#fun">Function buttons</a>
<li style="margin: 0.4em 0;"><a href="#cpu">CPU and Verbose</a>
</ol>

<td style="vertical-align: text-top;">
2. <a href="#projParams"><u>Project Parameters</u></a> 			<!-- in code -->
<ol type="a">
<li style="margin: 0.4em 0;"><a href="#start">Parameter panel</a>
<li style="margin: 0.4em 0;"><a href="#display">Display</a>
<li style="margin: 0.4em 0;"><a href="#project">Load project</a>
<li style="margin: 0.4em 0;"><a href="#anno">Load annotation</a>
<li style="margin: 0.4em 0;"><a href="#annot">GFF Attributes</a>
<li style="margin: 0.4em 0;"><a href="#save1">Save project parameters</a>

</ol>
<td style="vertical-align: text-top;">
3. <a href="#pairParams"><u>Pair Parameters</u></a>  			<!-- in code pre-v546 -->
<ol type="a">
<li style="margin: 0.4em 0;"><a href="#over2">Parameter panel</a>
<li style="margin: 0.4em 0;"><a href="#align2">Alignment</a>		<!-- in code v546 -->
<li style="margin: 0.4em 0;"><a href="#clust">Cluster Hits</a>		<!-- in code v546 -->
<li style="margin: 0.4em 0;"><a href="#syn">Synteny</a>				<!-- in code v546 -->
<li style="margin: 0.4em 0;"><a href="#save2">Save pair parameters</a>
<li style="margin: 0.4em 0;"><a href="#mum">MUMmer parameters</a>
</ol>
</table>
This document provide details of parameters and functions.

<h2>Start SyMAP</h2>
<table class="tx"><tr>
<td style="vertical-align: text-top;">To start SyMAP, type at the command line: <tt>./symap</tt>
<p>To view the command line options: <tt>./symap -h</tt>

<td>&nbsp;For the first time user of SyMAP, see:
<ul>
<li><a href="SystemGuide.html#machine" class="ext" target="_blank">System Guide</a> for setup and system requirements.
<li><a href="SystemGuide.html#demo" class="ext" target="_blank">Demo</a> to run the demo (do not skip this step!).
<li><a href="SystemGuide.html#create" class="ext" target="_blank">Create a new project</a>.
</ul>
</table>

<p><ttb>Project Manager</ttb>
<br><a href="img/symapPlants.png">
	<img src="img/symapPlants.png" alt="manager" style="border: 1px solid black; width: 580px"></a>

<!---------------------------------------------------------------------------->
<a id="build"></a> <!-- # ref in code -->
<h2>1. Build database</h2>

<table class="tz">
<tr>
<td><a href="#selLoad">Selected</a>
<td><a href="#avail">Available Synteny</a>
<td><a href="#fun">Function buttons</a>
<td class="colrt"><a href="#top">Go to top</a></td></tr>
</table>

<a id="selLoad"></a>

<h3>1.a Selected</h3>
The projects selected from the <ttb>Projects</ttb> in
the left panel are listed in the <ttb>Selected</ttb> section on the right.
The possible functions vary with the state, as listed below:

<table class="ty">
<tr><td colspan=2>&diams; <i>If there is any project not loaded to the database, you will see:</i>
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Load All Projects</ttb>
    <td style="white-space: nowrap;">Load all projects that have not been loaded yet.

<tr><td colspan=2>&diams; <i>If a project <b>is not</b> loaded to the database, you will see:</i>

<tr><td style="vertical-align: text-top;">&nbsp;&nbsp;
	<ttb>Remove from disk</ttb>
    <td><ttx>Only: Remove alignment directories from disk</ttx>
    <br><ttx>All: Remove alignment and project directory from disk</ttx>
    <p><ttx>Remove alignments</ttx> removes alignments from <tt>data/seq_results</tt> for this project. You will
    be prompted for each one to confirm you want it removed. If there are no alignments, you will only see the prompt
    to remove the project directory.
    <p><ttx>Remove from disk</ttx> remove  <tt>data/seq/&lt;project-name&gt;</tt> from disk.
    You will be prompted to confirm you want it removed.
    If removed, the project will no longer be shown on the left.

<tr><td style="white-space: nowrap;vertical-align: text-top;">&nbsp;&nbsp;
	<ttb>Load project</ttb>
    <td>Loads the sequence and optional annotations to the database. After loading results,
    always verify them by selecting the <ttb>View</ttb> link, which provides a summary of what has been loaded.

<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Parameters</ttb>
    <td>This brings up a panel of parameters, see <a href="#projParams">Project Parameters</a>.
	After the project is loaded, you can still change the <ttl>Display</ttl> parameters.

<tr><td colspan=2>&diams; <i>If a project <b>is</b> loaded into the database, you will see:</i>
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Remove from database</ttb>
    <td>The projects and its synteny pairs will be removed from the database, but the files stay on disk.

<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Reload project</ttb>
    <td><ttx>Only: reload project only.</ttx>
    <br><ttx>All: reload project and remove alignments from disk.</ttx>
	<p>If <ttl>All</ttl> is selected, it first prompts for each alignment directory for this project before it is removed.
		It removes the alignments, but it leaves the <tt>params.txt</tt> file. You will need to remove the
		alignment(s) if (1) there is a change in sequence, or (2) there is a change in the <ttl>Minimal length</ttl> parameter;
		see <a href="#project">Load project</a> parameters.
		<p>For either option, it executes <ttl>Remove from database</ttl> followed by <ttl>Load project</ttl>.

<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Reload annotation</ttb>
    <td>Removes the annotation from the database then load the annotations.

	<p>This does not effect the alignment, so they do not have to be redone.
	The <ttl>Align&amp;Synteny</ttl> commands will recognize if there is an existing alignment
		and will only perform the synteny computation.

<tr><td style="white-space: nowrap;">&nbsp;&nbsp;
	<ttb>Parameters</ttb>
    <td>This brings up the <a href="#projParams">Project Parameters</a> panel.
</table>
<p>For any action that will remove the project or alignments from disk, a popup will occur
to confirm that you want this done! If it will be removing multiple alignment directories, it will prompt
on each one.

<a id="avail"></a>
<h3>1.b Available Syntenies</h3>

Sequence alignments are performed with
<a href="http://mummer.sourceforge.net/manual" class="ext" target="_blank">MUMmer3</a>,
but can be changed to use
<a href="https://mummer4.github.io"  class="ext" target="_blank">MUMmer4</a> (see SyMAP
<a href="./MUMmer.html" class="ext" target="_blank">MUMmer</a>).

<p>This section shows a table with the status of alignments between
the selected <i>loaded</i> projects.
Each cell in the table represents a pair of projects and the cell contains a status code showing
 whether or not that pair has been aligned (codes are listed below). Note that the table
shows each pair cell twice, but only the lower cells are activated.
<p>Clicking on a cell selects that pair of projects (the cell will be highlighted in green), and
the buttons that can be selected are activated.
<p>
<table class="ty">
<tr><td><i>Code</i>
	<td><i>Description</i>
<tr><td>
		&#10004;
	<td>Synteny for this project pair is ready to view.
<tr><td><b>A</b>
	<td>The MUMmer alignment has been performed but the synteny computation has not been run.
		This status occurs if a pair is completed but then annotations are re-loaded for one of the projects,
		or if the MUMmer files have been added by the user.
<tr><td><b>?</b>
	<td>The alignment have not been completed. In this case, select <ttl>Selected Redo</ttl> and the
		alignments will be completed followed by the Cluster&amp;synteny algorithm.
<tr><td>&nbsp;
	<td>The alignment has not been started.
</table>
<p>See <a href="#over2">Pair Parameters</a> for additional information on the <ttb>Available Syntenies</ttb> and codes.

<a id="fun"></a>
<h3>1.c Function Buttons</h3>

<table class="ty">
<tr><td colspan=2><i>Align&amp;Synteny (A&amp;S)</i> Synteny usually implies <i>Cluster&amp;Synteny</i>
as the process has 2 distinct algorithms.
<tr><td style="white-space: nowrap;vertical-align: text-top;">
		<ttb>All Pairs</ttb>
	<td>
		Run (or complete) the synteny computation for all pairs in the <ttb>Available Syntenies</ttb> table.
		However, this will not run 'self-syntenies'; those need to be done individually by selecting
		the pair.
<tr><td style="white-space: nowrap;vertical-align: text-top;">
		<ttb>Selected Pair</ttb>
		<br><ttb>Selected Redo</ttb>
	<td>Run (or complete) the <ttl>A&amp;S</ttl> computation for the selected project pair.
		<br>If the pair is already complete, the button label changes to <ttb>Redo</ttb>, and
		only the <ttx>Cluster&amp;Synteny</ttx> will be rerun (see <a href="#drop">Pair Parameters</a> for some
		variations).
		<br>If you wish to rerun the MUMmer alignment, first use the <ttb>Clear Pair</ttb> function.
<tr><td style="white-space: nowrap;vertical-align: text-top;">
		<ttb>Clear Pair</ttb>
	<td><ttx>Only: remove synteny from database</ttx>
	<br><ttx>All: remove synteny and alignments from disk for this pair</ttx>
	<p>If you have changed the <ttl>Alignment</ttl> parameters, or loaded new sequence for one
	of the projects, you need to have the alignments removed and redone; otherwise, you can just remove the synteny
	from database (where this step can be skip by selecting the <ttb>Selected Pair (Redo)</ttb>).
<tr><td style="white-space: nowrap;vertical-align: text-top;">
		<ttb>Parameters</ttb>
	<td>Set the <a href="#pairParams">pair parameters</a> for the selected pair cell .
</table>

<p>For the remaining display buttons, see <a href="UserGuide.html#views" class="ext" target="_blank">User Guide</a>.

<a id="cpu"></a>
<h3>1.d CPU and Verbose</h3>
<ttl>CPUs:</ttl> Enter the maximum number of CPUs to use for an alignment. SyMAP will use
up to that number (but may use fewer, depending on the number of sequences being aligned). Alternatively,
the number of CPUs may be entered in the <tt>symap.config</tt> file or
entered as a command line argument (-p N).

	<p><ttl>Verbose</ttl> checkbox:
	<ul><li>If checked, detailed summary information is written as
	it processes the MUMmer files. The information is written both to the terminal and the <tt>logs/&lt;proj-to-proj&gt;/symap.log</tt> file.
	<li>If this is not checked, it will write status information repeatedly on the same terminal line.
	<li>See <a href="Demo.html#slog" class="ext" target="_blank">Demo examples</a>.
	<li>It is not saved as a parameter. This can also be turned on using a command line argument:
	<tt>symap -v</tt>
	</ul>


<hr>
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<a id="projParams"></a> <!-- # ref in code -->
<h2>2. Project Parameters</h2>

<table class="tz">
<tr>
<td><a href="#start">Parameter<br>panel</a>&nbsp;&nbsp;&nbsp;
<td><a href="#display"><i>Display</i></a>
<td><a href="#project"><i>Load<br>project</i></a>
<td><a href="#anno"><i>Load<br>annotation</i></a>&nbsp;&nbsp;&nbsp;
<td><a href="#annot">GFF<br>Attributes</a>
<td><a href="#save1">Save</a>
<td class="colrt"><a href="#top">Go to top</a></td></tr>
</table>

<a id="start"></a>
<h3>2.a Parameter panel</h3>
<table>
<tr><td style="vertical-align: text-top;">
Click the <u>Parameters</u> link for a project to open the parameters panel shown on the lower right.

<p><i>Make sure these two parameters are correct before running the alignment</i>: <ttl>Minimum length</ttl>, <ttl>Sequence files</ttl>.
See <a href="#project">Load project</a>.

<p>&nbsp;
<p><ttl>Sequence</ttl> and <ttl>Anno</ttl> (annotation) <ttl>files</ttl>:
<br>If the files are put in the default locations, nothing needs to be entered for them.
Default locations:
<pre>
  data/seq/&lt;project-name&gt/sequence
  data/seq/&lt;project-name&gt/annotation
</pre>
See <a href="./input/index.html" class="ext" target="_blank">input</a> for a description of the input files.
<td>
<a href="img/parameters1.png"><img src="img/parameters1.png" alt="parameters1" style="border: 1px solid black; width: 430px"></a>
</table>

<!--------------------------------------------------------------->
<a id="display"></a>
<h3>2.b Display</h3>

<table>
<tr><td>New values take immediate effect on <ttb>Save</ttb>. Most of the values are shown in the <ttl>Selected</ttl>
section, as shown on the right. The
values are saved in the <tt>symap_5/data/seq/&lt;project-name&gt;/params.txt</tt> file.
<td><a href="img/lineLoad.png"><img src="img/lineLoad.png" alt="Project options" style="border: 1px solid black; width: 430px"></a>
</table>
<p>
<table class="ty">
<tr><td><b>Parameter</b><td><b>Description</b><td><b>Default Value</b><td><b>Shown</b>
<tr><td>&nbsp;&nbsp;<ttl>Category</ttl>
	<td style="width: 450px;">
		Category label for the project. This is only used to group projects on the
		left side of the <ttb>Manager</ttb> panel.
		<p>Category labels must be composed of only letters, numbers, dash, underscore, or period.

		<p>Either select an existing label from the drop-down or enter a new one in the text box.
		Do NOT enter the same label with different capitalization -- it may mess-up.
	<td>Uncategorized
	<td><ttl>Selected</ttl>
<tr><td>&nbsp;&nbsp;<ttl>Display name</ttl>
	<td>A user-friendly name for the project.
	<p>Shorter names will work better in the displays.
	Names must be composed of only letters, numbers, dash, underscore, period.
	<p><i>It must be unique over all case-insensitive <ttl>Display names</ttl>
	and <ttl>project-names</ttl></i>.

	<td><ttl>project-name</ttl>
	<td><ttl>Selected</ttl> and all displays.

<tr><td>&nbsp;&nbsp;<ttl>Abbreviation</ttl>
	<td>A name this is exactly 4 characters.
	Names must be composed of only letters, numbers, dash, underscore, period.

	<p>Uniqueness is not required over other <ttl>Abbreviation</ttl>.
	It can be the same as the corresponding <ttl>Display name</ttl>
	or <ttl>project-name</ttl> if they are only 4 characters.
	<td>Last 4 characters of <ttl>Display name</ttl>.
	<td><ttb>Queries</ttb> column headings, and other places where a guaranteed short name is needed.

<tr><td>&nbsp;&nbsp;<ttl>Description</ttl>
	<td>Description of the project.
	<p>Do NOT use quotes, backslash or #.
	<td>New project
	<td><ttl>Selected</ttl> and <ttb>View</ttb>

<tr><td>&nbsp;&nbsp;<ttl>Group type</ttl>
	<td>How to refer to the sequences.
	<td>Chromosome
	<td><ttl>Selected</ttl>
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Anno key count</ttl>
	<td>This applies to the annotation attributes columns shown in the <ttb>Queries</ttb> results table. See the
		<a href="#annot">GFF Attributes</a> section below.
	<td>50
	<td>--
</table>

<a id="project"></a>
<h3>2.c Load project</h3>
The following parameters are under the <ttb>Load project</ttb> section of the <a href="#start">parameters</a> panel.

<p><ttl>Group prefix</ttl>
<br>The term "Group" is used for any FASTA sequence type, e.g. chromosome, scaffold, contig.
This option sounds trivial but is important for a good display, so please read carefully the following.

<table class="tx">
<tr><td style="vertical-align: text-top;">
	<table>
	<tr><td style="vertical-align: text-top;">1. <td>When a <ttl>Group prefix</ttl> is entered, it allows SyMAP to remove the prefix from the chromosome
		names and use the remaining part as a shorter name (e.g. '1' instead of 'chr1', as shown on the lower right).
	<tr><td style="vertical-align: text-top;">2.<td>If the sequence file has a mix of prefixes (e.g. Chr01, Scaf2345):
	<ul>
		<li>If a <ttl>Group prefix</ttl> is entered, only sequences with that prefix will be loaded.
		<li>Leave the <ttl>Group prefix</ttl> blank will load all sequences; their prefix will not be removed
		(e.g. the 'chr3' in the image on the right).
	</ul>
	<tr><td style="vertical-align: text-top;">3.<td><i>Long names really clutter the display</i>; it is best to use a consistent prefix that SyMAP can remove,
	otherwise, use really short prefixes, e.g. 's' for scaffold, 'c' for chromosome.
	<tr><td style="vertical-align: text-top;">4.
	<td><i><u>You may remove the prefix after the project is loaded</u></i>. For example, if your sequences had
		names "Scaf1", "Scaf2", etc, and "Scaf" was NOT entered as the <ttl>Group prefix</ttl> before load, you may
		enter it later and it will be removed from all sequence names. However, <i><u>this is not reversible</u></i>
		and a prefix cannot be added to the sequence names.
	</table>
<td><a href="img/demoPrefix.png"><img src="img/demoPrefix.png" alt="Demo prefix" style="border: 1px solid black; width: 80px"></a>
</table>

This parameter is finicky; after loading a project, select <ttb>View</ttb> for a popup of the input,
and check the output to the terminal to make sure the annotation was loaded right. Also, see
<a href="./input/index.html#convert" class="ext" target="_blank">xToSymap</a>
as it may help you create the files with good prefixes.

<p><ttl>Minimum length</ttl>
<br>This must be an integer, commas are allowed (e.g. 1,000,000).
<br>This is the minimum length of the FASTA sequence that will be loaded; smaller sequences will be ignored. Note that annotations
for ignored sequences will also be ignored, but some warning messages will be printed to the terminal.
See <a href="input/index.html#ol" class="ext" target="_blank">xToSymap Length</a> for help with setting this parameter.

<p><ttl>Sequence files</ttl>
<br>
Select the input FASTA sequence file(s), or directories of sequence files.
For formatting, see <a href="./input/index.html#seqprep" class="ext" target="_blank">Sequence files</a>.
Default location: <tt>data/seq/&lt;project-name&gt/sequence</tt>

<p>If any either the sequence files or <ttl>Minimal length</ttl> parameters are changed:
<br>&#9656; If the project has already been loaded, <ttb>Reload Project</ttb>.
<br>&#9656; If <ttl>A&amp;S</ttl> has previously been run,
select <ttb>Clear Pair</ttb> and remove the alignment files, then run <ttl>A&amp;S</ttl>.

<!----------------------------------------------->
<a id="anno"></a>
<h3>2.d Load annotation</h3>
The following parameters are under the <ttb>Load annotation</ttb> section of the <a href="#start">parameters</a> panel.

<p><ttl>Anno keywords</ttl>
<br>A comma separated list of keywords. This can be used to reduce the annotation attribute keywords
shown in the <ttl>2D display</ttl> and <ttl>Queries</ttl> table.
	See the <a href="#annot">GFF Attributes</a> section below.
<p>
<ttl>Anno files</ttl>
<br>
Select the input GFF3-formatted annotation files corresponding to your sequences. Note, using
a GFF3 file directly can cause problems if it does not conform to what SyMAP expects; see
<a href="input/index.html#annot" class="ext" target="_blank">Annotation files</a>.
Annotation is optional but highly recommended. Default location: <tt>data/seq/&lt;project-name&gt/annotation</tt>

<p>If either of the above are changed:
<br>&#9656; If the annotation has already been loaded, <ttb>Reload Annotation</ttb>.
<br>&#9656; If <ttl>A&amp;S</ttl> has been previously run, re-run <ttl>A&amp;S</ttl> (the existing alignments files will be reused).

<!----------------------------------------------->
<a id="annot"></a>
<h3>2.e GFF Attributes</h3>
This section gives details on what GFF <u>attributes</u> are displayed in SyMAP,
which refers to them as <u>annotations</u>.
<p>The gene annotation is shown on the
<a href="UserGuide.html#view2D" class="ext" target="_blank">2D display</a> and as columns in the
<a href="Query.html#col" class="ext" target="_blank">Queries</a> results table. The attributes (annotations)
comes from the last column of the GFF file.
The attributes are a keyword=value list, e.g.
<p>
<pre>
   ID=gene-AT1G01010;Name=NAC001;ID=rna-NM_099983.2;product=NAC domain containing protein 1
</pre>
Defaults: Generally, all genes in a file have the same keywords, in which case, use the defaults.
This will cause the entire attribute to be shown for the gene in the <ttl>2D display</ttl>, and the
<ttl>Queries</ttl> table will have columns for each keyword that has over 50 occurrences.
In the example above, the columns will be ID, Name and product (the second ID will be ignored).
<p>
If there are many different keywords in the attribute list, this causes too many columns in the <ttl>Queries</ttl> table.
This can be reduced by one (or both) of the following:
<p>
<ttl>Anno key count</ttl>: If there are many different keywords in the attribute list,
set this count N to filter out all keywords with &lt;N occurrences.
The <ttl>Anno key count</ttl> can be modified at any time using <ttp>symap</ttp> (not <ttp>viewSymap</ttp>).
<p>
<ttl>Anno keywords</ttl>: The <tt>keyword=value</tt> pairs to be saved for each gene can be limited by listing
the desired keywords separated by commas.  Using this approach, it will also reduce
annotation description per gene in the <ttl>2D display</ttl>.
Referring to the example above, if the string "<tt>ID, product</tt>" was entered,
the <tt>Name=value</tt> would not be part of any gene annotation. This must be set before <ttb>Load Annotations</ttb>
is executed.
<!----------------------------------------------->
<a id="save1"></a>
<h3>2.f Saving project parameters</h3>

Before the project is loaded, the parameters are saved to the file
<pre>  data/seq/&lt;project-name&gt/params.txt</pre>
The <tt>params.txt</tt> file parameters are shown on the <ttl>Project Parameters</ttl> panel. These can only be
viewed and changed using <ttp>symap</ttp> (not <ttp>viewSymap</ttp>).

<p>When the project is loaded, all parameters are saved to the database except the <ttl>A&amp;S</ttl> parameters
(<ttl>Mask-genes</ttl> and <ttl>Order-against</ttl>).
When the <ttl>A&amp;S</ttl> is executed, then the corresponding parameters are saved to the database.

<p>Any parameter that does not have the default value will be shown on the <ttb>View</ttb> popup.

<hr>
<!----------------------------------------------------------------------->
<!----------------------------------------------------------------------->
<!----------------------------------------------------------------------->
<a id="pairParams"></a> <!-- # ref in code -->
<h2>3. Pair Parameters</h2>

<table class="tz">
<tr>
<td><a href="#over2">Parameter<br>panel</a>&nbsp;&nbsp;&nbsp;
<td><a href="#align2"><i>Alignment</i></a>
<td><a href="#clust"><i>Cluster<br>Hits</i></a>
<td><a href="#syn"><i>Synteny</i></a>&nbsp;&nbsp;&nbsp;
<td><a href="#save2">Save</a>
<td><a href="#mum">MUMmer<br>parameters</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<a id="over2"></a>
<h3>3.a Parameter panel</h3>
The <a href="#avail">Available Syntenies</a> section at the beginning explains the table in the lower right. The following provides
more information in the context of the pair parameters.
<table>
<tr>
	<td>
	The table on the right has cells that have the following completed:
	<p><table class="ty">
	<tr><td><i>Value</i><td><i>Alignment</i><td><i>Synteny</i>
	<tr><td>blank<td>No<td>No
	<tr><td><b>A</b><td>Yes<td>No
	<tr><td>&#10004;<td>Yes<td>Yes
	</table>

	<td><a href="img/demoTableA.png"><img src="img/demoTableA.png" alt="demo table with A" style="border: 1px solid black; width: 550px"></a>
</table>

<table>
<tr>
	<td style="vertical-align: text-top;">
	<p>&nbsp;<p>Alignment will not be redone if the cell contains an <b>A</b>. This is important because MUMmer
is very time-consuming, but the synteny computation is not
(see <a href="SystemGuide.html#time" class="ext" target="_blank">timing results</a>); hence,
one can make changes to the cluster or synteny parameters and re-run without redoing the alignments.
<p>
Select a pair cell in the <ttb>Available Syntenies</ttb> table followed by the <ttl>Parameters</ttl>
button, which will popup the panel shown on the right.
	<p>If a &#10004; or <b>A</b> in pair cell exists and the parameters for a section
	are changed, do the following:

	<table class="ty">
		<tr><td style="white-space: nowrap;"><i>Changed Section</i><td><i>Action</i>
		<tr><td><b><i>Alignment</i></b> <td>Select <ttb>Clear Pair</ttb> to remove the existing alignments,
		then <ttb>Selected Pair</ttb>.
		<tr><td><b><i>Cluster Hits</i></b> <td>Use <ttb>Selected Redo</ttb>
		directly; the existing alignments for the pair will be used.
		<tr><td><b><i>Synteny</i></b> <td>If only synteny parameters are changed,
		click the lower-left drop-down and select <ttx>Synteny Only</ttx>
		(see <a href="#drop">drop-down</a>).
	</table>
	<p>&nbsp;<p>The parameters are described in the following 3 sections: <a href="#align2">Alignment</a>,
	<a href="#clust">Cluster hits</a> and <a href="#syn">Synteny</a>.

	<td>&nbsp;&nbsp;&nbsp;
	<a href="img/parameters2.png"><img src="img/parameters2.png" alt="pair parameters" style="border: 1px solid black; width: 340px"></a>
</table>

<a id="align"></a>
<a id="align2"></a>
<h3>3.b Alignment</h3>

<a id="mask"></a>
<a id="concat"></a> <!--- this is referenced elsewhere -->
<h4>Preparing the sequences</h4>
<table class="ty">
<tr><td><i>Parameter</i><td><i>Description</i><td><i>Default</i>
<tr><td><ttl>Concat</ttl>
<td>By default, all sequences of the 1st genome are concatenated into one file
and then multiple files are created from the 2nd genome to be searched against the first.
For the 2nd genome, sequential short sequences are put into one file until the file length is &gt;60M.

<p><ttl>Concat</ttl> unchecked: To reduce memory usage, you can uncheck <ttl>Concat</ttl> so that
multiple files are created for each genomes,
and all files from the 1st genome are searched against the all files from the 2nd genome. For both genomes,
sequential short sequences are put into one file until the file length is &gt;60M.
<p>See below for timing differences.
<td>On

<tr><td style="white-space: nowrap;"><ttl>Mask</ttl> &lt;abbrev&gt;<td>Mask out all non-genic parts of the sequences before running MUMmer (gene annotation must
be provided).
<p>The &lt;abbrev&gt; is set in the <ttb>Project parameters</ttb> popup <ttl>Abbreviation</ttl> parameter,
and the corresponding project's sequence is masked.

<p>If <ttl>Mask</ttl> is changed after <ttl>A&amp;S</ttl>, the alignment files need to be
removed with <ttb>Clear Pair</ttb> and <ttl>A&amp;S</ttl> run again.
<td>Off
</table>

<p><ttl>Concat</ttl>: The following statistics are from comparing Arabidopsis thaliana (119M) against Brassica rapa (297M)
on a MacOS using 1 CPU.
<table>
<tr><td><u>Concatenated</u>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><u>Not concatenated</u>
<tr><td>
<pre>
48819 hits
  334 synteny blocks
46319 gene hits
38334 synteny hits

Finished in 1 hour 8 minutes
</pre>
<td>
<td>
<pre>
48846 hits
  334 synteny blocks
46348 gene hits
38345 synteny hits

Finished in 1 hour 35 minutes
</pre>
</table>

<h4>MUMmer parameters</h4>
The default MUMmer parameter seems to work fine with SyMAP, so probably do not need changing.
<table class="ty">
<tr>
<td><i>Parameter</i><td><i>Description</i><td><i>Default</i>
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>PROmer Args</ttl><sup>1</sup>
	<td>Arguments for PROmer. See <a href="#mum">MUMmer parameters</a>.
	<td>&nbsp;-
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>NUCmer Args</ttl><sup>1</sup>
	<td>Arguments for NUCmer. See <a href="#mum">MUMmer parameters</a>.
	<td>&nbsp;-
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Self Args</ttl><sup>2</sup>
	<td>Arguments to use when aligning a chromosome to itself.
	<td>&nbsp;-
<tr><td>&nbsp;&nbsp;<ttl>PROmer Only</ttl><sup>3</sup>
	<td>Use PROmer for all alignments.
	<td>Off
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>NUCmer Only</ttl><sup>3</sup>
	<td>Use NUCmer for all alignments.
	<td>Off
</table>
<br><sup>1</sup>
<small><font color="red">BEWARE</font>: Entered PROmer and NUCmer arguments are <b>NOT</b> checked for correctness.
</small>
<br><sup>2</sup><small>When self-alignment is performed, standard arguments are used
when comparing different chromosomes. However, additional arguments may be desired
when a chromosome sequence is run against itself, e.g. <tt>--nosimplify</tt>.
</small>
<br><sup>3</sup><small> By default, PROmer is used for alignments between different
projects, while NUCmer is used for self alignments.
</small>

<!------------------------------------------------------------------------>
<!------------------------------------------------------------------------>
<a id="clust"></a>
<h3>3.c Cluster Hits</h3>

<table class="tz">
<tr>
<td><a href="#a1a2">Algo1 vs Algo2 with hints</a>
<td><a href="#clParam">Parameter description</a>
<td><a href="#pseudo">Pseudo and Piles</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<a id="a1a2"></a>
<h4>3.c.I  Algo1 vs Algo2</h4>

<table class="ty">
<tr><td colspan="2"><ttl>Algorithm 1 (modified original, updated v5.4.0, abbreviated Algo1)</ttl>:
<tr><td>Pros<td>This is an generic algorithm that has knowledge of genes versus intergenic hits.
<br>It is recommended for ordering sequence contigs and when there are little or no gene annotation.
<br>It <u>must</u> be used for self-synteny.
<br>It has been used on 100's of genome comparisons.
<tr><td>Cons<td>It does not distinguish between exon and intron hits. It is more likely to miss good homologous gene pairs.
<tr><td>Parameters<td>It only has one parameter, which is easier to run but there is no control over what hits are filtered.

<tr><td colspan="2"><ttl>Algorithm 2 (exon-intron, last update v5.6.0, abbreviated Algo2)</ttl>:
<tr><td>Pros<td>This is a new algorithm with explicit knowledge of gene pairs and their exon-intron structure.
<br>When there is good gene annotations for both genomes, this is definitely the superior algorithm.

<tr><td>Cons<td>It does not perform self-synteny.
<br>It does not work when a given chromosome is split over multiple MUMmer files; this will NOT happen
<br>when SyMAP generates the MUMmer files. This is also a "Pro" as it takes less memory.

<tr><td>Parameters<td>It has two set of parameters, hence, more control over results than <ttl>Algo1</ttl>. See <a href="#hint">Hints</a>
below the parameter explanation. (As of v5.6.0, the parameters generally do not need adjusting.)
</table>
<p><ttl>Algo1</ttl> is the default since it works for all inputs.

<p><b>Wrong strand</b>
The wrong strand is when all hits in a cluster are to the same strand (++/--)
yet the cluster aligns to two genes on the different strands (+-/-+), or vice versa.

<p><tt>Algo1</tt> does include these hits. You can view them in the <ttp>Queries</ttp>
where the <ttx>Hit St</ttx> column will be different than the two gene <ttx>Gst</ttx> columns.

<p><tt>Algo2</tt> does NOT include these hits. You can request to view the potential hits
during the <ttl>A&amp;S</ttl> by running it with the "-wsp" flag, i.e.
<tt>   ./symap -wsp </tt>.
This will only show gene pairs with (1) multiple hits to exons (in one or multiple gene pairs), (2) at least one is not
an overlapping gene. It is up to the user to determine what is real.

<a id="hint"></a>
<p><b>Hints about parameter settings</b>

<p><u>Hint for <ttx>Algo1</ttx></u>: Increasing the <ttl>Top N</ttl> parameter can cause too many hits and reduce synteny.
Decreasing it can remove more gene-pair hits.
Hence, try <ttl>Algo2</ttl> if you want more gene pairs.

<p><u>Hint for <ttx>Algo2</ttx></u>: On the output to the terminal (in Verbose mode), if any chromosome pair shows over 10,000 hits, the
parameters probably need to be made more stringent; too many hits confuses the synteny algorithm, which results
in synteny blocks not being found; it also results in very long execution time.
<p>&rarr;&nbsp;<i>Suggestion</i>: For large genomes, experiment with the parameters on just one pair of the chromosomes.
(You can use <a href="input/index.html" class="ext" target="_blank">xToSymap</a> for the split.)
<p>&rarr;&nbsp;<i>Using the v5.6.0 algorithm, I have experimented with the datasets</i> (1) human, chimpanzee, mouse (2)
Arabidopsis, Brassica rapa, Brassica oleracea.
Only B. rapa to B. oleracea needed parameter adjustment: the number of G1 hits was over 200k,
which is way more than typical;
by increasing all parameters a small amount, this reduced to just over 100k.

<!-------------------------------------------------------------------------->
<a id="clParam"></a>
<h4>3.c.II  Parameter description</h4>

<table class="ty">
<tr><td><i>Parameter</i><td><i>Description</i><td><i>Default</i>
<tr><td style="white-space: nowrap;"><ttl>Number Pseudo</ttl><td>If selected, the un-annotated ends of hits will be assigned a
pseudo number. This is explained below in <a href="#pseudo">Pseudo genes</a>.
<td>No
<tr><td colspan=2><ttl>Algo1 (original)</ttl>
    <td>Yes
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Top N piles</ttl>
	<td>It will retain the top N hits of a pile of overlapping hits (<a href="#pileOfHits">Pile of Hits</a>), as
		well as all hits with score at least 80% of the Nth hit.
	<td>2
<tr><td colspan=2><ttl>Algo2 (gene-centric)</ttl>
    <td>No
<tr><td>&nbsp;&nbsp;<ttl>Scale</ttl>
	<td>
	Increase a scale to filter out more clustered hits, decrease to filter out less.
	<br><center><table class="ty">
		<tr><td><ttl>Gene</ttl>
			<td>Determines the percentage of required gene coverage.
		<tr><td><ttl>Exon</ttl>
			<td>Determines the percentage of required exon coverage.
		<tr><td><ttl>Len</ttl>
			<td>For G2 and G1<sup>*</sup>, minimum size hit unless it completely covers a gene.
			<br>Default minimum is 300bp (N*300bp is minimum).
		<tr><td><ttl>G0_Len</ttl>
			<td>For G0<sup>*</sup>, minimum size hit.
			<br>Default minimum is 1000bp (N*1000bp is minimum).
			<br>Suggestion: increase for closely related species.
		<tr><td colspan="2"><sup>*</sup><small>G2=gene to gene, G1=gene to non-gene, G0=non-gene to non-gene</small>
		</table></center>

	<i>A clustered hit must pass multiple rules regarding coverage</i>.
	<br>For example, the <ttl>Exon</ttl>
	coverage parameter may be made more stringent, yet there will still be some low exon scores due to
	good gene coverage, etc.
	<td>N=1.0
<tr><td>&nbsp;&nbsp;<ttl>Keep piles</ttl>
	<td><ttl>EE</ttl>, <ttl>EI</ttl>, <ttl>En</ttl>, <ttl>II</ttl>, <ttl>In</ttl> (E=exon, I=intron, n=non-gene)
		<br>&bullet;&nbsp;This ONLY applies when there is a pile of overlapping hits (<a href="#pileOfHits">Pile of Hits</a>);
		<br>&nbsp;&nbsp;it tells the algorithm what type of cluster hits to retain if they are in a pile.
		<br>&bullet;&nbsp;Hits are filtered before pile analysis.
		<br>&bullet;&nbsp;Intergenic-intergenic pile hits, and any unchecked categories, are filtered
		<br>&nbsp;&nbsp;as described in the next row.
	<td><ttl>EE, EI, En</ttl>
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Top N piles</ttl>
	<td><ttx>Algo2</ttx> uses <ttx>Algo1</ttx> <ttl>Top N</ttl> parameter
for any uncheck categories, but in a more conservative way. It will retain the <ttl>Top N</ttl> hits of a piled region that are within
80% of the length of the longest hit.
	<td>2
</table>

<a id="pseudo"></a>
<h4>3.c.III Pseudos and Piles</h4>

<b>Pseudo genes</b>
<table><tr><td>
The end of a hit may not overlap an annotated gene; by default, this
will just show a <ttx>Gene#</ttx> of 'N.~' where N is the chromosome number.
<p>If <ttl>Number Pseudo</ttl> is selected,
a <i>pseudo</i> <ttx>Gene#</ttx> will be created. The counts start after the annotated
gene numbers and are suffixed by "~". For example, if the last <ttx>Gene#</ttx> for
Chr03 is 5550 (e.g. 3.5550.), the first pseudo gene number will be 6000 (e.g. 3.6000.~).

<td><a href="img/query_pseudo.png"><img src="img/query_pseudo.png" alt="Pseudo genes" style="border: 1px solid black; width:  370px"></a>
</table>
<p>If <ttl>A&amp;S</ttl> was run without numbered pseudos, go to the <ttb>Pair Parameter</ttb> panel,
and select <ttx>Number Pseudo</ttx> in the lower-left <a href="#drop">drop-down</a>; only this algorithm
will be run. <i>This cannot be undone</i>;
you would need to re-run <ttl>A&amp;S</ttl> with the <ttb>Number Pseudo</ttb> unchecked to remove them.
<p>
<table class="ty">
<tr><td>Pros
<td>
<ul>
<li>If you would like the <ttl>Queries</ttl>
<a href="Query.html#clhit" class="ext" target="_blank">Cluster</a>  and
<a href="Query.html#report" class="ext" target="_blank">Report</a>
to include un-annotated hits.
<li>If you are exploring new candidate genes, numbered pseudos are easier to track.
<li>If your genome is not annotated, numbered pseudos are easier to track.
</ul>
<tr><td>Cons
<td>
<ul>
<li>The <ttl>Queries</ttl> results can be easier to view with the
'N.~' as it is more distinct from a real <ttx>Gene#</ttx>.
</ul>
</table>
<br>If comparing more than 2 species, it makes the most sense to have them all numbered or not numbered (though a mix
will work).

<a id="pileOfHits"></a>
<p><b>Piles of Hits</b>

<table>
<tr><td>The below image shows a pile of hits on the left (Cabb Chr02)
that link to repetitive genes on the right (Arab Chr05). These are important to keep.
<p>The right image shows a pile of hits in an intergenic region (Cabbagge Chr03) to multiple other regions (B.rapa Chr01).
There are MANY occurrence of repeats like this in the MUMmer file, which is why these piles
must be filtered; if they are not, the synteny algorithm does not perform well.
<br><a href="img/demoPileG2.png"><img src="img/demoPileG2.png" alt="Pile to repetitive gene" style="border: 1px solid black; width:  400px"></a>
<td><a href="img/demoPileG0.png"><img src="img/demoPileG0.png" alt="Pile to intergenic" style="border: 1px solid black; width:  400px"></a>
</table>

<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<!---------------------------------------------------------------------->
<a id="syn"></a>
<h3>3.d Synteny</h3>

<table class="ty">
<tr><td><i>Parameter</i><td><i>Description</i><td><i>Default</i>
<tr><td>&nbsp;&nbsp;<ttl>Min Hits</ttl>
	<td>Minimum number of anchors required to define a synteny block.
	<td>7
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Strict</ttl>
	<td>This uses the <ttl>Original</ttl> algorithm with the following changes:
	<ul>
	<li>It used smaller gap sizes and stricter PCC cutoff.
	<li>It first computes the blocks based on orientation.
	<li>If <ttl>Orient</ttl> is not checked, it checks blocks
	and collinear sets to see if they are contained in another block, and if so, merges them regardless of orientation;
	this merge happens regardless of the <ttl>Merge</ttl> setting.
	</ul>
	<p>By default, the <ttl>Original</ttl> algorithm is run.
	However,  <ttl>Strict</ttl>  typically works better than the <ttl>Original</ttl>
	for complete genome sequence projects;
	see <a href="demo.html#synteny" class="ext" target="_blank">Synteny results</a>.
	It is not recommended for draft sequence!
	<td>Off
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Orient</ttl>
	<td>All hits in a block must have hits of the same orientation ('+/+' or '-/-') or
	different orientation ('+/-' or '-/+').
	<td>Off
<tr><td style="white-space: nowrap;">&nbsp;&nbsp;<ttl>Merge</ttl>
	<td><ttx>Overlap</ttx>: The blocks must overlap to be merged.
	<br><ttx>Close</ttx>: Blocks that overlap or are close will be merged.
	<td>None
</table>

<p>The following three images show the same regions when evaluated with the following 3 parameter sets:
<br>Default (one block)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<ttl>Same orient</ttl> (three blocks)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<ttl>Same orient</ttl> with <ttl>Merge</ttl> (two blocks).
<a href="img/pairOrient.png"><img src="img/pairOrient.png" alt="Pair Orient" style="border: 1px solid black; width:  700px"></a>
<br>In the last image, the reverse orientation block is embedded in another block
<p>See <a href="demo.html#synteny" class="ext" target="_blank">Synteny results</a> for further comparison of using the different
synteny parameters.

<a id="order"></a>
<h4>Order against</h4>

For draft sequences, they may be ordered against another project.
See <a href="./SystemGuide.html#draft2" class="ext" target="_blank">Ordering details</a>.
<table>
<tr><td>The <ttx>Draf-&gt;Seq2</ttx> and <ttx>Seq2-&gt;Draf</ttx> use the
<ttl>Abbreviation</ttl> set in the <ttb>Project parameters</ttb> panel.
The "-&gt;" indicates that the first sequence will be ordered against the second.
<td><a href="img/parameters2c.png"><img src="img/parameters2c.png" alt="pair parameters synteny" style="border: 1px solid black; width: 280px"></a>
</table>

<p>If the draft has been aligned to the <ttl>Order against</ttl> sequence, but this option
was not set, it can be set and the <ttx>Synteny Only</ttx> (see below) run with the existing alignment files.

<a id="hint2"></a>
<p><b>Hints about synteny parameter settings</b>

<p>It is easier to experiment with synteny parameters since <ttx>Synteny Only</ttx> can be used to speed it up.
For two genome complete sequence synteny, it is strongly suggested you start with <ttx>Strict</ttx>.

<p>&rarr;&nbsp;<i>Suggestion</i>: For large genomes, experiment with the parameters on just one or two pairs of the chromosomes;
you can use <a href="input/index.html" class="ext" target="_blank">xToSymap</a> for the split. Note: the synteny results of
one or two chromosomes will be slightly different compared to whole genome synteny.

<a id="save2"></a>
<h3>3.e Save</h3>

<a id="drop"></a>
<h4>Lower-left drop-down</h4>
This is shown in the image right above on the right.
<table class="tx">
<tr><td><ttx>Align&Synteny</ttx>
	<td>If the pair has a blank in the <ttl>Available Syntenies</ttl> table,
	this will be the only option.
<tr><td><ttx>Clust&Synteny</ttx>
	<td>If the pair has an <b>A</b> in the <ttl>Available Syntenies</ttl> table,
	this option will replace the <ttx>Align&Synteny</ttx> in the drop-down.
	Use this option if you have changed the clustering parameters.
<tr><td><ttx>Synteny Only</ttx><td>
If the pair has a &#10004; in the <ttl>Available Syntenies</ttl> table, this option will
be available.
<br>If you have only changed synteny parameters, select this option.
<tr><td><ttx>Number Pseudo</ttx><td>
If the pair has a &#10004; in the <ttl>Available Syntenies</ttl> table, this option will
be available.
<br>If you only want the pseudos number, select this option (see <a href="#pseudo"> Pseudos</a>).
</table>
<ul>
<li>Select <ttb>Save</ttb> after changing this.
<p>
<li><ttp>Manager</ttp>: The <ttb>Selected pair</ttb> button
will have its label replace to reflect the drop-down setting, as follows:
<ttx>Align&amp;Synteny &rarr; Selected Pair
<br>Clust&amp;Synteny &rarr; Selected Redo
<br>Synteny Only &nbsp;&rarr; Synteny Only
<br>Number Pseudo &rarr; Pseudo Only</ttx>
</ul>
<!--------------------->
<h4>Saving pairs parameters</h4>

Before the <ttl>A&amp;S</ttl> is executed, the parameters are saved in
<pre>  data/seq_results/&lt;proj1-to-proj2&gt;/params.txt</pre>
Once the <ttl>A&amp;S</ttl> is executed, the parameters are stored in the database.

<p>The file parameters are shown on the pair <ttb>Parameter</ttb> panel. These can only be
viewed and changed using <ttp>symap</ttp> (not <ttp>viewSymap</ttp>).

<p>Any parameter not the default will be shown on the <ttl>Summary</ttl> page.

<p>BEWARE: If you run <ttl>A&amp;S</ttl>, then change the PROmer or NUCmer settings, but forget to <ttb>Clear Pair</ttb>
before running <ttl>A&amp;S</ttl> again, the parameters
on the <ttb>Summary</ttb> page will be wrong (SyMAP does not check for this situation).
<!------------------------------------------------------------------------>
<a id="mum"></a>
<h3>3.f MUMmer parameters</h3>

The arguments for MUMmer are <b>NOT</b> checked for correctness.
<p>
To see the parameters for the default MUMmer V3 on MacOS, from the <tt>symap</tt> directory:
<pre>
./ext/mummer/mac/mummer -h
./ext/mummer/mac/promer -h
./ext/mummer/mac/nucmer -h
</pre>
To see the parameters for the default MUMmer V3 on Linux:
<pre>
./ext/mummer/linux/mummer -h
./ext/mummer/linux/promer -h
./ext/mummer/linux/nucmer -h
</pre>

If  you compiled V4 in the /ext directory:
<pre>
./ext/mummer4/m4/bin/mummer -h
./ext/mummer4/m4/bin/promer -h
./ext/mummer4/m4/bin/nucmer -h
</pre>
For a detail discussion of MUMmer running in SyMAP,
see <a href="./MUMmer.html" class="ext" target="_blank">MUMmer</a>.
<p>After running MUMmer, all alignment files are removed except the ".mum" file; to prevent removal,
<br> execute
<tt>./symap -mum</tt>.

<!---- END BODY -->
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p class="trailer">  Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></p>
</div>
</div>
</body>
</html>


