<!DOCTYPE html>
<html>
<head>
<title>SyMAP User Guide</title>
<style>
	a:link {text-decoration: none;color:#6666CC;}
	a.current {font-family:verdana; font-size:12px; color:white; text-decoration: none;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 3px; }
    top {vertical-align:text-top;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica,  sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP User Guide
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="current">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

This guide is for the Seq-to-Seq (sequence) synteny user interface.
Note, the term <i>pseudo</i> and <i>sequence</i> are used interchangeably.
<ol>
<li>Most of the contents also apply to the FPC-to-Seq view, where the specific information is provided in <a href=UserGuideFPC.html class="ext" target="_blank">FPC specific</a>.
<li>For detailed information on installation, see
the <a href="SystemGuide.html" class="ext"  target="_blank">System Guide</a>.
<li>For a quick visual introduction, see the <a href="Tour.html" class="ext" target="_blank">Tour</a>.
<li>Please send bug reports and suggestions to symap@agcol.arizona.edu.
</ol>

<h2>Contents</h2>
The Dot Plot and Circle displays are available from various places.
<ul>
<li><a href="#mgr">A. Main Display</a>
<li>B. Display for Selected Pair (Two-Genome Views):
	<ul>
	<li><a href="#dotplot_display">Dot Plot (Two genome) #1</a> <!-- # ref in code -->
		<ul>
		<li><a href="#dotplot_control">Dotplot Control Panel</a>
		<li><a href="#dotplot_filter">Dotplot Filters</a>
	</ul>

	<li><a href="#block">Block Display</a>
	<li><a href="#circle">Circle Display (Two genome) #1</a> <!-- # ref in code -->
	<li><a href="#sum">Summary</a>
	</ul>

<li>C. Display for All Projects (Multi-Genome Views):
	<ul>
	<li><a href="#explorer">Chromosome Explorer</a> (Multi-chromosome views)
		<ul>
		<li><a href="#alignment_display_circ">Circle Display (Multi-chromosome) #2</a>
		<li><a href="#dot2">Dot Plot (Multi-chromosome) #2</a>
		<li><a href="#alignment_display_2d">2D Display</a> <!-- # ref in code -->
			<ul>
			<li><a href="#control">Control Panel and Navigation</a>
			<li><a href="#main">Tracks in Main Panel</a>
				<ul>
				<li><a href="#sequence_track">Sequence Track and Filter</a>
				<li><a href="#hit">Hit Filter</a>
				</ul>
			<li><a href="#base">Base View</a>

			</ul>
		</ul>
	<li><a href="#dot3">Dot Plot (Multi-genome) #3</a>
	<li><a href="#annotsearch">SyMAP Queries </a>(Seq-to-Seq only)
	</ul>
<li><a href="#self">Self-synteny Displays</a>
<li><a href="#data_download">Data Download</a>
<li><a href="#image">Print and Help</a>
</ul>

The SyMAP User Agreement requires you cite one of the following articles
if you use SyMAP in a paper, poster, or presentation.

<pre>
        C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
        SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps
        Genome Research 16:1159-1168.
</pre>
<pre>
        C. Soderlund, M. Bomhoff, and W. Nelson (2010)
        SyMAP: A turnkey synteny system with application to plant genomes.
        Nucleic Acids Res 39(10):e68.
</pre>

<a id="mgr"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h1>A. Main display</h1>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
The display in the image below is shown when the following is executed from the terminal:
<pre>   ./viewSymap</pre>
Is it also available from <tt>./symap</tt>, where this display will also have the alignment commands.
<p>
<table>
<tr><td class="top">
All projects in the database will be listed on the left panel. If a date is beside the project name,
it has been loaded into the database on that date. If a [n] follows the date, it is part of n computed syntenies.
<p>Selecting projects on the left panel shows them on the right panel. A check mark in the
<b>Available Syntenies</b> table indicates pairs that have computed synteny that can be viewed.
<p>By selecting a cell with synteny, the <b>Display for Selected Pair:</b> buttons are activated.

<td><a href="img/symapdb3.png"><img src="img/symapdb3.png"  alt="" style="border: 1px solid black; width: 500px"></a>
</table>

<h1>B. Display for Selected Pair</h1>
The four buttons beside this label are "whole genome" displays for the selected pair (highlighted yellow).
<!------------------------------------------------------------>
<a id="dotplot_display"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Dot Plot (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Dots represent anchors (also referred to as <i>hits</i>). A blue box indicates a
<i>Synteny Block</i>. The horizontal and vertical grid lines
delineate the chromosomes (or linkage groups) of the two genomes being compared.
<p>
<table>
<tr>
<td>
 The
individual grid boxes (cell) for each chromosome pair are referred to as <i>Regions</i>, and clicking
on a region brings up a larger-scale <i>Region view</i>.
<p>Within the region view:
<ul>
<li>Click on a blue box or drag the mouse to highlight a sub-region.</li>
<li>Then, click on the highlighted sub-region to bring up a SyMAP 2D view for that sub-region.</li>
<li>This is illustrated in the <a href="#dot2">Chromosome Dotplot</a>.
</ul>

<td><a href="img/dot-two-genome.png"><img src="img/dot-two-genome.png"  alt="" style="border: 1px solid black; width: 500px"></a>
</table>

<a id="dotplot_control"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Dotplot Control Panel</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/controlpanel2.png"><img src="img/controlpanel2.png"  alt="" style="border: 1px solid black;"></a>
<p>
<b>Home</b> Resets the zoom selection to the full grid view of the entire dot plot.
<p>
<b>Zoom</b> The plus (<b><big>+</big></b>) and minus (<b><big>-</big></b>) buttons shrink or grow the display.
<p>
<b>Scale</b> The check box resizes the species displayed on the y-axis according to the current scale of
the species displayed on the x-axis.
<p>
<b>Reference</b> The drop-down menu allows the reference species on the x-axis to be changed.
<p>
<b>Size</b> The slider increases/decreases the size of the dots in the display.


<a id="dotplot_filter"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Dotplot Filters</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr><td>
The Filter image on the right indicates the default settings.
<p><b>Show Only Block Hits</b>
A <i>block hit</i> is a hit that the SyMAP synteny-finding algorithm
used to create a synteny block.
<p>
<b>Show All Hits</b>
Displays all hits. Note that the SyMAP hit loading algorithm removes
some repetitive hits from the raw anchor set, so some hits from the raw BLAT
or MUMmer files will not be shown.
<p>
<b>Highlight Block Hits</b>
Highlight hits that make up a synteny block in blue.
<p>
<b>Show Blocks</b>
This option draws blue rectangles around the synteny blocks.
<p>
<b>Show Empty Regions</b>
When deselected, hides regions that do not contain any synteny blocks.
<td><a href="img/dotplot_filter.png"><img src="img/dotplot_filter.png" alt="" style="border: 1px solid black; width: 220px"></a>
</table>
<!------------------------------------------------->
<a id="block"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Blocks Display (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The image below shows the block display for Maize and Rice.
The 10 Maize chromosomes are color coded as shown at the top, and aligned to
the 12 rice chromosomes. The Rice to Maize block view can be viewed by
selecting <u>reverse</u>.
<table>
<tr>
<td><a href="img/block.png"><img src="img/block.png" alt="" style="border: 1px solid black; width: 500px"></a>
<td>Clicking a chromosome pops up a window of the chromosome as shown below. Clicking
a block pops up a 2D view.
<p><a href="img/block2.png"><img src="img/block2.png"  alt="" style="border: 1px solid black; width: 220px"></a>

</table>
<!------------------------------------------------->
<a id="circle"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Circle Display (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The circle display shows the chromosomes arranged in a circle,
with synteny blocks shown as colored
ribbons between the chromosomes.
The color of a ribbon comes from one of the two chromosomes it connects; for example,
in the image below, the Rice Chr1 is red and the corresponding ribbon is red.
Clicking on a species names causes its colors to be used preferentially (<i>sometimes it
is necessary to click at the very beginning or end of the name for this to work</i>).
Clicking on a chromosome (e.g. the the red part of the circle for Rice Chr1)
causes ribbons from that chromosome to be shown on top.
<p>
<table>
<tr><td>
The three controls on the left above the circle image control various aspects of the display.
The first two buttons (<b>+, -</b>) change the size, while the third rotates the image.
<p>
<b>Scale to genome size:</b> This checkbox causes the chromosomes to be drawn proportionately to their
actual length in basepairs. Otherwise, each species is allocated the same amount of space,
e.g. for two species each one gets exactly half the circle.
<p>
<b>Show all blocks:</b>
The drop-down menu has "Show inverted blocks", "Show non-inverted blocks",
and "Two-color scheme", where this last option shows inverted=green, non-inverted=red.
<p>
<b>Self-align: </b> This checkbox controls whether blocks are shown between chromosomes
of the same species (if there is a check mark in the projects self-synteny cell indicating
its been computed).
<p>
<b>Rotated text:</b> Depending on the system, it may be easier (or harder) to read with
the text rotated.

<td><a href="img/circle-two-genome.png">
	<img src="img/circle-two-genome.png" alt="" style="border: 1px solid black; width: 450px"></a>
</table>

<!------------------------------------------------->
<a id="sum"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Summary (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The <b>Help</b> button on this view explains the different columns.
<p><a href="img/sum-two-genome.png">
	<img src="img/sum-two-genome.png" alt="" style="border: 1px solid black; width: 650px"></a>
<!------------------------------------------------->

<h1>C. Display for All Projects</h1>
Select two or more projects and the "Chromosome Explorer" and "Dotplot" buttons
will be activated.

<a id="explorer"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Chromosome Explorer</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The Explorer is a two-panel display in which the left panel is used to select specific chromosomes, while the right panel
shows the synteny for the selected chromosomes. Three different right-panel views are available:
<a href="#alignment_display_circ">circle</a>,
<a href="#alignment_display_2d">2D</a> and
<a href="#dotplot_display">Dotplot</a>.
Below is a snapshot of the Explorer showing the circle format, which is the initial view.
<p>
<a id="circfig"></a>
<a href="img/explorer.png">
	<img src="img/explorer.png" alt="" style="border: 1px solid black; width: 650px"></a>

<p><b>Left panel: </b>The left panel controls which species and chromosomes are shown:
<ul>
<li>Click a chromosome rectangle to add it to the display in the right panel. Click it again to remove
it.
<li>Click the chromosome number above a chromosome to make it the <i>reference</i>. (The
choice of reference sequence does not matter for the circle view,
but it is important for the 2D and Dot Plot views; see below).
<li> Adding and removing chromosomes affects the Circle view instantly.
To add/remove from the 2D or Dot Plot
views, open the Circle view, make the change, and then re-open the 2D or Dot Plot.
</ul>
<p>
Instructions: As the mouse moves over various
components, instructions for that component are shown in the Instruction box.
<p>

The four buttons in the lower-left corner of the window change the view in the right panel,
where the choices are
<a href="#alignment_display_2d">2D</a>,
<a href="#dotplot_display">Dotplot</a>,
 and <a href="#alignment_display_circ">Circle</a>.
<p>
Download Blocks: Exports a table of all of the synteny block co-ordinates for the selected species; see also <a href="#data_download">Data Download</a>.
<p>
Click the minus (<b><big>-</big></b>) button next to a project name or the Instruction box,
and that section will be hidden;
the minus sign will change to a plus.
Click the plus (<b><big>+</big></b>) button to show the project or Instruction box..


<p><b>Right panel:</b> The right panel shows the synteny display for the species and chromosomes selected on the left.
There are three views:
<a href="#alignment_display_2d">2D</a>,
<a href="#dotplot_display">Dotplot</a>,
 and <a href="#alignment_display_circ">Circle</a>.
The first three modes are
essentially overview modes, while the 2D view allows zooming in to details, all the way
to the basepair level if desired.
Each right-panel display format has its own controls; see the individual sections for details.

<!----------------------------------------------------------------->
<a id="alignment_display_circ"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Circle Display (Multi-chromosome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

This view is shown in the above image.
All features are described in the first <a href="#circle">circle section</a>,
except it is chromosome-based instead of genome-based.

<!----------------------------------------------------------------->
<a id="dot2"></a>
<h3>Dot Plot (Multi-chromosome)</h3>

All features are described in the first <a href="#dotplot_display">dot plot section</a>,
except that it is chromosome-based instead of genome-based.
Selecting a region followed by a synteny block (or selected region)
brings up the 2D view.
<table>
<tr><td><a href="img/dot-chr.png">
	<img src="img/dot-chr.png" alt="" style="border: 1px solid black; width: 350px"></a>
<td><a href="img/dot-chr-2d.png">
	<img src="img/dot-chr-2d.png" alt="" style="border: 1px solid black; width: 350px"></a>
</table>

<!-------------------------------------------------------------------->
<a id="alignment_display_2d"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2D Display (Multi-chromosome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The 2D display is activated from the <a href="#explorer">Explorer</a> by
clicking the <b>2D</b> button (or by selecting a region or block from a dot plot).
<p>
<table>
<tr><td>
On the right is an image depicting an alignment from Rice Chr10 to Maize Chr1 to Sorghum Chr1.
Note that Maize Chr1 is the <i>reference</i>
chromosome, hence is placed in between the others.
<p>
The alignment window displays <i>tracks</i> (drawn as rectangles) of sequence information,
showing synteny <i>hits</i> (drawn as lines) between tracks.
<p>
Left click in a track shows the most commonly used filters and features (e.g.
flipping the region). They are also on the filter menu above each track
(e.g. Sequence Filter). See the Filter descriptions below.
<td><a href="img/symapAlign.png">
	<img src="img/symapAlign.png" alt="" style="border: 1px solid black; width: 550px"></a>
</table>

<a id="control"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Control Panel and Navigation</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The control panel provides buttons that allow for navigation between views, scaling, color customization, and help.
<p>
<a href="img/controlpanel.png"><img src="img/controlpanel.png" alt=""></a>
<p>
<b>History (Home, &lt;, &gt;)</b> SyMAP retains a record of the prior views (like a web browser).
The history navigation back (<b><big>&lt;</big></b>) and
forward (<b><big>&gt;</big></b>) buttons allow you to
move back and forth through the previous views.
The <b>Home</b> button returns to the initial view.
<p>
<b>Zoom Buttons (-, +)</b>
The zoom buttons allow for quickly expanding and narrowing the view region.
 The minus (<b><big>-</big></b>) button shrinks the view region by 50%, keeping
the same center, while the plus (<b><big>+</big></b>) button doubles the view region,
again keeping the same center.
<p>
<b>Scale Button</b>
The Scale button
(right of <b><big>+</big></b>) resizes the tracks so that they are in the same scale
(base pairs per pixel) as the reference sequence track in the view.
<p>
<a id="mouse_function_selector"></a>
<b>Selected region</b>
This drop-down selector assigns a function to the mouse's left button
click and drag actions:

<table class="ty" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr><td style="white-space: nowrap; vertical-align: text-top;">Zoom All Tracks<td>Zoom the track to the selected region and all other tracks with hits in the region.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Zoom Selected Track<td>Zoom to the selected region on the selected track; all other tracks do not change.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Align (Max 30000)<td>Open the base alignment view for the selected region (see <a href="#base">Base View</a>).
No more than 30000 bases can be selected.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Show Sequence<td>The selected sequence will be shown in a popup. This has limited utility, but is the
only way to the underlying sequence of a region than may or may not have hit.
</table>

<p><b>Colors</b> Opens a menu for customizing colors.
<p>
<b>Additional navigation:</b>
<table  class="ty" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;"><b>Resize Track</b>&nbsp;</td>
	<td>Drag bottom of track</td>
	<td>Position mouse at bottom of track (resize cursor appears), hold down left
	mouse button, and move mouse.</td></tr>
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;"><b>Scroll Track</b>&nbsp;</td>
	<td>Mouse wheel</td>
	<td>Position mouse over track and use mouse wheel.</td></tr>
<tr>
	<td><b>Filter</b>&nbsp;</td>
	<td>Right mouse button</td>
	<td>Position mouse over track or white space between tracks, and click right mouse button.</td></tr>
</table>


<a id="main"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Tracks in Main Panel</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The main panel is divided into <i>tracks</i>,
where each track represents a sequence. Each track
is aligned to the track on its immediate left or right, and the
individual pieces of evidence for synteny (referred to as <i>anchors</i> or <i>hits</i>) are drawn as
lines between the two tracks. Above each track
is a "filter" button which allows configuration of the content displayed for
that track (e.g. annotations), and above the space between each pair of tracks
is a "hit filter" button that allows selection of the types of hits displayed
between the tracks on either side (e.g. only show syntenic hits).

<a id="sequence_track"></a>
<h4><i>Sequence Track and Filter</i></h4>

A Sequence track may represent a sequenced chromosome, linkage group or scaffold.
The sequence length
and the coordinates of the displayed region are indicated above and below the track.
<p>
Annotation data of the following types may be loaded into SyMAP and displayed.
(Note that display colors <a href="#control">can be changed</a>.)
<p>
<table class="ty" style="width:650;">
	<tr>
		<td><i>Annotation Type</i></td>
		<td><i>Display</i></td>
	</tr>
	<tr>
		<td>Gaps</td>
		<td>Red band across the chromosome</td>
	</tr>
	<tr>
		<td>Centromere</td>
		<td>Blue "X" across the chromosome</td>
	</tr>
	<tr>
		<td>Predicted genes and exons</td>
		<td>Annotation strip in center of chromosome (see <a href="#seqtrackimg">image</a> )</td>
	</tr>
</table>

<p>
The sequence information displayed can be configured via the
<a href="#sequence_filter">Sequence Filter</a> by clicking the
<b>Sequence Filter</b> button above the sequence track, or by right-clicking the
mouse over the sequence track.
<p>
The example below shows a sequence-to-sequence alignment, which was derived by right-clicking
at the top of the region, pulling down the mouse and releasing in order to view a small region.

<p>
<a id="seqtrackimg"></a>
<table>
<tr><td>
The black lines with heavier blue bars interspersed down the center indicate exon/intron
predictions. A solid black line would indicate the absence of introns in the annotation.
<p>The alternating brown/gray lines indicate anchor clusters. The solid brown lines indicate
a single anchor (or multiple anchors very close together), while the gray lines are spaces with no alignment.
<p>The ruler on the right side of the sequence rectangle
shows the relative BP position along the sequence.
<p>Additional information about each
annotation may be seen by choosing "Show Annotation Description" from the Sequence Filter.

<td><a href="img/2d-closeup.png">
	<img src="img/2d-closeup.png" alt="" style="border: 1px solid black; width: 500px"></a>

<tr><td>Right clicking everywhere but in a yellow annotation box provides
a pop-up menu with most of the items found on the Sequence Filter. Right
clicking on a yellow box results in a popup, as shown on the right; this
is useful for copying all or part of the description, and when the box
is partially hidden.
<td><a href="img/2d-desc.png">
	<img src="img/2d-desc.png" alt="" style="border: 1px solid black; width: 250px"></a>
</table>

<p>

<a id="sequence_filter"></a>
The Sequence Filter allows the user to select the type of information shown for the
sequence.  It is accessed via the <b>Sequence Filter</b> button above the sequence track,
or by right-clicking the mouse over the sequence track.
<p>
<b>Start and End</b>
The positions of the sequence display can be set via the
corresponding text boxes.
The units of the values entered can be selected from the accompanying drop down menus
(BP, KB, MB, GB).
<p>
<b>Full Sequence</b> Sets the start and end positions of
the sequence display to encompass the whole chromosome.
<p>

<table>
<tr><td>
<p>
<b>Flip</b> Reverses the orientation of the sequence track.
<p>
<b>Show Genes</b>
Enables/disables the display of gene/exon annotations
 along the sequence.
<p>
<b>Show Annotation Descriptions</b> Enables/disables the display of the annotation descriptions along the
right side of sequence. NOTE: this only works if you are zoomed in close enough that they can clearly
be displayed.

<p>
<b>Show Gaps</b> Enables/disables the display of sequence gaps (drawn as solid red
rectangles) along the sequence.
<p>
<b>Show Centromere</b> Enables/disables the display of the centromere (drawn as a cyan "X") on
the sequence.
<p>
<b>Show Framework Markers</b> Typically genetic markers or radiation hybrid markers.
This option enables/disables the display of framework markers (drawn as solid green
rectangles) along the sequence.
<p>

<td>
<a href="img/sequencefilter.png">
	<img src="img/sequencefilter.png" alt="" style="border: 1px solid black; width: 300px"></a>
</table>

<p>
<b>Show Ruler</b> Enables/disables the display of the sequence ruler along the right side
of the sequence.
<p>
<b>Show Hit Length</b> Enables/disables the display of the hit length line next to each hit along the
sequence.  The hit length line denotes the start and end points of the hits relative
to the sequence (may not be visible for tiny hits).
<p>
<b>Show Hit Score Line</b> Enables/disables the display of the score line next to each hit along the
sequence.  The length of the line represents the magnitude of the % Identity value for
the hit.
<p>
<b>Show Hit Score Value</b> Enables/disables the display of the score value next to each hit along the
sequence.  The score value corresponds to the % Identity value for the hit.

<a id="hitfilter"></a>
<a id="hit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4><i>Hit Filter</i></h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The Hit Filter menu allows the user to select which types of hits are displayed.
<table>
<tr><td>
<p>
The types of filter are as follows:

<p>
<b>Show Only Synteny Hits</b> Shows only hits that are part of a synteny block.
<p>
<b>Show Only Hits to Genes</b> Show only hits which intersect a gene annotation.
<p>
<b>Show Only Non-Gene Hits</b> Show only hits which do not intersect a gene annotation.
<p>
<b>Show All Hits</b> hows all known hits.

<td><a href="img/hitfilterS.png">
	<img src="img/hitfilterS.png" alt="" style="border: 1px solid black; width: 250px"></a>
</table>

<a id="base"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Base View</h4></td>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Select "Base View" from the "Mouse" drop-down, then select a region of maximum 50kb.
To select a region,  drag the
mouse along the sequence and release when the desired range is highlighted.
<p>
<a href="img/baseview.png">
	<img src="img/baseview.png" alt="" style="border: 1px solid black; width: 750px"></a>
<p>
If there are multiple hits in the selected region, than the base view of those hits
appears in a new dialog.  This view consists of a ruler along the top showing the area
of the sequence covered, the hits, and the genes.
<p>
<b>Hits</b><br>
Hits are displayed as lines with an arrow on one end showing the direction of the hit.
Clicking on a hit brings up the hit's
BLAST view in the bottom of the dialog. The hit presently shown in the BLAST view is
shown in gray.  A vertical red line along a hit represents a mismatch .
A vertical green line
along a hit represents a deletion. An arrow pointing down (i.e. 'v') along a hit
represents an insertion.
<p>
<b>Genes</b><br>
Annotated genes are displayed below the hits.  Exons are represented by a blue box, and
the leftmost or rightmost exon box will have an arrow tip indicating whether the gene
is on the + strand (right-pointing) or - strand (left-pointing) relative to the
sequence. The system attempts to expand the view to show the full gene that has
hits. Genes that overlap this gene will be partially shown.

<!-------------------------------------------------------------------->
<a id="dot3"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Dot Plot (Multi-genome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

All selected projects will be shown in the Dot Plot. The interface is the same as
discussed in <a href="#dotplot_display">Dot Plot (Two Genome)</a>. The dot plot in the image
shows three genomes.
<p><a href="img/dotplot.png"><img src="img/dotplot.png" alt="" style="border: 1px solid black; width: 600px"></a>
<!-------------------------------------------------------------------->

<a id="annotsearch"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>SyMAP Queries</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The query interface only works for multi-genome Seq-to-Seq synteny
(i.e does not work for self-synteny).
This interface has two basic
functions:
<br>
A. Locate homologous (or paralogous) regions based on annotation and location
<br>
B. Create putative gene families spanning multiple species, and apply family-based filters
<p>
Some sample queries which are possible through the interface:
<p>
<i>&bull; Find un-annotated regions on one genome which are homologous to regions on another genome
which have already been annotated. </i>
<p>
<i> &bull; Find putative gene families which are present in one lineage but absent in another.</i>
<p>
<i>&bull; Identify the likely orthologs on genome A of a given gene on genome B, by using synteny blocks
or exactly collinear pairs to filter out probable paralogs.  </i>
<p>
<i> &bull; Find orphan genes which are specific to one genome.</i>
<p>
To open the query interface, first select two or more sequence projects in the Project Manager. Then press the "SyMAP Queries" button to open the interface:
<p>
<a href="img/query_oview.png"><img src="img/query_oview.png" alt="" style="border: 1px solid black; width: 500px"></a>
<p>
The Overview window (above) lists the projects which were selected for querying. To set up a query, open the
"Query Setup" window by clicking on its title in the left panel. This opens the window below:
<p>
<a href="img/query_query.png"><img src="img/query_query.png" alt="" style="border: 1px solid black; width: 550px"></a>
<p>
Here you set up the query and then press "Run Search" to execute it. <p>
The query occurs in two stages, corresponding to the section numbers
on the Query Page. The second start, PgeneF, only runs if the "Compute PgeneF" is checked.

<h3>1. Filter hits</h3>

The first stage is to retrieve hits (anchors) from the database, based on
filters such as chromosome and annotation string. Note that each anchor
connects two species and hence represents a <i>pair</i> of putatively-related
regions on the two species. The filters are described in the <b>Help</b> in the upper corner.

<p><b>Annotation Description</b>
<br>
Enter a search string for any annotation keyword
(e.g. the gene name or description). Hits will be returned which overlap a matching annotation
on either side.
<p>When looking for a particular gene, say "gene-LOC9268236", enter this into the text box
and "Run Query". If there are no results, select "Only orphan genes" and run it again.

<p><b>Collinear genes N&gt;= N</b>
<br>
Return only hits which are part of a pair of N aligning genes having no intervening
non-aligning genes. There is a column called "RunSize" that provides the number
of adjacent aligning genes. Note that this is not the same as synteny blocks, which
allows intervening non-aligning genes.

<p><b>Show orphan genes</b>
<br>
If checked, this shows all orphan genes for the project selected (or all projects).
That is, it shows genes that have no overlapping hits to ANY other project for which
it has been compared.


<h3>2. Filter putative gene families (PgeneFs)</h3>

Using the hits that pass the Stage 1 filters, SyMAP constructs putative gene families (PgeneFs)
spanning the selected species. This is done by grouping hits whose hit regions overlap on
at least one genome. <i>Note, if you have more than 6 species selected, this stage can take
an hour or more.</i>
<p>
Each PgeneF is given a number, which is shown in the Query Results table (column name "PgeneF").
The size of the PgeneF is also shown (column "PgFSize").
<p>
The filters are described in the Query <b>Help</b> page.

<h3>SyMAP Query Result Table</h3>
<p>
When the query is complete, the Query Results page opens showing the table of results:
<p>

<a href="img/queryres1.png"><img src="img/queryres1.png" alt="" style="border: 1px solid black; width: 550px"></a>

<p>
The table contains all the hits (anchors) resulting from the query. Each hit connects
two species and you can see the respective chromosomes and start/end locations of
the hits, as well as gene annotations overlapped by the hits.
<p>
Note that the table contains columns for <i>all</i> of the selected
species (see snapshot below), but each hit only connects two species, and the other species columns are empty.
If the query specified orphan genes, then each row represents one gene and shows data only
for one species.
<p>
The columns can be sorted by clicking on the column name,
and rearrange them by dragging the header boxes.

<h4>Button functions</h4>

<div style="width:700px;margin-top:10px; margin-left:15px;">
<b>Align sequences</b>:
Sequences of the selected hit are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32).
<p>
<b>Show Synteny</b> :
Brings up a SyMAP 2-track view for each selected entry, so the hit
can be seen in their full chromosomal context. The hit is initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.
<p>
<b>Export CSV</b>:
Saves the selected hits (or entire table) using the selected set of columns to a CSV format suitable for import into Excel.
<p>
<p>
<b>Save as Fasta</b>: Sequences from the selected hits (or entire table) are written to a Fasta file. Both sides of each hit
are written, using the start/end coordinates shown in the table.
<p>
</div>

<h3>Results Table Columns</h3>
You can add/remove columns using the <b>Select Columns</b> button at the bottom. This
opens up a column-selection section. The columns are explain the the <b>Help</b> button on the table.

<a id="self"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Self-synteny Display</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Self-synteny can be viewed in all views (see <a href="Demo.html#self" class="ext" target="_blank">Demo Dot Plot</a>).

<p>The "SyMAP queries" does not work for self-synteny.

<a id="data_download"></a>
<h2>Data Download</h2>

There are two ways to download data for SyMAP synteny blocks, individual anchors, and annotations:
<ul>
<li> Through the <a href="#explorer">Explorer</a>. Select the species of interest,
 open the Explorer, and the download button is at the lower left.
 This exports a table of all the computed synteny block co-ordinates for all the
 selected species, including their self-alignments, if those were computed.
<li> Through the <a href="#annotsearch">Query interface</a>. For example,
 choosing two species and executing a query with the default parameters
 will show a table of all the anchors found between the two species, plus
their annotations and synteny block membership, if any.
</ul>

<a id="image"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Print and Help</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Most displays shown have a <img src="img/print.png" alt="" style="border: 1px solid black; width: 50px"> set of icons.
<p>The left icon is for printing the image. However, this does not work with all versions
of Java. If it does not work, use the system "Screen Capture" (all the images in this
document were created with screen capture, along with the images in the SyMAP publications).

<p>The right
icon brings up the this web page, typically to the correct section (obviously, there needs
to be an internet connection).

<!-------------------------- this link is in SyMap--------------------------------------->
<a id="alignment_display_3d"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3D Display (if available)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
The 3D display is generally obsolete, but if there is a "3D" button on your display,  see
<a href="./3D.html">3D</a>.

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
