<!DOCTYPE html>
<html>
<head>
<title>SyMAP User Guide</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style.css">
</head>

<body>
<!------ left side index - customize ----------------------------------------------------->
<div class="sidenav">
<a href="http://www.agcol.arizona.edu">
		<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 160px; border:1px solid black; background: white;"></a>
<p class=sidetit>&nbsp;&nbsp;User Guide</p>
<p>
<a href="#term">Terminology</a>
<a href="#mgr">Project Manager</a>
<a href="#pair">Selected Pair<br>(Two-Genome)</a>
<a href="#circle">&nbsp;&nbsp;&nbsp;Circle #1</a>
<a href="#dotplot_display">&nbsp;&nbsp;&nbsp;Dot Plot #1</a>
<a href="#block">&nbsp;&nbsp;&nbsp;Blocks</a>
<a href="#report">&nbsp;&nbsp;&nbsp;Reports...</a>

<a href="#all">Chromosome Explorer<br>(Multi-chromosome)</a>
<a href="#alignment_display_circ">&nbsp;&nbsp;&nbsp;Circle #2</a>
<a href="#alignment_display_dot">&nbsp;&nbsp;&nbsp;Dot Plot #2</a>
<a href="#alignment_display_2d">&nbsp;&nbsp;&nbsp;2D Display</a>
<a href="#align">&nbsp;&nbsp;&nbsp;Base Align</a>
<a href="#dot3">Dot Plot #3<br>(Multi-Genome)</a>

<a href="#gen">General</a>

<a href="#top">Go to top</a>
</div>

<!------ right side header and body ----------------------------------------------------->
<div class="rightSide">
<a id="top"></a>

<div class="topnav"> <!-- header same in all pages -->
    <br>
	<table>
	<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="http://www.arizona.edu">
		<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
	<td><a href="http://www.bio5.org">
		<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="https://csoderlund.github.io/SyMAP">SyMAP<br>Home</a>&nbsp;&nbsp;
	<td><a href="https://github.com/csoderlund/SyMAP/releases">Download</a>&nbsp;&nbsp;
	<td><a href="./Docs.html">Docs</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./SystemGuide.html">System<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./input/index.html">Input</a>&nbsp;&nbsp;
	<td><a href="./SystemHelp.html">Parameters</a>&nbsp;&nbsp;
	<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<td><a href="./UserGuide.html">User<br>Guide</a>&nbsp;&nbsp;
	<td><a href="./Query.html">Queries</a>
	</table>
 </div>

<!---------------------------- bulk of page ------------------------------------------->
<div class="main">

<p>This guide is for the <ttp>viewSymap</ttp> user interface;
<a href="Query.html" target="_blank" class="ext">Queries</a> user guide is a separate HTML page.
<p>
<ul>
	<li>Please send bug reports and suggestions to <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>.
	<li>The documentation always applies to the latest release.
</ul>

<!------ See SyMAP code util.Jhtml for refs in code; see all 'ref in code' -->
<b>Contents</b>
<table>
<tr>
<td style="width: 400px; vertical-align: text-top;">
	<ul>
	<li><a href="#term">Terminology</a>
	<li><a href="#mgr">Project Manager</a>
	<li><a href="#pair"><i>Selected Pair (Two-Genome Views)</i></a>
		<ul>
		<li><a href="#circle">Circle #1</a> - <span style="color:purple">details of Circle</span> <!-- # ref in code -->
		<li><a href="#dotplot_display">Dot Plot #1</a> - <span style="color:purple">details of Dot plot</span> <!-- # ref in code -->
		<li><a href="#block">Blocks</a>
		<li><a href="#report">Reports...</a>
		</ul>

<!-- # ref in code; see util.Jhtml.java, which has all links into this document -->
	<li><a href="#all"><i>All Pairs (Multi-Genome Views)</i></a>
		<ul>
		<li><a href="#explorer">Chromosome Explorer  (Multi-chromosome)</a>
			<ul>
			<li><a href="#alignment_display_circ">Circle #2</a>
			<li><a href="#alignment_display_dot">Dot Plot #2</a>
			<li><a href="#alignment_display_2d">2D Display</a> <!-- # ref in code -->
				<ul>
				<li><a href="#view2D">View a region</a>
				<li><a href="#info">Hovers and popups</a>
				<br>&nbsp;&nbsp;<a href="#hinfo">Hit</a>,
				          &nbsp;<a href="#ginfo">Gene</a>,
				          &nbsp;<a href="#binfo">Gene-Hit</a>
				<li><a href="#control">Control panel and navigation</a> <!-- # ref in code -->
				<li><a href="#sequence_filter">Sequence Filter</a>,&nbsp;
				<a href="#hitfilter">Hit Filter</a>
				<li><a href="#align">Base align</a>
				<br>&nbsp;&nbsp;<a href="#base">Graphical</a>,
				          &nbsp;<a href="#btext">Text</a>,
				          &nbsp;<a href="#alt">Alternatives</a>
				</ul>
			</ul>
		<li><a href="#dot3">Dot Plot #3</a>
		<li><a href="Query.html" target="_blank" class="ext">Queries</a>
		<li><a href="Details.html" target="_blank" class="ext">Details</a> <small>Collinear Sets, Hit popup, 3-chr conserved</small>
		</ul>
	</ul>

<td style="vertical-align: text-top;">
	&bull; <a href="#gen">General</a>
	<ul>
	<li><a href="#self">Self-synteny</a>
	<li><a href="#data_download">Data Download</a>
	<li><a href="#image">Color Wheel, Print and Help (?) Icons</a> <!-- # ref 'wheel' in code -->
	</ul>
	<i>All <a href="#wheel">colors</a> mentioned in this document are the defaults.</i>
	<p>

	<table>
	<tr><td>&nbsp;
	<tr><td>
	    <small><i>The <ttp>Dot Plot</ttp> and <ttp>Circle </ttp> are available from various places.</i></small>
	<tr><td>
		<small>Project Manager</small>
		<br><a href="./img/view1.png">
				<img src="./img/view1.png" alt="Dots and Circle" title="Dots and Circle"  style="width: 380px; border:1px solid black"></a>
	<tr><td>
		<small>Chromosome Explorer</small>
		<br><a href="./img/view2.png">
			<img src="./img/view2.png" alt="Dots and Circle" title="Dots and Circle" style="width: 170px; border:1px solid black"></a>
	<tr><td>&nbsp;
	<tr><td>&nbsp;
	<tr><td><i>Click on any image to see the closeup.</i>
	</table>
</table>

<p><b>Citation</b>: The SyMAP User Agreement requires you to cite the following publication
if you use SyMAP results in a paper, poster, or presentation.
<p>
<pre>
	C. Soderlund, M. Bomhoff, and W. Nelson (2010)
	SyMAP: A turnkey synteny system with application to plant genomes.
	Nucleic Acids Res 39(10):e68. <a href="https://academic.oup.com/nar/article/39/10/e68/1310457" target="_blank" class="ext">Link</a>
</pre>
<pre>
	Also cite the MUMmer publication: <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005944" target="_blank" class="ext">Marcais et al. 2018</a> or <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2004-5-2-r12"  target="_blank" class="ext">Kurtz et al. 2004</a>
</pre>

<a id="term"></a>
<h1>Terminology</h1>

<table class="ty">
	<tr><td>Chromosome<br>(Group)
		<td>For simplicity, sequence groups will be referred to as "Chromosome",
		though they may be scaffolds, linkage groups, etc. In some cases, "group" is used as
		a generic term in place of chromosome.
	<tr><td style="white-space: nowrap;">Hit (anchor)
	<td>A matched sequence between two chromosomes, computed by MUMmer.
	<tr><td style="white-space: nowrap;">Clustered hits<td>
	SyMAP clusters hits that are close or overlapping into a single hit
	for display; the MUMmer hits within the clustered hits are referred to as <i>subhits</i>.
	<tr><td style="white-space: nowrap;">Paired genes<td>Two genes joined by a hit.
	<tr><td>Synteny block<td>An approximately-collinear sequences of anchors; a block can
	have intervening genes which do not align. There may be small inverted regions within
	a block.  By default, blocks have at least 7 anchors.
	<tr><td>Collinear set<td>A sequence of paired aligned genes with no intervening non-paired genes or inverted gene
	pairs. A set can consist of 2 or more genes. See <a href="Details.html" target="_blank" class="ext">details</a>.
</table>

<a id="term2"></a>
<p>The following is more detailed terminology:
<table class="ty">
	<tr><td style="white-space: nowrap;">Cluster Hit Algorithm:
		<br><ttx>Algo1</ttx> or <ttx>Algo2</ttx><small><sup>1</sup></small>
		<td>During the synteny computation, either the <ttx>Algo1</ttx> or <ttx>Algo2</ttx> algorithm could have
		been selected to cluster hits.
		<br>&bull; The original <ttx>Algo1</ttx> was updated v5.4.0 to be more gene-aware but does not know exon-intron structure.
		The new <ttx>Algo2</ttx> knows exon-intron structure, last updated v5.6.0.
		 See <a href="SystemHelp.html#clust" class=ext target="_blank">System Help</a> for details.

		<br>&bull; The <a href="#sum">Summary</a> specifies which was used.
		There are statistics in SyMAP that have different values based on which algorithm was run.
	<tr><td><ttx>Algo1</ttx>:
		<br>g2, g1, g0
		<td>A hit overlaps a gene on <i>both</i> sides (g2), a hit overlaps a gene on <i>one</i> side (g1),
		a hit overlaps <i>no</i> genes (g0). These are also written G2, G1, G0.
	<tr><td><ttx>Algo2</ttx>:
		<br>EE, EI, En, In, nn
		<td>E is exon, I is intron, n is intergenic. So a EI is a hit that at least partially aligns to an exon of one gene and
		an intron on the other, etc.
	<tr><td><ttx>Gene#</ttx><td>This is the SyMAP assigned number representing the order of the gene.
	When genes overlap, they are given the same gene number with different suffixes.
	For example, see <a href="#view2D">View a region</a>. Gene numbering starts at 1 for each chromosome.
	<tr><td><ttx>Major</ttx> or <ttx>minor</ttx><br> assigned gene
		<td>When genes overlap, they may share hits in the overlapping regions. A clustered hit
		can only be assigned directly to one gene pair, which is the <ttx>Major</ttx>  pair; the
		other gene pair is assigned to the hit as  <ttx>Minor</ttx> and only shown in the
		<ttb>Queries</ttb> results.
	<tr><td><ttx>Hit#</ttx>
		<td>The hits are numbered sequentially along the chromosome of the alphabetically lesser
		<ttx>Directory</ttx> name. For example, if on the <ttp>Project Manager</ttp> panel (see below),
		you select <ttl>A.thal</ttl> and <ttl>B.rapa</ttl>,
		which have <ttx>Directory</ttx> names <ttx>Arab</ttx> and <ttx>Brap</ttx>,
		the hit numbering will be sequential down the <ttx>Arab</ttx> chromosomes.
	<tr><td>Block and !Block
		<td>Hits that are in a block and hits that are not within a block, respectively.
</table>

<!---------------------------------------------------------->
<a id="mgr"></a>
<h1>Project Manager</h1>

The display in the image on the lower right is shown when the following is executed from the terminal:
<pre>   ./viewSymap</pre>
Is it also available from <tt>./symap</tt>, which will include the alignment commands.
<p>

<a id="views"></a> <!-- # ref in code -->
<table>
	<tr><td style="vertical-align: text-top;">
	<p>All projects in the database will be listed on the left panel. The date beside the project name
	is when it was loaded. If a [n] follows the date, it is part of n computed syntenies.

	<p>Selecting projects on the left panel shows them on the right panel. Clicking the <ttb>View</ttb> popups
	a window with the information about the project.

	<p>A check mark (&#10003) in the <ttb>Available Syntenies</ttb> table indicates the pair has computed synteny that can be viewed.

	<td>
	<small>Project Manager</small>
	<br><a href="img/viewPlants.png"><img src="img/viewPlants.png"  alt="viewPlants" style="border: 1px solid black; width: 550px"></a>
</table>

<table class="ty">
	<tr><td style="white-space: nowrap;vertical-align: text-top;"><ttb>Selected Pair</ttb>
		<td>By selecting a cell with synteny (the cell will turn green), whole genome views are activated for the
	selected synteny pair.
	<br>The views are <a href="#circle">Circle</a>, <a href="#dotplot_display">Dot Plot</a>,
	<a href="#block">Block View</a>, and <a href="#sum">Summary</a>.
	<tr><td style="white-space: nowrap;vertical-align: text-top;"><ttb>All Pairs</ttb>
		<td>The views beside this label will be applied to all checked syntenies.
		<br>The views are <a href="#explorer">Chromosome Explorer</a>,
		<a href="#dot3">Dot Plot</a> and <a href="Query.html">Queries</a>.
</table>

<!------------------------------------------------->
<a id="pair"></a>
<h1>Selected Pair (Two-genome views)</h1>
<table class="tz"><tr>
<td><a href="#circle">Circle</a>
<td><a href="#dotplot_display">Dot Plot</a>
<td><a href="#block">Blocks</a>
<td><a href="#report">Report</a>
<td><a href="#sum">Summary</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<!------------------------------------------------->
<a id="circle"></a>
<h2>Circle #1 (Two Genome)</h2>
<table class="tz"><tr>
<td><a href="#ccp">Control Panel</a>
<td><a href="#circColors">Colors</a>
<td><a href="#"></a>
<td><a href="#"></a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>The <ttp>Circle</ttp> can be viewed for two-genomes or <a href="UserGuide.html#explorer">multi-chromosomes</a>.
This section described the two-genome <ttp>Circle</ttp>, which also applies to the multi-chromosomes <ttp>Circle</ttp>.
<p>
The <ttp>Circle</ttp> display shows the chromosomes arranged in a circle,
	with synteny blocks shown as <u>colored blocks</u> between the chromosomes.
	The color of a block comes from one of the two chromosomes it connects; for example,
	in the image below, the A.thal Chr05 <u>color arc</u> is orange and the corresponding
	blocks are orange.
<table>
	<tr><td style="vertical-align:text-top;">
	<i>Project name</i>: Move the mouse over a project name and click once:
	The project name will be in shown in bold italics
	and the projects colors will be used for the blocks.
	<p>For example, in the image on the right,
		<i>A.thal</i> is in italics and all the block are colored with its arc colors;
		click Cabbage and the block will be colored with its arc colors.

	<p><i>Chromosome color arc</i>: Move the mouse over an arc, followed by:
	<table>
	<tr><td style="vertical-align:text-top;">1.<td>Click: The blocks from the chromosome will be shown in front.
	<tr><td style="vertical-align:text-top;">2.<td>Double-click: Only the blocks of the selected
	chromosome will be shown. To undo, click an arc or a project name.
	</table>

	<td><a href="img/circle-2g.png">
		<img src="img/circle-2g.png" alt="circle-two-genome" style="border: 1px solid black; width: 450px"></a>
</table>

<a id="ccp"></a>
<h4>Circle Control Panel</h4>
<a href="img/circle-control.png"><img src="img/circle-control.png"  alt="circle controls" style="border: 1px solid black;width: 650px"></a>

<p>First group of buttons:
<table  class="tx">
<tr><td><img src="img/home.gif" alt="Home" style="width: 20px;">
     <td>Takes the display back to its original display settings, i.e. for all buttons up to the
     the first "|".
<tr><td><img src="img/plus.gif" alt="Plus" style="width: 20px;">
	<td>Increase the circle size.
<tr><td><img src="img/minus.gif" alt="Minus" style="width: 20px;">
	<td>Decrease the circle size.
<tr><td><img src="img/rotate-right.png" alt="Rotate Right" style="width: 20px;">
	<td>Rotate the circle clock-wise.
<tr><td><img src="img/rotate-left.png" alt="Rotate Left" style="width: 20px;">
	<td>Rotate the circle counter-clock-wise.
<tr><td style="vertical-align:text-top;"><ttl>Rotate</ttl>
	<td>Toggle: Rotate the text to face the circle. Un-toggled: As shown in the above image.
<tr><td style="vertical-align:text-top;"><img src="img/scale.gif" alt="Scale" style="width: 20px;">
	<td>Toggle: Draw the chromosomes proportionately to their actual length in basepairs.
<br>Un-toggled: Each species is allocated the same amount of space,
e.g. for two species each one gets exactly half the circle.
</table>

<p>The <ttl>All blocks</ttl> (default) popup menu has:
<table class="tx">
	<tr><td><ttl>All blocks</ttl><td>Show all blocks
	<tr><td><ttl>Inverted</ttl><td>Only show the inverted blocks
	<tr><td><ttl>Non-inverted</ttl><td>Only show the non-inverted blocks
	<tr><td><ttl>Two-color</ttl><td>Color the blocks as inverted=green, non-inverted=red.
</table>

<p>Remaining buttons:
<table class="tx">
	<tr><td style="white-space: nowrap; vertical-align:text-top;"><ttl>Self-align</ttl>
	<td>(If self-synteny has been computed) When toggled, the self-synteny blocks are shown.
<tr><td style="vertical-align:text-top;"><ttl>Reverse</ttl>
	<td>(Only available for 2-genome display) When toggled,
	the reference will be reversed, i.e. the bottom project is shown on top and uses the reference colors
	(you may need to click the top projects name to use its colors).

<tr><td><img src="img/colorPalette.png" alt="Color" style="width: 20px;">
	<td>Change the colors of the chromosome blocks, as described below.<td>
<tr><td><img src="img/print.gif" alt="Print" style="width: 20px;">
	<td>Prints the circle graphics as displayed.<td>
<tr><td><img src="img/helpSm.png" alt="Help" style="width: 20px;">
	<td>Displays the online help.<td>
<tr><td><img src="img/info.png" alt="Info" style="width: 20px;">
	<td>A popup window of quick help.<td>
</table>

<a id="circColors"></a>
<h4>Circle Colors</h4>
Circle Colors <img src="img/colorPalette.png" alt="Color" style="width: 20px;">:
On <ttl>Save</ttl>, the color settings are saved and will be reused on subsequent circle displays.
Since it is not reasonable to set colors for each ribbon, this panel gives options to tailor
the group circle of ribbon colors to your desire.
<table class="tx">
	<tr><td>
	<table class="tx">
	<tr><td><img src="img/colorchooser.gif" alt="Color Wheel" style="width: 20px;">
		<td>Pull-down option <ttl>Two-color all blocks</ttl> :
		This brings up the <ttx>Color Wheel</ttx>, which allows changing the colors
		of the inverted/non-inverted ribbons.
	<tr><td colspan=2>All other pull-down options:
	<tr><td style="white-space: nowrap; vertical-align:text-top;">
	   <ttl>Color Set</ttl><td>Choose from two different sets of 100 colors.
	   <br>The next 4 color options alter the selected set.
	<tr><td  style="white-space: nowrap; vertical-align:text-top;">
		<ttl>Scale</ttl><td>Set the scale between &gt;0 and &lt;1 to darken the colors,
		and &gt;1 to lighten the colors.
	<tr><td  style="white-space: nowrap; vertical-align:text-top;">
		<ttl>Order</ttl><td>Sort the colors so that the blue-green colors take precedence.
	<tr><td  style="white-space: nowrap; vertical-align:text-top;">
		<ttl>Reverse</ttl><td>Sort the colors so that the yellow-red colors take precedence.
	<tr><td  style="white-space: nowrap; vertical-align:text-top;">
		<ttl>Shuffle</ttl><td>Shuffle the colors. Setting a different integer will produce
	a different set.
	</table>
	<td><a href="img/circle-colors.png">
		<img src="img/circle-colors.png" alt="circle-colors" style="border: 1px solid black; width: 320px"></a>
</table>


<!------------------------------------------------------------>
<a id="dotplot_display"></a>
<h2>Dot Plot #1 (Two Genome)</h2>

<table class="tz"><tr>
<td><a href="#dotplot_control">Control Panel</a>
<td><a href="#dotplot_filter">Filter</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<p>
<table>
	<tr>
	<td style="vertical-align:text-top;">
	<ttp>Dot plots</ttp> are available for two-genomes, <a href="#dot3">multiple-genomes</a> and
<a href="#alignment_display_dot">multiple-chromosomes</a>. This describes the two-genome view,
which also applies to the other two views.
<p>Dots represent anchors (<i>hits</i>). A blue box indicates a
	<i>Synteny Block</i>. A chromosome pair is represented by a <i>cell</i>.

	<td><a href="img/dotplot-2g.png">
	<img src="img/dotplot-2g.png"  alt="dotplot-2g" style="border: 1px solid black; width: 550px"></a>
</table>

<table>
	<tr>
	<td style="vertical-align:text-top;">
	Clicking a cell will replace the whole genome view with the <ttp>Dot Plot</ttp> for the chromosome pair,
	as shown on the right.
	<p>From the cell <ttp>Dot Plot</ttp>, display the <ttp>2D</ttp>:
	<table class="tx">
	<tr><td>Click a block once and it will turn beige,
	click it again to view the <ttp>2D</ttp> display of the block.
	Filter: Show <i>block</i> hits
    <tr><td>Or, use the mouse to define a region, which will put a box around it.
		Then click the box to view the <ttp>2D</ttp> display.
	Filter: Show <i>all</i> hits
	</table>
    <td><a href="img/dotplot-2g-cell.png">
    <img src="img/dotplot-2g-cell.png"  alt="dotplot-2g-cell" style="border: 1px solid black; width: 280px"></a>
</table>

<!------------------------------------------------------------>
<a id="dotplot_control"></a>
<h4>Control Panel</h4>

<a href="img/dotplot-control.png"><img src="img/dotplot-control.png"  alt="dotplot controls" style="border: 1px solid black;width: 625px"></a>

<p>The following buttons change the size of the display:
<table  class="tx">
<tr><td><img src="img/home.gif" alt="Home" style="width: 20px;">
     <td>Revert the display back to full genome and its original size.

<tr><td style="white-space: nowrap; vertical-align: text-top;">
    <img src="img/plus.gif" alt="Plus" style="width: 20px;"><img src="img/minus.gif" alt="Minus" style="width: 20px;">
	<td>Increase/decrease the dot plot size.
	<br>&rarr;Change the "1" in the text box to increase the size changed. Values of 1 to 10 are valid.
	<br>&rarr;When showing a cell dotplot, if a block or region is selected (turning color beige),
	the dotplot will keep the beige region in view. Click outside the box to remove the highlighting.
<tr><td style="vertical-align:text-top;"><img src="img/scale.gif" alt="Scale" style="width: 20px;">
	<td>Toggle: Resizes the species displayed on the y-axis according to the current scale of the species displayed on the x-axis.
	Un-toggled: Each species is allocated the same amount of space.
</table>

<p>The following also change the display:
<table class="tx">
<tr><td>Drop-down
		<td>The drop-down menu allows the reference species on the x-axis to be changed.
<tr><td><ttl>Filter</ttl>
		<td>See <a href="#dotplot_filter">Dot Plot filter</a> below.
<tr><td><img src="img/colorchooser.gif" alt="Color Wheel" style="width: 20px;">
		<td>See <a href="#image">Color Wheel, Print and Help</a>.
</table>

<p>Information buttons:
<table>
<tr><td><img src="img/helpSm.png" alt="Help" style="width: 20px;">
	<td>Displays the online help.<td>
<tr><td><img src="img/info.png" alt="Info" style="width: 20px;">
	<td>Popup of quick help.
<tr><td style="vertical-align:text-top;"><img src="img/s.png" alt="Stats" style="width: 20px;">
	<td>Popup with more detailed statistics, i.e. more detail than shown at the bottom
	of the display above. This text can be copied. An example popup is as follows:
	<br>
	<a href="img/dotplot-stats.png"><img src="img/dotplot-stats.png"  alt="dotplot stats" style="border: 1px solid black;width: 650px"></a>
	<br>The <ttx>%Id=31</ttx> is the filtered percent identity of the hits,
	which can be changed in the <ttp>Dot Plot</ttp> <ttb>Filter</ttb>.
	<br>For the 3 <ttx>Annotated</ttx> columns (<ttx>Both, One, None</ttx>),
	the two percentages in parenthesis are the column (<ttx>Annotated</ttx>) percent and row (<ttx>Block Hits</ttx>) percent.

</table>
<!------------------------------------------------------------>
<a id="dotplot_filter"></a> <!-- # ref in code -->
<h4>Dot Plot Filter</h4>

The <ttb>Filter</ttb> settings are used for the duration of the given <ttp>Dot Plot</ttp> display. In contrast,
the <ttx>Color Icon</ttx> settings are stored for subsequent <ttp>Dot Plot</ttp> displays.
<p>
<table>
	<tr><td  style="vertical-align: text-top;">
	As soon as a <ttb>Filter</ttb> option is changed, it will be reflected on the dotplot. However, no other panel can be
	accessed until one of the following buttons is clicked, which will close the <ttb>Filter</ttb> window.
	<p>
	<table class="ty">
		<tr><td><ttb>Save</ttb><td>Keep the current settings.
		<tr><td><ttb>Cancel</ttb><td>Revert back to the settings used when the <ttb>Filter</ttb> window was opened.
		<tr><td><ttb>Defaults</ttb><td>Reset to the defaults, which are shown in the image on the right.
	</table>
	<td><a href="img/dotplot-filter.png">
	<img src="img/dotplot-filter.png" alt="dotplot filter" style="border: 1px solid black; width: 350px"></a>
</table>

<table  class="ty">
	<tr><td style="white-space: nowrap;"><ttl>Hit %Identity</ttl>
		<td>Show all hits that have &gt;N% identity, where the %Identity (%Id) is the MUMmer assigned value.
		The default label on the left starts at the lowest %Id in the plot (in this case 23%).
	<tr><td><ttl>Dot Size</ttl><sup>a</sup>
		<td>Increase the size of the dots in the plot.
	<tr><td style="white-space: nowrap;"><ttl>Scale dot by hit</ttl><sup>a</sup>
		<td>
		<table class="ty txy">
			<tr><td><ttx>Length</ttx><td>The dot is a rectangle relative to the hit's length.
			<tr><td><ttx>%Id</ttx><td> The dot is scaled by the hit's %Id (Identity).
			<tr><td><ttx>None</ttx><td> All dots are the same size.
		</table>
	<tr><td><ttl>Show Hits</ttl>
		<td>
		<table class="ty txy">
			<tr><td><ttx>All</ttx><td>Show all hits.
			<tr><td><ttx>Mix</ttx><td>Show all hits, but non-blocks hits will be displayed as a small dot.
			<tr><td><ttx>Block</ttx><td>Show only hits assigned to blocks.
		</table>
	<tr><td><ttl>Gene Only</ttl>
		<td>
		<table class="ty txy">
			<tr><td><ttx>Ignore</ttx><td>Show all.
			<tr><td><ttx>Both</ttx><td>Only show hits that have a gene on both ends to the hit (g2).
			<tr><td><ttx>One</ttx><td>Only show hits that have a gene on one end of hit (g1).
			<tr><td><ttx>None</ttx><td>Only show hits that do not have a gene on either end (g0).
		</table>
	<tr><td><ttl>Show Block</ttl>
		<td>
		<table class="ty txy">
			<tr><td><ttx>Boundary</ttx><td>Draw a blue rectangle around each block of hits.
			<i>This must be on in order to select a block for <ttp>2D</ttp> display.</i>
			<tr><td><ttx>Number</ttx><td>Draw the block number in or near the block.
		</table>
</table>
<small><sup>a</sup> These options can make the dots flow over the borders,
see <u>Size of dots</u> below.</small>

<p><ttl>Show Empty Regions</ttl> (Not shown in above <ttb>Filter</ttb> image) This option is only present
if at least one of the genomes displayed has &gt;=20 chromosomes or scaffolds; this is most relevant
when one genome is draft sequence, as there can be rows or columns with no blocks.
When the option is deselected, it hides rows or columns that do not contain any synteny blocks.

<p><u>Size of dots</u>
<table><tr><td>
	When the dots are all the same size (<ttl>Scale dot by hit</ttl>=<ttx>None</ttx> and <ttl>Dot Size</ttl>=1),
	they fit inside of their respective block. However,
	when they are changed to rectangles or their sizes increase, they can overlap the border of their block.
	<p>This is illustrated on the right, where the filters are set the <ttl>Length</ttl>, <ttl>Dot Size</ttl>=3.
	The black and grey dots are non-block hits. Some of the blue block hits are on the border of the block blue box.
	<td><a href="img/dotplot-block.png">
	<img src="img/dotplot-block.png" alt="dotplot filter" style="border: 1px solid black; width: 200px"></a>
</table>

<!------------------------------------------------->
<a id="block"></a>
<h2>Blocks (Two Genome)</h2>

The image below shows the <ttp>Block</ttp> display for  Cabbage to A.thal.
The 9 Cabbage chromosomes are color coded as shown at the top, and aligned to
the 5 A.thal chromosomes. The view can be reversed (i.e. A.thal to Cabbage) by
selecting <u>reverse</u>.
<table>
	<tr>
	<td><a href="img/block.png"><img src="img/block.png" alt="block" style="border: 1px solid black; width: 500px"></a>
	<td>Clicking a chromosome displays a window of the chromosome as shown below; the dotted border
	indicates an inverted block (i.e. the majority of hits are inverted). Clicking
	a block displays its <ttp>2D</ttp> view in a popup window.
	<p><a href="img/block-chr.png"><img src="img/block-chr.png"  alt="block chr" style="border: 1px solid black; width: 250px"></a>
</table>

<!------------------------------------------------->
<a id="report"></a>
<h2>Reports... (Two Genome)</h2>
<table>
<tr><td>Select <ttb>Reports...</ttb> which
popups a menu as shown on the right.
<p>The options are <a href="#sum">Summary</a> and <a href="#breport">Block Report</a>.
<td><a href="img/reportMenu.png">
	<img src="img/reportMenu.png" alt="report menu" style="border: 1px solid black; width: 100px"></a>
</table>

<!------------------------------------------------->
<a id="sum"></a>
<h3>Summary</h3>

The <ttb>Explain</ttb> button on this view explains the different columns.
<p><a href="img/sum-2g.png">
	<img src="img/sum-2g.png" alt="sum-two-genome" style="border: 1px solid black; width: 550px"></a>

<a id="breport"></a>
<h3>Block Report</h3>

<table style="width:90%">
<tr><td style="vertical-align: text-top;">

On the <ttp>Manager</ttp> panel, select a pair cell followed by <ttp>Report</ttp>.
This creates a report on the <ttl>Blocks</ttl>.

<p><ttl>Reference</ttl>: The <ttl>Reference</ttl> determines what species chromosomes and start coordinates will be sorted for display.
In the HTML file, the reference columns headings for <ttx>Species</ttx> and <ttx>Chr</ttx> will
be in bold; in the TSV file, they will be in upper case.
<p><ttl>Output</ttl>: The <ttx>Popup</ttx> popups a window with an HTML formated report.
The <ttx>HTML File</ttx> is written in a way that is easily edited, and the <ttx>TSV File</ttx> can be loaded into Excel.
<p><ttl>Combine C1.C2.B</ttl> If checked, the values for <ttx>Chr1</ttx>, <ttx>Chr2</ttx> and <ttx>Block</ttx>
will be one column where the values are separated by ".". If unchecked, these 3 values will have
their own column.
<td><a href="img/report.png">
	<img src="img/report.png" alt="sum-two-genome" style="border: 1px solid black; width: 250px"></a>
</table>
<p>The following 3 options are available, where the top of an HTML table is shown for each:

<p>&#8858; <ttl>Blocks plus</ttl>: <small>(<ttl>Reference</ttl> is A.thal, <ttl>Combined</ttl> unchecked)</small>
<table class='ty' style="font-size: 10px; text-align: right;">
<tr style="background-color: lightgrey;"><td style="text-align: center"><b>Species1</b><td style="text-align: center">Species2<td style="text-align: center"><b>Chr01</b><td style="text-align: center">ChrX<td style="text-align: center">Block<td style="text-align: center">#Hits<td style="text-align: center">PearsonR<td style="text-align: center">AvgGap1<td style="text-align: center">AvgGap2<td style="text-align: center">Len1<td style="text-align: center">Len2<td style="text-align: center">Start1<td style="text-align: center">Start2
<tr><td>A.thal<td>B.rapa<td>01<td>01<td>1<td>26<td>+0.992<td>30,798<td>98,780<td>826,961<td>2,531,385<td>3,104,633<td>17,277,988
<tr><td>A.thal<td>B.rapa<td>01<td>01<td>2<td>8<td>-0.988<td>100,814<td>85,161<td>721,937<td>613,967<td>10,855,813<td>5,391,401
</table>
<p>
<ul>
<li>The <ttx>PearsonR</ttx> is the approximate linearity of the hits in the chain as measured
by the Pearson correlation coefficient; a negative PCC is an inverted block.
<li>The <ttx>AvgGap</ttx> is <tt>sum(gaps-between-genes)/number-gaps</tt>.
<li><ttx>Chr01</ttx> and <ttx>ChrX</ttx>: The column heading are displayed again each time the reference chromosome changes. The
reference chromosome column heading will specify the chormosome (e.g. <ttx>Chr01</ttx>)
where the non-reference will use X for the chromosome as it will change.
</ul>
<p>&#8858; <ttl>Blocks with gene info</ttl>: <small>(<ttl>Reference</ttl> is B.rapa, <ttl>Combined</ttl> checked)</small>
<table class='ty' style="font-size: 10px; text-align: right;">
<tr style="background-color: lightgrey;"><td style="text-align: center">Species1<td style="text-align: center"><b>Species2</b><td style="text-align: center">Cx.<b>01</b>.B<td style="text-align: center">#Hits<td style="text-align: center">PearsonR<td style="text-align: center">Start1<td style="text-align: center">End1<td style="text-align: center">Start2<td style="text-align: center">End2<td style="text-align: center">#Genes1<td style="text-align: center">%Genes1<td style="text-align: center">#Genes2<td style="text-align: center">%Genes2
<tr><td>A.thal<td>B.rapa<td>01.01.1<td>26<td>0.992<td>3,104,633<td>3,931,593<td>17,277,988<td>19,809,372<td>236<td>0.110<td>218<td>0.119
<tr><td>A.thal<td>B.rapa<td>01.01.2<td>8<td>-0.988<td>10,855,813<td>11,577,749<td>5,391,401<td>6,005,367<td>159<td>0.050<td>102<td>0.078
</table>

<p>The <ttx>Genes1</ttx> column is the number of genes from <ttx>Species1</ttx> in the block and <ttx>%Genes1</ttx>
is the percentage of genes that have a hit.


<p>&#8858; <ttl>Blocks summary</ttl>: <small>(<ttl>Reference</ttl> does not matter, <ttl>Combined</ttl> unchecked)</small>
<table class='ty' style="font-size: 10px; text-align: right;">
<tr style="background-color: lightgrey;"><td><b>Species1</b><td>Species2<td><b>Chr1</b><td>Chr2<td>#Blocks<td>#Hits
<tr><td>A.thal<td>B.rapa<td>01<td>01<td>8<td>412
</table>
<p>The <ttx>#Blocks</ttx> and <ttx>#Hits</ttx> are the sum of blocks and hits for the chromosome pair.
<!--------------------------------------------------------------------------------------->
<!------------------------------------------------------------------------------------->
<!------------------------------------------------------------------------------------->
<hr>
<a id="all"></a>
<h1>All Pairs</h1>

<table class="tz"><tr>
<td><a href="#explorer">Chromosome Explorer</a>
<td><a href="#dot3">Dot Plot #3</a>
<td><a href="Query.html" class="ext" target="_blank">Queries</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>
<table>
<tr><td>From the <a href="#mgr">Project Manager</a>, select two or more projects, which will
activate the <ttb>All Pairs</ttb> buttons.

<td><a href="img/view1.png">
	<img src="img/view1.png" alt="Displays for all projects" style="border: 1px solid black; width: 350px"></a>

</table>
<a id="explorer"></a>
<h2>Chromosome Explorer</h2>

<table class="tz"><tr>
<td><a href="#alignment_display_circ">Circle #2</a>
<td><a href="#alignment_display_dot">Dot plot #2</a>
<td><a href="#alignment_display_2d">2D Display</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>
The <ttp>Explorer</ttp> is a two-panel display in which the left panel is used to select specific chromosomes,
while the right panel shows the synteny for the selected chromosomes.
Below is a snapshot of the <ttp>Explorer</ttp> showing the <ttp>Circle</ttp>, which is the initial view.
<p>
<a href="img/explorer.png">
	<img src="img/explorer.png" alt="explorer" style="border: 1px solid black; width: 650px"></a>

<p><u>Left panel:</u>The left panel controls which species and chromosomes are shown:
<ul>
<li>Click a chromosome rectangle to add it to the display in the right panel. Click it again to remove
it.
<li>Click the chromosome number above a chromosome to make it the <i>reference</i>. (The
choice of reference sequence does not matter much for the <ttp>Circle</ttp> view,
but it is important for the <ttp>2D</ttp> and <ttp>Dot Plot</ttp> views; see below).
<li> Adding and removing chromosomes changes the <ttp>Circle</ttp> view instantly.
To add/remove from the <ttp>2D</ttp> or <ttp>Dot Plot</ttp>
views, open the <ttp>Circle</ttp> view, make the change, and then re-open the <ttp>2D</ttp> or <ttp>Dot Plot</ttp>.
</ul>
<p>
<ttl>Information</ttl>: As the mouse moves over various
components, instructions or information for that component are shown in the <ttl>Information</ttl> box.
<p>
<ttl>Views:</ttl> The buttons in the lower-left corner of the window change the synteny view in the right panel.

<p>
Click the minus (<ttl><big>-</big></ttl>) button next to a project name or the <ttl>Information</ttl> box,
and that section will be hidden;
the minus sign will change to a plus.
Click the plus (<ttl><big>+</big></ttl>) button to show the project or <ttl>Information</ttl> box.

<p><u>Right panel:</u>
The right panel shows the synteny display for the species and chromosomes selected on the left.
The three view for the right panel are
<a href="#alignment_display_circ">Circle</a>,
<a href="#alignment_display_2d">2D</a> and
<a href="#alignment_display_dot">Dot Plot</a>.
The <ttp>Circle</ttp> and <ttp>Dot Plot</ttp> are
essentially overviews, while the <ttp>2D</ttp> view allows zooming down to details, all the way
to the basepair level.

<!----------------------------------------------------------------->
<a id="alignment_display_circ"></a>
<h3>Circle #2 (Multi-chromosome)</h3>

This view is illustrated in the above <a href="#explorer">Chromosome Explorer</a> image.
All features are described in the first <a href="#circle">Circle section</a>,
except it is chromosome-based instead of genome-based.

<!----------------------------------------------------------------->
<a id="alignment_display_dot"></a>
<h3>Dot Plot #2 (Multi-chromosome)</h3>

All features are described in the first <a href="#dotplot_display">Dot Plot section</a>,
except that it is chromosome-based instead of genome-based.
Selecting a chromosome pair (cell) from the image on the left will replace it with the
<ttp>Dot Plot</ttp> for the cell, shown on the right. Clicking on a synteny block (or selected region)
highlights it in beige, clicking it again brings up the <ttp>2D</ttp> view.
<table>
<tr><td><a href="img/dotplot-3chr.png">
	<img src="img/dotplot-3chr.png" alt="Dot Plot 3chr" style="border: 1px solid black; width: 400px"></a>
<td><a href="img/dotplot-chr-2d.png">
	<img src="img/dotplot-chr-2d.png" alt="dotplot-chr-2d" style="border: 1px solid black; width: 360px"></a>
</table>

<!-------------------------------------------------------------------->
<a id="alignment_display_2d"></a>
<h3>2D Display (Multi-chromosome)</h3>

<table class="tz"><tr>
<td><a href="#2dfull">Full<br>display</a>
<td><a href="#view2D">View<br>a region</a>
<td><a href="#info">Hovers<br>and popups</a>
<td><a href="#control">Control<br>panel</a>
<td><a href="#sequence_filter">Sequence<br>Filter</a>
<td><a href="#hitfilter">Hit<br>Filter</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<a id="2dfull"></a>
<h3>Full display</h3>
The <ttp>2D</ttp> display is activated from the <a href="#explorer">Explorer</a> by
clicking the <ttb>2D</ttb> button (or by selecting a region or block from a <ttp>Dot Plot</ttp>).
<p>
Below is an image depicting an alignment from A.thal Chr04 to Cabbage Chr01 to B.rapa Chr02.
Note that Cabbage Chr1 is the <i>reference</i>
chromosome, hence is placed in between the others.
<p>Each chromosome is drawn as a light-blue rectangle, called <i>tracks</i>.
The sequence length is shown at the bottom of a track; the coordinates are displayed on the side.
If the sequence has been flipped (i.e. A.thal Chr04), its sequence length is shown at the top of the track.

<p>
<table>
<tr><td style="vertical-align:text-top;">
<p>The <i>gene annotations</i> (if exists) are drawn down the middle of the rectangle in dark-blue (positive strand)
and purple (negative strand). Overlapping genes are horizontally staggered.

<p>The <i>hits</i> are the lines between tracks, referred to <i>hit-wires</i>, which connect the hit region
on each sequence.
Hit-wires colored brown align to the same strand on both sequences (i.e. ++,--) and light forest-green
align to different strands (i.e. +-, -+).
<p>
The display colors can be changed using the <a href="#wheel">Color Icon</a> in the upper right
of the <ttl>Control Panel</ttl> (shown at the top of the 2D display).
The tracks and hits can be filtered; see
 <a href="#sequence_filter">Sequence Filter</a> and  <a href="#hitfilter">Hit Filter</a>.
<td><a href="img/2d-3chr.png">
	<img src="img/2d-3chr.png" alt="2d-3chr" style="border: 1px solid black; width: 550px"></a>
</table>

<p>
<u>Graphics</u>: Data from the input <tt>.gff</tt> files may specify the following types for display:
<table class="ty" style="width:650;">
	<tr><td><i>Type</i><td><i>Display</i>
	<tr><td>Gaps<td>Red band across the chromosome
	<tr><td>Centromere<td>Blue "X" across the chromosome
	<tr><td>Predicted genes and exons
		<td>Annotation strip in center of chromosome, as described above.
</table>

<!------------------------------------------------------->
<a id="view2D"></a> <!-- used in SystemHelp -->

<h3>View a region</h3>

The selected region shown in the image below was created by left-clicking
on one chromosome near coordinate 11,638.48kb, dragging the mouse to 11,642.75kb and releasing.
The <ttb>Sequence Filter</ttb> was set to display <ttl>Annotations</ttl>.
<p>
<table>
<tr><td  style="vertical-align:text-top;">
Genes:
The introns are drawn as thinner black rectangles. The exons are drawn as thicker rectangles,
colored dark blue (positive strand) or purple (negative strand).
<p>The staggered genes, where one is
further out then the other, indicate overlapping genes; they have the same gene# with a different suffix.
<p>Hits: The <i>hit-wires</i> connects the aligned regions on the two sequences.
The aligned region graphics, located on the track at the end of the hit-wire, extends the length of the hit.
<p><i>Clustered hits</i> are multiple <i>subhits</i> clustered together with gaps in between,
where the gaps are represented by a grey area.

<td><a href="img/2d-closeup2.png">
	<img src="img/2d-closeup2.png" alt="2d-closeup" style="border: 1px solid black; width: 550px"></a>
</table>
<i>Major and Minor genes</i>: If one end of the hit-wire aligns to multiple genes, only one gene is assigned to
it, which is called the <ttx>major</ttx> gene. The other genes are the <ttx>minor</ttx> genes.
In the above example, gene 1910.b is the major gene for the top hit-wire and gene 1910.a is the minor gene.

<!------------------------------------------------------------------>
<a id="info"></a>
<h3>Hovers and popups</h3>
<table class="tz"><tr>
<td><a href="#hinfo">Hit Details</a>
<td><a href="#ginfo">Gene Details</a>
<td><a href="#binfo">Hit-Gene Details</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<p>Hovering the mouse over a hit-wire or gene shows information about them in the <ttl>Information</ttl> box, either
the box on the left side of the window (full <ttp>2D</ttp>) or at the bottom of the window (<ttp>2D</ttp> from <ttp>Dot Plot</ttp>).

<p>Right clicking on a hit-wire or gene popups a window with additional information.
The information in a popup can be copied by dragging the mouse over it followed by the copy command.
The popup does not go away until the <ttb>OK</ttb> button is selected or the <ttp>SyMAP</ttp> is exited.
Any number of gene popups can be present.
<p>Note: On Linux Ubuntu, the popup information windows will not stay in front, so get hidden behind the
<ttl>Explorer</ttl> window; they can be moved to the side.
<!------------------------------------------------------------------>
<a id="hinfo"></a>
<h4>Hit Details</h4>

<u>Information:</u> Hovering the mouse over a hit-wire will show its information in the <ttl>Information</ttl> box,
plus show similar information in the hit popup.
<p>
<table class="txy">
	<tr><td style="vertical-align: text-top;">
	<table class="txy">
	<tr><td><ttx>Hit #</ttx><td> The number representing the hit for the chromosome pair.
	<tr><td style="white-space: nowrap; vertical-align: text-top;">
	<ttx>gN</ttx> (<ttx>N={0,1,2}</ttx>)
	<br><ttx>XX</ttx> (<ttx>X={E,I,n}</ttx>)
	<td> <ttx>Algo1</ttx>: Hit overlaps 0 gene (g0), 1 gene (g1), 2 genes (g2).
	<br> <ttx>Algo2</ttx>: Hit overlaps two exons (EE), exon-intron (EI), intron-intron (II), exon-intergenic (En),
	intron-intergenic (In).
	<br> See <a href="#term2">Terminology</a> for an explanation of <ttx>Algo1</ttx> versus <ttx>Algo2</ttx>.

	<tr><td><ttx>Block #</ttx><td>The block number (0 indicates it is not in a block).
	<tr><td><ttx>Inv</ttx><td> Indicates that it is part of an inverted block.

	<tr><td  style="white-space: nowrap; vertical-align: text-top;"><ttx>cN.M</ttx> (<ttx>N&ge;2</ttx>)
	<td> It is part of a collinear set of size N; M is the unique set number for the chromosome pair.

	<tr><td  style="white-space: nowrap; vertical-align: text-top;">
	<ttx>Id</ttx>, <ttx>Sim</ttx>
	<td>These values are from the MUMmer file.
	For <ttx>Subhits&gt;1</ttx>, the values are approximate.

	<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttx>Cov</ttx>
	<td>The largest sum of the merged subhits between the two species.

	<tr><td  style="white-space: nowrap; vertical-align: text-top;">
	<ttx>Subhits</ttx>=(<ttx>N&ge;1</ttx>)
	<td> (Popup only) The number of MUMmer hits clustered into one hit, where there may be gaps
	or overlaps between hits.
	<tr><td>&nbsp;
	<tr><td style="white-space: nowrap; vertical-align: text-top;">
	<ttx>L</ttx> and <ttx>R</ttx><td>The coordinates for the hit on the left and right tracks, respectively.
<tr><td>&nbsp;
	<tr><td style="white-space: nowrap; vertical-align: text-top;">
	<ttx>L#</ttx> and <ttx>R#</ttx><td> (Hover only) The hover shows the <ttx>Gene#</ttx> whereas the popup shows additional
	hit information along with the <ttx>Gene#</ttx>.
</table>

<td><a href="img/2Dhitinfo.png"><img src="img/2Dhitinfo.png" alt="2d-hit information" style="border: 1px solid black; width: 230px"></a>
<br><a href="img/2Dhitwire.png">
<img src="img/2Dhitwire.png" alt="2D hit wire popup" style="border: 1px solid black; width: 290px"></a>
</table>

<p><u>Popup</u>: To view the popup of a hit-wire, right click on it;
the hit-wire must be highlighted in red from hovering over it to be clickable.
The hit-wire and the major genes (0,1 or 2) will be highlighted
magenta as shown on the upper right, and the hit-wire will be drawn as a thicker line;
the corresponding full popup is shown below.

<table>
<tr><td>
The top part of the popup is similar to the hit information shown in the image above.
The bottom part shows the coordinates of the subhits:
	<ol>
	<li>If the hits align to a gene at either end, the major gene(s) are stated.
	<li>The coordinates are relative to the input genome sequences.
	<li>Both lists are sorted by start coordinate.
	<li>The '#' column in the two subhit tables align to each other,
	e.g. the rows #2 coordinates align to each other (a MUMmer hit).
	<li>
	The sequence name that is alphabetically lower (e.g. arab&lt;cabb) is the <i>query</i> and the
	other is the <i>target</i>. The query is numbered 1-N; the target is ordered to match the first.
	<li>Two subhits may overlap (Gap&lt;=0) on one chromosome but not the other, as shown on the right
	and in <u>Gap vs no-gap</u> below.
	</ol>
<td>
<a href="img/2Dhitpopup.png">
<img src="img/2Dhitpopup.png" alt="hit popup" style="border: 1px solid black; width: 250px"></a>
</table>
The lower buttons are as follows:
<table class="ty">
<tr><td><ttl>Align</ttl><td>Select to view the subhits text alignment. See <a href="#btext">Align</a>.
<tr><td colspan=2>Conditional:
<tr><td><ttl>Merge</ttl><td>If there are overlapping hits (Gap&lt;=0), this button will be present.
	<br>Select to view a popup of the merged hits.
	See <a href="Details.html#hit" class="ext" target="_blank">Hits</a>.
<tr><td><ttl>Order</ttl><td>If the '#' column for the target genome (e.g. cabb) are not in sequential order,
	this button will be present.
	<br>The <ttl>Order</ttl> function will sort by '#' and
	then the subhits will not be in sorted order; these are referred to as <i>disordered</i> subhits.
	See <a href="Details.html#hit" class="ext" target="_blank">Hits</a>.
</table>

<p><u>Gap vs no-gap</u> Cabbage shows the distance between #1 and #2 as -35.
Viewing the Cabbage alignment, #327.1 and #327.2 overlap but they do not overlap for A.thal.

<table><tr>
<td><small>Selected Cabbage alignment</small>
<br><a href="img/2Dhitalign2.png">
<img src="img/2Dhitalign2.png" alt="Align graphics" style="border: 1px solid black; height: 200px"></a>
<td><small>Selected A.thal alignment</small>
<br><a href="img/2Dhitalign1.png">
<img src="img/2Dhitalign1.png" alt="Align graphics" style="border: 1px solid black; height: 200px"></a>
</table>

<!------------------------------------------------------------------------------->
<a id="ginfo"></a>
<h4>Gene Details</h4>

<u>Information</u>: Hovering the mouse over a gene will show its information in the <ttl>Information</ttl> box.
<table>
<tr><td>
<table class="txy">
<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttx>Gene #</ttx>
<td>is a sequential number along the chromosome. If genes overlap, they will have the same number
but a different suffix (e.g. a, b,...).
<tr><td><ttx>#Exon=</ttx><td>is the number of exons followed by their summed sizes.
<tr><td colspan=2>The next line is the location followed by its total length.
<tr><td colspan=2>The lines (<tt>ID</tt>...) are the annotation provided by the GFF file.
<tr><td colspan=2>If the mouse is over an exon, the last line will be number, e.g. <ttx>Exon #4</ttx>.
</table>
<td>
<a href="img/2Dgeneinfo.png"><img src="img/2Dgeneinfo.png" alt="2d gene information" style="border: 1px solid black; width: 260px"></a>
</table>

<table>
<tr>
  <td>
  <a href="img/2Dgeneview.png"><img src="img/2Dgeneview.png" alt="2d gene graphics" style="border: 1px solid black; width: 30px"></a>
	<td>&nbsp;&nbsp;
  <td><u>Popup:</u> To view the popup of the gene description, right-click on the gene in the track; the selected gene will turn cyan.
   <ul>
   <li style="margin: 1em 0;">The top text is the same as shown in the gene information.
   <li  style="margin: 1em 0;">All hits for the gene for the chromosome pair will be listed.
   In this case, the gene has two hits, where the first is a minor hit to the gene and is not part of a block
   (<i>if the <ttx>All hits</ttx> filter is not on, this hit will not be shown</i>).
   <li  style="margin: 1em 0;"><ttx>Algo1</ttx>: Only has (Gene N%).
   This is the <tt>hit-overlap/gene-length</tt>, without regard to gaps or introns.
   <br><ttx>Algo2</ttx>: Has (Gene N% Exon N%). The gene overlap will be the same
   as in <ttx>Algo1</ttx>. The exon overlap is the amount of hit that overlap exons divided by the sum
   of the exons.
   </ul>

  <td><a href="img/2Dgenepopup.png"><img src="img/2Dgenepopup.png" alt="2d gene popup" style="border: 1px solid black; width: 260px"></a>
</table>

<!----------------------------------------------------------->
<a id="binfo"></a>
<h4>Hit-Gene Details</h4>

The example below shows the two genes that are connected by the hit-wire. For the gene popups,
the exons are numbered differently; this is due to differences between NCBI and Ensembl annotation files.
<p><a href="img/2d-desc.png"><img src="img/2d-desc.png" alt="2d-desc" style="border: 1px solid black; width: 750px"></a>

<hr>
<!----------------------------------------------------------->

<a id="control"></a>
<h3>Control panel and navigation</h3>

<table class="tz"><tr>
<td><a href="#hist">History</a>
<td><a href="#mouse_function_selector">Zoom and Show</a>
<td><a href="#rem">Color wheel, etc</a>
<td><a href="#navi">Navigate</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>
<a href="img/2d-control.png"><img src="img/2d-control.png" alt="2d controlpanel"></a>
<p>The buttons are described in the order found in the control panel image above.
<p>
<a id="hist"></a>
<u>History Event</u>: SyMAP retains a record of the prior views (like a web browser).
<ul>
<li>A History Event occurs when the coordinates or tracks have
been changed by (1) any of the four middle buttons (-,+,shrink,scale),
(2) select a region with the mouse, (3) use the <ttl>Sequence Filter</ttl> to change coordinates.
<li>Any annotation, highlighting, or hit-wire display changes
by <ttl>Sequence</ttl>  or <ttl>Hit Filter</ttl> are NOT a history events.
<li>When a new coordinate change is applied,
     it becomes the next history event in the list and all previous history events are cleared.
<li>The first 3 buttons describe below apply to the history events.
</ul>
<table  class="tx">
<tr><td style="vertical-align: text-top;"><img src="img/home.gif" alt="Home" style="width: 20px;">
     <td>Go back to its original display.
<tr><td><img src="img/back.gif" alt="Back button" style="width: 20px;">
	<td>Go back one history event.
<tr><td><img src="img/forward.gif" alt="Forward button" style="width: 20px;">
	<td>Go forward one history event.
<tr><td><img src="img/plus.gif" alt="Plus" style="width: 20px;">
	<td>Doubles the base pair view region, keeping the same center.
<tr><td><img src="img/minus.gif" alt="Minus" style="width: 20px;">
	<td>Shrinks the base pair view region by 50%, keeping the same center.
<tr><td style="vertical-align: text-top;"><img src="img/shrink.png" alt="shrink" style="width: 20px;">
	<td>The shrink button reduces the space between tracks. This is useful
when viewing more than two tracks, e.g. click <a href="img/Exp_2D_B02.png">Exp_2D</a>.
<tr><td style="vertical-align: text-top;"><img src="img/scale.gif" alt="scale" style="width: 20px;">
	<td>The scale button resizes the tracks so that they are in the same scale
(base pairs per pixel) as the reference sequence track in the view. This is a toggle button,
so click it again to undo the scale.
</table>

<a id="mouse_function_selector"></a>
<p><ttl>Zoom</ttl> and <ttl>Show</ttl>: These two buttons performs a function on the selected region.
<ul>
<li>Check one of the buttons, then click the button and select from the menu (unless the default is desired).
<li>Select a region by
clicking the mouse left button on a location of the chromosome, drag and release - the function
will be performed when the mouse is released.
<li>The functions are described as follows:
</ul>
<table class=tx>
<tr><td><ttl>Zoom</ttl>
<td>
	<table class="ty txy">
	<tr><td><i>Function</i><td><i>Description</i><td><i>Button label</i>
	<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>Zoom All Tracks</ttl>
		<td>
			Zoom into the selected region and all other tracks with hits in the region.
			<p>Note: If the selection clips part of the hit, the hit-wire will be shown at the top or bottom of
			the chromosome track. The hit-wire of the clipped hit can be highlighted just like a fully shown hit.
		<td><ttl>Zoom</ttl>
	<tr><td style="white-space: nowrap;"><ttl>Zoom Selected Track</ttl>
		<td>Zoom into the selected region; all other tracks do not change.
		<td><ttl>Zoom Select</ttl>
	</table>
<tr><td><ttl>Show</ttl>
<td>
	<table class="ty txy">
	<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>Align (Max 50kb)</ttl>
		<td>Open the base alignment view for the selected region (see <a href="#base">Base Align</a>).
		There MUST be at least a partial hit in the region to align.
		<td><ttl>Show</ttl>
	<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttl>Show Seq Options</ttl>
		<td>The allows viewing the actual basepair sequence in a track.
		A sequence must first be selected, then a popup menu will have the options:
			<table class="ty">
			<tr><td><ttx>Hit</ttx><td>Show the sequences of the clustered subhits on the positive strand.
			If the hit is on the negative strand, it will be reverse complemented (RC).
			<tr><td><ttx>Reverse Hit</ttx><td>Show the sequences of the subhits on the negative strand.
			If the hit is on the positive strand, it will be reverse complemented.
			<tr><td><ttx>Full Hit</ttx><td>Concatenate and show the sequence from the start of the first
			cluster hit to the end of the last cluster subhit.
			<tr><td><ttx>Region</ttx><td>Show the sequence from the region selected.
			</table>
		<td><ttl>Show Seq</ttl>
		</table>
</table>

<a id="rem"></a>
<p><u>Color wheel, etc:</u>
<table class="tx">
<tr><td><img src="img/colorchooser.gif" alt="Color Wheel" style="width: 20px;">
		<td>See <a href="#image">Color Wheel, Print and Help</a>.
<tr><td><img src="img/print.gif" alt="print" style="width: 20px;">
		<td>Prints the graphics 2D region to an image file (note, this does not include the
		control bar, etc).

<tr><td><img src="img/helpSm.png" alt="Help" style="width: 20px;">
	<td>Displays the online help.
<tr><td><img src="img/info.png" alt="Info" style="width: 20px;">
	<td>Popup of quick help.
</table>

<a id=navi></a>
<p>
<u>Mouse navigation</u>:
<table  class="ty txy">
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">View region&nbsp;
	<td style="white-space: nowrap; vertical-align: text-top;">Drag left mouse button
	<td>Press down on the left mouse button on a chromosome track, drag the mouse and release it.
	By default, this will zoom into the selected region for both chromosomes.
	This action can be changed with the <ttl>Zoom</ttl> and <ttl>Show</ttl> as discussed above.
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">Resize Track&nbsp;</td>
	<td style="white-space: nowrap; vertical-align: text-top;">Drag bottom of track</td>
	<td>Position mouse at bottom of track (resize cursor appears), hold down left
	mouse button, and move mouse.</td></tr>
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">Scroll One Track&nbsp;</td>
	<td style="white-space: nowrap; vertical-align: text-top;">Mouse wheel</td>
	<td>Position mouse over track and use mouse wheel to scroll up/down the track.

	</td></tr>
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">Scroll Both Tracks&nbsp;</td>
	<td style="white-space: nowrap; vertical-align: text-top;">Mouse wheel</td>
	<td>Position mouse in the hit space to scroll up/down both tracks.
	</td></tr>
<tr>
	<td style="vertical-align: text-top;">Popup of partial set of filters&nbsp;</td>
	<td style="white-space: nowrap; vertical-align: text-top;">Right mouse button</td>
	<td>Position mouse over track area (for sequence filter) or white space between tracks (for hit filter),
	and click right mouse button.
	</td></tr>
</table>

<!----------------------------------------------------------->
<a id="sequence_track"></a> <!-- # ref in fpc html -->
<a id="sequence_filter"></a> <!-- # ref in code -->
<h3>Sequence Filter</h3>

<table class="tz"><tr>
<td><a href="#sfhigh">Highlight and Show</a>
<td><a href="#sfgrap">Graphics</a>
<td><a href="#sfhit">Hit Text</a>
<td><a href="#sfcoor">Coordinates</a>
<td><a href="#sfg2">3-track</a>
<td><a href="#sfpop">Popup</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

Select <ttb>Sequence Filter</ttb> button above the sequence track to change the display for the
respective chromosome. If the sequence track does not have a given annotation (e.g. Gaps), then that item
will be un-selectable.
<p>&rarr; All options in the first 3 sections occur immediately; <ttx>Coordinates</ttx> changes occur on <ttl>Save</ttl>.
<br>&rarr; Only <ttx>Coordinates</ttx> changes are <a href="#hist">History events</a>.
<p>
<a id="sfhigh"></a>
<table>
<tr><td>
<table class="ty">
<tr><td colspan="2"><ttl>Highlight</ttl>
<tr><td><ttx>Gene popup</ttx>
<td> Highlight the gene when its <a href="#ginfo">Gene Popup</a> is displayed.
The highlight color can be changed with the <a href="#wheel">Color Wheel</a>.

<tr><td colspan="2"><ttl>Show</ttl>
<tr><td style="vertical-align: text-top;"><ttx>Gene#</ttx><td>Display the <ttx>Gene#</ttx>
beside each gene. This is the SyMAP assigned number;
see <a href="#term2">Terminology</a> for more details.

<tr><td style="vertical-align: text-top;"><ttx>Annotation</ttx>
<td> Display of the annotation descriptions along the right side of sequence
(see <a href="img/2d-closeup2.png">example</a>).
<br>NOTE: this only works if you are zoomed in close enough that the annotation boxes can clearly be displayed.

<tr><td style="vertical-align: text-top;"><ttx>Has Hit</ttx><td>
This <ttx>Gene#</ttx> and <ttx>Annotation</ttx> will only show if the gene has a hit.

<tr><td style="vertical-align: text-top;"><ttx>Gene delimiter</ttx><td> When the zoom is close enough to view the exons, a black line will
be drawn over the top end of each gene. This helps delimit genes that do not overlap (so are not staggered),
but are close enough to appear one gene (see <a href="#ex2d">example below</a>).
</table>
<td>
<a href="img/seqFilter.png">
	<img src="img/seqFilter.png" alt="sequence filter" style="border: 1px solid black; width: 350px"></a>
</table>

<a id="sfgrap"></a>
<p>
<table class="ty">
<tr><td colspan="2"><ttl>Graphics</ttl>
<tr><td><ttx>Ruler</ttx>
  <td>Display of the ruler along the side of the sequence track.

<tr><td style="vertical-align: text-top;"><ttx>Gaps</ttx><sup>1</sup>
   <td>Display of sequence gaps (drawn as solid red rectangles) along the sequence.
<tr><td style="vertical-align: text-top;"><ttx>Centromere</ttx><sup>1</sup>
   <td> Display of the centromere (drawn as a cyan "X") on the sequence.
   <tr><td><ttx>Genes</ttx>
<td> Display of the gene with exons along the sequence. (If unchecked,
the other gene options are disabled).
<tr><td style="vertical-align: text-top;"><ttx>Hit length</ttx>
   <td> Display the hit length line along the inner boundary of the sequence. The subhits are colored
   brown and the gaps are colored gray (see <a href="img/2d-closeup2.png">example</a>).
<tr><td style="vertical-align: text-top;"><ttx>Hit %Id bar</ttx>
   <td>The placement of the <ttx>Hit length</ttx> line is proportional to the <ttx>%Identity</ttx>.
</table>
<sup>1</sup> The <ttx>Gaps</ttx> and <ttx>Centromere</ttx> options are disabled in the image above, as there are none for this chromosome.
<p>

<a id="sfhit"></a>
<table class="ty">
<tr><td><ttl>Hit Text</ttl><td> Show the value along the edge of the sequence track over
the hit-wire near the edge of the track. A value of '0' is not shown.
<br><b>Note</b>: The code will not draw text if it will be on top of another; however,
this gets messed up when there are inverted hits among non-inverted hits.
<td>Symbol
<tr><td><ttx>Block# 1st</ttx><td>Put the block number above the first hit of the visible block. <td>&nbsp;&nbsp;bN
<tr><td><ttx>Block# All</ttx><td>If the hit is in a block, its block number will be shown. <td>&nbsp;&nbsp;bN
<tr><td><ttx>Collinear#</ttx><td>If the gene-hit is assigned to a collinear set,
its collinear set number will be shown. <td>&nbsp;&nbsp;cN
<tr><td><ttx>Hit#</ttx><td>The number assigned by SyMAP for the hit.
<br>These are numbered sequentially
along the chromosome with the alphabetically lesser name; see <a href="#term2">Terminology</a>.<td>&nbsp;&nbsp;#N
<tr><td><ttx>Hit %Id</ttx><td>The score value corresponds to the %Identity value
for the hit from the MUMmer anchor file. <td>&nbsp;&nbsp;N%
<tr><td><ttx>None</ttx><td>Nothing is shown (default).<td>&nbsp;
</table>

<a id="sfcoor"></a>
<br>
<table class="ty">
<tr><td><ttl>Coordinates</ttl><td> All the following coordinate changes take effect on <ttb>Save</ttb>.
<tr><td><ttx>Flip sequence</ttx><td>Reverses the orientation of the sequence track.
<tr><td><ttx>Full</ttx><td> Sets the start and end positions of the sequence display to encompass the whole chromosome.
<tr><td>
<ttx>Start</ttx> and <ttx>End</ttx><td>
The positions of the sequence display can be set via the corresponding text boxes.
The units of the values entered can be selected from the accompanying drop down menus
(BP, KB, MB, GB).
<tr><td><ttx>Gene#</ttx><td>
&bull; Select the checkbox and enter a <ttx>Gene#</ttx> assigned by SyMAP. It
must be exact, i.e. if it has a suffix, the suffix must be included (e.g. 1234.b).
<br>&bull; When <ttb>Save</ttb> is selected, the 30kb region around the gene will be shown and
the exons of the <ttx>Gene#</ttx> will be highlighted.
<br>&bull; BEWARE: If the <a href="#hitfilter">Hit Filter</a> <ttx>All hits</ttx> filter is not turned on,
you may not see the corresponding gene hits as they may be filtered.
</table>

<a id="sfg2"></a>
<p><u>3-track</u>:
The <ttb>Sequence Filter</ttb> has two extra options, <ttx>g2g2</ttx> and <ttx>g2g1</ttx>,
for the <i>reference</i> track when it has both a left and right track.
The usage of this feature is complex, see
<a href="Details.html#2dg2" target="_blank" class="ext">3-chr conserved</a> for details.
<table>
<tr><td>
<table class="ty">
<tr><td><ttx>g2x2</ttx><td>The genes conserved across the 3-chromosomes are highlighted along with the hit-wires.
<tr><td><ttx>g2x1</ttx><td>The genes conserved in one chr-pair are highlighted along with the hit-wires.
<tr><td><ttx>Show</ttx><td>Only the g2x2 or g2x1 hit-wires will be shown.
</table>

<td><a href="img/seqFilterRefOnly.png">
	<img src="img/seqFilterRefOnly.png" alt="sequence filter reference" style="border: 1px solid black; width: 300px"></a>
</table>

<a id="sfpop"></a>
<p>
<table class="tx" style="width: 95%">
<tr><td class="coltop">
<u>Sequence Filter Popup</u>: A subset of the filters are available by right clicking in the empty space
of the chromosome rectangle, as shown on the right.
<p>&nbsp;
<p>During a <ttp>Chromosome Explorer</ttp> session,
<ttl>Sequence Filters</ttl> are reset to defaults between
different <ttp>2D</ttp> displays.


<p>&nbsp;
<p><i>Highlighting gene summary</i>:
<table class="tx txy">
<tr><td style="white-space: nowrap;">Filter <ttx>Gene#</ttx><td>The exons, introns and gene delimiter are highlighted.
<tr><td style="white-space: nowrap;">Gene popup<td>The exons, introns and gene delimiter are highlighted.
<tr><td>Hit popup<td>The exons, introns and gene delimiter are highlighted the same color as the hit-wire color.
<tr><td>g2x2, g2x1<td>(Reference, 3-track only) The exons are highlighted.
</table>

<td>
<a href="img/seqFilterPopup.png">
	<img src="img/seqFilterPopup.png" alt="sequence filter popup" style="border: 1px solid black; width: 180px"></a>
</table>

<!----------------------------------------------------------->
<a id="hitfilter"></a> <!-- # ref in code -->
<h3>Hit Filter</h3>

<table class="tz"><tr>
<td><a href="#hfhigh">Highlight</a>
<td><a href="#hfshow">Show</a>
<td><a href="#hfother">Other</a>
<td><a href="#hfpop">Popup</a>
<td><a href="#ex2d">Example</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>
<p>The <ttb>Hit Filter</ttb> menu allows the user to select which types of hits are displayed.
Multiple <ttl>Show</ttl> options can be selected, but only one <ttl>Highlight</ttl>.

<a id="hfhigh"></a>
<table>
<tr><td>
	<table class="ty">
	<tr><td colspan=2><ttl>Highlight</ttl>
	<tr><td><ttx>Synteny blocks</ttx><td>Hits that are part of a synteny block.
	<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttx>Collinear sets</ttx>
		<td>Hits that are part of a collinear set. See <a href="#ex2d">example below</a>
			and <a href="Details.html#col" target="_blank" class="ext">Details</a>.
	<tr><td><ttx>Hit =2 genes</ttx> <td>Hits that align to genes on both ends (g2).
	<tr><td><ttx>Hit =1 genes</ttx> <td>Hits that align to a gene on one end (g1).
	<tr><td><ttx>Hit =0 genes</ttx>   <td>Hits that do not align to any gene (g0).

	<tr><td><ttx>Hit popup<br>(or Query)</ttx> <td>
	The following two situations are highlighted with the <ttx>Popup-Query</ttx> color:
	<br>(1) When a <a href="#hinfo">Hit Popup</a> is displayed, the hit-wire and any aligned major genes are highlighted.
	<br>(2) When the <ttb>2D</ttb> is displayed with <ttb>Query</ttb>
	<a href="Query.html#view2D" class="ext" target="_blank">View 2D</a> option, the selected hit(s)
	will be highlighted.
	</table>
<td><a href="img/hitfilterS.png">
	<img src="img/hitfilterS.png" alt="hitfilterS" style="border: 1px solid black; width: 300px">
<tr><td>
</table>

<p>
<a id="hfshow"></a>
<table  class="ty">
<tr><td colspan=2><ttl>Show</ttl>
<tr><td><ttx>Synteny blocks</ttx><td>Hits that are part of a synteny block.
	<table class="ty">
	<tr><td><ttx>And</ttx><td>The hit will be shown if it has
	<u>all</u> the selected <ttx>Show</ttx> options.
	<tr><td><ttx>Or</ttx><td>The hit will be shown if it has <u>any</u> of the selected <ttx>Show</ttx> options.
	</table>
<tr><td style="white-space: nowrap; vertical-align: text-top;"><ttx>Collinear sets</ttx>
    <td>Hits that are part of a collinear set. See <a href="#ex2d">example below</a>
        and <a href="Details.html#col" target="_blank" class="ext">Details</a>.
<tr><td><ttx>Hit =2 genes</ttx> <td>Hits that align to genes on both ends (g2).
<tr><td><ttx>Hit =1 genes</ttx> <td>Hits that align to a gene on one end (g1).
<tr><td><ttx>Hit =0 genes</ttx> <td>Hits that do not align to any gene (g0).
<tr><td><ttx>All hits</ttx>   <td>Show all hits.
</table>

<p>
<a id="hfother"></a>
<table class="ty">
<tr><td colspan=2><ttl>Other</ttl>
<tr><td><ttx>Identity</ttx><td>Move the slider to view only the hits with &gt;= the specified identity.
<i>Precedence:</i> This takes precedence over all other <ttl>Show</ttl> options.
<tr><td><ttx>Block#</ttx><td>Enter a block number and only hits from that block will be shown.
<i>Precedence:</i> This takes precedence over everything but the <ttl>Identity</ttl>.
<p>While the filter window is showing, the user can enter a block number following by return
to see the block, then try the next one to see the next block. If an invalid block number is
entered, no hits will be shown.
<tr><td><ttx>G2xN</ttx><td>This is a hit filter that occurs in the <ttb>Sequence Filter</ttb> of the
reference track for a 3-track display; see <a href="Details.html#2dg2"  target="_blank" class="ext">3-chr conserved</a>.
It takes precedence over all other hit filters except <ttx>Identity</ttx>.
</table>

<a id="hfpop"></a>
<p><u>Hit Popup</u>
<table class="tx" style="width: 70%">
<tr>
<td>
A subset of the highlight filters are available by right clicking in the hit white space,
as shown on the right.
<p>&nbsp;
<p>During a <ttp>Chromosome Explorer</ttp> session, any <ttl>Hit Filter</ttl> settings are not retained between
different <ttp>2D</ttp> displays.

<td>
<a href="img/hitFilterPopup.png">
	<img src="img/hitFilterPopup.png" alt="hit filter popup" style="border: 1px solid black; width: 140px"></a>
</table>

<!-------------------------------------------------------------------->
<a id="ex2d"></a>
<h4>Filter Example</h4>

<table>
<tr><td>The image on the right has the <ttb>Sequence Filter</ttb> option of <ttl>Gene Delimiter</ttl>
and <ttl>Collinear#</ttl>, and the <ttb>Hit Filter</ttb> option of <ttl>Highlight Collinear Hits</ttl>.

<p><ttx>Show Gene Delimiter</ttx>: Note that the 2nd gene on the left track would blend in with the 3rd if
it was not for the black line at the top.
<p><ttx>Highlight Collinear Hits</ttx>:
When hits belonging to <ttx>Blocks</ttx> or <ttx>Collinear sets</ttx> are highlighted, subsequent blocks/sets
will have an alternating color relative to one chromosome of the pair. For example, in the image on the right,
the green highlighted hit-wires are part of one collinear set (c629) and the pink highlighted hit-wire are
part of another (c630).
See <a href="Details.html#col" class="ext" target="_blank">Details</a> for
more information.

<td><a href="img/colSetGL.png">
	<img src="img/colSetGL.png" alt="Collinear Highlight and Gene Lines" style="border: 1px solid black; width: 300px"></a>
</table>
<!--------------------------------------------------------->
<!--------------------------------------------------------->
<a id="align"></a>
<h3>Base Align</h3>

<table class="tz"><tr>
<td><a href="#base">Graphical Align<br>(from pull-down)</a>
<td><a href="#btext">Text align<br>(from hit popup)</a>
<td><a href="#alt">Alternatives</a>
<td class="colrt"><a href="#top">Go to top</a>
</table>

<p>To build the synteny, SyMAP uses Promer by default for analysis of two different genomes.
Promer translates sequences, uses a fast suffix tree alignment algorithm, then converts the coordinates back to
nucleotide. To display the alignment in SyMAP, SyMAP performs a semi-global dynamic programming (DP) algorithm on the nucleotide sequences using the MUMmer coordinates,
which can produce a slightly different alignment.

<a id="base"></a>
<h4>Graphical Align (from pull-down)</h4>

Check <ttb>Show</ttb> on the <a href="#control">2D control bar</a>, then select a region of maximum 50kb,
<i>which must have hits</i>. To select a region,  drag the
mouse along the sequence and release when the desired range is highlighted.
<p><a href="img/baseview2.png">
	<img src="img/baseview2.png" alt="baseview" style="border: 1px solid black; width: 700px"></a>
<p>The aligned base view of the clustered subhits will appears in a new window.
This view consists of a ruler along the top showing the area
of the sequence covered, the hits, and the genes. In the lower text panel, it indicates that
has been <a href="#trim">trimmed</a>.

<p><u>Subhits</u><br>
Subhits are displayed as lines, where red horizontal lines are mismatches,
forest green is deletion and a downward arrow is insertion.
When two hits overlap, one will be shown above the other. Sometimes one appears to overlap another when they do not;
this happens due to gaps.
<p>Clicking on a hit line shows its base alignment view in the bottom of the dialog; if the input genome
sequence was soft-masked, the masking is retained in the alignment.

<p><u>Genes</u><br>
Annotated genes are displayed below the hits. Blue exons are on + strand and burgundy are on the - strand.

<a id="strand"></a>
<p>
<table>
<tr>
<td><u>Strand for Algo1</u>: Sometimes it is wrong in the <ttb>Show</ttb> <ttl>Align</ttl> view.
In this image, the A.thal gene is on the negative strand,
but its aligned hit is shown on the positive strand. Most of these
occurrences have been fixed in v5.4.8, but a few will have slipped through.
<td><a href="img/alignStrand.png">
<img src="img/alignStrand.png" alt="Align strand" style="border: 1px solid black; width: 330px"></a>
</table>
<!--------------------------------------------------->
<a id="btext"></a>
<h4>Text Align (from hit popup)</h4>

Hover over a hit-wire and right click for the information popup shown on the lower left.
The <ttl>Align Hit</ttl> option aligns the hit as shown on the lower right. Use the <ttl>Reverse</ttl> button to view the reverse complement
alignment.
<table>
<tr>
<td><a href="img/hitInfo.png">
<img src="img/hitInfo.png" alt="Hit Information" style="border: 1px solid black; width: 230px"></a>
<td>&nbsp;&nbsp;
<td><a href="img/basetext.png">
	<img src="img/basetext.png" alt="Text view of alignment" style="border: 1px solid black; width: 430px"></a>
</table>
<a id="trim"></a>
<u>Trimmed alignments:</u> SyMAP extracts the sequence between the coordinates provided by MUMmer and
aligns the sequence using a dynamic programming algorithm.
When the sequences do not align fully, the alignment is trimmed. This is illustrated above for subhit 5, where
the <tt>Trim 0,13</tt> indicates that 13 bases were trimmed off the end. The un-trimmed
alignment can be viewed by selected <ttl>No trim</ttl>.

<!--------------------------------------------------->
<a id="alt"></a>
<h4>Alternative base views</h4>

To view the base alignment, use the <ttl>Show Seq Option</ttl> (from the <ttb>2D</ttb>
<a href="#mouse_function_selector">Show</a> pulldown).
Select the 1st region to be aligned, then select the 2nd region. These two sequences can be copied
from the popups, and aligned with online tools such as
<a href="https://www.bioinformatics.nl/cgi-bin/emboss/matcher" class="ext" target=_blank>EMBOSS Matcher</a>
or <a href="https://www.ezbiocloud.net/tools/pairAlign" class="ext" target=_blank>EZ BioCloud</a> (max 5000bp).
<p>Another alternative is to find the hit in the
<ttp>Queries</ttp>, then view the
<a href="Query.html#msaAlign" class="ext" target="_blank">MSA</a> (either MUSCLE or MAFFT can be used).

<!-------------------------------------------------------------------->
<a id="dot3"></a>
<h2>Dot Plot #3 (Multi-genome)</h2>

All selected projects will be shown in the <ttp>Dot Plot</ttp>. The interface is the same as
discussed in <a href="#dotplot_display">Dot Plot (Two Genome)</a>. The <ttp>Dot Plot</ttp> in the image
below shows three genomes; A.thal is the reference, which can be changed with the pull-down at the top.
<p><a href="img/dotplot-3g.png"><img src="img/dotplot-3g.png" alt="" style="border: 1px solid black; width: 600px"></a>

<!-------------------------------------------------------------------->
<a id="gen"></a>
<h2>General</h2>

<table class="tz">
<tr><td><a href="#self">Self-synteny</a>
	<td><a href="#data_download">Data Download</a>
	<td><a href="#image">Color Wheel, Print and Help (?) Icons</a> <!-- # ref 'wheel' in code -->
<td class="colrt"><a href="#top">Go to top</a>
</table>

<a id="self"></a>
<h3>Self-synteny</h3>

Self-synteny can be viewed in all views
(e.g. <a href="Demo.html#self" class="ext" target="_blank">Demo Dot Plot</a>), with the following exception:
<p>
<ul>
<li>The <ttp>Circle</ttp> view works, but it ribbon colors work a bit strangely since there is only one genome.
<p>
<li>The <ttp>2D</ttp> does not directly work for a single chromosome compared
to itself from the <ttp>Chromosome Explorer</ttp>. However, by going through the <ttp>Dotplot</ttp> or <ttp>Blocks view</ttp>, you can click on a self-chromosome block and view it in
the <ttp>2D</ttp> view.
<p>
<li><ttp>Queries</ttp> does not work for self-synteny.
<p>
<li><ttp>Summary</ttp> stats are not accurate since the dots above and below the diagonal are the same.
</ul>

<a id="data_download"></a>
<h3>Data Download</h3>

There are three ways to download data for SyMAP synteny blocks, individual hits, and annotations:
<p><ttp>Manager</ttp>:
<ul>
<li><a href="#breport">Block Report</a>: Select the species of interest, select <ttb>Report...</ttb> on the <ttp>Manager</ttp>,
then select <ttl>Block Report</ttl>
</ul>

<p><ttp>Queries</ttp>:
<ul>
<li><a href="Query.html#export" class="ext" target="_blank">Query Export</a>:
The results of a query are shown in a table,
which can be exported as a CSV or HTML file.  The hit sequences can be exported in FASTA format.
<p>
<li><a href="Query.html#export" class="ext" target="_blank">Gene Report</a>: A report that combines rows of a query can be created and exported in HTML or TSV format.
</ul>

<a id="image"></a>
<h3>Color Wheel, Print and Help (?) Icons</h3>

Most displays have one or more of the following icons (Color, Print, Help, respectively):
<br><img src="img/icons.png" alt="icons" style="border: 1px solid black; width: 150px">.

<p><u>Print</u> The print icon is for printing the image, which only saves the graphics area
(not control buttons). If this does not provide the view you want,
use the system "Screen Capture" (all the images in this
document were created with screen capture, along with the images in the SyMAP publications).

<p><u>Help</u> The <big>?</big> icon brings up this web page, typically to the correct section (obviously, there needs
to be an internet connection).

<a id="wheel"></a> <!-- # ref in code -->
<p><u>Color Wheel</u>
This icon opens a menu for customizing colors; the colors changes are saved in a file
called <tt>.symap_saved_props</tt> which resides in the user's home directory. Hence, the
changes are preserved between sessions and for different SyMAP databases. In contrast, <ttl>Filters</ttl>
changes only exist for the duration of the display.
<p>Click the <ttl>Color Wheel</ttl> in any of the displays, then:
<ul>
<li>Click a tab (e.g. <ttx>Hit</ttx>) to view the associated colors.
<li>Click a color box and a color chooser (far right image) will popup
that lets you change the color.
<li><ttl>Ok</ttl>: Instantiate all changes and close the menu.
<li><ttl>Cancel</ttl>: Cancel any changes and close the menu.
<li><ttl>Default</ttl>: For the selected tab, changes all colors back to the default colors.
</ul>
<table class="tx">
<tr><td>
<tr>
<td style="vertical-align: text-top;"><a href="img/2d-color.png">
  <img src="img/2d-color.png" alt="Track Colors" style="border: 1px solid black; width: 270px"></a>
<td style="vertical-align: text-top;"><a href="img/2d-color2.png">
  <img src="img/2d-color2.png" alt="Hit Colors" style="border: 1px solid black; width: 260px"></a>
<td><a href="img/2d-color3.png">
   <img src="img/2d-color3.png" alt="Change Colors" style="border: 1px solid black; width: 200px"></a>
</table>

<p><u>Dot Plot tab</u>:
<table>
	<tr><td>
	<table  class="ty txy">
		<tr><td><ttx>!Block g0</ttx><td>Non-block hit to 0 genes
		<tr><td><ttx>!Block g1</ttx><td>Non-block hit to 1 gene
		<tr><td><ttx>!Block g2</ttx><td>Non-block hit to 2 genes
		<tr><td><ttx>Block g0</ttx><td>Block hit to 0 genes
		<tr><td><ttx>Block g1</ttx><td>Block hit to 1 gene
		<tr><td><ttx>Block g2</ttx><td>Block hit to 2 genes
		<tr><td><ttx>Block Rect</ttx><td>Rectangle around block
	</table>
	<td>&nbsp;
	<td><a href="img/dotplot-color.png">
	<img src="img/dotplot-color.png" alt="dotplot color" style="border: 1px solid black; width: 320px"></a>
</table>

<!---- END BODY -->
<table style="width: 95%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<p class="trailer">  Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a></p>
</div>
</div>
</body>
</html>
