<!DOCTYPE html>
<html>
<head>
<title>SyMAP User Guide</title>
<style>
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	ttb  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: Black;}
	a:link {text-decoration: none;color:#6666CC;}
	a:link {text-decoration: none;color:#6666CC;}
	a.current {font-family:verdana; font-size:12px; color:white; text-decoration: none;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
    .ty td {border: 1px  solid black; padding: 3px; }
    top {vertical-align:text-top;}
    tt {font-size:13px}
	pre {font-size:13px}
	body {font-family: Verdana, Arial, Helvetica,  sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					SyMAP User Guide
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/SyMAP" class="white">SyMAP Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/SyMAP/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="./SystemGuide.html" class="white">System Guide</a> <mark class="white">|</mark>
						<a href="./UserGuide.html" class="current">User Guide</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

This guide is for the Seq-to-Seq (sequence) synteny user interface.
<ol>
<li>Most of the contents also apply to the FPC-to-Seq view. For specific information see <a href=UserGuideFPC.html class="ext" target="_blank">FPC specific</a>.
<li>For detailed information on installation, see
the <a href="SystemGuide.html" class="ext"  target="_blank">System Guide</a>.
<li>For a quick visual introduction, see the <a href="Tour.html" class="ext" target="_blank">Tour</a>.
<li>Please send bug reports and suggestions to symap@agcol.arizona.edu.
</ol>

<h2>Contents</h2>
The Dot Plot and Circle displays are available from various places.
<ul>
<li><a href="#mgr">Main Display</a>
<li><a href="#pair"><i>Displays for Selected Pair (Two-Genome Views)</i></a>
	<ul>
	<li><a href="#circle">Circle Display #1</a> <!-- # ref in code -->
	<li><a href="#dotplot_display">Dot Plot #1</a> <!-- # ref in code -->
	<li><a href="#block">Block Display</a>
	<li><a href="#sum">Summary</a>
	</ul>

<li><a href="#all"><i>Displays for All Projects (Multi-Genome Views)</i></a>
	<ul>
	<li><a href="#explorer">Chromosome Explorer  (Multi-chromosome)</a>
		<ul>
		<li><a href="#alignment_display_circ">Circle Display  #2</a>
		<li><a href="#alignment_display_dot">Dot Plot #2</a>
		<li><a href="#alignment_display_2d">2D Display</a> <!-- # ref in code -->
			<ul>
			<li><a href="#base">Align (Base View)</a>
			</ul>
		</ul>
	<li><a href="#dot3">Dot Plot #3</a>
	<li><a href="#annotsearch">SyMAP Queries </a>
	</ul>
<li><a href="#self">Self-synteny Displays</a>
<li><a href="#data_download">Data Download</a>
<li><a href="#image">Print and Help</a>
</ul>
<h4>Citations</h4>
The SyMAP User Agreement requires you cite one of the following articles
if you use SyMAP in a paper, poster, or presentation.
<pre>
	C. Soderlund, M. Bomhoff, and W. Nelson (2010)
	SyMAP: A turnkey synteny system with application to plant genomes.
	Nucleic Acids Res 39(10):e68. <a href="https://academic.oup.com/nar/article/39/10/e68/1310457" target="_blank" class="ext">Link</a>
</pre>
<pre>
	C. Soderlund,  W. Nelson, A. Shoemaker and A. Paterson (2006)
	SyMAP: A System for Discovering and Viewing Syntenic Regions of FPC maps
	Genome Research 16:1159-1168. <a href="https://genome.cshlp.org/content/16/9/1159.abstract" target="_blank" class="ext">Link</a>
</pre>
<pre>
	Also cite the MUMmer publication: <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005944" target="_blank" class="ext">Marcais et al. 2018</a> or <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2004-5-2-r12"  target="_blank" class="ext">Kurtz et al. 2004.</a>
</pre>
<a id="mgr"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h1>Main display</h1>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
&rarr;For simplicity, sequence groups will be referred to as "Chromosome"", though they may be scaffolds or linkage groups.
<p>
The display in the image on the right is shown when the following is executed from the terminal:
<pre>   ./viewSymap</pre>
Is it also available from <tt>./symap</tt>, which will also have the alignment commands.
<p>
<table>
<tr><td style="vertical-align:text-top;">

All projects in the database will be listed on the left panel. The date beside the project name
is when it was loaded. If a [n] follows the date, it is part of n computed syntenies.

<p>Selecting projects on the left panel shows them on the right panel.
A check mark in the <ttb>Available Syntenies</ttb> table indicates pairs that have computed synteny that can be viewed.
The two view options beside <ttl>Display for All Pairs</ttl> will be activated.

<p>By selecting a cell with synteny, the <ttl>Display for Selected Pair</ttl> buttons are activated.

<td><a href="img/viewRosid.png"><img src="img/viewRosid.png"  alt="viewRosid" style="border: 1px solid black; width: 550px"></a>
</table>

<a id="pair"></a>
<h1>Displays for Selected Pair</h1>
The buttons beside this label are "whole genome" displays for the selected pair (highlighted yellow).
<!------------------------------------------------->
<a id="circle"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Circle Display #1 (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
The two-genome and multi-chromosome Circle displays work the same;
see <a href="file:///Users/cari/Sites/symap/docs/UserGuide.html#explorer">Explorer</a>
for the chromosome display.
<p>
<table>
<tr><td style="vertical-align:text-top;">
The circle display shows the chromosomes arranged in a circle,
with synteny blocks shown as colored ribbons between the chromosomes.
The color of a ribbon comes from one of the two chromosomes it connects; for example,
in the image on the right, the Cabbage Chr1 is red and the corresponding ribbon is red.
<p>Clicking on a species names causes its colors to be used preferentially (<i>sometimes it
is necessary to click at the very beginning or end of the name for this to work</i>).
<p>Clicking on a chromosome (e.g. the red part of the circle for Cabbage Chr1)
causes ribbons from that chromosome to be shown on top.

<h4>Control Panel</h4>
Referring to top of the display, the first two buttons (<b>+, -</b>) change the size, while the third rotates the image.
<p>
<ttl>Scale to genome size:</ttl> This checkbox causes the chromosomes to be drawn proportionately to their
actual length in basepairs. Otherwise, each species is allocated the same amount of space,
e.g. for two species each one gets exactly half the circle.
<td><a href="img/circle-2g.png">
	<img src="img/circle-2g.png" alt="circle-two-genome" style="border: 1px solid black; width: 450px"></a>
</table>
The drop-down menu has: <ttl>Show all blocks</ttl>,
<ttl>Show all blocks</ttl>,
<ttl>Show inverted blocks</ttl>,
<ttl>Show non-inverted block</ttl>,
<ttl>Two-color scheme</ttl>, which shows inverted=green, non-inverted=red.

<p><ttl>Self-align</ttl> (If self-synteny has been computed) This checkbox controls whether blocks are shown between chromosomes
of the same species.
<p>
<ttl>Rotated text</ttl> Depending on the system, it may be easier (or harder) to read with
the text rotated.

<!------------------------------------------------------------>
<a id="dotplot_display"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Dot Plot #1 (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr>
<td style="vertical-align:text-top;">
Dots represent anchors (also referred to as <i>hits</i>). A blue box indicates a
<i>Synteny Block</i>. A chromosome pair is represented by a <i>cell</i>.
<p>
Clicking a cell will replace the whole genome view with the dotplot for the chromosome pair.
<p>Within the cell dotplot:
<ul>
<li>Click a block to view the 2D display of the block.
<li>Or, use the mouse to define a region, which will put a box around it. Then click the box to view the 2D display.
<li>This is illustrated in the <a href="#dot2">Chromosome Dotplot</a>.
</ul>

<td><a href="img/dotplot-2g.png"><img src="img/dotplot-2g.png"  alt="dotplot-2g" style="border: 1px solid black; width: 550px"></a>
</table>

<a id="dotplot_control"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Control Panel</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/dotplot-control.png"><img src="img/dotplot-control.png"  alt="dotplot controls" style="border: 1px solid black;width: 600px"></a>
<p>
<ttl>Home</ttl> Resets the zoom selection to the whole genome dot plot.
<p>
<ttl>Zoom</ttl> The plus (<ttl><big>+</big></ttl>) and minus (<ttl><big>-</big></ttl>) buttons shrink or grow the display.
<p>
<ttl>Scale</ttl> The check box resizes the species displayed on the y-axis according to the current scale of
the species displayed on the x-axis.
<p>
<ttl>Reference</ttl> The drop-down menu allows the reference species on the x-axis to be changed.
<p>
<ttl>Size</ttl> The slider increases/decreases the size of the dots in the display.


<a id="dotplot_filter"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Dotplot Filters</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr><td>
The Filter image on the right indicates the default settings.
<p><ttl>%Identity</ttl> Show all hits that have &gt;N% identity, where the %Identity is from the MUMmer anchor file.
<p><ttl>Show Only Block Hits</ttl>
A <i>block hit</i> is a hit that the SyMAP synteny-finding algorithm
used to create a synteny block.
<p><ttl>Show All Hits</ttl>
Displays all hits. Note that the SyMAP hit loading algorithm removes
some repetitive hits from the raw anchor set, so some hits from the MUMmer files will not be shown.
<p><ttl>Highlight Block Hits</ttl>
Highlight hits that make up a synteny block in blue.
<p><ttl>Show Blocks</ttl>
This option draws blue rectangles around the synteny blocks.
<p><ttl>Show Empty Regions</ttl>
When deselected, hides regions that do not contain any synteny blocks.
<td><a href="img/dotplot-filter.png"><img src="img/dotplot-filter.png" alt="dotplot filter" style="border: 1px solid black; width: 220px"></a>
</table>
<!------------------------------------------------->
<a id="block"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Blocks Display (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The image below shows the block display for B.rapa and Cabbage.
The 10 B.rapa chromosomes are color coded as shown at the top, and aligned to
the 9 Cabbage chromosomes. The Cabbage to B.rapa block view can be viewed by
selecting <u>reverse</u>.
<table>
<tr>
<td><a href="img/block.png"><img src="img/block.png" alt="block" style="border: 1px solid black; width: 500px"></a>
<td>Clicking a chromosome pops up a window of the chromosome as shown below. Clicking
a block pops up a 2D view.
<p><a href="img/block-chr.png"><img src="img/block-chr.png"  alt="block chr" style="border: 1px solid black; width: 290px"></a>
</table>

<!------------------------------------------------->
<a id="sum"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Summary (Two Genome)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The <ttb>Help</ttb> button on this view explains the different columns.
<p><a href="img/sum-2g.png">
	<img src="img/sum-2g.png" alt="sum-two-genome" style="border: 1px solid black; width: 650px"></a>
<!--------------------------------------------------------------------------------------->
<!------------------------------------------------------------------------------------->
<a id="all"></a>
<h1>Displays for All Projects</h1>
From the <a href="#mgr"> main SyMAP display</a>, select two or more projects, which will
activate the "Chromosome Explorer" and "Dotplot" buttons.

<a id="explorer"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Chromosome Explorer</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The Explorer is a two-panel display in which the left panel is used to select specific chromosomes, while the right panel
shows the synteny for the selected chromosomes using one of the following views:
<a href="#alignment_display_circ">Circle</a>,
<a href="#alignment_display_2d">2D</a> and
<a href="#alignment_display_dot">Dotplot</a>.
Below is a snapshot of the Explorer showing the circle, which is the initial view.
<p>
<a href="img/explorer.png">
	<img src="img/explorer.png" alt="explorer" style="border: 1px solid black; width: 650px"></a>

<p><b>Left panel: </b>The left panel controls which species and chromosomes are shown:
<ul>
<li>Click a chromosome rectangle to add it to the display in the right panel. Click it again to remove
it.
<li>Click the chromosome number above a chromosome to make it the <i>reference</i>. (The
choice of reference sequence does not matter for the circle view,
but it is important for the 2D and Dot Plot views; see below).
<li> Adding and removing chromosomes affects the Circle view instantly.
To add/remove from the 2D or Dot Plot
views, open the Circle view, make the change, and then re-open the 2D or Dot Plot.
</ul>
<p>
<ttl>Instructions</ttl>: As the mouse moves over various
components, instructions or information for that component are shown in the Instruction box.
<p>
<ttl>Views:</ttl> The buttons in the lower-left corner of the window change the synteny view in the right panel.
<p>
<ttl>Download Blocks</ttl>: Exports a table of all of the synteny block coordinates for the selected species; see also <a href="#data_download">Data Download</a>.
<p>
Click the minus (<ttl><big>-</big></ttl>) button next to a project name or the Instruction box,
and that section will be hidden;
the minus sign will change to a plus.
Click the plus (<ttl><big>+</big></ttl>) button to show the project or Instruction box..


<p><b>Right panel:</b>
The right panel shows the synteny display for the species and chromosomes selected on the left.
The three view for the right panel are
<a href="#alignment_display_circ">Circ</a>,
<a href="#alignment_display_2d">2D</a> and
<a href="#alignment_display_dot">Dotplot</a>.
The first two views are
essentially overviews, while the 2D view allows zooming in to details, all the way
to the basepair level if desired.
Each right-panel display format has its own controls; see the individual sections for details.

<!----------------------------------------------------------------->
<a id="alignment_display_circ"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Circle Display #2 (Multi-chromosome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

This view is shown in the above <a href="#explorer">Chromosome Explorer</a> image.
All features are described in the first <a href="#circle">circle section</a>,
except it is chromosome-based instead of genome-based.

<!----------------------------------------------------------------->
<a id="alignment_display_dot"></a>
<h3>Dot Plot #2 (Multi-chromosome)</h3>

All features are described in the first <a href="#dotplot_display">dot plot section</a>,
except that it is chromosome-based instead of genome-based.
Selecting a chromosome pair (cell) from the image on the left will replace it with the dotplot for the cell
on the right. Clicking on a synteny block (or selected region) brings up the 2D view shown furthest right below.
<table>
<tr><td><a href="img/dotplot-3chr.png">
	<img src="img/dotplot-3chr.png" alt="dotplot 3chr" style="border: 1px solid black; width: 350px"></a>
<td><a href="img/dotplot-chr-2d.png">
	<img src="img/dotplot-chr-2d.png" alt="dotplot-chr-2d" style="border: 1px solid black; width: 360px"></a>
</table>

<!-------------------------------------------------------------------->
<a id="alignment_display_2d"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>2D Display (Multi-chromosome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table>
<tr><td>Contents:&nbsp;&nbsp;&nbsp;
<td><a href="#view">View a region</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#control">Control Panel</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#sequence_filter">Sequence Filter</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td><a href="#hitfilter">Hit Filter</a>&nbsp;&nbsp;&nbsp;
</table>

<p>
The 2D display is activated from the <a href="#explorer">Explorer</a> by
clicking the <ttb>2D</ttb> button (or by selecting a region or block from a dot plot).
<p>
<table>
<tr><td style="vertical-align:text-top;">
On the right is an image depicting an alignment from A.thal Chr4 to Cabbage Chr1 to B.rapa Chr2.
Note that Cabbage Chr1 is the <i>reference</i>
chromosome, hence is placed in between the others.
<p>
Each chromosome is drawn as a light-blue rectangle, called <i>tracks</i>.
The sequence length is shown at the bottom of a track; the coordinates are displayed on the side.
<p>The gene annotations (if exists) are drawn down the middle of the rectangle in dark-blue.
<p>The synteny <i>hits</i> are the brown lines between tracks.
The colors can be changed using the Color Icon in the upper right.
<p>
Right click in the sequence track will cause a popup of the most commonly used filters and features (e.g.
flipping the region). They are also on the filter menu above each track
(see <a href="#sequence_filter">Sequence Filter</a>).

<td><a href="img/2d-3chr.png">
	<img src="img/2d-3chr.png" alt="2d-3chr" style="border: 1px solid black; width: 550px"></a>
</table>


<!------------------------------------------------------->
<a id="view"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>View a region</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The example below was derived from a two chromosome display of B.rapa Chr2 and Cabbage Chr1. The region
was displayed by right-clicking
near the top of the chromosome, dragging the mouse and releasing it in order to view the selected region.
The <ttb>Sequence Filter</ttb> was set to display <ttl>Annotation Descriptions</ttl>.

<p>
<table>
<tr><td  style="vertical-align:text-top;">
Genes: The black lines with heavier blue bars interspersed down the center indicate exon/intron
predictions. A solid black line would indicate the absence of introns in the annotation.

<p>Hits: The alternating brown/gray lines indicate anchor hit clusters; e.g. the lower gene shown on the right
has 16 separate MUMmer hits that are clustered together to represent one hit (#269294).
The solid brown lines indicate
a single hit (or multiple hits very close together), while the gray lines are spaces with no alignment.

<p>Ruler: The ruler on the right side of the sequence rectangle shows the relative BP position along the sequence.

<td><a href="img/2d-closeup.png">
	<img src="img/2d-closeup.png" alt="2d-closeup" style="border: 1px solid black; width: 500px"></a>
</table>

<table style="width: 70%;">
<tr><td>
Hovering the mouse over a hit or gene will show its information in the <ttl>Instruction</ttl> box,
as shown here for Hit #269294. Hovering the mouse over a gene will shown the gene information in
the instruction box.
<td><a href="img/2d-hit.png"><img src="img/2d-hit.png" alt="2d-hit" style="border: 1px solid black; width: 230px"></a>
<tr><td>Right clicking on a yellow annotation box results in a popup, as shown on the far right. Besides the
gene information, it also shows the list of exons. It is useful for copying all or part of the description, and viewing the
information when the box is partially hidden.
<td><a href="img/2d-desc.png"><img src="img/2d-desc.png" alt="2d-desc" style="border: 1px solid black; width: 250px"></a>
</table>
<p>
Annotation data of the following types may be loaded into SyMAP and displayed.
<table class="ty" style="width:650;">
	<tr><td><i>Annotation Type</i><td><i>Display</i>
	<tr><td>Gaps<td>Red band across the chromosome
	<tr><td>Centromere<td>Blue "X" across the chromosome
	<tr><td>Predicted genes and exons
		<td>Annotation strip in center of chromosome, as described above.
</table>
The display colors can be changed using the Color Icon in the upper right (color panel not shown).
<!----->

<a id="control"></a>
<a id="2dctrl"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Control Panel and Navigation</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a href="img/2d-control.png"><img src="img/2d-control.png" alt="2d controlpanel"></a>
<p>
<ttl>History (Home, &lt;, &gt;)</ttl> SyMAP retains a record of the prior views (like a web browser).
The history navigation back (<ttl><big>&lt;</big></ttl>) and
forward (<ttl><big>&gt;</big></ttl>) buttons allow you to
move back and forth through the previous views.
The <ttl>Home Icon</ttl> button returns to the initial view.
<p>
<ttl>Zoom Buttons (-, +)</ttl>
The zoom buttons allow for quickly expanding and narrowing the view region.
 The minus (<ttl><big>-</big></ttl>) button shrinks the view region by 50%, keeping
the same center, while the plus (<ttl><big>+</big></ttl>) button doubles the view region,
again keeping the same center.
<p>
<ttl>Scale Icon Button</ttl>
The Scale button
(right of <ttl><big>+</big></ttl>) resizes the tracks so that they are in the same scale
(base pairs per pixel) as the reference sequence track in the view.
<p>
<a id="mouse_function_selector"></a>
<ttl>Selected:</ttl>
This drop-down selector assigns a function to the mouse's left button
click and drag actions:

<table class="ty" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr><td style="white-space: nowrap; vertical-align: text-top;">Zoom All Tracks<td>Zoom the track to the selected region and all other tracks with hits in the region.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Zoom Selected Track<td>Zoom to the selected region on the selected track; all other tracks do not change.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Align (Max 30000)<td>
Open the base alignment view for the selected region (see <a href="#base">Base View</a>).
No more than 30000 bases can be selected. There MUST be hits in the region to align.
<tr><td style="white-space: nowrap; vertical-align: text-top;">Show Sequence<td>
The selected sequence will be shown in a popup, which can be copied. This has limited utility, but is the
only way to view the entire underlying sequence (the align view shows the bases of hit regions only).
</table>

<p><ttl>Print, Colors, <big>?</big></ttl>. See <a href="#image">Help and Print</a> for the 1st and 3rd button.
The <ttl>Color Icon Button</ttl> opens a menu for customizing colors.
<p>
Additional navigation:
<table  class="ty" style="font-family: Verdana, Arial, Helvetica, sans-serif; font-size:9pt">
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">Resize Track&nbsp;</td>
	<td>Drag bottom of track</td>
	<td>Position mouse at bottom of track (resize cursor appears), hold down left
	mouse button, and move mouse.</td></tr>
<tr>
	<td style="white-space: nowrap; vertical-align: text-top;">Scroll Track&nbsp;</td>
	<td>Mouse wheel</td>
	<td>Position mouse over track and use mouse wheel.</td></tr>
<tr>
	<td>Filter&nbsp;</td>
	<td>Right mouse button</td>
	<td>Position mouse over track or white space between tracks, and click right mouse button.</td></tr>
</table>

<p>
<a id="sequence_filter"></a>
<a id="sequence_track"></a>
<h4>Sequence Filter</h4>

Select <ttb>Sequence Filter</ttb> button above the sequence track to change the display for the
respective chromosome.
<p>
<ttl>Start</ttl> and <ttl>End</ttl>
The positions of the sequence display can be set via the corresponding text boxes.
The units of the values entered can be selected from the accompanying drop down menus
(BP, KB, MB, GB).

<table>
<tr><td>
<ttl>Full Sequence</ttl> Sets the start and end positions of
the sequence display to encompass the whole chromosome.
<p>
<ttl>Flip</ttl> Reverses the orientation of the sequence track.
<p>
<ttl>Show Genes</ttl> Enables/disables the display of gene/exon annotations along the sequence.
<p>
<ttl>Show Annotation Descriptions</ttl> Enables/disables the display of the annotation descriptions along the
right side of sequence. NOTE: this only works if you are zoomed in close enough that they can clearly
be displayed.
<p>
<ttl>Show Gaps</ttl> Enables/disables the display of sequence gaps (drawn as solid red
rectangles) along the sequence.
<p>
<ttl>Show Centromere</ttl> Enables/disables the display of the centromere (drawn as a cyan "X") on the sequence.
<p>
<ttl>Show Framework Markers</ttl> Ignore, not relevant.
<td>
<a href="img/sequencefilter.png">
	<img src="img/sequencefilter.png" alt="sequencefilter" style="border: 1px solid black; width: 300px"></a>
</table>

<p>
<ttl>Show Ruler</ttl> Enables/disables the display of the sequence ruler along the right side
of the sequence.
<p>
<ttl>Show Hit Length</ttl> Enables/disables the display of the hit length line next to each hit along the
sequence.  The hit length line denotes the start and end points of the hits relative
to the sequence (may not be visible for tiny hits).
<p>
<ttl>Show Hit %Id Line</ttl> Enables/disables the display of the %Id line next to each hit along the
sequence.  The length of the line represents the magnitude of the % Identity value for
the hit.
<p>
<ttl>Show Hit %Id Value</ttl> Enables/disables the display of the %Id value next to each hit along the
sequence.  The score value corresponds to the % Identity value for the hit from the MUMmer anchor file.

<table style="width:50%;">
<tr><td>A subset of the filters are available by right
clicking in the light blue area of the chromosome rectangle.
<td><a href="img/sequencefilter2.png">
	<img src="img/sequencefilter2.png" alt="sequencefilter" style="border: 1px solid black; width: 150px"></a>
</table>

<!------------->
<a id="hitfilter"></a>
<a id="hit"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h4>Hit Filter</h4>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The <ttb>Hit Filter</ttb> menu allows the user to select which types of hits are displayed.
<table>
<tr><td>
<ttl>Show Only Synteny Hits</ttl> Shows only hits that are part of a synteny block.
<p>
<ttl>Show Only Hits to Genes</ttl> Show only hits which intersect a gene annotation.
<p>
<ttl>Show Only Non-Gene Hits</ttl> Show only hits which do not intersect a gene annotation.
<p>
<ttl>Show All Hits</ttl> hows all known hits.

<td><a href="img/hitfilterS.png">
	<img src="img/hitfilterS.png" alt="hitfilterS" style="border: 1px solid black; width: 250px"></a>
</table>

<a id="base"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Align (Base View)</h3></td>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Select <ttl>Align (Max 30000bp)</ttl> from the <ttb>Selected:</ttb> drop-down, then select a region of maximum 30kb,
<i>which must have hits</i>.
To select a region,  drag the
mouse along the sequence and release when the desired range is highlighted.
<p>
<a href="img/baseview.png">
	<img src="img/baseview.png" alt="baseview" style="border: 1px solid black; width: 750px"></a>
<p>
The aligned base view of the hits will
appears in a new window.  This view consists of a ruler along the top showing the area
of the sequence covered, the hits, and the genes.
<p>
<b>Hits</b><br>
Hits are displayed as lines with an arrow on one end showing the direction of the hit;
a hit may represent multiple MUMmer hits, and is further broken into sub-hits where there are indels.
A vertical red line along a represents a mismatch.
Clicking on a sub-hit line brings up the its base alignment view in the bottom of the dialog.
<p>
<b>Genes</b><br>
Annotated genes are displayed below the hits.  Exons are represented by a blue box, and
the leftmost or rightmost exon box will have an arrow tip indicating whether the gene
is on the + strand (right-pointing) or - strand (left-pointing) relative to the
sequence. The system attempts to expand the view to show the full gene that has
hits. Genes that overlap this gene will be partially shown.

<!-------------------------------------------------------------------->
<a id="dot3"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>Dot Plot #3 (Multi-genome)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

All selected projects will be shown in the Dot Plot. The interface is the same as
discussed in <a href="#dotplot_display">Dot Plot (Two Genome)</a>. The dot plot in the image
shows three genomes.
<p><a href="img/dotplot-3g.png"><img src="img/dotplot-3g.png" alt="" style="border: 1px solid black; width: 600px"></a>
<!-------------------------------------------------------------------->

<a id="annotsearch"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>SyMAP Queries</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The query interface only works for multi-genome Seq-to-Seq synteny
(i.e does not work for self-synteny).
<p>
To open the query interface, first select two or more sequence projects in the Project Manager.
Then select the <ttb>SyMAP Queries</ttb> button to open the interface:
<p>
<a href="img/query_oview.png"><img src="img/query_oview.png" alt="query_oview" style="border: 1px solid black; width: 500px"></a>
<p>
The Overview window (above) lists the projects which were selected for querying. To set up a query, open the
<ttl>Query Setup</ttl> window by clicking on its title in the left panel. This opens the window below:
<p>
<a href="img/query_query.png"><img src="img/query_query.png" alt="query_query" style="border: 1px solid black; width: 550px"></a>
<p>
Set up the query and then select <ttb>Run Search</ttb> to execute it.
The Query option are fully explained in the <ttb>Help</ttb> popup.
The following describes two special features:
<p><ttl>Collinear genes N&gt;= N</ttl>
<br>
Return only hits which are part of a pair of N aligning genes having no intervening
non-aligning genes. There is a column called "RunSize" that provides the number
of adjacent aligning genes. Note that this is not the same as synteny blocks, which
allows intervening non-aligning genes.
<p><ttl>Filter putative gene families (PgeneFs)</ttl>
<br>
Using the hits that pass the hit filters, SyMAP constructs putative gene families (PgeneFs)
spanning the selected species. This is done by grouping hits whose hit regions overlap on
at least one genome. <i>Note, if you have more than 6 species selected, this stage can take
an hour or more.</i>
<p>
Each PgeneF is given a number, which is shown in the Query Results table (column name "PgeneF").
The size of the PgeneF is also shown (column "PgFSize").

<h3>SyMAP Query Result Table</h3>
<p>
When the query is complete, the Query Results page opens showing the table of results:
<p>
<a href="img/query_result.png"><img src="img/query_result.png" alt="query results" style="border: 1px solid black; width: 550px"></a>

<p>
The table contains all the hits (anchors) resulting from the query. Each hit connects
two species and you can see the respective chromosomes and start/end locations of
the hits, as well as gene annotations overlapped by the hits.
<p>
Note that the table contains columns for <i>all</i> of the selected
species, but each hit only connects two species, and the other species columns are empty.
If the query specified orphan genes, then each row represents one gene and shows data only
for one species.
<p>
The columns can be sorted by clicking on the column name,
and rearrange them by dragging the header boxes.

<h4>Button functions</h4>

<div style="width:700px;margin-top:10px; margin-left:15px;">
<ttb>Align sequences</ttb>:
Sequences of the selected hit are written out and a multiple alignment is created
using MUSCLE (Edgar 2004 NAR:32).
<p>
<ttb>Show Synteny</ttb> :
Brings up a SyMAP 2-track view for each selected entry, so the hit
can be seen in their full chromosomal context. The hit is initially padded to each side
by the margin amount indicated (default 50kb), but you can easily zoom further out in the 2-track views.
<p>
<ttb>Export CSV</ttb>:
Saves the selected hits (or entire table) using the selected set of columns to a CSV format suitable for import into Excel.
<p>
<ttb>Save as Fasta</ttb>: Sequences from the selected hits (or entire table) are written to a Fasta file. Both sides of each hit
are written, using the start/end coordinates shown in the table.
</div>

<h3>Results Table Columns</h3>
You can add/remove columns using the <ttb>Select Columns</ttb> button at the bottom. This
opens up a column-selection section. The columns are explain the the <ttb>Help</ttb> button on the table.

<a id="self"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Self-synteny Display</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Self-synteny can be viewed in all views
(see <a href="Demo.html#self" class="ext" target="_blank">Demo Dot Plot</a>), with the following exception:
The 2D does not work for a single chromosome compared
to itself. However, by going through the dotplot or blocks view, you can click on a self-chromosome block and view it in
the 2D view.

<p>The <ttb>SyMAP queries</ttb> does not work for self-synteny.

<a id="data_download"></a>
<h2>Data Download</h2>

There are two ways to download data for SyMAP synteny blocks, individual hit anchors, and annotations:
<ul>
<li> Through the <a href="#explorer">Explorer</a>. Select the species of interest,
 open the Explorer, and the download button is at the lower left.
 This exports a table of all the computed synteny block coordinates for all the
 selected species, including their self-alignments, if those were computed.
<li> Through the <a href="#annotsearch">SyMAP Queries</a>. For example,
 choosing two species and executing a query with the default parameters
 will show a table of all the hits found between the two species, plus
their annotations and synteny block membership, if any.
</ul>

<a id="image"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h2>Print and Help</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Most displays shown have a <img src="img/print.png" alt="print" style="border: 1px solid black; width: 50px"> set of icons.
<p>The left icon is for printing the image. However, this does not work with all versions
of Java. If it does not work, use the system "Screen Capture" (all the images in this
document were created with screen capture, along with the images in the SyMAP publications).

<p>The right
icon brings up this web page, typically to the correct section (obviously, there needs
to be an internet connection).

<!-------------------------- this link is in SyMap--------------------------------------->
<a id="alignment_display_3d"></a>
<table style="width: 95%"><tr><td style="text-align: left">
<h3>3D Display (if available)</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
The 3D display is generally obsolete, but if there is a "3D" button on your display,  see
<a href="./3D.html">3D</a>.

<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:symap@agcol.arizona.edu">symap@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
